<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>DNA Hydroxymethylation in Hepatocellular Carcinoma</title>
  <meta name="description" content="Data from Cai et al. (2019) paper are explored" />
  <meta name="generator" content="bookdown #bookdown:version# and GitBook 2.6.7" />

  <meta property="og:title" content="DNA Hydroxymethylation in Hepatocellular Carcinoma" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Data from Cai et al. (2019) paper are explored" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="DNA Hydroxymethylation in Hepatocellular Carcinoma" />
  
  <meta name="twitter:description" content="Data from Cai et al. (2019) paper are explored" />
  

<meta name="author" content="Francois Collin" />


<meta name="date" content="2020-01-01" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<!--bookdown:link_prev-->
<!--bookdown:link_next-->
<script src="libs/header-attrs-2.2/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<script type="text/javascript">
// source:https://stackoverflow.com/questions/45360998/code-folding-in-bookdown
// toggle visibility of R source blocks in R Markdown output
function toggle_R() {
  var x = document.getElementsByClassName('r');
  if (x.length == 0) return;
  function toggle_vis(o) {
    var d = o.style.display;
    o.style.display = (d == 'block' || d == '') ? 'none':'block';
  }

  for (i = 0; i < x.length; i++) {
    var y = x[i];
    if (y.tagName.toLowerCase() === 'pre') toggle_vis(y);
  }

    var elem = document.getElementById("myButton1");
    if (elem.value === "Hide Global") elem.value = "Show Global";
    else elem.value = "Hide Global";
}

document.write('<input onclick="toggle_R();" type="button" value="Hide Global" id="myButton1" style="position: absolute; top: 10%; right: 2%; z-index: 200"></input>')

</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



<!--bookdown:title:start-->
<div id="header">
<h1 class="title">DNA Hydroxymethylation in Hepatocellular Carcinoma</h1>
<p class="author"><em>Francois Collin</em></p>
<p class="date"><em>2020</em></p>
</div>
<!--bookdown:title:end-->

<!--bookdown:toc:start-->
  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">
<!--bookdown:toc2:start-->
<ul>
<li><a href="#index">Preamble</a>
<ul>
<li><a href="#license">License</a></li>
</ul></li>
<li><a href="#intro"><span class="toc-section-number">1</span> Introduction</a></li>
<li><a href="#modeling-background"><span class="toc-section-number">2</span> Classification Analysis in the <code>n &lt;&lt; p</code> Context</a>
<ul>
<li><a href="#caret-for-model-evaluation"><span class="toc-section-number">2.1</span> caret for model evaluation</a></li>
<li><a href="#the-glmnet-r-package"><span class="toc-section-number">2.2</span> The <code>glmnet</code> R package</a>
<ul>
<li><a href="#alpha-hyper-parameter"><strong>alpha</strong> hyper-parameter</a></li>
</ul></li>
<li><a href="#signal-to-noise"><span class="toc-section-number">2.3</span> Signal-to-noise Ratio</a></li>
<li><a href="#lasso-vs-best-sub"><span class="toc-section-number">2.4</span> Lasso vs Best Subset</a></li>
</ul></li>
<li><a href="#hcc-5hmcseq-preproc"><span class="toc-section-number">3</span> HCC 5hmC-Seq : Preprocessing</a>
<ul>
<li><a href="#load-the-data"><span class="toc-section-number">3.1</span> Load the data</a></li>
<li><a href="#dra"><span class="toc-section-number">3.2</span> Differential representation analysis</a>
<ul>
<li><a href="#remove-lowly-expressed-genes">Remove lowly expressed genes</a></li>
<li><a href="#creating-a-design-matrix-and-contrasts">Creating a design matrix and contrasts</a></li>
<li><a href="#removing-heteroscedasticity-from-the-count-data">Removing heteroscedasticity from the count data</a></li>
<li><a href="#fit-linear-models-and-examine-the-results">Fit linear models and examine the results</a></li>
<li><a href="#graphical-representations-of-de-results-md-plots">Graphical representations of DE results: MD Plots</a></li>
<li><a href="#de-genes-at-10-fold-change">DE genes at 10% fold change</a></li>
</ul></li>
<li><a href="#snr-regime"><span class="toc-section-number">3.3</span> Signal-to-noise ratio regime</a></li>
</ul></li>
<li><a href="#hcc-5hmcseq-explore-sparsity"><span class="toc-section-number">4</span> HCC 5hmC-Seq: Exploring Sparsity</a>
<ul>
<li><a href="#cross-validation-analysis-setup"><span class="toc-section-number">4.1</span> Cross-validation analysis setup</a></li>
<li><a href="#fit-and-compare-models"><span class="toc-section-number">4.2</span> Fit and compare models</a>
<ul>
<li><a href="#logistic-regression-in-glmnet"><span class="toc-section-number">4.2.1</span> Logistic regression in <code>glmnet</code></a></li>
</ul></li>
<li><a href="#the-relaxed-lasso-and-blended-mix-models"><span class="toc-section-number">4.3</span> The relaxed lasso and blended mix models</a></li>
<li><a href="#examination-of-sensitivity-vs-specificity"><span class="toc-section-number">4.4</span> Examination of sensitivity vs specificity</a>
<ul>
<li><a href="#training-data-out-of-fold-roc-curves"><span class="toc-section-number">4.4.1</span> Training data out-of-fold ROC curves</a></li>
</ul></li>
<li><a href="#compare-predictions-at-misclassified-samples"><span class="toc-section-number">4.5</span> Compare predictions at misclassified samples</a></li>
<li><a href="#compare-coefficient-profiles"><span class="toc-section-number">4.6</span> Compare coefficient profiles</a></li>
<li><a href="#examine-feature-selection"><span class="toc-section-number">4.7</span> Examine feature selection</a></li>
</ul></li>
<li><a href="#hcc-5hmcseq-model-suite"><span class="toc-section-number">5</span> HCC 5hmC-Seq: Sample size investigation</a>
<ul>
<li><a href="#full-data-set-fit"><span class="toc-section-number">5.1</span> Full data set fit</a>
<ul>
<li><a href="#get-sample-consistency-scores">Get sample consistency scores</a></li>
</ul></li>
<li><a href="#selected-feature-list-stability"><span class="toc-section-number">5.2</span> Selected feature list stability</a></li>
<li><a href="#simulation-design"><span class="toc-section-number">5.3</span> Simulation Design</a></li>
<li><a href="#setup-simulation"><span class="toc-section-number">5.4</span> Setup simulation</a></li>
<li><a href="#run-simulations"><span class="toc-section-number">5.5</span> Run simulations</a></li>
<li><a href="#simulation-results"><span class="toc-section-number">5.6</span> Simulation results</a>
<ul>
<li><a href="#simulation-results---look-at-one-simulation"><span class="toc-section-number">5.6.1</span> Simulation Results - look at one simulation</a></li>
<li><a href="#summarize-results-across-simulation-runs."><span class="toc-section-number">5.6.2</span> Summarize results across simulation runs.</a></li>
</ul></li>
<li><a href="#effect-of-sample-consistency"><span class="toc-section-number">5.7</span> Effect of sample consistency</a></li>
</ul></li>
<li><a href="#brca-rnaseq-preproc"><span class="toc-section-number">6</span> BrCa RNA-Seq: Preprocessing</a>
<ul>
<li><a href="#load-the-data-1"><span class="toc-section-number">6.1</span> Load the data</a></li>
<li><a href="#breast-rnaseq-dra"><span class="toc-section-number">6.2</span> Differential representation analysis</a>
<ul>
<li><a href="#remove-lowly-expressed-genes-1">Remove lowly expressed genes</a></li>
<li><a href="#creating-a-design-matrix-and-contrasts-1">Creating a design matrix and contrasts</a></li>
<li><a href="#fit-linear-models-and-examine-the-results-1">Fit linear models and examine the results</a></li>
<li><a href="#graphical-representations-of-de-results-md-plots-1">Graphical representations of DE results: MD Plots</a></li>
<li><a href="#de-genes-at-10-fold-change-1">DE genes at 10% fold change</a></li>
</ul></li>
<li><a href="#breast-rnaseq-snr-regime"><span class="toc-section-number">6.3</span> Signal-to-noise ratio regime</a></li>
</ul></li>
<li><a href="#brca-rnaseseq-explore-sparsity"><span class="toc-section-number">7</span> BrCa RNA-Seq: Exploring Sparsity</a>
<ul>
<li><a href="#cross-validation-analysis-setup-1"><span class="toc-section-number">7.1</span> Cross-validation analysis setup</a></li>
<li><a href="#fit-and-compare-models-1"><span class="toc-section-number">7.2</span> Fit and compare models</a></li>
<li><a href="#the-relaxed-lasso-and-blended-mix-models-1"><span class="toc-section-number">7.3</span> The relaxed lasso and blended mix models</a></li>
<li><a href="#examination-of-sensitivity-vs-specificity-1"><span class="toc-section-number">7.4</span> Examination of sensitivity vs specificity</a>
<ul>
<li><a href="#training-data-out-of-fold-roc-curves-1"><span class="toc-section-number">7.4.1</span> Training data out-of-fold ROC curves</a></li>
</ul></li>
<li><a href="#compare-predictions-at-misclassified-samples-1"><span class="toc-section-number">7.5</span> Compare predictions at misclassified samples</a></li>
<li><a href="#compare-coefficient-profiles-1"><span class="toc-section-number">7.6</span> Compare coefficient profiles</a></li>
<li><a href="#examine-feature-selection-1"><span class="toc-section-number">7.7</span> Examine feature selection</a></li>
</ul></li>
<li><a href="#brca-rnaseq-model-suite"><span class="toc-section-number">8</span> BrCa RNA-Seq: Sample size investigation</a>
<ul>
<li><a href="#full-data-set-fit-1"><span class="toc-section-number">8.1</span> Full data set fit</a>
<ul>
<li><a href="#get-sample-consistency-scores-1">Get sample consistency scores</a></li>
</ul></li>
<li><a href="#selected-feature-list-stability-1"><span class="toc-section-number">8.2</span> Selected feature list stability</a></li>
<li><a href="#simulation-design-1"><span class="toc-section-number">8.3</span> Simulation Design</a></li>
<li><a href="#setup-simulation-1"><span class="toc-section-number">8.4</span> Setup simulation</a></li>
<li><a href="#run-simulations-1"><span class="toc-section-number">8.5</span> Run simulations</a></li>
<li><a href="#simulation-results-1"><span class="toc-section-number">8.6</span> Simulation results</a>
<ul>
<li><a href="#simulation-results---look-at-one-simulation-1"><span class="toc-section-number">8.6.1</span> Simulation Results - look at one simulation</a></li>
<li><a href="#summarize-results-across-simulation-runs.-1"><span class="toc-section-number">8.6.2</span> Summarize results across simulation runs.</a></li>
</ul></li>
</ul></li>
<li><a href="#conclusions"><span class="toc-section-number">9</span> Conclusions</a></li>
<li><a href="#references">References</a></li>
<li><a href="#appendix-1">Appendix 1 - Sample size in elastic net fits to HCC 5hmC-Seq Data</a>
<ul>
<li><a href="#selected-feature-list-stability-2">Selected feature list stability</a></li>
<li><a href="#run-simulations---enet">Run simulations - enet</a></li>
<li><a href="#enet-simulation-results">enet Simulation results</a>
<ul>
<li><a href="#simulation-results---look-at-one-simulation-2">Simulation Results - look at one simulation</a></li>
<li><a href="#summarize-results-across-simulation-runs">Summarize results across simulation runs</a></li>
</ul></li>
</ul></li>
</ul>
<!--bookdown:toc2:end-->
      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">DNA Hydroxymethylation in Hepatocellular Carcinoma</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<!--bookdown:toc:end-->
<!--bookdown:body:start-->
<!-- ONLY THIS FILE SHOULD HAVE YAML -->
<!-- THIS FILE DOESN'T HAVE TO HAVE ANY CONTENT ... -->
<style>

.watermark {
  opacity: 0.2;
  position: fixed;
  top: 50%;
  left: 50%;
  font-size: 500%;
  color: #00407d;
}

</style>
<!-- THIS DIDN'T DO ANYTHING
<div class="watermark">DRAFT</div>
-->
<!--chapter:end:index.Rmd-->
<div id="index" class="section level1 unnumbered" number="">
<h1 class="unnumbered" number="">Preamble</h1>
<p>This vignette offers some exploratory data analyses of
DNA Hydroxymethylation data available from
NCBI GEO <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE112679">Series GSE112679</a>.
These data can be conveniently accessed through an R data package.
See <a href="https://12379monty.github.io/GSE112679/">GSE112679 R Data Package page</a>.</p>
<div id="license" class="section level2 unnumbered" number="">
<h2 class="unnumbered" number="">License</h2>
<!-- From https://github.com/santisoler/cc-licenses -->
<!-- THis doesnt work with pdf -->
<!-- COMMENT OUT FOR bookdown::pdf_book ????
![](CC_4_0.png)
![](https://i.creativecommons.org/l/by/4.0/88x31.png)

-->
<p><img src="Static/images/CC_4_0.png" width="84" /></p>
<p>This work by Francois Collin is licensed under a
<a href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a></p>
</div>
</div>
<div id="intro" class="section level1" number="1">
<h1 number="1"><span class="header-section-number">1</span> Introduction</h1>
<p>The goal of detecting
cancer at the earliest stage of development with a non-invasive procedure
has busied many groups with the task of perfecting techniques to support
what has become commonly known as a
liquid biopsy - the analysis of biomarkers circulating in fluids such as blood,
saliva or urine. Epigenetic biomarkers present themselves as good candidates for this application
(Gai and Sun (2019) <span class="citation">[<a href="#ref-Gai:2019aa" role="doc-biblioref">1</a>]</span>). In particular,
given their prevalence in the human genome,
close correlation with gene expression and high chemical stability,
DNA modifications such as 5-methylcytosine (5mC) and 5-hydroxymethylcytosine (5hmC)
are DNA epigenetic marks that provide much promise as
cancer diagnosis biomarkers that could be profitably analyzed in liquid biopsies
<span class="citation">[<a href="#ref-Cai:2019aa" role="doc-biblioref">2</a>–<a href="#ref-Collin:2018aa" role="doc-biblioref">5</a>]</span>.</p>
<!--
This work has already led to some commercial products.
Thrive Earlier Detection Corp.  launched CancerSEEK, a liquid biopsy test to detect multiple cancers early in 2019.  Guardant Health offers a liquid biopsy test, Guardant360, for advanced solid 
tumor cancers. Others in the space include Karius, which focuses on a liquid biopsy for 
infectious disease, and GRAIL Bio, which was launched by Illumina in January 2016. 
In 2017, Verily Life Sciences, one of Google/Alphabet’s companies, invested in Freenome, 
another liquid biopsy company.
-->
<p>Li et al. (2017) <span class="citation">[<a href="#ref-Li:2017aa" role="doc-biblioref">3</a>]</span> used a sensitive and selective chemical labeling technology
to extract genome-wide 5hmC profiles from circulating cell-free DNA (cfDNA)
as well as from genomic DNA (gDNA)
collected from a cohort of 260 patients recently diagnosed with colorectal,
gastric, pancreatic, liver or thyroid cancer and normal tissues from 90 healthy individuals
They found 5hmC-based biomarkers of circulating cfDNA to be highly predictive of some cancer types.
Similar small sample size findings were reported in Song et al. (2017) <span class="citation">[<a href="#ref-Song:2017aa" role="doc-biblioref">4</a>]</span>.</p>
<p>Focusing on hepatocellular carcinoma, Cai et al. (2019) <span class="citation">[<a href="#ref-Cai:2019aa" role="doc-biblioref">2</a>]</span> assembled a sizable dataset
to demonstrate the feasibility of using features derived from
5-hydroxymethylcytosines marks in circulating cell-free DNA as
a non-invasive approach for the early detection of
hepatocellular carcinoma. The data that are the basis of that
report are available on the NCBI GEO web site
(<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE112679">Series GSE112679</a>).
The data have also been bundled in a R data package which can be installed from github:</p>
<pre><code>if (!requireNamespace(&quot;devtools&quot;, quietly = TRUE))
    install.packages(&quot;devtools&quot;)
devtools::install_github(&quot;12379Monty/GSE112679&quot;)</code></pre>
<p>An important question in the early development of classifiers of the sorts
that are the basis of any liquid biopsy diagnostic tool is how many samples
should be collected to make properly informed decisions. In this
report we will explore the GSE112679 data to shed some light on
the relationship between sample size and model performance
in the context classifying samples based on 5hmC data.
The same analyses are repeated on Breast Cancer RNA-Seq data
to provide a different signal to noise context in which to evaluate
classification model performance in relation to sample size.</p>
<p>The layout of the report is the following:</p>
<ul>
<li><p>In Section @ref(modeling-background) we provide background on fitting and
analyzing predictive models in the <code>n &lt;&lt; p</code> context and
give some detail about the primary tool used in this report.</p></li>
<li><p>In Section @ref(hcc-5hmcseq-preproc) we pre-process the 5hmC data that
we will use for the classification analysis and perform some light QC analyses.</p></li>
<li><p>In Section @ref(hcc-5hmcseq-explore-sparsity) we explore sparse models
that discriminate between early stage HCC and control samples based on
5hmC gene body intensity data.</p></li>
<li><p>In Section @ref(hcc-5hmcseq-model-suite) we examine the results of fitting a
suite of models to the HCC 5hmC data investigate the effect of sample
size on model performance.</p></li>
<li><p>Sections @ref(brca-rnaseq-preproc), @ref(brca-rnaseseq-explore-sparsity)
and @ref(brca-rnaseq-model-suite) repeat the analyses run on the HCC 5hmC data
on breast cancer RNA-Seq data.</p></li>
</ul>
<!-- LEAVE FEATURE OUT IS NOT INFORMATIVE
* In Section \@ref(variable-importance) we look at the question of assessing variable importance.
-->
<ul>
<li>Concluding remarks are in Section @ref(conclusions).</li>
</ul>
<!--chapter:end:01-intro.Rmd-->
</div>
<div id="modeling-background" class="section level1" number="2">
<h1 number="2"><span class="header-section-number">2</span> Classification Analysis in the <code>n &lt;&lt; p</code> Context</h1>
<!--
Refer to [first pass study](https://hcc-5hmc-analysis.netlify.app/) for
relevant exploratory data  analysis results.
-->
<!--
In the section we look at  some models fitted to discriminate between
early stage HCC and healthy and benign samples (grouped as Controls here)
from the GSE112679 data set.  


* Some questions to address with the baseline model 
   - how separable are the data: what accuracy do we expect 
   - individual sample quality scores: which samples are hard to classify?  Compute a score
in [0, 1], where 1 is perfectly good classification and 0 is perfectly bad.

-->
<p>The main challenge in calibrating predictive models to genomic data is that
there are many more features than there are example cases to fit to;
the now classic <span class="math inline">\(n &lt;&lt; p\)</span> problem.
In this scenario, fitting methods tend to over fit. The problem
can be addressed by selecting variables, regularizing the fit or both.</p>
<!--
See the Trevor Hastie talk: 
[Statistical Learning with Big Data - Trevor Hastie](https://web.stanford.edu/~hastie/TALKS/SLBD_new.pdf)
for a good discussion of this problem and potential solutions.
-->
<p>In this report, we use genomic data to illustrate analyses of regularized regression
models in the <code>n &lt;&lt; p</code> context. Much of the analysis focuses on lasso models fitted
and analyzed using tools from the <code>glmnet</code> R package.</p>
<p>Before providing some background on that glmnet package and the
models that it supports, we briefly describe another R package
which is an essential tool for anyone who is doing predictive analytics in R - <code>caret</code> R package.</p>
<div id="caret-for-model-evaluation" class="section level2" number="2.1">
<h2 number="2.1"><span class="header-section-number">2.1</span> caret for model evaluation</h2>
<p><a href="https://topepo.github.io/caret/index.html">The <code>caret</code> Package</a>
provides a rich set of functions that streamline the process for fitting and
evaluating a large number of predictive models in parallel. The package contains tools for:</p>
<ul>
<li>data splitting<br />
</li>
<li>pre-processing<br />
</li>
<li>feature selection<br />
</li>
<li>model tuning using re-sampling<br />
</li>
<li>variable importance estimation</li>
</ul>
<p>The tools facilitate the process of automating randomly splitting data sets into training,
testing and evaluating so that predictive models can be evaluated on a comparable and
exhaustive basis. Especially useful is the functionality that is provided to
repeatedly randomly stratify samples into train and test set so that any
sample selection bias is removed.</p>
<p>What makes the <code>caret</code> package extremely useful is that it provides a common interface
to an exhaustive collection of fitting procedures. Without
this common interface one has to learn the specific syntax that
used in each fitting procedure to be included in a comparative analysis,
which can be quite burdensome.</p>
<p>Some of the models which can be evaluated with caret include:
(only some of these can be used with multinomial responses)</p>
<ul>
<li>FDA - Flexible Discriminant Analysis<br />
</li>
<li>stepLDA - Linear Discriminant Analysis with Stepwise Feature Selection<br />
</li>
<li>stepQDA - Quadratic Discriminant Analysis with Stepwise Feature Selection<br />
</li>
<li>knn - k nearest neighbors<br />
</li>
<li>pam - Nearest shrunken centroids<br />
</li>
<li>rf - Random forests<br />
</li>
<li>svmRadial - Support vector machines (RBF kernel)<br />
</li>
<li>gbm - Boosted trees<br />
</li>
<li>xgbLinear - eXtreme Gradient Boosting<br />
</li>
<li>xgbTree - eXtreme Gradient Boosting<br />
</li>
<li>neuralnet - neural network</li>
</ul>
<p>Many more models can be implemented and evaluated with <code>caret</code>,
including some <code>deep learning</code> methods, <code>Simulated Annealing Feature Selection</code>
and <code>Genetic Algorithms</code>.
Many other methods found <a href="https://topepo.github.io/caret/available-models.html">here</a>
are also worth investigating.</p>
<p>We mention <code>caret</code> here because it is an extremely useful tool for
anyone interested in exploring and comparing the performance of
many predictive models which could be applied to a given predictive
analysis context. We have gone through this exercise with a number
of genomic scale, RNA-Seq data sets in the past and have found
that in this context regularized regression models perform
as well as any. For this report, we will only consider regularized classification
models and will focus on the particular set of tools for fitting and analyzing
these models provided by the <code>glmnet</code> R package <span class="citation">[<a href="#ref-Friedman:2010aa" role="doc-biblioref">6</a>]</span>.</p>
</div>
<div id="the-glmnet-r-package" class="section level2" number="2.2">
<h2 number="2.2"><span class="header-section-number">2.2</span> The <code>glmnet</code> R package</h2>
<p>Several factors favor using the <code>glmnet</code> R package to analyze
classification models fitted to genomic scale data:</p>
<ul>
<li><p>the glmnet package is a well supported package providing
extensive functionality for regularized regression and classification models.</p></li>
<li><p>the hyper-parameters of the elastic net enable us to explore
the relationship between model size, or sparsity, and predictive accuracy.
ie. we can investigate the “bet on sparsity” principle:
<em>Use a procedure that does well in sparse problems, since no procedure
does well in dense problems</em>.</p></li>
<li><p>in our experience building classifiers from genomic scale data, regularized
classification models using the elastic net penalty do as well as any other,
and are more economical in terms of computing time, especially in comparison to
the more exotic boosting algorithms.</p></li>
<li><p>the <code>lasso</code> has been shown to be near optimal for the <span class="math inline">\(n&lt;&lt;p\)</span> problem
over a wide range of signal-to-noise regiments (Hastie et al. (2017) <span class="citation">[<a href="#ref-Hastie:2017aa" role="doc-biblioref">7</a>]</span>).</p></li>
</ul>
<hr />
<p>Much of the following comes from the
<a href="https://web.stanford.edu/~hastie/glmnet/glmnet_alpha.html">Glmnet Vignette</a>.</p>
<p>Glmnet is a package that fits a generalized linear model via penalized maximum likelihood.
The regularization path is computed for the lasso or elastic net penalty at a
grid of values for the regularization parameter lambda
(<span class="citation">[<a href="#ref-Friedman:2010aa" role="doc-biblioref">6</a>,<a href="#ref-Tibshirani:2012aa" role="doc-biblioref">8</a>–<a href="#ref-Simon:2013aa" role="doc-biblioref">10</a>]</span>).</p>
<p><code>glmnet</code> solves the following problem:</p>
<p><span class="math display">\[\min_{\beta_0,\beta} \frac{1}{N} \sum_{i=1}^{N} w_i l(y_i,\beta_0+\beta^T x_i) + \lambda\left[(1-\alpha)||\beta||_2^2/2 + \alpha ||\beta||_1\right],\]</span></p>
<p>over a grid of values of <span class="math inline">\(\lambda\)</span>.
Here <span class="math inline">\(l(y,\eta)\)</span> is the negative log-likelihood contribution for observation i;
e.g. for the Gaussian case it is <span class="math inline">\(\frac{1}{2}(y-\eta)^2\)</span>.</p>
<div id="alpha-hyper-parameter" class="section level3 unnumbered" number="">
<h3 class="unnumbered" number=""><strong>alpha</strong> hyper-parameter</h3>
<p>The elastic-net penalty is controlled by <span class="math inline">\(\alpha\)</span>, and bridges the gap between
lasso (<span class="math inline">\(\alpha\)</span>=1, the default) and ridge (<span class="math inline">\(\alpha\)</span>=0).
The tuning parameter <span class="math inline">\(\lambda\)</span> controls the overall strength of the penalty.</p>
<p>It is known that the ridge penalty shrinks the coefficients of correlated predictors
towards each other while the lasso tends to pick one of them and discard the others.
The elastic-net penalty mixes these two; if predictors are correlated in groups,
an <span class="math inline">\(\alpha\)</span>=0.5 tends to select the groups in or out together.
This is a higher level parameter, and users might pick a value upfront,
else experiment with a few different values. One use of <span class="math inline">\(\alpha\)</span> is for numerical stability;
for example, the <em>elastic net with <span class="math inline">\(\alpha = 1 - \epsilon\)</span> for some small <span class="math inline">\(\epsilon\)</span>&gt;0
performs much like the lasso, but removes any degeneracies and wild behavior caused
by extreme correlations</em>.</p>
</div>
</div>
<div id="signal-to-noise" class="section level2" number="2.3">
<h2 number="2.3"><span class="header-section-number">2.3</span> Signal-to-noise Ratio</h2>
<p>A key characteristic of classification problems is the
prevailing signal-to-noise ratio (SNR) of the problem at hand.</p>
<p>To define SNR, let <span class="math inline">\((x_0, y_0) \in \mathbb{R}^p \times \mathbb{R}\)</span>
be a pair of predcitor and response variables and define
<span class="math inline">\(f(x_0) = \mathbb{E}(y_0|x_0)\)</span> and <span class="math inline">\(\epsilon = y_0 - f(x_0)\)</span> so that</p>
<p><span class="math display">\[y_0 = f(x_0) + \epsilon_0.\]</span></p>
<p>The signal-to-noise ratio in this model is defined as</p>
<p><span class="math display">\[SNR=\frac{var(f(x_0))}{var(\epsilon_0)}.\]</span></p>
<p>It is useful to relate the SNR of a model to the proportion of variance explained (PVE).
For a given prediction function g — eg. one trained on n samples
<span class="math inline">\((x_i, y_i) i = 1, \dots, n\)</span> that are i.i.d. to <span class="math inline">\((x_0, y_0)\)</span> — its
associated proportion of variance explained is defined as</p>
<p><span class="math display">\[PVE(g)=1 - \frac{\mathbb{E}(y_0-g(x_0))^2}{Var(y_0)}.\]</span></p>
<p>This is maximized when we take <span class="math inline">\(g\)</span> to be the mean function <span class="math inline">\(f\)</span> itself,
in which case</p>
<p><span class="math display">\[PVE(f) = 1 - \frac{Var(\epsilon_0)}{Var(y_0)} = \frac{SNR}{1+SNR}.\]</span></p>
<p>Or equivalently,</p>
<p><span class="math display">\[SNR = \frac{PVE}{1-PVE}.\]</span></p>
<p>Hastie, Tibshirani, and Tibshirani (2017) <span class="citation">[<a href="#ref-Hastie:2017aa" role="doc-biblioref">7</a>]</span>, point out that
PVE is typically in the 0.2 range, and much lower in financial data. It
is also much lower in 5hmC data, as we will see in the next section.</p>
<p>Note that the SNR is a different characterization of noise level than the
coefficient of variation:</p>
<p><span class="math display">\[c_v = \frac{\sigma}{\mu}=\frac{Var(y)}{\mathbb{E}(y)}\]</span></p>
<p>Note that for small SNR, SNR <span class="math inline">\(\approx\)</span> PVE.</p>
<p>See Xiang et al. (2020) <span class="citation">[<a href="#ref-Xiang:2020aa" role="doc-biblioref">11</a>]</span>, Lozoya et al. (2018) <span class="citation">[<a href="#ref-Lozoya:2018aa" role="doc-biblioref">12</a>]</span>,
Simonson et al. (2018) <span class="citation">[<a href="#ref-Simonsen:2018aa" role="doc-biblioref">13</a>]</span> and
Rapaport et al. (2013) <span class="citation">[<a href="#ref-Rapaport:2013aa" role="doc-biblioref">14</a>]</span> for SNR in RNA-Seq</p>
</div>
<div id="lasso-vs-best-sub" class="section level2" number="2.4">
<h2 number="2.4"><span class="header-section-number">2.4</span> Lasso vs Best Subset</h2>
<p>Best subset selection finds the subset of k predictors that
produces the best fit in terms of squared error, solving the non-convex problem:</p>
<p><span class="math display">\[\begin{equation}
 \min_{\beta \in \mathcal{R}^p} ||Y - X\beta||^2_2 \, \, subject \, to \, \, ||\beta||_0 \leq k
 (\#eq:bestSub)
\end{equation}\]</span></p>
<p>The lasso solves a convex relaxation of the above where we replace the
<span class="math inline">\(l_0\)</span> norm by the <span class="math inline">\(l_1\)</span> norm, namely</p>
<p><span class="math display">\[\begin{equation}

 \min_{\beta \in \mathcal{R}^p} ||Y - X\beta||^2_2 \, \, subject \, to \, \, ||\beta||_1 \leq t

 (\#eq:lasso)
\end{equation}\]</span></p>
<p>where <span class="math inline">\(||\beta||_1 = \sum_{i=1}^{p} |\beta_i|\)</span>, and <span class="math inline">\(t \geq 0\)</span> is a tuning parameter.</p>
<p>Bertsimas et al. (2016) <span class="citation">[<a href="#ref-Bertsimas:2016aa" role="doc-biblioref">15</a>]</span> presented a mixed integer optimization (MIO)
formulation for the best subset selection problem. Using these MIO solvers,
one can solve problems with p in the hundreds and even thousands.
Bertsimas et al. showed evidence that
best subset selection generally gives superior prediction accuracy compared
to forward stepwise selection and the lasso, over a variety of problem setups.</p>
<p>In Hastie et al. (2017) <span class="citation">[<a href="#ref-Hastie:2017aa" role="doc-biblioref">7</a>]</span>, the authors countered by arguing that
neither best subset selection nor the lasso uniformly dominate the other,
with best subset selection generally performing better in high signal-to-noise (SNR)
ratio regimes, and <strong>the lasso better in low SNR regimes</strong>.
Best subset selection and forward stepwise perform quite similarly over
a range of SNR contexts, but the relaxed lasso is the overall best option,
performing just about as well as the lasso in low SNR scenarios,
and as well as best subset selection in high SNR scenarios.
Hastie et al. conclude that a blended mix of lasso and relaxed lasso estimator,
the <em>shrunken relaxed lasso fit</em>, is able to use its auxiliary shrinkage
parameter (<span class="math inline">\(\gamma\)</span>) to get the “best of both worlds”:
it accepts the heavy shrinkage from the lasso when such shrinkage is helpful, and reverses it when it is not.</p>
<!--
* relaxed lasso

$$\hat{\beta}^{relax}(\lambda, \gamma) = \gamma \beta^{lasso}(\lambda) + (1 - \gamma)(\beta^{LS}(\lambda)$$

* shrunken relaxed lasso (aka the blended fit)
-->
<p>Suppose the <strong>glmnet</strong> fitted linear predictor at <span class="math inline">\(\lambda\)</span> is <span class="math inline">\(\hat{\eta}_\lambda(x)\)</span>
and the relaxed version is <span class="math inline">\(\tilde{\eta}_\lambda(x)\)</span>, then the shrunken relaxed lasso fit is</p>
<p><span class="math display">\[\begin{equation}

\tilde{\eta}_{\lambda,\gamma}(x)=(1-\gamma)\tilde{\eta}_\lambda(x) + \gamma \hat{\eta}_\lambda(x)

 (\#eq:blended)
\end{equation}\]</span></p>
<p><span class="math inline">\(\gamma \in [0,\, 1]\)</span> is an additional tuning parameter which can be selected by cross validation.</p>
<p>The de-biasing will potentially improve prediction performance, and
cross-validation will typically select a model with a smaller number of variables.
This procedure is very competitive with forward-stepwise and
best-subset regression, and has a considerable speed advantage when the
number of variables is large. This is especially true for best-subset,
but even so for forward stepwise. The latter has to plod through the
variables one-at-a-time, while glmnet will just plunge in and find a good active set.</p>
<p>Further details may be found in
Friedman, Hastie, and Tibshirani (2010),
Tibshirani et al. (2012),
Simon et al. (2011),
Simon, Friedman, and Hastie (2013) and
Hastie, Tibshirani, and Tibshirani (2017)
(<span class="citation">[<a href="#ref-Friedman:2010aa" role="doc-biblioref">6</a>–<a href="#ref-Simon:2013aa" role="doc-biblioref">10</a>]</span>).</p>
<!--chapter:end:02-modelingBackground.Rmd-->
</div>
</div>
<div id="hcc-5hmcseq-preproc" class="section level1" number="3">
<h1 number="3"><span class="header-section-number">3</span> HCC 5hmC-Seq : Preprocessing</h1>
<p>Reader note:</p>
<ul>
<li>The most interesting part of this section is Subsection @ref(snr-regime)
which tells us where on the SNR spectrum the 5hmC data lies.
In view of the work by Hastie et al. (2017) <span class="citation">[<a href="#ref-Hastie:2017aa" role="doc-biblioref">7</a>]</span>
summarized in Subsection @ref(lasso-vs-best-sub), we would expect to
lasso model, and the relaxed version, to do very well with this
classification problem.</li>
</ul>
<!--
 FN <- 'tmp'
 # Shotcuts for knitting and redering while in R session (Invoke interactive R from R/Scripts folder)
 kk <- function(n='') knitr::knit2html(paste("t", n, sep=''), envir=globalenv(),
       output=paste(FN,".html", sep=''))

 rr <- function(n='') rmarkdown::render(paste("t", n, sep=''), envir=globalenv(),
       output_file=paste(FN,".html", sep='')) ##, output_dir='Scripts')

 bb <- function(n='') browseURL(paste(FN,".html", sep=''))

 # The usual shotcuts
 zz <- function(n='') source(paste("t", n, sep=''))
-->
<div id="load-the-data" class="section level2" number="3.1">
<h2 number="3.1"><span class="header-section-number">3.1</span> Load the data</h2>
<!-- THIS ENSURES NO EVALUATION TAKES PLACE BY DEFAULT -->
<!-- TO TURN ON, SET eval=T                            -->
<!-- Add base libraries -->
<p>The data that are available from NCBI GEO
<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE112679">Series GSE112679</a>
can be conveniently accessed through an R data package.
Attaching the GSE112679 package makes the count data tables
available as well as a gene annotation table and a sample description table.
See <a href="https://12379monty.github.io/GSE112679/">GSE112679 R Data Package page</a>.
In the Cai et al. <span class="citation">[<a href="#ref-Cai:2019aa" role="doc-biblioref">2</a>]</span> paper, samples were separated into
<code>Train</code> and <code>Val-1</code> subsets for model fitting and analysis.
<code>Val-2</code> was used as an external validation set.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb2-2"><a href="#cb2-2"></a></span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="cf">if</span> (<span class="op">!</span>(<span class="st">&quot;GSE112679&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()))) {</span>
<span id="cb2-4"><a href="#cb2-4"></a>  <span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;devtools&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)) {</span>
<span id="cb2-5"><a href="#cb2-5"></a>    <span class="kw">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span>
<span id="cb2-6"><a href="#cb2-6"></a>  }</span>
<span id="cb2-7"><a href="#cb2-7"></a>  devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;12379Monty/GSE112679&quot;</span>)</span>
<span id="cb2-8"><a href="#cb2-8"></a>}</span>
<span id="cb2-9"><a href="#cb2-9"></a></span>
<span id="cb2-10"><a href="#cb2-10"></a><span class="kw">library</span>(GSE112679)</span>
<span id="cb2-11"><a href="#cb2-11"></a></span>
<span id="cb2-12"><a href="#cb2-12"></a><span class="co"># Strip GSE_ID prefix from object names</span></span>
<span id="cb2-13"><a href="#cb2-13"></a><span class="co">### MOD: (2020.09.21) keep &#39;hcc5hmC_&#39; as prefix to object names. </span></span>
<span id="cb2-14"><a href="#cb2-14"></a><span class="co">###     Otherwise we run into problems when different datasets are </span></span>
<span id="cb2-15"><a href="#cb2-15"></a><span class="co">###     analyzed</span></span>
<span id="cb2-16"><a href="#cb2-16"></a><span class="cf">for</span>(OBJ <span class="cf">in</span> <span class="kw">data</span>(<span class="dt">package=</span><span class="st">&#39;GSE112679&#39;</span>)<span class="op">$</span>results[, <span class="st">&#39;Item&#39;</span>])</span>
<span id="cb2-17"><a href="#cb2-17"></a><span class="kw">assign</span>(<span class="kw">sub</span>(<span class="st">&#39;GSE112679_&#39;</span>,<span class="st">&#39;hcc5hmC_&#39;</span>,OBJ), <span class="kw">get</span>(OBJ))</span>
<span id="cb2-18"><a href="#cb2-18"></a></span>
<span id="cb2-19"><a href="#cb2-19"></a><span class="kw">detach</span>(package<span class="op">:</span>GSE112679, <span class="dt">unload =</span> T )</span>
<span id="cb2-20"><a href="#cb2-20"></a></span>
<span id="cb2-21"><a href="#cb2-21"></a><span class="co"># Continue</span></span>
<span id="cb2-22"><a href="#cb2-22"></a></span>
<span id="cb2-23"><a href="#cb2-23"></a>hcc5hmC_sampDesc<span class="op">$</span>DxStage &lt;-<span class="st"> </span><span class="kw">with</span>(hcc5hmC_sampDesc, <span class="kw">ifelse</span>(outcome<span class="op">==</span><span class="st">&#39;HCC&#39;</span>, </span>
<span id="cb2-24"><a href="#cb2-24"></a>   <span class="kw">paste0</span>(outcome,<span class="st">&#39;:&#39;</span>,stage), outcome))</span>
<span id="cb2-25"><a href="#cb2-25"></a></span>
<span id="cb2-26"><a href="#cb2-26"></a><span class="kw">with</span>(</span>
<span id="cb2-27"><a href="#cb2-27"></a>  hcc5hmC_sampDesc <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(sampType <span class="op">==</span><span class="st"> &quot;blood&quot;</span>),</span>
<span id="cb2-28"><a href="#cb2-28"></a>  knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">table</span>(DxStage, trainValGroup, <span class="dt">exclude =</span> <span class="ot">NULL</span>),</span>
<span id="cb2-29"><a href="#cb2-29"></a>   <span class="dt">caption=</span><span class="st">&quot;GSE112679 Samples by Dx Group and Subset&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-30"><a href="#cb2-30"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span>
<span id="cb2-31"><a href="#cb2-31"></a>)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-preproc-loadData)GSE112679 Samples by Dx Group and Subset
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Train
</th>
<th style="text-align:right;">
Val-1
</th>
<th style="text-align:right;">
Val-2
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Benign
</td>
<td style="text-align:right;">
253
</td>
<td style="text-align:right;">
132
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
CHB
</td>
<td style="text-align:right;">
190
</td>
<td style="text-align:right;">
96
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Cirrhosis
</td>
<td style="text-align:right;">
73
</td>
<td style="text-align:right;">
33
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
HCC:Early
</td>
<td style="text-align:right;">
335
</td>
<td style="text-align:right;">
220
</td>
<td style="text-align:right;">
24
</td>
</tr>
<tr>
<td style="text-align:left;">
HCC:Late
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
442
</td>
<td style="text-align:right;">
13
</td>
</tr>
<tr>
<td style="text-align:left;">
HCC:NA
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
147
</td>
<td style="text-align:right;">
23
</td>
</tr>
<tr>
<td style="text-align:left;">
Healthy
</td>
<td style="text-align:right;">
269
</td>
<td style="text-align:right;">
124
</td>
<td style="text-align:right;">
177
</td>
</tr>
</tbody>
</table>
<p>For this analysis, we will consider early stage cancer samples
and healthy or benign samples from the <code>Train</code> or <code>Val-1</code> subsets.
The appropriate outcome variable will be renamed or aliased <code>group</code></p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb3-2"><a href="#cb3-2"></a></span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="co"># Subset analysis samples</span></span>
<span id="cb3-4"><a href="#cb3-4"></a>hcc5hmC_sampDesc &lt;-</span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="st">  </span>hcc5hmC_sampDesc <span class="op">%&gt;%</span></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(sampType <span class="op">==</span><span class="st"> &quot;blood&quot;</span> <span class="op">&amp;</span></span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="st">    </span>(trainValGroup <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Train&quot;</span>, <span class="st">&quot;Val-1&quot;</span>)) <span class="op">&amp;</span></span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="st">    </span>((outcome2 <span class="op">==</span><span class="st"> &quot;BenignHealthy&quot;</span>) <span class="op">|</span></span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="st">      </span>(outcome2 <span class="op">==</span><span class="st"> &quot;HCC&quot;</span> <span class="op">&amp;</span><span class="st"> </span>stage <span class="op">==</span><span class="st"> &quot;Early&quot;</span>))) <span class="op">%&gt;%</span></span>
<span id="cb3-10"><a href="#cb3-10"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">rename</span>(<span class="dt">group =</span> outcome2) <span class="op">%&gt;%</span></span>
<span id="cb3-11"><a href="#cb3-11"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">arrange</span>(group, sampID)</span>
<span id="cb3-12"><a href="#cb3-12"></a><span class="co"># Recode group</span></span>
<span id="cb3-13"><a href="#cb3-13"></a>hcc5hmC_sampDesc<span class="op">$</span>group &lt;-<span class="st"> </span><span class="kw">with</span>(</span>
<span id="cb3-14"><a href="#cb3-14"></a>  hcc5hmC_sampDesc,</span>
<span id="cb3-15"><a href="#cb3-15"></a>  <span class="kw">ifelse</span>(group <span class="op">==</span><span class="st"> &quot;BenignHealthy&quot;</span>, <span class="st">&quot;Control&quot;</span>, group)</span>
<span id="cb3-16"><a href="#cb3-16"></a>)</span>
<span id="cb3-17"><a href="#cb3-17"></a><span class="co"># set hcc5hmC_groupCol for later</span></span>
<span id="cb3-18"><a href="#cb3-18"></a>hcc5hmC_groupCol &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;#F3C300&quot;</span>, <span class="st">&quot;#875692&quot;</span>)</span>
<span id="cb3-19"><a href="#cb3-19"></a><span class="kw">names</span>(hcc5hmC_groupCol) &lt;-<span class="st"> </span><span class="kw">unique</span>(hcc5hmC_sampDesc<span class="op">$</span>group)</span>
<span id="cb3-20"><a href="#cb3-20"></a></span>
<span id="cb3-21"><a href="#cb3-21"></a><span class="kw">with</span>(hcc5hmC_sampDesc, </span>
<span id="cb3-22"><a href="#cb3-22"></a> knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">table</span>(group, <span class="dt">exclude =</span> <span class="ot">NULL</span>),</span>
<span id="cb3-23"><a href="#cb3-23"></a>  <span class="dt">caption=</span><span class="st">&quot;Samples used in this analysis&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb3-24"><a href="#cb3-24"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span>
<span id="cb3-25"><a href="#cb3-25"></a>)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-preproc-subsetSamples)Samples used in this analysis
</caption>
<thead>
<tr>
<th style="text-align:left;">
group
</th>
<th style="text-align:right;">
Freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Control
</td>
<td style="text-align:right;">
778
</td>
</tr>
<tr>
<td style="text-align:left;">
HCC
</td>
<td style="text-align:right;">
555
</td>
</tr>
</tbody>
</table>
<p>The features are counts of reads captured by chemical labeling, and indicate
the level of 5-hydroxymethylcytosines within each gene body. Cai et al. (2019),
Li et al. (2017) and Song et al. (2017) <span class="citation">[<a href="#ref-Cai:2019aa" role="doc-biblioref">2</a>–<a href="#ref-Song:2017aa" role="doc-biblioref">4</a>]</span>
all analyze 5hmC gene body counts using standard RNA-Seq methodologies, and we will
do the same here.</p>
<p>Note that before conducting any substantive analyses, the data would normally
be very carefully examined for any sign of quality variation between groups
of samples. This analysis would integrate sample meta data - where and when were
the blood samples collected - as well as library preparation and sequencing metrics
in order to detect any sign of processing artifacts that may be present in the dataset.
This is particularly important when dealing with blood samples as variable
DNA quality degradation is a well known challenge that is encountered when dealing with
such samples <span class="citation">[<a href="#ref-Huang:2017aa" role="doc-biblioref">16</a>]</span>. Although blood specimen handling protocols can be
put in place to minimize quality variation <span class="citation">[<a href="#ref-Permenter:2015aa" role="doc-biblioref">17</a>]</span>, variability
can never be completely eradicated, especially in the context of blood samples
collected by different groups, working in different environments. The problem
of variable DNA quality becomes particularly pernicious when it is compounded
with a confounding factor that sneaks in when the control sample collection
events are separated in time and space from the cancer sample collection events;
an all too common occurrence.</p>
<p>As proper data QC requires an intimate familiarity with the details of
data collection and processing, such a task cannot be undertaken here.
We will simply run a <em>minimal set of QC sanity checks</em> to make sure that
there are no apparent systematic effects in the data.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb4-2"><a href="#cb4-2"></a></span>
<span id="cb4-3"><a href="#cb4-3"></a>hcc5hmC_featureCounts &lt;-<span class="st"> </span><span class="kw">cbind</span>(</span>
<span id="cb4-4"><a href="#cb4-4"></a>  hcc5hmC_Train_featureCount,</span>
<span id="cb4-5"><a href="#cb4-5"></a>  hcc5hmC_Val1_featureCount,</span>
<span id="cb4-6"><a href="#cb4-6"></a>  hcc5hmC_Val2_featureCount</span>
<span id="cb4-7"><a href="#cb4-7"></a>)[, <span class="kw">rownames</span>(hcc5hmC_sampDesc)]</span></code></pre></div>
<p>We first look at coverage - make sure there isn’t too much disparity of coverage
across samples. To detect shared variability, samples can be annotated and ordered
according to sample features that may be linked to sample batch processing. Here we
the samples have been ordered by group and sample id (an alias of geoAcc).</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb5-2"><a href="#cb5-2"></a></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="kw">boxplot</span>(<span class="kw">log2</span>(hcc5hmC_featureCounts <span class="op">+</span><span class="st"> </span><span class="dv">1</span>),</span>
<span id="cb5-5"><a href="#cb5-5"></a>  <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">3</span>, <span class="dv">11</span>), <span class="dt">ylab=</span><span class="st">&#39;log2 Count&#39;</span>,</span>
<span id="cb5-6"><a href="#cb5-6"></a>  <span class="dt">staplewex =</span> <span class="dv">0</span>,       <span class="co"># remove horizontal whisker lines</span></span>
<span id="cb5-7"><a href="#cb5-7"></a>  <span class="dt">staplecol =</span> <span class="st">&quot;white&quot;</span>, <span class="co"># just to be totally sure :)</span></span>
<span id="cb5-8"><a href="#cb5-8"></a>  <span class="dt">outline =</span> F,         <span class="co"># remove outlying points</span></span>
<span id="cb5-9"><a href="#cb5-9"></a>  <span class="dt">whisklty =</span> <span class="dv">0</span>,        <span class="co"># remove vertical whisker lines</span></span>
<span id="cb5-10"><a href="#cb5-10"></a>  <span class="dt">las =</span> <span class="dv">2</span>, <span class="dt">horizontal =</span> F, <span class="dt">xaxt =</span> <span class="st">&quot;n&quot;</span>,</span>
<span id="cb5-11"><a href="#cb5-11"></a>  <span class="dt">border =</span> hcc5hmC_groupCol[hcc5hmC_sampDesc<span class="op">$</span>group]</span>
<span id="cb5-12"><a href="#cb5-12"></a>)</span>
<span id="cb5-13"><a href="#cb5-13"></a><span class="kw">legend</span>(<span class="st">&quot;top&quot;</span>, <span class="dt">legend =</span> <span class="kw">names</span>(hcc5hmC_groupCol), <span class="dt">text.col =</span> hcc5hmC_groupCol, </span>
<span id="cb5-14"><a href="#cb5-14"></a>  <span class="dt">ncol =</span> <span class="dv">2</span>, <span class="dt">bty =</span> <span class="st">&quot;n&quot;</span>)</span>
<span id="cb5-15"><a href="#cb5-15"></a><span class="co"># Add reference lines</span></span>
<span id="cb5-16"><a href="#cb5-16"></a>SampleMedian &lt;-<span class="st"> </span><span class="kw">apply</span>(<span class="kw">log2</span>(hcc5hmC_featureCounts <span class="op">+</span><span class="st"> </span><span class="dv">1</span>), <span class="dv">2</span>, median)</span>
<span id="cb5-17"><a href="#cb5-17"></a><span class="kw">abline</span>(<span class="dt">h =</span> <span class="kw">median</span>(SampleMedian), <span class="dt">col =</span> <span class="st">&quot;grey&quot;</span>)</span>
<span id="cb5-18"><a href="#cb5-18"></a><span class="kw">axis</span>(<span class="dt">side =</span> <span class="dv">4</span>, <span class="dt">at =</span> <span class="kw">round</span>(<span class="kw">median</span>(SampleMedian), <span class="dv">1</span>), </span>
<span id="cb5-19"><a href="#cb5-19"></a>  <span class="dt">las =</span> <span class="dv">2</span>, <span class="dt">col =</span> <span class="st">&quot;grey&quot;</span>, <span class="dt">line =</span> <span class="dv">-1</span>, <span class="dt">tick =</span> F)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-preproc-lcpmBxp-1.png" alt="Sample log2 count boxplots" width="960" />
<p class="caption">
(#fig:hcc5hmC-preproc-lcpmBxp)Sample log2 count boxplots
</p>
</div>
<!--
Coverage level looks fairly comparable across samples.  It is sometimes helpful to
keep track of the actual coverage which can be adequately tracked by distribution
quantiles.
-->
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb6-2"><a href="#cb6-2"></a></span>
<span id="cb6-3"><a href="#cb6-3"></a>hcc5hmC_featureCounts_quant &lt;-<span class="st"> </span><span class="kw">apply</span>(hcc5hmC_featureCounts, <span class="dv">2</span>, <span class="cf">function</span>(CC) {</span>
<span id="cb6-4"><a href="#cb6-4"></a>  <span class="kw">c</span>(<span class="kw">quantile</span>(CC, <span class="dt">prob =</span> <span class="kw">c</span>(.<span class="dv">15</span>, (<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>)), <span class="dt">totCovM =</span> <span class="kw">sum</span>(CC) <span class="op">/</span><span class="st"> </span><span class="fl">1e6</span>)</span>
<span id="cb6-5"><a href="#cb6-5"></a>})</span>
<span id="cb6-6"><a href="#cb6-6"></a></span>
<span id="cb6-7"><a href="#cb6-7"></a>hcc5hmC_featureCounts_quant2 &lt;-<span class="st"> </span><span class="kw">apply</span>(hcc5hmC_featureCounts_quant, <span class="dv">1</span>, <span class="cf">function</span>(RR) {</span>
<span id="cb6-8"><a href="#cb6-8"></a>  <span class="kw">quantile</span>(RR, <span class="dt">prob =</span> (<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>)</span>
<span id="cb6-9"><a href="#cb6-9"></a>})</span>
<span id="cb6-10"><a href="#cb6-10"></a></span>
<span id="cb6-11"><a href="#cb6-11"></a>knitr<span class="op">::</span><span class="kw">kable</span>(hcc5hmC_featureCounts_quant2,</span>
<span id="cb6-12"><a href="#cb6-12"></a>  <span class="dt">digits =</span> <span class="dv">1</span>,</span>
<span id="cb6-13"><a href="#cb6-13"></a>  <span class="dt">caption =</span> <span class="kw">paste</span>(</span>
<span id="cb6-14"><a href="#cb6-14"></a>    <span class="st">&quot;Coverage Summary - Columns are sample coverage quantiles and total coverage&quot;</span>,</span>
<span id="cb6-15"><a href="#cb6-15"></a>    <span class="st">&quot;</span><span class="ch">\n</span><span class="st">Rows are quartiles across samples&quot;</span></span>
<span id="cb6-16"><a href="#cb6-16"></a>  )</span>
<span id="cb6-17"><a href="#cb6-17"></a>) <span class="op">%&gt;%</span><span class="st"> </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-preproc-hcc5hmC-quantCounts)Coverage Summary - Columns are sample coverage quantiles and total coverage
Rows are quartiles across samples
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
15%
</th>
<th style="text-align:right;">
25%
</th>
<th style="text-align:right;">
50%
</th>
<th style="text-align:right;">
75%
</th>
<th style="text-align:right;">
totCovM
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
25%
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
24
</td>
<td style="text-align:right;">
111
</td>
<td style="text-align:right;">
321
</td>
<td style="text-align:right;">
5.5
</td>
</tr>
<tr>
<td style="text-align:left;">
50%
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
135
</td>
<td style="text-align:right;">
391
</td>
<td style="text-align:right;">
6.7
</td>
</tr>
<tr>
<td style="text-align:left;">
75%
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
35
</td>
<td style="text-align:right;">
162
</td>
<td style="text-align:right;">
468
</td>
<td style="text-align:right;">
8.0
</td>
</tr>
</tbody>
</table>
<p>From this table, we see that 25% of the samples have total coverage exceeding
8M reads, 25% of samples
have a 15 percentile of coverage lower than
4, etc.</p>
<!-- SKIP
We next look at relative log representation (RLR) (in the context of measuring the density of 
5hmC marks in genes, we refer to `representation` as opposed to `expression`; the
two can be used interchangeably) -
make sure the shapes of the distributions are not widely different.
-->
<!-- SKIPPED
We note that the HCC samples have slightly more variable coverage distribution.
A few samples are quite different.
-->
</div>
<div id="dra" class="section level2" number="3.2">
<h2 number="3.2"><span class="header-section-number">3.2</span> Differential representation analysis</h2>
<p>In the remainder of this section, we will process the data and
perform differential expression analysis as outlined in
Law et al. (2018) <span class="citation">[<a href="#ref-Law:2018aa" role="doc-biblioref">18</a>]</span>. The main analysis steps are:</p>
<ul>
<li>remove lowly expressed genes</li>
<li>normalize gene expression distributions</li>
<li>remove heteroscedasticity</li>
<li>fit linear models and examine DE results</li>
</ul>
<p>It is good practice to perform this differential expression analysis prior to
fitting models to get an idea of how difficult it will be to discriminate
between samples belonging to the different subgroups. The pipeline
outlined in Law et al. (2018) <span class="citation">[<a href="#ref-Law:2018aa" role="doc-biblioref">18</a>]</span> also provides some
basic quality assessment opportunities.</p>
<div id="remove-lowly-expressed-genes" class="section level3 unnumbered" number="">
<h3 class="unnumbered" number="">Remove lowly expressed genes</h3>
<p>Genes that are not expressed at a biologically
meaningful level in any condition should be discarded to reduce the
subset of genes to those that are of interest, and to reduce the number of tests
carried out downstream when looking at differential expression. Carrying
uninformative genes may also be a hindrance to classification and other
downstream analyses.</p>
<p>To determine a sensible threshold we can begin by examining the shapes of the distributions.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb7-2"><a href="#cb7-2"></a></span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="kw">plot</span>(<span class="kw">density</span>(hcc5hmC_lcpm_mtx[, <span class="dv">1</span>]),</span>
<span id="cb7-5"><a href="#cb7-5"></a>  <span class="dt">col =</span> hcc5hmC_groupCol[hcc5hmC_sampDesc<span class="op">$</span>group[<span class="dv">1</span>]],</span>
<span id="cb7-6"><a href="#cb7-6"></a>  <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">.25</span>), <span class="dt">las =</span> <span class="dv">2</span>, <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;log2 CPM&quot;</span></span>
<span id="cb7-7"><a href="#cb7-7"></a>)</span>
<span id="cb7-8"><a href="#cb7-8"></a><span class="kw">abline</span>(<span class="dt">v =</span> <span class="dv">0</span>, <span class="dt">col =</span> <span class="dv">3</span>)</span>
<span id="cb7-9"><a href="#cb7-9"></a><span class="co"># After verifying no outliers, can plot a random subset </span></span>
<span id="cb7-10"><a href="#cb7-10"></a><span class="cf">for</span> (JJ <span class="cf">in</span> <span class="kw">sample</span>(<span class="dv">2</span><span class="op">:</span><span class="kw">ncol</span>(hcc5hmC_lcpm_mtx), <span class="dt">size =</span> <span class="dv">100</span>)) {</span>
<span id="cb7-11"><a href="#cb7-11"></a>  den &lt;-<span class="st"> </span><span class="kw">density</span>(hcc5hmC_lcpm_mtx[, JJ])</span>
<span id="cb7-12"><a href="#cb7-12"></a>  <span class="kw">lines</span>(den<span class="op">$</span>x, den<span class="op">$</span>y, <span class="dt">col =</span> hcc5hmC_groupCol[hcc5hmC_sampDesc<span class="op">$</span>group[JJ]], <span class="dt">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb7-13"><a href="#cb7-13"></a>} <span class="co"># for(JJ</span></span>
<span id="cb7-14"><a href="#cb7-14"></a><span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="dt">legend =</span> <span class="kw">names</span>(hcc5hmC_groupCol), </span>
<span id="cb7-15"><a href="#cb7-15"></a>  <span class="dt">text.col =</span> hcc5hmC_groupCol, <span class="dt">bty =</span> <span class="st">&quot;n&quot;</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-preproc-densityLcpm-1.png" alt="Sample $log_2$ CPM densities" width="960" />
<p class="caption">
(#fig:hcc5hmC-preproc-densityLcpm)Sample <span class="math inline">\(log_2\)</span> CPM densities
</p>
</div>
<p>As is typically the case with RNA-Seq data, we notice many weakly represented genes
in this dataset. A cpm value of 1 appears to adequatly separate
the expressed from the un-expressed genes, but we will be slightly more strict here
and require a CPM threshold of 3 . Using a nominal CPM value of
3, genes are deeemed to be <code>represented</code> if their expression is
above this threshold, and not represented otherwise.
For this analysis we will require that genes be <code>represented</code> in at least
25 samples across the entire dataset to be retained for downstream analysis.
Here, a CPM value of 3 means that a gene is represented if it
has at least 9 reads in the sample with the
lowest sequencing depth (library size 2.9 million).
Note that the thresholds used here are arbitrary as there are no hard and fast
rules to set these by.
The voom-plot, which is part of analyses done to remove heteroscedasticity,
can be examined to verify that the filtering performed is adequate.</p>
<!-- or at least 
40 counts in the sample with the 
greatest sequencing depth (library size 13.3 million).
-->
<p>Remove weakly represented genes and re-plot densities.</p>
<p>Removing 17.5% of genes…</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb8-2"><a href="#cb8-2"></a></span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="kw">plot</span>(<span class="kw">density</span>(hcc5hmC_F_lcpm_mtx[, <span class="dv">1</span>]),</span>
<span id="cb8-5"><a href="#cb8-5"></a>  <span class="dt">col =</span> hcc5hmC_groupCol[hcc5hmC_sampDesc<span class="op">$</span>group[<span class="dv">1</span>]],</span>
<span id="cb8-6"><a href="#cb8-6"></a>  <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">.25</span>), <span class="dt">las =</span> <span class="dv">2</span>, <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;log2 CPM&quot;</span></span>
<span id="cb8-7"><a href="#cb8-7"></a>)</span>
<span id="cb8-8"><a href="#cb8-8"></a><span class="co">#abline(v = 0, col = 3)</span></span>
<span id="cb8-9"><a href="#cb8-9"></a><span class="co"># After verifying no outliers, can plot a random subset </span></span>
<span id="cb8-10"><a href="#cb8-10"></a><span class="cf">for</span> (JJ <span class="cf">in</span> <span class="kw">sample</span>(<span class="dv">2</span><span class="op">:</span><span class="kw">ncol</span>(hcc5hmC_F_lcpm_mtx), <span class="dt">size =</span> <span class="dv">100</span>)) {</span>
<span id="cb8-11"><a href="#cb8-11"></a>  den &lt;-<span class="st"> </span><span class="kw">density</span>(hcc5hmC_F_lcpm_mtx[, JJ])</span>
<span id="cb8-12"><a href="#cb8-12"></a>  <span class="kw">lines</span>(den<span class="op">$</span>x, den<span class="op">$</span>y, <span class="dt">col =</span> hcc5hmC_groupCol[hcc5hmC_sampDesc<span class="op">$</span>group[JJ]], <span class="dt">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb8-13"><a href="#cb8-13"></a>} <span class="co"># for(JJ</span></span>
<span id="cb8-14"><a href="#cb8-14"></a><span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="dt">legend =</span> <span class="kw">names</span>(hcc5hmC_groupCol), </span>
<span id="cb8-15"><a href="#cb8-15"></a>  <span class="dt">text.col =</span> hcc5hmC_groupCol, <span class="dt">bty =</span> <span class="st">&quot;n&quot;</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-preproc-densityLcpm2-1.png" alt="Sample $log_2$ CPM densities after removing weak genes" width="960" />
<p class="caption">
(#fig:hcc5hmC-preproc-densityLcpm2)Sample <span class="math inline">\(log_2\)</span> CPM densities after removing weak genes
</p>
</div>
<!--
Note that the $log_2(CMP)$ distribution is not quite symmetric.
-->
<p>As another sanity check, we will look at a
multidimensional scaling plot of distances between gene expression
profiles. We use <code>plotMDS</code> in limma package <span class="citation">[<a href="#ref-Ritchie:2015aa" role="doc-biblioref">19</a>]</span>),
which plots samples on a two-dimensional scatterplot so that distances on
the plot approximate the typical log2 fold changes between the
samples.</p>
<p>Before producing the MDS plot we will normalize the distributions.
We will store the data into s <code>DGEList</code> object as this is convenient
when running many of the analyses implemented in the edgeR and limma packages.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb9-2"><a href="#cb9-2"></a></span>
<span id="cb9-3"><a href="#cb9-3"></a>hcc5hmC_F_dgel &lt;-<span class="st"> </span>edgeR<span class="op">::</span><span class="kw">DGEList</span>(</span>
<span id="cb9-4"><a href="#cb9-4"></a>  <span class="dt">counts =</span> hcc5hmC_featureCounts_F,</span>
<span id="cb9-5"><a href="#cb9-5"></a>  <span class="dt">genes =</span> hcc5hmC_genes_annot_F,</span>
<span id="cb9-6"><a href="#cb9-6"></a>  <span class="dt">samples =</span> hcc5hmC_sampDesc,</span>
<span id="cb9-7"><a href="#cb9-7"></a>  <span class="dt">group =</span> hcc5hmC_sampDesc<span class="op">$</span>group</span>
<span id="cb9-8"><a href="#cb9-8"></a>)</span>
<span id="cb9-9"><a href="#cb9-9"></a>hcc5hmC_F_dgel &lt;-<span class="st"> </span>edgeR<span class="op">::</span><span class="kw">calcNormFactors</span>(hcc5hmC_F_dgel)</span>
<span id="cb9-10"><a href="#cb9-10"></a>hcc5hmC_F_lcpm_mtx &lt;-<span class="st"> </span>edgeR<span class="op">::</span><span class="kw">cpm</span>(hcc5hmC_F_dgel, <span class="dt">log =</span> T)</span>
<span id="cb9-11"><a href="#cb9-11"></a></span>
<span id="cb9-12"><a href="#cb9-12"></a><span class="co"># Save hcc5hmC_F_dgel to facilitate restarting</span></span>
<span id="cb9-13"><a href="#cb9-13"></a><span class="co"># remove from final version</span></span>
<span id="cb9-14"><a href="#cb9-14"></a><span class="kw">save</span>(<span class="dt">list =</span> <span class="st">&quot;hcc5hmC_F_dgel&quot;</span>, <span class="dt">file =</span> <span class="st">&quot;RData/hcc5hmC_F_dgel&quot;</span>)</span></code></pre></div>
<p>Verify that the counts are properly normalized.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb10-2"><a href="#cb10-2"></a></span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="kw">boxplot</span>(hcc5hmC_F_lcpm_mtx,</span>
<span id="cb10-5"><a href="#cb10-5"></a>  <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">8</span>), <span class="dt">ylab=</span><span class="st">&#39;Normalized Log CPM&#39;</span>,</span>
<span id="cb10-6"><a href="#cb10-6"></a>  <span class="dt">staplewex =</span> <span class="dv">0</span>,       <span class="co"># remove horizontal whisker lines</span></span>
<span id="cb10-7"><a href="#cb10-7"></a>  <span class="dt">staplecol =</span> <span class="st">&quot;white&quot;</span>, <span class="co"># just to be totally sure :)</span></span>
<span id="cb10-8"><a href="#cb10-8"></a>  <span class="dt">outline =</span> F,         <span class="co"># remove outlying points</span></span>
<span id="cb10-9"><a href="#cb10-9"></a>  <span class="dt">whisklty =</span> <span class="dv">0</span>,        <span class="co"># remove vertical whisker lines</span></span>
<span id="cb10-10"><a href="#cb10-10"></a>  <span class="dt">las =</span> <span class="dv">2</span>, <span class="dt">horizontal =</span> F, <span class="dt">xaxt =</span> <span class="st">&quot;n&quot;</span>,</span>
<span id="cb10-11"><a href="#cb10-11"></a>  <span class="dt">border =</span> hcc5hmC_groupCol[hcc5hmC_sampDesc<span class="op">$</span>group]</span>
<span id="cb10-12"><a href="#cb10-12"></a>)</span>
<span id="cb10-13"><a href="#cb10-13"></a><span class="kw">legend</span>(<span class="st">&quot;top&quot;</span>, <span class="dt">legend =</span> <span class="kw">names</span>(hcc5hmC_groupCol), <span class="dt">text.col =</span> hcc5hmC_groupCol,</span>
<span id="cb10-14"><a href="#cb10-14"></a>  <span class="dt">ncol =</span> <span class="dv">2</span>, <span class="dt">bty =</span> <span class="st">&quot;n&quot;</span>)</span>
<span id="cb10-15"><a href="#cb10-15"></a><span class="co"># Add reference lines</span></span>
<span id="cb10-16"><a href="#cb10-16"></a>SampleMedian &lt;-<span class="st"> </span><span class="kw">apply</span>(hcc5hmC_F_lcpm_mtx, <span class="dv">2</span>, median)</span>
<span id="cb10-17"><a href="#cb10-17"></a><span class="kw">abline</span>(<span class="dt">h =</span> <span class="kw">median</span>(SampleMedian), <span class="dt">col =</span> <span class="st">&quot;grey&quot;</span>)</span>
<span id="cb10-18"><a href="#cb10-18"></a><span class="kw">axis</span>(<span class="dt">side =</span> <span class="dv">4</span>, <span class="dt">at =</span> <span class="kw">round</span>(<span class="kw">median</span>(SampleMedian), <span class="dv">1</span>),</span>
<span id="cb10-19"><a href="#cb10-19"></a>  <span class="dt">las =</span> <span class="dv">2</span>, <span class="dt">col =</span> <span class="st">&quot;grey&quot;</span>, <span class="dt">line =</span> <span class="dv">-1</span>, <span class="dt">tick =</span> F)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-preproc-normedLcpmBxp-1.png" alt="Sample log2 count boxplots" width="960" />
<p class="caption">
(#fig:hcc5hmC-preproc-normedLcpmBxp)Sample log2 count boxplots
</p>
</div>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb11-2"><a href="#cb11-2"></a></span>
<span id="cb11-3"><a href="#cb11-3"></a><span class="kw">par</span>(<span class="dt">mfcol =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">xpd =</span> <span class="ot">NA</span>, <span class="dt">oma =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">0</span>))</span>
<span id="cb11-4"><a href="#cb11-4"></a></span>
<span id="cb11-5"><a href="#cb11-5"></a><span class="co"># wo loss of generality, sample 500 samples</span></span>
<span id="cb11-6"><a href="#cb11-6"></a><span class="co"># simply a matter of convenience to save time</span></span>
<span id="cb11-7"><a href="#cb11-7"></a><span class="co"># remove from final version</span></span>
<span id="cb11-8"><a href="#cb11-8"></a><span class="kw">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb11-9"><a href="#cb11-9"></a>samp_ndx &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(hcc5hmC_F_lcpm_mtx), <span class="dt">size =</span> <span class="dv">500</span>)</span>
<span id="cb11-10"><a href="#cb11-10"></a>MDS.out &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">plotMDS</span>(hcc5hmC_F_lcpm_mtx[, samp_ndx],</span>
<span id="cb11-11"><a href="#cb11-11"></a>  <span class="dt">col =</span> hcc5hmC_groupCol[hcc5hmC_sampDesc<span class="op">$</span>group[samp_ndx]], <span class="dt">pch =</span> <span class="dv">1</span></span>
<span id="cb11-12"><a href="#cb11-12"></a>)</span>
<span id="cb11-13"><a href="#cb11-13"></a><span class="kw">legend</span>(<span class="st">&quot;topleft&quot;</span>,</span>
<span id="cb11-14"><a href="#cb11-14"></a>  <span class="dt">legend =</span> <span class="kw">names</span>(hcc5hmC_groupCol),</span>
<span id="cb11-15"><a href="#cb11-15"></a>  <span class="dt">text.col =</span> hcc5hmC_groupCol, <span class="dt">bty =</span> <span class="st">&quot;n&quot;</span></span>
<span id="cb11-16"><a href="#cb11-16"></a>)</span>
<span id="cb11-17"><a href="#cb11-17"></a></span>
<span id="cb11-18"><a href="#cb11-18"></a>MDS.out &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">plotMDS</span>(hcc5hmC_F_lcpm_mtx[, samp_ndx],</span>
<span id="cb11-19"><a href="#cb11-19"></a>  <span class="dt">col =</span> hcc5hmC_groupCol[hcc5hmC_sampDesc<span class="op">$</span>group[samp_ndx]], <span class="dt">pch =</span> <span class="dv">1</span>,</span>
<span id="cb11-20"><a href="#cb11-20"></a>  <span class="dt">dim.plot =</span> <span class="dv">3</span><span class="op">:</span><span class="dv">4</span></span>
<span id="cb11-21"><a href="#cb11-21"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-preproc-plotMDS-1.png" alt="MDS plots of log-CPM values" width="960" />
<p class="caption">
(#fig:hcc5hmC-preproc-plotMDS)MDS plots of log-CPM values
</p>
</div>
<p>The MDS plot, which is analogous to a PCA plot adapted to gene expression data,
does not indicate strong clustering of samples. The fanning pattern observed in the
first two dimensions indicates that a few samples are drifting way from the
core set, but in no particular direction. There is some structure in the
3rd and 4th dimension plot which should be investigated.<br />
<code>glMDSPlot</code> from package <code>Glimma</code> provides an interactive MDS
plot that can extremely usful for exploration</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb12-2"><a href="#cb12-2"></a></span>
<span id="cb12-3"><a href="#cb12-3"></a>Glimma<span class="op">::</span><span class="kw">glMDSPlot</span>(hcc5hmC_F_dgel[, samp_ndx],</span>
<span id="cb12-4"><a href="#cb12-4"></a>  <span class="dt">groups =</span> hcc5hmC_F_dgel<span class="op">$</span>samples[</span>
<span id="cb12-5"><a href="#cb12-5"></a>    samp_ndx,</span>
<span id="cb12-6"><a href="#cb12-6"></a>    <span class="kw">c</span>(<span class="st">&quot;group&quot;</span>, <span class="st">&quot;trainValGroup&quot;</span>, <span class="st">&quot;sampType&quot;</span>, <span class="st">&quot;tissue&quot;</span>, <span class="st">&quot;title&quot;</span>, <span class="st">&quot;stage&quot;</span>)</span>
<span id="cb12-7"><a href="#cb12-7"></a>  ],</span>
<span id="cb12-8"><a href="#cb12-8"></a>  <span class="dt">main =</span> <span class="kw">paste</span>(<span class="st">&quot;MDS plot: filtered counts&quot;</span>), <span class="co">#### , Excluding outlier samples&quot;),</span></span>
<span id="cb12-9"><a href="#cb12-9"></a>  <span class="dt">path =</span> <span class="st">&quot;.&quot;</span>, <span class="dt">folder =</span> figures_DIR,</span>
<span id="cb12-10"><a href="#cb12-10"></a>  <span class="dt">html =</span> <span class="kw">paste0</span>(<span class="st">&quot;GlMDSplot&quot;</span>), <span class="dt">launch =</span> F</span>
<span id="cb12-11"><a href="#cb12-11"></a>)</span></code></pre></div>
<p>Link to glMDSPlot:
<a href="Static/figures//GlMDSplot.html">Here</a></p>
<p>No obvious factor links the samples in the 3 clusters observed on the
4th MDS dimensions. The percent of variance explained by this dimension or
<span class="math inline">\(\approx\)</span> 4%. The glMDSPlot indicates further segregation along
the 6th dimension. The percent of variance explained by this dimension or
<span class="math inline">\(\approx\)</span> 2%. Tracking down this source of variability may be quite challenging,
especially without having the complete information about the sample attributes
and provenance.</p>
<p>Unwanted variability is a well-documented problem in the analysis of RNA-Seq data
(see Peixoto et al. (2015) <span class="citation">[<a href="#ref-Peixoto:2015aa" role="doc-biblioref">20</a>]</span>), and many procedures have been proposed
to reduce the effect of unwanted variation on RNA-Seq analsys results
(<span class="citation">[<a href="#ref-Peixoto:2015aa" role="doc-biblioref">20</a>–<a href="#ref-Risso:2014aa" role="doc-biblioref">22</a>]</span>). There are undoubtedly
some similar sources of systematic variation in the 5hmC data, but it is
beyond the scope of this work to investigate these in this particular dataset.
Given that the clustering of samples occurs in MDS dimensions that explain
a small fraction of variability, and that these is no association with the
factor of interest, HCC vs Control, these sources of variability should not
interfere too much with our classification analysis. It would nonetheless be interesting
to assess whether downstream results can be improved by removing this variability.</p>
</div>
<div id="creating-a-design-matrix-and-contrasts" class="section level3 unnumbered" number="">
<h3 class="unnumbered" number="">Creating a design matrix and contrasts</h3>
<p>Before proceeding with the statistical modeling used for the
differential expression analysis, we need to set up a
model design matrix.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb13-2"><a href="#cb13-2"></a></span>
<span id="cb13-3"><a href="#cb13-3"></a>Design_mtx &lt;-<span class="st"> </span><span class="kw">model.matrix</span>( <span class="op">~</span><span class="st">  </span><span class="dv">-1</span> <span class="op">+</span><span class="st"> </span>group, <span class="dt">data=</span>hcc5hmC_F_dgel<span class="op">$</span>samples)</span>
<span id="cb13-4"><a href="#cb13-4"></a><span class="kw">colnames</span>(Design_mtx) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&#39;group&#39;</span>, <span class="st">&#39;&#39;</span>, <span class="kw">colnames</span>(Design_mtx))</span>
<span id="cb13-5"><a href="#cb13-5"></a></span>
<span id="cb13-6"><a href="#cb13-6"></a><span class="kw">cat</span>(<span class="st">&quot;colSums(Design_mtx):</span><span class="ch">\n</span><span class="st">&quot;</span>)</span></code></pre></div>
<pre><code>## colSums(Design_mtx):</code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="kw">colSums</span>(Design_mtx)</span></code></pre></div>
<pre><code>## Control     HCC 
##     778     555</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>Contrasts_mtx &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">makeContrasts</span>(</span>
<span id="cb17-2"><a href="#cb17-2"></a>  <span class="dt">HCCvsControl =</span> HCC  <span class="op">-</span><span class="st"> </span>Control,</span>
<span id="cb17-3"><a href="#cb17-3"></a>  <span class="dt">levels=</span><span class="kw">colnames</span>(Design_mtx))</span>
<span id="cb17-4"><a href="#cb17-4"></a></span>
<span id="cb17-5"><a href="#cb17-5"></a><span class="kw">cat</span>(<span class="st">&quot;Contrasts:</span><span class="ch">\n</span><span class="st">&quot;</span>)</span></code></pre></div>
<pre><code>## Contrasts:</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>Contrasts_mtx</span></code></pre></div>
<pre><code>##          Contrasts
## Levels    HCCvsControl
##   Control           -1
##   HCC                1</code></pre>
</div>
<div id="removing-heteroscedasticity-from-the-count-data" class="section level3 unnumbered" number="">
<h3 class="unnumbered" number="">Removing heteroscedasticity from the count data</h3>
<p>As for RNA-Seq data, for 5hmC count data the variance is not independent of the mean.
In <code>limma</code>, the R package we are using for our analyses,
linear modeling is carried out on the log-CPM values which are assumed to be
normally distributed and the mean-variance relationship is accommodated using precision
weights calculated by the voom function. We apply this transformation next.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb21-2"><a href="#cb21-2"></a></span>
<span id="cb21-3"><a href="#cb21-3"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb21-4"><a href="#cb21-4"></a>hcc5hmC_F_voom &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">voom</span>(hcc5hmC_F_dgel, Design_mtx, <span class="dt">plot=</span>T)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-preproc-Voom1-1.png" alt="Removing heteroscedascity" width="1056" />
<p class="caption">
(#fig:hcc5hmC-preproc-Voom1)Removing heteroscedascity
</p>
</div>
<p>Note that the voom-plot provides a visual check on the level of filtering performed upstream.
If filtering of lowly-expressed genes is insufficient, a drop in variance levels can be
observed at the low end of the expression scale due to very small counts.</p>
<!--
Means (x-axis) and variances (y-axis) of each gene are plotted to show the dependence between the two before voom is applied to the data ( A) and how the trend is removed after voom precision weights are applied to the data ( B). The plot on the left is created within the voom function which extracts residual variances from fitting linear models to log-CPM transformed data. Variances are then re-scaled to quarter-root variances (or square-root of standard deviations) and plotted against the mean expression of each gene. The means are log 2-transformed mean-counts with an offset of 2. The plot on the right is created using plotSA which plots log 2 residual standard deviations against mean log-CPM values. The average log 2 residual standard deviation is marked by a horizontal blue line. In both plots, each black dot represents a gene and a red curve is fitted to these points.
-->
<!--

To get a sense of how this compares with RNA-Seq dsta, we can take a look at Figure 4 in Law et al. [@Law:2018aa]:

TICKr knitr::include_graphics(
  "Static/images/Law-fig4.gif",  dpi=100)TICK

We observe that the variability in the 5hmC data is quite a bit lower.  Statistical summaries would give
a better idea.  

-->
</div>
<div id="fit-linear-models-and-examine-the-results" class="section level3 unnumbered" number="">
<h3 class="unnumbered" number="">Fit linear models and examine the results</h3>
<p>Having properly filtered and normalized the data,
the linear models can be fitted to each gene and the results
examined to assess differential expression between the two groups
of interest, in our case HCC vs Control.</p>
<p>Table @ref(tab:lmFit) displays the counts of genes in each DE category:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb22-2"><a href="#cb22-2"></a></span>
<span id="cb22-3"><a href="#cb22-3"></a></span>
<span id="cb22-4"><a href="#cb22-4"></a> hcc5hmC_F_voom_fit &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">lmFit</span>(hcc5hmC_F_voom, Design_mtx)</span>
<span id="cb22-5"><a href="#cb22-5"></a> <span class="kw">colnames</span>(hcc5hmC_F_voom_fit<span class="op">$</span>coefficients) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">(Intercept</span><span class="ch">\\</span><span class="st">)&quot;</span>, <span class="st">&quot;Intercept&quot;</span>,</span>
<span id="cb22-6"><a href="#cb22-6"></a> <span class="kw">colnames</span>(hcc5hmC_F_voom_fit<span class="op">$</span>coefficients) )</span>
<span id="cb22-7"><a href="#cb22-7"></a></span>
<span id="cb22-8"><a href="#cb22-8"></a> hcc5hmC_F_voom_fit &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">contrasts.fit</span>(</span>
<span id="cb22-9"><a href="#cb22-9"></a>    hcc5hmC_F_voom_fit, <span class="dt">contrasts=</span>Contrasts_mtx)</span>
<span id="cb22-10"><a href="#cb22-10"></a></span>
<span id="cb22-11"><a href="#cb22-11"></a> hcc5hmC_F_voom_efit &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">eBayes</span>(hcc5hmC_F_voom_fit)</span>
<span id="cb22-12"><a href="#cb22-12"></a></span>
<span id="cb22-13"><a href="#cb22-13"></a> hcc5hmC_F_voom_efit_dt &lt;-</span>
<span id="cb22-14"><a href="#cb22-14"></a><span class="st"> </span>limma<span class="op">::</span><span class="kw">decideTests</span>(hcc5hmC_F_voom_efit,<span class="dt">adjust.method =</span> <span class="st">&quot;BH&quot;</span>, <span class="dt">p.value =</span> <span class="fl">0.05</span>)</span>
<span id="cb22-15"><a href="#cb22-15"></a> </span>
<span id="cb22-16"><a href="#cb22-16"></a> knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">summary</span>(hcc5hmC_F_voom_efit_dt)),</span>
<span id="cb22-17"><a href="#cb22-17"></a>  <span class="dt">caption=</span><span class="st">&quot;DE Results at FDR = 0.05&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb22-18"><a href="#cb22-18"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-preproc-lmFit)DE Results at FDR = 0.05
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Down
</th>
<th style="text-align:right;">
NotSig
</th>
<th style="text-align:right;">
Up
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
HCCvsControl
</td>
<td style="text-align:right;">
5214
</td>
<td style="text-align:right;">
5280
</td>
<td style="text-align:right;">
5258
</td>
</tr>
</tbody>
</table>
</div>
<div id="graphical-representations-of-de-results-md-plots" class="section level3 unnumbered" number="">
<h3 class="unnumbered" number="">Graphical representations of DE results: MD Plots</h3>
<p>To summarise results for all genes visually, mean-difference plots
(aka MA plot), which display log-FCs from the linear model fit against
the average log-CPM values can be generated using the plotMD function,
with the differentially expressed genes highlighted.</p>
<p>We may also be interested in whether certain gene features are
related to gene identification. Gene GC content, for example, might be
of interest.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb23-2"><a href="#cb23-2"></a></span>
<span id="cb23-3"><a href="#cb23-3"></a></span>
<span id="cb23-4"><a href="#cb23-4"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="fl">4.5</span>,<span class="fl">4.5</span>,<span class="dv">2</span>,<span class="dv">1</span>),<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb23-5"><a href="#cb23-5"></a></span>
<span id="cb23-6"><a href="#cb23-6"></a><span class="co"># log-fold-change vs ave-expr</span></span>
<span id="cb23-7"><a href="#cb23-7"></a>limma<span class="op">::</span><span class="kw">plotMD</span>(hcc5hmC_F_voom_efit,</span>
<span id="cb23-8"><a href="#cb23-8"></a> <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.4</span>, <span class="fl">0.4</span>),</span>
<span id="cb23-9"><a href="#cb23-9"></a> <span class="dt">column=</span><span class="st">&#39;HCCvsControl&#39;</span>,</span>
<span id="cb23-10"><a href="#cb23-10"></a> <span class="dt">status=</span>hcc5hmC_F_voom_efit_dt[,<span class="st">&#39;HCCvsControl&#39;</span>],</span>
<span id="cb23-11"><a href="#cb23-11"></a> <span class="dt">hl.pch =</span> <span class="dv">16</span>, <span class="dt">hl.col =</span> <span class="kw">c</span>(<span class="st">&quot;lightblue&quot;</span>, <span class="st">&quot;pink&quot;</span>), <span class="dt">hl.cex =</span> <span class="fl">.5</span>,</span>
<span id="cb23-12"><a href="#cb23-12"></a> <span class="dt">bg.pch =</span> <span class="dv">16</span>, <span class="dt">bg.col =</span> <span class="st">&quot;grey&quot;</span>, <span class="dt">bg.cex =</span> <span class="fl">0.5</span>,</span>
<span id="cb23-13"><a href="#cb23-13"></a> <span class="dt">main =</span> <span class="st">&#39;&#39;</span>,</span>
<span id="cb23-14"><a href="#cb23-14"></a> <span class="dt">xlab =</span> <span class="kw">paste0</span>(</span>
<span id="cb23-15"><a href="#cb23-15"></a>    <span class="st">&quot;Average log-expression: IQR=&quot;</span>,</span>
<span id="cb23-16"><a href="#cb23-16"></a>    <span class="kw">paste</span>(<span class="kw">round</span>(<span class="kw">quantile</span>(hcc5hmC_F_voom_efit<span class="op">$</span>Amean, <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>), <span class="dv">2</span>),</span>
<span id="cb23-17"><a href="#cb23-17"></a>      <span class="dt">collapse =</span> <span class="st">&quot;, &quot;</span></span>
<span id="cb23-18"><a href="#cb23-18"></a>    )</span>
<span id="cb23-19"><a href="#cb23-19"></a>  ),</span>
<span id="cb23-20"><a href="#cb23-20"></a>  <span class="dt">ylab =</span> <span class="kw">paste0</span>(</span>
<span id="cb23-21"><a href="#cb23-21"></a>    <span class="st">&quot;log-fold-change: IQR=&quot;</span>,</span>
<span id="cb23-22"><a href="#cb23-22"></a>    <span class="kw">paste</span>(<span class="kw">round</span>(<span class="kw">quantile</span>(hcc5hmC_F_voom_efit<span class="op">$</span>coefficients[, <span class="st">&#39;HCCvsControl&#39;</span>], <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>), <span class="dv">2</span>),</span>
<span id="cb23-23"><a href="#cb23-23"></a>      <span class="dt">collapse =</span> <span class="st">&quot;, &quot;</span></span>
<span id="cb23-24"><a href="#cb23-24"></a>    )</span>
<span id="cb23-25"><a href="#cb23-25"></a>  ),</span>
<span id="cb23-26"><a href="#cb23-26"></a>  <span class="dt">legend =</span> F, <span class="dt">cex.lab=</span><span class="fl">1.5</span></span>
<span id="cb23-27"><a href="#cb23-27"></a>)</span>
<span id="cb23-28"><a href="#cb23-28"></a><span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">0</span>, <span class="dt">col =</span> <span class="st">&quot;black&quot;</span>)</span>
<span id="cb23-29"><a href="#cb23-29"></a><span class="kw">rug</span>(<span class="kw">quantile</span>(hcc5hmC_F_voom_efit<span class="op">$</span>coefficients[, <span class="st">&#39;HCCvsControl&#39;</span>], <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>),</span>
<span id="cb23-30"><a href="#cb23-30"></a>  <span class="dt">col =</span> <span class="st">&quot;purple&quot;</span>,</span>
<span id="cb23-31"><a href="#cb23-31"></a>  <span class="dt">ticksize =</span> <span class="fl">.03</span>, <span class="dt">side =</span> <span class="dv">2</span>, <span class="dt">lwd =</span> <span class="dv">2</span></span>
<span id="cb23-32"><a href="#cb23-32"></a>)</span>
<span id="cb23-33"><a href="#cb23-33"></a><span class="kw">rug</span>(<span class="kw">quantile</span>(hcc5hmC_F_voom_efit<span class="op">$</span>Amean, <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>),</span>
<span id="cb23-34"><a href="#cb23-34"></a>  <span class="dt">col =</span> <span class="st">&quot;purple&quot;</span>,</span>
<span id="cb23-35"><a href="#cb23-35"></a>  <span class="dt">ticksize =</span> <span class="fl">.03</span>, <span class="dt">side =</span> <span class="dv">1</span>, <span class="dt">lwd =</span> <span class="dv">2</span></span>
<span id="cb23-36"><a href="#cb23-36"></a>)</span>
<span id="cb23-37"><a href="#cb23-37"></a></span>
<span id="cb23-38"><a href="#cb23-38"></a><span class="co"># log-fold-change vs identification</span></span>
<span id="cb23-39"><a href="#cb23-39"></a></span>
<span id="cb23-40"><a href="#cb23-40"></a><span class="kw">boxplot</span>(<span class="kw">split</span>(</span>
<span id="cb23-41"><a href="#cb23-41"></a> hcc5hmC_F_voom_efit<span class="op">$</span>coefficients[, <span class="st">&#39;HCCvsControl&#39;</span>],</span>
<span id="cb23-42"><a href="#cb23-42"></a> hcc5hmC_F_voom_efit_dt[,<span class="st">&#39;HCCvsControl&#39;</span>]),</span>
<span id="cb23-43"><a href="#cb23-43"></a> <span class="dt">outline=</span>F,</span>
<span id="cb23-44"><a href="#cb23-44"></a> <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&quot;pink&quot;</span>, <span class="st">&quot;grey&quot;</span>, <span class="st">&quot;lightblue&quot;</span>), <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>,</span>
<span id="cb23-45"><a href="#cb23-45"></a> <span class="dt">ylab=</span><span class="st">&#39;log-fold-change&#39;</span>, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span>.<span class="dv">4</span>, <span class="fl">.4</span>),</span>
<span id="cb23-46"><a href="#cb23-46"></a> <span class="dt">cex.lab=</span><span class="fl">1.5</span></span>
<span id="cb23-47"><a href="#cb23-47"></a>)</span>
<span id="cb23-48"><a href="#cb23-48"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="kw">c</span>(<span class="st">&#39;down&#39;</span>, <span class="st">&#39;notDE&#39;</span>, <span class="st">&#39;up&#39;</span>), <span class="dt">cex.axis=</span><span class="fl">1.5</span>)</span>
<span id="cb23-49"><a href="#cb23-49"></a></span>
<span id="cb23-50"><a href="#cb23-50"></a><span class="co"># gc vs identification</span></span>
<span id="cb23-51"><a href="#cb23-51"></a>genes_ndx &lt;-<span class="st"> </span><span class="kw">match</span>(<span class="kw">rownames</span>(hcc5hmC_F_voom_efit), hcc5hmC_genes_annot_F<span class="op">$</span>Symbol)</span>
<span id="cb23-52"><a href="#cb23-52"></a><span class="cf">if</span>(<span class="kw">sum</span>(<span class="kw">is.na</span>(genes_ndx))) <span class="kw">stop</span>(<span class="st">&quot;hcc5hmC_F_voom_efit/hcc5hmC_genes_annot_F: genes mismatch&quot;</span>)</span>
<span id="cb23-53"><a href="#cb23-53"></a>GC_vec &lt;-<span class="st"> </span><span class="kw">with</span>(hcc5hmC_genes_annot_F[genes_ndx,],(G<span class="op">+</span>C)<span class="op">/</span>(A<span class="op">+</span>C<span class="op">+</span>G<span class="op">+</span>T))</span>
<span id="cb23-54"><a href="#cb23-54"></a></span>
<span id="cb23-55"><a href="#cb23-55"></a></span>
<span id="cb23-56"><a href="#cb23-56"></a><span class="kw">boxplot</span>(<span class="kw">split</span>(</span>
<span id="cb23-57"><a href="#cb23-57"></a> GC_vec,</span>
<span id="cb23-58"><a href="#cb23-58"></a> hcc5hmC_F_voom_efit_dt[,<span class="st">&#39;HCCvsControl&#39;</span>]),</span>
<span id="cb23-59"><a href="#cb23-59"></a> <span class="dt">outline=</span>F,</span>
<span id="cb23-60"><a href="#cb23-60"></a> <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&quot;pink&quot;</span>, <span class="st">&quot;grey&quot;</span>, <span class="st">&quot;lightblue&quot;</span>), <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>,</span>
<span id="cb23-61"><a href="#cb23-61"></a> <span class="dt">ylab=</span><span class="st">&#39;gene-gc&#39;</span>, <span class="dt">cex.lab=</span><span class="fl">1.5</span></span>
<span id="cb23-62"><a href="#cb23-62"></a>)</span>
<span id="cb23-63"><a href="#cb23-63"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="kw">c</span>(<span class="st">&#39;down&#39;</span>, <span class="st">&#39;notDE&#39;</span>, <span class="st">&#39;up&#39;</span>), <span class="dt">cex.axis=</span><span class="fl">1.5</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-preproc-mdPlotEfit-1.png" alt="HCC vs Control - Genes Identified at FDR = 0,05" width="1056" />
<p class="caption">
(#fig:hcc5hmC-preproc-mdPlotEfit)HCC vs Control - Genes Identified at FDR = 0,05
</p>
</div>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a> <span class="co">#mtext(side=3, outer=T, cex=1.25, &quot;Genes identified at adjusted p-value=0.05&quot;)</span></span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb25-2"><a href="#cb25-2"></a></span>
<span id="cb25-3"><a href="#cb25-3"></a>hcc5hmC_featureCounts_F_logFC_sum &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb25-4"><a href="#cb25-4"></a> <span class="kw">split</span>(</span>
<span id="cb25-5"><a href="#cb25-5"></a> hcc5hmC_F_voom_efit<span class="op">$</span>coefficients[, <span class="st">&#39;HCCvsControl&#39;</span>],</span>
<span id="cb25-6"><a href="#cb25-6"></a> hcc5hmC_F_voom_efit_dt[,<span class="st">&#39;HCCvsControl&#39;</span>]),</span>
<span id="cb25-7"><a href="#cb25-7"></a> quantile, <span class="dt">prob =</span> (<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>)</span>
<span id="cb25-8"><a href="#cb25-8"></a></span>
<span id="cb25-9"><a href="#cb25-9"></a><span class="kw">colnames</span>(hcc5hmC_featureCounts_F_logFC_sum) &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">factor</span>(</span>
<span id="cb25-10"><a href="#cb25-10"></a> <span class="kw">colnames</span>(hcc5hmC_featureCounts_F_logFC_sum), </span>
<span id="cb25-11"><a href="#cb25-11"></a> <span class="dt">levels=</span><span class="kw">c</span>(<span class="st">&quot;-1&quot;</span>, <span class="st">&quot;0&quot;</span>, <span class="st">&quot;1&quot;</span>),</span>
<span id="cb25-12"><a href="#cb25-12"></a> <span class="dt">labels=</span><span class="kw">c</span>(<span class="st">&#39;down&#39;</span>, <span class="st">&#39;notDE&#39;</span>, <span class="st">&#39;up&#39;</span>)</span>
<span id="cb25-13"><a href="#cb25-13"></a>))</span>
<span id="cb25-14"><a href="#cb25-14"></a></span>
<span id="cb25-15"><a href="#cb25-15"></a></span>
<span id="cb25-16"><a href="#cb25-16"></a>knitr<span class="op">::</span><span class="kw">kable</span>(hcc5hmC_featureCounts_F_logFC_sum,</span>
<span id="cb25-17"><a href="#cb25-17"></a>  <span class="dt">digits =</span> <span class="dv">2</span>,</span>
<span id="cb25-18"><a href="#cb25-18"></a>  <span class="dt">caption =</span> <span class="st">&quot;log FC quartiles by gene identification&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb25-19"><a href="#cb25-19"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-preproc-quantlogFC)log FC quartiles by gene identification
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
down
</th>
<th style="text-align:right;">
notDE
</th>
<th style="text-align:right;">
up
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
25%
</td>
<td style="text-align:right;">
-0.07
</td>
<td style="text-align:right;">
-0.01
</td>
<td style="text-align:right;">
0.04
</td>
</tr>
<tr>
<td style="text-align:left;">
50%
</td>
<td style="text-align:right;">
-0.05
</td>
<td style="text-align:right;">
0.00
</td>
<td style="text-align:right;">
0.06
</td>
</tr>
<tr>
<td style="text-align:left;">
75%
</td>
<td style="text-align:right;">
-0.03
</td>
<td style="text-align:right;">
0.01
</td>
<td style="text-align:right;">
0.09
</td>
</tr>
</tbody>
</table>
<p>While many genes are identified, the effect sizes are quite small,
which results in a low signal-to-noise ratio context. See
Section @ref(snr-regime) below.</p>
<p>The log-fold-change distribution for up-represented genes is long-tailed,
with many high log fold-change values.
By contrast, log-fold-change distribution for down-represented genes
closer to symmetric and has few genes with low log fold-change values.
We will see how this affects the results of identifying genes with
an effect size requirement.</p>
<p>The GC content of down regulated genes tends to be slightly lower than the
rest of the genes. A statistical test would find that the difference
between the mean of the down regulated gene population is significantly different
than the mean of the other gene population even though the difference is
quite small
(-0.028).</p>
<p>These asymmetries are minor, but it would still be good to establish that
they reflect biology rather than processing artifacts.</p>
</div>
<div id="de-genes-at-10-fold-change" class="section level3 unnumbered" number="">
<h3 class="unnumbered" number="">DE genes at 10% fold change</h3>
<p>For a stricter definition on significance, one may require log-fold-changes
(log-FCs) to be above a minimum value. The treat method
(McCarthy and Smyth 2009 <span class="citation">[<a href="#ref-McCarthy:2009aa" role="doc-biblioref">23</a>]</span>) can be used to calculate p-values
from empirical Bayes moderated t-statistics with a minimum log-FC requirement.
The number of differentially expressed genes are greatly reduced if we
impose a minimal fold-change requirement of 10%.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb26-2"><a href="#cb26-2"></a></span>
<span id="cb26-3"><a href="#cb26-3"></a>hcc5hmC_F_voom_tfit &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">treat</span>(hcc5hmC_F_voom_fit, <span class="dt">lfc=</span><span class="kw">log2</span>(<span class="fl">1.10</span>))</span>
<span id="cb26-4"><a href="#cb26-4"></a>hcc5hmC_F_voom_tfit_dt &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">decideTests</span>(hcc5hmC_F_voom_tfit)</span>
<span id="cb26-5"><a href="#cb26-5"></a></span>
<span id="cb26-6"><a href="#cb26-6"></a><span class="kw">cat</span>(<span class="st">&quot;10% FC Gene Identification Summary - voom, adjust.method = BH, p.value = 0.05:</span><span class="ch">\n</span><span class="st">&quot;</span>)</span></code></pre></div>
<pre><code>## 10% FC Gene Identification Summary - voom, adjust.method = BH, p.value = 0.05:</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a><span class="kw">summary</span>(hcc5hmC_F_voom_tfit_dt)</span></code></pre></div>
<pre><code>##        HCCvsControl
## Down              3
## NotSig        15550
## Up              199</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a><span class="co"># log-fold-change vs ave-expr</span></span>
<span id="cb30-2"><a href="#cb30-2"></a>limma<span class="op">::</span><span class="kw">plotMD</span>(hcc5hmC_F_voom_efit,</span>
<span id="cb30-3"><a href="#cb30-3"></a> <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.5</span>, <span class="fl">0.5</span>),</span>
<span id="cb30-4"><a href="#cb30-4"></a> <span class="dt">column=</span><span class="st">&#39;HCCvsControl&#39;</span>,</span>
<span id="cb30-5"><a href="#cb30-5"></a> <span class="dt">status=</span>hcc5hmC_F_voom_tfit_dt[,<span class="st">&#39;HCCvsControl&#39;</span>],</span>
<span id="cb30-6"><a href="#cb30-6"></a> <span class="dt">hl.pch =</span> <span class="dv">16</span>, <span class="dt">hl.col =</span> <span class="kw">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;red&quot;</span>), <span class="dt">hl.cex =</span> <span class="fl">.7</span>,</span>
<span id="cb30-7"><a href="#cb30-7"></a> <span class="dt">bg.pch =</span> <span class="dv">16</span>, <span class="dt">bg.col =</span> <span class="st">&quot;grey&quot;</span>, <span class="dt">bg.cex =</span> <span class="fl">0.5</span>,</span>
<span id="cb30-8"><a href="#cb30-8"></a> <span class="dt">main =</span> <span class="st">&#39;&#39;</span>,</span>
<span id="cb30-9"><a href="#cb30-9"></a> <span class="dt">xlab =</span> <span class="kw">paste0</span>(</span>
<span id="cb30-10"><a href="#cb30-10"></a>    <span class="st">&quot;Average log-expression: IQR=&quot;</span>,</span>
<span id="cb30-11"><a href="#cb30-11"></a>    <span class="kw">paste</span>(<span class="kw">round</span>(<span class="kw">quantile</span>(hcc5hmC_F_voom_efit<span class="op">$</span>Amean, <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>), <span class="dv">2</span>),</span>
<span id="cb30-12"><a href="#cb30-12"></a>      <span class="dt">collapse =</span> <span class="st">&quot;, &quot;</span></span>
<span id="cb30-13"><a href="#cb30-13"></a>    )</span>
<span id="cb30-14"><a href="#cb30-14"></a>  ),</span>
<span id="cb30-15"><a href="#cb30-15"></a>  <span class="dt">ylab =</span> <span class="kw">paste0</span>(</span>
<span id="cb30-16"><a href="#cb30-16"></a>    <span class="st">&quot;log-fold-change: IQR=&quot;</span>,</span>
<span id="cb30-17"><a href="#cb30-17"></a>    <span class="kw">paste</span>(<span class="kw">round</span>(<span class="kw">quantile</span>(hcc5hmC_F_voom_efit<span class="op">$</span>coefficients[, <span class="st">&#39;HCCvsControl&#39;</span>], <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>), <span class="dv">2</span>),</span>
<span id="cb30-18"><a href="#cb30-18"></a>      <span class="dt">collapse =</span> <span class="st">&quot;, &quot;</span></span>
<span id="cb30-19"><a href="#cb30-19"></a>    )</span>
<span id="cb30-20"><a href="#cb30-20"></a>  ),</span>
<span id="cb30-21"><a href="#cb30-21"></a>  <span class="dt">legend =</span> F</span>
<span id="cb30-22"><a href="#cb30-22"></a>)</span>
<span id="cb30-23"><a href="#cb30-23"></a><span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">0</span>, <span class="dt">col =</span> <span class="st">&quot;black&quot;</span>)</span>
<span id="cb30-24"><a href="#cb30-24"></a><span class="kw">rug</span>(<span class="kw">quantile</span>(hcc5hmC_F_voom_efit<span class="op">$</span>coefficients[, <span class="st">&#39;HCCvsControl&#39;</span>], <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>),</span>
<span id="cb30-25"><a href="#cb30-25"></a>  <span class="dt">col =</span> <span class="st">&quot;purple&quot;</span>,</span>
<span id="cb30-26"><a href="#cb30-26"></a>  <span class="dt">ticksize =</span> <span class="fl">.03</span>, <span class="dt">side =</span> <span class="dv">2</span>, <span class="dt">lwd =</span> <span class="dv">2</span></span>
<span id="cb30-27"><a href="#cb30-27"></a>)</span>
<span id="cb30-28"><a href="#cb30-28"></a><span class="kw">rug</span>(<span class="kw">quantile</span>(hcc5hmC_F_voom_efit<span class="op">$</span>Amean, <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>),</span>
<span id="cb30-29"><a href="#cb30-29"></a>  <span class="dt">col =</span> <span class="st">&quot;purple&quot;</span>,</span>
<span id="cb30-30"><a href="#cb30-30"></a>  <span class="dt">ticksize =</span> <span class="fl">.03</span>, <span class="dt">side =</span> <span class="dv">1</span>, <span class="dt">lwd =</span> <span class="dv">2</span></span>
<span id="cb30-31"><a href="#cb30-31"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-preproc-mdPlotTfit-1.png" alt="HCC vs Control - Identified Genes at FDR = 0,05 and logFC &gt; 10%" width="1056" />
<p class="caption">
(#fig:hcc5hmC-preproc-mdPlotTfit)HCC vs Control - Identified Genes at FDR = 0,05 and logFC &gt; 10%
</p>
</div>
<p>As noted above, the log-fold-change distribution for the up-represented genes
is long-tailed in comparison to log-fold-change distribution for the down-represented genes.
As a result fewer down-represented than up-regulated genes are identified when a
minimum log-FC requirement is imposed.</p>
</div>
</div>
<div id="snr-regime" class="section level2" number="3.3">
<h2 number="3.3"><span class="header-section-number">3.3</span> Signal-to-noise ratio regime</h2>
<p>In Hastie et al. (2017) <span class="citation">[<a href="#ref-Hastie:2017aa" role="doc-biblioref">7</a>]</span>) results from <code>lasso</code> fits are
compared with <code>best subset</code> and <code>forward selection</code> fits and it is argued
that while <code>best subset</code> is optimal for high signal-to-noise regimes,
the lasso gains some competitive advantage when the prevailing signal-to-noise
ratio of the dataset is lowered.</p>
<p>We can extract sigma and signal from the fit objects to get SNR values for each gene
to see in what SNR regime the 5hmC gene body data are.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb31-2"><a href="#cb31-2"></a>lib.size &lt;-<span class="st"> </span><span class="kw">colSums</span>(hcc5hmC_F_dgel<span class="op">$</span>counts)</span>
<span id="cb31-3"><a href="#cb31-3"></a></span>
<span id="cb31-4"><a href="#cb31-4"></a>fit &lt;-<span class="st"> </span>hcc5hmC_F_voom_efit</span>
<span id="cb31-5"><a href="#cb31-5"></a>sx &lt;-<span class="st"> </span>fit<span class="op">$</span>Amean <span class="op">+</span><span class="st"> </span><span class="kw">mean</span>(<span class="kw">log2</span>(lib.size <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)) <span class="op">-</span><span class="st"> </span><span class="kw">log2</span>(<span class="fl">1e+06</span>)</span>
<span id="cb31-6"><a href="#cb31-6"></a>sy &lt;-<span class="st"> </span><span class="kw">sqrt</span>(fit<span class="op">$</span>sigma)</span>
<span id="cb31-7"><a href="#cb31-7"></a></span>
<span id="cb31-8"><a href="#cb31-8"></a>CV &lt;-<span class="st"> </span>sy<span class="op">/</span>sx    </span></code></pre></div>
<!-- DEBUG BCV from Section \@ref(analysis-of-coverage-variability) vs CV
pairs(cbind(BCV_mtx, CV)) 
boxplot(cbind(BCV_mtx, CV), outline=F) 
-->
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a><span class="co">#### CLEAR CACHE</span></span>
<span id="cb32-2"><a href="#cb32-2"></a></span>
<span id="cb32-3"><a href="#cb32-3"></a>Effect &lt;-<span class="st"> </span><span class="kw">abs</span>(hcc5hmC_F_voom_efit<span class="op">$</span>coefficients[,<span class="st">&#39;HCCvsControl&#39;</span>])</span>
<span id="cb32-4"><a href="#cb32-4"></a>Noise &lt;-<span class="st"> </span>hcc5hmC_F_voom_efit<span class="op">$</span>sigma</span>
<span id="cb32-5"><a href="#cb32-5"></a>SNR &lt;-<span class="st"> </span>Effect<span class="op">/</span>Noise</span>
<span id="cb32-6"><a href="#cb32-6"></a></span>
<span id="cb32-7"><a href="#cb32-7"></a><span class="kw">plot</span>(spatstat<span class="op">::</span><span class="kw">CDF</span>(<span class="kw">density</span>(SNR)),</span>
<span id="cb32-8"><a href="#cb32-8"></a>  <span class="dt">col =</span> <span class="dv">1</span>, <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">ylab =</span> <span class="st">&quot;Prob(SNR&lt;x)&quot;</span>,</span>
<span id="cb32-9"><a href="#cb32-9"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.2</span>)</span>
<span id="cb32-10"><a href="#cb32-10"></a>)</span>
<span id="cb32-11"><a href="#cb32-11"></a></span>
<span id="cb32-12"><a href="#cb32-12"></a>SNR_quant &lt;-<span class="st"> </span><span class="kw">quantile</span>(SNR, <span class="dt">prob=</span><span class="kw">c</span>((<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>)<span class="op">/</span><span class="dv">4</span>,.<span class="dv">9</span>))</span>
<span id="cb32-13"><a href="#cb32-13"></a><span class="kw">rug</span>(SNR_quant,</span>
<span id="cb32-14"><a href="#cb32-14"></a>    <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">ticksize =</span> <span class="fl">0.05</span>, <span class="dt">col =</span> <span class="dv">1</span></span>
<span id="cb32-15"><a href="#cb32-15"></a>  )</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-preproc-plotSNR-1.png" alt="Cumulative Distribution of SNR - rug = 25, 50, 75 and 90th percentile" width="960" />
<p class="caption">
(#fig:hcc5hmC-preproc-plotSNR)Cumulative Distribution of SNR - rug = 25, 50, 75 and 90th percentile
</p>
</div>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">t</span>(SNR_quant),</span>
<span id="cb33-2"><a href="#cb33-2"></a>  <span class="dt">digits =</span> <span class="dv">3</span>,</span>
<span id="cb33-3"><a href="#cb33-3"></a>  <span class="dt">caption =</span> <span class="kw">paste</span>(</span>
<span id="cb33-4"><a href="#cb33-4"></a>    <span class="st">&quot;SNR Quantiles&quot;</span>) </span>
<span id="cb33-5"><a href="#cb33-5"></a>) <span class="op">%&gt;%</span><span class="st"> </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-preproc-plotSNR)SNR Quantiles
</caption>
<thead>
<tr>
<th style="text-align:right;">
25%
</th>
<th style="text-align:right;">
50%
</th>
<th style="text-align:right;">
75%
</th>
<th style="text-align:right;">
90%
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
0.018
</td>
<td style="text-align:right;">
0.036
</td>
<td style="text-align:right;">
0.06
</td>
<td style="text-align:right;">
0.082
</td>
</tr>
</tbody>
</table>
<p>These SNR values are in the range where the lasso and relaxed lasso gain some advantage over
best subset and forward selection fits (see Hastie et al. (2017) <span class="citation">[<a href="#ref-Hastie:2017aa" role="doc-biblioref">7</a>]</span>).</p>
<!--chapter:end:03-preprocHCC5hmC.Rmd-->
</div>
</div>
<div id="hcc-5hmcseq-explore-sparsity" class="section level1" number="4">
<h1 number="4"><span class="header-section-number">4</span> HCC 5hmC-Seq: Exploring Sparsity</h1>
<p>In this section we explore various models fitted to
the HCC 5hmC-Seq data set explored in Section @ref(hcc-5hmcseq-preproc).
We focus our analyses on lasso fits which tend to favor sparse models.
These fits can be computed and analyzed with tools provided in the <code>glmnet</code> package.
Refer to the <a href="https://web.stanford.edu/~hastie/glmnet/glmnet_alpha.html">Glmnet Vignette</a>
for a quick reference guide.</p>
<div id="cross-validation-analysis-setup" class="section level2" number="4.1">
<h2 number="4.1"><span class="header-section-number">4.1</span> Cross-validation analysis setup</h2>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a>K_FOLD &lt;-<span class="st"> </span><span class="dv">10</span></span>
<span id="cb34-2"><a href="#cb34-2"></a>trainP &lt;-<span class="st"> </span><span class="fl">0.8</span></span></code></pre></div>
<!-- NOT CURRENTLY USED 
EPS <- 0.05    # Have no idea what "small" epsilon means
-->
<p>First we divide the analysis dataset into <code>train</code> and <code>test</code> in a <span class="math inline">\(4\)</span>:1 ratio.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb35-2"><a href="#cb35-2"></a></span>
<span id="cb35-3"><a href="#cb35-3"></a><span class="kw">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb35-4"><a href="#cb35-4"></a>hcc5hmC_train_sampID_vec &lt;-<span class="st"> </span><span class="kw">with</span>(hcc5hmC_F_dgel<span class="op">$</span>samples,</span>
<span id="cb35-5"><a href="#cb35-5"></a>hcc5hmC_F_dgel<span class="op">$</span>samples<span class="op">$</span>sampID[caret<span class="op">::</span><span class="kw">createDataPartition</span>(<span class="dt">y=</span>group, <span class="dt">p=</span>trainP, <span class="dt">list=</span>F)]</span>
<span id="cb35-6"><a href="#cb35-6"></a>)</span>
<span id="cb35-7"><a href="#cb35-7"></a></span>
<span id="cb35-8"><a href="#cb35-8"></a>hcc5hmC_test_sampID_vec &lt;-<span class="st"> </span><span class="kw">with</span>(hcc5hmC_F_dgel<span class="op">$</span>samples,</span>
<span id="cb35-9"><a href="#cb35-9"></a><span class="kw">setdiff</span>(sampID, hcc5hmC_train_sampID_vec)</span>
<span id="cb35-10"><a href="#cb35-10"></a>)</span>
<span id="cb35-11"><a href="#cb35-11"></a></span>
<span id="cb35-12"><a href="#cb35-12"></a>hcc5hmC_train_group_vec &lt;-<span class="st"> </span>hcc5hmC_F_dgel<span class="op">$</span>samples[hcc5hmC_train_sampID_vec, <span class="st">&#39;group&#39;</span>]</span>
<span id="cb35-13"><a href="#cb35-13"></a><span class="kw">names</span>(hcc5hmC_train_group_vec) &lt;-<span class="st"> </span>hcc5hmC_F_dgel<span class="op">$</span>samples[hcc5hmC_train_sampID_vec, <span class="st">&#39;sampID&#39;</span>]</span>
<span id="cb35-14"><a href="#cb35-14"></a></span>
<span id="cb35-15"><a href="#cb35-15"></a>hcc5hmC_test_group_vec &lt;-<span class="st"> </span>hcc5hmC_F_dgel<span class="op">$</span>samples[hcc5hmC_test_sampID_vec, <span class="st">&#39;group&#39;</span>]</span>
<span id="cb35-16"><a href="#cb35-16"></a><span class="kw">names</span>(hcc5hmC_test_group_vec) &lt;-<span class="st"> </span>hcc5hmC_F_dgel<span class="op">$</span>samples[hcc5hmC_test_sampID_vec, <span class="st">&#39;sampID&#39;</span>]</span>
<span id="cb35-17"><a href="#cb35-17"></a></span>
<span id="cb35-18"><a href="#cb35-18"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">table</span>(hcc5hmC_train_group_vec),</span>
<span id="cb35-19"><a href="#cb35-19"></a>  <span class="dt">caption=</span><span class="st">&quot;Train set&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb35-20"><a href="#cb35-20"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-glmnetFit-getTrainVal)Train set
</caption>
<thead>
<tr>
<th style="text-align:left;">
hcc5hmC_train_group_vec
</th>
<th style="text-align:right;">
Freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Control
</td>
<td style="text-align:right;">
623
</td>
</tr>
<tr>
<td style="text-align:left;">
HCC
</td>
<td style="text-align:right;">
444
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">table</span>(hcc5hmC_test_group_vec),</span>
<span id="cb36-2"><a href="#cb36-2"></a>  <span class="dt">caption=</span><span class="st">&quot;Test set&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb36-3"><a href="#cb36-3"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-glmnetFit-getTrainVal)Test set
</caption>
<thead>
<tr>
<th style="text-align:left;">
hcc5hmC_test_group_vec
</th>
<th style="text-align:right;">
Freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Control
</td>
<td style="text-align:right;">
155
</td>
</tr>
<tr>
<td style="text-align:left;">
HCC
</td>
<td style="text-align:right;">
111
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a>hcc5hmC_train_lcpm_mtx &lt;-<span class="st"> </span><span class="kw">t</span>(hcc5hmC_lcpm_mtx[,hcc5hmC_train_sampID_vec])</span>
<span id="cb37-2"><a href="#cb37-2"></a>hcc5hmC_test_lcpm_mtx &lt;-<span class="st"> </span><span class="kw">t</span>(hcc5hmC_lcpm_mtx[,hcc5hmC_test_sampID_vec])</span></code></pre></div>
<p>We explore some glmnet fits and the “bet on sparsity”.
We consider three models, specified by the value of the
<strong>alpha</strong> parameter in the elastic net parametrization:<br />
- lasso: <span class="math inline">\(\alpha = 1.0\)</span> - sparse models<br />
- ridge <span class="math inline">\(\alpha = 0\)</span> - shrunken coefficients models<br />
- elastic net: <span class="math inline">\(\alpha = 0.5\)</span> - semi sparse model<br />
<!-- - lassoC: $\alpha = 1-\epsilon =$ $TICKr 1- EPSTICK - lasso for correlated predictors  --></p>
<p>Some questions of interest include:</p>
<ul>
<li>How sparse are models enabling good 5hmC classification of Early HCC vs Control samples?</li>
</ul>
<!--
    - The exploratory analysis done in the preprocessing step
indicates that differential gene body 5hmC representation effects are small.
With reasonably large sample sizes we would expect many genes to be selected.
???
-->
<!-- DROP THIS - having trouble getting at pure lassoR
* Does the relaxed lasso improve performance in this case?  
-->
<ul>
<li><p>Does the shrunken relaxed lasso (aka the blended mix) improve performance in this case?</p></li>
<li><p>Is the degree of sparsity, or the size of the model, a stable feature of the problem and data set?</p></li>
</ul>
<p>In this analysis, we will only evaluate models in terms of
model sparsity, stability and performance. We leave the question
of significance testing of hypotheses about model parameters
completely out. See Lockhart et al. (2014) <span class="citation">[<a href="#ref-Lockhart:2014aa" role="doc-biblioref">24</a>]</span>
and Wassermam (2014) <span class="citation">[<a href="#ref-Wasserman:2014aa" role="doc-biblioref">25</a>]</span> for a discussion of this topic.</p>
<p>In this section we look at the relative performance and sparsity of the models
considered. The effect of the size of the sample set on the level and
stability of performance will be investigated in the next section.</p>
<hr />
<p>First we create folds for <span class="math inline">\(10\)</span>-fold cross-validation of models fitted to
training data. We’ll use caret::createFolds to assign samples
to folds while keeping the outcome ratios constant across folds.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb38-2"><a href="#cb38-2"></a></span>
<span id="cb38-3"><a href="#cb38-3"></a><span class="co"># This is too variable, both in terms of fold size And composition</span></span>
<span id="cb38-4"><a href="#cb38-4"></a><span class="co">#foldid_vec &lt;- sample(1:10, size=length(hcc5hmC_train_group_vec), replace=T)</span></span>
<span id="cb38-5"><a href="#cb38-5"></a></span>
<span id="cb38-6"><a href="#cb38-6"></a><span class="kw">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb38-7"><a href="#cb38-7"></a>hcc5hmC_train_foldid_vec &lt;-<span class="st"> </span>caret<span class="op">::</span><span class="kw">createFolds</span>(</span>
<span id="cb38-8"><a href="#cb38-8"></a> <span class="kw">factor</span>(hcc5hmC_train_group_vec), </span>
<span id="cb38-9"><a href="#cb38-9"></a> <span class="dt">k=</span>K_FOLD,</span>
<span id="cb38-10"><a href="#cb38-10"></a> <span class="dt">list=</span>F)</span>
<span id="cb38-11"><a href="#cb38-11"></a></span>
<span id="cb38-12"><a href="#cb38-12"></a><span class="co"># hcc5hmC_train_foldid_vec contains the left-out IDs </span></span>
<span id="cb38-13"><a href="#cb38-13"></a><span class="co"># the rest are kept</span></span>
<span id="cb38-14"><a href="#cb38-14"></a>fold_out_tbl &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(hcc5hmC_train_group_vec, hcc5hmC_train_foldid_vec),</span>
<span id="cb38-15"><a href="#cb38-15"></a>  table)</span>
<span id="cb38-16"><a href="#cb38-16"></a><span class="kw">rownames</span>(fold_out_tbl) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">rownames</span>(fold_out_tbl), <span class="st">&#39;- Out&#39;</span>) </span>
<span id="cb38-17"><a href="#cb38-17"></a></span>
<span id="cb38-18"><a href="#cb38-18"></a>fold_in_tbl &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="st">&#39;cbind&#39;</span>, <span class="kw">lapply</span>(<span class="kw">sort</span>(<span class="kw">unique</span>(hcc5hmC_train_foldid_vec)),</span>
<span id="cb38-19"><a href="#cb38-19"></a>  <span class="cf">function</span>(FOLD) <span class="kw">table</span>(hcc5hmC_train_group_vec[hcc5hmC_train_foldid_vec <span class="op">!=</span><span class="st"> </span>FOLD])))</span>
<span id="cb38-20"><a href="#cb38-20"></a><span class="kw">rownames</span>(fold_in_tbl) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">rownames</span>(fold_in_tbl), <span class="st">&#39;- In&#39;</span>) </span>
<span id="cb38-21"><a href="#cb38-21"></a><span class="kw">colnames</span>(fold_in_tbl) &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">sort</span>(<span class="kw">unique</span>(hcc5hmC_train_foldid_vec)))</span>
<span id="cb38-22"><a href="#cb38-22"></a></span>
<span id="cb38-23"><a href="#cb38-23"></a></span>
<span id="cb38-24"><a href="#cb38-24"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">rbind</span>(fold_in_tbl, fold_out_tbl[,<span class="kw">colnames</span>(fold_in_tbl)]),</span>
<span id="cb38-25"><a href="#cb38-25"></a>  <span class="dt">caption=</span><span class="st">&quot;training samples fold composition&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb38-26"><a href="#cb38-26"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-glmnetFit-getTrainFolds)training samples fold composition
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
1
</th>
<th style="text-align:right;">
2
</th>
<th style="text-align:right;">
3
</th>
<th style="text-align:right;">
4
</th>
<th style="text-align:right;">
5
</th>
<th style="text-align:right;">
6
</th>
<th style="text-align:right;">
7
</th>
<th style="text-align:right;">
8
</th>
<th style="text-align:right;">
9
</th>
<th style="text-align:right;">
10
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Control - In
</td>
<td style="text-align:right;">
561
</td>
<td style="text-align:right;">
561
</td>
<td style="text-align:right;">
561
</td>
<td style="text-align:right;">
560
</td>
<td style="text-align:right;">
561
</td>
<td style="text-align:right;">
561
</td>
<td style="text-align:right;">
560
</td>
<td style="text-align:right;">
561
</td>
<td style="text-align:right;">
560
</td>
<td style="text-align:right;">
561
</td>
</tr>
<tr>
<td style="text-align:left;">
HCC - In
</td>
<td style="text-align:right;">
399
</td>
<td style="text-align:right;">
400
</td>
<td style="text-align:right;">
400
</td>
<td style="text-align:right;">
399
</td>
<td style="text-align:right;">
400
</td>
<td style="text-align:right;">
399
</td>
<td style="text-align:right;">
400
</td>
<td style="text-align:right;">
399
</td>
<td style="text-align:right;">
400
</td>
<td style="text-align:right;">
400
</td>
</tr>
<tr>
<td style="text-align:left;">
Control - Out
</td>
<td style="text-align:right;">
62
</td>
<td style="text-align:right;">
62
</td>
<td style="text-align:right;">
62
</td>
<td style="text-align:right;">
63
</td>
<td style="text-align:right;">
62
</td>
<td style="text-align:right;">
62
</td>
<td style="text-align:right;">
63
</td>
<td style="text-align:right;">
62
</td>
<td style="text-align:right;">
63
</td>
<td style="text-align:right;">
62
</td>
</tr>
<tr>
<td style="text-align:left;">
HCC - Out
</td>
<td style="text-align:right;">
45
</td>
<td style="text-align:right;">
44
</td>
<td style="text-align:right;">
44
</td>
<td style="text-align:right;">
45
</td>
<td style="text-align:right;">
44
</td>
<td style="text-align:right;">
45
</td>
<td style="text-align:right;">
44
</td>
<td style="text-align:right;">
45
</td>
<td style="text-align:right;">
44
</td>
<td style="text-align:right;">
44
</td>
</tr>
</tbody>
</table>
<p>Note that the folds identify samples that are left-out of the training
data for each fold fit.</p>
</div>
<div id="fit-and-compare-models" class="section level2" number="4.2">
<h2 number="4.2"><span class="header-section-number">4.2</span> Fit and compare models</h2>
<p><code>glmnet</code> provides cross-validation methods to pick the parameter <strong>lambda</strong> which
controls to size of the penalty function. The “one standard error rule”
produces a model with fewer predictors then the minimum cv error model.
On the training data, this usually results in increased MSE and more
biased parameter estimates
(see Engebretsen et al. (2019) <span class="citation">[<a href="#ref-Engebretsen:2019aa" role="doc-biblioref">26</a>]</span> for example).
The question of interest though is the performance on unseen data; not on the training data.
The cv error rates computed and stored in fit results by <code>glmnet</code> methods are
out-of-fold error rates computed from
the training data. The results show that these are good indicators of
test set error rates, and that
the one standard error rule models do as well as the minimum cv error models
for the lasso, which has the best overall performance.</p>
<div id="logistic-regression-in-glmnet" class="section level3" number="4.2.1">
<h3 number="4.2.1"><span class="header-section-number">4.2.1</span> Logistic regression in <code>glmnet</code></h3>
<p>It is sometimes necessary to compute cv error rates manually from
<code>glmnet</code> fit results. This section provides some necessary detail.</p>
<p><code>glmnet</code> provides functionality to extract various predicted of fitted values
from calibrated models. Note that some folks make a distinction between
<strong>fitted</strong> or <strong>estimated</strong> values for sample points in the training data
versus <strong>predicted</strong> values for sample points that
are not in the training dataset. <code>glmnet</code> makes no such distinction and the
<code>predict</code> function is used to produce both fitted as well as predicted values.
When predict is invoked to make predictions for design points that are part
of the training dataset, what is returned are fitted values.
When predict is invoked to make predictions for design points that are not part
of the training dataset, what is returned are predicted values.</p>
<p>For logistic regressions, which is the model fitted in a regularized fashion
when models are fitted by glmnet with the parameter <code>family='binomial'</code>, three
fitted or predicted values can be extracted at a given design point.
Suppose our response variable Y is either 0 or 1 (Control or HCC in our case).
These are specified by the <code>type</code> parameter. <code>type='resp'</code> returns
the fitted or predicted probability of <span class="math inline">\(Y=1\)</span>. <code>type='class'</code> returns the fitted or
predicted class for the design point, which is simply dichotomizing the
response: class = 1 if the fitted or predicted probability is greater than 0.5
(check to make sure class is no the Bayes estimate). <code>type='link'</code> returns
the fitted or predicted value of the linear predictor <span class="math inline">\(\beta&#39;x\)</span>. The relationship
between the linear predictor and the response can be derided from the
logistic regression model:</p>
<p><span class="math display">\[P(Y=1|x,\beta) = g^{-1}(\beta&#39;x) = h(\beta&#39;x) = \frac{e^{\beta&#39;x}}{1+e^{\beta&#39;x}}\]</span></p>
<p>where <span class="math inline">\(g\)</span> is the link function, <span class="math inline">\(g^{-1}\)</span> the mean function.
The link function is given by:</p>
<p><span class="math display">\[g(y) = h^{-1}(y) = ln(\frac{y}{1-y})\]</span></p>
<p>This link function is called the <em>logit</em> function, and its inverse the <em>logistic</em>
function.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a>logistic_f &lt;-<span class="st"> </span><span class="cf">function</span>(x) <span class="kw">ifelse</span>(<span class="kw">is.nan</span>(<span class="kw">exp</span>(x)<span class="op">/</span>(<span class="dv">1</span><span class="op">+</span><span class="kw">exp</span>(x))), <span class="dv">1</span>, <span class="kw">exp</span>(x)<span class="op">/</span>(<span class="dv">1</span><span class="op">+</span><span class="kw">exp</span>(x)))</span></code></pre></div>
<p>We should also point out that the cv error rates quoted in various <code>glmnet</code> summaries
are computed from out-of-fold predictions. In other words,
we cannot recover the <code>cvm</code> component of a glmnet fit object by comparing
class predictions extracted by invoking <code>predict()</code> with parameters <code>newx = train_data</code>
and <code>type = 'class'</code> to the true class labels.</p>
<p><code>glmnet</code> fitting functions have a
parameter, <em>keep</em>, which instructs the fitting function to keep the
<code>out-of-fold</code>, or <code>prevalidated</code>, predictions as part of the returned object. The
<code>out-of-fold</code> predictions are predicted values for the samples in the
left-out folds, pooled across all cv folds. For each hyper-parameter
specification, we get one full set of <code>out-of-fold</code> predictions for
the training set samples. Performance assessments based on these
values are usually more generalizable. The <code>cvm</code> component of glmnet fitted
objects are derived from these <code>out-of-fold</code> predictions.
See Hofling and Tibshirani (2008) <span class="citation">[<a href="#ref-Hofling:2008aa" role="doc-biblioref">27</a>]</span>
for a description of the use of pre-validation in model assessment.</p>
<!--
Because the `keep=T` option will store predicted values for 
all models evaluated in the cross-validation process, we will
limit the number of models tested by setting **nlambda=30**
when calling the fitting functions.  This has no effect on 
performance in this data set.
-->
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1"></a><span class="co">### CLEAR CACHE </span></span>
<span id="cb40-2"><a href="#cb40-2"></a></span>
<span id="cb40-3"><a href="#cb40-3"></a>start_time &lt;-<span class="st">  </span><span class="kw">proc.time</span>()</span>
<span id="cb40-4"><a href="#cb40-4"></a></span>
<span id="cb40-5"><a href="#cb40-5"></a>hcc5hmC_cv_lasso &lt;-<span class="st"> </span>glmnet<span class="op">::</span><span class="kw">cv.glmnet</span>(</span>
<span id="cb40-6"><a href="#cb40-6"></a> <span class="dt">x=</span>hcc5hmC_train_lcpm_mtx,</span>
<span id="cb40-7"><a href="#cb40-7"></a> <span class="dt">y=</span>hcc5hmC_train_group_vec,</span>
<span id="cb40-8"><a href="#cb40-8"></a> <span class="dt">foldid=</span>hcc5hmC_train_foldid_vec,</span>
<span id="cb40-9"><a href="#cb40-9"></a> <span class="dt">alpha=</span><span class="dv">1</span>,</span>
<span id="cb40-10"><a href="#cb40-10"></a> <span class="dt">family=</span><span class="st">&#39;binomial&#39;</span>, </span>
<span id="cb40-11"><a href="#cb40-11"></a> <span class="dt">type.measure =</span> <span class="st">&quot;class&quot;</span>,</span>
<span id="cb40-12"><a href="#cb40-12"></a> <span class="dt">keep=</span>T,</span>
<span id="cb40-13"><a href="#cb40-13"></a> <span class="dt">nlambda=</span><span class="dv">30</span></span>
<span id="cb40-14"><a href="#cb40-14"></a>)</span>
<span id="cb40-15"><a href="#cb40-15"></a></span>
<span id="cb40-16"><a href="#cb40-16"></a><span class="kw">message</span>(<span class="st">&quot;lasso time: &quot;</span>, <span class="kw">round</span>((<span class="kw">proc.time</span>() <span class="op">-</span><span class="st"> </span>start_time)[<span class="dv">3</span>],<span class="dv">2</span>),<span class="st">&quot;s&quot;</span>)</span></code></pre></div>
<pre><code>## lasso time: 13.94s</code></pre>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1"></a>start_time &lt;-<span class="st">  </span><span class="kw">proc.time</span>()</span>
<span id="cb42-2"><a href="#cb42-2"></a></span>
<span id="cb42-3"><a href="#cb42-3"></a>hcc5hmC_cv_ridge &lt;-<span class="st"> </span>glmnet<span class="op">::</span><span class="kw">cv.glmnet</span>(</span>
<span id="cb42-4"><a href="#cb42-4"></a> <span class="dt">x=</span>hcc5hmC_train_lcpm_mtx,</span>
<span id="cb42-5"><a href="#cb42-5"></a> <span class="dt">y=</span>hcc5hmC_train_group_vec,</span>
<span id="cb42-6"><a href="#cb42-6"></a> <span class="dt">foldid=</span>hcc5hmC_train_foldid_vec,</span>
<span id="cb42-7"><a href="#cb42-7"></a> <span class="dt">alpha=</span><span class="dv">0</span>,</span>
<span id="cb42-8"><a href="#cb42-8"></a> <span class="dt">family=</span><span class="st">&#39;binomial&#39;</span>, </span>
<span id="cb42-9"><a href="#cb42-9"></a> <span class="dt">type.measure =</span> <span class="st">&quot;class&quot;</span>,</span>
<span id="cb42-10"><a href="#cb42-10"></a> <span class="dt">keep=</span>T,</span>
<span id="cb42-11"><a href="#cb42-11"></a> <span class="dt">nlambda=</span><span class="dv">30</span></span>
<span id="cb42-12"><a href="#cb42-12"></a>)</span>
<span id="cb42-13"><a href="#cb42-13"></a></span>
<span id="cb42-14"><a href="#cb42-14"></a><span class="kw">message</span>(<span class="st">&quot;ridge time: &quot;</span>, <span class="kw">round</span>((<span class="kw">proc.time</span>() <span class="op">-</span><span class="st"> </span>start_time)[<span class="dv">3</span>],<span class="dv">2</span>),<span class="st">&quot;s&quot;</span>)</span></code></pre></div>
<pre><code>## ridge time: 131.08s</code></pre>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1"></a><span class="co">### CLEAR CACHE </span></span>
<span id="cb44-2"><a href="#cb44-2"></a>start_time &lt;-<span class="st">  </span><span class="kw">proc.time</span>()</span>
<span id="cb44-3"><a href="#cb44-3"></a></span>
<span id="cb44-4"><a href="#cb44-4"></a>hcc5hmC_cv_enet &lt;-<span class="st"> </span>glmnet<span class="op">::</span><span class="kw">cv.glmnet</span>(</span>
<span id="cb44-5"><a href="#cb44-5"></a> <span class="dt">x=</span>hcc5hmC_train_lcpm_mtx,</span>
<span id="cb44-6"><a href="#cb44-6"></a> <span class="dt">y=</span>hcc5hmC_train_group_vec,</span>
<span id="cb44-7"><a href="#cb44-7"></a> <span class="dt">foldid=</span>hcc5hmC_train_foldid_vec,</span>
<span id="cb44-8"><a href="#cb44-8"></a> <span class="dt">alpha=</span><span class="fl">0.5</span>,</span>
<span id="cb44-9"><a href="#cb44-9"></a> <span class="dt">family=</span><span class="st">&#39;binomial&#39;</span>,</span>
<span id="cb44-10"><a href="#cb44-10"></a> <span class="dt">type.measure =</span> <span class="st">&quot;class&quot;</span>,</span>
<span id="cb44-11"><a href="#cb44-11"></a> <span class="dt">keep=</span>T,</span>
<span id="cb44-12"><a href="#cb44-12"></a> <span class="dt">nlambda=</span><span class="dv">30</span></span>
<span id="cb44-13"><a href="#cb44-13"></a>)</span>
<span id="cb44-14"><a href="#cb44-14"></a></span>
<span id="cb44-15"><a href="#cb44-15"></a><span class="kw">message</span>(<span class="st">&quot;enet time: &quot;</span>, <span class="kw">round</span>((<span class="kw">proc.time</span>() <span class="op">-</span><span class="st"> </span>start_time)[<span class="dv">3</span>],<span class="dv">2</span>),<span class="st">&quot;s&quot;</span>)</span></code></pre></div>
<pre><code>## enet time: 15.71s</code></pre>
<!--
The ridge regression model takes over 10 times longer to compute.
-->
<!-- do not show
Define plotting function.
Maybe show in appendix??
-->
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1"></a>plot_cv_f &lt;-<span class="st"> </span><span class="cf">function</span>(cv_fit, <span class="dt">Nzero=</span>T, ...) {</span>
<span id="cb46-2"><a href="#cb46-2"></a> </span>
<span id="cb46-3"><a href="#cb46-3"></a> <span class="kw">suppressPackageStartupMessages</span>(<span class="kw">require</span>(glmnet))</span>
<span id="cb46-4"><a href="#cb46-4"></a></span>
<span id="cb46-5"><a href="#cb46-5"></a> <span class="co"># No nonger used</span></span>
<span id="cb46-6"><a href="#cb46-6"></a> <span class="co">#lambda.1se_p &lt;- cv_fit$nzero[cv_fit$lambda == cv_fit$lambda.1se]</span></span>
<span id="cb46-7"><a href="#cb46-7"></a> <span class="co">#lambda.min_p &lt;- cv_fit$nzero[cv_fit$lambda == cv_fit$lambda.min]</span></span>
<span id="cb46-8"><a href="#cb46-8"></a> </span>
<span id="cb46-9"><a href="#cb46-9"></a> <span class="co"># Get oof error - cv errors produced by extraction method ARE oof!!!</span></span>
<span id="cb46-10"><a href="#cb46-10"></a> ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se,cv_fit<span class="op">$</span>lambda)</span>
<span id="cb46-11"><a href="#cb46-11"></a> train_oofPred_1se_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb46-12"><a href="#cb46-12"></a>  <span class="kw">logistic_f</span>(cv_fit<span class="op">$</span>fit.preval[,ndx_1se]) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&#39;HCC&#39;</span>, <span class="st">&#39;Control&#39;</span>)</span>
<span id="cb46-13"><a href="#cb46-13"></a> train_oofPred_1se_error &lt;-<span class="st"> </span><span class="kw">mean</span>(train_oofPred_1se_vec <span class="op">!=</span><span class="st"> </span>hcc5hmC_train_group_vec)</span>
<span id="cb46-14"><a href="#cb46-14"></a></span>
<span id="cb46-15"><a href="#cb46-15"></a> ndx_min &lt;-<span class="st"> </span><span class="kw">match</span>(cv_fit<span class="op">$</span>lambda.min,cv_fit<span class="op">$</span>lambda)</span>
<span id="cb46-16"><a href="#cb46-16"></a> train_oofPred_min_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb46-17"><a href="#cb46-17"></a>  <span class="kw">logistic_f</span>(cv_fit<span class="op">$</span>fit.preval[,ndx_min]) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&#39;HCC&#39;</span>, <span class="st">&#39;Control&#39;</span>)</span>
<span id="cb46-18"><a href="#cb46-18"></a> train_oofPred_min_error &lt;-<span class="st"> </span><span class="kw">mean</span>(train_oofPred_min_vec <span class="op">!=</span><span class="st"> </span>hcc5hmC_train_group_vec)</span>
<span id="cb46-19"><a href="#cb46-19"></a></span>
<span id="cb46-20"><a href="#cb46-20"></a> <span class="co"># Get test set error</span></span>
<span id="cb46-21"><a href="#cb46-21"></a> test_pred_1se_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb46-22"><a href="#cb46-22"></a>  cv_fit, </span>
<span id="cb46-23"><a href="#cb46-23"></a>  <span class="dt">newx=</span>hcc5hmC_test_lcpm_mtx, </span>
<span id="cb46-24"><a href="#cb46-24"></a>  <span class="dt">s=</span><span class="st">&quot;lambda.1se&quot;</span>,</span>
<span id="cb46-25"><a href="#cb46-25"></a>  <span class="dt">type=</span><span class="st">&quot;class&quot;</span></span>
<span id="cb46-26"><a href="#cb46-26"></a> )</span>
<span id="cb46-27"><a href="#cb46-27"></a> test_pred_1se_error &lt;-<span class="st"> </span><span class="kw">mean</span>(test_pred_1se_vec <span class="op">!=</span><span class="st"> </span>hcc5hmC_test_group_vec)</span>
<span id="cb46-28"><a href="#cb46-28"></a> </span>
<span id="cb46-29"><a href="#cb46-29"></a> test_pred_min_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb46-30"><a href="#cb46-30"></a>  cv_fit, </span>
<span id="cb46-31"><a href="#cb46-31"></a>  <span class="dt">newx=</span>hcc5hmC_test_lcpm_mtx, </span>
<span id="cb46-32"><a href="#cb46-32"></a>  <span class="dt">s=</span><span class="st">&quot;lambda.min&quot;</span>,</span>
<span id="cb46-33"><a href="#cb46-33"></a>  <span class="dt">type=</span><span class="st">&quot;class&quot;</span></span>
<span id="cb46-34"><a href="#cb46-34"></a> )</span>
<span id="cb46-35"><a href="#cb46-35"></a> test_pred_min_error &lt;-<span class="st"> </span><span class="kw">mean</span>(test_pred_min_vec <span class="op">!=</span><span class="st"> </span>hcc5hmC_test_group_vec)</span>
<span id="cb46-36"><a href="#cb46-36"></a> </span>
<span id="cb46-37"><a href="#cb46-37"></a>  </span>
<span id="cb46-38"><a href="#cb46-38"></a> <span class="kw">plot</span>(</span>
<span id="cb46-39"><a href="#cb46-39"></a>  <span class="kw">log</span>(cv_fit<span class="op">$</span>lambda),</span>
<span id="cb46-40"><a href="#cb46-40"></a>  cv_fit<span class="op">$</span>cvm,</span>
<span id="cb46-41"><a href="#cb46-41"></a>  <span class="dt">pch=</span><span class="dv">16</span>,<span class="dt">col=</span><span class="st">&quot;red&quot;</span>,</span>
<span id="cb46-42"><a href="#cb46-42"></a>  <span class="dt">xlab=</span><span class="st">&#39;&#39;</span>,<span class="dt">ylab=</span><span class="st">&#39;&#39;</span>,</span>
<span id="cb46-43"><a href="#cb46-43"></a>  ...</span>
<span id="cb46-44"><a href="#cb46-44"></a> )</span>
<span id="cb46-45"><a href="#cb46-45"></a> <span class="kw">abline</span>(<span class="dt">v=</span><span class="kw">log</span>(<span class="kw">c</span>(cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se, cv_fit<span class="op">$</span>lambda.min)))</span>
<span id="cb46-46"><a href="#cb46-46"></a> <span class="cf">if</span>(Nzero)</span>
<span id="cb46-47"><a href="#cb46-47"></a> <span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">tick=</span>F, <span class="dt">at=</span><span class="kw">log</span>(cv_fit<span class="op">$</span>lambda), </span>
<span id="cb46-48"><a href="#cb46-48"></a>  <span class="dt">labels=</span>cv_fit<span class="op">$</span>nzero, <span class="dt">line =</span> <span class="dv">-1</span></span>
<span id="cb46-49"><a href="#cb46-49"></a> )</span>
<span id="cb46-50"><a href="#cb46-50"></a> LL &lt;-<span class="st"> </span><span class="dv">2</span></span>
<span id="cb46-51"><a href="#cb46-51"></a> <span class="co">#mtext(side=1, outer=F, line = LL, &quot;log(Lambda)&quot;)</span></span>
<span id="cb46-52"><a href="#cb46-52"></a> <span class="co">#LL &lt;- LL+1</span></span>
<span id="cb46-53"><a href="#cb46-53"></a> <span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">outer=</span>F, <span class="dt">line =</span> LL, <span class="kw">paste</span>(</span>
<span id="cb46-54"><a href="#cb46-54"></a>  <span class="co">#ifelse(Nzero, paste(&quot;1se p =&quot;, lambda.1se_p),&#39;&#39;),</span></span>
<span id="cb46-55"><a href="#cb46-55"></a>  <span class="st">&quot;1se: train =&quot;</span>, <span class="kw">round</span>(<span class="dv">100</span><span class="op">*</span>cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se], <span class="dv">1</span>),</span>
<span id="cb46-56"><a href="#cb46-56"></a>  <span class="co">##&quot;oof =&quot;, round(100*train_oofPred_1se_error, 1), </span><span class="al">###</span><span class="co"> REDUNDANT</span></span>
<span id="cb46-57"><a href="#cb46-57"></a>  <span class="st">&quot;test =&quot;</span>, <span class="kw">round</span>(<span class="dv">100</span><span class="op">*</span>test_pred_1se_error, <span class="dv">1</span>)</span>
<span id="cb46-58"><a href="#cb46-58"></a> ))</span>
<span id="cb46-59"><a href="#cb46-59"></a> LL &lt;-<span class="st"> </span>LL<span class="op">+</span><span class="dv">1</span></span>
<span id="cb46-60"><a href="#cb46-60"></a> <span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">outer=</span>F, <span class="dt">line =</span> LL, <span class="kw">paste</span>(</span>
<span id="cb46-61"><a href="#cb46-61"></a>  <span class="co">#ifelse(Nzero, paste(&quot;min p =&quot;, lambda.min_p),&#39;&#39;),</span></span>
<span id="cb46-62"><a href="#cb46-62"></a>  <span class="st">&quot;min: train =&quot;</span>, <span class="kw">round</span>(<span class="dv">100</span><span class="op">*</span>cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min], <span class="dv">1</span>),</span>
<span id="cb46-63"><a href="#cb46-63"></a>  <span class="co">##&quot;oof =&quot;, round(100*train_oofPred_min_error, 1),  </span><span class="al">###</span><span class="co"> REDUNDANT</span></span>
<span id="cb46-64"><a href="#cb46-64"></a>  <span class="st">&quot;test =&quot;</span>, <span class="kw">round</span>(<span class="dv">100</span><span class="op">*</span>test_pred_min_error, <span class="dv">1</span>)</span>
<span id="cb46-65"><a href="#cb46-65"></a> ))</span>
<span id="cb46-66"><a href="#cb46-66"></a> </span>
<span id="cb46-67"><a href="#cb46-67"></a> tmp &lt;-</span>
<span id="cb46-68"><a href="#cb46-68"></a><span class="st"> </span><span class="kw">cbind</span>(</span>
<span id="cb46-69"><a href="#cb46-69"></a>  <span class="dt">error_1se =</span> <span class="kw">c</span>(</span>
<span id="cb46-70"><a href="#cb46-70"></a>   <span class="dt">p =</span> cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se],</span>
<span id="cb46-71"><a href="#cb46-71"></a>   <span class="dt">train  =</span> <span class="dv">100</span><span class="op">*</span>cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se],</span>
<span id="cb46-72"><a href="#cb46-72"></a>   <span class="co">#train_oof = 100*train_oofPred_1se_error,  </span><span class="al">###</span><span class="co"> REDUNANT</span></span>
<span id="cb46-73"><a href="#cb46-73"></a>   <span class="dt">test =</span> <span class="dv">100</span><span class="op">*</span>test_pred_1se_error),</span>
<span id="cb46-74"><a href="#cb46-74"></a>  <span class="dt">error_min =</span> <span class="kw">c</span>(</span>
<span id="cb46-75"><a href="#cb46-75"></a>   <span class="dt">p =</span> cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min],</span>
<span id="cb46-76"><a href="#cb46-76"></a>   <span class="dt">train  =</span> <span class="dv">100</span><span class="op">*</span>cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min],</span>
<span id="cb46-77"><a href="#cb46-77"></a>   <span class="co">#train_oof = 100*train_oofPred_min_error, </span><span class="al">###</span><span class="co"> REDUNDSANT</span></span>
<span id="cb46-78"><a href="#cb46-78"></a>   <span class="dt">test =</span> <span class="dv">100</span><span class="op">*</span>test_pred_min_error)</span>
<span id="cb46-79"><a href="#cb46-79"></a>  )</span>
<span id="cb46-80"><a href="#cb46-80"></a>  <span class="co"># Need to fix names  </span></span>
<span id="cb46-81"><a href="#cb46-81"></a>  <span class="kw">rownames</span>(tmp) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;p&#39;</span>, <span class="st">&#39;train&#39;</span>, <span class="st">&#39;test&#39;</span>)</span>
<span id="cb46-82"><a href="#cb46-82"></a>  tmp </span>
<span id="cb46-83"><a href="#cb46-83"></a>}</span></code></pre></div>
<p>Examine model performance.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1"></a> <span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">5</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">0</span>)) </span>
<span id="cb47-2"><a href="#cb47-2"></a></span>
<span id="cb47-3"><a href="#cb47-3"></a> lasso_errors_mtx &lt;-<span class="st"> </span><span class="kw">plot_cv_f</span>(hcc5hmC_cv_lasso, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,.<span class="dv">5</span>))</span></code></pre></div>
<pre><code>## Warning: package &#39;glmnet&#39; was built under R version 4.0.2</code></pre>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1"></a> <span class="kw">title</span>(<span class="st">&#39;lasso&#39;</span>)</span>
<span id="cb49-2"><a href="#cb49-2"></a></span>
<span id="cb49-3"><a href="#cb49-3"></a> rifge_errors_mtx &lt;-<span class="st"> </span><span class="kw">plot_cv_f</span>(hcc5hmC_cv_ridge, <span class="dt">Nzero=</span>F, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,.<span class="dv">5</span>))</span>
<span id="cb49-4"><a href="#cb49-4"></a> <span class="kw">title</span>(<span class="st">&#39;ridge&#39;</span>)</span>
<span id="cb49-5"><a href="#cb49-5"></a></span>
<span id="cb49-6"><a href="#cb49-6"></a> enet_errors_mtx &lt;-<span class="st">  </span><span class="kw">plot_cv_f</span>(hcc5hmC_cv_enet, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,.<span class="dv">5</span>))</span>
<span id="cb49-7"><a href="#cb49-7"></a> <span class="kw">title</span>(<span class="st">&#39;enet&#39;</span>)</span>
<span id="cb49-8"><a href="#cb49-8"></a></span>
<span id="cb49-9"><a href="#cb49-9"></a> <span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, <span class="st">&#39;log(Lambda)&#39;</span>)</span>
<span id="cb49-10"><a href="#cb49-10"></a> <span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, hcc5hmC_cv_lasso<span class="op">$</span>name)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetFit-lookFits-1.png" alt="compare fits" width="1056" />
<p class="caption">
(#fig:hcc5hmC-glmnetFit-lookFits)compare fits
</p>
</div>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1"></a>errors_frm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb50-2"><a href="#cb50-2"></a>  <span class="dt">lasso =</span> lasso_errors_mtx, <span class="dt">ridge =</span> rifge_errors_mtx, <span class="dt">enet =</span> enet_errors_mtx</span>
<span id="cb50-3"><a href="#cb50-3"></a>)</span>
<span id="cb50-4"><a href="#cb50-4"></a><span class="kw">colnames</span>(errors_frm) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&#39;</span><span class="ch">\\</span><span class="st">.error&#39;</span>,<span class="st">&#39;&#39;</span>, <span class="kw">colnames</span>(errors_frm))</span>
<span id="cb50-5"><a href="#cb50-5"></a></span>
<span id="cb50-6"><a href="#cb50-6"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">t</span>(errors_frm),</span>
<span id="cb50-7"><a href="#cb50-7"></a> <span class="dt">caption =</span> <span class="st">&#39;Misclassifiaction error rates&#39;</span>,</span>
<span id="cb50-8"><a href="#cb50-8"></a> <span class="dt">digits=</span><span class="dv">1</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb50-9"><a href="#cb50-9"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-glmnetFit-printErrors)Misclassifiaction error rates
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
p
</th>
<th style="text-align:right;">
train
</th>
<th style="text-align:right;">
test
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
lasso_1se
</td>
<td style="text-align:right;">
68
</td>
<td style="text-align:right;">
7.9
</td>
<td style="text-align:right;">
10.9
</td>
</tr>
<tr>
<td style="text-align:left;">
lasso_min
</td>
<td style="text-align:right;">
226
</td>
<td style="text-align:right;">
7.0
</td>
<td style="text-align:right;">
10.2
</td>
</tr>
<tr>
<td style="text-align:left;">
ridge_1se
</td>
<td style="text-align:right;">
19100
</td>
<td style="text-align:right;">
15.0
</td>
<td style="text-align:right;">
18.8
</td>
</tr>
<tr>
<td style="text-align:left;">
ridge_min
</td>
<td style="text-align:right;">
19100
</td>
<td style="text-align:right;">
13.8
</td>
<td style="text-align:right;">
17.3
</td>
</tr>
<tr>
<td style="text-align:left;">
enet_1se
</td>
<td style="text-align:right;">
160
</td>
<td style="text-align:right;">
7.1
</td>
<td style="text-align:right;">
11.3
</td>
</tr>
<tr>
<td style="text-align:left;">
enet_min
</td>
<td style="text-align:right;">
284
</td>
<td style="text-align:right;">
6.6
</td>
<td style="text-align:right;">
9.4
</td>
</tr>
</tbody>
</table>
<p><br/></p>
<p>We see that the lasso and enet models do better than the ridge model.
The <em>one standard error rule</em> (1se) lambda lasso fit is only slightly more parsimonious than the
<em>1se</em> elastic net fit, but its test set accuracy is better.
The <em>min</em> lambda models have lower test set error rates than
the more parsimonious <em>1se</em> models in the ridge and elastic net case.
(for the lasso, 1se and minimum lambda rules give the same model).
The minimum lambda elastic net fit performs as well as the lasso model,
but is much less parsimonious.</p>
<p>Note that the cv estimates of misclassification error produced by
the glmnet extraction method are <code>out-of-fold</code> estimated error rates.
In this particular dataset, these estimates of error are
optimistic in comparison the test set error rates.</p>
</div>
</div>
<div id="the-relaxed-lasso-and-blended-mix-models" class="section level2" number="4.3">
<h2 number="4.3"><span class="header-section-number">4.3</span> The relaxed lasso and blended mix models</h2>
<p>Next we look at the so-called <code>relaxed lasso</code> model, and
the <code>blended mix</code> which is an optimized shrinkage
between the relaxed lasso and the regular lasso.
See @ref(eq:blended) in Section @ref(modeling-background).</p>
<!--
The relaxed fit takes quite a bit longer.  
-->
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1"></a><span class="kw">library</span>(glmnet)</span>
<span id="cb51-2"><a href="#cb51-2"></a></span>
<span id="cb51-3"><a href="#cb51-3"></a>hcc5hmC_cv_lassoR_sum &lt;-<span class="st"> </span><span class="kw">print</span>(hcc5hmC_cv_lassoR)</span></code></pre></div>
<pre><code>## 
## Call:  glmnet::cv.glmnet(x = hcc5hmC_train_lcpm_mtx, y = hcc5hmC_train_group_vec,      type.measure = &quot;class&quot;, foldid = hcc5hmC_train_foldid_vec,      keep = T, relax = T, alpha = 1, family = &quot;binomial&quot;, nlambda = 30) 
## 
## Measure: Misclassification Error 
## 
##     Gamma  Lambda Measure       SE Nonzero
## min   0.5 0.03790 0.06467 0.004819      38
## 1se   0.5 0.04442 0.06560 0.005303      33</code></pre>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1"></a><span class="kw">plot</span>(hcc5hmC_cv_lassoR)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetFit-lookLassoR-1.png" alt="Relaxed lasso fit" width="576" />
<p class="caption">
(#fig:hcc5hmC-glmnetFit-lookLassoR)Relaxed lasso fit
</p>
</div>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1"></a><span class="co"># only report  1se</span></span>
<span id="cb54-2"><a href="#cb54-2"></a>ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(hcc5hmC_cv_lassoR<span class="op">$</span>lambda<span class="fl">.1</span>se, hcc5hmC_cv_lassoR<span class="op">$</span>lambda)</span>
<span id="cb54-3"><a href="#cb54-3"></a>ndx_min &lt;-<span class="st"> </span><span class="kw">match</span>(hcc5hmC_cv_lassoR<span class="op">$</span>lambda.min, hcc5hmC_cv_lassoR<span class="op">$</span>lambda)</span>
<span id="cb54-4"><a href="#cb54-4"></a></span>
<span id="cb54-5"><a href="#cb54-5"></a><span class="co"># only show 1se anyway</span></span>
<span id="cb54-6"><a href="#cb54-6"></a><span class="co"># if(ndx_1se != ndx_min) stop(&quot;lambda.1se != lambda.min&quot;)</span></span>
<span id="cb54-7"><a href="#cb54-7"></a></span>
<span id="cb54-8"><a href="#cb54-8"></a></span>
<span id="cb54-9"><a href="#cb54-9"></a><span class="co"># train oof data - NOT CLEAR WHY THESE DIFFER FROM CV ERRORS EXTRACTED FROM MODEL</span></span>
<span id="cb54-10"><a href="#cb54-10"></a><span class="co"># Get relaxed lasso (gamma=0) oof error</span></span>
<span id="cb54-11"><a href="#cb54-11"></a>train_oofPred_relaxed_1se_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb54-12"><a href="#cb54-12"></a>  <span class="kw">logistic_f</span>(hcc5hmC_cv_lassoR<span class="op">$</span>fit.preval[[<span class="st">&quot;g:0&quot;</span>]][, ndx_1se]) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&quot;HCC&quot;</span>, <span class="st">&quot;Control&quot;</span></span>
<span id="cb54-13"><a href="#cb54-13"></a>)</span>
<span id="cb54-14"><a href="#cb54-14"></a>train_oofPred_relaxed_1se_error &lt;-<span class="st"> </span><span class="kw">mean</span>(train_oofPred_relaxed_1se_vec <span class="op">!=</span><span class="st"> </span>hcc5hmC_train_group_vec)</span>
<span id="cb54-15"><a href="#cb54-15"></a></span>
<span id="cb54-16"><a href="#cb54-16"></a><span class="co"># blended mix (gamma=0.5)</span></span>
<span id="cb54-17"><a href="#cb54-17"></a>train_oofPred_blended_1se_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb54-18"><a href="#cb54-18"></a>  <span class="kw">logistic_f</span>(hcc5hmC_cv_lassoR<span class="op">$</span>fit.preval[[<span class="st">&quot;g:0.5&quot;</span>]][, ndx_1se]) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&quot;HCC&quot;</span>, <span class="st">&quot;Control&quot;</span></span>
<span id="cb54-19"><a href="#cb54-19"></a>)</span>
<span id="cb54-20"><a href="#cb54-20"></a>train_oofPred_blended_1se_error &lt;-<span class="st"> </span><span class="kw">mean</span>(train_oofPred_blended_1se_vec <span class="op">!=</span><span class="st"> </span>hcc5hmC_train_group_vec)</span>
<span id="cb54-21"><a href="#cb54-21"></a></span>
<span id="cb54-22"><a href="#cb54-22"></a></span>
<span id="cb54-23"><a href="#cb54-23"></a><span class="co"># Test set error - relaxed</span></span>
<span id="cb54-24"><a href="#cb54-24"></a>test_pred_relaxed_1se_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb54-25"><a href="#cb54-25"></a>  hcc5hmC_cv_lassoR,</span>
<span id="cb54-26"><a href="#cb54-26"></a>  <span class="dt">newx =</span> hcc5hmC_test_lcpm_mtx,</span>
<span id="cb54-27"><a href="#cb54-27"></a>  <span class="dt">s =</span> <span class="st">&quot;lambda.1se&quot;</span>,</span>
<span id="cb54-28"><a href="#cb54-28"></a>  <span class="dt">type =</span> <span class="st">&quot;class&quot;</span>,</span>
<span id="cb54-29"><a href="#cb54-29"></a>  <span class="dt">gamma =</span> <span class="dv">0</span></span>
<span id="cb54-30"><a href="#cb54-30"></a>)</span>
<span id="cb54-31"><a href="#cb54-31"></a>test_pred_relaxed_1se_error &lt;-<span class="st"> </span><span class="kw">mean</span>(test_pred_relaxed_1se_vec <span class="op">!=</span><span class="st"> </span>hcc5hmC_test_group_vec)</span>
<span id="cb54-32"><a href="#cb54-32"></a></span>
<span id="cb54-33"><a href="#cb54-33"></a><span class="co"># Test set error - blended</span></span>
<span id="cb54-34"><a href="#cb54-34"></a>test_pred_blended_1se_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb54-35"><a href="#cb54-35"></a>  hcc5hmC_cv_lassoR,</span>
<span id="cb54-36"><a href="#cb54-36"></a>  <span class="dt">newx =</span> hcc5hmC_test_lcpm_mtx,</span>
<span id="cb54-37"><a href="#cb54-37"></a>  <span class="dt">s =</span> <span class="st">&quot;lambda.1se&quot;</span>,</span>
<span id="cb54-38"><a href="#cb54-38"></a>  <span class="dt">type =</span> <span class="st">&quot;class&quot;</span>,</span>
<span id="cb54-39"><a href="#cb54-39"></a>  <span class="dt">gamma =</span> <span class="fl">0.5</span></span>
<span id="cb54-40"><a href="#cb54-40"></a>)</span>
<span id="cb54-41"><a href="#cb54-41"></a>test_pred_blended_1se_error &lt;-<span class="st"> </span><span class="kw">mean</span>(test_pred_blended_1se_vec <span class="op">!=</span><span class="st"> </span>hcc5hmC_test_group_vec)</span>
<span id="cb54-42"><a href="#cb54-42"></a></span>
<span id="cb54-43"><a href="#cb54-43"></a></span>
<span id="cb54-44"><a href="#cb54-44"></a>hcc5hmC_cv_lassoR_1se_error &lt;-<span class="st"> </span>hcc5hmC_cv_lassoR<span class="op">$</span>cvm[hcc5hmC_cv_lassoR<span class="op">$</span>lambda<span class="op">==</span>hcc5hmC_cv_lassoR<span class="op">$</span>lambda.min]</span>
<span id="cb54-45"><a href="#cb54-45"></a></span>
<span id="cb54-46"><a href="#cb54-46"></a>cv_blended_statlist &lt;-<span class="st"> </span>hcc5hmC_cv_lassoR<span class="op">$</span>relaxed<span class="op">$</span>statlist[[<span class="st">&#39;g:0.5&#39;</span>]]</span>
<span id="cb54-47"><a href="#cb54-47"></a>cv_blended_1se_error &lt;-<span class="st"> </span>cv_blended_statlist<span class="op">$</span>cvm[cv_blended_statlist<span class="op">$</span>lambda<span class="op">==</span></span>
<span id="cb54-48"><a href="#cb54-48"></a><span class="st">   </span>hcc5hmC_cv_lassoR<span class="op">$</span>relaxed<span class="op">$</span>lambda<span class="fl">.1</span>se]</span>
<span id="cb54-49"><a href="#cb54-49"></a> </span>
<span id="cb54-50"><a href="#cb54-50"></a></span>
<span id="cb54-51"><a href="#cb54-51"></a></span>
<span id="cb54-52"><a href="#cb54-52"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">data.frame</span>(</span>
<span id="cb54-53"><a href="#cb54-53"></a>  <span class="dt">train_relaxed  =</span> hcc5hmC_cv_lassoR_1se_error,</span>
<span id="cb54-54"><a href="#cb54-54"></a>  <span class="dt">train_blended  =</span> cv_blended_1se_error,</span>
<span id="cb54-55"><a href="#cb54-55"></a>  <span class="co">#train_relaxed_oof = train_oofPred_relaxed_1se_error,</span></span>
<span id="cb54-56"><a href="#cb54-56"></a>  <span class="co">#train_blended_oof = train_oofPred_blended_1se_error,</span></span>
<span id="cb54-57"><a href="#cb54-57"></a>  <span class="dt">test_relaxed  =</span> test_pred_relaxed_1se_error,</span>
<span id="cb54-58"><a href="#cb54-58"></a>  <span class="dt">test_blended  =</span> test_pred_blended_1se_error</span>
<span id="cb54-59"><a href="#cb54-59"></a>)) <span class="op">*</span><span class="st"> </span><span class="dv">100</span>,</span>
<span id="cb54-60"><a href="#cb54-60"></a><span class="dt">digits =</span> <span class="dv">1</span>,</span>
<span id="cb54-61"><a href="#cb54-61"></a><span class="dt">caption =</span> <span class="st">&quot;Relaxed lasso and blended mix error rates&quot;</span></span>
<span id="cb54-62"><a href="#cb54-62"></a>) <span class="op">%&gt;%</span></span>
<span id="cb54-63"><a href="#cb54-63"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-glmnetFit-lookLassoR2)Relaxed lasso and blended mix error rates
</caption>
<tbody>
<tr>
<td style="text-align:left;">
train_relaxed
</td>
<td style="text-align:right;">
7.0
</td>
</tr>
<tr>
<td style="text-align:left;">
train_blended
</td>
<td style="text-align:right;">
6.6
</td>
</tr>
<tr>
<td style="text-align:left;">
test_relaxed
</td>
<td style="text-align:right;">
10.5
</td>
</tr>
<tr>
<td style="text-align:left;">
test_blended
</td>
<td style="text-align:right;">
9.8
</td>
</tr>
</tbody>
</table>
<p><br/></p>
<p>The relaxed lasso and blended mix error rates are comparable to the
regular lasso fit error rate. We see here too that the reported cv
error rates are optimistic.
<!-- , while out-of-fold error rates
continue to be good indicators of unseen data error rates, as captured
by the test set.  
--></p>
<p>The <em>1se</em> lambda rule applied to the relaxed lasso fit selected a model with
<span class="math inline">\(68\)</span> features,
while for the blended mix model
(See @ref(eq:blended) in Section @ref(modeling-background))
the <em>1se</em> lambda rule selected
<span class="math inline">\(33\)</span> features (vertical
dotted reference line in Figure @ref(fig:lookLassoR)).
This feature is pointed out in the
<a href="https://cran.r-project.org/web/packages/glmnet/vignettes/relax.pdf">glmnet 3.0 vignette</a>:
<em>The debiasing will potentially improve prediction performance,
and CV will typically select a model with a smaller number of variables.</em></p>
</div>
<div id="examination-of-sensitivity-vs-specificity" class="section level2" number="4.4">
<h2 number="4.4"><span class="header-section-number">4.4</span> Examination of sensitivity vs specificity</h2>
<p>In the results above we reported error rates without inspecting the
sensitivity versus specificity trade-off. ROC curves can be examined
to get a sense of the trade-off.</p>
<div id="training-data-out-of-fold-roc-curves" class="section level3" number="4.4.1">
<h3 number="4.4.1"><span class="header-section-number">4.4.1</span> Training data out-of-fold ROC curves</h3>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1"></a><span class="co"># train</span></span>
<span id="cb55-2"><a href="#cb55-2"></a><span class="co"># lasso</span></span>
<span id="cb55-3"><a href="#cb55-3"></a>ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(hcc5hmC_cv_lasso<span class="op">$</span>lambda<span class="fl">.1</span>se,hcc5hmC_cv_lasso<span class="op">$</span>lambda)</span>
<span id="cb55-4"><a href="#cb55-4"></a>train_lasso_oofProb_vec &lt;-<span class="st"> </span><span class="kw">logistic_f</span>(hcc5hmC_cv_lasso<span class="op">$</span>fit.preval[,ndx_1se])</span>
<span id="cb55-5"><a href="#cb55-5"></a>train_lasso_roc &lt;-<span class="st"> </span>pROC<span class="op">::</span><span class="kw">roc</span>(</span>
<span id="cb55-6"><a href="#cb55-6"></a> <span class="dt">response =</span> <span class="kw">as.numeric</span>(hcc5hmC_train_group_vec<span class="op">==</span><span class="st">&#39;HCC&#39;</span>),</span>
<span id="cb55-7"><a href="#cb55-7"></a> <span class="dt">predictor =</span> train_lasso_oofProb_vec)</span></code></pre></div>
<pre><code>## Setting levels: control = 0, case = 1</code></pre>
<pre><code>## Setting direction: controls &lt; cases</code></pre>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1"></a><span class="co"># enet</span></span>
<span id="cb58-2"><a href="#cb58-2"></a>ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(hcc5hmC_cv_enet<span class="op">$</span>lambda<span class="fl">.1</span>se,hcc5hmC_cv_enet<span class="op">$</span>lambda)</span>
<span id="cb58-3"><a href="#cb58-3"></a>train_enet_oofProb_vec &lt;-<span class="st"> </span><span class="kw">logistic_f</span>(hcc5hmC_cv_enet<span class="op">$</span>fit.preval[,ndx_1se])</span>
<span id="cb58-4"><a href="#cb58-4"></a>train_enet_roc &lt;-<span class="st"> </span>pROC<span class="op">::</span><span class="kw">roc</span>(</span>
<span id="cb58-5"><a href="#cb58-5"></a> <span class="dt">response =</span> <span class="kw">as.numeric</span>(hcc5hmC_train_group_vec<span class="op">==</span><span class="st">&#39;HCC&#39;</span>),</span>
<span id="cb58-6"><a href="#cb58-6"></a> <span class="dt">predictor =</span> train_enet_oofProb_vec)</span></code></pre></div>
<pre><code>## Setting levels: control = 0, case = 1
## Setting direction: controls &lt; cases</code></pre>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1"></a><span class="co"># lasso - relaxed</span></span>
<span id="cb60-2"><a href="#cb60-2"></a>ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(hcc5hmC_cv_lassoR<span class="op">$</span>lambda<span class="fl">.1</span>se,hcc5hmC_cv_lassoR<span class="op">$</span>lambda)</span>
<span id="cb60-3"><a href="#cb60-3"></a>train_relaxed_oofProb_vec &lt;-<span class="st"> </span><span class="kw">logistic_f</span>(hcc5hmC_cv_lassoR<span class="op">$</span>fit.preval[[<span class="st">&#39;g:0&#39;</span>]][,ndx_1se])</span>
<span id="cb60-4"><a href="#cb60-4"></a>train_relaxed_roc &lt;-<span class="st"> </span>pROC<span class="op">::</span><span class="kw">roc</span>(</span>
<span id="cb60-5"><a href="#cb60-5"></a> <span class="dt">response =</span> <span class="kw">as.numeric</span>(hcc5hmC_train_group_vec<span class="op">==</span><span class="st">&#39;HCC&#39;</span>),</span>
<span id="cb60-6"><a href="#cb60-6"></a> <span class="dt">predictor =</span> train_relaxed_oofProb_vec)</span></code></pre></div>
<pre><code>## Setting levels: control = 0, case = 1
## Setting direction: controls &lt; cases</code></pre>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1"></a><span class="co"># blended mix (gamma=0.5)</span></span>
<span id="cb62-2"><a href="#cb62-2"></a>ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(hcc5hmC_cv_lassoR<span class="op">$</span>lambda<span class="fl">.1</span>se,hcc5hmC_cv_lassoR<span class="op">$</span>lambda)</span>
<span id="cb62-3"><a href="#cb62-3"></a>train_blended_oofProb_vec &lt;-<span class="st"> </span><span class="kw">logistic_f</span>(hcc5hmC_cv_lassoR<span class="op">$</span>fit.preval[[<span class="st">&#39;g:0.5&#39;</span>]][,ndx_1se])</span>
<span id="cb62-4"><a href="#cb62-4"></a>train_blended_roc &lt;-<span class="st"> </span>pROC<span class="op">::</span><span class="kw">roc</span>(</span>
<span id="cb62-5"><a href="#cb62-5"></a> <span class="dt">response =</span> <span class="kw">as.numeric</span>(hcc5hmC_train_group_vec<span class="op">==</span><span class="st">&#39;HCC&#39;</span>),</span>
<span id="cb62-6"><a href="#cb62-6"></a> <span class="dt">predictor =</span> train_blended_oofProb_vec)</span></code></pre></div>
<pre><code>## Setting levels: control = 0, case = 1
## Setting direction: controls &lt; cases</code></pre>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1"></a><span class="kw">plot</span>(train_lasso_roc, <span class="dt">col=</span>col_vec[<span class="dv">1</span>])</span>
<span id="cb64-2"><a href="#cb64-2"></a><span class="kw">lines</span>(train_enet_roc, <span class="dt">col=</span>col_vec[<span class="dv">2</span>])</span>
<span id="cb64-3"><a href="#cb64-3"></a><span class="kw">lines</span>(train_relaxed_roc, <span class="dt">col=</span>col_vec[<span class="dv">3</span>])</span>
<span id="cb64-4"><a href="#cb64-4"></a><span class="kw">lines</span>(train_blended_roc, <span class="dt">col=</span>col_vec[<span class="dv">4</span>])</span>
<span id="cb64-5"><a href="#cb64-5"></a></span>
<span id="cb64-6"><a href="#cb64-6"></a><span class="kw">legend</span>(<span class="st">&#39;bottomright&#39;</span>, <span class="dt">title=</span><span class="st">&#39;AUC&#39;</span>,</span>
<span id="cb64-7"><a href="#cb64-7"></a> <span class="dt">legend=</span><span class="kw">c</span>(</span>
<span id="cb64-8"><a href="#cb64-8"></a>  <span class="kw">paste</span>(<span class="st">&#39;lasso =&#39;</span>, <span class="kw">round</span>(train_lasso_roc[[<span class="st">&#39;auc&#39;</span>]],<span class="dv">3</span>)),</span>
<span id="cb64-9"><a href="#cb64-9"></a>  <span class="kw">paste</span>(<span class="st">&#39;enet =&#39;</span>, <span class="kw">round</span>(train_enet_roc[[<span class="st">&#39;auc&#39;</span>]],<span class="dv">3</span>)),</span>
<span id="cb64-10"><a href="#cb64-10"></a>  <span class="kw">paste</span>(<span class="st">&#39;relaxed =&#39;</span>, <span class="kw">round</span>(train_relaxed_roc[[<span class="st">&#39;auc&#39;</span>]],<span class="dv">3</span>)),</span>
<span id="cb64-11"><a href="#cb64-11"></a>  <span class="kw">paste</span>(<span class="st">&#39;blended =&#39;</span>, <span class="kw">round</span>(train_blended_roc[[<span class="st">&#39;auc&#39;</span>]],<span class="dv">3</span>))</span>
<span id="cb64-12"><a href="#cb64-12"></a> ),</span>
<span id="cb64-13"><a href="#cb64-13"></a> <span class="dt">text.col =</span> col_vec[<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>],</span>
<span id="cb64-14"><a href="#cb64-14"></a> <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb64-15"><a href="#cb64-15"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetFit-trainROC-1.png" alt="Train data out-of-sample ROCs" width="480" />
<p class="caption">
(#fig:hcc5hmC-glmnetFit-trainROC)Train data out-of-sample ROCs
</p>
</div>
<p>Compare thresholds for 90% Specificity:</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1"></a> lasso_ndx &lt;-<span class="st"> </span><span class="kw">with</span>(<span class="kw">as.data.frame</span>(pROC<span class="op">::</span><span class="kw">coords</span>(train_lasso_roc, <span class="dt">transpose=</span>F)), </span>
<span id="cb65-2"><a href="#cb65-2"></a>   <span class="kw">min</span>(<span class="kw">which</span>(specificity <span class="op">&gt;=</span><span class="st"> </span><span class="fl">0.9</span>)))</span>
<span id="cb65-3"><a href="#cb65-3"></a></span>
<span id="cb65-4"><a href="#cb65-4"></a> enet_ndx &lt;-<span class="st"> </span><span class="kw">with</span>(<span class="kw">as.data.frame</span>(pROC<span class="op">::</span><span class="kw">coords</span>(train_enet_roc, <span class="dt">transpose=</span>F)), </span>
<span id="cb65-5"><a href="#cb65-5"></a>   <span class="kw">min</span>(<span class="kw">which</span>(specificity <span class="op">&gt;=</span><span class="st"> </span><span class="fl">0.9</span>)))</span>
<span id="cb65-6"><a href="#cb65-6"></a></span>
<span id="cb65-7"><a href="#cb65-7"></a> lassoR_ndx &lt;-<span class="st"> </span><span class="kw">with</span>(<span class="kw">as.data.frame</span>(pROC<span class="op">::</span><span class="kw">coords</span>(train_relaxed_roc, <span class="dt">transpose=</span>F)), </span>
<span id="cb65-8"><a href="#cb65-8"></a>   <span class="kw">min</span>(<span class="kw">which</span>(specificity <span class="op">&gt;=</span><span class="st"> </span><span class="fl">0.9</span>)))</span>
<span id="cb65-9"><a href="#cb65-9"></a></span>
<span id="cb65-10"><a href="#cb65-10"></a> blended_ndx &lt;-<span class="st"> </span><span class="kw">with</span>(<span class="kw">as.data.frame</span>(pROC<span class="op">::</span><span class="kw">coords</span>(train_blended_roc, <span class="dt">transpose=</span>F)), </span>
<span id="cb65-11"><a href="#cb65-11"></a>   <span class="kw">min</span>(<span class="kw">which</span>(specificity <span class="op">&gt;=</span><span class="st"> </span><span class="fl">0.9</span>)))</span>
<span id="cb65-12"><a href="#cb65-12"></a></span>
<span id="cb65-13"><a href="#cb65-13"></a>  spec90_frm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">rbind</span>(</span>
<span id="cb65-14"><a href="#cb65-14"></a>  <span class="dt">lasso=</span><span class="kw">as.data.frame</span>(pROC<span class="op">::</span><span class="kw">coords</span>(train_lasso_roc, <span class="dt">transpose=</span>F))[lasso_ndx,],</span>
<span id="cb65-15"><a href="#cb65-15"></a>  <span class="dt">enet=</span><span class="kw">as.data.frame</span>(pROC<span class="op">::</span><span class="kw">coords</span>(train_enet_roc, <span class="dt">transpose=</span>F))[enet_ndx,],</span>
<span id="cb65-16"><a href="#cb65-16"></a>  <span class="dt">relaxed=</span><span class="kw">as.data.frame</span>(pROC<span class="op">::</span><span class="kw">coords</span>(train_relaxed_roc, <span class="dt">transpose=</span>F))[lassoR_ndx,],</span>
<span id="cb65-17"><a href="#cb65-17"></a>  <span class="dt">blended=</span><span class="kw">as.data.frame</span>(pROC<span class="op">::</span><span class="kw">coords</span>(train_blended_roc, <span class="dt">transpose=</span>F))[blended_ndx,]</span>
<span id="cb65-18"><a href="#cb65-18"></a> ))</span>
<span id="cb65-19"><a href="#cb65-19"></a></span>
<span id="cb65-20"><a href="#cb65-20"></a></span>
<span id="cb65-21"><a href="#cb65-21"></a>knitr<span class="op">::</span><span class="kw">kable</span>(spec90_frm,</span>
<span id="cb65-22"><a href="#cb65-22"></a>  <span class="dt">digits=</span><span class="dv">3</span>,</span>
<span id="cb65-23"><a href="#cb65-23"></a>  <span class="dt">caption=</span><span class="st">&quot;Specificity = .90 Coordinates&quot;</span></span>
<span id="cb65-24"><a href="#cb65-24"></a>) <span class="op">%&gt;%</span></span>
<span id="cb65-25"><a href="#cb65-25"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-glmnetFit-thresh90)Specificity = .90 Coordinates
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
threshold
</th>
<th style="text-align:right;">
specificity
</th>
<th style="text-align:right;">
sensitivity
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
lasso
</td>
<td style="text-align:right;">
0.358
</td>
<td style="text-align:right;">
0.9
</td>
<td style="text-align:right;">
0.932
</td>
</tr>
<tr>
<td style="text-align:left;">
enet
</td>
<td style="text-align:right;">
0.350
</td>
<td style="text-align:right;">
0.9
</td>
<td style="text-align:right;">
0.937
</td>
</tr>
<tr>
<td style="text-align:left;">
relaxed
</td>
<td style="text-align:right;">
0.356
</td>
<td style="text-align:right;">
0.9
</td>
<td style="text-align:right;">
0.860
</td>
</tr>
<tr>
<td style="text-align:left;">
blended
</td>
<td style="text-align:right;">
0.281
</td>
<td style="text-align:right;">
0.9
</td>
<td style="text-align:right;">
0.867
</td>
</tr>
</tbody>
</table>
<p>This is strange.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1"></a><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>), <span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">3</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb66-2"><a href="#cb66-2"></a></span>
<span id="cb66-3"><a href="#cb66-3"></a><span class="co"># lasso</span></span>
<span id="cb66-4"><a href="#cb66-4"></a><span class="kw">plot</span>(<span class="kw">density</span>(train_lasso_oofProb_vec[hcc5hmC_train_group_vec <span class="op">==</span><span class="st"> &quot;Control&quot;</span>]),</span>
<span id="cb66-5"><a href="#cb66-5"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;green&quot;</span></span>
<span id="cb66-6"><a href="#cb66-6"></a>)</span>
<span id="cb66-7"><a href="#cb66-7"></a><span class="kw">lines</span>(<span class="kw">density</span>(train_lasso_oofProb_vec[hcc5hmC_train_group_vec <span class="op">==</span><span class="st"> &quot;HCC&quot;</span>]),</span>
<span id="cb66-8"><a href="#cb66-8"></a>  <span class="dt">col =</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb66-9"><a href="#cb66-9"></a>)</span>
<span id="cb66-10"><a href="#cb66-10"></a><span class="kw">title</span>(<span class="st">&quot;lasso&quot;</span>)</span>
<span id="cb66-11"><a href="#cb66-11"></a><span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&quot;Control&quot;</span>, <span class="st">&quot;HCC&quot;</span>), <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;red&quot;</span>))</span>
<span id="cb66-12"><a href="#cb66-12"></a></span>
<span id="cb66-13"><a href="#cb66-13"></a><span class="co"># enet</span></span>
<span id="cb66-14"><a href="#cb66-14"></a><span class="kw">plot</span>(<span class="kw">density</span>(train_enet_oofProb_vec[hcc5hmC_train_group_vec <span class="op">==</span><span class="st"> &quot;Control&quot;</span>]),</span>
<span id="cb66-15"><a href="#cb66-15"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;green&quot;</span></span>
<span id="cb66-16"><a href="#cb66-16"></a>)</span>
<span id="cb66-17"><a href="#cb66-17"></a><span class="kw">lines</span>(<span class="kw">density</span>(train_enet_oofProb_vec[hcc5hmC_train_group_vec <span class="op">==</span><span class="st"> &quot;HCC&quot;</span>]),</span>
<span id="cb66-18"><a href="#cb66-18"></a>  <span class="dt">col =</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb66-19"><a href="#cb66-19"></a>)</span>
<span id="cb66-20"><a href="#cb66-20"></a><span class="kw">title</span>(<span class="st">&quot;enet&quot;</span>)</span>
<span id="cb66-21"><a href="#cb66-21"></a></span>
<span id="cb66-22"><a href="#cb66-22"></a><span class="co"># lassoR</span></span>
<span id="cb66-23"><a href="#cb66-23"></a><span class="kw">plot</span>(<span class="kw">density</span>(train_relaxed_oofProb_vec[hcc5hmC_train_group_vec <span class="op">==</span><span class="st"> &quot;Control&quot;</span>]),</span>
<span id="cb66-24"><a href="#cb66-24"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;green&quot;</span></span>
<span id="cb66-25"><a href="#cb66-25"></a>)</span>
<span id="cb66-26"><a href="#cb66-26"></a><span class="kw">lines</span>(<span class="kw">density</span>(train_relaxed_oofProb_vec[hcc5hmC_train_group_vec <span class="op">==</span><span class="st"> &quot;HCC&quot;</span>]),</span>
<span id="cb66-27"><a href="#cb66-27"></a>  <span class="dt">col =</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb66-28"><a href="#cb66-28"></a>)</span>
<span id="cb66-29"><a href="#cb66-29"></a><span class="kw">title</span>(<span class="st">&quot;lassoR&quot;</span>)</span>
<span id="cb66-30"><a href="#cb66-30"></a></span>
<span id="cb66-31"><a href="#cb66-31"></a><span class="co"># blended</span></span>
<span id="cb66-32"><a href="#cb66-32"></a><span class="kw">plot</span>(<span class="kw">density</span>(train_blended_oofProb_vec[hcc5hmC_train_group_vec <span class="op">==</span><span class="st"> &quot;Control&quot;</span>]),</span>
<span id="cb66-33"><a href="#cb66-33"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;green&quot;</span></span>
<span id="cb66-34"><a href="#cb66-34"></a>)</span>
<span id="cb66-35"><a href="#cb66-35"></a><span class="kw">lines</span>(<span class="kw">density</span>(train_blended_oofProb_vec[hcc5hmC_train_group_vec <span class="op">==</span><span class="st"> &quot;HCC&quot;</span>]),</span>
<span id="cb66-36"><a href="#cb66-36"></a>  <span class="dt">col =</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb66-37"><a href="#cb66-37"></a>)</span>
<span id="cb66-38"><a href="#cb66-38"></a><span class="kw">title</span>(<span class="st">&quot;blended&quot;</span>)</span>
<span id="cb66-39"><a href="#cb66-39"></a></span>
<span id="cb66-40"><a href="#cb66-40"></a><span class="kw">mtext</span>(<span class="dt">side =</span> <span class="dv">1</span>, <span class="dt">outer =</span> T, <span class="st">&quot;out-of-fold predicted probability&quot;</span>, <span class="dt">cex =</span> <span class="fl">1.25</span>)</span>
<span id="cb66-41"><a href="#cb66-41"></a><span class="kw">mtext</span>(<span class="dt">side =</span> <span class="dv">2</span>, <span class="dt">outer =</span> T, <span class="st">&quot;density&quot;</span>, <span class="dt">cex =</span> <span class="fl">1.25</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetFit-trainOOFprops-1.png" alt="Train data out-of-fold predicted probabilities" width="960" />
<p class="caption">
(#fig:hcc5hmC-glmnetFit-trainOOFprops)Train data out-of-fold predicted probabilities
</p>
</div>
<p>The relaxed lasso fit results in essentially dichotomized predicted probability
distribution - predicted probabilities are very close to 0 or 1.</p>
<p>Look at test data ROC curves.</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1"></a><span class="co"># plot all</span></span>
<span id="cb67-2"><a href="#cb67-2"></a><span class="kw">plot</span>(test_lasso_roc, <span class="dt">col =</span> col_vec[<span class="dv">1</span>])</span>
<span id="cb67-3"><a href="#cb67-3"></a><span class="kw">lines</span>(test_enet_roc, <span class="dt">col =</span> col_vec[<span class="dv">2</span>])</span>
<span id="cb67-4"><a href="#cb67-4"></a><span class="kw">lines</span>(test_relaxed_roc, <span class="dt">col =</span> col_vec[<span class="dv">3</span>])</span>
<span id="cb67-5"><a href="#cb67-5"></a><span class="kw">lines</span>(test_blended_roc, <span class="dt">col =</span> col_vec[<span class="dv">4</span>])</span>
<span id="cb67-6"><a href="#cb67-6"></a></span>
<span id="cb67-7"><a href="#cb67-7"></a><span class="kw">legend</span>(<span class="st">&quot;bottomright&quot;</span>,</span>
<span id="cb67-8"><a href="#cb67-8"></a>  <span class="dt">title =</span> <span class="st">&quot;AUC&quot;</span>,</span>
<span id="cb67-9"><a href="#cb67-9"></a>  <span class="dt">legend =</span> <span class="kw">c</span>(</span>
<span id="cb67-10"><a href="#cb67-10"></a>    <span class="kw">paste</span>(<span class="st">&quot;lasso =&quot;</span>, <span class="kw">round</span>(test_lasso_roc[[<span class="st">&quot;auc&quot;</span>]], <span class="dv">3</span>)),</span>
<span id="cb67-11"><a href="#cb67-11"></a>    <span class="kw">paste</span>(<span class="st">&quot;enet =&quot;</span>, <span class="kw">round</span>(test_enet_roc[[<span class="st">&quot;auc&quot;</span>]], <span class="dv">3</span>)),</span>
<span id="cb67-12"><a href="#cb67-12"></a>    <span class="kw">paste</span>(<span class="st">&quot;relaxed =&quot;</span>, <span class="kw">round</span>(test_relaxed_roc[[<span class="st">&quot;auc&quot;</span>]], <span class="dv">3</span>)),</span>
<span id="cb67-13"><a href="#cb67-13"></a>    <span class="kw">paste</span>(<span class="st">&quot;blended =&quot;</span>, <span class="kw">round</span>(test_blended_roc[[<span class="st">&quot;auc&quot;</span>]], <span class="dv">3</span>))</span>
<span id="cb67-14"><a href="#cb67-14"></a>  ),</span>
<span id="cb67-15"><a href="#cb67-15"></a>  <span class="dt">text.col =</span> col_vec[<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>],</span>
<span id="cb67-16"><a href="#cb67-16"></a>  <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb67-17"><a href="#cb67-17"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetFit-testROC2-1.png" alt="Test data out-of-sample ROCs" width="480" />
<p class="caption">
(#fig:hcc5hmC-glmnetFit-testROC2)Test data out-of-sample ROCs
</p>
</div>
<p>Look at densities of predicted probabilities.</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1"></a><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>), <span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">3</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb68-2"><a href="#cb68-2"></a></span>
<span id="cb68-3"><a href="#cb68-3"></a><span class="co"># lasso</span></span>
<span id="cb68-4"><a href="#cb68-4"></a><span class="kw">plot</span>(<span class="kw">density</span>(test_lasso_predProb_vec[hcc5hmC_test_group_vec <span class="op">==</span><span class="st"> &quot;Control&quot;</span>]),</span>
<span id="cb68-5"><a href="#cb68-5"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;green&quot;</span></span>
<span id="cb68-6"><a href="#cb68-6"></a>)</span>
<span id="cb68-7"><a href="#cb68-7"></a><span class="kw">lines</span>(<span class="kw">density</span>(test_lasso_predProb_vec[hcc5hmC_test_group_vec <span class="op">==</span><span class="st"> &quot;HCC&quot;</span>]),</span>
<span id="cb68-8"><a href="#cb68-8"></a>  <span class="dt">col =</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb68-9"><a href="#cb68-9"></a>)</span>
<span id="cb68-10"><a href="#cb68-10"></a><span class="kw">title</span>(<span class="st">&quot;lasso&quot;</span>)</span>
<span id="cb68-11"><a href="#cb68-11"></a><span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&quot;Control&quot;</span>, <span class="st">&quot;HCC&quot;</span>), <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;red&quot;</span>))</span>
<span id="cb68-12"><a href="#cb68-12"></a></span>
<span id="cb68-13"><a href="#cb68-13"></a><span class="co"># enet</span></span>
<span id="cb68-14"><a href="#cb68-14"></a><span class="kw">plot</span>(<span class="kw">density</span>(test_enet_predProb_vec[hcc5hmC_test_group_vec <span class="op">==</span><span class="st"> &quot;Control&quot;</span>]),</span>
<span id="cb68-15"><a href="#cb68-15"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;green&quot;</span></span>
<span id="cb68-16"><a href="#cb68-16"></a>)</span>
<span id="cb68-17"><a href="#cb68-17"></a><span class="kw">lines</span>(<span class="kw">density</span>(test_enet_predProb_vec[hcc5hmC_test_group_vec <span class="op">==</span><span class="st"> &quot;HCC&quot;</span>]),</span>
<span id="cb68-18"><a href="#cb68-18"></a>  <span class="dt">col =</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb68-19"><a href="#cb68-19"></a>)</span>
<span id="cb68-20"><a href="#cb68-20"></a><span class="kw">title</span>(<span class="st">&quot;enet&quot;</span>)</span>
<span id="cb68-21"><a href="#cb68-21"></a></span>
<span id="cb68-22"><a href="#cb68-22"></a><span class="co"># relaxed</span></span>
<span id="cb68-23"><a href="#cb68-23"></a><span class="kw">plot</span>(<span class="kw">density</span>(test_relaxed_predProb_vec[hcc5hmC_test_group_vec <span class="op">==</span><span class="st"> &quot;Control&quot;</span>]),</span>
<span id="cb68-24"><a href="#cb68-24"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;green&quot;</span></span>
<span id="cb68-25"><a href="#cb68-25"></a>)</span>
<span id="cb68-26"><a href="#cb68-26"></a><span class="kw">lines</span>(<span class="kw">density</span>(test_relaxed_predProb_vec[hcc5hmC_test_group_vec <span class="op">==</span><span class="st"> &quot;HCC&quot;</span>]),</span>
<span id="cb68-27"><a href="#cb68-27"></a>  <span class="dt">col =</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb68-28"><a href="#cb68-28"></a>)</span>
<span id="cb68-29"><a href="#cb68-29"></a><span class="kw">title</span>(<span class="st">&quot;relaxed&quot;</span>)</span>
<span id="cb68-30"><a href="#cb68-30"></a></span>
<span id="cb68-31"><a href="#cb68-31"></a><span class="co">#sapply(split(test_relaxed_predProb_vec, hcc5hmC_test_group_vec), summary)</span></span>
<span id="cb68-32"><a href="#cb68-32"></a></span>
<span id="cb68-33"><a href="#cb68-33"></a></span>
<span id="cb68-34"><a href="#cb68-34"></a><span class="co"># blended</span></span>
<span id="cb68-35"><a href="#cb68-35"></a><span class="kw">plot</span>(<span class="kw">density</span>(test_blended_predProb_vec[hcc5hmC_test_group_vec <span class="op">==</span><span class="st"> &quot;Control&quot;</span>]),</span>
<span id="cb68-36"><a href="#cb68-36"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;green&quot;</span></span>
<span id="cb68-37"><a href="#cb68-37"></a>)</span>
<span id="cb68-38"><a href="#cb68-38"></a><span class="kw">lines</span>(<span class="kw">density</span>(test_blended_predProb_vec[hcc5hmC_test_group_vec <span class="op">==</span><span class="st"> &quot;HCC&quot;</span>]),</span>
<span id="cb68-39"><a href="#cb68-39"></a>  <span class="dt">col =</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb68-40"><a href="#cb68-40"></a>)</span>
<span id="cb68-41"><a href="#cb68-41"></a><span class="kw">title</span>(<span class="st">&quot;blended&quot;</span>)</span>
<span id="cb68-42"><a href="#cb68-42"></a></span>
<span id="cb68-43"><a href="#cb68-43"></a><span class="kw">mtext</span>(<span class="dt">side =</span> <span class="dv">1</span>, <span class="dt">outer =</span> T, <span class="st">&quot;test set predicted probability&quot;</span>, <span class="dt">cex =</span> <span class="fl">1.25</span>)</span>
<span id="cb68-44"><a href="#cb68-44"></a><span class="kw">mtext</span>(<span class="dt">side =</span> <span class="dv">2</span>, <span class="dt">outer =</span> T, <span class="st">&quot;density&quot;</span>, <span class="dt">cex =</span> <span class="fl">1.25</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetFit-testOOFprobs-1.png" alt="Test data out-of-fold predicted probabilities" width="960" />
<p class="caption">
(#fig:hcc5hmC-glmnetFit-testOOFprobs)Test data out-of-fold predicted probabilities
</p>
</div>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1"></a><span class="co"># Define plotting function</span></span>
<span id="cb69-2"><a href="#cb69-2"></a>bxpPredProb_f &lt;-<span class="st"> </span><span class="cf">function</span>(cv_fit, <span class="dt">Gamma=</span><span class="ot">NULL</span>) {</span>
<span id="cb69-3"><a href="#cb69-3"></a>  <span class="co"># Train - preval is out-of-fold linear predictor for training design points</span></span>
<span id="cb69-4"><a href="#cb69-4"></a>  onese_ndx &lt;-<span class="st"> </span><span class="kw">match</span>(cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se, cv_fit<span class="op">$</span>lambda)</span>
<span id="cb69-5"><a href="#cb69-5"></a>  <span class="cf">if</span>(<span class="kw">is.null</span>(Gamma)) </span>
<span id="cb69-6"><a href="#cb69-6"></a>   train_1se_preval_vec &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>fit.preval[, onese_ndx] <span class="cf">else</span></span>
<span id="cb69-7"><a href="#cb69-7"></a>   train_1se_preval_vec &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>fit.preval[[Gamma]][, onese_ndx] </span>
<span id="cb69-8"><a href="#cb69-8"></a></span>
<span id="cb69-9"><a href="#cb69-9"></a>  train_1se_predProb_vec &lt;-<span class="st"> </span><span class="kw">logistic_f</span>(train_1se_preval_vec)</span>
<span id="cb69-10"><a href="#cb69-10"></a></span>
<span id="cb69-11"><a href="#cb69-11"></a>  <span class="co"># Test</span></span>
<span id="cb69-12"><a href="#cb69-12"></a>  test_1se_predProb_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb69-13"><a href="#cb69-13"></a>    cv_fit,</span>
<span id="cb69-14"><a href="#cb69-14"></a>    <span class="dt">newx =</span> hcc5hmC_test_lcpm_mtx,</span>
<span id="cb69-15"><a href="#cb69-15"></a>    <span class="dt">s =</span> <span class="st">&quot;lambda.1se&quot;</span>,</span>
<span id="cb69-16"><a href="#cb69-16"></a>    <span class="dt">type =</span> <span class="st">&quot;resp&quot;</span></span>
<span id="cb69-17"><a href="#cb69-17"></a>  )</span>
<span id="cb69-18"><a href="#cb69-18"></a></span>
<span id="cb69-19"><a href="#cb69-19"></a>  tmp &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb69-20"><a href="#cb69-20"></a>    <span class="dt">train =</span> <span class="kw">split</span>(train_1se_predProb_vec, hcc5hmC_train_group_vec),</span>
<span id="cb69-21"><a href="#cb69-21"></a>    <span class="dt">test =</span> <span class="kw">split</span>(test_1se_predProb_vec, hcc5hmC_test_group_vec)</span>
<span id="cb69-22"><a href="#cb69-22"></a>  )</span>
<span id="cb69-23"><a href="#cb69-23"></a>  <span class="kw">names</span>(tmp) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>, <span class="kw">sub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">.&quot;</span>, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>, <span class="kw">names</span>(tmp)))</span>
<span id="cb69-24"><a href="#cb69-24"></a></span>
<span id="cb69-25"><a href="#cb69-25"></a>  <span class="kw">boxplot</span>(tmp)</span>
<span id="cb69-26"><a href="#cb69-26"></a>}</span>
<span id="cb69-27"><a href="#cb69-27"></a></span>
<span id="cb69-28"><a href="#cb69-28"></a><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>), <span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">5</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb69-29"><a href="#cb69-29"></a></span>
<span id="cb69-30"><a href="#cb69-30"></a><span class="kw">bxpPredProb_f</span>(hcc5hmC_cv_lasso)</span>
<span id="cb69-31"><a href="#cb69-31"></a><span class="kw">title</span>(<span class="st">&#39;lasso&#39;</span>)</span>
<span id="cb69-32"><a href="#cb69-32"></a></span>
<span id="cb69-33"><a href="#cb69-33"></a><span class="kw">bxpPredProb_f</span>(hcc5hmC_cv_enet)</span>
<span id="cb69-34"><a href="#cb69-34"></a><span class="kw">title</span>(<span class="st">&#39;enet&#39;</span>)</span>
<span id="cb69-35"><a href="#cb69-35"></a></span>
<span id="cb69-36"><a href="#cb69-36"></a><span class="kw">bxpPredProb_f</span>(hcc5hmC_cv_lassoR, <span class="dt">Gamma=</span><span class="st">&#39;g:0&#39;</span>)</span>
<span id="cb69-37"><a href="#cb69-37"></a><span class="kw">title</span>(<span class="st">&#39;relaxed&#39;</span>)</span>
<span id="cb69-38"><a href="#cb69-38"></a></span>
<span id="cb69-39"><a href="#cb69-39"></a><span class="kw">bxpPredProb_f</span>(hcc5hmC_cv_lassoR, <span class="dt">Gamma=</span><span class="st">&#39;g:0.5&#39;</span>)</span>
<span id="cb69-40"><a href="#cb69-40"></a><span class="kw">title</span>(<span class="st">&#39;blended&#39;</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetFit-fitPrevalByGroup-1.png" alt="Predicted Probabilities - Train and Test" width="768" />
<p class="caption">
(#fig:hcc5hmC-glmnetFit-fitPrevalByGroup)Predicted Probabilities - Train and Test
</p>
</div>
<!--
Another look - plot train and test set logistic curves with annotation.

The following shows that predicted classes come from fitted
probabilities - not out of sample probabilities.

Also shows that threshold is at 0.5 

SKIP
-->
<!-- SKIP ALL THIS 
We have seen above that assessments of model performance based on the out-of-fold 
predicted values are close to the test set assessments, and that
assessments based on prediction extracted from glmnet object are optimistic.
Here we look at confusion matrices to see how this affects the
classification results.

Here we use a threshold of 0.5 to dichotomize the predicted
probabilities into a class prediction, as is done in the
glmnet predictions.

-->
<!-- SKIPPED
The out-of-fold error rates are larger for the relaxed lasso and blended fit models.
On the test set, errors are slightly higher for the elastic net model.
-->
</div>
</div>
<div id="compare-predictions-at-misclassified-samples" class="section level2" number="4.5">
<h2 number="4.5"><span class="header-section-number">4.5</span> Compare predictions at misclassified samples</h2>
<p>It is useful to examine classification errors more carefully.
If models have different failure modes, one might get improved
performance by combining model predictions. Note that the models
considered here are not expected to compliment each other usefully
as they are too similar in nature.</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1"></a><span class="co"># </span><span class="al">NOTE</span><span class="co">: here we use computred oofClass rather than predClass </span></span>
<span id="cb70-2"><a href="#cb70-2"></a><span class="co"># as predClass extracted from predict() are fitted values.</span></span>
<span id="cb70-3"><a href="#cb70-3"></a></span>
<span id="cb70-4"><a href="#cb70-4"></a><span class="co"># train - oof</span></span>
<span id="cb70-5"><a href="#cb70-5"></a>ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(hcc5hmC_cv_lasso<span class="op">$</span>lambda<span class="fl">.1</span>se,hcc5hmC_cv_lasso<span class="op">$</span>lambda)</span>
<span id="cb70-6"><a href="#cb70-6"></a>train_lasso_oofProb_vec &lt;-<span class="st"> </span><span class="kw">logistic_f</span>(hcc5hmC_cv_lasso<span class="op">$</span>fit.preval[,ndx_1se])</span>
<span id="cb70-7"><a href="#cb70-7"></a>train_lasso_oofClass_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb70-8"><a href="#cb70-8"></a>   train_lasso_oofProb_vec <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&#39;HCC&#39;</span>, <span class="st">&#39;Control&#39;</span>)</span>
<span id="cb70-9"><a href="#cb70-9"></a></span>
<span id="cb70-10"><a href="#cb70-10"></a>ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(hcc5hmC_cv_enet<span class="op">$</span>lambda<span class="fl">.1</span>se,hcc5hmC_cv_enet<span class="op">$</span>lambda)</span>
<span id="cb70-11"><a href="#cb70-11"></a>train_enet_oofProb_vec &lt;-<span class="st"> </span><span class="kw">logistic_f</span>(hcc5hmC_cv_enet<span class="op">$</span>fit.preval[,ndx_1se])</span>
<span id="cb70-12"><a href="#cb70-12"></a>train_enet_oofClass_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb70-13"><a href="#cb70-13"></a>   train_enet_oofProb_vec <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&#39;HCC&#39;</span>, <span class="st">&#39;Control&#39;</span>)</span>
<span id="cb70-14"><a href="#cb70-14"></a></span>
<span id="cb70-15"><a href="#cb70-15"></a><span class="co"># RECALL: hcc5hmC_cv_lassoR$nzero[hcc5hmC_cv_lassoR$lambda==hcc5hmC_cv_lassoR$lambda.1se]</span></span>
<span id="cb70-16"><a href="#cb70-16"></a><span class="co"># train - oof</span></span>
<span id="cb70-17"><a href="#cb70-17"></a>ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(hcc5hmC_cv_lassoR<span class="op">$</span>lambda<span class="fl">.1</span>se,hcc5hmC_cv_lassoR<span class="op">$</span>lambda)</span>
<span id="cb70-18"><a href="#cb70-18"></a>train_relaxed_oofProb_vec &lt;-<span class="st"> </span><span class="kw">logistic_f</span>(hcc5hmC_cv_lassoR<span class="op">$</span>fit.preval[[<span class="st">&#39;g:0&#39;</span>]][,ndx_1se])</span>
<span id="cb70-19"><a href="#cb70-19"></a>train_relaxed_oofClass_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb70-20"><a href="#cb70-20"></a>   train_relaxed_oofProb_vec <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&#39;HCC&#39;</span>, <span class="st">&#39;Control&#39;</span>)</span>
<span id="cb70-21"><a href="#cb70-21"></a></span>
<span id="cb70-22"><a href="#cb70-22"></a></span>
<span id="cb70-23"><a href="#cb70-23"></a><span class="co"># RECALL $`r hcc5hmC_cv_lassoR$relaxed$nzero.1se`$ features (vertical</span></span>
<span id="cb70-24"><a href="#cb70-24"></a><span class="co">#  cv_blended_statlist &lt;- hcc5hmC_cv_lassoR$relaxed$statlist[[&#39;g:0.5&#39;]]</span></span>
<span id="cb70-25"><a href="#cb70-25"></a><span class="co">#  cv_blended_1se_error &lt;- cv_blended_statlist$cvm[cv_blended_statlist$lambda==</span></span>
<span id="cb70-26"><a href="#cb70-26"></a>      <span class="co">#hcc5hmC_cv_lassoR$relaxed$lambda.1se]</span></span>
<span id="cb70-27"><a href="#cb70-27"></a></span>
<span id="cb70-28"><a href="#cb70-28"></a><span class="co"># train - oof</span></span>
<span id="cb70-29"><a href="#cb70-29"></a>cv_blended_statlist &lt;-<span class="st"> </span>hcc5hmC_cv_lassoR<span class="op">$</span>relaxed<span class="op">$</span>statlist[[<span class="st">&#39;g:0.5&#39;</span>]]</span>
<span id="cb70-30"><a href="#cb70-30"></a>ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(hcc5hmC_cv_lassoR<span class="op">$</span>relaxed<span class="op">$</span>lambda<span class="fl">.1</span>se, cv_blended_statlist<span class="op">$</span>lambda)</span>
<span id="cb70-31"><a href="#cb70-31"></a>train_blended_oofProb_vec &lt;-<span class="st"> </span><span class="kw">logistic_f</span>(hcc5hmC_cv_lassoR<span class="op">$</span>fit.preval[[<span class="st">&#39;g:0.5&#39;</span>]][,ndx_1se])</span>
<span id="cb70-32"><a href="#cb70-32"></a>train_blended_oofClass_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb70-33"><a href="#cb70-33"></a>   train_blended_oofProb_vec <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&#39;HCC&#39;</span>, <span class="st">&#39;Control&#39;</span>)</span>
<span id="cb70-34"><a href="#cb70-34"></a></span>
<span id="cb70-35"><a href="#cb70-35"></a></span>
<span id="cb70-36"><a href="#cb70-36"></a>misclass_id_vec &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">c</span>(</span>
<span id="cb70-37"><a href="#cb70-37"></a> <span class="kw">names</span>(train_lasso_oofClass_vec)[train_lasso_oofClass_vec <span class="op">!=</span><span class="st"> </span>hcc5hmC_train_group_vec],</span>
<span id="cb70-38"><a href="#cb70-38"></a> <span class="kw">names</span>(train_enet_oofClass_vec)[train_enet_oofClass_vec <span class="op">!=</span><span class="st"> </span>hcc5hmC_train_group_vec],</span>
<span id="cb70-39"><a href="#cb70-39"></a> <span class="kw">names</span>(train_relaxed_oofClass_vec)[train_relaxed_oofClass_vec <span class="op">!=</span><span class="st"> </span>hcc5hmC_train_group_vec],</span>
<span id="cb70-40"><a href="#cb70-40"></a> <span class="kw">names</span>(train_blended_oofClass_vec)[train_blended_oofClass_vec <span class="op">!=</span><span class="st"> </span>hcc5hmC_train_group_vec]</span>
<span id="cb70-41"><a href="#cb70-41"></a> )</span>
<span id="cb70-42"><a href="#cb70-42"></a>)</span>
<span id="cb70-43"><a href="#cb70-43"></a></span>
<span id="cb70-44"><a href="#cb70-44"></a><span class="co"># Dont know why NAs creeo in here</span></span>
<span id="cb70-45"><a href="#cb70-45"></a></span>
<span id="cb70-46"><a href="#cb70-46"></a>missclass_oofProb_mtx &lt;-<span class="st"> </span><span class="kw">cbind</span>(</span>
<span id="cb70-47"><a href="#cb70-47"></a> train_lasso_oofProb_vec[misclass_id_vec],</span>
<span id="cb70-48"><a href="#cb70-48"></a> train_enet_oofProb_vec[misclass_id_vec],</span>
<span id="cb70-49"><a href="#cb70-49"></a> train_relaxed_oofProb_vec[misclass_id_vec],</span>
<span id="cb70-50"><a href="#cb70-50"></a> train_blended_oofProb_vec[misclass_id_vec]</span>
<span id="cb70-51"><a href="#cb70-51"></a>)</span>
<span id="cb70-52"><a href="#cb70-52"></a><span class="kw">colnames</span>(missclass_oofProb_mtx) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;lasso&#39;</span>,<span class="st">&#39;enet&#39;</span>, <span class="st">&#39;lassoR&#39;</span>, <span class="st">&#39;blended&#39;</span>)</span>
<span id="cb70-53"><a href="#cb70-53"></a></span>
<span id="cb70-54"><a href="#cb70-54"></a></span>
<span id="cb70-55"><a href="#cb70-55"></a>row_med_vec &lt;-<span class="st"> </span><span class="kw">apply</span>(missclass_oofProb_mtx, <span class="dv">1</span>, median)</span>
<span id="cb70-56"><a href="#cb70-56"></a>missclass_oofProb_mtx &lt;-<span class="st"> </span>missclass_oofProb_mtx[</span>
<span id="cb70-57"><a href="#cb70-57"></a>  <span class="kw">order</span>(hcc5hmC_train_group_vec[<span class="kw">rownames</span>(missclass_oofProb_mtx)], row_med_vec),]</span>
<span id="cb70-58"><a href="#cb70-58"></a></span>
<span id="cb70-59"><a href="#cb70-59"></a><span class="kw">plot</span>(</span>
<span id="cb70-60"><a href="#cb70-60"></a> <span class="dt">x=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="kw">nrow</span>(missclass_oofProb_mtx)), <span class="dt">xlab=</span><span class="st">&#39;samples&#39;</span>,</span>
<span id="cb70-61"><a href="#cb70-61"></a> <span class="dt">y=</span><span class="kw">range</span>(missclass_oofProb_mtx), <span class="dt">ylab=</span><span class="st">&#39;out-of-fold predicted probability&#39;</span>,</span>
<span id="cb70-62"><a href="#cb70-62"></a> <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>, <span class="dt">type=</span><span class="st">&#39;n&#39;</span>)</span>
<span id="cb70-63"><a href="#cb70-63"></a></span>
<span id="cb70-64"><a href="#cb70-64"></a><span class="cf">for</span>(RR <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(missclass_oofProb_mtx))</span>
<span id="cb70-65"><a href="#cb70-65"></a><span class="kw">points</span>(</span>
<span id="cb70-66"><a href="#cb70-66"></a> <span class="kw">rep</span>(RR, <span class="kw">ncol</span>(missclass_oofProb_mtx)), </span>
<span id="cb70-67"><a href="#cb70-67"></a> missclass_oofProb_mtx[RR,],</span>
<span id="cb70-68"><a href="#cb70-68"></a> <span class="dt">col=</span><span class="kw">ifelse</span>(hcc5hmC_train_group_vec[<span class="kw">rownames</span>(missclass_oofProb_mtx)[RR]] <span class="op">==</span><span class="st"> &#39;Control&#39;</span>,</span>
<span id="cb70-69"><a href="#cb70-69"></a>  <span class="st">&#39;green&#39;</span>, <span class="st">&#39;red&#39;</span>),</span>
<span id="cb70-70"><a href="#cb70-70"></a> <span class="dt">pch=</span><span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(missclass_oofProb_mtx))</span>
<span id="cb70-71"><a href="#cb70-71"></a></span>
<span id="cb70-72"><a href="#cb70-72"></a><span class="kw">legend</span>(<span class="st">&#39;top&#39;</span>, <span class="dt">ncol=</span><span class="dv">2</span>, <span class="dt">legend=</span><span class="kw">colnames</span>(missclass_oofProb_mtx), </span>
<span id="cb70-73"><a href="#cb70-73"></a> <span class="dt">pch=</span><span class="dv">1</span><span class="op">:</span><span class="dv">4</span>, <span class="dt">bty=</span><span class="st">&#39;n&#39;</span>)</span>
<span id="cb70-74"><a href="#cb70-74"></a></span>
<span id="cb70-75"><a href="#cb70-75"></a><span class="kw">abline</span>(<span class="dt">h=</span><span class="fl">0.5</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetFit-misclassTrain-1.png" alt="out-of-fold predicted probabilities at miscassified samples" width="768" />
<p class="caption">
(#fig:hcc5hmC-glmnetFit-misclassTrain)out-of-fold predicted probabilities at miscassified samples
</p>
</div>
<p>As we’ve seen above, predictions from lassoR and the blended mix model
are basically dichotomous; 0 or 1. Samples have been order by group, and
median P(HCC) within group. For the Controls (green), predicted probabilities
less than 0.5 are considered correct here. For the HCC (red) samples,
predicted probabilities greater than 0.5 are considered correct here.</p>
<p>Now look at the same plot on the test data set.</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1"></a>test_lasso_predClass_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb71-2"><a href="#cb71-2"></a> hcc5hmC_cv_lasso,</span>
<span id="cb71-3"><a href="#cb71-3"></a> <span class="dt">newx=</span>hcc5hmC_test_lcpm_mtx,</span>
<span id="cb71-4"><a href="#cb71-4"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span>,</span>
<span id="cb71-5"><a href="#cb71-5"></a> <span class="dt">type=</span><span class="st">&#39;class&#39;</span></span>
<span id="cb71-6"><a href="#cb71-6"></a>)</span>
<span id="cb71-7"><a href="#cb71-7"></a></span>
<span id="cb71-8"><a href="#cb71-8"></a>test_enet_predClass_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb71-9"><a href="#cb71-9"></a> hcc5hmC_cv_enet,</span>
<span id="cb71-10"><a href="#cb71-10"></a> <span class="dt">newx=</span>hcc5hmC_test_lcpm_mtx,</span>
<span id="cb71-11"><a href="#cb71-11"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span>,</span>
<span id="cb71-12"><a href="#cb71-12"></a> <span class="dt">type=</span><span class="st">&#39;class&#39;</span></span>
<span id="cb71-13"><a href="#cb71-13"></a>)</span>
<span id="cb71-14"><a href="#cb71-14"></a></span>
<span id="cb71-15"><a href="#cb71-15"></a>test_relaxed_predClass_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb71-16"><a href="#cb71-16"></a> hcc5hmC_cv_lassoR,</span>
<span id="cb71-17"><a href="#cb71-17"></a> <span class="dt">g=</span><span class="dv">0</span>,</span>
<span id="cb71-18"><a href="#cb71-18"></a> <span class="dt">newx=</span>hcc5hmC_test_lcpm_mtx,</span>
<span id="cb71-19"><a href="#cb71-19"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span>,</span>
<span id="cb71-20"><a href="#cb71-20"></a> <span class="dt">type=</span><span class="st">&#39;class&#39;</span></span>
<span id="cb71-21"><a href="#cb71-21"></a>)</span>
<span id="cb71-22"><a href="#cb71-22"></a></span>
<span id="cb71-23"><a href="#cb71-23"></a>test_blended_predClass_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb71-24"><a href="#cb71-24"></a> hcc5hmC_cv_lassoR,</span>
<span id="cb71-25"><a href="#cb71-25"></a> <span class="dt">g=</span><span class="fl">0.5</span>,</span>
<span id="cb71-26"><a href="#cb71-26"></a> <span class="dt">newx=</span>hcc5hmC_test_lcpm_mtx,</span>
<span id="cb71-27"><a href="#cb71-27"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span>,</span>
<span id="cb71-28"><a href="#cb71-28"></a> <span class="dt">type=</span><span class="st">&#39;class&#39;</span></span>
<span id="cb71-29"><a href="#cb71-29"></a>)</span>
<span id="cb71-30"><a href="#cb71-30"></a></span>
<span id="cb71-31"><a href="#cb71-31"></a>misclass_id_vec &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">c</span>(</span>
<span id="cb71-32"><a href="#cb71-32"></a> <span class="kw">names</span>(test_lasso_predClass_vec[,<span class="dv">1</span>])[test_lasso_predClass_vec <span class="op">!=</span><span class="st"> </span>hcc5hmC_test_group_vec],</span>
<span id="cb71-33"><a href="#cb71-33"></a> <span class="kw">names</span>(test_enet_predClass_vec[,<span class="dv">1</span>])[test_enet_predClass_vec <span class="op">!=</span><span class="st"> </span>hcc5hmC_test_group_vec],</span>
<span id="cb71-34"><a href="#cb71-34"></a> <span class="kw">names</span>(test_relaxed_predClass_vec[,<span class="dv">1</span>])[test_relaxed_predClass_vec <span class="op">!=</span><span class="st"> </span>hcc5hmC_test_group_vec],</span>
<span id="cb71-35"><a href="#cb71-35"></a> <span class="kw">names</span>(test_blended_predClass_vec[,<span class="dv">1</span>])[test_blended_predClass_vec <span class="op">!=</span><span class="st"> </span>hcc5hmC_test_group_vec]</span>
<span id="cb71-36"><a href="#cb71-36"></a> )</span>
<span id="cb71-37"><a href="#cb71-37"></a>)</span>
<span id="cb71-38"><a href="#cb71-38"></a></span>
<span id="cb71-39"><a href="#cb71-39"></a></span>
<span id="cb71-40"><a href="#cb71-40"></a>missclass_oofProb_mtx &lt;-<span class="st"> </span><span class="kw">cbind</span>(</span>
<span id="cb71-41"><a href="#cb71-41"></a> test_lasso_predProb_vec[misclass_id_vec,],</span>
<span id="cb71-42"><a href="#cb71-42"></a> test_enet_predProb_vec[misclass_id_vec,],</span>
<span id="cb71-43"><a href="#cb71-43"></a> test_relaxed_predProb_vec[misclass_id_vec,],</span>
<span id="cb71-44"><a href="#cb71-44"></a> test_blended_predProb_vec[misclass_id_vec,]</span>
<span id="cb71-45"><a href="#cb71-45"></a>)</span>
<span id="cb71-46"><a href="#cb71-46"></a><span class="kw">colnames</span>(missclass_oofProb_mtx) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;lasso&#39;</span>,<span class="st">&#39;enet&#39;</span>, <span class="st">&#39;lassoR&#39;</span>, <span class="st">&#39;blended&#39;</span>)</span>
<span id="cb71-47"><a href="#cb71-47"></a></span>
<span id="cb71-48"><a href="#cb71-48"></a>row_med_vec &lt;-<span class="st"> </span><span class="kw">apply</span>(missclass_oofProb_mtx, <span class="dv">1</span>, median)</span>
<span id="cb71-49"><a href="#cb71-49"></a>missclass_oofProb_mtx &lt;-<span class="st"> </span>missclass_oofProb_mtx[</span>
<span id="cb71-50"><a href="#cb71-50"></a>  <span class="kw">order</span>(hcc5hmC_test_group_vec[<span class="kw">rownames</span>(missclass_oofProb_mtx)], row_med_vec),]</span>
<span id="cb71-51"><a href="#cb71-51"></a></span>
<span id="cb71-52"><a href="#cb71-52"></a><span class="kw">plot</span>(</span>
<span id="cb71-53"><a href="#cb71-53"></a> <span class="dt">x=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="kw">nrow</span>(missclass_oofProb_mtx)), <span class="dt">xlab=</span><span class="st">&#39;samples&#39;</span>,</span>
<span id="cb71-54"><a href="#cb71-54"></a> <span class="dt">y=</span><span class="kw">range</span>(missclass_oofProb_mtx), <span class="dt">ylab=</span><span class="st">&#39;out-of-fold predicted probability&#39;</span>,</span>
<span id="cb71-55"><a href="#cb71-55"></a> <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>, <span class="dt">type=</span><span class="st">&#39;n&#39;</span>)</span>
<span id="cb71-56"><a href="#cb71-56"></a></span>
<span id="cb71-57"><a href="#cb71-57"></a><span class="cf">for</span>(RR <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(missclass_oofProb_mtx))</span>
<span id="cb71-58"><a href="#cb71-58"></a><span class="kw">points</span>(</span>
<span id="cb71-59"><a href="#cb71-59"></a> <span class="kw">rep</span>(RR, <span class="kw">ncol</span>(missclass_oofProb_mtx)), </span>
<span id="cb71-60"><a href="#cb71-60"></a> missclass_oofProb_mtx[RR,],</span>
<span id="cb71-61"><a href="#cb71-61"></a> <span class="dt">col=</span><span class="kw">ifelse</span>(hcc5hmC_test_group_vec[<span class="kw">rownames</span>(missclass_oofProb_mtx)[RR]] <span class="op">==</span><span class="st"> &#39;Control&#39;</span>,</span>
<span id="cb71-62"><a href="#cb71-62"></a>  <span class="st">&#39;green&#39;</span>, <span class="st">&#39;red&#39;</span>),</span>
<span id="cb71-63"><a href="#cb71-63"></a> <span class="dt">pch=</span><span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(missclass_oofProb_mtx))</span>
<span id="cb71-64"><a href="#cb71-64"></a></span>
<span id="cb71-65"><a href="#cb71-65"></a><span class="kw">legend</span>(<span class="st">&#39;top&#39;</span>, <span class="dt">ncol=</span><span class="dv">2</span>, <span class="dt">legend=</span><span class="kw">colnames</span>(missclass_oofProb_mtx), </span>
<span id="cb71-66"><a href="#cb71-66"></a> <span class="dt">pch=</span><span class="dv">1</span><span class="op">:</span><span class="dv">4</span>, <span class="dt">bty=</span><span class="st">&#39;n&#39;</span>)</span>
<span id="cb71-67"><a href="#cb71-67"></a></span>
<span id="cb71-68"><a href="#cb71-68"></a><span class="kw">abline</span>(<span class="dt">h=</span><span class="fl">0.5</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetFit-misclassTest-1.png" alt="Test data predicted probabilities at miscassified samples" width="768" />
<p class="caption">
(#fig:hcc5hmC-glmnetFit-misclassTest)Test data predicted probabilities at miscassified samples
</p>
</div>
<p>The relaxed lasso fit results in essentially dichotomized predicted probability
distribution - predicted probabilities are very close to 0 or 1.</p>
<p>We see that for design points in the training set, the predicted probabilities from the relaxed lasso
are essentially dichotomized to be tightly distributed at the extremes of the
response range. For design points in the test set, the predicted probabilities
from the relaxed lasso are comparable to the lasso model predicted probabilities.
This seems to indicate over-fitting in the relaxed lasso fit.</p>
</div>
<div id="compare-coefficient-profiles" class="section level2" number="4.6">
<h2 number="4.6"><span class="header-section-number">4.6</span> Compare coefficient profiles</h2>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1"></a><span class="co"># lasso </span></span>
<span id="cb72-2"><a href="#cb72-2"></a><span class="co">##########################</span></span>
<span id="cb72-3"><a href="#cb72-3"></a><span class="co"># train - cv predicted</span></span>
<span id="cb72-4"><a href="#cb72-4"></a>lasso_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb72-5"><a href="#cb72-5"></a> hcc5hmC_cv_lasso,</span>
<span id="cb72-6"><a href="#cb72-6"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span></span>
<span id="cb72-7"><a href="#cb72-7"></a>)</span>
<span id="cb72-8"><a href="#cb72-8"></a>lasso_coef_frm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb72-9"><a href="#cb72-9"></a> <span class="dt">gene=</span>lasso_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][<span class="kw">c</span>(<span class="dv">1</span>, lasso_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>])],</span>
<span id="cb72-10"><a href="#cb72-10"></a> <span class="dt">lasso=</span>lasso_coef<span class="op">@</span>x)</span>
<span id="cb72-11"><a href="#cb72-11"></a></span>
<span id="cb72-12"><a href="#cb72-12"></a></span>
<span id="cb72-13"><a href="#cb72-13"></a><span class="co"># enet</span></span>
<span id="cb72-14"><a href="#cb72-14"></a><span class="co">##########################</span></span>
<span id="cb72-15"><a href="#cb72-15"></a>enet_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb72-16"><a href="#cb72-16"></a> hcc5hmC_cv_enet,</span>
<span id="cb72-17"><a href="#cb72-17"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span></span>
<span id="cb72-18"><a href="#cb72-18"></a>)</span>
<span id="cb72-19"><a href="#cb72-19"></a>enet_coef_frm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb72-20"><a href="#cb72-20"></a> <span class="dt">gene=</span>enet_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][<span class="kw">c</span>(<span class="dv">1</span>, enet_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>])],</span>
<span id="cb72-21"><a href="#cb72-21"></a> <span class="dt">enet=</span>enet_coef<span class="op">@</span>x)</span>
<span id="cb72-22"><a href="#cb72-22"></a></span>
<span id="cb72-23"><a href="#cb72-23"></a><span class="co"># THESE ARE NOT CORRECT - SKIP</span></span>
<span id="cb72-24"><a href="#cb72-24"></a><span class="co"># relaxed lasso (gamma=0)</span></span>
<span id="cb72-25"><a href="#cb72-25"></a><span class="co">##########################</span></span>
<span id="cb72-26"><a href="#cb72-26"></a>SKIP &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb72-27"><a href="#cb72-27"></a>lassoR_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb72-28"><a href="#cb72-28"></a> hcc5hmC_cv_lassoR,</span>
<span id="cb72-29"><a href="#cb72-29"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span>,</span>
<span id="cb72-30"><a href="#cb72-30"></a> <span class="dt">g=</span><span class="dv">0</span></span>
<span id="cb72-31"><a href="#cb72-31"></a>)</span>
<span id="cb72-32"><a href="#cb72-32"></a>lassoR_coef_frm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb72-33"><a href="#cb72-33"></a> <span class="dt">gene=</span>lassoR_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][<span class="kw">c</span>(<span class="dv">1</span>, lassoR_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>])],</span>
<span id="cb72-34"><a href="#cb72-34"></a> <span class="dt">lassoR=</span>lassoR_coef<span class="op">@</span>x)</span>
<span id="cb72-35"><a href="#cb72-35"></a>}</span>
<span id="cb72-36"><a href="#cb72-36"></a></span>
<span id="cb72-37"><a href="#cb72-37"></a><span class="co"># blended mix (gamma=0.5)</span></span>
<span id="cb72-38"><a href="#cb72-38"></a><span class="co">###############################</span></span>
<span id="cb72-39"><a href="#cb72-39"></a>blended_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb72-40"><a href="#cb72-40"></a> hcc5hmC_cv_lassoR,</span>
<span id="cb72-41"><a href="#cb72-41"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span>,</span>
<span id="cb72-42"><a href="#cb72-42"></a> <span class="dt">g=</span><span class="fl">0.5</span></span>
<span id="cb72-43"><a href="#cb72-43"></a>)</span>
<span id="cb72-44"><a href="#cb72-44"></a>blended_coef_frm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb72-45"><a href="#cb72-45"></a> <span class="dt">gene=</span>blended_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][<span class="kw">c</span>(<span class="dv">1</span>, blended_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>])],</span>
<span id="cb72-46"><a href="#cb72-46"></a> <span class="dt">blended=</span>blended_coef<span class="op">@</span>x)</span>
<span id="cb72-47"><a href="#cb72-47"></a></span>
<span id="cb72-48"><a href="#cb72-48"></a></span>
<span id="cb72-49"><a href="#cb72-49"></a><span class="co"># put it all together</span></span>
<span id="cb72-50"><a href="#cb72-50"></a>all_coef_frm &lt;-<span class="st"> </span></span>
<span id="cb72-51"><a href="#cb72-51"></a><span class="st"> </span>base<span class="op">::</span><span class="kw">merge</span>(</span>
<span id="cb72-52"><a href="#cb72-52"></a> <span class="dt">x =</span> lasso_coef_frm, </span>
<span id="cb72-53"><a href="#cb72-53"></a> <span class="dt">y =</span> base<span class="op">::</span><span class="kw">merge</span>(</span>
<span id="cb72-54"><a href="#cb72-54"></a>     <span class="dt">x =</span> enet_coef_frm,</span>
<span id="cb72-55"><a href="#cb72-55"></a>     <span class="dt">y =</span> blended_coef_frm,</span>
<span id="cb72-56"><a href="#cb72-56"></a>         <span class="dt">by=</span><span class="st">&#39;gene&#39;</span>, <span class="dt">all=</span>T),</span>
<span id="cb72-57"><a href="#cb72-57"></a> <span class="dt">by=</span><span class="st">&#39;gene&#39;</span>, <span class="dt">all=</span>T)</span>
<span id="cb72-58"><a href="#cb72-58"></a></span>
<span id="cb72-59"><a href="#cb72-59"></a><span class="co"># SKIPPED</span></span>
<span id="cb72-60"><a href="#cb72-60"></a><span class="co">#base::merge(</span></span>
<span id="cb72-61"><a href="#cb72-61"></a>         <span class="co">#x = lassoR_coef_frm,</span></span>
<span id="cb72-62"><a href="#cb72-62"></a>         <span class="co">#y = blended_coef_frm,</span></span>
<span id="cb72-63"><a href="#cb72-63"></a>         <span class="co">#by=&#39;gene&#39;, all=T),</span></span>
<span id="cb72-64"><a href="#cb72-64"></a></span>
<span id="cb72-65"><a href="#cb72-65"></a><span class="co"># save gene name into rownames</span></span>
<span id="cb72-66"><a href="#cb72-66"></a><span class="kw">rownames</span>(all_coef_frm) &lt;-<span class="st"> </span>all_coef_frm<span class="op">$</span>gene</span>
<span id="cb72-67"><a href="#cb72-67"></a>all_coef_frm  <span class="op">%&lt;&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(<span class="op">-</span>gene)</span>
<span id="cb72-68"><a href="#cb72-68"></a></span>
<span id="cb72-69"><a href="#cb72-69"></a><span class="co"># set na to 0</span></span>
<span id="cb72-70"><a href="#cb72-70"></a>all_coef_frm[<span class="kw">is.na</span>(all_coef_frm)] &lt;-<span class="st"> </span><span class="dv">0</span></span>
<span id="cb72-71"><a href="#cb72-71"></a></span>
<span id="cb72-72"><a href="#cb72-72"></a><span class="co"># remove intercept</span></span>
<span id="cb72-73"><a href="#cb72-73"></a>all_coef_frm &lt;-<span class="st"> </span>all_coef_frm[<span class="op">!</span><span class="kw">rownames</span>(all_coef_frm)<span class="op">==</span><span class="st">&#39;(Intercept)&#39;</span>,]</span>
<span id="cb72-74"><a href="#cb72-74"></a></span>
<span id="cb72-75"><a href="#cb72-75"></a><span class="co"># order rows by size</span></span>
<span id="cb72-76"><a href="#cb72-76"></a>all_coef_frm &lt;-<span class="st"> </span>all_coef_frm[<span class="kw">order</span>(<span class="kw">rowMeans</span>(<span class="kw">abs</span>(all_coef_frm)), <span class="dt">decreasing=</span>T), ]</span>
<span id="cb72-77"><a href="#cb72-77"></a></span>
<span id="cb72-78"><a href="#cb72-78"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="kw">ncol</span>(all_coef_frm),<span class="dv">1</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">5</span>,<span class="fl">0.5</span>,<span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb72-79"><a href="#cb72-79"></a></span>
<span id="cb72-80"><a href="#cb72-80"></a><span class="cf">for</span>(CC <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(all_coef_frm)) {</span>
<span id="cb72-81"><a href="#cb72-81"></a> <span class="kw">plot</span>(</span>
<span id="cb72-82"><a href="#cb72-82"></a>  <span class="dt">x=</span><span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(all_coef_frm), <span class="dt">xlab=</span><span class="st">&#39;&#39;</span>,</span>
<span id="cb72-83"><a href="#cb72-83"></a>  <span class="dt">y=</span>all_coef_frm[, CC], <span class="dt">ylab=</span><span class="kw">colnames</span>(all_coef_frm)[CC],</span>
<span id="cb72-84"><a href="#cb72-84"></a>  <span class="dt">type=</span><span class="st">&#39;h&#39;</span>, <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>)</span>
<span id="cb72-85"><a href="#cb72-85"></a>}</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetFit-compCoeffProf-1.png" alt="Coefficient Profiles" width="768" />
<p class="caption">
(#fig:hcc5hmC-glmnetFit-compCoeffProf)Coefficient Profiles
</p>
</div>
<p>Note that there is little difference between the elastic net and the lasso
in the selected features, and when the coefficient is zero in one set, it
is small in the other. By contrast, the blended fit produces more shrinkage.</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1"></a>knitr<span class="op">::</span><span class="kw">kable</span>(</span>
<span id="cb73-2"><a href="#cb73-2"></a><span class="kw">with</span>(all_coef_frm, <span class="kw">table</span>(<span class="dt">lassoZero=</span>lasso<span class="op">==</span><span class="dv">0</span>, <span class="dt">enetZero=</span>enet<span class="op">==</span><span class="dv">0</span>)),</span>
<span id="cb73-3"><a href="#cb73-3"></a> <span class="dt">caption=</span><span class="st">&#39;Zero Ceofficient: rows are lasso, columns enet&#39;</span>) <span class="op">%&gt;%</span></span>
<span id="cb73-4"><a href="#cb73-4"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-glmnetFit-zreros)Zero Ceofficient: rows are lasso, columns enet
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
FALSE
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
68
</td>
</tr>
<tr>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
92
</td>
</tr>
</tbody>
</table>
<!--
Coefficients in the relaxed lasso fit are much larger than those in the
lasso fit, or zero.  As a consequence, the blended fit coefficients look 
like a shrunken version of the relaxed lasso fit coefficients.  
-->
<p>Coefficients in the blended fit are larger than those in the
lasso fit, or zero.</p>
<p>We can also examine these with a scatter plot matrix.</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1"></a><span class="kw">pairs</span>(all_coef_frm,</span>
<span id="cb74-2"><a href="#cb74-2"></a>  <span class="dt">lower.panel =</span> <span class="ot">NULL</span>,</span>
<span id="cb74-3"><a href="#cb74-3"></a>  <span class="dt">panel =</span> <span class="cf">function</span>(x, y) {</span>
<span id="cb74-4"><a href="#cb74-4"></a>    <span class="kw">points</span>(x, y, <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">col =</span> <span class="st">&quot;blue&quot;</span>)</span>
<span id="cb74-5"><a href="#cb74-5"></a>  }</span>
<span id="cb74-6"><a href="#cb74-6"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetFit-pairsCoeffProf-1.png" alt="Coefficients from fits" width="768" />
<p class="caption">
(#fig:hcc5hmC-glmnetFit-pairsCoeffProf)Coefficients from fits
</p>
</div>
</div>
<div id="examine-feature-selection" class="section level2" number="4.7">
<h2 number="4.7"><span class="header-section-number">4.7</span> Examine feature selection</h2>
<p>Recall from <code>glmnet</code> vignette:</p>
<pre><code>It is known that the ridge penalty shrinks the coefficients of correlated predictors
towards each other while the lasso tends to pick one of them and discard the others.
The elastic-net penalty mixes these two; if predictors are correlated in groups,
an $\alpha$=0.5 tends to select the groups in or out together.
This is a higher level parameter, and users might pick a value upfront,
else experiment with a few different values. One use of $\alpha$ is for numerical stability;
for example, the *elastic net with $\alpha = 1 - \epsilon$ for some small $\epsilon$&gt;0
performs much like the lasso, but removes any degeneracies and wild behavior caused
by extreme correlations*.</code></pre>
<p>To see how this plays out in this dataset, we can look at feature expression
heat maps.</p>
<p>Reader notes:</p>
<pre><code>Heat maps are rarely useful other than to display the obvious.
Here too heat maps fail to yield any insights, or confirmation
of the relationship between feature correlation and lasso vs enet
feature selection.</code></pre>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1"></a> <span class="kw">suppressPackageStartupMessages</span>(<span class="kw">require</span>(gplots))</span>
<span id="cb77-2"><a href="#cb77-2"></a></span>
<span id="cb77-3"><a href="#cb77-3"></a><span class="co"># train - cv predicted</span></span>
<span id="cb77-4"><a href="#cb77-4"></a>lasso_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb77-5"><a href="#cb77-5"></a> hcc5hmC_cv_lasso,</span>
<span id="cb77-6"><a href="#cb77-6"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span></span>
<span id="cb77-7"><a href="#cb77-7"></a>)</span>
<span id="cb77-8"><a href="#cb77-8"></a>lasso_coef_frm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb77-9"><a href="#cb77-9"></a> <span class="dt">gene=</span>lasso_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][<span class="kw">c</span>(<span class="dv">1</span>, lasso_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>])],</span>
<span id="cb77-10"><a href="#cb77-10"></a> <span class="dt">lasso=</span>lasso_coef<span class="op">@</span>x)</span>
<span id="cb77-11"><a href="#cb77-11"></a></span>
<span id="cb77-12"><a href="#cb77-12"></a> </span>
<span id="cb77-13"><a href="#cb77-13"></a>  Mycol &lt;-<span class="st"> </span><span class="kw">colorpanel</span>(<span class="dv">1000</span>, <span class="st">&quot;blue&quot;</span>, <span class="st">&quot;red&quot;</span>)</span>
<span id="cb77-14"><a href="#cb77-14"></a>  <span class="kw">heatmap.2</span>(</span>
<span id="cb77-15"><a href="#cb77-15"></a>    <span class="dt">x=</span><span class="kw">t</span>(hcc5hmC_train_lcpm_mtx[,lasso_coef_frm<span class="op">$</span>gene[<span class="op">-</span><span class="dv">1</span>]]),</span>
<span id="cb77-16"><a href="#cb77-16"></a>    <span class="dt">scale=</span><span class="st">&quot;row&quot;</span>,</span>
<span id="cb77-17"><a href="#cb77-17"></a>    <span class="dt">labRow=</span>lasso_coef_frm<span class="op">$</span>gene,</span>
<span id="cb77-18"><a href="#cb77-18"></a>    <span class="dt">labCol=</span>hcc5hmC_train_group_vec,</span>
<span id="cb77-19"><a href="#cb77-19"></a>    <span class="dt">col=</span>Mycol, </span>
<span id="cb77-20"><a href="#cb77-20"></a>    <span class="dt">trace=</span><span class="st">&quot;none&quot;</span>, <span class="dt">density.info=</span><span class="st">&quot;none&quot;</span>, </span>
<span id="cb77-21"><a href="#cb77-21"></a>    <span class="co">#margin=c(8,6), lhei=c(2,10), </span></span>
<span id="cb77-22"><a href="#cb77-22"></a>    <span class="co">#lwid=c(0.1,4), #lhei=c(0.1,4)</span></span>
<span id="cb77-23"><a href="#cb77-23"></a>    <span class="dt">key=</span>F,</span>
<span id="cb77-24"><a href="#cb77-24"></a>    <span class="dt">ColSideColors=</span><span class="kw">ifelse</span>(hcc5hmC_train_group_vec<span class="op">==</span><span class="st">&#39;Control&#39;</span>, <span class="st">&#39;green&#39;</span>,<span class="st">&#39;red&#39;</span>),</span>
<span id="cb77-25"><a href="#cb77-25"></a>    <span class="dt">dendrogram=</span><span class="st">&quot;both&quot;</span>,</span>
<span id="cb77-26"><a href="#cb77-26"></a>    <span class="dt">main=</span><span class="kw">paste</span>(<span class="st">&#39;lasso genes - N =&#39;</span>, <span class="kw">nrow</span>(lasso_coef_frm)<span class="op">-</span><span class="dv">1</span>))</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetFit-heatmapLasso-1.png" alt="Lasso Model Genes" width="768" />
<p class="caption">
(#fig:hcc5hmC-glmnetFit-heatmapLasso)Lasso Model Genes
</p>
</div>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1"></a> <span class="kw">suppressPackageStartupMessages</span>(<span class="kw">require</span>(gplots))</span>
<span id="cb78-2"><a href="#cb78-2"></a></span>
<span id="cb78-3"><a href="#cb78-3"></a><span class="co"># train - cv predicted</span></span>
<span id="cb78-4"><a href="#cb78-4"></a>enet_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb78-5"><a href="#cb78-5"></a> hcc5hmC_cv_enet,</span>
<span id="cb78-6"><a href="#cb78-6"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span></span>
<span id="cb78-7"><a href="#cb78-7"></a>)</span>
<span id="cb78-8"><a href="#cb78-8"></a>enet_coef_frm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb78-9"><a href="#cb78-9"></a> <span class="dt">gene=</span>enet_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][<span class="kw">c</span>(<span class="dv">1</span>, enet_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>])],</span>
<span id="cb78-10"><a href="#cb78-10"></a> <span class="dt">enet=</span>enet_coef<span class="op">@</span>x)</span>
<span id="cb78-11"><a href="#cb78-11"></a></span>
<span id="cb78-12"><a href="#cb78-12"></a> </span>
<span id="cb78-13"><a href="#cb78-13"></a>  Mycol &lt;-<span class="st"> </span><span class="kw">colorpanel</span>(<span class="dv">1000</span>, <span class="st">&quot;blue&quot;</span>, <span class="st">&quot;red&quot;</span>)</span>
<span id="cb78-14"><a href="#cb78-14"></a>  <span class="kw">heatmap.2</span>(</span>
<span id="cb78-15"><a href="#cb78-15"></a>    <span class="dt">x=</span><span class="kw">t</span>(hcc5hmC_train_lcpm_mtx[,enet_coef_frm<span class="op">$</span>gene[<span class="op">-</span><span class="dv">1</span>]]),</span>
<span id="cb78-16"><a href="#cb78-16"></a>    <span class="dt">scale=</span><span class="st">&quot;row&quot;</span>,</span>
<span id="cb78-17"><a href="#cb78-17"></a>    <span class="dt">labRow=</span>enet_coef_frm<span class="op">$</span>gene,</span>
<span id="cb78-18"><a href="#cb78-18"></a>    <span class="dt">labCol=</span>hcc5hmC_train_group_vec,</span>
<span id="cb78-19"><a href="#cb78-19"></a>    <span class="dt">col=</span>Mycol, </span>
<span id="cb78-20"><a href="#cb78-20"></a>    <span class="dt">trace=</span><span class="st">&quot;none&quot;</span>, <span class="dt">density.info=</span><span class="st">&quot;none&quot;</span>, </span>
<span id="cb78-21"><a href="#cb78-21"></a>    <span class="co">#margin=c(8,6), lhei=c(2,10), </span></span>
<span id="cb78-22"><a href="#cb78-22"></a>    <span class="co">#lwid=c(0.1,4), #lhei=c(0.1,4)</span></span>
<span id="cb78-23"><a href="#cb78-23"></a>    <span class="dt">key=</span>F,</span>
<span id="cb78-24"><a href="#cb78-24"></a>    <span class="dt">ColSideColors=</span><span class="kw">ifelse</span>(hcc5hmC_train_group_vec<span class="op">==</span><span class="st">&#39;Control&#39;</span>, <span class="st">&#39;green&#39;</span>,<span class="st">&#39;red&#39;</span>),</span>
<span id="cb78-25"><a href="#cb78-25"></a>    <span class="dt">dendrogram=</span><span class="st">&quot;both&quot;</span>,</span>
<span id="cb78-26"><a href="#cb78-26"></a>    <span class="dt">main=</span><span class="kw">paste</span>(<span class="st">&#39;enet genes - N =&#39;</span>, <span class="kw">nrow</span>(enet_coef_frm)<span class="op">-</span><span class="dv">1</span>))</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetFit-heatmapEnet-1.png" alt="Enet Model Genes" width="768" />
<p class="caption">
(#fig:hcc5hmC-glmnetFit-heatmapEnet)Enet Model Genes
</p>
</div>
<!--chapter:end:04-glmnetFitsHCC5hmC.Rmd-->
</div>
</div>
<div id="hcc-5hmcseq-model-suite" class="section level1" number="5">
<h1 number="5"><span class="header-section-number">5</span> HCC 5hmC-Seq: Sample size investigation</h1>
<p>We now examine the results of fitting a suite of lasso models to
investigate the effect of sample size on
various aspects of model performance:</p>
<ul>
<li><p>assessed accuracy: out-of-fold estimates of precision and variability vs train set estimates</p></li>
<li><p>selected feature profile stability - to what extent does the
feature set implicitly selected by the lasso vary across random
sampling and what is the effect of sample size.</p></li>
</ul>
<p>It is hypothesized that below a certain threshold,
sample sizes are too small to provide reliable estimates
of performance or stable selected feature profiles.</p>
<p><strong>Reader Note</strong></p>
<pre><code>It&#39;s not clear how to display the effect of sample set composition
on model performance, or that it merits special attention as it
may be obvious to anyone who is paying attention that not all
sample points are created equal and that having a substantial
number of mislabeled or otherwise hard to classify samples
can throw off estimates of the separability of subgroups
when sample sizes are small.  

Keep this stuff on consistency for now.</code></pre>
<p>In examining the relationship between sample size and
model performance, including variability and reliability
of performance indicators derived from fits, we should bear in
mind that in addition to sample size, sample composition might
also play a factor. If a training data set contains many outliers
or otherwise <em>hard to properly classify</em> samples, the performance
of models fitted to such a data set is expected to be negatively impacted.</p>
<p>In the simulations that we run below we will track sample <em>consistency</em>
to examine its impact on fitted model performance. We use
<em>consistency</em> to refer to a measure of how hard it is to
appropriately classify individual samples.
Predicted probabilities from fitted model can be transformed into sample
<em>consistency</em> scores: <span class="math inline">\(Q_i = p_i^{y_i}(1-p_i)^{1-y_i}\)</span>, where <span class="math inline">\(p_i\)</span> is the
estimated probability of HCC for sample i and <span class="math inline">\(y_i\)</span> is 1 for HCC samples and
0 for Controls. ie. we use the fitted sample contribution to the
likelihood function as a sample <em>consistency</em> score,
which is <span class="math inline">\(P(HCC)\)</span>, the predicted probability of HCC for HCC samples
and <span class="math inline">\(1 - P(HCC)\)</span> for Controls. To derive the sample <em>consistency</em> scores,
we will use the predicted response from a lasso model fitted to the entire data set.</p>
<p>Hard to classify samples will have low <em>consistency</em> scores.
In the results that we discuss below, when we look at variability across repeated
random sampling of different sizes, we can use sample <em>consistency</em> scores to investigate
how much of the variability is due to sample selection.
Note that <em>consistency</em> here is not used to say anything about the sample data quality.
Low <em>consistency</em> here only means that a sample is different from the
core of the data set in a way that makes it hard to properly classify.
That could happen if the sample were mislabeled, in which case we could
think of this sample as being poor quality of course.</p>
<p>The variability in sample set <em>consistency</em> that we get from simple random sampling,
as is done in the simulation below, is not expected to adequately reflect
sample set <em>consistency</em> variation encountered in practice when data for a particular
study are accumulated over extensive time horizons and across varied
physical settings. In order to accrue study samples at an acceptable rate,
it is not uncommon for the study sponsor to work with several clinics, medical centers,
or other tissue or blood sample provider.
Or the sponsor may sub-contract the sample acquisition
task to a third party who may have suppliers distributed across the globe. This
is great news for the rate of sample acquisition, but not such great news for
the uniformity of consistency of samples. In such a context, the variability
in sample set consistency as the data set grows over time would not
be adequately captured by simple random sampling variability;
the variability would be more akin to cluster sampling with potential confounding
batch effects. The impact that these effects can have on the classification analysis
results cannot be understated, especially in the context of a new technology that
is exploiting biological processes that are still not fully understood. All this
to make the point that the variability the we are studying here has to be regarded
as a lower bound on the variability that is to be expected in practice, and that without
an external validation data set to verify results, one should be cautious
when interpreting empirical findings, especially in the absence of solid
biological underpinnings.</p>
<p><strong>Reader Note</strong></p>
<pre><code>We still need to demonstrate the effect of sample set composition
on results.  With random sampling the effects are subtle and not
immediately obvious.  We may have to fabricate sample sets with 
low consistency scores - ie. made up of many hard to classify samples -
in order to show this impact.</code></pre>
<div id="full-data-set-fit" class="section level2" number="5.1">
<h2 number="5.1"><span class="header-section-number">5.1</span> Full data set fit</h2>
<p>We begin by fitting a model to the entire data set in order to:</p>
<ul>
<li><p>obtain a baseline classification performance against which to judge the performance
obtained from the fits to smaller sample sets,</p></li>
<li><p>obtain sample consistency scores which can be used to explain variability
in the performance of model fitted to data sets of a fixed size, and</p></li>
<li><p>produce a <em>full model</em> gene signature which can be used to evaluate
the stability of selected features in models fitted to data sets of different
sizes.</p></li>
</ul>
<p>First assemble the data set. This entails simply re-combining the
train and test data.</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1"></a><span class="co"># combine train and test </span></span>
<span id="cb81-2"><a href="#cb81-2"></a>hcc5hmC_all_lcpm_mtx &lt;-<span class="st"> </span><span class="kw">rbind</span>(hcc5hmC_train_lcpm_mtx, hcc5hmC_test_lcpm_mtx)</span>
<span id="cb81-3"><a href="#cb81-3"></a></span>
<span id="cb81-4"><a href="#cb81-4"></a><span class="co"># we have to be careful with factors!</span></span>
<span id="cb81-5"><a href="#cb81-5"></a><span class="co"># We&#39;ll keep as a character and change to factor when needed</span></span>
<span id="cb81-6"><a href="#cb81-6"></a>hcc5hmC_all_group_vec &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb81-7"><a href="#cb81-7"></a> <span class="kw">as.character</span>(hcc5hmC_train_group_vec), </span>
<span id="cb81-8"><a href="#cb81-8"></a> <span class="kw">as.character</span>(hcc5hmC_test_group_vec)</span>
<span id="cb81-9"><a href="#cb81-9"></a>)</span>
<span id="cb81-10"><a href="#cb81-10"></a><span class="co"># I suspect adding names to vectors breaks one of the tidy commandments,</span></span>
<span id="cb81-11"><a href="#cb81-11"></a><span class="co"># but then again I am sure I have already offended the creed beyond salvation</span></span>
<span id="cb81-12"><a href="#cb81-12"></a><span class="kw">names</span>(hcc5hmC_all_group_vec) &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb81-13"><a href="#cb81-13"></a> <span class="kw">names</span>(hcc5hmC_train_group_vec),</span>
<span id="cb81-14"><a href="#cb81-14"></a> <span class="kw">names</span>(hcc5hmC_test_group_vec)</span>
<span id="cb81-15"><a href="#cb81-15"></a>)</span>
<span id="cb81-16"><a href="#cb81-16"></a></span>
<span id="cb81-17"><a href="#cb81-17"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">table</span>(<span class="dt">group =</span> hcc5hmC_all_group_vec),</span>
<span id="cb81-18"><a href="#cb81-18"></a>  <span class="dt">caption =</span> <span class="st">&quot;samples by group&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb81-19"><a href="#cb81-19"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-glmnetSuite-get-all-data)samples by group
</caption>
<thead>
<tr>
<th style="text-align:left;">
group
</th>
<th style="text-align:right;">
Freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Control
</td>
<td style="text-align:right;">
778
</td>
</tr>
<tr>
<td style="text-align:left;">
HCC
</td>
<td style="text-align:right;">
555
</td>
</tr>
</tbody>
</table>
<p>Now fit the lasso model through cross-validation.
Note that the results of a cv fit are random due to the
random allocation of samples to folds. We can reduce this
variability by properly averaging results over repeated cv fits.
Here we will obtain sample consistency scores by averaging results
over 30 cv runs.</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1"></a><span class="kw">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb82-2"><a href="#cb82-2"></a></span>
<span id="cb82-3"><a href="#cb82-3"></a>start_time &lt;-<span class="st">  </span><span class="kw">proc.time</span>()</span>
<span id="cb82-4"><a href="#cb82-4"></a></span>
<span id="cb82-5"><a href="#cb82-5"></a>hcc5hmC_cv_lassoAll_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">30</span>, <span class="cf">function</span>(REP) {</span>
<span id="cb82-6"><a href="#cb82-6"></a>glmnet<span class="op">::</span><span class="kw">cv.glmnet</span>(</span>
<span id="cb82-7"><a href="#cb82-7"></a> <span class="dt">x =</span> hcc5hmC_all_lcpm_mtx,</span>
<span id="cb82-8"><a href="#cb82-8"></a> <span class="dt">y =</span> <span class="kw">factor</span>(hcc5hmC_all_group_vec,<span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&#39;Control&#39;</span>, <span class="st">&#39;HCC&#39;</span>)),</span>
<span id="cb82-9"><a href="#cb82-9"></a> <span class="dt">alpha =</span> <span class="dv">1</span>,</span>
<span id="cb82-10"><a href="#cb82-10"></a> <span class="dt">family =</span> <span class="st">&#39;binomial&#39;</span>,</span>
<span id="cb82-11"><a href="#cb82-11"></a> <span class="dt">type.measure  =</span>  <span class="st">&quot;class&quot;</span>,</span>
<span id="cb82-12"><a href="#cb82-12"></a> <span class="dt">keep =</span> T,</span>
<span id="cb82-13"><a href="#cb82-13"></a> <span class="dt">nlambda =</span> <span class="dv">100</span></span>
<span id="cb82-14"><a href="#cb82-14"></a>)</span>
<span id="cb82-15"><a href="#cb82-15"></a>}</span>
<span id="cb82-16"><a href="#cb82-16"></a>)</span>
<span id="cb82-17"><a href="#cb82-17"></a></span>
<span id="cb82-18"><a href="#cb82-18"></a><span class="kw">message</span>(<span class="st">&quot;lassoAll time: &quot;</span>, <span class="kw">round</span>((<span class="kw">proc.time</span>() <span class="op">-</span><span class="st"> </span>start_time)[<span class="dv">3</span>],<span class="dv">2</span>),<span class="st">&quot;s&quot;</span>)</span></code></pre></div>
<!-- lasso-fit-all takes a while - save results -->
<!-- DO THIS ONCE -->
<p>Examine the fits.</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb83-2"><a href="#cb83-2"></a><span class="kw">plot</span>(</span>
<span id="cb83-3"><a href="#cb83-3"></a> <span class="kw">log</span>(hcc5hmC_cv_lassoAll_lst[[<span class="dv">1</span>]]<span class="op">$</span>lambda),</span>
<span id="cb83-4"><a href="#cb83-4"></a> hcc5hmC_cv_lassoAll_lst[[<span class="dv">1</span>]]<span class="op">$</span>cvm,</span>
<span id="cb83-5"><a href="#cb83-5"></a> <span class="dt">lwd=</span><span class="dv">2</span>,</span>
<span id="cb83-6"><a href="#cb83-6"></a> <span class="dt">xlab=</span><span class="st">&#39;log(Lambda)&#39;</span>, <span class="dt">ylab=</span><span class="st">&#39;CV Misclassification Error&#39;</span>, <span class="dt">type=</span><span class="st">&#39;l&#39;</span>, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">.5</span>)</span>
<span id="cb83-7"><a href="#cb83-7"></a>)</span>
<span id="cb83-8"><a href="#cb83-8"></a></span>
<span id="cb83-9"><a href="#cb83-9"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">2</span><span class="op">:</span><span class="kw">length</span>(hcc5hmC_cv_lassoAll_lst))</span>
<span id="cb83-10"><a href="#cb83-10"></a> <span class="kw">lines</span>(</span>
<span id="cb83-11"><a href="#cb83-11"></a>  <span class="kw">log</span>(hcc5hmC_cv_lassoAll_lst[[JJ]]<span class="op">$</span>lambda),</span>
<span id="cb83-12"><a href="#cb83-12"></a>  hcc5hmC_cv_lassoAll_lst[[JJ]]<span class="op">$</span>cvm,</span>
<span id="cb83-13"><a href="#cb83-13"></a>  <span class="dt">lwd=</span><span class="dv">2</span></span>
<span id="cb83-14"><a href="#cb83-14"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuite-plot-lassoAll-1.png" alt="Repeated cv lasso models fitted to all samples" width="576" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuite-plot-lassoAll)Repeated cv lasso models fitted to all samples
</p>
</div>
<p>These cv curves are remarkably consistent meaning that the determination of the size or sparsity
of the model fitted through cross validation to the full data set is fairly precise:</p>
<!-- DONT CACHE THIS ??? -->
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1"></a><span class="kw">library</span>(magrittr)</span>
<span id="cb84-2"><a href="#cb84-2"></a></span>
<span id="cb84-3"><a href="#cb84-3"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb84-4"><a href="#cb84-4"></a></span>
<span id="cb84-5"><a href="#cb84-5"></a><span class="co"># nzero</span></span>
<span id="cb84-6"><a href="#cb84-6"></a>nzero_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_lassoAll_lst,</span>
<span id="cb84-7"><a href="#cb84-7"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb84-8"><a href="#cb84-8"></a></span>
<span id="cb84-9"><a href="#cb84-9"></a>nzero_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_lassoAll_lst,</span>
<span id="cb84-10"><a href="#cb84-10"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb84-11"><a href="#cb84-11"></a></span>
<span id="cb84-12"><a href="#cb84-12"></a><span class="kw">boxplot</span>(<span class="kw">list</span>(<span class="st">`</span><span class="dt">1se</span><span class="st">`</span>=nzero_1se_vec, <span class="dt">min =</span> nzero_min_vec), <span class="dt">ylab=</span><span class="st">&quot;Selected Features&quot;</span>)</span>
<span id="cb84-13"><a href="#cb84-13"></a></span>
<span id="cb84-14"><a href="#cb84-14"></a><span class="co"># error</span></span>
<span id="cb84-15"><a href="#cb84-15"></a>error_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_lassoAll_lst,</span>
<span id="cb84-16"><a href="#cb84-16"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb84-17"><a href="#cb84-17"></a></span>
<span id="cb84-18"><a href="#cb84-18"></a>error_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_lassoAll_lst,</span>
<span id="cb84-19"><a href="#cb84-19"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb84-20"><a href="#cb84-20"></a></span>
<span id="cb84-21"><a href="#cb84-21"></a><span class="kw">boxplot</span>(</span>
<span id="cb84-22"><a href="#cb84-22"></a> <span class="kw">list</span>(<span class="st">`</span><span class="dt">1se</span><span class="st">`</span>=error_1se_vec, <span class="dt">min =</span> error_min_vec), </span>
<span id="cb84-23"><a href="#cb84-23"></a> <span class="dt">ylab=</span>hcc5hmC_cv_lassoAll_lst[[<span class="dv">1</span>]]<span class="op">$</span>name,</span>
<span id="cb84-24"><a href="#cb84-24"></a> <span class="dt">ylim=</span><span class="kw">c</span>(<span class="fl">0.06</span>, <span class="fl">.10</span>)</span>
<span id="cb84-25"><a href="#cb84-25"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuite-model-size-lassoAll-1.png" alt="Feature selection and estimated error by repeated cv lasso models" width="768" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuite-model-size-lassoAll)Feature selection and estimated error by repeated cv lasso models
</p>
</div>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1"></a><span class="co"># tabular format</span></span>
<span id="cb85-2"><a href="#cb85-2"></a>tmp &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">rbind</span>(</span>
<span id="cb85-3"><a href="#cb85-3"></a> <span class="st">`</span><span class="dt">features_1se</span><span class="st">`</span> =<span class="st"> </span><span class="kw">summary</span>(nzero_1se_vec),</span>
<span id="cb85-4"><a href="#cb85-4"></a> <span class="dt">features_min =</span> <span class="kw">summary</span>(nzero_min_vec),</span>
<span id="cb85-5"><a href="#cb85-5"></a> <span class="st">`</span><span class="dt">features:min-1se</span><span class="st">`</span> =<span class="st"> </span><span class="kw">summary</span>(nzero_min_vec <span class="op">-</span><span class="st"> </span>nzero_1se_vec),</span>
<span id="cb85-6"><a href="#cb85-6"></a> <span class="st">`</span><span class="dt">cv_error_1se</span><span class="st">`</span> =<span class="st"> </span><span class="kw">summary</span>(<span class="dv">100</span><span class="op">*</span>error_1se_vec),</span>
<span id="cb85-7"><a href="#cb85-7"></a> <span class="dt">cv_error_min =</span> <span class="kw">summary</span>(<span class="dv">100</span><span class="op">*</span>error_min_vec),</span>
<span id="cb85-8"><a href="#cb85-8"></a> <span class="st">`</span><span class="dt">cv_error:1se-min</span><span class="st">`</span> =<span class="st"> </span><span class="kw">summary</span>(<span class="dv">100</span><span class="op">*</span>(error_1se_vec<span class="op">-</span>error_min_vec))</span>
<span id="cb85-9"><a href="#cb85-9"></a>))</span>
<span id="cb85-10"><a href="#cb85-10"></a></span>
<span id="cb85-11"><a href="#cb85-11"></a>knitr<span class="op">::</span><span class="kw">kable</span>(tmp <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(<span class="op">-</span>Mean),</span>
<span id="cb85-12"><a href="#cb85-12"></a>  <span class="dt">caption =</span> <span class="st">&quot;Number of selected features&quot;</span>,</span>
<span id="cb85-13"><a href="#cb85-13"></a>  <span class="dt">digits=</span><span class="dv">1</span>) <span class="op">%&gt;%</span></span>
<span id="cb85-14"><a href="#cb85-14"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-glmnetSuite-model-size-lassoAll)Number of selected features
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Min.
</th>
<th style="text-align:right;">
X1st.Qu.
</th>
<th style="text-align:right;">
Median
</th>
<th style="text-align:right;">
X3rd.Qu.
</th>
<th style="text-align:right;">
Max.
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
features_1se
</td>
<td style="text-align:right;">
54.0
</td>
<td style="text-align:right;">
102.0
</td>
<td style="text-align:right;">
108.0
</td>
<td style="text-align:right;">
132.5
</td>
<td style="text-align:right;">
194.0
</td>
</tr>
<tr>
<td style="text-align:left;">
features_min
</td>
<td style="text-align:right;">
118.0
</td>
<td style="text-align:right;">
175.2
</td>
<td style="text-align:right;">
200.0
</td>
<td style="text-align:right;">
261.0
</td>
<td style="text-align:right;">
365.0
</td>
</tr>
<tr>
<td style="text-align:left;">
features:min-1se
</td>
<td style="text-align:right;">
10.0
</td>
<td style="text-align:right;">
55.0
</td>
<td style="text-align:right;">
86.0
</td>
<td style="text-align:right;">
141.5
</td>
<td style="text-align:right;">
247.0
</td>
</tr>
<tr>
<td style="text-align:left;">
cv_error_1se
</td>
<td style="text-align:right;">
6.9
</td>
<td style="text-align:right;">
7.6
</td>
<td style="text-align:right;">
7.7
</td>
<td style="text-align:right;">
8.1
</td>
<td style="text-align:right;">
8.9
</td>
</tr>
<tr>
<td style="text-align:left;">
cv_error_min
</td>
<td style="text-align:right;">
6.7
</td>
<td style="text-align:right;">
7.1
</td>
<td style="text-align:right;">
7.1
</td>
<td style="text-align:right;">
7.3
</td>
<td style="text-align:right;">
8.0
</td>
</tr>
<tr>
<td style="text-align:left;">
cv_error:1se-min
</td>
<td style="text-align:right;">
0.2
</td>
<td style="text-align:right;">
0.5
</td>
<td style="text-align:right;">
0.6
</td>
<td style="text-align:right;">
0.7
</td>
<td style="text-align:right;">
1.0
</td>
</tr>
</tbody>
</table>
<p>The number of features selected by the minimum lambda models are larger
than the number selected by the “one standard error” rule models by a median
of 86.
The cv error rates obtained from the minimum lambda models are lower
then “one standard error” rule models error rates by a median of
0.6%.</p>
<p>The cv error rates observed in this set are comparable to the
rates observed in the lasso models fitted to the training sample set
which consisted of 80% of the samples in this set. In other words,
there is no obvious gain in performance in moving from
a data set with
622 vs 444 samples
to a data set with
778 vs 555 samples.
See Table @ref(tab:printErrors).</p>
<p>It’s not clear at this point whether the minimum lambda model is truly better than
the “one standard error” rule model. We would need and external validation
set to make this determination. We can compare the two sets
of out-of-fold predicted values, averaged across cv replicates, to see if
there is a meaningful difference between the two.</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1"></a><span class="co"># predicted probs - 1se</span></span>
<span id="cb86-2"><a href="#cb86-2"></a>lassoAll_predResp_1se_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_lassoAll_lst, <span class="cf">function</span>(cv_fit) { </span>
<span id="cb86-3"><a href="#cb86-3"></a>  ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se,cv_fit<span class="op">$</span>lambda)</span>
<span id="cb86-4"><a href="#cb86-4"></a>  <span class="kw">logistic_f</span>(cv_fit<span class="op">$</span>fit.preval[,ndx_1se])</span>
<span id="cb86-5"><a href="#cb86-5"></a> })</span>
<span id="cb86-6"><a href="#cb86-6"></a>lassoAll_predResp_1se_vec &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(lassoAll_predResp_1se_mtx)</span>
<span id="cb86-7"><a href="#cb86-7"></a></span>
<span id="cb86-8"><a href="#cb86-8"></a><span class="co"># predicted probs - min</span></span>
<span id="cb86-9"><a href="#cb86-9"></a>lassoAll_predResp_min_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_lassoAll_lst, <span class="cf">function</span>(cv_fit) { </span>
<span id="cb86-10"><a href="#cb86-10"></a>  ndx_min &lt;-<span class="st"> </span><span class="kw">match</span>(cv_fit<span class="op">$</span>lambda.min,cv_fit<span class="op">$</span>lambda)</span>
<span id="cb86-11"><a href="#cb86-11"></a>  <span class="kw">logistic_f</span>(cv_fit<span class="op">$</span>fit.preval[,ndx_min])</span>
<span id="cb86-12"><a href="#cb86-12"></a> })</span>
<span id="cb86-13"><a href="#cb86-13"></a>lassoAll_predResp_min_vec &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(lassoAll_predResp_min_mtx)</span>
<span id="cb86-14"><a href="#cb86-14"></a></span>
<span id="cb86-15"><a href="#cb86-15"></a><span class="co"># plot</span></span>
<span id="cb86-16"><a href="#cb86-16"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">5</span>,<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb86-17"><a href="#cb86-17"></a>tmp &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb86-18"><a href="#cb86-18"></a> <span class="st">`</span><span class="dt">1se</span><span class="st">`</span> =<span class="st"> </span><span class="kw">split</span>(lassoAll_predResp_1se_vec, hcc5hmC_all_group_vec),</span>
<span id="cb86-19"><a href="#cb86-19"></a> <span class="dt">min =</span> <span class="kw">split</span>(lassoAll_predResp_min_vec, hcc5hmC_all_group_vec)</span>
<span id="cb86-20"><a href="#cb86-20"></a>)</span>
<span id="cb86-21"><a href="#cb86-21"></a><span class="kw">names</span>(tmp) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&#39;</span><span class="ch">\\</span><span class="st">.&#39;</span>, <span class="st">&#39;</span><span class="ch">\n</span><span class="st">&#39;</span>, <span class="kw">names</span>(tmp))</span>
<span id="cb86-22"><a href="#cb86-22"></a></span>
<span id="cb86-23"><a href="#cb86-23"></a><span class="kw">boxplot</span>(</span>
<span id="cb86-24"><a href="#cb86-24"></a> tmp,</span>
<span id="cb86-25"><a href="#cb86-25"></a> <span class="dt">ylab=</span><span class="st">&#39;Predicted oof probability&#39;</span>,</span>
<span id="cb86-26"><a href="#cb86-26"></a> <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;green&#39;</span>, <span class="st">&#39;red&#39;</span>),</span>
<span id="cb86-27"><a href="#cb86-27"></a> <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb86-28"><a href="#cb86-28"></a>)</span>
<span id="cb86-29"><a href="#cb86-29"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(tmp), <span class="dt">tick=</span>F, <span class="kw">names</span>(tmp))</span>
<span id="cb86-30"><a href="#cb86-30"></a></span>
<span id="cb86-31"><a href="#cb86-31"></a></span>
<span id="cb86-32"><a href="#cb86-32"></a><span class="co"># compare the two</span></span>
<span id="cb86-33"><a href="#cb86-33"></a><span class="kw">plot</span>(</span>
<span id="cb86-34"><a href="#cb86-34"></a> <span class="dt">x =</span> lassoAll_predResp_1se_vec, <span class="dt">xlab=</span><span class="st">&#39;1se model oof Prob&#39;</span>,</span>
<span id="cb86-35"><a href="#cb86-35"></a> <span class="dt">y =</span> lassoAll_predResp_min_vec, <span class="dt">ylab=</span><span class="st">&#39;min lambda model oof Prob&#39;</span>,</span>
<span id="cb86-36"><a href="#cb86-36"></a> <span class="dt">col =</span> <span class="kw">ifelse</span>(hcc5hmC_all_group_vec <span class="op">==</span><span class="st"> &#39;HCC&#39;</span>, <span class="st">&#39;red&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb86-37"><a href="#cb86-37"></a>)</span>
<span id="cb86-38"><a href="#cb86-38"></a> </span>
<span id="cb86-39"><a href="#cb86-39"></a><span class="co"># Add referecne lines at 10% false positive</span></span>
<span id="cb86-40"><a href="#cb86-40"></a>thres_1se &lt;-<span class="st"> </span><span class="kw">quantile</span>(lassoAll_predResp_1se_vec[hcc5hmC_all_group_vec <span class="op">==</span><span class="st"> &#39;Control&#39;</span>], <span class="dt">prob=</span>.<span class="dv">9</span>)</span>
<span id="cb86-41"><a href="#cb86-41"></a>thres_min &lt;-<span class="st"> </span><span class="kw">quantile</span>(lassoAll_predResp_min_vec[hcc5hmC_all_group_vec <span class="op">==</span><span class="st"> &#39;Control&#39;</span>], <span class="dt">prob=</span>.<span class="dv">9</span>)</span>
<span id="cb86-42"><a href="#cb86-42"></a><span class="kw">abline</span>(<span class="dt">v =</span> thres_1se, <span class="dt">h =</span> thres_min, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuite-get-sample-pred-1.png" alt="Predicted probabilities - averaged over cv replicates" width="960" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuite-get-sample-pred)Predicted probabilities - averaged over cv replicates
</p>
</div>
<!-- THIS PARAGRAPH REFERRED TO THE FITTED PROBS; NOT THE OOF PRED PROBS
We see that the minimum lambda models provide a better fit to the data,
which is to be expected as the minimum lambda models have more estimated
parameters than the one standard error rule models.  
-->
<p>We see that there isn’t a big difference in out-of-fold predicted
probabilities between the one-standard-error rule and the minimum lambda models.
One way to quantify
the difference in classification errors is to classify samples
according to each vector of predicted probabilities, setting
the thresholds to achieve a fixed false positive rate, 10% say.
These thresholds are indicated by the grey lines in the scatter plot
on the right side of Figure @ref(fig:get-sample-pred).</p>
<!-- APPLIED TO THE FITTED VALUES
We note
that although predicted probability distributions are quite different
for the two models, the class predictions at a 10% false discovery threshold
are largely in agreement.
-->
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1"></a>lassoAll_predClass_1se_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb87-2"><a href="#cb87-2"></a> lassoAll_predResp_1se_vec <span class="op">&gt;</span><span class="st"> </span>thres_1se, <span class="st">&#39;HCC&#39;</span>, <span class="st">&#39;Control&#39;</span>)</span>
<span id="cb87-3"><a href="#cb87-3"></a></span>
<span id="cb87-4"><a href="#cb87-4"></a>lassoAll_predClass_min_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb87-5"><a href="#cb87-5"></a> lassoAll_predResp_min_vec <span class="op">&gt;</span><span class="st"> </span>thres_min, <span class="st">&#39;HCC&#39;</span>, <span class="st">&#39;Control&#39;</span>)</span>
<span id="cb87-6"><a href="#cb87-6"></a></span>
<span id="cb87-7"><a href="#cb87-7"></a>tmp &lt;-<span class="st"> </span><span class="kw">cbind</span>(</span>
<span id="cb87-8"><a href="#cb87-8"></a> <span class="kw">table</span>(<span class="dt">truth=</span>hcc5hmC_all_group_vec, <span class="st">`</span><span class="dt">1se-pred</span><span class="st">`</span>=lassoAll_predClass_1se_vec),</span>
<span id="cb87-9"><a href="#cb87-9"></a> <span class="kw">table</span>(<span class="dt">truth=</span>hcc5hmC_all_group_vec, <span class="st">`</span><span class="dt">min-pred</span><span class="st">`</span>=lassoAll_predClass_min_vec)</span>
<span id="cb87-10"><a href="#cb87-10"></a>) </span>
<span id="cb87-11"><a href="#cb87-11"></a><span class="co"># Hack for printing</span></span>
<span id="cb87-12"><a href="#cb87-12"></a><span class="kw">colnames</span>(tmp) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;1se-Control&#39;</span>, <span class="st">&#39;1se-HCC&#39;</span>, <span class="st">&#39;min-Control&#39;</span>, <span class="st">&#39;min-HCC&#39;</span>)</span>
<span id="cb87-13"><a href="#cb87-13"></a></span>
<span id="cb87-14"><a href="#cb87-14"></a>knitr<span class="op">::</span><span class="kw">kable</span>(tmp,</span>
<span id="cb87-15"><a href="#cb87-15"></a>  <span class="dt">caption =</span> <span class="st">&quot;Classifications: rows are truth&quot;</span>,</span>
<span id="cb87-16"><a href="#cb87-16"></a>  <span class="dt">digits=</span><span class="dv">1</span>) <span class="op">%&gt;%</span></span>
<span id="cb87-17"><a href="#cb87-17"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-glmnetSuite-get-sample-class)Classifications: rows are truth
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
1se-Control
</th>
<th style="text-align:right;">
1se-HCC
</th>
<th style="text-align:right;">
min-Control
</th>
<th style="text-align:right;">
min-HCC
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Control
</td>
<td style="text-align:right;">
700
</td>
<td style="text-align:right;">
78
</td>
<td style="text-align:right;">
700
</td>
<td style="text-align:right;">
78
</td>
</tr>
<tr>
<td style="text-align:left;">
HCC
</td>
<td style="text-align:right;">
39
</td>
<td style="text-align:right;">
516
</td>
<td style="text-align:right;">
32
</td>
<td style="text-align:right;">
523
</td>
</tr>
</tbody>
</table>
<p>When we fix the false positive rate at 10% (ie. the control samples error rates are fixed),
the <code>1se</code> model makes 39 false negative calls whereas the minimum lambda model makes 32. A difference
of 1.3%</p>
<!-- APPLIED TO THE FITTED PROBABILITIES
We see that the min lambda model, makes no false negative calls at a 90% sensitivity
setting, and the sensitivity could be increased substantially at no false negative
cost.  This is definitely over-fitting the data set.  For the purpose
of computing sample consistency scores - what do these differences mean? 
-->
<div id="get-sample-consistency-scores" class="section level3 unnumbered" number="">
<h3 class="unnumbered" number="">Get sample consistency scores</h3>
<p>To compute consistency scores, we will use the out-of-fold predicted probabilities.</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1"></a><span class="co"># get qual scores</span></span>
<span id="cb88-2"><a href="#cb88-2"></a></span>
<span id="cb88-3"><a href="#cb88-3"></a>y &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(hcc5hmC_all_group_vec <span class="op">==</span><span class="st"> &#39;HCC&#39;</span>)</span>
<span id="cb88-4"><a href="#cb88-4"></a><span class="co"># 1se</span></span>
<span id="cb88-5"><a href="#cb88-5"></a>p &lt;-<span class="st"> </span>lassoAll_predResp_1se_vec</span>
<span id="cb88-6"><a href="#cb88-6"></a>hcc5hmC_sample_1se_qual_vec &lt;-<span class="st"> </span>p<span class="op">^</span>y<span class="op">*</span>(<span class="dv">1</span><span class="op">-</span>p)<span class="op">^</span>(<span class="dv">1</span><span class="op">-</span>y)</span>
<span id="cb88-7"><a href="#cb88-7"></a></span>
<span id="cb88-8"><a href="#cb88-8"></a><span class="co"># min</span></span>
<span id="cb88-9"><a href="#cb88-9"></a>p &lt;-<span class="st"> </span>lassoAll_predResp_min_vec</span>
<span id="cb88-10"><a href="#cb88-10"></a>hcc5hmC_sample_min_qual_vec &lt;-<span class="st"> </span>p<span class="op">^</span>y<span class="op">*</span>(<span class="dv">1</span><span class="op">-</span>p)<span class="op">^</span>(<span class="dv">1</span><span class="op">-</span>y)</span></code></pre></div>
<p>We can examine consistency scores as a function of classification bin.</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1"></a>y &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(hcc5hmC_all_group_vec <span class="op">==</span><span class="st"> &#39;HCC&#39;</span>)</span>
<span id="cb89-2"><a href="#cb89-2"></a></span>
<span id="cb89-3"><a href="#cb89-3"></a><span class="co"># 1se</span></span>
<span id="cb89-4"><a href="#cb89-4"></a>lassoAll_1se_conf_vec &lt;-<span class="st"> </span><span class="kw">paste</span>(</span>
<span id="cb89-5"><a href="#cb89-5"></a> y, </span>
<span id="cb89-6"><a href="#cb89-6"></a> <span class="kw">as.numeric</span>(lassoAll_predClass_1se_vec<span class="op">==</span><span class="st">&#39;HCC&#39;</span>),</span>
<span id="cb89-7"><a href="#cb89-7"></a> <span class="dt">sep =</span> <span class="st">&#39;:&#39;</span></span>
<span id="cb89-8"><a href="#cb89-8"></a>)</span>
<span id="cb89-9"><a href="#cb89-9"></a></span>
<span id="cb89-10"><a href="#cb89-10"></a><span class="co"># min</span></span>
<span id="cb89-11"><a href="#cb89-11"></a>lassoAll_min_conf_vec &lt;-<span class="st"> </span><span class="kw">paste</span>(</span>
<span id="cb89-12"><a href="#cb89-12"></a> y, </span>
<span id="cb89-13"><a href="#cb89-13"></a> <span class="kw">as.numeric</span>(lassoAll_predClass_min_vec<span class="op">==</span><span class="st">&#39;HCC&#39;</span>),</span>
<span id="cb89-14"><a href="#cb89-14"></a> <span class="dt">sep =</span> <span class="st">&#39;:&#39;</span></span>
<span id="cb89-15"><a href="#cb89-15"></a>)</span>
<span id="cb89-16"><a href="#cb89-16"></a></span>
<span id="cb89-17"><a href="#cb89-17"></a></span>
<span id="cb89-18"><a href="#cb89-18"></a>tmp &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb89-19"><a href="#cb89-19"></a> <span class="kw">split</span>(hcc5hmC_sample_1se_qual_vec, lassoAll_1se_conf_vec), </span>
<span id="cb89-20"><a href="#cb89-20"></a> <span class="kw">split</span>(hcc5hmC_sample_min_qual_vec, lassoAll_min_conf_vec)</span>
<span id="cb89-21"><a href="#cb89-21"></a>)</span>
<span id="cb89-22"><a href="#cb89-22"></a></span>
<span id="cb89-23"><a href="#cb89-23"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">3</span>,<span class="dv">3</span>,<span class="dv">2</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb89-24"><a href="#cb89-24"></a>gplots<span class="op">::</span><span class="kw">boxplot2</span>(<span class="kw">split</span>(hcc5hmC_sample_1se_qual_vec, lassoAll_1se_conf_vec), </span>
<span id="cb89-25"><a href="#cb89-25"></a>  <span class="dt">outline=</span>F, <span class="dt">ylab =</span> <span class="st">&#39;&#39;</span>, </span>
<span id="cb89-26"><a href="#cb89-26"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;green&#39;</span>, <span class="st">&#39;green&#39;</span>, <span class="st">&#39;red&#39;</span>, <span class="st">&#39;red&#39;</span>),</span>
<span id="cb89-27"><a href="#cb89-27"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>))</span>
<span id="cb89-28"><a href="#cb89-28"></a><span class="kw">title</span>(<span class="st">&#39;1se Model&#39;</span>)</span>
<span id="cb89-29"><a href="#cb89-29"></a></span>
<span id="cb89-30"><a href="#cb89-30"></a>gplots<span class="op">::</span><span class="kw">boxplot2</span>(<span class="kw">split</span>(hcc5hmC_sample_min_qual_vec, lassoAll_min_conf_vec), </span>
<span id="cb89-31"><a href="#cb89-31"></a>  <span class="dt">outline=</span>F, <span class="dt">ylab =</span> <span class="st">&#39;&#39;</span>,</span>
<span id="cb89-32"><a href="#cb89-32"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;green&#39;</span>, <span class="st">&#39;green&#39;</span>, <span class="st">&#39;red&#39;</span>, <span class="st">&#39;red&#39;</span>),</span>
<span id="cb89-33"><a href="#cb89-33"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>))</span>
<span id="cb89-34"><a href="#cb89-34"></a><span class="kw">title</span>(<span class="st">&#39;min lambda Model&#39;</span>)</span>
<span id="cb89-35"><a href="#cb89-35"></a></span>
<span id="cb89-36"><a href="#cb89-36"></a></span>
<span id="cb89-37"><a href="#cb89-37"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.5</span>, <span class="st">&#39;Classification - Truth:Predicted&#39;</span>)</span>
<span id="cb89-38"><a href="#cb89-38"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.5</span>, <span class="st">&#39;Consistency Score&#39;</span>)</span>
<span id="cb89-39"><a href="#cb89-39"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.5</span>, <span class="st">&#39;Sample Consistency vs Classification Outcome&#39;</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuite-plot-qual-conf-1.png" alt="quality scores by classification - Control=0, HCC=1" width="768" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuite-plot-qual-conf)quality scores by classification - Control=0, HCC=1
</p>
</div>
<p>This figure shows that for false positive cases (0:1 or classifying a
control as an affected case), the algorithm is <em>less certain</em> of its predicted
outcome than for the false negative cases (1:0 or classifying an affected case as a control).
ie. the misclassified HCC samples are quite similar to Controls, whereas there
is more ambiguity in the misclassified Control samples.</p>
<p>We will use the minimum lambda model to provide
the fitted probabilities used to compute quality scores,
but we could have used either one.</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1"></a>sample_qual_vec &lt;-<span class="st"> </span>hcc5hmC_sample_min_qual_vec</span></code></pre></div>
</div>
</div>
<div id="selected-feature-list-stability" class="section level2" number="5.2">
<h2 number="5.2"><span class="header-section-number">5.2</span> Selected feature list stability</h2>
<p>Before moving on to the simulation, let’s examine gene selection stability on the
full data set. We have two sets of selected features - one for the
one standard deviation rule model, and one for the minimum lambda model.
We saw in Table @ref(tab:model-size-lassoAll) that the number of features
selected by the minimum lambda models had an IQR of
175.25-261,
while the one standard error rule models had an IQR of
102-132.5.</p>
<p>Let’s examine the stability of the gene lists across cv replicates.</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb91-2"><a href="#cb91-2"></a></span>
<span id="cb91-3"><a href="#cb91-3"></a></span>
<span id="cb91-4"><a href="#cb91-4"></a><span class="co"># 1se</span></span>
<span id="cb91-5"><a href="#cb91-5"></a>lassoAll_coef_1se_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(hcc5hmC_cv_lassoAll_lst, <span class="cf">function</span>(cv_fit){</span>
<span id="cb91-6"><a href="#cb91-6"></a> cv_fit_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb91-7"><a href="#cb91-7"></a> cv_fit,</span>
<span id="cb91-8"><a href="#cb91-8"></a> <span class="dt">s =</span> <span class="st">&quot;lambda.1se&quot;</span></span>
<span id="cb91-9"><a href="#cb91-9"></a> )</span>
<span id="cb91-10"><a href="#cb91-10"></a> cv_fit_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][cv_fit_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>]]</span>
<span id="cb91-11"><a href="#cb91-11"></a> })</span>
<span id="cb91-12"><a href="#cb91-12"></a></span>
<span id="cb91-13"><a href="#cb91-13"></a><span class="co"># put into matrix</span></span>
<span id="cb91-14"><a href="#cb91-14"></a>lassoAll_coef_1se_all &lt;-<span class="st"> </span><span class="kw">Reduce</span>(union, lassoAll_coef_1se_lst)</span>
<span id="cb91-15"><a href="#cb91-15"></a>lassoAll_coef_1se_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(lassoAll_coef_1se_lst, </span>
<span id="cb91-16"><a href="#cb91-16"></a>  <span class="cf">function</span>(LL) <span class="kw">is.element</span>(lassoAll_coef_1se_all, LL)</span>
<span id="cb91-17"><a href="#cb91-17"></a>)</span>
<span id="cb91-18"><a href="#cb91-18"></a><span class="kw">rownames</span>(lassoAll_coef_1se_mtx) &lt;-<span class="st"> </span>lassoAll_coef_1se_all</span>
<span id="cb91-19"><a href="#cb91-19"></a></span>
<span id="cb91-20"><a href="#cb91-20"></a>genes_by_rep_1se_tbl &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="kw">rowSums</span>(lassoAll_coef_1se_mtx))</span>
<span id="cb91-21"><a href="#cb91-21"></a><span class="kw">barplot</span>(</span>
<span id="cb91-22"><a href="#cb91-22"></a> genes_by_rep_1se_tbl,</span>
<span id="cb91-23"><a href="#cb91-23"></a> <span class="dt">xlab=</span><span class="st">&#39;Number of Replicates&#39;</span>,</span>
<span id="cb91-24"><a href="#cb91-24"></a> <span class="dt">ylab=</span><span class="st">&#39;Number of features&#39;</span></span>
<span id="cb91-25"><a href="#cb91-25"></a></span>
<span id="cb91-26"><a href="#cb91-26"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuite-feature-list-1se-1.png" alt="Feature list stability for one standard error rule models" width="768" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuite-feature-list-1se)Feature list stability for one standard error rule models
</p>
</div>
<p>We see that 44 features are included in every
cv replicate. These make up between
33%
and
43%
(Q1 and Q3) of the cv replicate one standard error rule models feature lists.</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb92-2"><a href="#cb92-2"></a></span>
<span id="cb92-3"><a href="#cb92-3"></a></span>
<span id="cb92-4"><a href="#cb92-4"></a><span class="co"># min</span></span>
<span id="cb92-5"><a href="#cb92-5"></a>lassoAll_coef_min_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(hcc5hmC_cv_lassoAll_lst, <span class="cf">function</span>(cv_fit){</span>
<span id="cb92-6"><a href="#cb92-6"></a> cv_fit_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb92-7"><a href="#cb92-7"></a> cv_fit,</span>
<span id="cb92-8"><a href="#cb92-8"></a> <span class="dt">s =</span> <span class="st">&quot;lambda.min&quot;</span></span>
<span id="cb92-9"><a href="#cb92-9"></a> )</span>
<span id="cb92-10"><a href="#cb92-10"></a> cv_fit_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][cv_fit_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>]]</span>
<span id="cb92-11"><a href="#cb92-11"></a> })</span>
<span id="cb92-12"><a href="#cb92-12"></a></span>
<span id="cb92-13"><a href="#cb92-13"></a><span class="co"># put into matrix</span></span>
<span id="cb92-14"><a href="#cb92-14"></a>lassoAll_coef_min_all &lt;-<span class="st"> </span><span class="kw">Reduce</span>(union, lassoAll_coef_min_lst)</span>
<span id="cb92-15"><a href="#cb92-15"></a>lassoAll_coef_min_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(lassoAll_coef_min_lst, </span>
<span id="cb92-16"><a href="#cb92-16"></a>  <span class="cf">function</span>(LL) <span class="kw">is.element</span>(lassoAll_coef_min_all, LL)</span>
<span id="cb92-17"><a href="#cb92-17"></a>)</span>
<span id="cb92-18"><a href="#cb92-18"></a><span class="kw">rownames</span>(lassoAll_coef_min_mtx) &lt;-<span class="st"> </span>lassoAll_coef_min_all</span>
<span id="cb92-19"><a href="#cb92-19"></a></span>
<span id="cb92-20"><a href="#cb92-20"></a>genes_by_rep_min_tbl &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="kw">rowSums</span>(lassoAll_coef_min_mtx))</span>
<span id="cb92-21"><a href="#cb92-21"></a><span class="kw">barplot</span>(</span>
<span id="cb92-22"><a href="#cb92-22"></a> genes_by_rep_min_tbl,</span>
<span id="cb92-23"><a href="#cb92-23"></a> <span class="dt">xlab=</span><span class="st">&#39;Number of Replicates&#39;</span>,</span>
<span id="cb92-24"><a href="#cb92-24"></a> <span class="dt">ylab=</span><span class="st">&#39;Number of features&#39;</span></span>
<span id="cb92-25"><a href="#cb92-25"></a></span>
<span id="cb92-26"><a href="#cb92-26"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuite-feature-list-min-1.png" alt="Feature list stability for minimum lambda models" width="768" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuite-feature-list-min)Feature list stability for minimum lambda models
</p>
</div>
<p>We see that 106 features are included in every
cv replicate. These make up between
41%
and
60%
(Q1 and Q3) of the cv replicate min feature lists.
We will consider the genes that are selected in all cv replicates as a
gene signature produced by each model.</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1"></a>lasso_gene_sign_1se_vec &lt;-<span class="st"> </span><span class="kw">rownames</span>(lassoAll_coef_1se_mtx)[<span class="kw">rowSums</span>(lassoAll_coef_1se_mtx)<span class="op">==</span><span class="dv">30</span>]</span>
<span id="cb93-2"><a href="#cb93-2"></a>lasso_gene_sign_min_vec &lt;-<span class="st"> </span><span class="kw">rownames</span>(lassoAll_coef_min_mtx)[<span class="kw">rowSums</span>(lassoAll_coef_min_mtx)<span class="op">==</span><span class="dv">30</span>]</span></code></pre></div>
<p>44 out of
44 of the genes in the 1se model gene signature
are contained in the min lambda model gene signature.</p>
</div>
<div id="simulation-design" class="section level2" number="5.3">
<h2 number="5.3"><span class="header-section-number">5.3</span> Simulation Design</h2>
<p>We are now ready to run the simulations.</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1"></a> SIM &lt;-<span class="st"> </span><span class="dv">30</span></span>
<span id="cb94-2"><a href="#cb94-2"></a> SIZE &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">25</span>, <span class="dv">50</span>, <span class="dv">100</span>, <span class="dv">200</span>, <span class="dv">300</span>)</span>
<span id="cb94-3"><a href="#cb94-3"></a> CV_REP &lt;-<span class="st"> </span><span class="dv">30</span></span></code></pre></div>
<p>Simluation parameters:</p>
<ul>
<li><p>Number of simulations : SIM = 30</p></li>
<li><p>Sample sizes: SIZE = 25, 50, 100, 200, 300</p></li>
<li><p>Number of CV Replicates: CV_REP = 30</p></li>
</ul>
<p>We will repeat the simulation process SIM = 30 times.
For each simulation iteration, we will select 300 Control and
300 HCC samples at random. Models will be fitted and analyzed
to balanced subsets of SIZE = 25, 50, 100, 200, 300, in a <code>Matryoshka doll</code> manner to
emulate a typical sample accrual process. Note that in this accrual process
there is no time effect - the accrual process is completely randomized. In practice,
there could be significant time effects. For example, the first 25 HCC samples could come
from Center A, while the next 25 could come from Center B. And
affected and control samples could be acquired from different clinics
or in different time intervals. In other words,
there is no batch effect or shared variability in our simulation,
while these are almost always present in real data, including
batch effects that are associated with class labels - controls being in
different batches than affected samples is an all too common occurrence,
for example. One should be especially watchful of potential batch effects
when dealing with blood samples as blood is notoriously finicky in
character [<span class="citation">[<a href="#ref-Huang:2017aa" role="doc-biblioref">16</a>]</span>; <span class="citation">[<a href="#ref-Permenter:2015aa" role="doc-biblioref">17</a>]</span>;].
Presented with results that look impressively good based on a small data set,
one should definitely be skeptical of the promise of future equally good results.</p>
<p>For a given simulation and a given sample size, we will obtain
CV_REP = 30 cross-validated lasso fits. From these fits,
we can obtain 30 out-of-fold assessments of classification accuracy
to get a sense if its variability. From each cv replicate, we also obtain
an estimated model size and a set of selected features. We will want
to examine how these stabilize as the sample size increases.</p>
<p>Note that we limit the simulations to a maximum of sample size of 300 in
order to to have simulations with low overlap. With 300
randomly selected HCC samples, the expected overlap between two randomly
selected sets of HCC samples is 29.2%.
For Controls the expected overlap is 14.9%.</p>
<p><strong>Reader Note</strong></p>
<pre><code>We are currently not following individual performance paths
as a study set grows over time.  We currently simply summarize
results across simulations.  Exmaining individual paths would
show how chaotic the assessments of performance can be when
sample sizes are small.

To Do.</code></pre>
</div>
<div id="setup-simulation" class="section level2" number="5.4">
<h2 number="5.4"><span class="header-section-number">5.4</span> Setup simulation</h2>
<p>To setup the simulation, we only need two master tables: one for the selection of Controls
and one for the selection of HCC samples.</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1"></a>hcc5hmC_all_control_vec &lt;-<span class="st"> </span><span class="kw">names</span>(hcc5hmC_all_group_vec[hcc5hmC_all_group_vec<span class="op">==</span><span class="st">&#39;Control&#39;</span>]) </span>
<span id="cb96-2"><a href="#cb96-2"></a>hcc5hmC_all_affected_vec &lt;-<span class="st"> </span><span class="kw">names</span>(hcc5hmC_all_group_vec[hcc5hmC_all_group_vec<span class="op">==</span><span class="st">&#39;HCC&#39;</span>])  </span></code></pre></div>
<p>We have 778 control sample IDs stored in <code>hcc5hmC_all_control_vec</code>
and 555 affected sample IDs stored in <code>hcc5hmC_all_affected_vec</code>.
To create a suite of random samples from these, we only need to randomly select indices from
each vector.</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1"></a><span class="kw">set.seed</span>(<span class="dv">12379</span>)</span>
<span id="cb97-2"><a href="#cb97-2"></a></span>
<span id="cb97-3"><a href="#cb97-3"></a>hcc5hmC_sim_control_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb97-4"><a href="#cb97-4"></a> <span class="dv">1</span><span class="op">:</span>SIM, </span>
<span id="cb97-5"><a href="#cb97-5"></a> <span class="cf">function</span>(dummy) </span>
<span id="cb97-6"><a href="#cb97-6"></a>   <span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(hcc5hmC_all_control_vec), <span class="dt">size =</span>  <span class="kw">max</span>(SIZE))</span>
<span id="cb97-7"><a href="#cb97-7"></a>)</span>
<span id="cb97-8"><a href="#cb97-8"></a></span>
<span id="cb97-9"><a href="#cb97-9"></a></span>
<span id="cb97-10"><a href="#cb97-10"></a>hcc5hmC_sim_affected_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb97-11"><a href="#cb97-11"></a> <span class="dv">1</span><span class="op">:</span>SIM, </span>
<span id="cb97-12"><a href="#cb97-12"></a> <span class="cf">function</span>(dummy) </span>
<span id="cb97-13"><a href="#cb97-13"></a>   <span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(hcc5hmC_all_affected_vec), <span class="dt">size =</span>  <span class="kw">max</span>(SIZE))</span>
<span id="cb97-14"><a href="#cb97-14"></a>)</span></code></pre></div>
<p>Each simulation is specified by a given column of the simulation design matrices:
<code>hcc5hmC_sim_control_mtx</code> and <code>hcc5hmC_sim_affected_mtx</code>, each with dimensions 300, 30.
Within each simulation, we can run the analyses of size 25, 50, 100, 200, 300 by simply selecting
samples specified in the appropriate rows of each design matrix.</p>
<p>We can examine how much variability we have in the quality scores of the selected samples.
Here we show results for the small sample sizes where variability will be the greatest.</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb98-2"><a href="#cb98-2"></a></span>
<span id="cb98-3"><a href="#cb98-3"></a>all_control_qual_vec &lt;-<span class="st"> </span>sample_qual_vec[hcc5hmC_all_control_vec]</span>
<span id="cb98-4"><a href="#cb98-4"></a>sim_control_qual_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb98-5"><a href="#cb98-5"></a>  <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(hcc5hmC_sim_control_mtx), </span>
<span id="cb98-6"><a href="#cb98-6"></a>  <span class="cf">function</span>(CC) all_control_qual_vec[hcc5hmC_sim_control_mtx[,CC]]</span>
<span id="cb98-7"><a href="#cb98-7"></a> )</span>
<span id="cb98-8"><a href="#cb98-8"></a></span>
<span id="cb98-9"><a href="#cb98-9"></a>all_affected_qual_vec &lt;-<span class="st"> </span>sample_qual_vec[hcc5hmC_all_affected_vec]</span>
<span id="cb98-10"><a href="#cb98-10"></a>sim_affected_qual_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb98-11"><a href="#cb98-11"></a>  <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(hcc5hmC_sim_affected_mtx),  </span>
<span id="cb98-12"><a href="#cb98-12"></a>  <span class="cf">function</span>(CC) all_affected_qual_vec[hcc5hmC_sim_affected_mtx[,CC]]</span>
<span id="cb98-13"><a href="#cb98-13"></a> )</span>
<span id="cb98-14"><a href="#cb98-14"></a></span>
<span id="cb98-15"><a href="#cb98-15"></a><span class="co"># ONLY LOOK AT SAMPLE SIZE == 50</span></span>
<span id="cb98-16"><a href="#cb98-16"></a>NN &lt;-<span class="st"> </span><span class="dv">50</span></span>
<span id="cb98-17"><a href="#cb98-17"></a></span>
<span id="cb98-18"><a href="#cb98-18"></a><span class="co"># PLOT</span></span>
<span id="cb98-19"><a href="#cb98-19"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">1</span>), <span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">5</span>,<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb98-20"><a href="#cb98-20"></a><span class="co"># control</span></span>
<span id="cb98-21"><a href="#cb98-21"></a><span class="kw">boxplot</span>(</span>
<span id="cb98-22"><a href="#cb98-22"></a>  sim_control_qual_mtx[<span class="dv">1</span><span class="op">:</span>NN,],</span>
<span id="cb98-23"><a href="#cb98-23"></a>  <span class="dt">outline =</span> T, </span>
<span id="cb98-24"><a href="#cb98-24"></a>  <span class="dt">ylab =</span> <span class="st">&#39;Consistency Score&#39;</span>,</span>
<span id="cb98-25"><a href="#cb98-25"></a>  <span class="dt">xaxt =</span> <span class="st">&#39;n&#39;</span></span>
<span id="cb98-26"><a href="#cb98-26"></a>)</span>
<span id="cb98-27"><a href="#cb98-27"></a><span class="kw">title</span>(<span class="st">&quot;Control sample consistency across simulations&quot;</span>)</span>
<span id="cb98-28"><a href="#cb98-28"></a></span>
<span id="cb98-29"><a href="#cb98-29"></a><span class="co"># affected</span></span>
<span id="cb98-30"><a href="#cb98-30"></a><span class="kw">boxplot</span>(</span>
<span id="cb98-31"><a href="#cb98-31"></a>  sim_affected_qual_mtx[<span class="dv">1</span><span class="op">:</span>NN,],</span>
<span id="cb98-32"><a href="#cb98-32"></a>  <span class="dt">outline =</span> T, </span>
<span id="cb98-33"><a href="#cb98-33"></a>  <span class="dt">ylab =</span> <span class="st">&#39;Consistency Score&#39;</span></span>
<span id="cb98-34"><a href="#cb98-34"></a>)</span>
<span id="cb98-35"><a href="#cb98-35"></a><span class="kw">title</span>(<span class="st">&quot;Affected sample consistency across simulations&quot;</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuite-look-sim-qual-0-50ONLY-1.png" alt="sample consistency by simulation run for size = 50 " width="960" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuite-look-sim-qual-0-50ONLY)sample consistency by simulation run for size = 50
</p>
</div>
<p>In this figure, we are summarizing the quality measures of 50 samples per group across
30 simulations, or random selections of control and affected samples.
We see significant variability in sample consistency, especially in the affected cases.
This may lead an unwary observer to be overly optimistic, or overly pessimistic,
in the early accrual stages of a study.</p>
</div>
<div id="run-simulations" class="section level2" number="5.5">
<h2 number="5.5"><span class="header-section-number">5.5</span> Run simulations</h2>
<p>As these take a while to run,
we will save the results of each simulation to a different
object and store to disk. These can be easily read from disk
when needed for analysis.</p>
<p>The simulation results are saved to the file system and
only needs to be run once. The simulation takes <span class="math inline">\(\approx\)</span> 8 to 10 minutes
per iteration, or 4 to 5 hours of run time on a laptop.
(Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Mojave 10.14.6)</p>
<!-- RUN ONCE - THEN GET FROM MEMORY -->
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1"></a>start_time &lt;-<span class="st"> </span><span class="kw">proc.time</span>()</span>
<span id="cb99-2"><a href="#cb99-2"></a></span>
<span id="cb99-3"><a href="#cb99-3"></a><span class="co"># Get stage from SIZE</span></span>
<span id="cb99-4"><a href="#cb99-4"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sim_control_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>, SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb99-5"><a href="#cb99-5"></a></span>
<span id="cb99-6"><a href="#cb99-6"></a><span class="cf">for</span> (SIMno <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(sim_control_qual_mtx)) {</span>
<span id="cb99-7"><a href="#cb99-7"></a></span>
<span id="cb99-8"><a href="#cb99-8"></a>  <span class="co">#cat(&quot;Running simulation &quot;, SIMno, &quot;\n&quot;)</span></span>
<span id="cb99-9"><a href="#cb99-9"></a></span>
<span id="cb99-10"><a href="#cb99-10"></a>  sim_cv_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(<span class="kw">levels</span>(stage_vec)), <span class="cf">function</span>(STGno) {</span>
<span id="cb99-11"><a href="#cb99-11"></a>    Stage_rows_vec &lt;-<span class="st"> </span><span class="kw">which</span>(stage_vec <span class="op">%in%</span><span class="st"> </span><span class="kw">levels</span>(stage_vec)[<span class="dv">1</span><span class="op">:</span>STGno])</span>
<span id="cb99-12"><a href="#cb99-12"></a>    <span class="co">#cat(&quot;Stage &quot;, STGno, &quot;- analyzing&quot;, length(Stage_rows_vec), &quot;paired samples.\n&quot;)</span></span>
<span id="cb99-13"><a href="#cb99-13"></a></span>
<span id="cb99-14"><a href="#cb99-14"></a>    sim_stage_samples_vec &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb99-15"><a href="#cb99-15"></a>      hcc5hmC_all_control_vec[hcc5hmC_sim_control_mtx[Stage_rows_vec, SIMno]],</span>
<span id="cb99-16"><a href="#cb99-16"></a>      hcc5hmC_all_affected_vec[hcc5hmC_sim_affected_mtx[Stage_rows_vec, SIMno]]</span>
<span id="cb99-17"><a href="#cb99-17"></a>    )</span>
<span id="cb99-18"><a href="#cb99-18"></a>    sim_stage_lcpm_mtx &lt;-<span class="st"> </span>hcc5hmC_all_lcpm_mtx[sim_stage_samples_vec, ]</span>
<span id="cb99-19"><a href="#cb99-19"></a>    sim_stage_group_vec &lt;-<span class="st"> </span>hcc5hmC_all_group_vec[sim_stage_samples_vec]</span>
<span id="cb99-20"><a href="#cb99-20"></a>    <span class="co">#print(table(sim_stage_group_vec))</span></span>
<span id="cb99-21"><a href="#cb99-21"></a></span>
<span id="cb99-22"><a href="#cb99-22"></a>    sim_stage_cv_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span>CV_REP, <span class="cf">function</span>(CV) {</span>
<span id="cb99-23"><a href="#cb99-23"></a>      cv_fit &lt;-<span class="st"> </span>glmnet<span class="op">::</span><span class="kw">cv.glmnet</span>(</span>
<span id="cb99-24"><a href="#cb99-24"></a>        <span class="dt">x =</span> sim_stage_lcpm_mtx,</span>
<span id="cb99-25"><a href="#cb99-25"></a>        <span class="dt">y =</span> sim_stage_group_vec,</span>
<span id="cb99-26"><a href="#cb99-26"></a>        <span class="dt">alpha =</span> <span class="dv">1</span>,</span>
<span id="cb99-27"><a href="#cb99-27"></a>        <span class="dt">family =</span> <span class="st">&quot;binomial&quot;</span>,</span>
<span id="cb99-28"><a href="#cb99-28"></a>        <span class="dt">type.measure =</span> <span class="st">&quot;class&quot;</span>,</span>
<span id="cb99-29"><a href="#cb99-29"></a>        <span class="dt">keep =</span> T,</span>
<span id="cb99-30"><a href="#cb99-30"></a>        <span class="dt">nlambda =</span> <span class="dv">30</span></span>
<span id="cb99-31"><a href="#cb99-31"></a>      )</span>
<span id="cb99-32"><a href="#cb99-32"></a></span>
<span id="cb99-33"><a href="#cb99-33"></a>      <span class="co"># Extract 1se metrics from cv_fit</span></span>
<span id="cb99-34"><a href="#cb99-34"></a>      <span class="co">#######################</span></span>
<span id="cb99-35"><a href="#cb99-35"></a>      ndx_1se &lt;-<span class="st"> </span><span class="kw">which</span>(cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se)</span>
<span id="cb99-36"><a href="#cb99-36"></a></span>
<span id="cb99-37"><a href="#cb99-37"></a>      nzero_1se &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>nzero[ndx_1se]</span>
<span id="cb99-38"><a href="#cb99-38"></a>      cvm_1se &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>cvm[ndx_1se]</span>
<span id="cb99-39"><a href="#cb99-39"></a></span>
<span id="cb99-40"><a href="#cb99-40"></a>      <span class="co"># test error</span></span>
<span id="cb99-41"><a href="#cb99-41"></a>      sim_stage_test_samples_vec &lt;-<span class="st"> </span><span class="kw">setdiff</span>(<span class="kw">rownames</span>(hcc5hmC_all_lcpm_mtx), sim_stage_samples_vec)</span>
<span id="cb99-42"><a href="#cb99-42"></a>      sim_stage_hcc5hmC_test_lcpm_mtx &lt;-<span class="st"> </span>hcc5hmC_all_lcpm_mtx[sim_stage_test_samples_vec,]</span>
<span id="cb99-43"><a href="#cb99-43"></a>      sim_stage_hcc5hmC_test_group_vec &lt;-<span class="st"> </span>hcc5hmC_all_group_vec[sim_stage_test_samples_vec]</span>
<span id="cb99-44"><a href="#cb99-44"></a></span>
<span id="cb99-45"><a href="#cb99-45"></a>      test_pred_1se_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb99-46"><a href="#cb99-46"></a>       cv_fit,</span>
<span id="cb99-47"><a href="#cb99-47"></a>       <span class="dt">newx=</span>sim_stage_hcc5hmC_test_lcpm_mtx,</span>
<span id="cb99-48"><a href="#cb99-48"></a>       <span class="dt">s=</span><span class="st">&quot;lambda.1se&quot;</span>,</span>
<span id="cb99-49"><a href="#cb99-49"></a>       <span class="dt">type=</span><span class="st">&quot;class&quot;</span></span>
<span id="cb99-50"><a href="#cb99-50"></a>      )</span>
<span id="cb99-51"><a href="#cb99-51"></a>      test_1se_error &lt;-<span class="st"> </span><span class="kw">mean</span>(test_pred_1se_vec <span class="op">!=</span><span class="st"> </span>sim_stage_hcc5hmC_test_group_vec)</span>
<span id="cb99-52"><a href="#cb99-52"></a></span>
<span id="cb99-53"><a href="#cb99-53"></a>      <span class="co"># genes</span></span>
<span id="cb99-54"><a href="#cb99-54"></a>      coef_1se &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb99-55"><a href="#cb99-55"></a>        cv_fit,</span>
<span id="cb99-56"><a href="#cb99-56"></a>        <span class="dt">s =</span> <span class="st">&quot;lambda.1se&quot;</span></span>
<span id="cb99-57"><a href="#cb99-57"></a>      )</span>
<span id="cb99-58"><a href="#cb99-58"></a>      genes_1se &lt;-<span class="st"> </span>coef_1se<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][coef_1se<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>]]</span>
<span id="cb99-59"><a href="#cb99-59"></a></span>
<span id="cb99-60"><a href="#cb99-60"></a>      <span class="co"># Extract min metrics from cv_fit</span></span>
<span id="cb99-61"><a href="#cb99-61"></a>      <span class="co">#######################</span></span>
<span id="cb99-62"><a href="#cb99-62"></a>      ndx_min &lt;-<span class="st"> </span><span class="kw">which</span>(cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min)</span>
<span id="cb99-63"><a href="#cb99-63"></a></span>
<span id="cb99-64"><a href="#cb99-64"></a>      nzero_min &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>nzero[ndx_min]</span>
<span id="cb99-65"><a href="#cb99-65"></a>      cvm_min &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>cvm[ndx_min]</span>
<span id="cb99-66"><a href="#cb99-66"></a></span>
<span id="cb99-67"><a href="#cb99-67"></a>      <span class="co"># test error</span></span>
<span id="cb99-68"><a href="#cb99-68"></a>      sim_stage_test_samples_vec &lt;-<span class="st"> </span><span class="kw">setdiff</span>(<span class="kw">rownames</span>(hcc5hmC_all_lcpm_mtx), sim_stage_samples_vec)</span>
<span id="cb99-69"><a href="#cb99-69"></a>      sim_stage_hcc5hmC_test_lcpm_mtx &lt;-<span class="st"> </span>hcc5hmC_all_lcpm_mtx[sim_stage_test_samples_vec,]</span>
<span id="cb99-70"><a href="#cb99-70"></a>      sim_stage_hcc5hmC_test_group_vec &lt;-<span class="st"> </span>hcc5hmC_all_group_vec[sim_stage_test_samples_vec]</span>
<span id="cb99-71"><a href="#cb99-71"></a></span>
<span id="cb99-72"><a href="#cb99-72"></a>      test_pred_min_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb99-73"><a href="#cb99-73"></a>       cv_fit,</span>
<span id="cb99-74"><a href="#cb99-74"></a>       <span class="dt">newx=</span>sim_stage_hcc5hmC_test_lcpm_mtx,</span>
<span id="cb99-75"><a href="#cb99-75"></a>       <span class="dt">s=</span><span class="st">&quot;lambda.min&quot;</span>,</span>
<span id="cb99-76"><a href="#cb99-76"></a>       <span class="dt">type=</span><span class="st">&quot;class&quot;</span></span>
<span id="cb99-77"><a href="#cb99-77"></a>      )</span>
<span id="cb99-78"><a href="#cb99-78"></a>      test_min_error &lt;-<span class="st"> </span><span class="kw">mean</span>(test_pred_min_vec <span class="op">!=</span><span class="st"> </span>sim_stage_hcc5hmC_test_group_vec)</span>
<span id="cb99-79"><a href="#cb99-79"></a></span>
<span id="cb99-80"><a href="#cb99-80"></a>      <span class="co"># genes</span></span>
<span id="cb99-81"><a href="#cb99-81"></a>      coef_min &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb99-82"><a href="#cb99-82"></a>        cv_fit,</span>
<span id="cb99-83"><a href="#cb99-83"></a>        <span class="dt">s =</span> <span class="st">&quot;lambda.min&quot;</span></span>
<span id="cb99-84"><a href="#cb99-84"></a>      )</span>
<span id="cb99-85"><a href="#cb99-85"></a>      genes_min &lt;-<span class="st"> </span>coef_min<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][coef_min<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>]]</span>
<span id="cb99-86"><a href="#cb99-86"></a></span>
<span id="cb99-87"><a href="#cb99-87"></a>      <span class="co"># return cv_fit summary metrics</span></span>
<span id="cb99-88"><a href="#cb99-88"></a>      <span class="kw">list</span>(</span>
<span id="cb99-89"><a href="#cb99-89"></a>       <span class="dt">p_1se =</span> nzero_1se, </span>
<span id="cb99-90"><a href="#cb99-90"></a>       <span class="dt">p_min =</span> nzero_min, </span>
<span id="cb99-91"><a href="#cb99-91"></a>       <span class="dt">cv_1se =</span> cvm_1se, </span>
<span id="cb99-92"><a href="#cb99-92"></a>       <span class="dt">cv_min =</span> cvm_min, </span>
<span id="cb99-93"><a href="#cb99-93"></a>       <span class="dt">test_1se=</span>test_1se_error, </span>
<span id="cb99-94"><a href="#cb99-94"></a>       <span class="dt">test_min=</span>test_min_error, </span>
<span id="cb99-95"><a href="#cb99-95"></a>       <span class="dt">genes_1se =</span> genes_1se,</span>
<span id="cb99-96"><a href="#cb99-96"></a>       <span class="dt">genes_min =</span> genes_min)</span>
<span id="cb99-97"><a href="#cb99-97"></a>    })</span>
<span id="cb99-98"><a href="#cb99-98"></a>    sim_stage_cv_lst</span>
<span id="cb99-99"><a href="#cb99-99"></a>  })</span>
<span id="cb99-100"><a href="#cb99-100"></a></span>
<span id="cb99-101"><a href="#cb99-101"></a>  <span class="co"># save  sim_cv_lst</span></span>
<span id="cb99-102"><a href="#cb99-102"></a>  fName &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;hcc5hmC_sim_&quot;</span>, SIMno, <span class="st">&quot;_cv_lst&quot;</span>)</span>
<span id="cb99-103"><a href="#cb99-103"></a>  <span class="kw">assign</span>(fName, sim_cv_lst)</span>
<span id="cb99-104"><a href="#cb99-104"></a>  <span class="kw">save</span>(<span class="dt">list =</span> fName, <span class="dt">file=</span><span class="kw">file.path</span>(<span class="st">&quot;RData&quot;</span>, fName))</span>
<span id="cb99-105"><a href="#cb99-105"></a></span>
<span id="cb99-106"><a href="#cb99-106"></a>}</span></code></pre></div>
<!-- DEBUG - rename after the fact -->
</div>
<div id="simulation-results" class="section level2" number="5.6">
<h2 number="5.6"><span class="header-section-number">5.6</span> Simulation results</h2>
<p>Recall the we have 30 simulations, or randomly selected sets of HCC and Control samples,
analyzed in increasing sizes of 25, 50, 100, 200, 300, with
30 repeated cross-validated lasso fits:</p>
<ul>
<li><p>Sample sizes: SIZE = 25, 50, 100, 200, 300</p></li>
<li><p>Number of CV Replicates: CV_REP = 30</p></li>
</ul>
<p>First we extract simulation results and store into one big table
(only showing the top of table here):</p>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb100-2"><a href="#cb100-2"></a>sim_files_vec &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="st">&#39;RData&#39;</span>, <span class="st">&#39;^hcc5hmC_sim_&#39;</span>)</span>
<span id="cb100-3"><a href="#cb100-3"></a></span>
<span id="cb100-4"><a href="#cb100-4"></a></span>
<span id="cb100-5"><a href="#cb100-5"></a><span class="co"># define extraction methods</span></span>
<span id="cb100-6"><a href="#cb100-6"></a></span>
<span id="cb100-7"><a href="#cb100-7"></a><span class="co"># Each sumulation is a list of cv results </span></span>
<span id="cb100-8"><a href="#cb100-8"></a><span class="co">## nested in a list of replicates</span></span>
<span id="cb100-9"><a href="#cb100-9"></a><span class="co">##############################################</span></span>
<span id="cb100-10"><a href="#cb100-10"></a></span>
<span id="cb100-11"><a href="#cb100-11"></a><span class="co"># cvList2frm_f makes a frame out of the inner list</span></span>
<span id="cb100-12"><a href="#cb100-12"></a>cvList2frm_f &lt;-<span class="st"> </span><span class="cf">function</span>(cv_lst) {</span>
<span id="cb100-13"><a href="#cb100-13"></a> frm1 &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">t</span>(<span class="kw">sapply</span>(cv_lst, <span class="cf">function</span>(x) x)))</span>
<span id="cb100-14"><a href="#cb100-14"></a> frm2 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb100-15"><a href="#cb100-15"></a>  <span class="kw">unlist</span>(frm1[[<span class="dv">1</span>]]), <span class="kw">unlist</span>(frm1[[<span class="dv">2</span>]]),</span>
<span id="cb100-16"><a href="#cb100-16"></a>  <span class="kw">unlist</span>(frm1[[<span class="dv">3</span>]]), <span class="kw">unlist</span>(frm1[[<span class="dv">4</span>]]),</span>
<span id="cb100-17"><a href="#cb100-17"></a>  <span class="kw">unlist</span>(frm1[[<span class="dv">5</span>]]), <span class="kw">unlist</span>(frm1[[<span class="dv">6</span>]]),</span>
<span id="cb100-18"><a href="#cb100-18"></a>  frm1[<span class="dv">7</span>], frm1[<span class="dv">8</span>])</span>
<span id="cb100-19"><a href="#cb100-19"></a>  <span class="kw">names</span>(frm2) &lt;-<span class="st"> </span><span class="kw">names</span>(frm1)</span>
<span id="cb100-20"><a href="#cb100-20"></a>  <span class="kw">data.frame</span>(<span class="dt">Rep=</span><span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(frm2), frm2)}</span>
<span id="cb100-21"><a href="#cb100-21"></a></span>
<span id="cb100-22"><a href="#cb100-22"></a><span class="co"># cv_lst_to_frm loop over replicates, concatenating the inner list frames</span></span>
<span id="cb100-23"><a href="#cb100-23"></a>cv_lst_to_frm &lt;-<span class="st"> </span><span class="cf">function</span>(sim_cv_lst) {</span>
<span id="cb100-24"><a href="#cb100-24"></a> <span class="kw">do.call</span>(<span class="st">&#39;rbind&#39;</span>, <span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(sim_cv_lst),</span>
<span id="cb100-25"><a href="#cb100-25"></a>  <span class="cf">function</span>(JJ) {</span>
<span id="cb100-26"><a href="#cb100-26"></a>    siz_frm &lt;-<span class="st"> </span><span class="kw">cvList2frm_f</span>(sim_cv_lst[[JJ]])</span>
<span id="cb100-27"><a href="#cb100-27"></a>    <span class="kw">data.frame</span>(<span class="dt">Size=</span>SIZE[JJ], siz_frm)</span>
<span id="cb100-28"><a href="#cb100-28"></a>  }))</span>
<span id="cb100-29"><a href="#cb100-29"></a>}</span>
<span id="cb100-30"><a href="#cb100-30"></a></span>
<span id="cb100-31"><a href="#cb100-31"></a><span class="co"># we loop across simulations to combine all results into one big table</span></span>
<span id="cb100-32"><a href="#cb100-32"></a>hcc5hmC_lasso_sim_results_frm &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="st">&#39;rbind&#39;</span>, <span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(sim_files_vec),</span>
<span id="cb100-33"><a href="#cb100-33"></a> <span class="cf">function</span>(SIM_NO) {</span>
<span id="cb100-34"><a href="#cb100-34"></a>  <span class="kw">load</span>(<span class="dt">file=</span><span class="kw">file.path</span>(<span class="st">&#39;RData&#39;</span>, sim_files_vec[SIM_NO]))</span>
<span id="cb100-35"><a href="#cb100-35"></a>  <span class="kw">assign</span>(<span class="st">&#39;sim_cv_lst&#39;</span>, <span class="kw">get</span>(sim_files_vec[SIM_NO]))</span>
<span id="cb100-36"><a href="#cb100-36"></a>  <span class="kw">rm</span>(<span class="dt">list=</span>sim_files_vec[SIM_NO])</span>
<span id="cb100-37"><a href="#cb100-37"></a>  </span>
<span id="cb100-38"><a href="#cb100-38"></a>  <span class="kw">data.frame</span>(<span class="dt">SimNo=</span><span class="kw">paste0</span>(<span class="st">&#39;Sim_&#39;</span>,<span class="kw">formatC</span>(SIM_NO,<span class="dt">width =</span> <span class="dv">2</span>,<span class="dt">flag =</span> <span class="dv">0</span>)), <span class="kw">cv_lst_to_frm</span>(sim_cv_lst))</span>
<span id="cb100-39"><a href="#cb100-39"></a>} </span>
<span id="cb100-40"><a href="#cb100-40"></a>)) </span></code></pre></div>
<!-- 
Have a table of simulation results - `hcc5hmC_lasso_sim_results_frm`:
-->
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb101-2"><a href="#cb101-2"></a> </span>
<span id="cb101-3"><a href="#cb101-3"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">head</span>(<span class="kw">with</span>(hcc5hmC_lasso_sim_results_frm, <span class="kw">table</span>(SimNo, Size))),</span>
<span id="cb101-4"><a href="#cb101-4"></a>  <span class="dt">caption =</span> <span class="kw">paste</span>(<span class="st">&quot;Simulation Results - N Sim =&quot;</span>, SIM)) <span class="op">%&gt;%</span></span>
<span id="cb101-5"><a href="#cb101-5"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-glmnetSuite-sum-table)Simulation Results - N Sim = 30
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
25
</th>
<th style="text-align:right;">
50
</th>
<th style="text-align:right;">
100
</th>
<th style="text-align:right;">
200
</th>
<th style="text-align:right;">
300
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Sim_01
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
</tr>
<tr>
<td style="text-align:left;">
Sim_02
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
</tr>
<tr>
<td style="text-align:left;">
Sim_03
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
</tr>
<tr>
<td style="text-align:left;">
Sim_04
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
</tr>
<tr>
<td style="text-align:left;">
Sim_05
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
</tr>
<tr>
<td style="text-align:left;">
Sim_06
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
30
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">head</span>(hcc5hmC_lasso_sim_results_frm) <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(<span class="op">-</span><span class="kw">c</span>(genes_1se, genes_min)),</span>
<span id="cb102-2"><a href="#cb102-2"></a>    <span class="dt">caption =</span> <span class="kw">paste</span>(<span class="st">&quot;Simulation Results - not showing genes column&quot;</span>),</span>
<span id="cb102-3"><a href="#cb102-3"></a>    <span class="dt">digits=</span><span class="dv">2</span>) <span class="op">%&gt;%</span></span>
<span id="cb102-4"><a href="#cb102-4"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-glmnetSuite-sum-table)Simulation Results - not showing genes column
</caption>
<thead>
<tr>
<th style="text-align:left;">
SimNo
</th>
<th style="text-align:right;">
Size
</th>
<th style="text-align:right;">
Rep
</th>
<th style="text-align:right;">
p_1se
</th>
<th style="text-align:right;">
p_min
</th>
<th style="text-align:right;">
cv_1se
</th>
<th style="text-align:right;">
cv_min
</th>
<th style="text-align:right;">
test_1se
</th>
<th style="text-align:right;">
test_min
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Sim_01
</td>
<td style="text-align:right;">
25
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
20
</td>
<td style="text-align:right;">
40
</td>
<td style="text-align:right;">
0.32
</td>
<td style="text-align:right;">
0.26
</td>
<td style="text-align:right;">
0.30
</td>
<td style="text-align:right;">
0.31
</td>
</tr>
<tr>
<td style="text-align:left;">
Sim_01
</td>
<td style="text-align:right;">
25
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
26
</td>
<td style="text-align:right;">
26
</td>
<td style="text-align:right;">
0.22
</td>
<td style="text-align:right;">
0.22
</td>
<td style="text-align:right;">
0.29
</td>
<td style="text-align:right;">
0.29
</td>
</tr>
<tr>
<td style="text-align:left;">
Sim_01
</td>
<td style="text-align:right;">
25
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
32
</td>
<td style="text-align:right;">
0.34
</td>
<td style="text-align:right;">
0.30
</td>
<td style="text-align:right;">
0.35
</td>
<td style="text-align:right;">
0.31
</td>
</tr>
<tr>
<td style="text-align:left;">
Sim_01
</td>
<td style="text-align:right;">
25
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
24
</td>
<td style="text-align:right;">
0.36
</td>
<td style="text-align:right;">
0.30
</td>
<td style="text-align:right;">
0.32
</td>
<td style="text-align:right;">
0.29
</td>
</tr>
<tr>
<td style="text-align:left;">
Sim_01
</td>
<td style="text-align:right;">
25
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
27
</td>
<td style="text-align:right;">
32
</td>
<td style="text-align:right;">
0.26
</td>
<td style="text-align:right;">
0.22
</td>
<td style="text-align:right;">
0.30
</td>
<td style="text-align:right;">
0.31
</td>
</tr>
<tr>
<td style="text-align:left;">
Sim_01
</td>
<td style="text-align:right;">
25
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
20
</td>
<td style="text-align:right;">
27
</td>
<td style="text-align:right;">
0.36
</td>
<td style="text-align:right;">
0.32
</td>
<td style="text-align:right;">
0.30
</td>
<td style="text-align:right;">
0.30
</td>
</tr>
</tbody>
</table>
<div id="simulation-results---look-at-one-simulation" class="section level3" number="5.6.1">
<h3 number="5.6.1"><span class="header-section-number">5.6.1</span> Simulation Results - look at one simulation</h3>
<div id="model-accuracy-assessment" class="section level4" number="5.6.1.1">
<h4 number="5.6.1.1"><span class="header-section-number">5.6.1.1</span> Model Accuracy Assessment</h4>
<p>First examine results for one simulation run. In the figures that follow,
each boxplot summarized 30 repeated cross validation runs performed on a
fixed random selection of Control and Affected samples. Recall that as
we move from 25 to 50, etc., the sample sets are growing to emulate an
accrual of samples over time.</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb103-2"><a href="#cb103-2"></a></span>
<span id="cb103-3"><a href="#cb103-3"></a><span class="co"># get full model cv error ref</span></span>
<span id="cb103-4"><a href="#cb103-4"></a>error_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_lassoAll_lst,</span>
<span id="cb103-5"><a href="#cb103-5"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb103-6"><a href="#cb103-6"></a>error_1se_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(error_1se_vec, <span class="dt">prob=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>)        </span>
<span id="cb103-7"><a href="#cb103-7"></a></span>
<span id="cb103-8"><a href="#cb103-8"></a>error_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_lassoAll_lst,</span>
<span id="cb103-9"><a href="#cb103-9"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb103-10"><a href="#cb103-10"></a>error_min_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(error_min_vec, <span class="dt">prob=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>)        </span>
<span id="cb103-11"><a href="#cb103-11"></a></span>
<span id="cb103-12"><a href="#cb103-12"></a><span class="co"># Utility objects</span></span>
<span id="cb103-13"><a href="#cb103-13"></a>SIZE0 &lt;-<span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_pad</span>(SIZE, <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>)</span>
<span id="cb103-14"><a href="#cb103-14"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sim_control_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>,SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb103-15"><a href="#cb103-15"></a></span>
<span id="cb103-16"><a href="#cb103-16"></a></span>
<span id="cb103-17"><a href="#cb103-17"></a><span class="co">#SIM &lt;- &quot;Sim_01&quot;</span></span>
<span id="cb103-18"><a href="#cb103-18"></a></span>
<span id="cb103-19"><a href="#cb103-19"></a><span class="cf">for</span>(SIM <span class="cf">in</span> <span class="kw">unique</span>(hcc5hmC_lasso_sim_results_frm<span class="op">$</span>SimNo)[<span class="dv">1</span>]){</span>
<span id="cb103-20"><a href="#cb103-20"></a></span>
<span id="cb103-21"><a href="#cb103-21"></a>SimNum &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">sub</span>(<span class="st">&#39;Sim_&#39;</span>,<span class="st">&#39;&#39;</span>,SIM))</span>
<span id="cb103-22"><a href="#cb103-22"></a></span>
<span id="cb103-23"><a href="#cb103-23"></a>simNo_results_frm &lt;-<span class="st"> </span>hcc5hmC_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(SimNo<span class="op">==</span>SIM)</span>
<span id="cb103-24"><a href="#cb103-24"></a></span>
<span id="cb103-25"><a href="#cb103-25"></a></span>
<span id="cb103-26"><a href="#cb103-26"></a><span class="co"># errors</span></span>
<span id="cb103-27"><a href="#cb103-27"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb103-28"><a href="#cb103-28"></a><span class="co">###################</span></span>
<span id="cb103-29"><a href="#cb103-29"></a><span class="co"># 1se</span></span>
<span id="cb103-30"><a href="#cb103-30"></a><span class="co">####################</span></span>
<span id="cb103-31"><a href="#cb103-31"></a>cv_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb103-32"><a href="#cb103-32"></a> <span class="kw">split</span>(cv_1se, Size))</span>
<span id="cb103-33"><a href="#cb103-33"></a><span class="kw">names</span>(cv_1se_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(cv_1se_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb103-34"><a href="#cb103-34"></a></span>
<span id="cb103-35"><a href="#cb103-35"></a>test_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb103-36"><a href="#cb103-36"></a> <span class="kw">split</span>(test_1se, Size))</span>
<span id="cb103-37"><a href="#cb103-37"></a><span class="kw">names</span>(test_1se_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(test_1se_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb103-38"><a href="#cb103-38"></a></span>
<span id="cb103-39"><a href="#cb103-39"></a>error_1se_lst &lt;-<span class="st"> </span><span class="kw">c</span>(cv_1se_lst, test_1se_lst)</span>
<span id="cb103-40"><a href="#cb103-40"></a>error_1se_lst &lt;-<span class="st"> </span>error_1se_lst[<span class="kw">order</span>(<span class="kw">names</span>(error_1se_lst))]</span>
<span id="cb103-41"><a href="#cb103-41"></a></span>
<span id="cb103-42"><a href="#cb103-42"></a><span class="kw">boxplot</span>(error_1se_lst, </span>
<span id="cb103-43"><a href="#cb103-43"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>), </span>
<span id="cb103-44"><a href="#cb103-44"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>),</span>
<span id="cb103-45"><a href="#cb103-45"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb103-46"><a href="#cb103-46"></a>)</span>
<span id="cb103-47"><a href="#cb103-47"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T,  <span class="st">&#39;Misclassification Error&#39;</span>)</span>
<span id="cb103-48"><a href="#cb103-48"></a></span>
<span id="cb103-49"><a href="#cb103-49"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb103-50"><a href="#cb103-50"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb103-51"><a href="#cb103-51"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_lst)), </span>
<span id="cb103-52"><a href="#cb103-52"></a>  SIZE0</span>
<span id="cb103-53"><a href="#cb103-53"></a> )</span>
<span id="cb103-54"><a href="#cb103-54"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb103-55"><a href="#cb103-55"></a><span class="kw">abline</span>(<span class="dt">h=</span> error_1se_q2, <span class="dt">col =</span> <span class="st">&#39;red&#39;</span>)</span>
<span id="cb103-56"><a href="#cb103-56"></a><span class="kw">legend</span>(<span class="st">&#39;topright&#39;</span>, </span>
<span id="cb103-57"><a href="#cb103-57"></a>   <span class="co">#title=&#39;1se errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb103-58"><a href="#cb103-58"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb103-59"><a href="#cb103-59"></a>   <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&#39;cv error&#39;</span>, <span class="st">&#39;test set&#39;</span>),</span>
<span id="cb103-60"><a href="#cb103-60"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb103-61"><a href="#cb103-61"></a> )</span>
<span id="cb103-62"><a href="#cb103-62"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;one se lambda models&#39;</span>))</span>
<span id="cb103-63"><a href="#cb103-63"></a></span>
<span id="cb103-64"><a href="#cb103-64"></a>SKIP  &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb103-65"><a href="#cb103-65"></a><span class="co"># Add qual annotation</span></span>
<span id="cb103-66"><a href="#cb103-66"></a>control_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(sim_control_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb103-67"><a href="#cb103-67"></a>affected_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(sim_affected_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb103-68"><a href="#cb103-68"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb103-69"><a href="#cb103-69"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb103-70"><a href="#cb103-70"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_lst)),</span>
<span id="cb103-71"><a href="#cb103-71"></a>  <span class="kw">round</span>(control_qual_vec, <span class="dv">2</span>)</span>
<span id="cb103-72"><a href="#cb103-72"></a> )</span>
<span id="cb103-73"><a href="#cb103-73"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb103-74"><a href="#cb103-74"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb103-75"><a href="#cb103-75"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_lst)),</span>
<span id="cb103-76"><a href="#cb103-76"></a>  <span class="kw">round</span>(affected_qual_vec, <span class="dv">2</span>)</span>
<span id="cb103-77"><a href="#cb103-77"></a> )</span>
<span id="cb103-78"><a href="#cb103-78"></a>}<span class="co">#SKIP</span></span>
<span id="cb103-79"><a href="#cb103-79"></a></span>
<span id="cb103-80"><a href="#cb103-80"></a><span class="co"># min</span></span>
<span id="cb103-81"><a href="#cb103-81"></a><span class="co">####################</span></span>
<span id="cb103-82"><a href="#cb103-82"></a>cv_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb103-83"><a href="#cb103-83"></a> <span class="kw">split</span>(cv_min, Size))</span>
<span id="cb103-84"><a href="#cb103-84"></a><span class="kw">names</span>(cv_min_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(cv_min_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb103-85"><a href="#cb103-85"></a></span>
<span id="cb103-86"><a href="#cb103-86"></a>test_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb103-87"><a href="#cb103-87"></a> <span class="kw">split</span>(test_min, Size))</span>
<span id="cb103-88"><a href="#cb103-88"></a><span class="kw">names</span>(test_min_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(test_min_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb103-89"><a href="#cb103-89"></a></span>
<span id="cb103-90"><a href="#cb103-90"></a>error_min_lst &lt;-<span class="st"> </span><span class="kw">c</span>(cv_min_lst, test_min_lst)</span>
<span id="cb103-91"><a href="#cb103-91"></a>error_min_lst &lt;-<span class="st"> </span>error_min_lst[<span class="kw">order</span>(<span class="kw">names</span>(error_min_lst))]</span>
<span id="cb103-92"><a href="#cb103-92"></a></span>
<span id="cb103-93"><a href="#cb103-93"></a><span class="kw">boxplot</span>(error_min_lst, </span>
<span id="cb103-94"><a href="#cb103-94"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>), </span>
<span id="cb103-95"><a href="#cb103-95"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>),</span>
<span id="cb103-96"><a href="#cb103-96"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb103-97"><a href="#cb103-97"></a>)</span>
<span id="cb103-98"><a href="#cb103-98"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T,  <span class="st">&#39;Misclassification Error&#39;</span>)</span>
<span id="cb103-99"><a href="#cb103-99"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb103-100"><a href="#cb103-100"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb103-101"><a href="#cb103-101"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_lst)), </span>
<span id="cb103-102"><a href="#cb103-102"></a>  SIZE0</span>
<span id="cb103-103"><a href="#cb103-103"></a> )</span>
<span id="cb103-104"><a href="#cb103-104"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb103-105"><a href="#cb103-105"></a><span class="kw">abline</span>(<span class="dt">h=</span> error_min_q2, <span class="dt">col =</span> <span class="st">&#39;red&#39;</span>)</span>
<span id="cb103-106"><a href="#cb103-106"></a><span class="kw">legend</span>(<span class="st">&#39;topright&#39;</span>, </span>
<span id="cb103-107"><a href="#cb103-107"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb103-108"><a href="#cb103-108"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb103-109"><a href="#cb103-109"></a>   <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&#39;cv error&#39;</span>, <span class="st">&#39;test set&#39;</span>),</span>
<span id="cb103-110"><a href="#cb103-110"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb103-111"><a href="#cb103-111"></a> )</span>
<span id="cb103-112"><a href="#cb103-112"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;min lambda models&#39;</span>))</span>
<span id="cb103-113"><a href="#cb103-113"></a></span>
<span id="cb103-114"><a href="#cb103-114"></a>SKIP  &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb103-115"><a href="#cb103-115"></a><span class="co"># Add qual annotation</span></span>
<span id="cb103-116"><a href="#cb103-116"></a>control_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(sim_control_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb103-117"><a href="#cb103-117"></a>affected_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(sim_affected_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb103-118"><a href="#cb103-118"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb103-119"><a href="#cb103-119"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb103-120"><a href="#cb103-120"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_lst)),</span>
<span id="cb103-121"><a href="#cb103-121"></a>  <span class="kw">round</span>(control_qual_vec, <span class="dv">2</span>)</span>
<span id="cb103-122"><a href="#cb103-122"></a> )</span>
<span id="cb103-123"><a href="#cb103-123"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb103-124"><a href="#cb103-124"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb103-125"><a href="#cb103-125"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_lst)),</span>
<span id="cb103-126"><a href="#cb103-126"></a>  <span class="kw">round</span>(affected_qual_vec, <span class="dv">2</span>)</span>
<span id="cb103-127"><a href="#cb103-127"></a> )</span>
<span id="cb103-128"><a href="#cb103-128"></a>}<span class="co">#SKIP</span></span>
<span id="cb103-129"><a href="#cb103-129"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, <span class="kw">paste</span>(<span class="st">&#39;Sim =&#39;</span>,  SIM))</span>
<span id="cb103-130"><a href="#cb103-130"></a></span>
<span id="cb103-131"><a href="#cb103-131"></a>} <span class="co"># for(SIM</span></span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuite-lasso-simRes-errors-bySim-1.png" alt="lasso Model Errors by Sample Size" width="960" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuite-lasso-simRes-errors-bySim)lasso Model Errors by Sample Size
</p>
</div>
<p>In this one simulation, we see:</p>
<ul>
<li><p>Model accuracy increases with sample size, with minimal improvement going from N=200 to N=300.</p></li>
<li><p>CV error rates tend to be pessimistic, especially for the small sample sizes. This is odd
and may be related to sample consistency.</p></li>
<li><p>There isn’t much to chose from between the one standard error and the minimum lambda models. The
latter may show lower propensity to produce optimistic cv error rates.</p></li>
</ul>
</div>
<div id="feature-selection" class="section level4" number="5.6.1.2">
<h4 number="5.6.1.2"><span class="header-section-number">5.6.1.2</span> Feature Selection</h4>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb104-2"><a href="#cb104-2"></a></span>
<span id="cb104-3"><a href="#cb104-3"></a><span class="co"># get full model nzero ref</span></span>
<span id="cb104-4"><a href="#cb104-4"></a>nzero_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_lassoAll_lst,</span>
<span id="cb104-5"><a href="#cb104-5"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb104-6"><a href="#cb104-6"></a>nzero_1se_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(nzero_1se_vec, <span class="dt">prob=</span><span class="kw">c</span>(<span class="dv">2</span>)<span class="op">/</span><span class="dv">4</span>)</span>
<span id="cb104-7"><a href="#cb104-7"></a></span>
<span id="cb104-8"><a href="#cb104-8"></a>nzero_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_lassoAll_lst,</span>
<span id="cb104-9"><a href="#cb104-9"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb104-10"><a href="#cb104-10"></a>nzero_min_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(nzero_min_vec, <span class="dt">prob=</span><span class="kw">c</span>(<span class="dv">2</span>)<span class="op">/</span><span class="dv">4</span>)</span>
<span id="cb104-11"><a href="#cb104-11"></a></span>
<span id="cb104-12"><a href="#cb104-12"></a><span class="co"># Utility objects</span></span>
<span id="cb104-13"><a href="#cb104-13"></a>SIZE0 &lt;-<span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_pad</span>(SIZE, <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>)</span>
<span id="cb104-14"><a href="#cb104-14"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sim_control_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>,SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb104-15"><a href="#cb104-15"></a></span>
<span id="cb104-16"><a href="#cb104-16"></a></span>
<span id="cb104-17"><a href="#cb104-17"></a><span class="co">#SIM &lt;- &quot;Sim_01&quot;</span></span>
<span id="cb104-18"><a href="#cb104-18"></a></span>
<span id="cb104-19"><a href="#cb104-19"></a><span class="cf">for</span>(SIM <span class="cf">in</span> <span class="kw">unique</span>(hcc5hmC_lasso_sim_results_frm<span class="op">$</span>SimNo)[<span class="dv">1</span>]){</span>
<span id="cb104-20"><a href="#cb104-20"></a></span>
<span id="cb104-21"><a href="#cb104-21"></a>SimNum &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">sub</span>(<span class="st">&#39;Sim_&#39;</span>,<span class="st">&#39;&#39;</span>,SIM))</span>
<span id="cb104-22"><a href="#cb104-22"></a></span>
<span id="cb104-23"><a href="#cb104-23"></a>simNo_results_frm &lt;-<span class="st"> </span>hcc5hmC_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(SimNo<span class="op">==</span>SIM)</span>
<span id="cb104-24"><a href="#cb104-24"></a></span>
<span id="cb104-25"><a href="#cb104-25"></a></span>
<span id="cb104-26"><a href="#cb104-26"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb104-27"><a href="#cb104-27"></a><span class="co">###################</span></span>
<span id="cb104-28"><a href="#cb104-28"></a><span class="co"># 1se</span></span>
<span id="cb104-29"><a href="#cb104-29"></a><span class="co">####################</span></span>
<span id="cb104-30"><a href="#cb104-30"></a><span class="co"># selected feature counts</span></span>
<span id="cb104-31"><a href="#cb104-31"></a>p_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb104-32"><a href="#cb104-32"></a> <span class="kw">split</span>(p_1se, Size))</span>
<span id="cb104-33"><a href="#cb104-33"></a><span class="kw">names</span>(p_1se_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(p_1se_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_p&#39;</span>)</span>
<span id="cb104-34"><a href="#cb104-34"></a></span>
<span id="cb104-35"><a href="#cb104-35"></a><span class="co"># get selected features that are part of lasso_gene_sign_1se_vec</span></span>
<span id="cb104-36"><a href="#cb104-36"></a><span class="co"># - the signature selected genes</span></span>
<span id="cb104-37"><a href="#cb104-37"></a>sign_genes_1se_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(simNo_results_frm), <span class="cf">function</span>(RR)</span>
<span id="cb104-38"><a href="#cb104-38"></a>    <span class="kw">intersect</span>(<span class="kw">unlist</span>(simNo_results_frm[RR, <span class="st">&#39;genes_1se&#39;</span>]), lasso_gene_sign_1se_vec))</span>
<span id="cb104-39"><a href="#cb104-39"></a></span>
<span id="cb104-40"><a href="#cb104-40"></a>sign_p_1se_lst &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="kw">sapply</span>(sign_genes_1se_lst, length), simNo_results_frm<span class="op">$</span>Size)</span>
<span id="cb104-41"><a href="#cb104-41"></a><span class="kw">names</span>(sign_p_1se_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(sign_p_1se_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_signP&#39;</span>)</span>
<span id="cb104-42"><a href="#cb104-42"></a></span>
<span id="cb104-43"><a href="#cb104-43"></a></span>
<span id="cb104-44"><a href="#cb104-44"></a>p_singP_1se_lst &lt;-<span class="st"> </span><span class="kw">c</span>(p_1se_lst, sign_p_1se_lst)</span>
<span id="cb104-45"><a href="#cb104-45"></a>p_singP_1se_lst &lt;-<span class="st"> </span>p_singP_1se_lst[<span class="kw">order</span>(<span class="kw">names</span>(p_singP_1se_lst))]</span>
<span id="cb104-46"><a href="#cb104-46"></a></span>
<span id="cb104-47"><a href="#cb104-47"></a><span class="kw">boxplot</span>(p_singP_1se_lst,</span>
<span id="cb104-48"><a href="#cb104-48"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb104-49"><a href="#cb104-49"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">200</span>),</span>
<span id="cb104-50"><a href="#cb104-50"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb104-51"><a href="#cb104-51"></a>)</span>
<span id="cb104-52"><a href="#cb104-52"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T,  <span class="st">&#39;number of selected features&#39;</span>)</span>
<span id="cb104-53"><a href="#cb104-53"></a></span>
<span id="cb104-54"><a href="#cb104-54"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb104-55"><a href="#cb104-55"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb104-56"><a href="#cb104-56"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_lst)),</span>
<span id="cb104-57"><a href="#cb104-57"></a>  SIZE0</span>
<span id="cb104-58"><a href="#cb104-58"></a> )</span>
<span id="cb104-59"><a href="#cb104-59"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb104-60"><a href="#cb104-60"></a><span class="co">#abline(h= nzero_1se_q2, col = &#39;red&#39;)</span></span>
<span id="cb104-61"><a href="#cb104-61"></a><span class="kw">legend</span>(<span class="st">&#39;topleft&#39;</span>,</span>
<span id="cb104-62"><a href="#cb104-62"></a>   <span class="co">#title=&#39;1se errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb104-63"><a href="#cb104-63"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>),</span>
<span id="cb104-64"><a href="#cb104-64"></a>   <span class="dt">legend=</span> <span class="kw">c</span>(<span class="st">&#39;selected genes&#39;</span>,<span class="st">&#39;signature genes&#39;</span>),</span>
<span id="cb104-65"><a href="#cb104-65"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb104-66"><a href="#cb104-66"></a> )</span>
<span id="cb104-67"><a href="#cb104-67"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;one se lambda models&#39;</span>))</span>
<span id="cb104-68"><a href="#cb104-68"></a></span>
<span id="cb104-69"><a href="#cb104-69"></a></span>
<span id="cb104-70"><a href="#cb104-70"></a>SKIP  &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb104-71"><a href="#cb104-71"></a><span class="co"># Add qual annotation</span></span>
<span id="cb104-72"><a href="#cb104-72"></a>control_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(sim_control_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb104-73"><a href="#cb104-73"></a>affected_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(sim_affected_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb104-74"><a href="#cb104-74"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb104-75"><a href="#cb104-75"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb104-76"><a href="#cb104-76"></a>  <span class="dt">at =</span>  <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_lst)),</span>
<span id="cb104-77"><a href="#cb104-77"></a>  <span class="kw">round</span>(control_qual_vec, <span class="dv">2</span>)</span>
<span id="cb104-78"><a href="#cb104-78"></a> )</span>
<span id="cb104-79"><a href="#cb104-79"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb104-80"><a href="#cb104-80"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb104-81"><a href="#cb104-81"></a>  <span class="dt">at =</span>  <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_lst)),</span>
<span id="cb104-82"><a href="#cb104-82"></a>  <span class="kw">round</span>(affected_qual_vec, <span class="dv">2</span>)</span>
<span id="cb104-83"><a href="#cb104-83"></a> )</span>
<span id="cb104-84"><a href="#cb104-84"></a>}<span class="co">#SKIP</span></span>
<span id="cb104-85"><a href="#cb104-85"></a></span>
<span id="cb104-86"><a href="#cb104-86"></a><span class="co">###################</span></span>
<span id="cb104-87"><a href="#cb104-87"></a><span class="co"># min</span></span>
<span id="cb104-88"><a href="#cb104-88"></a><span class="co">####################</span></span>
<span id="cb104-89"><a href="#cb104-89"></a><span class="co"># selected feature counts</span></span>
<span id="cb104-90"><a href="#cb104-90"></a>p_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb104-91"><a href="#cb104-91"></a> <span class="kw">split</span>(p_min, Size))</span>
<span id="cb104-92"><a href="#cb104-92"></a><span class="kw">names</span>(p_min_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(p_min_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_p&#39;</span>)</span>
<span id="cb104-93"><a href="#cb104-93"></a></span>
<span id="cb104-94"><a href="#cb104-94"></a><span class="co"># get selected features that are part of lasso_gene_sign_min_vec</span></span>
<span id="cb104-95"><a href="#cb104-95"></a><span class="co"># - the signature selected genes</span></span>
<span id="cb104-96"><a href="#cb104-96"></a>sign_genes_min_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(simNo_results_frm), <span class="cf">function</span>(RR)</span>
<span id="cb104-97"><a href="#cb104-97"></a>    <span class="kw">intersect</span>(<span class="kw">unlist</span>(simNo_results_frm[RR, <span class="st">&#39;genes_min&#39;</span>]), lasso_gene_sign_min_vec))</span>
<span id="cb104-98"><a href="#cb104-98"></a></span>
<span id="cb104-99"><a href="#cb104-99"></a>sign_p_min_lst &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="kw">sapply</span>(sign_genes_min_lst, length), simNo_results_frm<span class="op">$</span>Size)</span>
<span id="cb104-100"><a href="#cb104-100"></a><span class="kw">names</span>(sign_p_min_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(sign_p_min_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_signP&#39;</span>)</span>
<span id="cb104-101"><a href="#cb104-101"></a></span>
<span id="cb104-102"><a href="#cb104-102"></a></span>
<span id="cb104-103"><a href="#cb104-103"></a>p_singP_min_lst &lt;-<span class="st"> </span><span class="kw">c</span>(p_min_lst, sign_p_min_lst)</span>
<span id="cb104-104"><a href="#cb104-104"></a>p_singP_min_lst &lt;-<span class="st"> </span>p_singP_min_lst[<span class="kw">order</span>(<span class="kw">names</span>(p_singP_min_lst))]</span>
<span id="cb104-105"><a href="#cb104-105"></a></span>
<span id="cb104-106"><a href="#cb104-106"></a><span class="kw">boxplot</span>(p_singP_min_lst,</span>
<span id="cb104-107"><a href="#cb104-107"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb104-108"><a href="#cb104-108"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">200</span>),</span>
<span id="cb104-109"><a href="#cb104-109"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb104-110"><a href="#cb104-110"></a>)</span>
<span id="cb104-111"><a href="#cb104-111"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb104-112"><a href="#cb104-112"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb104-113"><a href="#cb104-113"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_lst)),</span>
<span id="cb104-114"><a href="#cb104-114"></a>  SIZE0</span>
<span id="cb104-115"><a href="#cb104-115"></a> )</span>
<span id="cb104-116"><a href="#cb104-116"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb104-117"><a href="#cb104-117"></a><span class="co">#abline(h= nzero_min_q2, col = &#39;red&#39;)</span></span>
<span id="cb104-118"><a href="#cb104-118"></a><span class="kw">legend</span>(<span class="st">&#39;topleft&#39;</span>,</span>
<span id="cb104-119"><a href="#cb104-119"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb104-120"><a href="#cb104-120"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>),</span>
<span id="cb104-121"><a href="#cb104-121"></a>   <span class="dt">legend=</span> <span class="kw">c</span>(<span class="st">&#39;selected genes&#39;</span>,<span class="st">&#39;signature genes&#39;</span>),</span>
<span id="cb104-122"><a href="#cb104-122"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb104-123"><a href="#cb104-123"></a> )</span>
<span id="cb104-124"><a href="#cb104-124"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;min lambda models&#39;</span>))</span>
<span id="cb104-125"><a href="#cb104-125"></a></span>
<span id="cb104-126"><a href="#cb104-126"></a>SKIP  &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb104-127"><a href="#cb104-127"></a><span class="co"># Add qual annotation</span></span>
<span id="cb104-128"><a href="#cb104-128"></a>control_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(sim_control_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb104-129"><a href="#cb104-129"></a>affected_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(sim_affected_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb104-130"><a href="#cb104-130"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb104-131"><a href="#cb104-131"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb104-132"><a href="#cb104-132"></a>  <span class="dt">at =</span>  <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_lst)),</span>
<span id="cb104-133"><a href="#cb104-133"></a>  <span class="kw">round</span>(control_qual_vec, <span class="dv">2</span>)</span>
<span id="cb104-134"><a href="#cb104-134"></a> )</span>
<span id="cb104-135"><a href="#cb104-135"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb104-136"><a href="#cb104-136"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb104-137"><a href="#cb104-137"></a>  <span class="dt">at =</span>  <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_lst)),</span>
<span id="cb104-138"><a href="#cb104-138"></a>  <span class="kw">round</span>(affected_qual_vec, <span class="dv">2</span>)</span>
<span id="cb104-139"><a href="#cb104-139"></a> )</span>
<span id="cb104-140"><a href="#cb104-140"></a>}<span class="co">#SKIP</span></span>
<span id="cb104-141"><a href="#cb104-141"></a></span>
<span id="cb104-142"><a href="#cb104-142"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, <span class="kw">paste</span>(<span class="st">&#39;Sim =&#39;</span>,  SIM))</span>
<span id="cb104-143"><a href="#cb104-143"></a></span>
<span id="cb104-144"><a href="#cb104-144"></a>} <span class="co"># for(SIM</span></span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuite-lasso-simRes-features-bySim-1.png" alt="lasso Models Selected Features by Sample Size" width="960" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuite-lasso-simRes-features-bySim)lasso Models Selected Features by Sample Size
</p>
</div>
<p>In this one simulation, we see:</p>
<ul>
<li><p>The selected number of features in the smaller sample size analyses
are low, with few features belonging to the core signature identified in the
full data set.</p></li>
<li><p>As the sample size increases the number of features selected in the minimum
lambda model remains variable, but the number of core signature features
selected in the samples of sizes 200 and 300 is stable and between 40 and 50.</p></li>
</ul>
</div>
</div>
<div id="summarize-results-across-simulation-runs." class="section level3" number="5.6.2">
<h3 number="5.6.2"><span class="header-section-number">5.6.2</span> Summarize results across simulation runs.</h3>
<div id="model-accuracy-assessment-1" class="section level4" number="5.6.2.1">
<h4 number="5.6.2.1"><span class="header-section-number">5.6.2.1</span> Model Accuracy Assessment</h4>
<p>Now look across all simulations. In the figures that follow, each boxplot
summarizes the results of 30 simulations. For a give sample size and a
given simulation, each data point is the median across 30 repeated cv runs.</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb105-2"><a href="#cb105-2"></a></span>
<span id="cb105-3"><a href="#cb105-3"></a><span class="co"># get full model cv error ref</span></span>
<span id="cb105-4"><a href="#cb105-4"></a>error_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_lassoAll_lst,</span>
<span id="cb105-5"><a href="#cb105-5"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb105-6"><a href="#cb105-6"></a>error_1se_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(error_1se_vec, <span class="dt">prob=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>)        </span>
<span id="cb105-7"><a href="#cb105-7"></a></span>
<span id="cb105-8"><a href="#cb105-8"></a>error_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_lassoAll_lst,</span>
<span id="cb105-9"><a href="#cb105-9"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb105-10"><a href="#cb105-10"></a>error_min_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(error_min_vec, <span class="dt">prob=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>)        </span>
<span id="cb105-11"><a href="#cb105-11"></a></span>
<span id="cb105-12"><a href="#cb105-12"></a><span class="co"># Utility objects</span></span>
<span id="cb105-13"><a href="#cb105-13"></a>SIZE0 &lt;-<span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_pad</span>(SIZE, <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>)</span>
<span id="cb105-14"><a href="#cb105-14"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sim_control_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>,SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb105-15"><a href="#cb105-15"></a></span>
<span id="cb105-16"><a href="#cb105-16"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb105-17"><a href="#cb105-17"></a><span class="co"># 1se</span></span>
<span id="cb105-18"><a href="#cb105-18"></a><span class="co">#########################################</span></span>
<span id="cb105-19"><a href="#cb105-19"></a><span class="co">## cv</span></span>
<span id="cb105-20"><a href="#cb105-20"></a>cv_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(hcc5hmC_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb105-21"><a href="#cb105-21"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb105-22"><a href="#cb105-22"></a> sizeVal_results_frm &lt;-<span class="st"> </span>hcc5hmC_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb105-23"><a href="#cb105-23"></a> sizeVal_cv_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(cv_1se, SimNo))</span>
<span id="cb105-24"><a href="#cb105-24"></a> <span class="kw">sapply</span>(sizeVal_cv_1se_lst, median)</span>
<span id="cb105-25"><a href="#cb105-25"></a>})</span>
<span id="cb105-26"><a href="#cb105-26"></a><span class="kw">names</span>(cv_1se_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb105-27"><a href="#cb105-27"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(hcc5hmC_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb105-28"><a href="#cb105-28"></a></span>
<span id="cb105-29"><a href="#cb105-29"></a><span class="co">## test</span></span>
<span id="cb105-30"><a href="#cb105-30"></a>test_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(hcc5hmC_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb105-31"><a href="#cb105-31"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb105-32"><a href="#cb105-32"></a> sizeVal_results_frm &lt;-<span class="st"> </span>hcc5hmC_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb105-33"><a href="#cb105-33"></a> sizeVal_test_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(test_1se, SimNo))</span>
<span id="cb105-34"><a href="#cb105-34"></a> <span class="kw">sapply</span>(sizeVal_test_1se_lst, median)</span>
<span id="cb105-35"><a href="#cb105-35"></a>})</span>
<span id="cb105-36"><a href="#cb105-36"></a><span class="kw">names</span>(test_1se_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb105-37"><a href="#cb105-37"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(hcc5hmC_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_test&#39;</span>)</span>
<span id="cb105-38"><a href="#cb105-38"></a></span>
<span id="cb105-39"><a href="#cb105-39"></a></span>
<span id="cb105-40"><a href="#cb105-40"></a>error_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">c</span>(cv_1se_Bysize_lst, test_1se_Bysize_lst)</span>
<span id="cb105-41"><a href="#cb105-41"></a>error_1se_Bysize_lst &lt;-<span class="st"> </span>error_1se_Bysize_lst[<span class="kw">order</span>(<span class="kw">names</span>(error_1se_Bysize_lst))]</span>
<span id="cb105-42"><a href="#cb105-42"></a></span>
<span id="cb105-43"><a href="#cb105-43"></a><span class="kw">boxplot</span>(error_1se_Bysize_lst,</span>
<span id="cb105-44"><a href="#cb105-44"></a>  <span class="dt">col=</span><span class="dv">0</span>,</span>
<span id="cb105-45"><a href="#cb105-45"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb105-46"><a href="#cb105-46"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">.5</span>),</span>
<span id="cb105-47"><a href="#cb105-47"></a>  <span class="dt">outline=</span>F,</span>
<span id="cb105-48"><a href="#cb105-48"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb105-49"><a href="#cb105-49"></a>)</span>
<span id="cb105-50"><a href="#cb105-50"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T,  <span class="st">&#39;Misclassification Error&#39;</span>)</span>
<span id="cb105-51"><a href="#cb105-51"></a></span>
<span id="cb105-52"><a href="#cb105-52"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(error_1se_Bysize_lst))</span>
<span id="cb105-53"><a href="#cb105-53"></a><span class="kw">points</span>(</span>
<span id="cb105-54"><a href="#cb105-54"></a>   <span class="dt">x=</span><span class="kw">jitter</span>(<span class="kw">rep</span>(JJ, <span class="kw">length</span>(error_1se_Bysize_lst[[JJ]])), <span class="dt">amount=</span><span class="fl">0.25</span>), </span>
<span id="cb105-55"><a href="#cb105-55"></a>   <span class="dt">y=</span>error_1se_Bysize_lst[[JJ]], <span class="dt">cex=</span><span class="fl">0.5</span>,</span>
<span id="cb105-56"><a href="#cb105-56"></a>   <span class="dt">col=</span><span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">&#39;cv&#39;</span>, <span class="kw">names</span>(error_1se_Bysize_lst)[JJ]),<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb105-57"><a href="#cb105-57"></a>)</span>
<span id="cb105-58"><a href="#cb105-58"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb105-59"><a href="#cb105-59"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb105-60"><a href="#cb105-60"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_Bysize_lst)),</span>
<span id="cb105-61"><a href="#cb105-61"></a>  SIZE0</span>
<span id="cb105-62"><a href="#cb105-62"></a> )</span>
<span id="cb105-63"><a href="#cb105-63"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_Bysize_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb105-64"><a href="#cb105-64"></a><span class="kw">abline</span>(<span class="dt">h=</span> error_min_q2, <span class="dt">col =</span> <span class="st">&#39;red&#39;</span>)</span>
<span id="cb105-65"><a href="#cb105-65"></a><span class="kw">legend</span>(<span class="st">&#39;topright&#39;</span>,</span>
<span id="cb105-66"><a href="#cb105-66"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb105-67"><a href="#cb105-67"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb105-68"><a href="#cb105-68"></a>   <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&#39;cv error&#39;</span>, <span class="st">&#39;test set&#39;</span>),</span>
<span id="cb105-69"><a href="#cb105-69"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb105-70"><a href="#cb105-70"></a> )</span>
<span id="cb105-71"><a href="#cb105-71"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;one se lambda models&#39;</span>))</span>
<span id="cb105-72"><a href="#cb105-72"></a></span>
<span id="cb105-73"><a href="#cb105-73"></a></span>
<span id="cb105-74"><a href="#cb105-74"></a><span class="co"># min</span></span>
<span id="cb105-75"><a href="#cb105-75"></a><span class="co">#########################################</span></span>
<span id="cb105-76"><a href="#cb105-76"></a><span class="co">## cv</span></span>
<span id="cb105-77"><a href="#cb105-77"></a>cv_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(hcc5hmC_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb105-78"><a href="#cb105-78"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb105-79"><a href="#cb105-79"></a> sizeVal_results_frm &lt;-<span class="st"> </span>hcc5hmC_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb105-80"><a href="#cb105-80"></a> sizeVal_cv_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(cv_min, SimNo))</span>
<span id="cb105-81"><a href="#cb105-81"></a> <span class="kw">sapply</span>(sizeVal_cv_min_lst, median)</span>
<span id="cb105-82"><a href="#cb105-82"></a>})</span>
<span id="cb105-83"><a href="#cb105-83"></a><span class="kw">names</span>(cv_min_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb105-84"><a href="#cb105-84"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(hcc5hmC_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb105-85"><a href="#cb105-85"></a></span>
<span id="cb105-86"><a href="#cb105-86"></a><span class="co">## test</span></span>
<span id="cb105-87"><a href="#cb105-87"></a>test_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(hcc5hmC_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb105-88"><a href="#cb105-88"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb105-89"><a href="#cb105-89"></a> sizeVal_results_frm &lt;-<span class="st"> </span>hcc5hmC_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb105-90"><a href="#cb105-90"></a> sizeVal_test_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(test_min, SimNo))</span>
<span id="cb105-91"><a href="#cb105-91"></a> <span class="kw">sapply</span>(sizeVal_test_min_lst, median)</span>
<span id="cb105-92"><a href="#cb105-92"></a>})</span>
<span id="cb105-93"><a href="#cb105-93"></a><span class="kw">names</span>(test_min_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb105-94"><a href="#cb105-94"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(hcc5hmC_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_test&#39;</span>)</span>
<span id="cb105-95"><a href="#cb105-95"></a></span>
<span id="cb105-96"><a href="#cb105-96"></a></span>
<span id="cb105-97"><a href="#cb105-97"></a>error_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">c</span>(cv_min_Bysize_lst, test_min_Bysize_lst)</span>
<span id="cb105-98"><a href="#cb105-98"></a>error_min_Bysize_lst &lt;-<span class="st"> </span>error_min_Bysize_lst[<span class="kw">order</span>(<span class="kw">names</span>(error_min_Bysize_lst))]</span>
<span id="cb105-99"><a href="#cb105-99"></a></span>
<span id="cb105-100"><a href="#cb105-100"></a><span class="kw">boxplot</span>(error_min_Bysize_lst,</span>
<span id="cb105-101"><a href="#cb105-101"></a>  <span class="dt">col=</span><span class="dv">0</span>,</span>
<span id="cb105-102"><a href="#cb105-102"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb105-103"><a href="#cb105-103"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">.5</span>),</span>
<span id="cb105-104"><a href="#cb105-104"></a>  <span class="dt">outline=</span>F,</span>
<span id="cb105-105"><a href="#cb105-105"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb105-106"><a href="#cb105-106"></a>)</span>
<span id="cb105-107"><a href="#cb105-107"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(error_min_Bysize_lst))</span>
<span id="cb105-108"><a href="#cb105-108"></a><span class="kw">points</span>(</span>
<span id="cb105-109"><a href="#cb105-109"></a>   <span class="dt">x=</span><span class="kw">jitter</span>(<span class="kw">rep</span>(JJ, <span class="kw">length</span>(error_min_Bysize_lst[[JJ]])), <span class="dt">amount=</span><span class="fl">0.25</span>), </span>
<span id="cb105-110"><a href="#cb105-110"></a>   <span class="dt">y=</span>error_min_Bysize_lst[[JJ]], <span class="dt">cex=</span><span class="fl">0.5</span>,</span>
<span id="cb105-111"><a href="#cb105-111"></a>   <span class="dt">col=</span><span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">&#39;cv&#39;</span>, <span class="kw">names</span>(error_min_Bysize_lst)[JJ]),<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb105-112"><a href="#cb105-112"></a>)</span>
<span id="cb105-113"><a href="#cb105-113"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb105-114"><a href="#cb105-114"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb105-115"><a href="#cb105-115"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_Bysize_lst)),</span>
<span id="cb105-116"><a href="#cb105-116"></a>  SIZE0</span>
<span id="cb105-117"><a href="#cb105-117"></a> )</span>
<span id="cb105-118"><a href="#cb105-118"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_Bysize_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb105-119"><a href="#cb105-119"></a><span class="kw">abline</span>(<span class="dt">h=</span> error_min_q2, <span class="dt">col =</span> <span class="st">&#39;red&#39;</span>)</span>
<span id="cb105-120"><a href="#cb105-120"></a><span class="kw">legend</span>(<span class="st">&#39;topright&#39;</span>,</span>
<span id="cb105-121"><a href="#cb105-121"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb105-122"><a href="#cb105-122"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb105-123"><a href="#cb105-123"></a>   <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&#39;cv error&#39;</span>, <span class="st">&#39;test set&#39;</span>),</span>
<span id="cb105-124"><a href="#cb105-124"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb105-125"><a href="#cb105-125"></a> )</span>
<span id="cb105-126"><a href="#cb105-126"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;min lambda models&#39;</span>))</span>
<span id="cb105-127"><a href="#cb105-127"></a></span>
<span id="cb105-128"><a href="#cb105-128"></a></span>
<span id="cb105-129"><a href="#cb105-129"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, <span class="kw">paste</span>(<span class="st">&#39;lasso fit error rates summarized across simulations&#39;</span>))</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuite-lasso-simRes-errors-overSim-1.png" alt="lasso Model Errors by Sample Size" width="960" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuite-lasso-simRes-errors-overSim)lasso Model Errors by Sample Size
</p>
</div>
<p><br/></p>
<ul>
<li><p>For the smaller samples sizes, cv error rates for the minimum lambda models tend to be
optimistic.</p></li>
<li><p><strong>For lower sample sizes, assess performance is quite variable</strong>. This
is key: with N=25 cases and controls, the estimated classification error
rate produced by a given cohort can range from below 0.2 to higher than 0.5,
or basically no indication of discrimination value in the analyzed features.</p></li>
</ul>
<p><br/></p>
<p>To appreciate how much variability can be encountered as samples are accrued over time
we need to look at a typical path the assessed model accuracy estimates might take.</p>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb106-2"><a href="#cb106-2"></a></span>
<span id="cb106-3"><a href="#cb106-3"></a>error_1se_Bysize_mtx &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="st">&#39;cbind&#39;</span>, <span class="kw">lapply</span>(error_1se_Bysize_lst, <span class="cf">function</span>(LL) LL))</span>
<span id="cb106-4"><a href="#cb106-4"></a></span>
<span id="cb106-5"><a href="#cb106-5"></a>cv_error_1se_Bysize_mtx &lt;-<span class="st"> </span>error_1se_Bysize_mtx[,<span class="kw">grep</span>(<span class="st">&#39;_cv&#39;</span>, <span class="kw">colnames</span>(error_1se_Bysize_mtx))]</span>
<span id="cb106-6"><a href="#cb106-6"></a></span>
<span id="cb106-7"><a href="#cb106-7"></a><span class="kw">plot</span>(<span class="dt">x=</span><span class="kw">c</span>(<span class="dv">1</span>, <span class="kw">ncol</span>(cv_error_1se_Bysize_mtx)), <span class="dt">y=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.6</span>), </span>
<span id="cb106-8"><a href="#cb106-8"></a>  <span class="dt">xlab=</span><span class="st">&#39;sample size&#39;</span>, <span class="dt">ylab=</span><span class="st">&#39;Misclassification Error&#39;</span>, </span>
<span id="cb106-9"><a href="#cb106-9"></a>  <span class="dt">type=</span><span class="st">&#39;n&#39;</span>, <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>)</span>
<span id="cb106-10"><a href="#cb106-10"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(cv_error_1se_Bysize_mtx), </span>
<span id="cb106-11"><a href="#cb106-11"></a> <span class="dt">labels=</span><span class="kw">sub</span>(<span class="st">&#39;_cv&#39;</span>,<span class="st">&#39;&#39;</span>,<span class="kw">colnames</span>(cv_error_1se_Bysize_mtx)))</span>
<span id="cb106-12"><a href="#cb106-12"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">15</span>)</span>
<span id="cb106-13"><a href="#cb106-13"></a><span class="kw">lines</span>(<span class="dt">x=</span><span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(cv_error_1se_Bysize_mtx), <span class="dt">y=</span>cv_error_1se_Bysize_mtx[JJ,],</span>
<span id="cb106-14"><a href="#cb106-14"></a> <span class="dt">type=</span><span class="st">&#39;b&#39;</span>, <span class="dt">pch=</span>JJ, <span class="dt">col=</span>JJ)</span>
<span id="cb106-15"><a href="#cb106-15"></a><span class="kw">title</span>(<span class="st">&#39;Example Misclassification Error Paths&#39;</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuite-lasso-simRes-errorsPath-overSim-1.png" alt="lasso Model Error Paths" width="960" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuite-lasso-simRes-errorsPath-overSim)lasso Model Error Paths
</p>
</div>
<p>We see how erratic the assessed model accuracy can be when sample sizes are small,
and that it would be hard to guess the ultimate level of accuracy the
is achievable, or the number of samples required to get a reasonable
estimate of the achievable level of accuracy.</p>
<p><br/></p>
<p><br/></p>
</div>
<div id="feature-selection-1" class="section level4" number="5.6.2.2">
<h4 number="5.6.2.2"><span class="header-section-number">5.6.2.2</span> Feature Selection</h4>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb107-2"><a href="#cb107-2"></a></span>
<span id="cb107-3"><a href="#cb107-3"></a><span class="co"># Utility objects</span></span>
<span id="cb107-4"><a href="#cb107-4"></a>SIZE0 &lt;-<span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_pad</span>(SIZE, <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>)</span>
<span id="cb107-5"><a href="#cb107-5"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sim_control_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>,SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb107-6"><a href="#cb107-6"></a></span>
<span id="cb107-7"><a href="#cb107-7"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb107-8"><a href="#cb107-8"></a><span class="co"># 1se</span></span>
<span id="cb107-9"><a href="#cb107-9"></a><span class="co">#########################################</span></span>
<span id="cb107-10"><a href="#cb107-10"></a><span class="co"># selected features</span></span>
<span id="cb107-11"><a href="#cb107-11"></a>p_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(hcc5hmC_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb107-12"><a href="#cb107-12"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb107-13"><a href="#cb107-13"></a> sizeVal_results_frm &lt;-<span class="st"> </span>hcc5hmC_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb107-14"><a href="#cb107-14"></a> sizeVal_p_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(p_1se, SimNo))</span>
<span id="cb107-15"><a href="#cb107-15"></a> <span class="kw">sapply</span>(sizeVal_p_1se_lst, median)</span>
<span id="cb107-16"><a href="#cb107-16"></a>})</span>
<span id="cb107-17"><a href="#cb107-17"></a><span class="kw">names</span>(p_1se_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb107-18"><a href="#cb107-18"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(hcc5hmC_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_p&#39;</span>)</span>
<span id="cb107-19"><a href="#cb107-19"></a></span>
<span id="cb107-20"><a href="#cb107-20"></a><span class="co"># selected signatue features</span></span>
<span id="cb107-21"><a href="#cb107-21"></a>sign_p_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(hcc5hmC_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb107-22"><a href="#cb107-22"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb107-23"><a href="#cb107-23"></a> sizeVal_results_frm &lt;-<span class="st"> </span>hcc5hmC_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb107-24"><a href="#cb107-24"></a> </span>
<span id="cb107-25"><a href="#cb107-25"></a>  </span>
<span id="cb107-26"><a href="#cb107-26"></a> sizeVal_sign_genes_1se_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sizeVal_results_frm), <span class="cf">function</span>(RR)</span>
<span id="cb107-27"><a href="#cb107-27"></a>    <span class="kw">intersect</span>(<span class="kw">unlist</span>(sizeVal_results_frm[RR, <span class="st">&#39;genes_1se&#39;</span>]), lasso_gene_sign_1se_vec))</span>
<span id="cb107-28"><a href="#cb107-28"></a></span>
<span id="cb107-29"><a href="#cb107-29"></a> sizeVal_sign_p_1se_lst &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="kw">sapply</span>(sizeVal_sign_genes_1se_lst, length),</span>
<span id="cb107-30"><a href="#cb107-30"></a>    sizeVal_results_frm<span class="op">$</span>SimNo)</span>
<span id="cb107-31"><a href="#cb107-31"></a> </span>
<span id="cb107-32"><a href="#cb107-32"></a> <span class="kw">sapply</span>(sizeVal_sign_p_1se_lst, median)</span>
<span id="cb107-33"><a href="#cb107-33"></a>})</span>
<span id="cb107-34"><a href="#cb107-34"></a><span class="kw">names</span>(sign_p_1se_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb107-35"><a href="#cb107-35"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(hcc5hmC_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_signP&#39;</span>)</span>
<span id="cb107-36"><a href="#cb107-36"></a></span>
<span id="cb107-37"><a href="#cb107-37"></a></span>
<span id="cb107-38"><a href="#cb107-38"></a>p_singP_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">c</span>(p_1se_Bysize_lst, sign_p_1se_Bysize_lst)</span>
<span id="cb107-39"><a href="#cb107-39"></a>p_singP_1se_Bysize_lst &lt;-<span class="st"> </span>p_singP_1se_Bysize_lst[<span class="kw">order</span>(<span class="kw">names</span>(p_singP_1se_Bysize_lst))]</span>
<span id="cb107-40"><a href="#cb107-40"></a></span>
<span id="cb107-41"><a href="#cb107-41"></a><span class="kw">boxplot</span>(p_singP_1se_Bysize_lst,</span>
<span id="cb107-42"><a href="#cb107-42"></a>  <span class="dt">col=</span><span class="dv">0</span>,</span>
<span id="cb107-43"><a href="#cb107-43"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb107-44"><a href="#cb107-44"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">200</span>),</span>
<span id="cb107-45"><a href="#cb107-45"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb107-46"><a href="#cb107-46"></a>)</span>
<span id="cb107-47"><a href="#cb107-47"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T,  <span class="st">&#39;number of selected features&#39;</span>)</span>
<span id="cb107-48"><a href="#cb107-48"></a></span>
<span id="cb107-49"><a href="#cb107-49"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(p_singP_1se_Bysize_lst))</span>
<span id="cb107-50"><a href="#cb107-50"></a><span class="kw">points</span>(</span>
<span id="cb107-51"><a href="#cb107-51"></a>   <span class="dt">x=</span><span class="kw">jitter</span>(<span class="kw">rep</span>(JJ, <span class="kw">length</span>(p_singP_1se_Bysize_lst[[JJ]])), <span class="dt">amount=</span><span class="fl">0.25</span>),</span>
<span id="cb107-52"><a href="#cb107-52"></a>   <span class="dt">y=</span>p_singP_1se_Bysize_lst[[JJ]], <span class="dt">cex=</span><span class="fl">0.5</span>,</span>
<span id="cb107-53"><a href="#cb107-53"></a>   <span class="dt">col=</span><span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">&#39;_p&#39;</span>, <span class="kw">names</span>(p_singP_1se_Bysize_lst)[JJ]),<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb107-54"><a href="#cb107-54"></a>)</span>
<span id="cb107-55"><a href="#cb107-55"></a></span>
<span id="cb107-56"><a href="#cb107-56"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb107-57"><a href="#cb107-57"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb107-58"><a href="#cb107-58"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_Bysize_lst)),</span>
<span id="cb107-59"><a href="#cb107-59"></a>  SIZE0</span>
<span id="cb107-60"><a href="#cb107-60"></a> )</span>
<span id="cb107-61"><a href="#cb107-61"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_Bysize_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb107-62"><a href="#cb107-62"></a><span class="co">#abline(h= nzero_1se_q2, col = &#39;red&#39;)</span></span>
<span id="cb107-63"><a href="#cb107-63"></a><span class="kw">legend</span>(<span class="st">&#39;topleft&#39;</span>,</span>
<span id="cb107-64"><a href="#cb107-64"></a>   <span class="co">#title=&#39;1se errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb107-65"><a href="#cb107-65"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>),</span>
<span id="cb107-66"><a href="#cb107-66"></a>   <span class="dt">legend=</span> <span class="kw">c</span>(<span class="st">&#39;selected genes&#39;</span>,<span class="st">&#39;signature genes&#39;</span>),</span>
<span id="cb107-67"><a href="#cb107-67"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb107-68"><a href="#cb107-68"></a> )</span>
<span id="cb107-69"><a href="#cb107-69"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;one se lamdba models&#39;</span>))</span>
<span id="cb107-70"><a href="#cb107-70"></a></span>
<span id="cb107-71"><a href="#cb107-71"></a></span>
<span id="cb107-72"><a href="#cb107-72"></a><span class="co"># min</span></span>
<span id="cb107-73"><a href="#cb107-73"></a><span class="co">#########################################</span></span>
<span id="cb107-74"><a href="#cb107-74"></a><span class="co"># selected features</span></span>
<span id="cb107-75"><a href="#cb107-75"></a>p_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(hcc5hmC_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb107-76"><a href="#cb107-76"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb107-77"><a href="#cb107-77"></a> sizeVal_results_frm &lt;-<span class="st"> </span>hcc5hmC_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb107-78"><a href="#cb107-78"></a> sizeVal_p_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(p_min, SimNo))</span>
<span id="cb107-79"><a href="#cb107-79"></a> <span class="kw">sapply</span>(sizeVal_p_min_lst, median)</span>
<span id="cb107-80"><a href="#cb107-80"></a>})</span>
<span id="cb107-81"><a href="#cb107-81"></a><span class="kw">names</span>(p_min_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb107-82"><a href="#cb107-82"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(hcc5hmC_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_p&#39;</span>)</span>
<span id="cb107-83"><a href="#cb107-83"></a></span>
<span id="cb107-84"><a href="#cb107-84"></a><span class="co"># selected signatue features</span></span>
<span id="cb107-85"><a href="#cb107-85"></a>sign_p_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(hcc5hmC_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb107-86"><a href="#cb107-86"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb107-87"><a href="#cb107-87"></a> sizeVal_results_frm &lt;-<span class="st"> </span>hcc5hmC_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb107-88"><a href="#cb107-88"></a> </span>
<span id="cb107-89"><a href="#cb107-89"></a>  </span>
<span id="cb107-90"><a href="#cb107-90"></a> sizeVal_sign_genes_min_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sizeVal_results_frm), <span class="cf">function</span>(RR)</span>
<span id="cb107-91"><a href="#cb107-91"></a>    <span class="kw">intersect</span>(<span class="kw">unlist</span>(sizeVal_results_frm[RR, <span class="st">&#39;genes_min&#39;</span>]), lasso_gene_sign_min_vec))</span>
<span id="cb107-92"><a href="#cb107-92"></a></span>
<span id="cb107-93"><a href="#cb107-93"></a> sizeVal_sign_p_min_lst &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="kw">sapply</span>(sizeVal_sign_genes_min_lst, length),</span>
<span id="cb107-94"><a href="#cb107-94"></a>    sizeVal_results_frm<span class="op">$</span>SimNo)</span>
<span id="cb107-95"><a href="#cb107-95"></a> </span>
<span id="cb107-96"><a href="#cb107-96"></a> <span class="kw">sapply</span>(sizeVal_sign_p_min_lst, median)</span>
<span id="cb107-97"><a href="#cb107-97"></a>})</span>
<span id="cb107-98"><a href="#cb107-98"></a><span class="kw">names</span>(sign_p_min_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb107-99"><a href="#cb107-99"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(hcc5hmC_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_signP&#39;</span>)</span>
<span id="cb107-100"><a href="#cb107-100"></a></span>
<span id="cb107-101"><a href="#cb107-101"></a></span>
<span id="cb107-102"><a href="#cb107-102"></a>p_singP_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">c</span>(p_min_Bysize_lst, sign_p_min_Bysize_lst)</span>
<span id="cb107-103"><a href="#cb107-103"></a>p_singP_min_Bysize_lst &lt;-<span class="st"> </span>p_singP_min_Bysize_lst[<span class="kw">order</span>(<span class="kw">names</span>(p_singP_min_Bysize_lst))]</span>
<span id="cb107-104"><a href="#cb107-104"></a></span>
<span id="cb107-105"><a href="#cb107-105"></a><span class="kw">boxplot</span>(p_singP_min_Bysize_lst,</span>
<span id="cb107-106"><a href="#cb107-106"></a>  <span class="dt">col=</span><span class="dv">0</span>,</span>
<span id="cb107-107"><a href="#cb107-107"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb107-108"><a href="#cb107-108"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">200</span>),</span>
<span id="cb107-109"><a href="#cb107-109"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb107-110"><a href="#cb107-110"></a>)</span>
<span id="cb107-111"><a href="#cb107-111"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(p_singP_min_Bysize_lst))</span>
<span id="cb107-112"><a href="#cb107-112"></a><span class="kw">points</span>(</span>
<span id="cb107-113"><a href="#cb107-113"></a>   <span class="dt">x=</span><span class="kw">jitter</span>(<span class="kw">rep</span>(JJ, <span class="kw">length</span>(p_singP_min_Bysize_lst[[JJ]])), <span class="dt">amount=</span><span class="fl">0.25</span>),</span>
<span id="cb107-114"><a href="#cb107-114"></a>   <span class="dt">y=</span>p_singP_min_Bysize_lst[[JJ]], <span class="dt">cex=</span><span class="fl">0.5</span>,</span>
<span id="cb107-115"><a href="#cb107-115"></a>   <span class="dt">col=</span><span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">&#39;_p&#39;</span>, <span class="kw">names</span>(p_singP_min_Bysize_lst)[JJ]),<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb107-116"><a href="#cb107-116"></a>)</span>
<span id="cb107-117"><a href="#cb107-117"></a></span>
<span id="cb107-118"><a href="#cb107-118"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb107-119"><a href="#cb107-119"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb107-120"><a href="#cb107-120"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_Bysize_lst)),</span>
<span id="cb107-121"><a href="#cb107-121"></a>  SIZE0</span>
<span id="cb107-122"><a href="#cb107-122"></a> )</span>
<span id="cb107-123"><a href="#cb107-123"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_Bysize_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb107-124"><a href="#cb107-124"></a><span class="co">#abline(h= nzero_min_q2, col = &#39;red&#39;)</span></span>
<span id="cb107-125"><a href="#cb107-125"></a><span class="kw">legend</span>(<span class="st">&#39;topleft&#39;</span>,</span>
<span id="cb107-126"><a href="#cb107-126"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb107-127"><a href="#cb107-127"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>),</span>
<span id="cb107-128"><a href="#cb107-128"></a>   <span class="dt">legend=</span> <span class="kw">c</span>(<span class="st">&#39;selected genes&#39;</span>,<span class="st">&#39;signature genes&#39;</span>),</span>
<span id="cb107-129"><a href="#cb107-129"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb107-130"><a href="#cb107-130"></a> )</span>
<span id="cb107-131"><a href="#cb107-131"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;min lambda models&#39;</span>))</span>
<span id="cb107-132"><a href="#cb107-132"></a></span>
<span id="cb107-133"><a href="#cb107-133"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, <span class="kw">paste</span>(<span class="st">&#39;lasso fit feature selection summarized across simulations&#39;</span>))</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuite-lasso-simRes-features-OverSim-1.png" alt="lasso Models Selected Features by Sample Size" width="960" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuite-lasso-simRes-features-OverSim)lasso Models Selected Features by Sample Size
</p>
</div>
<ul>
<li><p>The number of selected features increase with sample size.</p></li>
<li><p>The number of selected features is quite variable, even for larger sample sizes.</p></li>
<li><p>The number of core signature features among selected features is stable for larger sample
sizes and represents 30 to 40% of the selected features (for N=200 and 300 respectively).</p></li>
</ul>
</div>
</div>
</div>
<div id="effect-of-sample-consistency" class="section level2" number="5.7">
<h2 number="5.7"><span class="header-section-number">5.7</span> Effect of sample consistency</h2>
<p>To see the effect of sample consistency on classification results, we will
focus on the performance of the small sample fits (N=25) where variability is the
greatest, and the context where investigators should be most aware of the
effect of sample selection over and beyond sample size concerns.</p>
<p>In the plot below,</p>
<ul>
<li><p>control_Q and affected_Q are summary measures of
sample set consistencies for the control and affected groups in the different
cv runs.</p></li>
<li><p>1se_cv and 1se_test are the cross-validation and test set error rates, respectively.</p></li>
</ul>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1"></a>control_samp25_qual_vec &lt;-<span class="st"> </span><span class="kw">apply</span>(sim_control_qual_mtx[<span class="dv">1</span><span class="op">:</span><span class="dv">25</span>, ], <span class="dv">2</span>, median)</span>
<span id="cb108-2"><a href="#cb108-2"></a>affected_samp25_qual_vec &lt;-<span class="st"> </span><span class="kw">apply</span>(sim_affected_qual_mtx[<span class="dv">1</span><span class="op">:</span><span class="dv">25</span>, ], <span class="dv">2</span>, median)</span>
<span id="cb108-3"><a href="#cb108-3"></a></span>
<span id="cb108-4"><a href="#cb108-4"></a>samp25_qual_error_mtx &lt;-<span class="st"> </span><span class="kw">cbind</span>(</span>
<span id="cb108-5"><a href="#cb108-5"></a>   <span class="st">`</span><span class="dt">1se_cv</span><span class="st">`</span> =<span class="st"> </span>error_1se_Bysize_lst[[<span class="st">&#39;025_cv&#39;</span>]],</span>
<span id="cb108-6"><a href="#cb108-6"></a>   <span class="st">`</span><span class="dt">1se_test</span><span class="st">`</span> =<span class="st"> </span>error_1se_Bysize_lst[[<span class="st">&#39;025_test&#39;</span>]],</span>
<span id="cb108-7"><a href="#cb108-7"></a>   <span class="st">`</span><span class="dt">control_Q</span><span class="st">`</span> =<span class="st"> </span>control_samp25_qual_vec,</span>
<span id="cb108-8"><a href="#cb108-8"></a>   <span class="st">`</span><span class="dt">affected_Q</span><span class="st">`</span> =<span class="st"> </span>affected_samp25_qual_vec)</span>
<span id="cb108-9"><a href="#cb108-9"></a> </span>
<span id="cb108-10"><a href="#cb108-10"></a></span>
<span id="cb108-11"><a href="#cb108-11"></a><span class="co"># Correlation panel</span></span>
<span id="cb108-12"><a href="#cb108-12"></a>panel.cor &lt;-<span class="st"> </span><span class="cf">function</span>(x, y){</span>
<span id="cb108-13"><a href="#cb108-13"></a>    usr &lt;-<span class="st"> </span><span class="kw">par</span>(<span class="st">&quot;usr&quot;</span>); <span class="kw">on.exit</span>(<span class="kw">par</span>(usr))</span>
<span id="cb108-14"><a href="#cb108-14"></a>    <span class="kw">par</span>(<span class="dt">usr =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>))</span>
<span id="cb108-15"><a href="#cb108-15"></a>    r &lt;-<span class="st"> </span><span class="kw">round</span>(<span class="kw">cor</span>(x, y), <span class="dt">digits=</span><span class="dv">2</span>)</span>
<span id="cb108-16"><a href="#cb108-16"></a>    txt &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;R = &quot;</span>, r)</span>
<span id="cb108-17"><a href="#cb108-17"></a>    cex.cor &lt;-<span class="st"> </span><span class="fl">0.8</span><span class="op">/</span><span class="kw">strwidth</span>(txt)</span>
<span id="cb108-18"><a href="#cb108-18"></a>    <span class="kw">text</span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>, txt, <span class="dt">cex =</span> <span class="fl">1.5</span>) <span class="co">###cex.cor * r)</span></span>
<span id="cb108-19"><a href="#cb108-19"></a>}</span>
<span id="cb108-20"><a href="#cb108-20"></a></span>
<span id="cb108-21"><a href="#cb108-21"></a><span class="kw">pairs</span>(samp25_qual_error_mtx,</span>
<span id="cb108-22"><a href="#cb108-22"></a> <span class="dt">lower.panel =</span> panel.cor)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuite-small-sample-qual-1.png" alt="sample consistency vs classifier performance" width="672" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuite-small-sample-qual)sample consistency vs classifier performance
</p>
</div>
<p>We see that although the sample consistency effect is not strong, the consistency of the
affected sample cohorts does have a measurable impact on the 1se cv and test set error
rates (cor = -0.32 and -0.27, respectively).</p>
<p>This plot is somewhat under-whelming.</p>
<!--chapter:end:05-glmnetSuiteHCC5hmC.Rmd-->
</div>
</div>
<div id="brca-rnaseq-preproc" class="section level1" number="6">
<h1 number="6"><span class="header-section-number">6</span> BrCa RNA-Seq: Preprocessing</h1>
<p>In this section we examine some RNA-Seq data from a breast cancer study
to provide a different SNR context from the HCC 5hmC data in which
to examine the relationship between sample size and model performance.
For this exercise, we will look at discriminating between breast
cancer samples of subtype LumA and all other subtypes among ER+/HER2-
breast cancer samples. This may
be a somewhat artificial and uninteresting classification problem,
but it does provide a binary classification problem with many examples
in the two subgroups allowing is to run simulations analogous to
those that were run on the HCC 5hmC data.</p>
<!--
Reader note:  

* The most interesting part of this section is Subsection \@ref(snr-regime)
which tells us where on the SNR spectrum the 5hmC data lies.
In view of the work by  Hastie et al. (2017) [@Hastie:2017aa]
summarized in Subsection \@ref(lasso-vs-best-sub), we would expect to
lasso model, and the relaxed version, to do very well with this
classifiaction problem.
-->
<!--

 FN <- 'tmp'
 # Shotcuts for knitting and redering while in R session (Invoke interactive R from R/Scripts folder)
 kk <- function(n='') knitr::knit2html(paste("t", n, sep=''), envir=globalenv(),
       output=paste(FN,".html", sep=''))

 rr <- function(n='') rmarkdown::render(paste("t", n, sep=''), envir=globalenv(),
       output_file=paste(FN,".html", sep='')) ##, output_dir='Scripts')

 bb <- function(n='') browseURL(paste(FN,".html", sep=''))

 # The usual shotcuts
 zz <- function(n='') source(paste("t", n, sep=''))

-->
<div id="load-the-data-1" class="section level2" number="6.1">
<h2 number="6.1"><span class="header-section-number">6.1</span> Load the data</h2>
<!-- THIS ENSURES NO EVALUATION TAKES PLACE BY DEFAULT -->
<!-- TO TURN ON, SET eval=T                            -->
<!-- Add base libraries -->
<p>The data that are available from NCBI GEO
<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE96058">Series GSE96058</a>
can be conveniently accessed through an R data package.
Attaching the GSE96058 package makes the count data tables
available as well as a gene annotation table and a sample description table.
See <a href="https://12379monty.github.io/GSE96058/">GSE96058 R Data Package page</a>.</p>
<p>The data in GSE96058 have been separated into various data frames and
matrices. The separation of the gene expression data into separate matrices
was designed to get around github’s file size restrictions. See
<a href="https://github.com/12379Monty/GSE96058">GSE96058 Package Page</a>
for details.</p>
<!--
**Programming Note**:  

* When the GSE96058 package is attached with the `library(GSE96058)`
function call, any previously loaded object with the same
name will mask the corresponding objects lazy-loaded by GSE96058.
To get around this problem, R data packages should have
unique object names.

-->
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb109-2"><a href="#cb109-2"></a><span class="cf">if</span> (<span class="op">!</span>(<span class="st">&quot;GSE96058&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()))) {</span>
<span id="cb109-3"><a href="#cb109-3"></a>  <span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;devtools&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)) {</span>
<span id="cb109-4"><a href="#cb109-4"></a>    <span class="kw">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span>
<span id="cb109-5"><a href="#cb109-5"></a>  }</span>
<span id="cb109-6"><a href="#cb109-6"></a>  devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;12379Monty/GSE96058&quot;</span>)</span>
<span id="cb109-7"><a href="#cb109-7"></a>}</span>
<span id="cb109-8"><a href="#cb109-8"></a></span>
<span id="cb109-9"><a href="#cb109-9"></a><span class="kw">library</span>(GSE96058)</span>
<span id="cb109-10"><a href="#cb109-10"></a></span>
<span id="cb109-11"><a href="#cb109-11"></a><span class="co"># Strip GSE_ID prefix form object names</span></span>
<span id="cb109-12"><a href="#cb109-12"></a><span class="co">### MOD: (2020.09.21) keep &#39;brcaRna_&#39; as prefix to object names.</span></span>
<span id="cb109-13"><a href="#cb109-13"></a><span class="co">###     Otherwise we run into problems when different datasets are</span></span>
<span id="cb109-14"><a href="#cb109-14"></a><span class="co">###     analyzed</span></span>
<span id="cb109-15"><a href="#cb109-15"></a></span>
<span id="cb109-16"><a href="#cb109-16"></a><span class="cf">for</span>(OBJ <span class="cf">in</span> <span class="kw">data</span>(<span class="dt">package=</span><span class="st">&#39;GSE96058&#39;</span>)<span class="op">$</span>results[, <span class="st">&#39;Item&#39;</span>])</span>
<span id="cb109-17"><a href="#cb109-17"></a><span class="kw">assign</span>(<span class="kw">sub</span>(<span class="st">&#39;GSE96058_&#39;</span>,<span class="st">&#39;brcaRna_&#39;</span>,OBJ), <span class="kw">get</span>(OBJ))</span>
<span id="cb109-18"><a href="#cb109-18"></a></span>
<span id="cb109-19"><a href="#cb109-19"></a><span class="kw">detach</span>(package<span class="op">:</span>GSE96058, <span class="dt">unload =</span> T )</span></code></pre></div>
<p>For this analysis, we will consider ER+/HER2- samples only and
examine <code>pam50_subtype = 'LumA' vs 'Other'</code> as outcome.
This choice is predicated on the desire to have a fairly
balanced binary outcome to work with when we examine the effect
of sample size on classification performance.
The outcome in this dataset will be aliased as <code>group</code>.</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1"></a>ER_HER2_tbl &lt;-<span class="st"> </span><span class="kw">with</span>(brcaRna_sampDesc <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb110-2"><a href="#cb110-2"></a><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">grepl</span>(<span class="st">&#39;repl$&#39;</span>, title) <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span>isRepl <span class="op">&amp;</span><span class="st"> </span>er_Status<span class="op">!=</span><span class="st">&#39;NA&#39;</span>   <span class="op">&amp;</span><span class="st"> </span>her2_Status<span class="op">!=</span><span class="st">&#39;NA&#39;</span>),</span>
<span id="cb110-3"><a href="#cb110-3"></a> <span class="kw">table</span>(<span class="dt">ER_HER2=</span><span class="kw">paste</span>(er_Status, her2_Status, <span class="dt">sep=</span><span class="st">&#39;_&#39;</span>),</span>
<span id="cb110-4"><a href="#cb110-4"></a>  <span class="dt">exclude=</span><span class="ot">NULL</span>)</span>
<span id="cb110-5"><a href="#cb110-5"></a> )</span>
<span id="cb110-6"><a href="#cb110-6"></a></span>
<span id="cb110-7"><a href="#cb110-7"></a> knitr<span class="op">::</span><span class="kw">kable</span>(ER_HER2_tbl,</span>
<span id="cb110-8"><a href="#cb110-8"></a>  <span class="dt">caption=</span><span class="st">&quot;ER, PR and HER2 Status&quot;</span>)  <span class="op">%&gt;%</span></span>
<span id="cb110-9"><a href="#cb110-9"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:brcaRna-preproc-subsetSamples)ER, PR and HER2 Status
</caption>
<thead>
<tr>
<th style="text-align:left;">
ER_HER2
</th>
<th style="text-align:right;">
Freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
0_0
</td>
<td style="text-align:right;">
165
</td>
</tr>
<tr>
<td style="text-align:left;">
0_1
</td>
<td style="text-align:right;">
63
</td>
</tr>
<tr>
<td style="text-align:left;">
1_0
</td>
<td style="text-align:right;">
2425
</td>
</tr>
<tr>
<td style="text-align:left;">
1_1
</td>
<td style="text-align:right;">
310
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1"></a><span class="co"># Subset analysis samples</span></span>
<span id="cb111-2"><a href="#cb111-2"></a>brcaRna_sampDesc &lt;-</span>
<span id="cb111-3"><a href="#cb111-3"></a><span class="st">  </span>brcaRna_sampDesc <span class="op">%&gt;%</span></span>
<span id="cb111-4"><a href="#cb111-4"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">grepl</span>(<span class="st">&#39;repl$&#39;</span>, title) <span class="op">&amp;</span><span class="st"> </span>er_Status <span class="op">==</span><span class="st"> &#39;1&#39;</span> <span class="op">&amp;</span><span class="st"> </span>her2_Status <span class="op">==</span><span class="st"> &#39;0&#39;</span>) <span class="op">%&gt;%</span></span>
<span id="cb111-5"><a href="#cb111-5"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">mutate</span>(<span class="dt">group =</span> pam50_subtype) <span class="op">%&gt;%</span></span>
<span id="cb111-6"><a href="#cb111-6"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">rename</span>(<span class="dt">sampID =</span> title) <span class="op">%&gt;%</span></span>
<span id="cb111-7"><a href="#cb111-7"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">arrange</span>(group, sampID)</span>
<span id="cb111-8"><a href="#cb111-8"></a></span>
<span id="cb111-9"><a href="#cb111-9"></a><span class="co">#with(brcaRna_sampDesc, table(group, exclude=NULL))</span></span>
<span id="cb111-10"><a href="#cb111-10"></a></span>
<span id="cb111-11"><a href="#cb111-11"></a><span class="co"># Recode group</span></span>
<span id="cb111-12"><a href="#cb111-12"></a>brcaRna_sampDesc<span class="op">$</span>group &lt;-<span class="st"> </span><span class="kw">with</span>(</span>
<span id="cb111-13"><a href="#cb111-13"></a>  brcaRna_sampDesc,</span>
<span id="cb111-14"><a href="#cb111-14"></a>  <span class="kw">ifelse</span>(group <span class="op">==</span><span class="st"> &quot;LumA&quot;</span>, <span class="st">&quot;LumA&quot;</span>, <span class="st">&quot;Other&quot;</span>)  <span class="co">## Luma &gt; Other</span></span>
<span id="cb111-15"><a href="#cb111-15"></a>)</span>
<span id="cb111-16"><a href="#cb111-16"></a></span>
<span id="cb111-17"><a href="#cb111-17"></a><span class="co"># Re-order and name rows</span></span>
<span id="cb111-18"><a href="#cb111-18"></a>o.v &lt;-<span class="st"> </span><span class="kw">with</span>(brcaRna_sampDesc, <span class="kw">order</span>(group, sampID))</span>
<span id="cb111-19"><a href="#cb111-19"></a>brcaRna_sampDesc  &lt;-<span class="st"> </span>brcaRna_sampDesc[o.v, ]</span>
<span id="cb111-20"><a href="#cb111-20"></a></span>
<span id="cb111-21"><a href="#cb111-21"></a><span class="kw">rownames</span>(brcaRna_sampDesc) &lt;-<span class="st"> </span>brcaRna_sampDesc<span class="op">$</span>sampID</span>
<span id="cb111-22"><a href="#cb111-22"></a></span>
<span id="cb111-23"><a href="#cb111-23"></a><span class="co"># set brcaRna_groupCol for later</span></span>
<span id="cb111-24"><a href="#cb111-24"></a>brcaRna_groupCol &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;#F3C300&quot;</span>, <span class="st">&quot;#875692&quot;</span>)</span>
<span id="cb111-25"><a href="#cb111-25"></a><span class="kw">names</span>(brcaRna_groupCol) &lt;-<span class="st"> </span><span class="kw">unique</span>(brcaRna_sampDesc<span class="op">$</span>group)</span>
<span id="cb111-26"><a href="#cb111-26"></a></span>
<span id="cb111-27"><a href="#cb111-27"></a><span class="kw">with</span>(brcaRna_sampDesc, </span>
<span id="cb111-28"><a href="#cb111-28"></a> knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">table</span>(group, <span class="dt">exclude =</span> <span class="ot">NULL</span>),</span>
<span id="cb111-29"><a href="#cb111-29"></a>  <span class="dt">caption=</span><span class="st">&quot;Samples used in this analysis&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb111-30"><a href="#cb111-30"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span>
<span id="cb111-31"><a href="#cb111-31"></a>)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:brcaRna-preproc-subsetSamples)Samples used in this analysis
</caption>
<thead>
<tr>
<th style="text-align:left;">
group
</th>
<th style="text-align:right;">
Freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
LumA
</td>
<td style="text-align:right;">
1492
</td>
</tr>
<tr>
<td style="text-align:left;">
Other
</td>
<td style="text-align:right;">
933
</td>
</tr>
</tbody>
</table>
<p>The features in this dataset are gene expression indicators:</p>
<pre><code>
Gene expression data in FPKM were generated using cufflinks 2.2.1 
(default parameters except –GTF, –frag-bias-correct GRCh38.fa, –multi-read-correct, 
–library-type fr-firststrand, –total-hits-norm, –max-bundle-frags 10000000). 
The resulting data was was post-processed by collapsing on 30,865 unique gene symbols 
(sum of FPKM values of each matching transcript), adding to each expression measurement 
0.1 FPKM, and performing a log2 transformation.
</code></pre>
<p><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>.</p>
<p>We now proceed to run a <em>minimal set of QC sanity checks</em> to make sure that
there are no apparent systematic effects in the data.</p>
<p>Assemble feature matrix for <code>brcaRna_sampDesc</code> samples.</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1"></a><span class="co"># Start for geneExpression_repl1 data</span></span>
<span id="cb113-2"><a href="#cb113-2"></a>repl1_ndx &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">is.element</span>(<span class="kw">colnames</span>(brcaRna_geneExpression_repl1), brcaRna_sampDesc<span class="op">$</span>sampID))</span>
<span id="cb113-3"><a href="#cb113-3"></a>brcaRna_geneExpression &lt;-<span class="st"> </span>brcaRna_geneExpression_repl1[, repl1_ndx]</span>
<span id="cb113-4"><a href="#cb113-4"></a></span>
<span id="cb113-5"><a href="#cb113-5"></a><span class="co"># add samples from brcaRna_geneExpression_subJJ</span></span>
<span id="cb113-6"><a href="#cb113-6"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>) {</span>
<span id="cb113-7"><a href="#cb113-7"></a>  sub_ndx &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">is.element</span>(</span>
<span id="cb113-8"><a href="#cb113-8"></a>   <span class="kw">colnames</span>(<span class="kw">get</span>(<span class="kw">paste0</span>(<span class="st">&#39;brcaRna_geneExpression_sub&#39;</span>,JJ))),  brcaRna_sampDesc<span class="op">$</span>sampID))</span>
<span id="cb113-9"><a href="#cb113-9"></a>  brcaRna_geneExpression &lt;-<span class="st"> </span><span class="kw">cbind</span>(brcaRna_geneExpression,</span>
<span id="cb113-10"><a href="#cb113-10"></a>   <span class="kw">get</span>(<span class="kw">paste0</span>(<span class="st">&#39;brcaRna_geneExpression_sub&#39;</span>,JJ))[, sub_ndx])</span>
<span id="cb113-11"><a href="#cb113-11"></a>}</span>
<span id="cb113-12"><a href="#cb113-12"></a></span>
<span id="cb113-13"><a href="#cb113-13"></a><span class="co"># align brcaRna_geneExpression columns to brcaRna_sampDesc samples order</span></span>
<span id="cb113-14"><a href="#cb113-14"></a>geneExpr_ndx &lt;-<span class="st"> </span><span class="kw">match</span>(brcaRna_sampDesc<span class="op">$</span>sampID, <span class="kw">colnames</span>(brcaRna_geneExpression))</span>
<span id="cb113-15"><a href="#cb113-15"></a><span class="cf">if</span>(<span class="kw">any</span>(<span class="kw">is.na</span>(geneExpr_ndx)))</span>
<span id="cb113-16"><a href="#cb113-16"></a><span class="kw">stop</span>(<span class="st">&quot;brcaRna_sampDesc/brcaRna_geneExpression: sample mismatch&quot;</span>) <span class="cf">else</span></span>
<span id="cb113-17"><a href="#cb113-17"></a>brcaRna_geneExpression &lt;-<span class="st"> </span>brcaRna_geneExpression[, <span class="kw">rownames</span>(brcaRna_sampDesc)]</span></code></pre></div>
<p>We first look at coverage - make sure there isn’t too much disparity of coverage
across samples. To detect shared variability, samples can be annotated and ordered
according to sample features that may be linked to sample batch processing. Here we
the samples have been ordered by group and sample id (an alias of geoAcc).</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1"></a><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">3</span>))</span>
<span id="cb114-2"><a href="#cb114-2"></a><span class="kw">boxplot</span>(brcaRna_geneExpression,</span>
<span id="cb114-3"><a href="#cb114-3"></a>  <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">4</span>, <span class="dv">4</span>), <span class="dt">ylab=</span><span class="st">&#39;log2 Count&#39;</span>,</span>
<span id="cb114-4"><a href="#cb114-4"></a>  <span class="dt">staplewex =</span> <span class="dv">0</span>,       <span class="co"># remove horizontal whisker lines</span></span>
<span id="cb114-5"><a href="#cb114-5"></a>  <span class="dt">staplecol =</span> <span class="st">&quot;white&quot;</span>, <span class="co"># just to be totally sure :)</span></span>
<span id="cb114-6"><a href="#cb114-6"></a>  <span class="dt">outline =</span> F,         <span class="co"># remove outlying points</span></span>
<span id="cb114-7"><a href="#cb114-7"></a>  <span class="dt">whisklty =</span> <span class="dv">0</span>,        <span class="co"># remove vertical whisker lines</span></span>
<span id="cb114-8"><a href="#cb114-8"></a>  <span class="dt">las =</span> <span class="dv">2</span>, <span class="dt">horizontal =</span> F, <span class="dt">xaxt =</span> <span class="st">&quot;n&quot;</span>,</span>
<span id="cb114-9"><a href="#cb114-9"></a>  <span class="dt">border =</span> brcaRna_groupCol[brcaRna_sampDesc<span class="op">$</span>group]</span>
<span id="cb114-10"><a href="#cb114-10"></a>)</span>
<span id="cb114-11"><a href="#cb114-11"></a><span class="kw">legend</span>(<span class="st">&quot;top&quot;</span>, <span class="dt">legend =</span> <span class="kw">names</span>(brcaRna_groupCol), <span class="dt">text.col =</span> brcaRna_groupCol, </span>
<span id="cb114-12"><a href="#cb114-12"></a>  <span class="dt">ncol =</span> <span class="dv">2</span>, <span class="dt">bty =</span> <span class="st">&quot;n&quot;</span>)</span>
<span id="cb114-13"><a href="#cb114-13"></a><span class="co"># Add reference lines</span></span>
<span id="cb114-14"><a href="#cb114-14"></a>SampleMedian &lt;-<span class="st"> </span><span class="kw">apply</span>(brcaRna_geneExpression, <span class="dv">2</span>, median)</span>
<span id="cb114-15"><a href="#cb114-15"></a><span class="kw">abline</span>(<span class="dt">h =</span> <span class="kw">median</span>(SampleMedian), <span class="dt">col =</span> <span class="st">&quot;grey&quot;</span>)</span>
<span id="cb114-16"><a href="#cb114-16"></a><span class="kw">axis</span>(<span class="dt">side =</span> <span class="dv">4</span>, <span class="dt">at =</span> <span class="kw">round</span>(<span class="kw">median</span>(SampleMedian), <span class="dv">2</span>), </span>
<span id="cb114-17"><a href="#cb114-17"></a>  <span class="dt">las =</span> <span class="dv">2</span>, <span class="dt">col =</span> <span class="st">&quot;grey&quot;</span>, <span class="dt">line =</span> <span class="dv">0</span>, <span class="dt">tick =</span> F)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-preproc-geneExprBxp-1.png" alt="Sample log2 count boxplots" width="960" />
<p class="caption">
(#fig:brcaRna-preproc-geneExprBxp)Sample log2 count boxplots
</p>
</div>
<!-- DNA - we're not looking at COVERAGE!!!! 
Coverage/brcaRna_geneExpression level looks fairly comparable across samples.  It is sometimes helpful to
keep track of the actual coverage which can be adequetely tracked by distribution
quantiles.
-->
<!-- SKIP
From this table, we see that 25% of the samples have total coverage exceeding
$TICKr round(brcaRna_geneExpression_quant2["75%", "totCovM"],1)TICK$M reads, 25% of samples
have a 15 percentile of coverage lower than
$TICKr brcaRna_geneExpression_quant2["25%", "15%"]TICK$, etc.  
-->
<!-- SKIP
We next look at relative log representation (RLR) (in the context of measuring the density of 
5hmC marks in genes, we refer to `representation` as opposed to `expression`; the
two can be used interchangibly) -
make sure the shapes of the distributions are not widely different.
-->
<!-- SKIPPED
We note that the HCC samples have slightly more variable coverage distribution.
A few samples are quite different.
-->
</div>
<div id="breast-rnaseq-dra" class="section level2" number="6.2">
<h2 number="6.2"><span class="header-section-number">6.2</span> Differential representation analysis</h2>
<p>In the remainder of this section, we will process the data and
perform differential expression analysis. If we had counts,
the analysis pipeline outlined in Law et al. (2018) <span class="citation">[<a href="#ref-Law:2018aa" role="doc-biblioref">18</a>]</span>
would be appropriate. The main analysis steps of the pipeline are:</p>
<ul>
<li>remove lowly expressed genes</li>
<li>normalize gene expression distributions</li>
<li>remove heteroscedascity</li>
<li>fit linear models and examine DE results</li>
</ul>
<p>We will adapt this pipeline to the FPKM data that we have to work with - basically
skip the 2nd and 3rd steps in the above pipeline.</p>
<div id="remove-lowly-expressed-genes-1" class="section level3 unnumbered" number="">
<h3 class="unnumbered" number="">Remove lowly expressed genes</h3>
<p>To determine a sensible threshold we can begin by examining the shapes of the distributions.</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1"></a><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb115-2"><a href="#cb115-2"></a><span class="kw">plot</span>(<span class="kw">density</span>(brcaRna_geneExpression[, <span class="dv">1</span>]), </span>
<span id="cb115-3"><a href="#cb115-3"></a>  <span class="dt">col =</span> brcaRna_groupCol[brcaRna_sampDesc<span class="op">$</span>group[<span class="dv">1</span>]], <span class="dt">lwd =</span> <span class="dv">2</span>, </span>
<span id="cb115-4"><a href="#cb115-4"></a>  <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">.30</span>), <span class="dt">xlim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">5</span>, <span class="dv">10</span>),</span>
<span id="cb115-5"><a href="#cb115-5"></a>  <span class="dt">las =</span> <span class="dv">2</span>, <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;gene expression&quot;</span></span>
<span id="cb115-6"><a href="#cb115-6"></a>)</span>
<span id="cb115-7"><a href="#cb115-7"></a><span class="kw">abline</span>(<span class="dt">v =</span> <span class="dv">0</span>, <span class="dt">col =</span> <span class="dv">3</span>)</span>
<span id="cb115-8"><a href="#cb115-8"></a><span class="co"># After verifying no outliers, can plot a random subset </span></span>
<span id="cb115-9"><a href="#cb115-9"></a><span class="cf">for</span> (JJ <span class="cf">in</span> <span class="kw">sample</span>(<span class="dv">2</span><span class="op">:</span><span class="kw">ncol</span>(brcaRna_geneExpression), <span class="dt">size =</span> <span class="dv">100</span>)) {</span>
<span id="cb115-10"><a href="#cb115-10"></a>  den &lt;-<span class="st"> </span><span class="kw">density</span>(brcaRna_geneExpression[, JJ])</span>
<span id="cb115-11"><a href="#cb115-11"></a>  <span class="kw">lines</span>(den<span class="op">$</span>x, den<span class="op">$</span>y, <span class="dt">col =</span> brcaRna_groupCol[brcaRna_sampDesc<span class="op">$</span>group[JJ]], <span class="dt">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb115-12"><a href="#cb115-12"></a>} <span class="co"># for(JJ</span></span>
<span id="cb115-13"><a href="#cb115-13"></a><span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="dt">legend =</span> <span class="kw">names</span>(brcaRna_groupCol), </span>
<span id="cb115-14"><a href="#cb115-14"></a>  <span class="dt">text.col =</span> brcaRna_groupCol, <span class="dt">bty =</span> <span class="st">&quot;n&quot;</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-preproc-densityGeneExpr-1.png" alt="Sample gene expression densities" width="960" />
<p class="caption">
(#fig:brcaRna-preproc-densityGeneExpr)Sample gene expression densities
</p>
</div>
<p>In this analysis we will be using a nominal gene expression value of
0, genes are deeemed to be <code>represented</code> if their expression is
above this threshold, and not represented otherwise.
For this analysis we will require that genes be <code>represented</code> in at least
25 samples across the entire dataset to be retained for downstream analysis.</p>
<!-- or at least 
TICKr round(EXPR_THR*LibSizeSum['Max.'])TICK counts in the sample with the 
greatest sequencing depth (library size TICKr round(LibSizeSum['Max.'],1)TICK million).
-->
<p>Remove weakly represented genes and replot densities.</p>
<p>Removing 37.1% of genes…</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb116-2"><a href="#cb116-2"></a></span>
<span id="cb116-3"><a href="#cb116-3"></a><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb116-4"><a href="#cb116-4"></a><span class="kw">plot</span>(<span class="kw">density</span>(brcaRna_geneExpression_F[, <span class="dv">1</span>]),</span>
<span id="cb116-5"><a href="#cb116-5"></a>  <span class="dt">col =</span> brcaRna_groupCol[brcaRna_sampDesc<span class="op">$</span>group[<span class="dv">1</span>]],</span>
<span id="cb116-6"><a href="#cb116-6"></a>  <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">las =</span> <span class="dv">2</span>, <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;gene expression&quot;</span>,</span>
<span id="cb116-7"><a href="#cb116-7"></a>  <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">.30</span>), <span class="dt">xlim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">5</span>, <span class="dv">10</span>)</span>
<span id="cb116-8"><a href="#cb116-8"></a>)</span>
<span id="cb116-9"><a href="#cb116-9"></a><span class="co">#abline(v = 0, col = 3)</span></span>
<span id="cb116-10"><a href="#cb116-10"></a><span class="co"># After verifying no outliers, can plot a random subset </span></span>
<span id="cb116-11"><a href="#cb116-11"></a><span class="cf">for</span> (JJ <span class="cf">in</span> <span class="kw">sample</span>(<span class="dv">2</span><span class="op">:</span><span class="kw">ncol</span>(brcaRna_geneExpression_F), <span class="dt">size =</span> <span class="dv">100</span>)) {</span>
<span id="cb116-12"><a href="#cb116-12"></a>  den &lt;-<span class="st"> </span><span class="kw">density</span>(brcaRna_geneExpression_F[, JJ])</span>
<span id="cb116-13"><a href="#cb116-13"></a>  <span class="kw">lines</span>(den<span class="op">$</span>x, den<span class="op">$</span>y, <span class="dt">col =</span> brcaRna_groupCol[brcaRna_sampDesc<span class="op">$</span>group[JJ]], <span class="dt">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb116-14"><a href="#cb116-14"></a>} <span class="co"># for(JJ</span></span>
<span id="cb116-15"><a href="#cb116-15"></a><span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="dt">legend =</span> <span class="kw">names</span>(brcaRna_groupCol), </span>
<span id="cb116-16"><a href="#cb116-16"></a>  <span class="dt">text.col =</span> brcaRna_groupCol, <span class="dt">bty =</span> <span class="st">&quot;n&quot;</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-preproc-densityGeneExpr2-1.png" alt="Sample gene expression densities after removing weak genes" width="960" />
<p class="caption">
(#fig:brcaRna-preproc-densityGeneExpr2)Sample gene expression densities after removing weak genes
</p>
</div>
<!--
Note that the $log_2(CMP)$ distribution is not quite symmetric.
-->
<p>As another sanity check, we will look at a
multidimensional scaling plot of distances between gene expression
profiles. We use <code>plotMDS</code> in limma package <span class="citation">[<a href="#ref-Ritchie:2015aa" role="doc-biblioref">19</a>]</span>),
which plots samples on a two-dimensional scatterplot so that distances on
the plot approximate the typical log2 fold changes between the
samples.</p>
<!-- SKIP - data are $\approx$ normalized 
Before producing the MDS plot we will normalize the distributions.
We will store the data into s `DGEList` object as this is convenient
when running many of the analyses implemented in the edgeR and limma packages.
Call the set 'AF', for set 'A', 'Filtered'.
-->
<!-- SKIP
Verify that the counts are properly normalized.
-->
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb117-2"><a href="#cb117-2"></a></span>
<span id="cb117-3"><a href="#cb117-3"></a><span class="kw">par</span>(<span class="dt">mfcol =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">xpd =</span> <span class="ot">NA</span>, <span class="dt">oma =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">0</span>))</span>
<span id="cb117-4"><a href="#cb117-4"></a></span>
<span id="cb117-5"><a href="#cb117-5"></a><span class="co"># wo loss of generality, sample 500 samples</span></span>
<span id="cb117-6"><a href="#cb117-6"></a><span class="co"># simply a matter of convenience to save time</span></span>
<span id="cb117-7"><a href="#cb117-7"></a><span class="co"># remove from final version</span></span>
<span id="cb117-8"><a href="#cb117-8"></a><span class="kw">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb117-9"><a href="#cb117-9"></a>samp_ndx &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(brcaRna_geneExpression_F), <span class="dt">size =</span> <span class="dv">500</span>)</span>
<span id="cb117-10"><a href="#cb117-10"></a>MDS.out &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">plotMDS</span>(brcaRna_geneExpression_F[, samp_ndx],</span>
<span id="cb117-11"><a href="#cb117-11"></a>  <span class="dt">col =</span> brcaRna_groupCol[brcaRna_sampDesc<span class="op">$</span>group[samp_ndx]], <span class="dt">pch =</span> <span class="dv">1</span></span>
<span id="cb117-12"><a href="#cb117-12"></a>)</span>
<span id="cb117-13"><a href="#cb117-13"></a><span class="kw">legend</span>(<span class="st">&quot;topleft&quot;</span>,</span>
<span id="cb117-14"><a href="#cb117-14"></a>  <span class="dt">legend =</span> <span class="kw">names</span>(brcaRna_groupCol),</span>
<span id="cb117-15"><a href="#cb117-15"></a>  <span class="dt">text.col =</span> brcaRna_groupCol, <span class="dt">bty =</span> <span class="st">&quot;n&quot;</span></span>
<span id="cb117-16"><a href="#cb117-16"></a>)</span>
<span id="cb117-17"><a href="#cb117-17"></a></span>
<span id="cb117-18"><a href="#cb117-18"></a>MDS.out &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">plotMDS</span>(brcaRna_geneExpression_F[, samp_ndx],</span>
<span id="cb117-19"><a href="#cb117-19"></a>  <span class="dt">col =</span> brcaRna_groupCol[brcaRna_sampDesc<span class="op">$</span>group[samp_ndx]], <span class="dt">pch =</span> <span class="dv">1</span>,</span>
<span id="cb117-20"><a href="#cb117-20"></a>  <span class="dt">dim.plot =</span> <span class="dv">3</span><span class="op">:</span><span class="dv">4</span></span>
<span id="cb117-21"><a href="#cb117-21"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-preproc-plotMDS-1.png" alt="MDS plots of gene expression values" width="960" />
<p class="caption">
(#fig:brcaRna-preproc-plotMDS)MDS plots of gene expression values
</p>
</div>
<p>The MDS plot, which is analogous to a PCA plot adapted to gene exression data,
does not indicate strong clustering of samples.</p>
<!-- SKIP
`glMDSPlot` from package `Glimma` provides an interactive MDS 
plot that can extremely usful for exploration
-->
<!-- SKIP
Link to glMDSPlot: 
[Here]($TICKr file.path(figures_DIR, paste0("GlMDSplot.html"))TICK  

No obvious factor links the samples in the 3 clusters observed on the
4th MDS dimensions. The percent of variance exaplained by this dimension or 
$\approx$ 4%.   The glMDSPlot indicates further segregation along
the 6th dimension.  The percent of variance exaplained by this dimension or 
$\approx$ 2%.  Tracking down this source of variability may be quite challenging,
especially without having the complete information about the sample attributes 
and provenance.  

Unwanted variability is a well-documented problem in the analysis of RNA-Seq data
(see Peixoto et al. (2015) [@Peixoto:2015aa]), and many procedures have been proposed
to reduce the effect of unwanted variation on RNA-Seq analsys results 
([@Gandolfo:2018aa;@Peixoto:2015aa;@Risso:2014aa]).  There are undoubtedly
some similar sources of systematic variation in the 5hmC data, but it is
beyond the scope of this work to investigate these in this particular dataset.
Given that the clustering of samples occurs in MDS dimensions that explain
a small fraction of variability, and that these is no assocation with the
factor of interest, LumA vs Other, these sources of variability should not
interfere too much with our classification analysis.  It would nonetheless be interesting
to assess whether downstream results can be improved by removing this variability.

-->
</div>
<div id="creating-a-design-matrix-and-contrasts-1" class="section level3 unnumbered" number="">
<h3 class="unnumbered" number="">Creating a design matrix and contrasts</h3>
<p>Before proceeding with the statistical modeling used for the
differential expression analysis, we need to set up a
model design matrix.</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb118-2"><a href="#cb118-2"></a></span>
<span id="cb118-3"><a href="#cb118-3"></a>Design_mtx &lt;-<span class="st"> </span><span class="kw">model.matrix</span>( <span class="op">~</span><span class="st">  </span><span class="dv">-1</span> <span class="op">+</span><span class="st"> </span>group, <span class="dt">data=</span>brcaRna_sampDesc)</span>
<span id="cb118-4"><a href="#cb118-4"></a><span class="kw">colnames</span>(Design_mtx) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&#39;group&#39;</span>, <span class="st">&#39;&#39;</span>, <span class="kw">colnames</span>(Design_mtx))</span>
<span id="cb118-5"><a href="#cb118-5"></a></span>
<span id="cb118-6"><a href="#cb118-6"></a><span class="kw">cat</span>(<span class="st">&quot;colSums(Design_mtx):</span><span class="ch">\n</span><span class="st">&quot;</span>)</span></code></pre></div>
<pre><code>## colSums(Design_mtx):</code></pre>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1"></a><span class="kw">colSums</span>(Design_mtx)</span></code></pre></div>
<pre><code>##  LumA Other 
##  1492   933</code></pre>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1"></a>Contrasts_mtx &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">makeContrasts</span>(</span>
<span id="cb122-2"><a href="#cb122-2"></a>  <span class="dt">LumAvsOther =</span> LumA  <span class="op">-</span><span class="st"> </span>Other,</span>
<span id="cb122-3"><a href="#cb122-3"></a>  <span class="dt">levels=</span><span class="kw">colnames</span>(Design_mtx))</span>
<span id="cb122-4"><a href="#cb122-4"></a></span>
<span id="cb122-5"><a href="#cb122-5"></a><span class="kw">cat</span>(<span class="st">&quot;Contrasts:</span><span class="ch">\n</span><span class="st">&quot;</span>)</span></code></pre></div>
<pre><code>## Contrasts:</code></pre>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1"></a>Contrasts_mtx</span></code></pre></div>
<pre><code>##        Contrasts
## Levels  LumAvsOther
##   LumA            1
##   Other          -1</code></pre>
<!-- DOES NOT APPLY - NEED COUNTS!!!!
### Removing heteroscedasticity from the count data {-}


As for RNA-Seq data, for 5hmC count data the variance is not independent of the mean.
In `limma`, the R package we are using for our analyses, 
linear modeling is carried out on the gene expression values which are assumed to be 
normally distributed and the mean-variance relationship is accommodated using precision 
weights calculated by the voom function.  We apply this transformation next.

-->
<!--

Note that the voom-plot provides a visual check on the level of filtering performed upstream.
If filtering of lowly-expressed genes is insufficient, a drop in variance levels can be 
observed at the low end of the expression scale due to very small counts. 

-->
<!--
Means (x-axis) and variances (y-axis) of each gene are plotted to show the dependence between the two before voom is applied to the data ( A) and how the trend is removed after voom precision weights are applied to the data ( B). The plot on the left is created within the voom function which extracts residual variances from fitting linear models to gene expression transformed data. Variances are then re-scaled to quarter-root variances (or square-root of standard deviations) and plotted against the mean expression of each gene. The means are log 2-transformed mean-counts with an offset of 2. The plot on the right is created using plotSA which plots log 2 residual standard deviations against mean gene expression values. The average log 2 residual standard deviation is marked by a horizontal blue line. In both plots, each black dot represents a gene and a red curve is fitted to these points.
-->
<!--

To get a sense of how this compares with RNA-Seq dsta, we can take a look at Figure 4 in Law et al. [@Law:2018aa]:

TICKr knitr::include_graphics(
  "Static/images/Law-fig4.gif",  dpi=100)TICK

We observe that the variability in the 5hmC data is quite a bit lower.  Statistical summaries would give
a better idea.  

-->
</div>
<div id="fit-linear-models-and-examine-the-results-1" class="section level3 unnumbered" number="">
<h3 class="unnumbered" number="">Fit linear models and examine the results</h3>
<p>Having properly filtered and normalized the data,
the linear models can be fitted to each gene and the results
examined to assess differential expression between the two groups
of interest, in our case LumA vs Other.</p>
<p>Table @ref(tab:lmFit) displays the counts of genes in each DE category:</p>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1"></a> brcaRna_geneExpression_F_fit &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">lmFit</span>(brcaRna_geneExpression_F, Design_mtx)</span>
<span id="cb126-2"><a href="#cb126-2"></a> <span class="kw">colnames</span>(brcaRna_geneExpression_F_fit<span class="op">$</span>coefficients) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">(Intercept</span><span class="ch">\\</span><span class="st">)&quot;</span>, <span class="st">&quot;Intercept&quot;</span>,</span>
<span id="cb126-3"><a href="#cb126-3"></a> <span class="kw">colnames</span>(brcaRna_geneExpression_F_fit<span class="op">$</span>coefficients) )</span>
<span id="cb126-4"><a href="#cb126-4"></a></span>
<span id="cb126-5"><a href="#cb126-5"></a> brcaRna_geneExpression_F_fit &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">contrasts.fit</span>(</span>
<span id="cb126-6"><a href="#cb126-6"></a>    brcaRna_geneExpression_F_fit, <span class="dt">contrasts=</span>Contrasts_mtx)</span>
<span id="cb126-7"><a href="#cb126-7"></a></span>
<span id="cb126-8"><a href="#cb126-8"></a> brcaRna_geneExpression_F_efit &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">eBayes</span>(brcaRna_geneExpression_F_fit)</span>
<span id="cb126-9"><a href="#cb126-9"></a></span>
<span id="cb126-10"><a href="#cb126-10"></a> brcaRna_geneExpression_F_efit_dt &lt;-</span>
<span id="cb126-11"><a href="#cb126-11"></a><span class="st"> </span>limma<span class="op">::</span><span class="kw">decideTests</span>(brcaRna_geneExpression_F_efit,<span class="dt">adjust.method =</span> <span class="st">&quot;BH&quot;</span>, <span class="dt">p.value =</span> <span class="fl">0.05</span>)</span>
<span id="cb126-12"><a href="#cb126-12"></a> </span>
<span id="cb126-13"><a href="#cb126-13"></a> knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">summary</span>(brcaRna_geneExpression_F_efit_dt)),</span>
<span id="cb126-14"><a href="#cb126-14"></a>  <span class="dt">caption=</span><span class="st">&quot;DE Results at FDR = 0.05&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb126-15"><a href="#cb126-15"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:brcaRna-preproc-lmFit)DE Results at FDR = 0.05
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Down
</th>
<th style="text-align:right;">
NotSig
</th>
<th style="text-align:right;">
Up
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
LumAvsOther
</td>
<td style="text-align:right;">
6301
</td>
<td style="text-align:right;">
4498
</td>
<td style="text-align:right;">
8607
</td>
</tr>
</tbody>
</table>
</div>
<div id="graphical-representations-of-de-results-md-plots-1" class="section level3 unnumbered" number="">
<h3 class="unnumbered" number="">Graphical representations of DE results: MD Plots</h3>
<p>To summarise results for all genes visually, mean-difference plots
(aka MA plot), which display log-FCs from the linear model fit against
the average gene expression values can be generated using the plotMD function,
with the differentially expressed genes highlighted.</p>
<p>We may also be interested in whether certain gene features are
related to gene identification. Gene GC content, for example, might be
of interest. We don’t have GC content here - skip this for now.</p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1"></a><span class="co">###  CLEAR CACHE</span></span>
<span id="cb127-2"><a href="#cb127-2"></a></span>
<span id="cb127-3"><a href="#cb127-3"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="fl">4.5</span>,<span class="fl">4.5</span>,<span class="dv">2</span>,<span class="dv">1</span>),<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb127-4"><a href="#cb127-4"></a></span>
<span id="cb127-5"><a href="#cb127-5"></a><span class="co"># log-fold-change vs ave-expr</span></span>
<span id="cb127-6"><a href="#cb127-6"></a>limma<span class="op">::</span><span class="kw">plotMD</span>(brcaRna_geneExpression_F_efit,</span>
<span id="cb127-7"><a href="#cb127-7"></a> <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.5</span>, <span class="fl">0.5</span>),</span>
<span id="cb127-8"><a href="#cb127-8"></a> <span class="dt">column=</span><span class="st">&#39;LumAvsOther&#39;</span>,</span>
<span id="cb127-9"><a href="#cb127-9"></a> <span class="dt">status=</span>brcaRna_geneExpression_F_efit_dt[,<span class="st">&#39;LumAvsOther&#39;</span>],</span>
<span id="cb127-10"><a href="#cb127-10"></a> <span class="dt">hl.pch =</span> <span class="dv">16</span>, <span class="dt">hl.col =</span> <span class="kw">c</span>(<span class="st">&quot;lightblue&quot;</span>, <span class="st">&quot;pink&quot;</span>), <span class="dt">hl.cex =</span> <span class="fl">.5</span>,</span>
<span id="cb127-11"><a href="#cb127-11"></a> <span class="dt">bg.pch =</span> <span class="dv">16</span>, <span class="dt">bg.col =</span> <span class="st">&quot;grey&quot;</span>, <span class="dt">bg.cex =</span> <span class="fl">0.5</span>,</span>
<span id="cb127-12"><a href="#cb127-12"></a> <span class="dt">main =</span> <span class="st">&#39;&#39;</span>,</span>
<span id="cb127-13"><a href="#cb127-13"></a> <span class="dt">xlab =</span> <span class="kw">paste0</span>(</span>
<span id="cb127-14"><a href="#cb127-14"></a>    <span class="st">&quot;Average log-expression: IQR=&quot;</span>,</span>
<span id="cb127-15"><a href="#cb127-15"></a>    <span class="kw">paste</span>(<span class="kw">round</span>(<span class="kw">quantile</span>(brcaRna_geneExpression_F_efit<span class="op">$</span>Amean, <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>), <span class="dv">2</span>),</span>
<span id="cb127-16"><a href="#cb127-16"></a>      <span class="dt">collapse =</span> <span class="st">&quot;, &quot;</span></span>
<span id="cb127-17"><a href="#cb127-17"></a>    )</span>
<span id="cb127-18"><a href="#cb127-18"></a>  ),</span>
<span id="cb127-19"><a href="#cb127-19"></a>  <span class="dt">ylab =</span> <span class="kw">paste0</span>(</span>
<span id="cb127-20"><a href="#cb127-20"></a>    <span class="st">&quot;log-fold-change: IQR=&quot;</span>,</span>
<span id="cb127-21"><a href="#cb127-21"></a>    <span class="kw">paste</span>(<span class="kw">round</span>(<span class="kw">quantile</span>(brcaRna_geneExpression_F_efit<span class="op">$</span>coefficients[, <span class="st">&#39;LumAvsOther&#39;</span>], <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>), <span class="dv">2</span>),</span>
<span id="cb127-22"><a href="#cb127-22"></a>      <span class="dt">collapse =</span> <span class="st">&quot;, &quot;</span></span>
<span id="cb127-23"><a href="#cb127-23"></a>    )</span>
<span id="cb127-24"><a href="#cb127-24"></a>  ),</span>
<span id="cb127-25"><a href="#cb127-25"></a>  <span class="dt">legend =</span> F, <span class="dt">cex.lab=</span><span class="fl">1.5</span></span>
<span id="cb127-26"><a href="#cb127-26"></a>)</span>
<span id="cb127-27"><a href="#cb127-27"></a><span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">0</span>, <span class="dt">col =</span> <span class="st">&quot;black&quot;</span>)</span>
<span id="cb127-28"><a href="#cb127-28"></a><span class="kw">rug</span>(<span class="kw">quantile</span>(brcaRna_geneExpression_F_efit<span class="op">$</span>coefficients[, <span class="st">&#39;LumAvsOther&#39;</span>], <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>),</span>
<span id="cb127-29"><a href="#cb127-29"></a>  <span class="dt">col =</span> <span class="st">&quot;purple&quot;</span>,</span>
<span id="cb127-30"><a href="#cb127-30"></a>  <span class="dt">ticksize =</span> <span class="fl">.03</span>, <span class="dt">side =</span> <span class="dv">2</span>, <span class="dt">lwd =</span> <span class="dv">2</span></span>
<span id="cb127-31"><a href="#cb127-31"></a>)</span>
<span id="cb127-32"><a href="#cb127-32"></a><span class="kw">rug</span>(<span class="kw">quantile</span>(brcaRna_geneExpression_F_efit<span class="op">$</span>Amean, <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>),</span>
<span id="cb127-33"><a href="#cb127-33"></a>  <span class="dt">col =</span> <span class="st">&quot;purple&quot;</span>,</span>
<span id="cb127-34"><a href="#cb127-34"></a>  <span class="dt">ticksize =</span> <span class="fl">.03</span>, <span class="dt">side =</span> <span class="dv">1</span>, <span class="dt">lwd =</span> <span class="dv">2</span></span>
<span id="cb127-35"><a href="#cb127-35"></a>)</span>
<span id="cb127-36"><a href="#cb127-36"></a></span>
<span id="cb127-37"><a href="#cb127-37"></a><span class="co"># log-fold-change vs identification</span></span>
<span id="cb127-38"><a href="#cb127-38"></a></span>
<span id="cb127-39"><a href="#cb127-39"></a><span class="kw">boxplot</span>(<span class="kw">split</span>(</span>
<span id="cb127-40"><a href="#cb127-40"></a> brcaRna_geneExpression_F_efit<span class="op">$</span>coefficients[, <span class="st">&#39;LumAvsOther&#39;</span>],</span>
<span id="cb127-41"><a href="#cb127-41"></a> brcaRna_geneExpression_F_efit_dt[,<span class="st">&#39;LumAvsOther&#39;</span>]),</span>
<span id="cb127-42"><a href="#cb127-42"></a> <span class="dt">outline=</span>F,</span>
<span id="cb127-43"><a href="#cb127-43"></a> <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&quot;pink&quot;</span>, <span class="st">&quot;grey&quot;</span>, <span class="st">&quot;lightblue&quot;</span>), <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>,</span>
<span id="cb127-44"><a href="#cb127-44"></a> <span class="dt">ylab=</span><span class="st">&#39;log-fold-change&#39;</span>, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span>.<span class="dv">4</span>, <span class="fl">.4</span>),</span>
<span id="cb127-45"><a href="#cb127-45"></a> <span class="dt">cex.lab=</span><span class="fl">1.5</span></span>
<span id="cb127-46"><a href="#cb127-46"></a>)</span>
<span id="cb127-47"><a href="#cb127-47"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="kw">c</span>(<span class="st">&#39;down&#39;</span>, <span class="st">&#39;notDE&#39;</span>, <span class="st">&#39;up&#39;</span>), <span class="dt">cex.axis=</span><span class="fl">1.5</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-preproc-mdPlotEfit-1.png" alt="LumA vs Other - Genes Identified at FDR = 0,05" width="1056" />
<p class="caption">
(#fig:brcaRna-preproc-mdPlotEfit)LumA vs Other - Genes Identified at FDR = 0,05
</p>
</div>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1"></a><span class="co">#SKIP - DONT HAVE ACGT </span></span>
<span id="cb128-2"><a href="#cb128-2"></a>SKIP &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb128-3"><a href="#cb128-3"></a><span class="co"># gc vs identification</span></span>
<span id="cb128-4"><a href="#cb128-4"></a>genes_ndx &lt;-<span class="st"> </span><span class="kw">match</span>(<span class="kw">rownames</span>(brcaRna_geneExpression_F_efit), brcaRna_genes_annotF<span class="op">$</span>geneSymbol)</span>
<span id="cb128-5"><a href="#cb128-5"></a><span class="cf">if</span>(<span class="kw">sum</span>(<span class="kw">is.na</span>(genes_ndx))) <span class="kw">stop</span>(<span class="st">&quot;brcaRna_geneExpression_F_efit/brcaRna_genes_annotF: genes mismatch&quot;</span>)</span>
<span id="cb128-6"><a href="#cb128-6"></a>GC_vec &lt;-<span class="st"> </span><span class="kw">with</span>(brcaRna_genes_annotF[genes_ndx,],(G<span class="op">+</span>C)<span class="op">/</span>(A<span class="op">+</span>C<span class="op">+</span>G<span class="op">+</span>T))</span>
<span id="cb128-7"><a href="#cb128-7"></a></span>
<span id="cb128-8"><a href="#cb128-8"></a></span>
<span id="cb128-9"><a href="#cb128-9"></a><span class="kw">boxplot</span>(<span class="kw">split</span>(</span>
<span id="cb128-10"><a href="#cb128-10"></a> GC_vec,</span>
<span id="cb128-11"><a href="#cb128-11"></a> brcaRna_geneExpression_F_efit_dt[,<span class="st">&#39;LumAvsOther&#39;</span>]),</span>
<span id="cb128-12"><a href="#cb128-12"></a> <span class="dt">outline=</span>F,</span>
<span id="cb128-13"><a href="#cb128-13"></a> <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&quot;pink&quot;</span>, <span class="st">&quot;grey&quot;</span>, <span class="st">&quot;lightblue&quot;</span>), <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>,</span>
<span id="cb128-14"><a href="#cb128-14"></a> <span class="dt">ylab=</span><span class="st">&#39;gene-gc&#39;</span>, <span class="dt">cex.lab=</span><span class="fl">1.5</span></span>
<span id="cb128-15"><a href="#cb128-15"></a>)</span>
<span id="cb128-16"><a href="#cb128-16"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="kw">c</span>(<span class="st">&#39;down&#39;</span>, <span class="st">&#39;notDE&#39;</span>, <span class="st">&#39;up&#39;</span>), <span class="dt">cex.axis=</span><span class="fl">1.5</span>)</span>
<span id="cb128-17"><a href="#cb128-17"></a></span>
<span id="cb128-18"><a href="#cb128-18"></a> <span class="co">#mtext(side=3, outer=T, cex=1.25, &quot;Genes identified at adjusted p-value=0.05&quot;)</span></span>
<span id="cb128-19"><a href="#cb128-19"></a>}<span class="co">#SKIP</span></span></code></pre></div>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1"></a><span class="co">###  CLEAR CACHE</span></span>
<span id="cb129-2"><a href="#cb129-2"></a></span>
<span id="cb129-3"><a href="#cb129-3"></a>brcaRna_geneExpression_F_logFC_sum &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb129-4"><a href="#cb129-4"></a> <span class="kw">split</span>(</span>
<span id="cb129-5"><a href="#cb129-5"></a> brcaRna_geneExpression_F_efit<span class="op">$</span>coefficients[, <span class="st">&#39;LumAvsOther&#39;</span>],</span>
<span id="cb129-6"><a href="#cb129-6"></a> brcaRna_geneExpression_F_efit_dt[,<span class="st">&#39;LumAvsOther&#39;</span>]),</span>
<span id="cb129-7"><a href="#cb129-7"></a> quantile, <span class="dt">prob =</span> (<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>)</span>
<span id="cb129-8"><a href="#cb129-8"></a></span>
<span id="cb129-9"><a href="#cb129-9"></a><span class="kw">colnames</span>(brcaRna_geneExpression_F_logFC_sum) &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">factor</span>(</span>
<span id="cb129-10"><a href="#cb129-10"></a> <span class="kw">colnames</span>(brcaRna_geneExpression_F_logFC_sum), </span>
<span id="cb129-11"><a href="#cb129-11"></a> <span class="dt">levels=</span><span class="kw">c</span>(<span class="st">&quot;-1&quot;</span>, <span class="st">&quot;0&quot;</span>, <span class="st">&quot;1&quot;</span>),</span>
<span id="cb129-12"><a href="#cb129-12"></a> <span class="dt">labels=</span><span class="kw">c</span>(<span class="st">&#39;down&#39;</span>, <span class="st">&#39;notDE&#39;</span>, <span class="st">&#39;up&#39;</span>)</span>
<span id="cb129-13"><a href="#cb129-13"></a>))</span>
<span id="cb129-14"><a href="#cb129-14"></a></span>
<span id="cb129-15"><a href="#cb129-15"></a></span>
<span id="cb129-16"><a href="#cb129-16"></a>knitr<span class="op">::</span><span class="kw">kable</span>(brcaRna_geneExpression_F_logFC_sum,</span>
<span id="cb129-17"><a href="#cb129-17"></a>  <span class="dt">digits =</span> <span class="dv">2</span>,</span>
<span id="cb129-18"><a href="#cb129-18"></a>  <span class="dt">caption =</span> <span class="st">&quot;log FC quartiles by gene identification&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb129-19"><a href="#cb129-19"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:brcaRna-preproc-quantlogFC)log FC quartiles by gene identification
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
down
</th>
<th style="text-align:right;">
notDE
</th>
<th style="text-align:right;">
up
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
25%
</td>
<td style="text-align:right;">
-0.28
</td>
<td style="text-align:right;">
-0.03
</td>
<td style="text-align:right;">
0.11
</td>
</tr>
<tr>
<td style="text-align:left;">
50%
</td>
<td style="text-align:right;">
-0.17
</td>
<td style="text-align:right;">
0.00
</td>
<td style="text-align:right;">
0.20
</td>
</tr>
<tr>
<td style="text-align:left;">
75%
</td>
<td style="text-align:right;">
-0.10
</td>
<td style="text-align:right;">
0.03
</td>
<td style="text-align:right;">
0.37
</td>
</tr>
</tbody>
</table>
<p>Many genes are identified, and the effect sizes are substantial
in comparison to the 5hmc data - compare tables
@ref(tab:preprocBrCa-quantlogFC) and
@ref(tab:quantlogFC).
This will result in a higher signal-to-noise ratio context. See
Section @ref(breast-rnaseq-snr-regime).</p>
<!-- 5hmc comments
The log-fold-change distribution for up-represented genes is long-tailed,
with many high log fold-change values.
By contrast, log-fold-change distribution for down-represented genes
closer to symmetric and has few genes with low log fold-change values.
We will see how this affects the results of identifying genes with
an effect size requirement.

The GC content of down regulated genes tends to be slightly lower than the
rest of the genes.  A statistical test would find that the difference
between the mean of the down regulated gene population is singificantly different
than the mean of the other gene population even though the difference is
quite small
(TICKr round( 
mean(GC_vec[brcaRna_geneExpression_F_efit_dt[,'LumAvsOther']=='-1']) -
mean(GC_vec[brcaRna_geneExpression_F_efit_dt[,'LumAvsOther']!='-1']),
3)TICK

These asymmetries are minor, but it would still be good to establish that
they relfect biology rather than processing artifacts.  
-->
</div>
<div id="de-genes-at-10-fold-change-1" class="section level3 unnumbered" number="">
<h3 class="unnumbered" number="">DE genes at 10% fold change</h3>
<p>For a stricter definition on significance, one may require log-fold-changes
(log-FCs) to be above a minimum value. The treat method
(McCarthy and Smyth 2009 <span class="citation">[<a href="#ref-McCarthy:2009aa" role="doc-biblioref">23</a>]</span>) can be used to calculate p-values
from empirical Bayes moderated t-statistics with a minimum log-FC requirement.
The number of differentially expressed genes are greatly reduced if we
impose a minimal fold-change requirement of 10%.</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1"></a><span class="co">###  CLEAR CACHE</span></span>
<span id="cb130-2"><a href="#cb130-2"></a></span>
<span id="cb130-3"><a href="#cb130-3"></a>brcaRna_geneExpression_F_tfit &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">treat</span>(brcaRna_geneExpression_F_fit, <span class="dt">lfc=</span><span class="kw">log2</span>(<span class="fl">1.10</span>))</span>
<span id="cb130-4"><a href="#cb130-4"></a>brcaRna_geneExpression_F_tfit_dt &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">decideTests</span>(brcaRna_geneExpression_F_tfit)</span>
<span id="cb130-5"><a href="#cb130-5"></a></span>
<span id="cb130-6"><a href="#cb130-6"></a><span class="kw">cat</span>(<span class="st">&quot;10% FC Gene Identification Summary - voom, adjust.method = BH, p.value = 0.05:</span><span class="ch">\n</span><span class="st">&quot;</span>)</span></code></pre></div>
<pre><code>## 10% FC Gene Identification Summary - voom, adjust.method = BH, p.value = 0.05:</code></pre>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1"></a><span class="kw">summary</span>(brcaRna_geneExpression_F_tfit_dt)</span></code></pre></div>
<pre><code>##        LumAvsOther
## Down          2569
## NotSig       12519
## Up            4318</code></pre>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1"></a><span class="co"># log-fold-change vs ave-expr</span></span>
<span id="cb134-2"><a href="#cb134-2"></a>limma<span class="op">::</span><span class="kw">plotMD</span>(brcaRna_geneExpression_F_efit,</span>
<span id="cb134-3"><a href="#cb134-3"></a> <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.6</span>, <span class="fl">0.6</span>),</span>
<span id="cb134-4"><a href="#cb134-4"></a> <span class="dt">column=</span><span class="st">&#39;LumAvsOther&#39;</span>,</span>
<span id="cb134-5"><a href="#cb134-5"></a> <span class="dt">status=</span>brcaRna_geneExpression_F_tfit_dt[,<span class="st">&#39;LumAvsOther&#39;</span>],</span>
<span id="cb134-6"><a href="#cb134-6"></a> <span class="dt">hl.pch =</span> <span class="dv">16</span>, <span class="dt">hl.col =</span> <span class="kw">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;red&quot;</span>), <span class="dt">hl.cex =</span> <span class="fl">.7</span>,</span>
<span id="cb134-7"><a href="#cb134-7"></a> <span class="dt">bg.pch =</span> <span class="dv">16</span>, <span class="dt">bg.col =</span> <span class="st">&quot;grey&quot;</span>, <span class="dt">bg.cex =</span> <span class="fl">0.5</span>,</span>
<span id="cb134-8"><a href="#cb134-8"></a> <span class="dt">main =</span> <span class="st">&#39;&#39;</span>,</span>
<span id="cb134-9"><a href="#cb134-9"></a> <span class="dt">xlab =</span> <span class="kw">paste0</span>(</span>
<span id="cb134-10"><a href="#cb134-10"></a>    <span class="st">&quot;Average log-expression: IQR=&quot;</span>,</span>
<span id="cb134-11"><a href="#cb134-11"></a>    <span class="kw">paste</span>(<span class="kw">round</span>(<span class="kw">quantile</span>(brcaRna_geneExpression_F_efit<span class="op">$</span>Amean, <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>), <span class="dv">2</span>),</span>
<span id="cb134-12"><a href="#cb134-12"></a>      <span class="dt">collapse =</span> <span class="st">&quot;, &quot;</span></span>
<span id="cb134-13"><a href="#cb134-13"></a>    )</span>
<span id="cb134-14"><a href="#cb134-14"></a>  ),</span>
<span id="cb134-15"><a href="#cb134-15"></a>  <span class="dt">ylab =</span> <span class="kw">paste0</span>(</span>
<span id="cb134-16"><a href="#cb134-16"></a>    <span class="st">&quot;log-fold-change: IQR=&quot;</span>,</span>
<span id="cb134-17"><a href="#cb134-17"></a>    <span class="kw">paste</span>(<span class="kw">round</span>(<span class="kw">quantile</span>(brcaRna_geneExpression_F_efit<span class="op">$</span>coefficients[, <span class="st">&#39;LumAvsOther&#39;</span>], <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>), <span class="dv">2</span>),</span>
<span id="cb134-18"><a href="#cb134-18"></a>      <span class="dt">collapse =</span> <span class="st">&quot;, &quot;</span></span>
<span id="cb134-19"><a href="#cb134-19"></a>    )</span>
<span id="cb134-20"><a href="#cb134-20"></a>  ),</span>
<span id="cb134-21"><a href="#cb134-21"></a>  <span class="dt">legend =</span> F</span>
<span id="cb134-22"><a href="#cb134-22"></a>)</span>
<span id="cb134-23"><a href="#cb134-23"></a><span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">0</span>, <span class="dt">col =</span> <span class="st">&quot;black&quot;</span>)</span>
<span id="cb134-24"><a href="#cb134-24"></a><span class="kw">rug</span>(<span class="kw">quantile</span>(brcaRna_geneExpression_F_efit<span class="op">$</span>coefficients[, <span class="st">&#39;LumAvsOther&#39;</span>], <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>),</span>
<span id="cb134-25"><a href="#cb134-25"></a>  <span class="dt">col =</span> <span class="st">&quot;purple&quot;</span>,</span>
<span id="cb134-26"><a href="#cb134-26"></a>  <span class="dt">ticksize =</span> <span class="fl">.03</span>, <span class="dt">side =</span> <span class="dv">2</span>, <span class="dt">lwd =</span> <span class="dv">2</span></span>
<span id="cb134-27"><a href="#cb134-27"></a>)</span>
<span id="cb134-28"><a href="#cb134-28"></a><span class="kw">rug</span>(<span class="kw">quantile</span>(brcaRna_geneExpression_F_efit<span class="op">$</span>Amean, <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>),</span>
<span id="cb134-29"><a href="#cb134-29"></a>  <span class="dt">col =</span> <span class="st">&quot;purple&quot;</span>,</span>
<span id="cb134-30"><a href="#cb134-30"></a>  <span class="dt">ticksize =</span> <span class="fl">.03</span>, <span class="dt">side =</span> <span class="dv">1</span>, <span class="dt">lwd =</span> <span class="dv">2</span></span>
<span id="cb134-31"><a href="#cb134-31"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-preproc-mdPlotTfit-1.png" alt="LumA vs Other - Identified Genes at FDR = 0,05 and logFC &gt; 10%" width="1056" />
<p class="caption">
(#fig:brcaRna-preproc-mdPlotTfit)LumA vs Other - Identified Genes at FDR = 0,05 and logFC &gt; 10%
</p>
</div>
<!-- 5hmc comment - does not apply to this RNA-Seq dataset
As noted above, the log-fold-change distribution for the up-represented genes
is long-tailes in  comparison to log-fold-change distribution for the down-represented genes.
As a result fewer down-represented than up-regulated genes are identified when a 
minimum log-FC requirement is imposed.
-->
</div>
</div>
<div id="breast-rnaseq-snr-regime" class="section level2" number="6.3">
<h2 number="6.3"><span class="header-section-number">6.3</span> Signal-to-noise ratio regime</h2>
<p>In Hastie et al. (2017) <span class="citation">[<a href="#ref-Hastie:2017aa" role="doc-biblioref">7</a>]</span>) results from <code>lasso</code> fits are
compared with <code>best subset</code> and <code>forward selection</code> fits and it is argued
that while <code>best subset</code> is optimal for high signal-to-noise regimes,
the lasso gains some competitive advantage when the prevailing signal-to-noise
ratio of the dataset is lowered.</p>
<!-- DNA ???
We can extract sigma and signal from the fit objects to get SNR values for each gene
to see in what SNR regime the 5hmC gene body data are.

-->
<!-- DEBUG BCV from Section \@ref(analysis-of-coverage-variability) vs CV
pairs(cbind(BCV_mtx, CV)) 
boxplot(cbind(BCV_mtx, CV), outline=F) 
-->
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1"></a><span class="co">###  CLEAR CACHE</span></span>
<span id="cb135-2"><a href="#cb135-2"></a></span>
<span id="cb135-3"><a href="#cb135-3"></a>Effect &lt;-<span class="st"> </span><span class="kw">abs</span>(brcaRna_geneExpression_F_efit<span class="op">$</span>coefficients[,<span class="st">&#39;LumAvsOther&#39;</span>])</span>
<span id="cb135-4"><a href="#cb135-4"></a>Noise &lt;-<span class="st"> </span>brcaRna_geneExpression_F_efit<span class="op">$</span>sigma</span>
<span id="cb135-5"><a href="#cb135-5"></a>SNR &lt;-<span class="st"> </span>Effect<span class="op">/</span>Noise</span>
<span id="cb135-6"><a href="#cb135-6"></a></span>
<span id="cb135-7"><a href="#cb135-7"></a><span class="kw">plot</span>(spatstat<span class="op">::</span><span class="kw">CDF</span>(<span class="kw">density</span>(SNR)),</span>
<span id="cb135-8"><a href="#cb135-8"></a>  <span class="dt">col =</span> <span class="dv">1</span>, <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">ylab =</span> <span class="st">&quot;Prob(SNR&lt;x)&quot;</span>,</span>
<span id="cb135-9"><a href="#cb135-9"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">1.5</span>)</span>
<span id="cb135-10"><a href="#cb135-10"></a>)</span>
<span id="cb135-11"><a href="#cb135-11"></a></span>
<span id="cb135-12"><a href="#cb135-12"></a>SNR_quant &lt;-<span class="st"> </span><span class="kw">quantile</span>(SNR, <span class="dt">prob=</span><span class="kw">c</span>((<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>)<span class="op">/</span><span class="dv">4</span>,.<span class="dv">9</span>))</span>
<span id="cb135-13"><a href="#cb135-13"></a><span class="kw">rug</span>(SNR_quant,</span>
<span id="cb135-14"><a href="#cb135-14"></a>    <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">ticksize =</span> <span class="fl">0.05</span>, <span class="dt">col =</span> <span class="dv">1</span></span>
<span id="cb135-15"><a href="#cb135-15"></a>  )</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-preproc-plotSNR-1.png" alt="Cumulative Distribution of SNR - rug = 25, 50, 75 and 90th percentile" width="960" />
<p class="caption">
(#fig:brcaRna-preproc-plotSNR)Cumulative Distribution of SNR - rug = 25, 50, 75 and 90th percentile
</p>
</div>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">t</span>(SNR_quant),</span>
<span id="cb136-2"><a href="#cb136-2"></a>  <span class="dt">digits =</span> <span class="dv">3</span>,</span>
<span id="cb136-3"><a href="#cb136-3"></a>  <span class="dt">caption =</span> <span class="kw">paste</span>(</span>
<span id="cb136-4"><a href="#cb136-4"></a>    <span class="st">&quot;SNR Quantiles&quot;</span>) </span>
<span id="cb136-5"><a href="#cb136-5"></a>) <span class="op">%&gt;%</span><span class="st"> </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:brcaRna-preproc-plotSNR)SNR Quantiles
</caption>
<thead>
<tr>
<th style="text-align:right;">
25%
</th>
<th style="text-align:right;">
50%
</th>
<th style="text-align:right;">
75%
</th>
<th style="text-align:right;">
90%
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
0.094
</td>
<td style="text-align:right;">
0.217
</td>
<td style="text-align:right;">
0.384
</td>
<td style="text-align:right;">
0.56
</td>
</tr>
</tbody>
</table>
<!--
These SNR values are in the range where the lasso and relaxed lasso gain some advantage over
best subset and forward selection fits (see  Hastie et al. (2017) [@Hastie:2017aa]).
-->
<!--chapter:end:06-preprocBrCaRNASeq.Rmd-->
</div>
</div>
<div id="brca-rnaseseq-explore-sparsity" class="section level1" number="7">
<h1 number="7"><span class="header-section-number">7</span> BrCa RNA-Seq: Exploring Sparsity</h1>
<p>In this section we explore various models fitted to
the Breast Cancer RNA-Seq data set explored in Section @ref(brca-rnaseq-preproc).
We focus our analyses on lasso fits which tend to favor sparse models.</p>
<div id="cross-validation-analysis-setup-1" class="section level2" number="7.1">
<h2 number="7.1"><span class="header-section-number">7.1</span> Cross-validation analysis setup</h2>
<p>We use the same CV set-up as was used with the HCC 5hmC-Seq data set
(Section @ref(hcc-5hmcseq-explore-sparsity)).</p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1"></a>K_FOLD &lt;-<span class="st"> </span><span class="dv">10</span></span>
<span id="cb137-2"><a href="#cb137-2"></a>trainP &lt;-<span class="st"> </span><span class="fl">0.8</span></span></code></pre></div>
<!-- NOT CURRENTLY USED 
EPS <- 0.05    # Have no idea what "small" epsilon means
-->
<p>First we divide the analysis dataset into <code>train</code> and <code>test</code> in a <span class="math inline">\(4\)</span>:1 ratio.</p>
<div class="sourceCode" id="cb138"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb138-2"><a href="#cb138-2"></a></span>
<span id="cb138-3"><a href="#cb138-3"></a><span class="kw">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb138-4"><a href="#cb138-4"></a>brcaRna_train_sampID_vec &lt;-<span class="st"> </span><span class="kw">with</span>(brcaRna_sampDesc,</span>
<span id="cb138-5"><a href="#cb138-5"></a>brcaRna_sampDesc<span class="op">$</span>sampID[caret<span class="op">::</span><span class="kw">createDataPartition</span>(<span class="dt">y=</span>group, <span class="dt">p=</span>trainP, <span class="dt">list=</span>F)]</span>
<span id="cb138-6"><a href="#cb138-6"></a>)</span>
<span id="cb138-7"><a href="#cb138-7"></a></span>
<span id="cb138-8"><a href="#cb138-8"></a>brcaRna_test_sampID_vec &lt;-<span class="st"> </span><span class="kw">with</span>(brcaRna_sampDesc,</span>
<span id="cb138-9"><a href="#cb138-9"></a><span class="kw">setdiff</span>(sampID, brcaRna_train_sampID_vec)</span>
<span id="cb138-10"><a href="#cb138-10"></a>)</span>
<span id="cb138-11"><a href="#cb138-11"></a></span>
<span id="cb138-12"><a href="#cb138-12"></a>brcaRna_train_group_vec &lt;-<span class="st"> </span><span class="kw">factor</span>(brcaRna_sampDesc[brcaRna_train_sampID_vec, <span class="st">&#39;group&#39;</span>],</span>
<span id="cb138-13"><a href="#cb138-13"></a><span class="dt">levels=</span><span class="kw">c</span>(<span class="st">&#39;Other&#39;</span>, <span class="st">&#39;LumA&#39;</span>))</span>
<span id="cb138-14"><a href="#cb138-14"></a><span class="kw">names</span>(brcaRna_train_group_vec) &lt;-<span class="st"> </span>brcaRna_sampDesc[brcaRna_train_sampID_vec, <span class="st">&#39;sampID&#39;</span>]</span>
<span id="cb138-15"><a href="#cb138-15"></a></span>
<span id="cb138-16"><a href="#cb138-16"></a>brcaRna_test_group_vec &lt;-<span class="st"> </span><span class="kw">factor</span>(brcaRna_sampDesc[brcaRna_test_sampID_vec, <span class="st">&#39;group&#39;</span>],</span>
<span id="cb138-17"><a href="#cb138-17"></a><span class="dt">levels=</span><span class="kw">c</span>(<span class="st">&#39;Other&#39;</span>, <span class="st">&#39;LumA&#39;</span>))</span>
<span id="cb138-18"><a href="#cb138-18"></a><span class="kw">names</span>(brcaRna_test_group_vec) &lt;-<span class="st"> </span>brcaRna_sampDesc[brcaRna_test_sampID_vec, <span class="st">&#39;sampID&#39;</span>]</span>
<span id="cb138-19"><a href="#cb138-19"></a></span>
<span id="cb138-20"><a href="#cb138-20"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">table</span>(brcaRna_train_group_vec),</span>
<span id="cb138-21"><a href="#cb138-21"></a>  <span class="dt">caption=</span><span class="st">&quot;Train set&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb138-22"><a href="#cb138-22"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:brcaRna-glmnetFit-getTrainVal)Train set
</caption>
<thead>
<tr>
<th style="text-align:left;">
brcaRna_train_group_vec
</th>
<th style="text-align:right;">
Freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Other
</td>
<td style="text-align:right;">
747
</td>
</tr>
<tr>
<td style="text-align:left;">
LumA
</td>
<td style="text-align:right;">
1194
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">table</span>(brcaRna_test_group_vec),</span>
<span id="cb139-2"><a href="#cb139-2"></a>  <span class="dt">caption=</span><span class="st">&quot;Test set&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb139-3"><a href="#cb139-3"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:brcaRna-glmnetFit-getTrainVal)Test set
</caption>
<thead>
<tr>
<th style="text-align:left;">
brcaRna_test_group_vec
</th>
<th style="text-align:right;">
Freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Other
</td>
<td style="text-align:right;">
186
</td>
</tr>
<tr>
<td style="text-align:left;">
LumA
</td>
<td style="text-align:right;">
298
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb140"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb140-1"><a href="#cb140-1"></a>brcaRna_train_geneExpr_mtx &lt;-<span class="st"> </span><span class="kw">t</span>(brcaRna_geneExpression_F[,brcaRna_train_sampID_vec])</span>
<span id="cb140-2"><a href="#cb140-2"></a>brcaRna_test_geneExpr_mtx &lt;-<span class="st"> </span><span class="kw">t</span>(brcaRna_geneExpression_F[,brcaRna_test_sampID_vec])</span></code></pre></div>
<p>We explore some glmnet fits and the “bet on sparsity”.
We consider three models, specified by the value of the
<strong>alpha</strong> parameter in the elastic net parametrization:<br />
- lasso: <span class="math inline">\(\alpha = 1.0\)</span> - sparse models<br />
- ridge <span class="math inline">\(\alpha = 0\)</span> - shrunken coefficients models<br />
- elastic net: <span class="math inline">\(\alpha = 0.5\)</span> - semi sparse model<br />
<!-- - lassoC: $\alpha = 1-\epsilon =$ $TICKr 1- EPSTICK - lasso for correlated predictors  --></p>
<p>In this analysis, we will only evaluate models in terms of
model sparsity, stability and performance. We leave the question
of significance testing of hypotheses about model parameters
completely out. See Lockhart et al. (2014) <span class="citation">[<a href="#ref-Lockhart:2014aa" role="doc-biblioref">24</a>]</span>
and Wassermam (2014) <span class="citation">[<a href="#ref-Wasserman:2014aa" role="doc-biblioref">25</a>]</span> for a discussion of this topic.</p>
<p>In this section we look at the relative performance and sparsity of the models
considered. The effect of the size of the sample set on the level and
stability of performance will be investigated in the next section.</p>
<hr />
<p>First we create folds for <span class="math inline">\(10\)</span>-fold cross-validation of models fitted to
training data. We’ll use caret::createFolds to assign samples
to folds while keeping the outcome ratios constant across folds.</p>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1"></a><span class="co"># This is too variable, both in terms of fold size And composition</span></span>
<span id="cb141-2"><a href="#cb141-2"></a><span class="co">#foldid_vec &lt;- sample(1:10, size=length(brcaRna_train_group_vec), replace=T)</span></span>
<span id="cb141-3"><a href="#cb141-3"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb141-4"><a href="#cb141-4"></a></span>
<span id="cb141-5"><a href="#cb141-5"></a><span class="kw">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb141-6"><a href="#cb141-6"></a>brcaRna_train_foldid_vec &lt;-<span class="st"> </span>caret<span class="op">::</span><span class="kw">createFolds</span>(</span>
<span id="cb141-7"><a href="#cb141-7"></a> <span class="kw">factor</span>(brcaRna_train_group_vec), </span>
<span id="cb141-8"><a href="#cb141-8"></a> <span class="dt">k=</span>K_FOLD,</span>
<span id="cb141-9"><a href="#cb141-9"></a> <span class="dt">list=</span>F)</span>
<span id="cb141-10"><a href="#cb141-10"></a></span>
<span id="cb141-11"><a href="#cb141-11"></a><span class="co"># brcaRna_train_foldid_vec contains the left-out IDs </span></span>
<span id="cb141-12"><a href="#cb141-12"></a><span class="co"># the rest are kept</span></span>
<span id="cb141-13"><a href="#cb141-13"></a>fold_out_tbl &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(brcaRna_train_group_vec, brcaRna_train_foldid_vec),</span>
<span id="cb141-14"><a href="#cb141-14"></a>  table)</span>
<span id="cb141-15"><a href="#cb141-15"></a><span class="kw">rownames</span>(fold_out_tbl) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">rownames</span>(fold_out_tbl), <span class="st">&#39;- Out&#39;</span>) </span>
<span id="cb141-16"><a href="#cb141-16"></a></span>
<span id="cb141-17"><a href="#cb141-17"></a>fold_in_tbl &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="st">&#39;cbind&#39;</span>, <span class="kw">lapply</span>(<span class="kw">sort</span>(<span class="kw">unique</span>(brcaRna_train_foldid_vec)),</span>
<span id="cb141-18"><a href="#cb141-18"></a>  <span class="cf">function</span>(FOLD) <span class="kw">table</span>(brcaRna_train_group_vec[brcaRna_train_foldid_vec <span class="op">!=</span><span class="st"> </span>FOLD])))</span>
<span id="cb141-19"><a href="#cb141-19"></a><span class="kw">rownames</span>(fold_in_tbl) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">rownames</span>(fold_in_tbl), <span class="st">&#39;- In&#39;</span>) </span>
<span id="cb141-20"><a href="#cb141-20"></a><span class="kw">colnames</span>(fold_in_tbl) &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">sort</span>(<span class="kw">unique</span>(brcaRna_train_foldid_vec)))</span>
<span id="cb141-21"><a href="#cb141-21"></a></span>
<span id="cb141-22"><a href="#cb141-22"></a></span>
<span id="cb141-23"><a href="#cb141-23"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">rbind</span>(fold_in_tbl, fold_out_tbl[,<span class="kw">colnames</span>(fold_in_tbl)]),</span>
<span id="cb141-24"><a href="#cb141-24"></a>  <span class="dt">caption=</span><span class="st">&quot;training samples fold composition&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb141-25"><a href="#cb141-25"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:brcaRna-glmnetFit-getTrainFolds)training samples fold composition
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
1
</th>
<th style="text-align:right;">
2
</th>
<th style="text-align:right;">
3
</th>
<th style="text-align:right;">
4
</th>
<th style="text-align:right;">
5
</th>
<th style="text-align:right;">
6
</th>
<th style="text-align:right;">
7
</th>
<th style="text-align:right;">
8
</th>
<th style="text-align:right;">
9
</th>
<th style="text-align:right;">
10
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Other - In
</td>
<td style="text-align:right;">
673
</td>
<td style="text-align:right;">
673
</td>
<td style="text-align:right;">
672
</td>
<td style="text-align:right;">
672
</td>
<td style="text-align:right;">
672
</td>
<td style="text-align:right;">
672
</td>
<td style="text-align:right;">
672
</td>
<td style="text-align:right;">
672
</td>
<td style="text-align:right;">
673
</td>
<td style="text-align:right;">
672
</td>
</tr>
<tr>
<td style="text-align:left;">
LumA - In
</td>
<td style="text-align:right;">
1074
</td>
<td style="text-align:right;">
1075
</td>
<td style="text-align:right;">
1075
</td>
<td style="text-align:right;">
1074
</td>
<td style="text-align:right;">
1075
</td>
<td style="text-align:right;">
1075
</td>
<td style="text-align:right;">
1074
</td>
<td style="text-align:right;">
1075
</td>
<td style="text-align:right;">
1074
</td>
<td style="text-align:right;">
1075
</td>
</tr>
<tr>
<td style="text-align:left;">
Other - Out
</td>
<td style="text-align:right;">
74
</td>
<td style="text-align:right;">
74
</td>
<td style="text-align:right;">
75
</td>
<td style="text-align:right;">
75
</td>
<td style="text-align:right;">
75
</td>
<td style="text-align:right;">
75
</td>
<td style="text-align:right;">
75
</td>
<td style="text-align:right;">
75
</td>
<td style="text-align:right;">
74
</td>
<td style="text-align:right;">
75
</td>
</tr>
<tr>
<td style="text-align:left;">
LumA - Out
</td>
<td style="text-align:right;">
120
</td>
<td style="text-align:right;">
119
</td>
<td style="text-align:right;">
119
</td>
<td style="text-align:right;">
120
</td>
<td style="text-align:right;">
119
</td>
<td style="text-align:right;">
119
</td>
<td style="text-align:right;">
120
</td>
<td style="text-align:right;">
119
</td>
<td style="text-align:right;">
120
</td>
<td style="text-align:right;">
119
</td>
</tr>
</tbody>
</table>
<p>Note that the folds identify samples that are left-out of the training
data for each fold fit.</p>
</div>
<div id="fit-and-compare-models-1" class="section level2" number="7.2">
<h2 number="7.2"><span class="header-section-number">7.2</span> Fit and compare models</h2>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb142-2"><a href="#cb142-2"></a></span>
<span id="cb142-3"><a href="#cb142-3"></a>start_time &lt;-<span class="st">  </span><span class="kw">proc.time</span>()</span>
<span id="cb142-4"><a href="#cb142-4"></a></span>
<span id="cb142-5"><a href="#cb142-5"></a>brcaRna_cv_lasso &lt;-<span class="st"> </span>glmnet<span class="op">::</span><span class="kw">cv.glmnet</span>(</span>
<span id="cb142-6"><a href="#cb142-6"></a> <span class="dt">x=</span>brcaRna_train_geneExpr_mtx,</span>
<span id="cb142-7"><a href="#cb142-7"></a> <span class="dt">y=</span>brcaRna_train_group_vec,</span>
<span id="cb142-8"><a href="#cb142-8"></a> <span class="dt">foldid=</span>brcaRna_train_foldid_vec,</span>
<span id="cb142-9"><a href="#cb142-9"></a> <span class="dt">alpha=</span><span class="dv">1</span>,</span>
<span id="cb142-10"><a href="#cb142-10"></a> <span class="dt">family=</span><span class="st">&#39;binomial&#39;</span>, </span>
<span id="cb142-11"><a href="#cb142-11"></a> <span class="dt">type.measure =</span> <span class="st">&quot;class&quot;</span>,</span>
<span id="cb142-12"><a href="#cb142-12"></a> <span class="dt">keep=</span>T,</span>
<span id="cb142-13"><a href="#cb142-13"></a> <span class="dt">nlambda=</span><span class="dv">30</span></span>
<span id="cb142-14"><a href="#cb142-14"></a>)</span>
<span id="cb142-15"><a href="#cb142-15"></a></span>
<span id="cb142-16"><a href="#cb142-16"></a><span class="kw">message</span>(<span class="st">&quot;lasso time: &quot;</span>, <span class="kw">round</span>((<span class="kw">proc.time</span>() <span class="op">-</span><span class="st"> </span>start_time)[<span class="dv">3</span>],<span class="dv">2</span>),<span class="st">&quot;s&quot;</span>)</span></code></pre></div>
<pre><code>## lasso time: 31.48s</code></pre>
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb144-1"><a href="#cb144-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb144-2"><a href="#cb144-2"></a></span>
<span id="cb144-3"><a href="#cb144-3"></a>start_time &lt;-<span class="st">  </span><span class="kw">proc.time</span>()</span>
<span id="cb144-4"><a href="#cb144-4"></a></span>
<span id="cb144-5"><a href="#cb144-5"></a>brcaRna_cv_ridge &lt;-<span class="st"> </span>glmnet<span class="op">::</span><span class="kw">cv.glmnet</span>(</span>
<span id="cb144-6"><a href="#cb144-6"></a> <span class="dt">x=</span>brcaRna_train_geneExpr_mtx,</span>
<span id="cb144-7"><a href="#cb144-7"></a> <span class="dt">y=</span>brcaRna_train_group_vec,</span>
<span id="cb144-8"><a href="#cb144-8"></a> <span class="dt">foldid=</span>brcaRna_train_foldid_vec,</span>
<span id="cb144-9"><a href="#cb144-9"></a> <span class="dt">alpha=</span><span class="dv">0</span>,</span>
<span id="cb144-10"><a href="#cb144-10"></a> <span class="dt">family=</span><span class="st">&#39;binomial&#39;</span>, </span>
<span id="cb144-11"><a href="#cb144-11"></a> <span class="dt">type.measure =</span> <span class="st">&quot;class&quot;</span>,</span>
<span id="cb144-12"><a href="#cb144-12"></a> <span class="dt">keep=</span>T,</span>
<span id="cb144-13"><a href="#cb144-13"></a> <span class="dt">nlambda=</span><span class="dv">30</span></span>
<span id="cb144-14"><a href="#cb144-14"></a>)</span>
<span id="cb144-15"><a href="#cb144-15"></a></span>
<span id="cb144-16"><a href="#cb144-16"></a><span class="kw">message</span>(<span class="st">&quot;ridge time: &quot;</span>, <span class="kw">round</span>((<span class="kw">proc.time</span>() <span class="op">-</span><span class="st"> </span>start_time)[<span class="dv">3</span>],<span class="dv">2</span>),<span class="st">&quot;s&quot;</span>)</span></code></pre></div>
<pre><code>## ridge time: 258.93s</code></pre>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb146-2"><a href="#cb146-2"></a></span>
<span id="cb146-3"><a href="#cb146-3"></a>start_time &lt;-<span class="st">  </span><span class="kw">proc.time</span>()</span>
<span id="cb146-4"><a href="#cb146-4"></a></span>
<span id="cb146-5"><a href="#cb146-5"></a>brcaRna_cv_enet &lt;-<span class="st"> </span>glmnet<span class="op">::</span><span class="kw">cv.glmnet</span>(</span>
<span id="cb146-6"><a href="#cb146-6"></a> <span class="dt">x=</span>brcaRna_train_geneExpr_mtx,</span>
<span id="cb146-7"><a href="#cb146-7"></a> <span class="dt">y=</span>brcaRna_train_group_vec,</span>
<span id="cb146-8"><a href="#cb146-8"></a> <span class="dt">foldid=</span>brcaRna_train_foldid_vec,</span>
<span id="cb146-9"><a href="#cb146-9"></a> <span class="dt">alpha=</span><span class="fl">0.5</span>,</span>
<span id="cb146-10"><a href="#cb146-10"></a> <span class="dt">family=</span><span class="st">&#39;binomial&#39;</span>,</span>
<span id="cb146-11"><a href="#cb146-11"></a> <span class="dt">type.measure =</span> <span class="st">&quot;class&quot;</span>,</span>
<span id="cb146-12"><a href="#cb146-12"></a> <span class="dt">keep=</span>T,</span>
<span id="cb146-13"><a href="#cb146-13"></a> <span class="dt">nlambda=</span><span class="dv">30</span></span>
<span id="cb146-14"><a href="#cb146-14"></a>)</span>
<span id="cb146-15"><a href="#cb146-15"></a></span>
<span id="cb146-16"><a href="#cb146-16"></a><span class="kw">message</span>(<span class="st">&quot;enet time: &quot;</span>, <span class="kw">round</span>((<span class="kw">proc.time</span>() <span class="op">-</span><span class="st"> </span>start_time)[<span class="dv">3</span>],<span class="dv">2</span>),<span class="st">&quot;s&quot;</span>)</span></code></pre></div>
<pre><code>## enet time: 32.59s</code></pre>
<!--
The ridge regression model takes over 10 times longer to compute.
-->
<!-- do not show
Define plotting function.
Maybe show in appendix??
-->
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb148-2"><a href="#cb148-2"></a></span>
<span id="cb148-3"><a href="#cb148-3"></a>plot_cv_f &lt;-<span class="st"> </span><span class="cf">function</span>(cv_fit, <span class="dt">Nzero=</span>T, ...) {</span>
<span id="cb148-4"><a href="#cb148-4"></a> </span>
<span id="cb148-5"><a href="#cb148-5"></a> <span class="kw">suppressPackageStartupMessages</span>(<span class="kw">require</span>(glmnet))</span>
<span id="cb148-6"><a href="#cb148-6"></a></span>
<span id="cb148-7"><a href="#cb148-7"></a> <span class="co"># No nonger used</span></span>
<span id="cb148-8"><a href="#cb148-8"></a> <span class="co">#lambda.1se_p &lt;- cv_fit$nzero[cv_fit$lambda == cv_fit$lambda.1se]</span></span>
<span id="cb148-9"><a href="#cb148-9"></a> <span class="co">#lambda.min_p &lt;- cv_fit$nzero[cv_fit$lambda == cv_fit$lambda.min]</span></span>
<span id="cb148-10"><a href="#cb148-10"></a> </span>
<span id="cb148-11"><a href="#cb148-11"></a> <span class="co"># Get oof error - cv errors produced by extraction method ARE oof!!!</span></span>
<span id="cb148-12"><a href="#cb148-12"></a> ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se,cv_fit<span class="op">$</span>lambda)</span>
<span id="cb148-13"><a href="#cb148-13"></a> train_oofPred_1se_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb148-14"><a href="#cb148-14"></a>  <span class="kw">logistic_f</span>(cv_fit<span class="op">$</span>fit.preval[,ndx_1se]) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&#39;LumA&#39;</span>, <span class="st">&#39;_Other&#39;</span>)</span>
<span id="cb148-15"><a href="#cb148-15"></a> train_oofPred_1se_error &lt;-<span class="st"> </span><span class="kw">mean</span>(train_oofPred_1se_vec <span class="op">!=</span><span class="st"> </span>brcaRna_train_group_vec)</span>
<span id="cb148-16"><a href="#cb148-16"></a></span>
<span id="cb148-17"><a href="#cb148-17"></a> ndx_min &lt;-<span class="st"> </span><span class="kw">match</span>(cv_fit<span class="op">$</span>lambda.min,cv_fit<span class="op">$</span>lambda)</span>
<span id="cb148-18"><a href="#cb148-18"></a> train_oofPred_min_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb148-19"><a href="#cb148-19"></a>  <span class="kw">logistic_f</span>(cv_fit<span class="op">$</span>fit.preval[,ndx_min]) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&#39;LumA&#39;</span>, <span class="st">&#39;_Other&#39;</span>)</span>
<span id="cb148-20"><a href="#cb148-20"></a> train_oofPred_min_error &lt;-<span class="st"> </span><span class="kw">mean</span>(train_oofPred_min_vec <span class="op">!=</span><span class="st"> </span>brcaRna_train_group_vec)</span>
<span id="cb148-21"><a href="#cb148-21"></a></span>
<span id="cb148-22"><a href="#cb148-22"></a> <span class="co"># Get test set error</span></span>
<span id="cb148-23"><a href="#cb148-23"></a> test_pred_1se_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb148-24"><a href="#cb148-24"></a>  cv_fit, </span>
<span id="cb148-25"><a href="#cb148-25"></a>  <span class="dt">newx=</span>brcaRna_test_geneExpr_mtx, </span>
<span id="cb148-26"><a href="#cb148-26"></a>  <span class="dt">s=</span><span class="st">&quot;lambda.1se&quot;</span>,</span>
<span id="cb148-27"><a href="#cb148-27"></a>  <span class="dt">type=</span><span class="st">&quot;class&quot;</span></span>
<span id="cb148-28"><a href="#cb148-28"></a> )</span>
<span id="cb148-29"><a href="#cb148-29"></a> test_pred_1se_error &lt;-<span class="st"> </span><span class="kw">mean</span>(test_pred_1se_vec <span class="op">!=</span><span class="st"> </span>brcaRna_test_group_vec)</span>
<span id="cb148-30"><a href="#cb148-30"></a> </span>
<span id="cb148-31"><a href="#cb148-31"></a> test_pred_min_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb148-32"><a href="#cb148-32"></a>  cv_fit, </span>
<span id="cb148-33"><a href="#cb148-33"></a>  <span class="dt">newx=</span>brcaRna_test_geneExpr_mtx, </span>
<span id="cb148-34"><a href="#cb148-34"></a>  <span class="dt">s=</span><span class="st">&quot;lambda.min&quot;</span>,</span>
<span id="cb148-35"><a href="#cb148-35"></a>  <span class="dt">type=</span><span class="st">&quot;class&quot;</span></span>
<span id="cb148-36"><a href="#cb148-36"></a> )</span>
<span id="cb148-37"><a href="#cb148-37"></a> test_pred_min_error &lt;-<span class="st"> </span><span class="kw">mean</span>(test_pred_min_vec <span class="op">!=</span><span class="st"> </span>brcaRna_test_group_vec)</span>
<span id="cb148-38"><a href="#cb148-38"></a> </span>
<span id="cb148-39"><a href="#cb148-39"></a>  </span>
<span id="cb148-40"><a href="#cb148-40"></a> <span class="kw">plot</span>(</span>
<span id="cb148-41"><a href="#cb148-41"></a>  <span class="kw">log</span>(cv_fit<span class="op">$</span>lambda),</span>
<span id="cb148-42"><a href="#cb148-42"></a>  cv_fit<span class="op">$</span>cvm,</span>
<span id="cb148-43"><a href="#cb148-43"></a>  <span class="dt">pch=</span><span class="dv">16</span>,<span class="dt">col=</span><span class="st">&quot;red&quot;</span>,</span>
<span id="cb148-44"><a href="#cb148-44"></a>  <span class="dt">xlab=</span><span class="st">&#39;&#39;</span>,<span class="dt">ylab=</span><span class="st">&#39;&#39;</span>,</span>
<span id="cb148-45"><a href="#cb148-45"></a>  ...</span>
<span id="cb148-46"><a href="#cb148-46"></a> )</span>
<span id="cb148-47"><a href="#cb148-47"></a> <span class="kw">abline</span>(<span class="dt">v=</span><span class="kw">log</span>(<span class="kw">c</span>(cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se, cv_fit<span class="op">$</span>lambda.min)))</span>
<span id="cb148-48"><a href="#cb148-48"></a> <span class="cf">if</span>(Nzero)</span>
<span id="cb148-49"><a href="#cb148-49"></a> <span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">tick=</span>F, <span class="dt">at=</span><span class="kw">log</span>(cv_fit<span class="op">$</span>lambda), </span>
<span id="cb148-50"><a href="#cb148-50"></a>  <span class="dt">labels=</span>cv_fit<span class="op">$</span>nzero, <span class="dt">line =</span> <span class="dv">-1</span></span>
<span id="cb148-51"><a href="#cb148-51"></a> )</span>
<span id="cb148-52"><a href="#cb148-52"></a> LL &lt;-<span class="st"> </span><span class="dv">2</span></span>
<span id="cb148-53"><a href="#cb148-53"></a> <span class="co">#mtext(side=1, outer=F, line = LL, &quot;log(Lambda)&quot;)</span></span>
<span id="cb148-54"><a href="#cb148-54"></a> <span class="co">#LL &lt;- LL+1</span></span>
<span id="cb148-55"><a href="#cb148-55"></a> <span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">outer=</span>F, <span class="dt">line =</span> LL, <span class="kw">paste</span>(</span>
<span id="cb148-56"><a href="#cb148-56"></a>  <span class="co">#ifelse(Nzero, paste(&quot;1se p =&quot;, lambda.1se_p),&#39;&#39;),</span></span>
<span id="cb148-57"><a href="#cb148-57"></a>  <span class="st">&quot;1se: train =&quot;</span>, <span class="kw">round</span>(<span class="dv">100</span><span class="op">*</span>cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se], <span class="dv">1</span>),</span>
<span id="cb148-58"><a href="#cb148-58"></a>  <span class="co">##&quot;oof =&quot;, round(100*train_oofPred_1se_error, 1), </span><span class="al">###</span><span class="co"> REDUNDANT</span></span>
<span id="cb148-59"><a href="#cb148-59"></a>  <span class="st">&quot;test =&quot;</span>, <span class="kw">round</span>(<span class="dv">100</span><span class="op">*</span>test_pred_1se_error, <span class="dv">1</span>)</span>
<span id="cb148-60"><a href="#cb148-60"></a> ))</span>
<span id="cb148-61"><a href="#cb148-61"></a> LL &lt;-<span class="st"> </span>LL<span class="op">+</span><span class="dv">1</span></span>
<span id="cb148-62"><a href="#cb148-62"></a> <span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">outer=</span>F, <span class="dt">line =</span> LL, <span class="kw">paste</span>(</span>
<span id="cb148-63"><a href="#cb148-63"></a>  <span class="co">#ifelse(Nzero, paste(&quot;min p =&quot;, lambda.min_p),&#39;&#39;),</span></span>
<span id="cb148-64"><a href="#cb148-64"></a>  <span class="st">&quot;min: train =&quot;</span>, <span class="kw">round</span>(<span class="dv">100</span><span class="op">*</span>cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min], <span class="dv">1</span>),</span>
<span id="cb148-65"><a href="#cb148-65"></a>  <span class="co">##&quot;oof =&quot;, round(100*train_oofPred_min_error, 1),  </span><span class="al">###</span><span class="co"> REDUNDANT</span></span>
<span id="cb148-66"><a href="#cb148-66"></a>  <span class="st">&quot;test =&quot;</span>, <span class="kw">round</span>(<span class="dv">100</span><span class="op">*</span>test_pred_min_error, <span class="dv">1</span>)</span>
<span id="cb148-67"><a href="#cb148-67"></a> ))</span>
<span id="cb148-68"><a href="#cb148-68"></a> </span>
<span id="cb148-69"><a href="#cb148-69"></a> tmp &lt;-</span>
<span id="cb148-70"><a href="#cb148-70"></a><span class="st"> </span><span class="kw">cbind</span>(</span>
<span id="cb148-71"><a href="#cb148-71"></a>  <span class="dt">error_1se =</span> <span class="kw">c</span>(</span>
<span id="cb148-72"><a href="#cb148-72"></a>   <span class="dt">p =</span> cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se],</span>
<span id="cb148-73"><a href="#cb148-73"></a>   <span class="dt">train  =</span> <span class="dv">100</span><span class="op">*</span>cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se],</span>
<span id="cb148-74"><a href="#cb148-74"></a>   <span class="co">#train_oof = 100*train_oofPred_1se_error,  </span><span class="al">###</span><span class="co"> REDUNANT</span></span>
<span id="cb148-75"><a href="#cb148-75"></a>   <span class="dt">test =</span> <span class="dv">100</span><span class="op">*</span>test_pred_1se_error),</span>
<span id="cb148-76"><a href="#cb148-76"></a>  <span class="dt">error_min =</span> <span class="kw">c</span>(</span>
<span id="cb148-77"><a href="#cb148-77"></a>   <span class="dt">p =</span> cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min],</span>
<span id="cb148-78"><a href="#cb148-78"></a>   <span class="dt">train  =</span> <span class="dv">100</span><span class="op">*</span>cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min],</span>
<span id="cb148-79"><a href="#cb148-79"></a>   <span class="co">#train_oof = 100*train_oofPred_min_error, </span><span class="al">###</span><span class="co"> REDUNDSANT</span></span>
<span id="cb148-80"><a href="#cb148-80"></a>   <span class="dt">test =</span> <span class="dv">100</span><span class="op">*</span>test_pred_min_error)</span>
<span id="cb148-81"><a href="#cb148-81"></a>  )</span>
<span id="cb148-82"><a href="#cb148-82"></a>  <span class="co"># Need to fix names  </span></span>
<span id="cb148-83"><a href="#cb148-83"></a>  <span class="kw">rownames</span>(tmp) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;p&#39;</span>, <span class="st">&#39;train&#39;</span>, <span class="st">&#39;test&#39;</span>)</span>
<span id="cb148-84"><a href="#cb148-84"></a>  tmp </span>
<span id="cb148-85"><a href="#cb148-85"></a>}</span></code></pre></div>
<p>Examine model performance.</p>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb149-2"><a href="#cb149-2"></a></span>
<span id="cb149-3"><a href="#cb149-3"></a> <span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">5</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">0</span>)) </span>
<span id="cb149-4"><a href="#cb149-4"></a></span>
<span id="cb149-5"><a href="#cb149-5"></a> lasso_errors_mtx &lt;-<span class="st"> </span><span class="kw">plot_cv_f</span>(brcaRna_cv_lasso, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,.<span class="dv">5</span>))</span>
<span id="cb149-6"><a href="#cb149-6"></a> <span class="kw">title</span>(<span class="st">&#39;lasso&#39;</span>)</span>
<span id="cb149-7"><a href="#cb149-7"></a></span>
<span id="cb149-8"><a href="#cb149-8"></a> rifge_errors_mtx &lt;-<span class="st"> </span><span class="kw">plot_cv_f</span>(brcaRna_cv_ridge, <span class="dt">Nzero=</span>F, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,.<span class="dv">5</span>))</span>
<span id="cb149-9"><a href="#cb149-9"></a> <span class="kw">title</span>(<span class="st">&#39;ridge&#39;</span>)</span>
<span id="cb149-10"><a href="#cb149-10"></a></span>
<span id="cb149-11"><a href="#cb149-11"></a> enet_errors_mtx &lt;-<span class="st">  </span><span class="kw">plot_cv_f</span>(brcaRna_cv_enet, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,.<span class="dv">5</span>))</span>
<span id="cb149-12"><a href="#cb149-12"></a> <span class="kw">title</span>(<span class="st">&#39;enet&#39;</span>)</span>
<span id="cb149-13"><a href="#cb149-13"></a></span>
<span id="cb149-14"><a href="#cb149-14"></a> <span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, <span class="st">&#39;log(Lambda)&#39;</span>)</span>
<span id="cb149-15"><a href="#cb149-15"></a> <span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, brcaRna_cv_lasso<span class="op">$</span>name)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-glmnetFit-lookFits-1.png" alt="compare fits" width="1056" />
<p class="caption">
(#fig:brcaRna-glmnetFit-lookFits)compare fits
</p>
</div>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb150-2"><a href="#cb150-2"></a></span>
<span id="cb150-3"><a href="#cb150-3"></a></span>
<span id="cb150-4"><a href="#cb150-4"></a>errors_frm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb150-5"><a href="#cb150-5"></a>  <span class="dt">lasso =</span> lasso_errors_mtx, <span class="dt">ridge =</span> rifge_errors_mtx, <span class="dt">enet =</span> enet_errors_mtx</span>
<span id="cb150-6"><a href="#cb150-6"></a>)</span>
<span id="cb150-7"><a href="#cb150-7"></a><span class="kw">colnames</span>(errors_frm) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&#39;</span><span class="ch">\\</span><span class="st">.error&#39;</span>,<span class="st">&#39;&#39;</span>, <span class="kw">colnames</span>(errors_frm))</span>
<span id="cb150-8"><a href="#cb150-8"></a></span>
<span id="cb150-9"><a href="#cb150-9"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">t</span>(errors_frm),</span>
<span id="cb150-10"><a href="#cb150-10"></a> <span class="dt">caption =</span> <span class="st">&#39;Misclassifiaction error rates&#39;</span>,</span>
<span id="cb150-11"><a href="#cb150-11"></a> <span class="dt">digits=</span><span class="dv">1</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb150-12"><a href="#cb150-12"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:brcaRna-glmnetFit-printErrors)Misclassifiaction error rates
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
p
</th>
<th style="text-align:right;">
train
</th>
<th style="text-align:right;">
test
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
lasso_1se
</td>
<td style="text-align:right;">
168
</td>
<td style="text-align:right;">
11.9
</td>
<td style="text-align:right;">
9.9
</td>
</tr>
<tr>
<td style="text-align:left;">
lasso_min
</td>
<td style="text-align:right;">
408
</td>
<td style="text-align:right;">
11.4
</td>
<td style="text-align:right;">
12.8
</td>
</tr>
<tr>
<td style="text-align:left;">
ridge_1se
</td>
<td style="text-align:right;">
19406
</td>
<td style="text-align:right;">
13.0
</td>
<td style="text-align:right;">
13.0
</td>
</tr>
<tr>
<td style="text-align:left;">
ridge_min
</td>
<td style="text-align:right;">
19406
</td>
<td style="text-align:right;">
12.6
</td>
<td style="text-align:right;">
12.4
</td>
</tr>
<tr>
<td style="text-align:left;">
enet_1se
</td>
<td style="text-align:right;">
315
</td>
<td style="text-align:right;">
11.6
</td>
<td style="text-align:right;">
11.2
</td>
</tr>
<tr>
<td style="text-align:left;">
enet_min
</td>
<td style="text-align:right;">
671
</td>
<td style="text-align:right;">
11.1
</td>
<td style="text-align:right;">
13.4
</td>
</tr>
</tbody>
</table>
<p><br/></p>
<ul>
<li><p>lasso and enet, 1se and min lambda, have comparable performance
on the training data.</p></li>
<li><p>The lasso 1se model gains performance in the test data, but
lasso min lambda and enet models stay at the same level or drop
in performance on the test set.</p></li>
<li><p>lasso 1se is most parsimonious by a factor of 2.</p></li>
</ul>
</div>
<div id="the-relaxed-lasso-and-blended-mix-models-1" class="section level2" number="7.3">
<h2 number="7.3"><span class="header-section-number">7.3</span> The relaxed lasso and blended mix models</h2>
<p>Next we look at the so-called <code>relaxed lasso</code> model, and
the <code>blended mix</code> which is an optimized shrinkage
between the relaxed lasso and the regular lasso.
See @ref(eq:blended) in Section @ref(modeling-background).</p>
<!--
The relaxed fit takes quite a bit longer.  
-->
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb151-2"><a href="#cb151-2"></a></span>
<span id="cb151-3"><a href="#cb151-3"></a><span class="kw">library</span>(glmnet)</span>
<span id="cb151-4"><a href="#cb151-4"></a></span>
<span id="cb151-5"><a href="#cb151-5"></a>brcaRna_cv_lassoR_sum &lt;-<span class="st"> </span><span class="kw">print</span>(brcaRna_cv_lassoR)</span></code></pre></div>
<pre><code>## 
## Call:  glmnet::cv.glmnet(x = brcaRna_train_geneExpr_mtx, y = brcaRna_train_group_vec,      type.measure = &quot;class&quot;, foldid = brcaRna_train_foldid_vec,      keep = T, relax = T, alpha = 1, family = &quot;binomial&quot;, nlambda = 30) 
## 
## Measure: Misclassification Error 
## 
##     Gamma   Lambda Measure       SE Nonzero
## min  1.00 0.005227  0.1139 0.006167     408
## 1se  0.75 0.021823  0.1180 0.008812      91</code></pre>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1"></a><span class="kw">plot</span>(brcaRna_cv_lassoR)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-glmnetFit-lookLassoR-1.png" alt="Relaxed lasso fit" width="576" />
<p class="caption">
(#fig:brcaRna-glmnetFit-lookLassoR)Relaxed lasso fit
</p>
</div>
<div class="sourceCode" id="cb154"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb154-1"><a href="#cb154-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb154-2"><a href="#cb154-2"></a><span class="co"># only report  1se</span></span>
<span id="cb154-3"><a href="#cb154-3"></a>ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(brcaRna_cv_lassoR<span class="op">$</span>lambda<span class="fl">.1</span>se, brcaRna_cv_lassoR<span class="op">$</span>lambda)</span>
<span id="cb154-4"><a href="#cb154-4"></a>ndx_min &lt;-<span class="st"> </span><span class="kw">match</span>(brcaRna_cv_lassoR<span class="op">$</span>lambda.min, brcaRna_cv_lassoR<span class="op">$</span>lambda)</span>
<span id="cb154-5"><a href="#cb154-5"></a></span>
<span id="cb154-6"><a href="#cb154-6"></a><span class="co"># only show 1se anyway</span></span>
<span id="cb154-7"><a href="#cb154-7"></a><span class="co"># if(ndx_1se != ndx_min) stop(&quot;lambda.1se != lambda.min&quot;)</span></span>
<span id="cb154-8"><a href="#cb154-8"></a></span>
<span id="cb154-9"><a href="#cb154-9"></a></span>
<span id="cb154-10"><a href="#cb154-10"></a><span class="co"># train oof data - NOT CLEAR WHY THESE DIFFER FROM CV ERRORS EXTRACTED FROM MODEL</span></span>
<span id="cb154-11"><a href="#cb154-11"></a><span class="co"># Get relaxed lasso (gamma=0) oof error</span></span>
<span id="cb154-12"><a href="#cb154-12"></a>train_oofPred_relaxed_1se_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb154-13"><a href="#cb154-13"></a>  <span class="kw">logistic_f</span>(brcaRna_cv_lassoR<span class="op">$</span>fit.preval[[<span class="st">&quot;g:0&quot;</span>]][, ndx_1se]) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&quot;LumA&quot;</span>, <span class="st">&quot;Other&quot;</span></span>
<span id="cb154-14"><a href="#cb154-14"></a>)</span>
<span id="cb154-15"><a href="#cb154-15"></a>train_oofPred_relaxed_1se_error &lt;-<span class="st"> </span><span class="kw">mean</span>(train_oofPred_relaxed_1se_vec <span class="op">!=</span><span class="st"> </span>brcaRna_train_group_vec)</span>
<span id="cb154-16"><a href="#cb154-16"></a></span>
<span id="cb154-17"><a href="#cb154-17"></a><span class="co"># blended mix (gamma=0.5)</span></span>
<span id="cb154-18"><a href="#cb154-18"></a>train_oofPred_blended_1se_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb154-19"><a href="#cb154-19"></a>  <span class="kw">logistic_f</span>(brcaRna_cv_lassoR<span class="op">$</span>fit.preval[[<span class="st">&quot;g:0.5&quot;</span>]][, ndx_1se]) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&quot;LumA&quot;</span>, <span class="st">&quot;Other&quot;</span></span>
<span id="cb154-20"><a href="#cb154-20"></a>)</span>
<span id="cb154-21"><a href="#cb154-21"></a>train_oofPred_blended_1se_error &lt;-<span class="st"> </span><span class="kw">mean</span>(train_oofPred_blended_1se_vec <span class="op">!=</span><span class="st"> </span>brcaRna_train_group_vec)</span>
<span id="cb154-22"><a href="#cb154-22"></a></span>
<span id="cb154-23"><a href="#cb154-23"></a></span>
<span id="cb154-24"><a href="#cb154-24"></a><span class="co"># Test set error - relaxed</span></span>
<span id="cb154-25"><a href="#cb154-25"></a>test_pred_relaxed_1se_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb154-26"><a href="#cb154-26"></a>  brcaRna_cv_lassoR,</span>
<span id="cb154-27"><a href="#cb154-27"></a>  <span class="dt">newx =</span> brcaRna_test_geneExpr_mtx,</span>
<span id="cb154-28"><a href="#cb154-28"></a>  <span class="dt">s =</span> <span class="st">&quot;lambda.1se&quot;</span>,</span>
<span id="cb154-29"><a href="#cb154-29"></a>  <span class="dt">type =</span> <span class="st">&quot;class&quot;</span>,</span>
<span id="cb154-30"><a href="#cb154-30"></a>  <span class="dt">gamma =</span> <span class="dv">0</span></span>
<span id="cb154-31"><a href="#cb154-31"></a>)</span>
<span id="cb154-32"><a href="#cb154-32"></a>test_pred_relaxed_1se_error &lt;-<span class="st"> </span><span class="kw">mean</span>(test_pred_relaxed_1se_vec <span class="op">!=</span><span class="st"> </span>brcaRna_test_group_vec)</span>
<span id="cb154-33"><a href="#cb154-33"></a></span>
<span id="cb154-34"><a href="#cb154-34"></a><span class="co"># Test set error - blended</span></span>
<span id="cb154-35"><a href="#cb154-35"></a>test_pred_blended_1se_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb154-36"><a href="#cb154-36"></a>  brcaRna_cv_lassoR,</span>
<span id="cb154-37"><a href="#cb154-37"></a>  <span class="dt">newx =</span> brcaRna_test_geneExpr_mtx,</span>
<span id="cb154-38"><a href="#cb154-38"></a>  <span class="dt">s =</span> <span class="st">&quot;lambda.1se&quot;</span>,</span>
<span id="cb154-39"><a href="#cb154-39"></a>  <span class="dt">type =</span> <span class="st">&quot;class&quot;</span>,</span>
<span id="cb154-40"><a href="#cb154-40"></a>  <span class="dt">gamma =</span> <span class="fl">0.5</span></span>
<span id="cb154-41"><a href="#cb154-41"></a>)</span>
<span id="cb154-42"><a href="#cb154-42"></a>test_pred_blended_1se_error &lt;-<span class="st"> </span><span class="kw">mean</span>(test_pred_blended_1se_vec <span class="op">!=</span><span class="st"> </span>brcaRna_test_group_vec)</span>
<span id="cb154-43"><a href="#cb154-43"></a></span>
<span id="cb154-44"><a href="#cb154-44"></a></span>
<span id="cb154-45"><a href="#cb154-45"></a>brcaRna_cv_lassoR_1se_error &lt;-<span class="st"> </span>brcaRna_cv_lassoR<span class="op">$</span>cvm[brcaRna_cv_lassoR<span class="op">$</span>lambda<span class="op">==</span>brcaRna_cv_lassoR<span class="op">$</span>lambda.min]</span>
<span id="cb154-46"><a href="#cb154-46"></a></span>
<span id="cb154-47"><a href="#cb154-47"></a>cv_blended_statlist &lt;-<span class="st"> </span>brcaRna_cv_lassoR<span class="op">$</span>relaxed<span class="op">$</span>statlist[[<span class="st">&#39;g:0.5&#39;</span>]]</span>
<span id="cb154-48"><a href="#cb154-48"></a>cv_blended_1se_error &lt;-<span class="st"> </span>cv_blended_statlist<span class="op">$</span>cvm[cv_blended_statlist<span class="op">$</span>lambda<span class="op">==</span></span>
<span id="cb154-49"><a href="#cb154-49"></a><span class="st">   </span>brcaRna_cv_lassoR<span class="op">$</span>relaxed<span class="op">$</span>lambda<span class="fl">.1</span>se]</span>
<span id="cb154-50"><a href="#cb154-50"></a> </span>
<span id="cb154-51"><a href="#cb154-51"></a></span>
<span id="cb154-52"><a href="#cb154-52"></a></span>
<span id="cb154-53"><a href="#cb154-53"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">data.frame</span>(</span>
<span id="cb154-54"><a href="#cb154-54"></a>  <span class="dt">train_relaxed  =</span> brcaRna_cv_lassoR_1se_error,</span>
<span id="cb154-55"><a href="#cb154-55"></a>  <span class="dt">train_blended  =</span> cv_blended_1se_error,</span>
<span id="cb154-56"><a href="#cb154-56"></a>  <span class="co">#train_relaxed_oof = train_oofPred_relaxed_1se_error,</span></span>
<span id="cb154-57"><a href="#cb154-57"></a>  <span class="co">#train_blended_oof = train_oofPred_blended_1se_error,</span></span>
<span id="cb154-58"><a href="#cb154-58"></a>  <span class="dt">test_relaxed  =</span> test_pred_relaxed_1se_error,</span>
<span id="cb154-59"><a href="#cb154-59"></a>  <span class="dt">test_blended  =</span> test_pred_blended_1se_error</span>
<span id="cb154-60"><a href="#cb154-60"></a>)) <span class="op">*</span><span class="st"> </span><span class="dv">100</span>,</span>
<span id="cb154-61"><a href="#cb154-61"></a><span class="dt">digits =</span> <span class="dv">1</span>,</span>
<span id="cb154-62"><a href="#cb154-62"></a><span class="dt">caption =</span> <span class="st">&quot;Relaxed lasso and blended mix error rates&quot;</span></span>
<span id="cb154-63"><a href="#cb154-63"></a>) <span class="op">%&gt;%</span></span>
<span id="cb154-64"><a href="#cb154-64"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:brcaRna-glmnetFit-lookLassoR2)Relaxed lasso and blended mix error rates
</caption>
<tbody>
<tr>
<td style="text-align:left;">
train_relaxed
</td>
<td style="text-align:right;">
11.4
</td>
</tr>
<tr>
<td style="text-align:left;">
train_blended
</td>
<td style="text-align:right;">
11.8
</td>
</tr>
<tr>
<td style="text-align:left;">
test_relaxed
</td>
<td style="text-align:right;">
11.4
</td>
</tr>
<tr>
<td style="text-align:left;">
test_blended
</td>
<td style="text-align:right;">
10.3
</td>
</tr>
</tbody>
</table>
<p><br/></p>
<p>The relaxed lasso and blended mix error rates are comparable to the
regular lasso fit error rate. We see here too that the reported cv
error rates are comparable to test set error rates.</p>
<!-- , while out-of-fold error rates
continue to be good indicators of unseen data error rates, as captured
by the test set.  
-->
<p>The <em>1se</em> lambda rule applied to the relaxed lasso fit selected a model with
<span class="math inline">\(168\)</span> features,
while for the blended mix model
(See @ref(eq:blended) in Section @ref(modeling-background))
the <em>1se</em> lambda rule selected
<span class="math inline">\(91\)</span> features (vertical
dotted reference line in Figure @ref(fig:lookLassoR)).
This feature is pointed out in the
<a href="https://cran.r-project.org/web/packages/glmnet/vignettes/relax.pdf">glmnet 3.0 vignette</a>:
<em>The debiasing will potentially improve prediction performance,
and CV will typically select a model with a smaller number of variables.</em></p>
</div>
<div id="examination-of-sensitivity-vs-specificity-1" class="section level2" number="7.4">
<h2 number="7.4"><span class="header-section-number">7.4</span> Examination of sensitivity vs specificity</h2>
<p>In the results above we reported error rates without inspecting the
sensitivity versus specificity trade-off. ROC curves can be examined
to get a sense of the trade-off.</p>
<div id="training-data-out-of-fold-roc-curves-1" class="section level3" number="7.4.1">
<h3 number="7.4.1"><span class="header-section-number">7.4.1</span> Training data out-of-fold ROC curves</h3>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="#cb155-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb155-2"><a href="#cb155-2"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb155-3"><a href="#cb155-3"></a></span>
<span id="cb155-4"><a href="#cb155-4"></a><span class="co"># train</span></span>
<span id="cb155-5"><a href="#cb155-5"></a><span class="co"># lasso</span></span>
<span id="cb155-6"><a href="#cb155-6"></a>ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(brcaRna_cv_lasso<span class="op">$</span>lambda<span class="fl">.1</span>se,brcaRna_cv_lasso<span class="op">$</span>lambda)</span>
<span id="cb155-7"><a href="#cb155-7"></a>train_lasso_oofProb_vec &lt;-<span class="st"> </span><span class="kw">logistic_f</span>(brcaRna_cv_lasso<span class="op">$</span>fit.preval[,ndx_1se])</span>
<span id="cb155-8"><a href="#cb155-8"></a>train_lasso_roc &lt;-<span class="st"> </span>pROC<span class="op">::</span><span class="kw">roc</span>(</span>
<span id="cb155-9"><a href="#cb155-9"></a> <span class="dt">response =</span> <span class="kw">as.numeric</span>(brcaRna_train_group_vec<span class="op">==</span><span class="st">&#39;LumA&#39;</span>),</span>
<span id="cb155-10"><a href="#cb155-10"></a> <span class="dt">predictor =</span> train_lasso_oofProb_vec)</span></code></pre></div>
<pre><code>## Setting levels: control = 0, case = 1</code></pre>
<pre><code>## Setting direction: controls &lt; cases</code></pre>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb158-1"><a href="#cb158-1"></a><span class="co"># enet</span></span>
<span id="cb158-2"><a href="#cb158-2"></a>ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(brcaRna_cv_enet<span class="op">$</span>lambda<span class="fl">.1</span>se,brcaRna_cv_enet<span class="op">$</span>lambda)</span>
<span id="cb158-3"><a href="#cb158-3"></a>train_enet_oofProb_vec &lt;-<span class="st"> </span><span class="kw">logistic_f</span>(brcaRna_cv_enet<span class="op">$</span>fit.preval[,ndx_1se])</span>
<span id="cb158-4"><a href="#cb158-4"></a>train_enet_roc &lt;-<span class="st"> </span>pROC<span class="op">::</span><span class="kw">roc</span>(</span>
<span id="cb158-5"><a href="#cb158-5"></a> <span class="dt">response =</span> <span class="kw">as.numeric</span>(brcaRna_train_group_vec<span class="op">==</span><span class="st">&#39;LumA&#39;</span>),</span>
<span id="cb158-6"><a href="#cb158-6"></a> <span class="dt">predictor =</span> train_enet_oofProb_vec)</span></code></pre></div>
<pre><code>## Setting levels: control = 0, case = 1
## Setting direction: controls &lt; cases</code></pre>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="#cb160-1"></a><span class="co"># lasso - relaxed</span></span>
<span id="cb160-2"><a href="#cb160-2"></a>ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(brcaRna_cv_lassoR<span class="op">$</span>lambda<span class="fl">.1</span>se,brcaRna_cv_lassoR<span class="op">$</span>lambda)</span>
<span id="cb160-3"><a href="#cb160-3"></a>train_relaxed_oofProb_vec &lt;-<span class="st"> </span><span class="kw">logistic_f</span>(brcaRna_cv_lassoR<span class="op">$</span>fit.preval[[<span class="st">&#39;g:0&#39;</span>]][,ndx_1se])</span>
<span id="cb160-4"><a href="#cb160-4"></a>train_relaxed_roc &lt;-<span class="st"> </span>pROC<span class="op">::</span><span class="kw">roc</span>(</span>
<span id="cb160-5"><a href="#cb160-5"></a> <span class="dt">response =</span> <span class="kw">as.numeric</span>(brcaRna_train_group_vec<span class="op">==</span><span class="st">&#39;LumA&#39;</span>),</span>
<span id="cb160-6"><a href="#cb160-6"></a> <span class="dt">predictor =</span> train_relaxed_oofProb_vec)</span></code></pre></div>
<pre><code>## Setting levels: control = 0, case = 1
## Setting direction: controls &lt; cases</code></pre>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="#cb162-1"></a><span class="co"># blended mix (gamma=0.5)</span></span>
<span id="cb162-2"><a href="#cb162-2"></a>ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(brcaRna_cv_lassoR<span class="op">$</span>lambda<span class="fl">.1</span>se,brcaRna_cv_lassoR<span class="op">$</span>lambda)</span>
<span id="cb162-3"><a href="#cb162-3"></a>train_blended_oofProb_vec &lt;-<span class="st"> </span><span class="kw">logistic_f</span>(brcaRna_cv_lassoR<span class="op">$</span>fit.preval[[<span class="st">&#39;g:0.5&#39;</span>]][,ndx_1se])</span>
<span id="cb162-4"><a href="#cb162-4"></a>train_blended_roc &lt;-<span class="st"> </span>pROC<span class="op">::</span><span class="kw">roc</span>(</span>
<span id="cb162-5"><a href="#cb162-5"></a> <span class="dt">response =</span> <span class="kw">as.numeric</span>(brcaRna_train_group_vec<span class="op">==</span><span class="st">&#39;LumA&#39;</span>),</span>
<span id="cb162-6"><a href="#cb162-6"></a> <span class="dt">predictor =</span> train_blended_oofProb_vec)</span></code></pre></div>
<pre><code>## Setting levels: control = 0, case = 1
## Setting direction: controls &lt; cases</code></pre>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="#cb164-1"></a><span class="kw">plot</span>(train_lasso_roc, <span class="dt">col =</span> col_vec[<span class="dv">1</span>])</span>
<span id="cb164-2"><a href="#cb164-2"></a><span class="kw">lines</span>(train_enet_roc, <span class="dt">col =</span> col_vec[<span class="dv">2</span>])</span>
<span id="cb164-3"><a href="#cb164-3"></a><span class="kw">lines</span>(train_relaxed_roc, <span class="dt">col =</span> col_vec[<span class="dv">3</span>])</span>
<span id="cb164-4"><a href="#cb164-4"></a><span class="kw">lines</span>(train_blended_roc, <span class="dt">col =</span> col_vec[<span class="dv">4</span>])</span>
<span id="cb164-5"><a href="#cb164-5"></a></span>
<span id="cb164-6"><a href="#cb164-6"></a><span class="kw">legend</span>(<span class="st">&#39;bottomright&#39;</span>, <span class="dt">title=</span><span class="st">&#39;AUC&#39;</span>,</span>
<span id="cb164-7"><a href="#cb164-7"></a> <span class="dt">legend=</span><span class="kw">c</span>(</span>
<span id="cb164-8"><a href="#cb164-8"></a>  <span class="kw">paste</span>(<span class="st">&#39;lasso =&#39;</span>, <span class="kw">round</span>(train_lasso_roc[[<span class="st">&#39;auc&#39;</span>]],<span class="dv">3</span>)),</span>
<span id="cb164-9"><a href="#cb164-9"></a>  <span class="kw">paste</span>(<span class="st">&#39;enet =&#39;</span>, <span class="kw">round</span>(train_enet_roc[[<span class="st">&#39;auc&#39;</span>]],<span class="dv">3</span>)),</span>
<span id="cb164-10"><a href="#cb164-10"></a>  <span class="kw">paste</span>(<span class="st">&#39;relaxed =&#39;</span>, <span class="kw">round</span>(train_relaxed_roc[[<span class="st">&#39;auc&#39;</span>]],<span class="dv">3</span>)),</span>
<span id="cb164-11"><a href="#cb164-11"></a>  <span class="kw">paste</span>(<span class="st">&#39;blended =&#39;</span>, <span class="kw">round</span>(train_blended_roc[[<span class="st">&#39;auc&#39;</span>]],<span class="dv">3</span>))</span>
<span id="cb164-12"><a href="#cb164-12"></a> ),</span>
<span id="cb164-13"><a href="#cb164-13"></a> <span class="dt">text.col =</span> col_vec[<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>],</span>
<span id="cb164-14"><a href="#cb164-14"></a> <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb164-15"><a href="#cb164-15"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-glmnetFit-trainROC-1.png" alt="Train data out-of-sample ROCs" width="480" />
<p class="caption">
(#fig:brcaRna-glmnetFit-trainROC)Train data out-of-sample ROCs
</p>
</div>
<p>Compare thresholds for 90% Specificity:</p>
<div class="sourceCode" id="cb165"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb165-1"><a href="#cb165-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb165-2"><a href="#cb165-2"></a></span>
<span id="cb165-3"><a href="#cb165-3"></a> lasso_ndx &lt;-<span class="st"> </span><span class="kw">with</span>(<span class="kw">as.data.frame</span>(pROC<span class="op">::</span><span class="kw">coords</span>(train_lasso_roc, <span class="dt">transpose=</span>F)), </span>
<span id="cb165-4"><a href="#cb165-4"></a>   <span class="kw">min</span>(<span class="kw">which</span>(specificity <span class="op">&gt;=</span><span class="st"> </span><span class="fl">0.9</span>)))</span>
<span id="cb165-5"><a href="#cb165-5"></a></span>
<span id="cb165-6"><a href="#cb165-6"></a> enet_ndx &lt;-<span class="st"> </span><span class="kw">with</span>(<span class="kw">as.data.frame</span>(pROC<span class="op">::</span><span class="kw">coords</span>(train_enet_roc, <span class="dt">transpose=</span>F)), </span>
<span id="cb165-7"><a href="#cb165-7"></a>   <span class="kw">min</span>(<span class="kw">which</span>(specificity <span class="op">&gt;=</span><span class="st"> </span><span class="fl">0.9</span>)))</span>
<span id="cb165-8"><a href="#cb165-8"></a></span>
<span id="cb165-9"><a href="#cb165-9"></a> lassoR_ndx &lt;-<span class="st"> </span><span class="kw">with</span>(<span class="kw">as.data.frame</span>(pROC<span class="op">::</span><span class="kw">coords</span>(train_relaxed_roc, <span class="dt">transpose=</span>F)), </span>
<span id="cb165-10"><a href="#cb165-10"></a>   <span class="kw">min</span>(<span class="kw">which</span>(specificity <span class="op">&gt;=</span><span class="st"> </span><span class="fl">0.9</span>)))</span>
<span id="cb165-11"><a href="#cb165-11"></a></span>
<span id="cb165-12"><a href="#cb165-12"></a> blended_ndx &lt;-<span class="st"> </span><span class="kw">with</span>(<span class="kw">as.data.frame</span>(pROC<span class="op">::</span><span class="kw">coords</span>(train_blended_roc, <span class="dt">transpose=</span>F)), </span>
<span id="cb165-13"><a href="#cb165-13"></a>   <span class="kw">min</span>(<span class="kw">which</span>(specificity <span class="op">&gt;=</span><span class="st"> </span><span class="fl">0.9</span>)))</span>
<span id="cb165-14"><a href="#cb165-14"></a></span>
<span id="cb165-15"><a href="#cb165-15"></a>  spec90_frm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">rbind</span>(</span>
<span id="cb165-16"><a href="#cb165-16"></a>  <span class="dt">lasso=</span><span class="kw">as.data.frame</span>(pROC<span class="op">::</span><span class="kw">coords</span>(train_lasso_roc, <span class="dt">transpose=</span>F))[lasso_ndx,],</span>
<span id="cb165-17"><a href="#cb165-17"></a>  <span class="dt">enet=</span><span class="kw">as.data.frame</span>(pROC<span class="op">::</span><span class="kw">coords</span>(train_enet_roc, <span class="dt">transpose=</span>F))[enet_ndx,],</span>
<span id="cb165-18"><a href="#cb165-18"></a>  <span class="dt">relaxed=</span><span class="kw">as.data.frame</span>(pROC<span class="op">::</span><span class="kw">coords</span>(train_relaxed_roc, <span class="dt">transpose=</span>F))[lassoR_ndx,],</span>
<span id="cb165-19"><a href="#cb165-19"></a>  <span class="dt">blended=</span><span class="kw">as.data.frame</span>(pROC<span class="op">::</span><span class="kw">coords</span>(train_blended_roc, <span class="dt">transpose=</span>F))[blended_ndx,]</span>
<span id="cb165-20"><a href="#cb165-20"></a> ))</span>
<span id="cb165-21"><a href="#cb165-21"></a></span>
<span id="cb165-22"><a href="#cb165-22"></a></span>
<span id="cb165-23"><a href="#cb165-23"></a>knitr<span class="op">::</span><span class="kw">kable</span>(spec90_frm,</span>
<span id="cb165-24"><a href="#cb165-24"></a>  <span class="dt">digits=</span><span class="dv">3</span>,</span>
<span id="cb165-25"><a href="#cb165-25"></a>  <span class="dt">caption=</span><span class="st">&quot;Specificity = .90 Coordinates&quot;</span></span>
<span id="cb165-26"><a href="#cb165-26"></a>) <span class="op">%&gt;%</span></span>
<span id="cb165-27"><a href="#cb165-27"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:brcaRna-glmnetFit-thresh90)Specificity = .90 Coordinates
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
threshold
</th>
<th style="text-align:right;">
specificity
</th>
<th style="text-align:right;">
sensitivity
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
lasso
</td>
<td style="text-align:right;">
0.672
</td>
<td style="text-align:right;">
0.901
</td>
<td style="text-align:right;">
0.850
</td>
</tr>
<tr>
<td style="text-align:left;">
enet
</td>
<td style="text-align:right;">
0.661
</td>
<td style="text-align:right;">
0.901
</td>
<td style="text-align:right;">
0.876
</td>
</tr>
<tr>
<td style="text-align:left;">
relaxed
</td>
<td style="text-align:right;">
1.000
</td>
<td style="text-align:right;">
0.901
</td>
<td style="text-align:right;">
0.625
</td>
</tr>
<tr>
<td style="text-align:left;">
blended
</td>
<td style="text-align:right;">
0.999
</td>
<td style="text-align:right;">
0.901
</td>
<td style="text-align:right;">
0.659
</td>
</tr>
</tbody>
</table>
<p>This is strange.</p>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="#cb166-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb166-2"><a href="#cb166-2"></a></span>
<span id="cb166-3"><a href="#cb166-3"></a><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>), <span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">3</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb166-4"><a href="#cb166-4"></a></span>
<span id="cb166-5"><a href="#cb166-5"></a><span class="co"># lasso</span></span>
<span id="cb166-6"><a href="#cb166-6"></a><span class="kw">plot</span>(<span class="kw">density</span>(train_lasso_oofProb_vec[brcaRna_train_group_vec <span class="op">==</span><span class="st"> &quot;Other&quot;</span>]),</span>
<span id="cb166-7"><a href="#cb166-7"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;green&quot;</span></span>
<span id="cb166-8"><a href="#cb166-8"></a>)</span>
<span id="cb166-9"><a href="#cb166-9"></a><span class="kw">lines</span>(<span class="kw">density</span>(train_lasso_oofProb_vec[brcaRna_train_group_vec <span class="op">==</span><span class="st"> &quot;LumA&quot;</span>]),</span>
<span id="cb166-10"><a href="#cb166-10"></a>  <span class="dt">col =</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb166-11"><a href="#cb166-11"></a>)</span>
<span id="cb166-12"><a href="#cb166-12"></a><span class="kw">title</span>(<span class="st">&quot;lasso&quot;</span>)</span>
<span id="cb166-13"><a href="#cb166-13"></a><span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&quot;Other&quot;</span>, <span class="st">&quot;LumA&quot;</span>), <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;red&quot;</span>))</span>
<span id="cb166-14"><a href="#cb166-14"></a></span>
<span id="cb166-15"><a href="#cb166-15"></a><span class="co"># enet</span></span>
<span id="cb166-16"><a href="#cb166-16"></a><span class="kw">plot</span>(<span class="kw">density</span>(train_enet_oofProb_vec[brcaRna_train_group_vec <span class="op">==</span><span class="st"> &quot;Other&quot;</span>]),</span>
<span id="cb166-17"><a href="#cb166-17"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;green&quot;</span></span>
<span id="cb166-18"><a href="#cb166-18"></a>)</span>
<span id="cb166-19"><a href="#cb166-19"></a><span class="kw">lines</span>(<span class="kw">density</span>(train_enet_oofProb_vec[brcaRna_train_group_vec <span class="op">==</span><span class="st"> &quot;LumA&quot;</span>]),</span>
<span id="cb166-20"><a href="#cb166-20"></a>  <span class="dt">col =</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb166-21"><a href="#cb166-21"></a>)</span>
<span id="cb166-22"><a href="#cb166-22"></a><span class="kw">title</span>(<span class="st">&quot;enet&quot;</span>)</span>
<span id="cb166-23"><a href="#cb166-23"></a></span>
<span id="cb166-24"><a href="#cb166-24"></a><span class="co"># lassoR</span></span>
<span id="cb166-25"><a href="#cb166-25"></a><span class="kw">plot</span>(<span class="kw">density</span>(train_relaxed_oofProb_vec[brcaRna_train_group_vec <span class="op">==</span><span class="st"> &quot;Other&quot;</span>]),</span>
<span id="cb166-26"><a href="#cb166-26"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;green&quot;</span></span>
<span id="cb166-27"><a href="#cb166-27"></a>)</span>
<span id="cb166-28"><a href="#cb166-28"></a><span class="kw">lines</span>(<span class="kw">density</span>(train_relaxed_oofProb_vec[brcaRna_train_group_vec <span class="op">==</span><span class="st"> &quot;LumA&quot;</span>]),</span>
<span id="cb166-29"><a href="#cb166-29"></a>  <span class="dt">col =</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb166-30"><a href="#cb166-30"></a>)</span>
<span id="cb166-31"><a href="#cb166-31"></a><span class="kw">title</span>(<span class="st">&quot;lassoR&quot;</span>)</span>
<span id="cb166-32"><a href="#cb166-32"></a></span>
<span id="cb166-33"><a href="#cb166-33"></a><span class="co"># blended</span></span>
<span id="cb166-34"><a href="#cb166-34"></a><span class="kw">plot</span>(<span class="kw">density</span>(train_blended_oofProb_vec[brcaRna_train_group_vec <span class="op">==</span><span class="st"> &quot;Other&quot;</span>]),</span>
<span id="cb166-35"><a href="#cb166-35"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;green&quot;</span></span>
<span id="cb166-36"><a href="#cb166-36"></a>)</span>
<span id="cb166-37"><a href="#cb166-37"></a><span class="kw">lines</span>(<span class="kw">density</span>(train_blended_oofProb_vec[brcaRna_train_group_vec <span class="op">==</span><span class="st"> &quot;LumA&quot;</span>]),</span>
<span id="cb166-38"><a href="#cb166-38"></a>  <span class="dt">col =</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb166-39"><a href="#cb166-39"></a>)</span>
<span id="cb166-40"><a href="#cb166-40"></a><span class="kw">title</span>(<span class="st">&quot;blended&quot;</span>)</span>
<span id="cb166-41"><a href="#cb166-41"></a></span>
<span id="cb166-42"><a href="#cb166-42"></a><span class="kw">mtext</span>(<span class="dt">side =</span> <span class="dv">1</span>, <span class="dt">outer =</span> T, <span class="st">&quot;out-of-fold predicted probability&quot;</span>, <span class="dt">cex =</span> <span class="fl">1.25</span>)</span>
<span id="cb166-43"><a href="#cb166-43"></a><span class="kw">mtext</span>(<span class="dt">side =</span> <span class="dv">2</span>, <span class="dt">outer =</span> T, <span class="st">&quot;density&quot;</span>, <span class="dt">cex =</span> <span class="fl">1.25</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-glmnetFit-trainOOFprops-1.png" alt="Train data out-of-fold predicted probabilities" width="960" />
<p class="caption">
(#fig:brcaRna-glmnetFit-trainOOFprops)Train data out-of-fold predicted probabilities
</p>
</div>
<p>The relaxed lasso fit results in essentially dichotomized predicted probability
distribution - predicted probabilities are very close to 0 or 1.</p>
<p>Look at test data ROC curves.</p>
<div class="sourceCode" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="#cb167-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb167-2"><a href="#cb167-2"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb167-3"><a href="#cb167-3"></a><span class="co"># plot all</span></span>
<span id="cb167-4"><a href="#cb167-4"></a><span class="kw">plot</span>(test_lasso_roc, <span class="dt">col =</span> col_vec[<span class="dv">1</span>])</span>
<span id="cb167-5"><a href="#cb167-5"></a><span class="kw">lines</span>(test_enet_roc, <span class="dt">col =</span> col_vec[<span class="dv">2</span>])</span>
<span id="cb167-6"><a href="#cb167-6"></a><span class="kw">lines</span>(test_relaxed_roc, <span class="dt">col =</span> col_vec[<span class="dv">3</span>])</span>
<span id="cb167-7"><a href="#cb167-7"></a><span class="kw">lines</span>(test_blended_roc, <span class="dt">col =</span> col_vec[<span class="dv">4</span>])</span>
<span id="cb167-8"><a href="#cb167-8"></a></span>
<span id="cb167-9"><a href="#cb167-9"></a><span class="kw">legend</span>(<span class="st">&quot;bottomright&quot;</span>,</span>
<span id="cb167-10"><a href="#cb167-10"></a>  <span class="dt">title =</span> <span class="st">&quot;AUC&quot;</span>,</span>
<span id="cb167-11"><a href="#cb167-11"></a>  <span class="dt">legend =</span> <span class="kw">c</span>(</span>
<span id="cb167-12"><a href="#cb167-12"></a>    <span class="kw">paste</span>(<span class="st">&quot;lasso =&quot;</span>, <span class="kw">round</span>(test_lasso_roc[[<span class="st">&quot;auc&quot;</span>]], <span class="dv">3</span>)),</span>
<span id="cb167-13"><a href="#cb167-13"></a>    <span class="kw">paste</span>(<span class="st">&quot;enet =&quot;</span>, <span class="kw">round</span>(test_enet_roc[[<span class="st">&quot;auc&quot;</span>]], <span class="dv">3</span>)),</span>
<span id="cb167-14"><a href="#cb167-14"></a>    <span class="kw">paste</span>(<span class="st">&quot;relaxed =&quot;</span>, <span class="kw">round</span>(test_relaxed_roc[[<span class="st">&quot;auc&quot;</span>]], <span class="dv">3</span>)),</span>
<span id="cb167-15"><a href="#cb167-15"></a>    <span class="kw">paste</span>(<span class="st">&quot;blended =&quot;</span>, <span class="kw">round</span>(test_blended_roc[[<span class="st">&quot;auc&quot;</span>]], <span class="dv">3</span>))</span>
<span id="cb167-16"><a href="#cb167-16"></a>  ),</span>
<span id="cb167-17"><a href="#cb167-17"></a>  <span class="dt">text.col =</span> col_vec[<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>],</span>
<span id="cb167-18"><a href="#cb167-18"></a>  <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb167-19"><a href="#cb167-19"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-glmnetFit-testROC2-1.png" alt="Test data out-of-sample ROCs" width="480" />
<p class="caption">
(#fig:brcaRna-glmnetFit-testROC2)Test data out-of-sample ROCs
</p>
</div>
<p>Look at densities of predicted probabilities.</p>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="#cb168-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb168-2"><a href="#cb168-2"></a></span>
<span id="cb168-3"><a href="#cb168-3"></a><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>), <span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">3</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb168-4"><a href="#cb168-4"></a></span>
<span id="cb168-5"><a href="#cb168-5"></a><span class="co"># lasso</span></span>
<span id="cb168-6"><a href="#cb168-6"></a><span class="kw">plot</span>(<span class="kw">density</span>(test_lasso_predProb_vec[brcaRna_test_group_vec <span class="op">==</span><span class="st"> &quot;Other&quot;</span>]),</span>
<span id="cb168-7"><a href="#cb168-7"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;green&quot;</span></span>
<span id="cb168-8"><a href="#cb168-8"></a>)</span>
<span id="cb168-9"><a href="#cb168-9"></a><span class="kw">lines</span>(<span class="kw">density</span>(test_lasso_predProb_vec[brcaRna_test_group_vec <span class="op">==</span><span class="st"> &quot;LumA&quot;</span>]),</span>
<span id="cb168-10"><a href="#cb168-10"></a>  <span class="dt">col =</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb168-11"><a href="#cb168-11"></a>)</span>
<span id="cb168-12"><a href="#cb168-12"></a><span class="kw">title</span>(<span class="st">&quot;lasso&quot;</span>)</span>
<span id="cb168-13"><a href="#cb168-13"></a><span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&quot;Other&quot;</span>, <span class="st">&quot;LumA&quot;</span>), <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;red&quot;</span>))</span>
<span id="cb168-14"><a href="#cb168-14"></a></span>
<span id="cb168-15"><a href="#cb168-15"></a><span class="co"># enet</span></span>
<span id="cb168-16"><a href="#cb168-16"></a><span class="kw">plot</span>(<span class="kw">density</span>(test_enet_predProb_vec[brcaRna_test_group_vec <span class="op">==</span><span class="st"> &quot;Other&quot;</span>]),</span>
<span id="cb168-17"><a href="#cb168-17"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;green&quot;</span></span>
<span id="cb168-18"><a href="#cb168-18"></a>)</span>
<span id="cb168-19"><a href="#cb168-19"></a><span class="kw">lines</span>(<span class="kw">density</span>(test_enet_predProb_vec[brcaRna_test_group_vec <span class="op">==</span><span class="st"> &quot;LumA&quot;</span>]),</span>
<span id="cb168-20"><a href="#cb168-20"></a>  <span class="dt">col =</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb168-21"><a href="#cb168-21"></a>)</span>
<span id="cb168-22"><a href="#cb168-22"></a><span class="kw">title</span>(<span class="st">&quot;enet&quot;</span>)</span>
<span id="cb168-23"><a href="#cb168-23"></a></span>
<span id="cb168-24"><a href="#cb168-24"></a><span class="co"># relaxed</span></span>
<span id="cb168-25"><a href="#cb168-25"></a><span class="kw">plot</span>(<span class="kw">density</span>(test_relaxed_predProb_vec[brcaRna_test_group_vec <span class="op">==</span><span class="st"> &quot;Other&quot;</span>]),</span>
<span id="cb168-26"><a href="#cb168-26"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;green&quot;</span></span>
<span id="cb168-27"><a href="#cb168-27"></a>)</span>
<span id="cb168-28"><a href="#cb168-28"></a><span class="kw">lines</span>(<span class="kw">density</span>(test_relaxed_predProb_vec[brcaRna_test_group_vec <span class="op">==</span><span class="st"> &quot;LumA&quot;</span>]),</span>
<span id="cb168-29"><a href="#cb168-29"></a>  <span class="dt">col =</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb168-30"><a href="#cb168-30"></a>)</span>
<span id="cb168-31"><a href="#cb168-31"></a><span class="kw">title</span>(<span class="st">&quot;relaxed&quot;</span>)</span>
<span id="cb168-32"><a href="#cb168-32"></a></span>
<span id="cb168-33"><a href="#cb168-33"></a><span class="co">#sapply(split(test_relaxed_predProb_vec, brcaRna_test_group_vec), summary)</span></span>
<span id="cb168-34"><a href="#cb168-34"></a></span>
<span id="cb168-35"><a href="#cb168-35"></a></span>
<span id="cb168-36"><a href="#cb168-36"></a><span class="co"># blended</span></span>
<span id="cb168-37"><a href="#cb168-37"></a><span class="kw">plot</span>(<span class="kw">density</span>(test_blended_predProb_vec[brcaRna_test_group_vec <span class="op">==</span><span class="st"> &quot;Other&quot;</span>]),</span>
<span id="cb168-38"><a href="#cb168-38"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;green&quot;</span></span>
<span id="cb168-39"><a href="#cb168-39"></a>)</span>
<span id="cb168-40"><a href="#cb168-40"></a><span class="kw">lines</span>(<span class="kw">density</span>(test_blended_predProb_vec[brcaRna_test_group_vec <span class="op">==</span><span class="st"> &quot;LumA&quot;</span>]),</span>
<span id="cb168-41"><a href="#cb168-41"></a>  <span class="dt">col =</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb168-42"><a href="#cb168-42"></a>)</span>
<span id="cb168-43"><a href="#cb168-43"></a><span class="kw">title</span>(<span class="st">&quot;blended&quot;</span>)</span>
<span id="cb168-44"><a href="#cb168-44"></a></span>
<span id="cb168-45"><a href="#cb168-45"></a><span class="kw">mtext</span>(<span class="dt">side =</span> <span class="dv">1</span>, <span class="dt">outer =</span> T, <span class="st">&quot;test set predicted probability&quot;</span>, <span class="dt">cex =</span> <span class="fl">1.25</span>)</span>
<span id="cb168-46"><a href="#cb168-46"></a><span class="kw">mtext</span>(<span class="dt">side =</span> <span class="dv">2</span>, <span class="dt">outer =</span> T, <span class="st">&quot;density&quot;</span>, <span class="dt">cex =</span> <span class="fl">1.25</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-glmnetFit-testOOFprobs-1.png" alt="Test data out-of-fold predicted probabilities" width="960" />
<p class="caption">
(#fig:brcaRna-glmnetFit-testOOFprobs)Test data out-of-fold predicted probabilities
</p>
</div>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb169-1"><a href="#cb169-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb169-2"><a href="#cb169-2"></a></span>
<span id="cb169-3"><a href="#cb169-3"></a><span class="co"># Define plotting function</span></span>
<span id="cb169-4"><a href="#cb169-4"></a>bxpPredProb_f &lt;-<span class="st"> </span><span class="cf">function</span>(cv_fit, <span class="dt">Gamma=</span><span class="ot">NULL</span>) {</span>
<span id="cb169-5"><a href="#cb169-5"></a>  <span class="co"># Train - preval is out-of-fold linear predictor for training design points</span></span>
<span id="cb169-6"><a href="#cb169-6"></a>  onese_ndx &lt;-<span class="st"> </span><span class="kw">match</span>(cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se, cv_fit<span class="op">$</span>lambda)</span>
<span id="cb169-7"><a href="#cb169-7"></a>  <span class="cf">if</span>(<span class="kw">is.null</span>(Gamma)) </span>
<span id="cb169-8"><a href="#cb169-8"></a>   train_1se_preval_vec &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>fit.preval[, onese_ndx] <span class="cf">else</span></span>
<span id="cb169-9"><a href="#cb169-9"></a>   train_1se_preval_vec &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>fit.preval[[Gamma]][, onese_ndx] </span>
<span id="cb169-10"><a href="#cb169-10"></a></span>
<span id="cb169-11"><a href="#cb169-11"></a>  train_1se_predProb_vec &lt;-<span class="st"> </span><span class="kw">logistic_f</span>(train_1se_preval_vec)</span>
<span id="cb169-12"><a href="#cb169-12"></a></span>
<span id="cb169-13"><a href="#cb169-13"></a>  <span class="co"># Test</span></span>
<span id="cb169-14"><a href="#cb169-14"></a>  test_1se_predProb_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb169-15"><a href="#cb169-15"></a>    cv_fit,</span>
<span id="cb169-16"><a href="#cb169-16"></a>    <span class="dt">newx =</span> brcaRna_test_geneExpr_mtx,</span>
<span id="cb169-17"><a href="#cb169-17"></a>    <span class="dt">s =</span> <span class="st">&quot;lambda.1se&quot;</span>,</span>
<span id="cb169-18"><a href="#cb169-18"></a>    <span class="dt">type =</span> <span class="st">&quot;resp&quot;</span></span>
<span id="cb169-19"><a href="#cb169-19"></a>  )</span>
<span id="cb169-20"><a href="#cb169-20"></a></span>
<span id="cb169-21"><a href="#cb169-21"></a>  tmp &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb169-22"><a href="#cb169-22"></a>    <span class="dt">train =</span> <span class="kw">split</span>(train_1se_predProb_vec, brcaRna_train_group_vec),</span>
<span id="cb169-23"><a href="#cb169-23"></a>    <span class="dt">test =</span> <span class="kw">split</span>(test_1se_predProb_vec, brcaRna_test_group_vec)</span>
<span id="cb169-24"><a href="#cb169-24"></a>  )</span>
<span id="cb169-25"><a href="#cb169-25"></a>  <span class="kw">names</span>(tmp) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>, <span class="kw">sub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">.&quot;</span>, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>, <span class="kw">names</span>(tmp)))</span>
<span id="cb169-26"><a href="#cb169-26"></a></span>
<span id="cb169-27"><a href="#cb169-27"></a>  <span class="kw">boxplot</span>(tmp)</span>
<span id="cb169-28"><a href="#cb169-28"></a>}</span>
<span id="cb169-29"><a href="#cb169-29"></a></span>
<span id="cb169-30"><a href="#cb169-30"></a><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>), <span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">5</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb169-31"><a href="#cb169-31"></a></span>
<span id="cb169-32"><a href="#cb169-32"></a><span class="kw">bxpPredProb_f</span>(brcaRna_cv_lasso)</span>
<span id="cb169-33"><a href="#cb169-33"></a><span class="kw">title</span>(<span class="st">&#39;lasso&#39;</span>)</span>
<span id="cb169-34"><a href="#cb169-34"></a></span>
<span id="cb169-35"><a href="#cb169-35"></a><span class="kw">bxpPredProb_f</span>(brcaRna_cv_enet)</span>
<span id="cb169-36"><a href="#cb169-36"></a><span class="kw">title</span>(<span class="st">&#39;enet&#39;</span>)</span>
<span id="cb169-37"><a href="#cb169-37"></a></span>
<span id="cb169-38"><a href="#cb169-38"></a><span class="kw">bxpPredProb_f</span>(brcaRna_cv_lassoR, <span class="dt">Gamma=</span><span class="st">&#39;g:0&#39;</span>)</span>
<span id="cb169-39"><a href="#cb169-39"></a><span class="kw">title</span>(<span class="st">&#39;relaxed&#39;</span>)</span>
<span id="cb169-40"><a href="#cb169-40"></a></span>
<span id="cb169-41"><a href="#cb169-41"></a><span class="kw">bxpPredProb_f</span>(brcaRna_cv_lassoR, <span class="dt">Gamma=</span><span class="st">&#39;g:0.5&#39;</span>)</span>
<span id="cb169-42"><a href="#cb169-42"></a><span class="kw">title</span>(<span class="st">&#39;blended&#39;</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-glmnetFit-fitPrevalByGroup-1.png" alt="Predicted Probabilities - Train and Test" width="768" />
<p class="caption">
(#fig:brcaRna-glmnetFit-fitPrevalByGroup)Predicted Probabilities - Train and Test
</p>
</div>
<!--
Another look - plot train and test set logistic curves with annotation.

The following shows that predicted classes come from fitted
probabilities - not out of sample probabilities.

Also shows that threshold is at 0.5 

SKIP
-->
<!-- SKIP ALL THIS 
We have seen above that assessments of model performance based on the out-of-fold 
predicted values are close to the test set assessments, and that
assessments based on prediction extracted from glmnet object are optimistic.
Here we look at confusion matrices to see how this affects the
classification results.

Here we use a threshold of 0.5 to dichotomize the predicted
probabilities into a class prediction, as is done in the
glmnet predictions.

-->
<!-- SKIPPED
The out-of-fold error rates are larger for the relaxed lasso and blended fit models.
On the test set, errors are slightly higher for the elastic net model.
-->
</div>
</div>
<div id="compare-predictions-at-misclassified-samples-1" class="section level2" number="7.5">
<h2 number="7.5"><span class="header-section-number">7.5</span> Compare predictions at misclassified samples</h2>
<p>It is useful to examine classification errors more carefully.
If models have different failure modes, one might get improved
performance by combining model predictions. Note that the models
considered here are not expected to compliment each other usefully
as they are too similar in nature.</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="#cb170-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb170-2"><a href="#cb170-2"></a></span>
<span id="cb170-3"><a href="#cb170-3"></a><span class="co"># </span><span class="al">NOTE</span><span class="co">: here we use computred oofClass rather than predClass </span></span>
<span id="cb170-4"><a href="#cb170-4"></a><span class="co"># as predClass extracted from predict() are fitted values.</span></span>
<span id="cb170-5"><a href="#cb170-5"></a></span>
<span id="cb170-6"><a href="#cb170-6"></a><span class="co"># train - oof</span></span>
<span id="cb170-7"><a href="#cb170-7"></a>ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(brcaRna_cv_lasso<span class="op">$</span>lambda<span class="fl">.1</span>se,brcaRna_cv_lasso<span class="op">$</span>lambda)</span>
<span id="cb170-8"><a href="#cb170-8"></a>train_lasso_oofProb_vec &lt;-<span class="st"> </span><span class="kw">logistic_f</span>(brcaRna_cv_lasso<span class="op">$</span>fit.preval[,ndx_1se])</span>
<span id="cb170-9"><a href="#cb170-9"></a>train_lasso_oofClass_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb170-10"><a href="#cb170-10"></a>   train_lasso_oofProb_vec <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&#39;LumA&#39;</span>, <span class="st">&#39;_Other&#39;</span>)</span>
<span id="cb170-11"><a href="#cb170-11"></a></span>
<span id="cb170-12"><a href="#cb170-12"></a>ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(brcaRna_cv_enet<span class="op">$</span>lambda<span class="fl">.1</span>se,brcaRna_cv_enet<span class="op">$</span>lambda)</span>
<span id="cb170-13"><a href="#cb170-13"></a>train_enet_oofProb_vec &lt;-<span class="st"> </span><span class="kw">logistic_f</span>(brcaRna_cv_enet<span class="op">$</span>fit.preval[,ndx_1se])</span>
<span id="cb170-14"><a href="#cb170-14"></a>train_enet_oofClass_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb170-15"><a href="#cb170-15"></a>   train_enet_oofProb_vec <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&#39;LumA&#39;</span>, <span class="st">&#39;_Other&#39;</span>)</span>
<span id="cb170-16"><a href="#cb170-16"></a></span>
<span id="cb170-17"><a href="#cb170-17"></a><span class="co"># RECALL: brcaRna_cv_lassoR$nzero[brcaRna_cv_lassoR$lambda==brcaRna_cv_lassoR$lambda.1se]</span></span>
<span id="cb170-18"><a href="#cb170-18"></a><span class="co"># train - oof</span></span>
<span id="cb170-19"><a href="#cb170-19"></a>ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(brcaRna_cv_lassoR<span class="op">$</span>lambda<span class="fl">.1</span>se,brcaRna_cv_lassoR<span class="op">$</span>lambda)</span>
<span id="cb170-20"><a href="#cb170-20"></a>train_relaxed_oofProb_vec &lt;-<span class="st"> </span><span class="kw">logistic_f</span>(brcaRna_cv_lassoR<span class="op">$</span>fit.preval[[<span class="st">&#39;g:0&#39;</span>]][,ndx_1se])</span>
<span id="cb170-21"><a href="#cb170-21"></a>train_relaxed_oofClass_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb170-22"><a href="#cb170-22"></a>   train_relaxed_oofProb_vec <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&#39;LumA&#39;</span>, <span class="st">&#39;_Other&#39;</span>)</span>
<span id="cb170-23"><a href="#cb170-23"></a></span>
<span id="cb170-24"><a href="#cb170-24"></a><span class="co"># RECALL $`r brcaRna_cv_lassoR$relaxed$nzero.1se`$ features (vertical</span></span>
<span id="cb170-25"><a href="#cb170-25"></a><span class="co">#  cv_blended_statlist &lt;- brcaRna_cv_lassoR$relaxed$statlist[[&#39;g:0.5&#39;]]</span></span>
<span id="cb170-26"><a href="#cb170-26"></a><span class="co">#  cv_blended_1se_error &lt;- cv_blended_statlist$cvm[cv_blended_statlist$lambda==</span></span>
<span id="cb170-27"><a href="#cb170-27"></a>      <span class="co">#brcaRna_cv_lassoR$relaxed$lambda.1se]</span></span>
<span id="cb170-28"><a href="#cb170-28"></a></span>
<span id="cb170-29"><a href="#cb170-29"></a><span class="co"># train - oof</span></span>
<span id="cb170-30"><a href="#cb170-30"></a>cv_blended_statlist &lt;-<span class="st"> </span>brcaRna_cv_lassoR<span class="op">$</span>relaxed<span class="op">$</span>statlist[[<span class="st">&#39;g:0.5&#39;</span>]]</span>
<span id="cb170-31"><a href="#cb170-31"></a>ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(brcaRna_cv_lassoR<span class="op">$</span>relaxed<span class="op">$</span>lambda<span class="fl">.1</span>se, cv_blended_statlist<span class="op">$</span>lambda)</span>
<span id="cb170-32"><a href="#cb170-32"></a>train_blended_oofProb_vec &lt;-<span class="st"> </span><span class="kw">logistic_f</span>(brcaRna_cv_lassoR<span class="op">$</span>fit.preval[[<span class="st">&#39;g:0.5&#39;</span>]][,ndx_1se])</span>
<span id="cb170-33"><a href="#cb170-33"></a>train_blended_oofClass_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb170-34"><a href="#cb170-34"></a>   train_blended_oofProb_vec <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&#39;LumA&#39;</span>, <span class="st">&#39;_Other&#39;</span>)</span>
<span id="cb170-35"><a href="#cb170-35"></a></span>
<span id="cb170-36"><a href="#cb170-36"></a></span>
<span id="cb170-37"><a href="#cb170-37"></a>misclass_id_vec &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">c</span>(</span>
<span id="cb170-38"><a href="#cb170-38"></a> <span class="kw">names</span>(train_lasso_oofClass_vec)[train_lasso_oofClass_vec <span class="op">!=</span><span class="st"> </span>brcaRna_train_group_vec],</span>
<span id="cb170-39"><a href="#cb170-39"></a> <span class="kw">names</span>(train_enet_oofClass_vec)[train_enet_oofClass_vec <span class="op">!=</span><span class="st"> </span>brcaRna_train_group_vec],</span>
<span id="cb170-40"><a href="#cb170-40"></a> <span class="kw">names</span>(train_relaxed_oofClass_vec)[train_relaxed_oofClass_vec <span class="op">!=</span><span class="st"> </span>brcaRna_train_group_vec],</span>
<span id="cb170-41"><a href="#cb170-41"></a> <span class="kw">names</span>(train_blended_oofClass_vec)[train_blended_oofClass_vec <span class="op">!=</span><span class="st"> </span>brcaRna_train_group_vec]</span>
<span id="cb170-42"><a href="#cb170-42"></a> )</span>
<span id="cb170-43"><a href="#cb170-43"></a>)</span>
<span id="cb170-44"><a href="#cb170-44"></a></span>
<span id="cb170-45"><a href="#cb170-45"></a></span>
<span id="cb170-46"><a href="#cb170-46"></a>missclass_oofProb_mtx &lt;-<span class="st"> </span><span class="kw">cbind</span>(</span>
<span id="cb170-47"><a href="#cb170-47"></a> train_lasso_oofProb_vec[misclass_id_vec],</span>
<span id="cb170-48"><a href="#cb170-48"></a> train_enet_oofProb_vec[misclass_id_vec],</span>
<span id="cb170-49"><a href="#cb170-49"></a> train_relaxed_oofProb_vec[misclass_id_vec],</span>
<span id="cb170-50"><a href="#cb170-50"></a> train_blended_oofProb_vec[misclass_id_vec]</span>
<span id="cb170-51"><a href="#cb170-51"></a>)</span>
<span id="cb170-52"><a href="#cb170-52"></a><span class="kw">colnames</span>(missclass_oofProb_mtx) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;lasso&#39;</span>,<span class="st">&#39;enet&#39;</span>, <span class="st">&#39;lassoR&#39;</span>, <span class="st">&#39;blended&#39;</span>)</span>
<span id="cb170-53"><a href="#cb170-53"></a></span>
<span id="cb170-54"><a href="#cb170-54"></a>row_med_vec &lt;-<span class="st"> </span><span class="kw">apply</span>(missclass_oofProb_mtx, <span class="dv">1</span>, median)</span>
<span id="cb170-55"><a href="#cb170-55"></a>missclass_oofProb_mtx &lt;-<span class="st"> </span>missclass_oofProb_mtx[</span>
<span id="cb170-56"><a href="#cb170-56"></a>  <span class="kw">order</span>(brcaRna_train_group_vec[<span class="kw">rownames</span>(missclass_oofProb_mtx)], row_med_vec),]</span>
<span id="cb170-57"><a href="#cb170-57"></a></span>
<span id="cb170-58"><a href="#cb170-58"></a><span class="kw">plot</span>(</span>
<span id="cb170-59"><a href="#cb170-59"></a> <span class="dt">x=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="kw">nrow</span>(missclass_oofProb_mtx)), <span class="dt">xlab=</span><span class="st">&#39;samples&#39;</span>,</span>
<span id="cb170-60"><a href="#cb170-60"></a> <span class="dt">y=</span><span class="kw">range</span>(missclass_oofProb_mtx), <span class="dt">ylab=</span><span class="st">&#39;out-of-fold predicted probability&#39;</span>,</span>
<span id="cb170-61"><a href="#cb170-61"></a> <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>, <span class="dt">type=</span><span class="st">&#39;n&#39;</span>)</span>
<span id="cb170-62"><a href="#cb170-62"></a></span>
<span id="cb170-63"><a href="#cb170-63"></a><span class="cf">for</span>(RR <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(missclass_oofProb_mtx))</span>
<span id="cb170-64"><a href="#cb170-64"></a><span class="kw">points</span>(</span>
<span id="cb170-65"><a href="#cb170-65"></a> <span class="kw">rep</span>(RR, <span class="kw">ncol</span>(missclass_oofProb_mtx)), </span>
<span id="cb170-66"><a href="#cb170-66"></a> missclass_oofProb_mtx[RR,],</span>
<span id="cb170-67"><a href="#cb170-67"></a> <span class="dt">col=</span><span class="kw">ifelse</span>(brcaRna_train_group_vec[<span class="kw">rownames</span>(missclass_oofProb_mtx)[RR]] <span class="op">==</span><span class="st"> &#39;Other&#39;</span>,</span>
<span id="cb170-68"><a href="#cb170-68"></a>  <span class="st">&#39;green&#39;</span>, <span class="st">&#39;red&#39;</span>),</span>
<span id="cb170-69"><a href="#cb170-69"></a> <span class="dt">pch=</span><span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(missclass_oofProb_mtx))</span>
<span id="cb170-70"><a href="#cb170-70"></a></span>
<span id="cb170-71"><a href="#cb170-71"></a><span class="kw">legend</span>(<span class="st">&#39;top&#39;</span>, <span class="dt">ncol=</span><span class="dv">2</span>, <span class="dt">legend=</span><span class="kw">colnames</span>(missclass_oofProb_mtx), </span>
<span id="cb170-72"><a href="#cb170-72"></a> <span class="dt">pch=</span><span class="dv">1</span><span class="op">:</span><span class="dv">4</span>, <span class="dt">bty=</span><span class="st">&#39;n&#39;</span>)</span>
<span id="cb170-73"><a href="#cb170-73"></a></span>
<span id="cb170-74"><a href="#cb170-74"></a><span class="kw">abline</span>(<span class="dt">h=</span><span class="fl">0.5</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-glmnetFit-misclassTrain-1.png" alt="out-of-fold predicted probabilities at miscassified samples" width="768" />
<p class="caption">
(#fig:brcaRna-glmnetFit-misclassTrain)out-of-fold predicted probabilities at miscassified samples
</p>
</div>
<p>As we’ve seen above, predictions from lassoR and the blended mix model
are basically dichotomous; 0 or 1. Samples have been order by group, and
median P(LumA) within group. For the Other (green), predicted probabilities
less than 0.5 are considered correct here. For the LumA (red) samples,
predicted probabilities greater than 0.5 are considered correct here.</p>
<p>Now look at the same plot on the test data set.</p>
<div class="sourceCode" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb171-1"><a href="#cb171-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb171-2"><a href="#cb171-2"></a></span>
<span id="cb171-3"><a href="#cb171-3"></a>test_lasso_predClass_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb171-4"><a href="#cb171-4"></a> brcaRna_cv_lasso,</span>
<span id="cb171-5"><a href="#cb171-5"></a> <span class="dt">newx=</span>brcaRna_test_geneExpr_mtx,</span>
<span id="cb171-6"><a href="#cb171-6"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span>,</span>
<span id="cb171-7"><a href="#cb171-7"></a> <span class="dt">type=</span><span class="st">&#39;class&#39;</span></span>
<span id="cb171-8"><a href="#cb171-8"></a>)</span>
<span id="cb171-9"><a href="#cb171-9"></a></span>
<span id="cb171-10"><a href="#cb171-10"></a>test_enet_predClass_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb171-11"><a href="#cb171-11"></a> brcaRna_cv_enet,</span>
<span id="cb171-12"><a href="#cb171-12"></a> <span class="dt">newx=</span>brcaRna_test_geneExpr_mtx,</span>
<span id="cb171-13"><a href="#cb171-13"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span>,</span>
<span id="cb171-14"><a href="#cb171-14"></a> <span class="dt">type=</span><span class="st">&#39;class&#39;</span></span>
<span id="cb171-15"><a href="#cb171-15"></a>)</span>
<span id="cb171-16"><a href="#cb171-16"></a></span>
<span id="cb171-17"><a href="#cb171-17"></a>test_relaxed_predClass_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb171-18"><a href="#cb171-18"></a> brcaRna_cv_lassoR,</span>
<span id="cb171-19"><a href="#cb171-19"></a> <span class="dt">g=</span><span class="dv">0</span>,</span>
<span id="cb171-20"><a href="#cb171-20"></a> <span class="dt">newx=</span>brcaRna_test_geneExpr_mtx,</span>
<span id="cb171-21"><a href="#cb171-21"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span>,</span>
<span id="cb171-22"><a href="#cb171-22"></a> <span class="dt">type=</span><span class="st">&#39;class&#39;</span></span>
<span id="cb171-23"><a href="#cb171-23"></a>)</span>
<span id="cb171-24"><a href="#cb171-24"></a></span>
<span id="cb171-25"><a href="#cb171-25"></a>test_blended_predClass_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb171-26"><a href="#cb171-26"></a> brcaRna_cv_lassoR,</span>
<span id="cb171-27"><a href="#cb171-27"></a> <span class="dt">g=</span><span class="fl">0.5</span>,</span>
<span id="cb171-28"><a href="#cb171-28"></a> <span class="dt">newx=</span>brcaRna_test_geneExpr_mtx,</span>
<span id="cb171-29"><a href="#cb171-29"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span>,</span>
<span id="cb171-30"><a href="#cb171-30"></a> <span class="dt">type=</span><span class="st">&#39;class&#39;</span></span>
<span id="cb171-31"><a href="#cb171-31"></a>)</span>
<span id="cb171-32"><a href="#cb171-32"></a></span>
<span id="cb171-33"><a href="#cb171-33"></a>misclass_id_vec &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">c</span>(</span>
<span id="cb171-34"><a href="#cb171-34"></a> <span class="kw">names</span>(test_lasso_predClass_vec[,<span class="dv">1</span>])[test_lasso_predClass_vec <span class="op">!=</span><span class="st"> </span>brcaRna_test_group_vec],</span>
<span id="cb171-35"><a href="#cb171-35"></a> <span class="kw">names</span>(test_enet_predClass_vec[,<span class="dv">1</span>])[test_enet_predClass_vec <span class="op">!=</span><span class="st"> </span>brcaRna_test_group_vec],</span>
<span id="cb171-36"><a href="#cb171-36"></a> <span class="kw">names</span>(test_relaxed_predClass_vec[,<span class="dv">1</span>])[test_relaxed_predClass_vec <span class="op">!=</span><span class="st"> </span>brcaRna_test_group_vec],</span>
<span id="cb171-37"><a href="#cb171-37"></a> <span class="kw">names</span>(test_blended_predClass_vec[,<span class="dv">1</span>])[test_blended_predClass_vec <span class="op">!=</span><span class="st"> </span>brcaRna_test_group_vec]</span>
<span id="cb171-38"><a href="#cb171-38"></a> )</span>
<span id="cb171-39"><a href="#cb171-39"></a>)</span>
<span id="cb171-40"><a href="#cb171-40"></a></span>
<span id="cb171-41"><a href="#cb171-41"></a></span>
<span id="cb171-42"><a href="#cb171-42"></a>missclass_oofProb_mtx &lt;-<span class="st"> </span><span class="kw">cbind</span>(</span>
<span id="cb171-43"><a href="#cb171-43"></a> test_lasso_predProb_vec[misclass_id_vec,],</span>
<span id="cb171-44"><a href="#cb171-44"></a> test_enet_predProb_vec[misclass_id_vec,],</span>
<span id="cb171-45"><a href="#cb171-45"></a> test_relaxed_predProb_vec[misclass_id_vec,],</span>
<span id="cb171-46"><a href="#cb171-46"></a> test_blended_predProb_vec[misclass_id_vec,]</span>
<span id="cb171-47"><a href="#cb171-47"></a>)</span>
<span id="cb171-48"><a href="#cb171-48"></a><span class="kw">colnames</span>(missclass_oofProb_mtx) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;lasso&#39;</span>,<span class="st">&#39;enet&#39;</span>, <span class="st">&#39;lassoR&#39;</span>, <span class="st">&#39;blended&#39;</span>)</span>
<span id="cb171-49"><a href="#cb171-49"></a></span>
<span id="cb171-50"><a href="#cb171-50"></a>row_med_vec &lt;-<span class="st"> </span><span class="kw">apply</span>(missclass_oofProb_mtx, <span class="dv">1</span>, median)</span>
<span id="cb171-51"><a href="#cb171-51"></a>missclass_oofProb_mtx &lt;-<span class="st"> </span>missclass_oofProb_mtx[</span>
<span id="cb171-52"><a href="#cb171-52"></a>  <span class="kw">order</span>(brcaRna_test_group_vec[<span class="kw">rownames</span>(missclass_oofProb_mtx)], row_med_vec),]</span>
<span id="cb171-53"><a href="#cb171-53"></a></span>
<span id="cb171-54"><a href="#cb171-54"></a><span class="kw">plot</span>(</span>
<span id="cb171-55"><a href="#cb171-55"></a> <span class="dt">x=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="kw">nrow</span>(missclass_oofProb_mtx)), <span class="dt">xlab=</span><span class="st">&#39;samples&#39;</span>,</span>
<span id="cb171-56"><a href="#cb171-56"></a> <span class="dt">y=</span><span class="kw">range</span>(missclass_oofProb_mtx), <span class="dt">ylab=</span><span class="st">&#39;out-of-fold predicted probability&#39;</span>,</span>
<span id="cb171-57"><a href="#cb171-57"></a> <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>, <span class="dt">type=</span><span class="st">&#39;n&#39;</span>)</span>
<span id="cb171-58"><a href="#cb171-58"></a></span>
<span id="cb171-59"><a href="#cb171-59"></a><span class="cf">for</span>(RR <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(missclass_oofProb_mtx))</span>
<span id="cb171-60"><a href="#cb171-60"></a><span class="kw">points</span>(</span>
<span id="cb171-61"><a href="#cb171-61"></a> <span class="kw">rep</span>(RR, <span class="kw">ncol</span>(missclass_oofProb_mtx)), </span>
<span id="cb171-62"><a href="#cb171-62"></a> missclass_oofProb_mtx[RR,],</span>
<span id="cb171-63"><a href="#cb171-63"></a> <span class="dt">col=</span><span class="kw">ifelse</span>(brcaRna_test_group_vec[<span class="kw">rownames</span>(missclass_oofProb_mtx)[RR]] <span class="op">==</span><span class="st"> &#39;Other&#39;</span>,</span>
<span id="cb171-64"><a href="#cb171-64"></a>  <span class="st">&#39;green&#39;</span>, <span class="st">&#39;red&#39;</span>),</span>
<span id="cb171-65"><a href="#cb171-65"></a> <span class="dt">pch=</span><span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(missclass_oofProb_mtx))</span>
<span id="cb171-66"><a href="#cb171-66"></a></span>
<span id="cb171-67"><a href="#cb171-67"></a><span class="kw">legend</span>(<span class="st">&#39;top&#39;</span>, <span class="dt">ncol=</span><span class="dv">2</span>, <span class="dt">legend=</span><span class="kw">colnames</span>(missclass_oofProb_mtx), </span>
<span id="cb171-68"><a href="#cb171-68"></a> <span class="dt">pch=</span><span class="dv">1</span><span class="op">:</span><span class="dv">4</span>, <span class="dt">bty=</span><span class="st">&#39;n&#39;</span>)</span>
<span id="cb171-69"><a href="#cb171-69"></a></span>
<span id="cb171-70"><a href="#cb171-70"></a><span class="kw">abline</span>(<span class="dt">h=</span><span class="fl">0.5</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-glmnetFit-misclassTest-1.png" alt="Test data predicted probabilities at miscassified samples" width="768" />
<p class="caption">
(#fig:brcaRna-glmnetFit-misclassTest)Test data predicted probabilities at miscassified samples
</p>
</div>
<p>The relaxed lasso fit results in essentially dichotomized predicted probability
distribution - predicted probabilities are very close to 0 or 1.</p>
<p>We see that for design points in the training set, the predicted probabilies from the relaxed lasso
are essentially dichotomized to be tightly distributed at the extremes of the
response range. For design points in the test set, the predicted probabilies from the relaxed lasso
are comparable to the lasso model predicted porbabilities. This seems to indicate over-fitting
in the relaxed lasso fit.</p>
</div>
<div id="compare-coefficient-profiles-1" class="section level2" number="7.6">
<h2 number="7.6"><span class="header-section-number">7.6</span> Compare coefficient profiles</h2>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="#cb172-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb172-2"><a href="#cb172-2"></a></span>
<span id="cb172-3"><a href="#cb172-3"></a><span class="co"># lasso </span></span>
<span id="cb172-4"><a href="#cb172-4"></a><span class="co">##########################</span></span>
<span id="cb172-5"><a href="#cb172-5"></a><span class="co"># train - cv predicted</span></span>
<span id="cb172-6"><a href="#cb172-6"></a>lasso_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb172-7"><a href="#cb172-7"></a> brcaRna_cv_lasso,</span>
<span id="cb172-8"><a href="#cb172-8"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span></span>
<span id="cb172-9"><a href="#cb172-9"></a>)</span>
<span id="cb172-10"><a href="#cb172-10"></a>lasso_coef_frm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb172-11"><a href="#cb172-11"></a> <span class="dt">gene=</span>lasso_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][<span class="kw">c</span>(<span class="dv">1</span>, lasso_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>])],</span>
<span id="cb172-12"><a href="#cb172-12"></a> <span class="dt">lasso=</span>lasso_coef<span class="op">@</span>x)</span>
<span id="cb172-13"><a href="#cb172-13"></a></span>
<span id="cb172-14"><a href="#cb172-14"></a></span>
<span id="cb172-15"><a href="#cb172-15"></a><span class="co"># enet</span></span>
<span id="cb172-16"><a href="#cb172-16"></a><span class="co">##########################</span></span>
<span id="cb172-17"><a href="#cb172-17"></a>enet_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb172-18"><a href="#cb172-18"></a> brcaRna_cv_enet,</span>
<span id="cb172-19"><a href="#cb172-19"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span></span>
<span id="cb172-20"><a href="#cb172-20"></a>)</span>
<span id="cb172-21"><a href="#cb172-21"></a>enet_coef_frm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb172-22"><a href="#cb172-22"></a> <span class="dt">gene=</span>enet_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][<span class="kw">c</span>(<span class="dv">1</span>, enet_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>])],</span>
<span id="cb172-23"><a href="#cb172-23"></a> <span class="dt">enet=</span>enet_coef<span class="op">@</span>x)</span>
<span id="cb172-24"><a href="#cb172-24"></a></span>
<span id="cb172-25"><a href="#cb172-25"></a><span class="co"># THESE ARE NOT CORRECT - SKIP</span></span>
<span id="cb172-26"><a href="#cb172-26"></a><span class="co"># relaxed lasso (gamma=0)</span></span>
<span id="cb172-27"><a href="#cb172-27"></a><span class="co">##########################</span></span>
<span id="cb172-28"><a href="#cb172-28"></a>SKIP &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb172-29"><a href="#cb172-29"></a>lassoR_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb172-30"><a href="#cb172-30"></a> brcaRna_cv_lassoR,</span>
<span id="cb172-31"><a href="#cb172-31"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span>,</span>
<span id="cb172-32"><a href="#cb172-32"></a> <span class="dt">g=</span><span class="dv">0</span></span>
<span id="cb172-33"><a href="#cb172-33"></a>)</span>
<span id="cb172-34"><a href="#cb172-34"></a>lassoR_coef_frm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb172-35"><a href="#cb172-35"></a> <span class="dt">gene=</span>lassoR_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][<span class="kw">c</span>(<span class="dv">1</span>, lassoR_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>])],</span>
<span id="cb172-36"><a href="#cb172-36"></a> <span class="dt">lassoR=</span>lassoR_coef<span class="op">@</span>x)</span>
<span id="cb172-37"><a href="#cb172-37"></a>}</span>
<span id="cb172-38"><a href="#cb172-38"></a></span>
<span id="cb172-39"><a href="#cb172-39"></a><span class="co"># blended mix (gamma=0.5)</span></span>
<span id="cb172-40"><a href="#cb172-40"></a><span class="co">###############################</span></span>
<span id="cb172-41"><a href="#cb172-41"></a>blended_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb172-42"><a href="#cb172-42"></a> brcaRna_cv_lassoR,</span>
<span id="cb172-43"><a href="#cb172-43"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span>,</span>
<span id="cb172-44"><a href="#cb172-44"></a> <span class="dt">g=</span><span class="fl">0.5</span></span>
<span id="cb172-45"><a href="#cb172-45"></a>)</span>
<span id="cb172-46"><a href="#cb172-46"></a>blended_coef_frm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb172-47"><a href="#cb172-47"></a> <span class="dt">gene=</span>blended_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][<span class="kw">c</span>(<span class="dv">1</span>, blended_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>])],</span>
<span id="cb172-48"><a href="#cb172-48"></a> <span class="dt">blended=</span>blended_coef<span class="op">@</span>x)</span>
<span id="cb172-49"><a href="#cb172-49"></a></span>
<span id="cb172-50"><a href="#cb172-50"></a></span>
<span id="cb172-51"><a href="#cb172-51"></a><span class="co"># put it all together</span></span>
<span id="cb172-52"><a href="#cb172-52"></a>all_coef_frm &lt;-<span class="st"> </span></span>
<span id="cb172-53"><a href="#cb172-53"></a><span class="st"> </span>base<span class="op">::</span><span class="kw">merge</span>(</span>
<span id="cb172-54"><a href="#cb172-54"></a> <span class="dt">x =</span> lasso_coef_frm, </span>
<span id="cb172-55"><a href="#cb172-55"></a> <span class="dt">y =</span> base<span class="op">::</span><span class="kw">merge</span>(</span>
<span id="cb172-56"><a href="#cb172-56"></a>     <span class="dt">x =</span> enet_coef_frm,</span>
<span id="cb172-57"><a href="#cb172-57"></a>     <span class="dt">y =</span> blended_coef_frm,</span>
<span id="cb172-58"><a href="#cb172-58"></a>         <span class="dt">by=</span><span class="st">&#39;gene&#39;</span>, <span class="dt">all=</span>T),</span>
<span id="cb172-59"><a href="#cb172-59"></a> <span class="dt">by=</span><span class="st">&#39;gene&#39;</span>, <span class="dt">all=</span>T)</span>
<span id="cb172-60"><a href="#cb172-60"></a></span>
<span id="cb172-61"><a href="#cb172-61"></a><span class="co"># SKIPPED</span></span>
<span id="cb172-62"><a href="#cb172-62"></a><span class="co">#base::merge(</span></span>
<span id="cb172-63"><a href="#cb172-63"></a>         <span class="co">#x = lassoR_coef_frm,</span></span>
<span id="cb172-64"><a href="#cb172-64"></a>         <span class="co">#y = blended_coef_frm,</span></span>
<span id="cb172-65"><a href="#cb172-65"></a>         <span class="co">#by=&#39;gene&#39;, all=T),</span></span>
<span id="cb172-66"><a href="#cb172-66"></a></span>
<span id="cb172-67"><a href="#cb172-67"></a><span class="co"># save gene name into rownames</span></span>
<span id="cb172-68"><a href="#cb172-68"></a><span class="kw">rownames</span>(all_coef_frm) &lt;-<span class="st"> </span>all_coef_frm<span class="op">$</span>gene</span>
<span id="cb172-69"><a href="#cb172-69"></a>all_coef_frm  <span class="op">%&lt;&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(<span class="op">-</span>gene)</span>
<span id="cb172-70"><a href="#cb172-70"></a></span>
<span id="cb172-71"><a href="#cb172-71"></a><span class="co"># set na to 0</span></span>
<span id="cb172-72"><a href="#cb172-72"></a>all_coef_frm[<span class="kw">is.na</span>(all_coef_frm)] &lt;-<span class="st"> </span><span class="dv">0</span></span>
<span id="cb172-73"><a href="#cb172-73"></a></span>
<span id="cb172-74"><a href="#cb172-74"></a><span class="co"># remove intercept</span></span>
<span id="cb172-75"><a href="#cb172-75"></a>all_coef_frm &lt;-<span class="st"> </span>all_coef_frm[<span class="op">!</span><span class="kw">rownames</span>(all_coef_frm)<span class="op">==</span><span class="st">&#39;(Intercept)&#39;</span>,]</span>
<span id="cb172-76"><a href="#cb172-76"></a></span>
<span id="cb172-77"><a href="#cb172-77"></a><span class="co"># order rows by size</span></span>
<span id="cb172-78"><a href="#cb172-78"></a>all_coef_frm &lt;-<span class="st"> </span>all_coef_frm[<span class="kw">order</span>(<span class="kw">rowMeans</span>(<span class="kw">abs</span>(all_coef_frm)), <span class="dt">decreasing=</span>T), ]</span>
<span id="cb172-79"><a href="#cb172-79"></a></span>
<span id="cb172-80"><a href="#cb172-80"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="kw">ncol</span>(all_coef_frm),<span class="dv">1</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">5</span>,<span class="fl">0.5</span>,<span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb172-81"><a href="#cb172-81"></a></span>
<span id="cb172-82"><a href="#cb172-82"></a><span class="cf">for</span>(CC <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(all_coef_frm)) {</span>
<span id="cb172-83"><a href="#cb172-83"></a> <span class="kw">plot</span>(</span>
<span id="cb172-84"><a href="#cb172-84"></a>  <span class="dt">x=</span><span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(all_coef_frm), <span class="dt">xlab=</span><span class="st">&#39;&#39;</span>, </span>
<span id="cb172-85"><a href="#cb172-85"></a>  <span class="dt">y=</span>all_coef_frm[, CC], <span class="dt">ylab=</span><span class="kw">colnames</span>(all_coef_frm)[CC],</span>
<span id="cb172-86"><a href="#cb172-86"></a>  <span class="dt">type=</span><span class="st">&#39;h&#39;</span>, <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>)</span>
<span id="cb172-87"><a href="#cb172-87"></a>}</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-glmnetFit-compCoeffProf-1.png" alt="Coefficient Profiles" width="768" />
<p class="caption">
(#fig:brcaRna-glmnetFit-compCoeffProf)Coefficient Profiles
</p>
</div>
<p>Note that there is little difference between the elastic net and the lasso
in the selected features, and when the coefficient is zero in one set, it
is smaell in the other. By contrast, the blended fit produces more shrinkage.</p>
<div class="sourceCode" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb173-1"><a href="#cb173-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb173-2"><a href="#cb173-2"></a></span>
<span id="cb173-3"><a href="#cb173-3"></a>knitr<span class="op">::</span><span class="kw">kable</span>(</span>
<span id="cb173-4"><a href="#cb173-4"></a><span class="kw">with</span>(all_coef_frm, <span class="kw">table</span>(<span class="dt">lassoZero=</span>lasso<span class="op">==</span><span class="dv">0</span>, <span class="dt">enetZero=</span>enet<span class="op">==</span><span class="dv">0</span>)),</span>
<span id="cb173-5"><a href="#cb173-5"></a> <span class="dt">caption=</span><span class="st">&#39;Zero Ceofficient: rows are lasso, columns enet&#39;</span>) <span class="op">%&gt;%</span></span>
<span id="cb173-6"><a href="#cb173-6"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:brcaRna-glmnetFit-zreros)Zero Ceofficient: rows are lasso, columns enet
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
FALSE
</th>
<th style="text-align:right;">
TRUE
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
168
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
147
</td>
<td style="text-align:right;">
6
</td>
</tr>
</tbody>
</table>
<!--
Coefficients in the relaxed lasso fit are much larger than those in the
lasso fit, or zero.  As a consequence, the blended fit coefficients look 
like a shrunken version of the relaxed lasso fit coefficients.  
-->
<p>Coefficients in the blended fit are larger than those in the
lasso fit, or zero.</p>
<p>We can also examine these with a scatter plot matrix.</p>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="#cb174-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb174-2"><a href="#cb174-2"></a></span>
<span id="cb174-3"><a href="#cb174-3"></a></span>
<span id="cb174-4"><a href="#cb174-4"></a><span class="kw">pairs</span>(all_coef_frm,</span>
<span id="cb174-5"><a href="#cb174-5"></a>  <span class="dt">lower.panel =</span> <span class="ot">NULL</span>,</span>
<span id="cb174-6"><a href="#cb174-6"></a>  <span class="dt">panel =</span> <span class="cf">function</span>(x, y) {</span>
<span id="cb174-7"><a href="#cb174-7"></a>    <span class="kw">points</span>(x, y, <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">col =</span> <span class="st">&quot;blue&quot;</span>)</span>
<span id="cb174-8"><a href="#cb174-8"></a>  }</span>
<span id="cb174-9"><a href="#cb174-9"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-glmnetFit-pairsCoeffProf-1.png" alt="Coefficients from fits" width="768" />
<p class="caption">
(#fig:brcaRna-glmnetFit-pairsCoeffProf)Coefficients from fits
</p>
</div>
</div>
<div id="examine-feature-selection-1" class="section level2" number="7.7">
<h2 number="7.7"><span class="header-section-number">7.7</span> Examine feature selection</h2>
<p>Recall from <code>glmnet</code> vignette:</p>
<pre><code>It is known that the ridge penalty shrinks the coefficients of correlated predictors
towards each other while the lasso tends to pick one of them and discard the others.
The elastic-net penalty mixes these two; if predictors are correlated in groups,
an $\alpha$=0.5 tends to select the groups in or out together.
This is a higher level parameter, and users might pick a value upfront,
else experiment with a few different values. One use of $\alpha$ is for numerical stability;
for example, the *elastic net with $\alpha = 1 - \epsilon$ for some small $\epsilon$&gt;0
performs much like the lasso, but removes any degeneracies and wild behavior caused
by extreme correlations*.</code></pre>
<p>To see how this plays out in this dataset, we can look at feature expression
heat maps.</p>
<p>Reader notes:</p>
<pre><code>Heat maps are rarely useful other than to display the obvious.
Here too heat maps fail to yield any insights, or confirmation
of the relationship between feature correlation and lasso vs enet
feature selection.</code></pre>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb177-1"><a href="#cb177-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb177-2"><a href="#cb177-2"></a> <span class="kw">suppressPackageStartupMessages</span>(<span class="kw">require</span>(gplots))</span>
<span id="cb177-3"><a href="#cb177-3"></a></span>
<span id="cb177-4"><a href="#cb177-4"></a><span class="co"># train - cv predicted</span></span>
<span id="cb177-5"><a href="#cb177-5"></a>lasso_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb177-6"><a href="#cb177-6"></a> brcaRna_cv_lasso,</span>
<span id="cb177-7"><a href="#cb177-7"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span></span>
<span id="cb177-8"><a href="#cb177-8"></a>)</span>
<span id="cb177-9"><a href="#cb177-9"></a>lasso_coef_frm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb177-10"><a href="#cb177-10"></a> <span class="dt">gene=</span>lasso_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][<span class="kw">c</span>(<span class="dv">1</span>, lasso_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>])],</span>
<span id="cb177-11"><a href="#cb177-11"></a> <span class="dt">lasso=</span>lasso_coef<span class="op">@</span>x)</span>
<span id="cb177-12"><a href="#cb177-12"></a></span>
<span id="cb177-13"><a href="#cb177-13"></a> </span>
<span id="cb177-14"><a href="#cb177-14"></a>  Mycol &lt;-<span class="st"> </span><span class="kw">colorpanel</span>(<span class="dv">1000</span>, <span class="st">&quot;blue&quot;</span>, <span class="st">&quot;red&quot;</span>)</span>
<span id="cb177-15"><a href="#cb177-15"></a>  <span class="kw">heatmap.2</span>(</span>
<span id="cb177-16"><a href="#cb177-16"></a>    <span class="dt">x=</span><span class="kw">t</span>(brcaRna_train_geneExpr_mtx[,lasso_coef_frm<span class="op">$</span>gene[<span class="op">-</span><span class="dv">1</span>]]),</span>
<span id="cb177-17"><a href="#cb177-17"></a>    <span class="dt">scale=</span><span class="st">&quot;row&quot;</span>,</span>
<span id="cb177-18"><a href="#cb177-18"></a>    <span class="dt">labRow=</span>lasso_coef_frm<span class="op">$</span>gene,</span>
<span id="cb177-19"><a href="#cb177-19"></a>    <span class="dt">labCol=</span>brcaRna_train_group_vec,</span>
<span id="cb177-20"><a href="#cb177-20"></a>    <span class="dt">col=</span>Mycol, </span>
<span id="cb177-21"><a href="#cb177-21"></a>    <span class="dt">trace=</span><span class="st">&quot;none&quot;</span>, <span class="dt">density.info=</span><span class="st">&quot;none&quot;</span>, </span>
<span id="cb177-22"><a href="#cb177-22"></a>    <span class="co">#margin=c(8,6), lhei=c(2,10), </span></span>
<span id="cb177-23"><a href="#cb177-23"></a>    <span class="co">#lwid=c(0.1,4), #lhei=c(0.1,4)</span></span>
<span id="cb177-24"><a href="#cb177-24"></a>    <span class="dt">key=</span>F,</span>
<span id="cb177-25"><a href="#cb177-25"></a>    <span class="dt">ColSideColors=</span><span class="kw">ifelse</span>(brcaRna_train_group_vec<span class="op">==</span><span class="st">&#39;_Other&#39;</span>, <span class="st">&#39;green&#39;</span>,<span class="st">&#39;red&#39;</span>),</span>
<span id="cb177-26"><a href="#cb177-26"></a>    <span class="dt">dendrogram=</span><span class="st">&quot;both&quot;</span>,</span>
<span id="cb177-27"><a href="#cb177-27"></a>    <span class="dt">main=</span><span class="kw">paste</span>(<span class="st">&#39;lasso genes - N =&#39;</span>, <span class="kw">nrow</span>(lasso_coef_frm)<span class="op">-</span><span class="dv">1</span>))</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-glmnetFit-heatmapLasso-1.png" alt="Lasso Model Genes" width="768" />
<p class="caption">
(#fig:brcaRna-glmnetFit-heatmapLasso)Lasso Model Genes
</p>
</div>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="#cb178-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb178-2"><a href="#cb178-2"></a> <span class="kw">suppressPackageStartupMessages</span>(<span class="kw">require</span>(gplots))</span>
<span id="cb178-3"><a href="#cb178-3"></a></span>
<span id="cb178-4"><a href="#cb178-4"></a><span class="co"># train - cv predicted</span></span>
<span id="cb178-5"><a href="#cb178-5"></a>enet_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb178-6"><a href="#cb178-6"></a> brcaRna_cv_enet,</span>
<span id="cb178-7"><a href="#cb178-7"></a> <span class="dt">s=</span><span class="st">&#39;lambda.1se&#39;</span></span>
<span id="cb178-8"><a href="#cb178-8"></a>)</span>
<span id="cb178-9"><a href="#cb178-9"></a>enet_coef_frm &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb178-10"><a href="#cb178-10"></a> <span class="dt">gene=</span>enet_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][<span class="kw">c</span>(<span class="dv">1</span>, enet_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>])],</span>
<span id="cb178-11"><a href="#cb178-11"></a> <span class="dt">enet=</span>enet_coef<span class="op">@</span>x)</span>
<span id="cb178-12"><a href="#cb178-12"></a></span>
<span id="cb178-13"><a href="#cb178-13"></a> </span>
<span id="cb178-14"><a href="#cb178-14"></a>  Mycol &lt;-<span class="st"> </span><span class="kw">colorpanel</span>(<span class="dv">1000</span>, <span class="st">&quot;blue&quot;</span>, <span class="st">&quot;red&quot;</span>)</span>
<span id="cb178-15"><a href="#cb178-15"></a>  <span class="kw">heatmap.2</span>(</span>
<span id="cb178-16"><a href="#cb178-16"></a>    <span class="dt">x=</span><span class="kw">t</span>(brcaRna_train_geneExpr_mtx[,enet_coef_frm<span class="op">$</span>gene[<span class="op">-</span><span class="dv">1</span>]]),</span>
<span id="cb178-17"><a href="#cb178-17"></a>    <span class="dt">scale=</span><span class="st">&quot;row&quot;</span>,</span>
<span id="cb178-18"><a href="#cb178-18"></a>    <span class="dt">labRow=</span>enet_coef_frm<span class="op">$</span>gene,</span>
<span id="cb178-19"><a href="#cb178-19"></a>    <span class="dt">labCol=</span>brcaRna_train_group_vec,</span>
<span id="cb178-20"><a href="#cb178-20"></a>    <span class="dt">col=</span>Mycol, </span>
<span id="cb178-21"><a href="#cb178-21"></a>    <span class="dt">trace=</span><span class="st">&quot;none&quot;</span>, <span class="dt">density.info=</span><span class="st">&quot;none&quot;</span>, </span>
<span id="cb178-22"><a href="#cb178-22"></a>    <span class="co">#margin=c(8,6), lhei=c(2,10), </span></span>
<span id="cb178-23"><a href="#cb178-23"></a>    <span class="co">#lwid=c(0.1,4), #lhei=c(0.1,4)</span></span>
<span id="cb178-24"><a href="#cb178-24"></a>    <span class="dt">key=</span>F,</span>
<span id="cb178-25"><a href="#cb178-25"></a>    <span class="dt">ColSideColors=</span><span class="kw">ifelse</span>(brcaRna_train_group_vec<span class="op">==</span><span class="st">&#39;_Other&#39;</span>, <span class="st">&#39;green&#39;</span>,<span class="st">&#39;red&#39;</span>),</span>
<span id="cb178-26"><a href="#cb178-26"></a>    <span class="dt">dendrogram=</span><span class="st">&quot;both&quot;</span>,</span>
<span id="cb178-27"><a href="#cb178-27"></a>    <span class="dt">main=</span><span class="kw">paste</span>(<span class="st">&#39;enet genes - N =&#39;</span>, <span class="kw">nrow</span>(enet_coef_frm)<span class="op">-</span><span class="dv">1</span>))</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brcaRna-glmnetFit-heatmapEnet-1.png" alt="Enet Model Genes" width="768" />
<p class="caption">
(#fig:brcaRna-glmnetFit-heatmapEnet)Enet Model Genes
</p>
</div>
<!--chapter:end:07-glmnetFitsBrCaRNASeq.Rmd-->
</div>
</div>
<div id="brca-rnaseq-model-suite" class="section level1" number="8">
<h1 number="8"><span class="header-section-number">8</span> BrCa RNA-Seq: Sample size investigation</h1>
<p>In this section we repeat the analyses ran on the HCC 5hmC data in
Section @ref(hcc-5hmcseq-model-suite) on the breast cancer
RNA-Seq data set to provide a contrasting SNR regime context.</p>
<div id="full-data-set-fit-1" class="section level2" number="8.1">
<h2 number="8.1"><span class="header-section-number">8.1</span> Full data set fit</h2>
<p>We begin by fitting a model to the entire data set in order to:</p>
<ul>
<li><p>obtain a baseline clssification performance against which to judge the performance
obtained from the fits to smaller sample sets,</p></li>
<li><p>obtain sample consistency scores which can be used to explain variability
in the performance of model fitted to data sets of a fixed size, and</p></li>
<li><p>produce a <em>full model</em> gene signature which can be used to evaluate
the stability of selected features in models fitted to data sets of diffferent
sizes.</p></li>
</ul>
<p>First assemble the data set. This entails simply re-combining the
train and test data.</p>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="#cb179-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb179-2"><a href="#cb179-2"></a></span>
<span id="cb179-3"><a href="#cb179-3"></a><span class="co"># combine train and test </span></span>
<span id="cb179-4"><a href="#cb179-4"></a>brcaRna_all_geneExpr_mtx &lt;-<span class="st"> </span><span class="kw">rbind</span>(brcaRna_train_geneExpr_mtx, brcaRna_test_geneExpr_mtx)</span>
<span id="cb179-5"><a href="#cb179-5"></a></span>
<span id="cb179-6"><a href="#cb179-6"></a><span class="co"># we have to be careful with factors!</span></span>
<span id="cb179-7"><a href="#cb179-7"></a><span class="co"># We&#39;ll keep as a character and change to factor when needed</span></span>
<span id="cb179-8"><a href="#cb179-8"></a>brcaRna_all_group_vec &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb179-9"><a href="#cb179-9"></a> <span class="kw">as.character</span>(brcaRna_train_group_vec), </span>
<span id="cb179-10"><a href="#cb179-10"></a> <span class="kw">as.character</span>(brcaRna_test_group_vec)</span>
<span id="cb179-11"><a href="#cb179-11"></a>)</span>
<span id="cb179-12"><a href="#cb179-12"></a><span class="co"># I suspect adding names to vectors breaks one of the tidy commandments,</span></span>
<span id="cb179-13"><a href="#cb179-13"></a><span class="co"># but then again I am sure I have already offended the creed beyond salvation</span></span>
<span id="cb179-14"><a href="#cb179-14"></a><span class="kw">names</span>(brcaRna_all_group_vec) &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb179-15"><a href="#cb179-15"></a> <span class="kw">names</span>(brcaRna_train_group_vec),</span>
<span id="cb179-16"><a href="#cb179-16"></a> <span class="kw">names</span>(brcaRna_test_group_vec)</span>
<span id="cb179-17"><a href="#cb179-17"></a>)</span>
<span id="cb179-18"><a href="#cb179-18"></a></span>
<span id="cb179-19"><a href="#cb179-19"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">table</span>(<span class="dt">group =</span> brcaRna_all_group_vec),</span>
<span id="cb179-20"><a href="#cb179-20"></a>  <span class="dt">caption =</span> <span class="st">&quot;samples by group&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb179-21"><a href="#cb179-21"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:brca-rnaseq-get-all-data)samples by group
</caption>
<thead>
<tr>
<th style="text-align:left;">
group
</th>
<th style="text-align:right;">
Freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
LumA
</td>
<td style="text-align:right;">
1492
</td>
</tr>
<tr>
<td style="text-align:left;">
Other
</td>
<td style="text-align:right;">
933
</td>
</tr>
</tbody>
</table>
<p>Now fit the losso model through cross-validation.
Note that the results of a cv fit are random due to the
random allocation of samples to folds. We can reduce this
varibility by properly averaging results over repeated cv fits.
Here we will obtain sample consistency scores by averaging results
over 30 cv runs.</p>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="#cb180-1"></a><span class="kw">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb180-2"><a href="#cb180-2"></a></span>
<span id="cb180-3"><a href="#cb180-3"></a>start_time &lt;-<span class="st">  </span><span class="kw">proc.time</span>()</span>
<span id="cb180-4"><a href="#cb180-4"></a></span>
<span id="cb180-5"><a href="#cb180-5"></a>brcaRna_cv_lassoAll_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">30</span>, <span class="cf">function</span>(REP) {</span>
<span id="cb180-6"><a href="#cb180-6"></a>glmnet<span class="op">::</span><span class="kw">cv.glmnet</span>(</span>
<span id="cb180-7"><a href="#cb180-7"></a> <span class="dt">x =</span> brcaRna_all_geneExpr_mtx,</span>
<span id="cb180-8"><a href="#cb180-8"></a> <span class="dt">y =</span> <span class="kw">factor</span>(brcaRna_all_group_vec,<span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&#39;Other&#39;</span>, <span class="st">&#39;LumA&#39;</span>)),</span>
<span id="cb180-9"><a href="#cb180-9"></a> <span class="dt">alpha =</span> <span class="dv">1</span>,</span>
<span id="cb180-10"><a href="#cb180-10"></a> <span class="dt">family =</span> <span class="st">&#39;binomial&#39;</span>,</span>
<span id="cb180-11"><a href="#cb180-11"></a> <span class="dt">type.measure  =</span>  <span class="st">&quot;class&quot;</span>,</span>
<span id="cb180-12"><a href="#cb180-12"></a> <span class="dt">keep =</span> T,</span>
<span id="cb180-13"><a href="#cb180-13"></a> <span class="dt">nlambda =</span> <span class="dv">100</span></span>
<span id="cb180-14"><a href="#cb180-14"></a>)</span>
<span id="cb180-15"><a href="#cb180-15"></a>}</span>
<span id="cb180-16"><a href="#cb180-16"></a>)</span>
<span id="cb180-17"><a href="#cb180-17"></a></span>
<span id="cb180-18"><a href="#cb180-18"></a><span class="kw">message</span>(<span class="st">&quot;lassoAll time: &quot;</span>, <span class="kw">round</span>((<span class="kw">proc.time</span>() <span class="op">-</span><span class="st"> </span>start_time)[<span class="dv">3</span>],<span class="dv">2</span>),<span class="st">&quot;s&quot;</span>)</span></code></pre></div>
<!-- lasso-fit-all takes a while - save results -->
<!-- DO THIS ONCE -->
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="#cb181-1"></a> <span class="kw">load</span>(<span class="dt">file=</span><span class="kw">file.path</span>(<span class="st">&quot;RData&quot;</span>,<span class="st">&#39;brcaRna_cv_lassoAll_lst&#39;</span>))</span></code></pre></div>
<p>Examine the fits.</p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="#cb182-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb182-2"><a href="#cb182-2"></a><span class="kw">plot</span>(</span>
<span id="cb182-3"><a href="#cb182-3"></a> <span class="kw">log</span>(brcaRna_cv_lassoAll_lst[[<span class="dv">1</span>]]<span class="op">$</span>lambda),</span>
<span id="cb182-4"><a href="#cb182-4"></a> brcaRna_cv_lassoAll_lst[[<span class="dv">1</span>]]<span class="op">$</span>cvm,</span>
<span id="cb182-5"><a href="#cb182-5"></a> <span class="dt">lwd=</span><span class="dv">2</span>,</span>
<span id="cb182-6"><a href="#cb182-6"></a> <span class="dt">xlab=</span><span class="st">&#39;log(Lambda)&#39;</span>, <span class="dt">ylab=</span><span class="st">&#39;CV Misclassification Error&#39;</span>, <span class="dt">type=</span><span class="st">&#39;l&#39;</span>, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">.5</span>)</span>
<span id="cb182-7"><a href="#cb182-7"></a>)</span>
<span id="cb182-8"><a href="#cb182-8"></a></span>
<span id="cb182-9"><a href="#cb182-9"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">2</span><span class="op">:</span><span class="kw">length</span>(brcaRna_cv_lassoAll_lst))</span>
<span id="cb182-10"><a href="#cb182-10"></a> <span class="kw">lines</span>(</span>
<span id="cb182-11"><a href="#cb182-11"></a>  <span class="kw">log</span>(brcaRna_cv_lassoAll_lst[[JJ]]<span class="op">$</span>lambda),</span>
<span id="cb182-12"><a href="#cb182-12"></a>  brcaRna_cv_lassoAll_lst[[JJ]]<span class="op">$</span>cvm,</span>
<span id="cb182-13"><a href="#cb182-13"></a>  <span class="dt">lwd=</span><span class="dv">2</span></span>
<span id="cb182-14"><a href="#cb182-14"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brca-rnaseq-plot-lassoAll-1.png" alt="Repeated cv lasso models fitted to all samples" width="576" />
<p class="caption">
(#fig:brca-rnaseq-plot-lassoAll)Repeated cv lasso models fitted to all samples
</p>
</div>
<p>These cv curves are remarkably consistent meaning that the determination of the size or sparsity
of the model fitted through cross validation to the full data set is fairly precise:</p>
<!-- DONT CACHE THIS ??? -->
<div class="sourceCode" id="cb183"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb183-1"><a href="#cb183-1"></a><span class="kw">library</span>(magrittr)</span>
<span id="cb183-2"><a href="#cb183-2"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb183-3"><a href="#cb183-3"></a></span>
<span id="cb183-4"><a href="#cb183-4"></a><span class="co"># nzero</span></span>
<span id="cb183-5"><a href="#cb183-5"></a>nzero_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb183-6"><a href="#cb183-6"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb183-7"><a href="#cb183-7"></a></span>
<span id="cb183-8"><a href="#cb183-8"></a>nzero_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb183-9"><a href="#cb183-9"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb183-10"><a href="#cb183-10"></a></span>
<span id="cb183-11"><a href="#cb183-11"></a><span class="kw">boxplot</span>(<span class="kw">list</span>(<span class="st">`</span><span class="dt">1se</span><span class="st">`</span>=nzero_1se_vec, <span class="dt">min =</span> nzero_min_vec), <span class="dt">ylab=</span><span class="st">&quot;Selected Features&quot;</span>)</span>
<span id="cb183-12"><a href="#cb183-12"></a></span>
<span id="cb183-13"><a href="#cb183-13"></a><span class="co"># error</span></span>
<span id="cb183-14"><a href="#cb183-14"></a>error_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb183-15"><a href="#cb183-15"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb183-16"><a href="#cb183-16"></a></span>
<span id="cb183-17"><a href="#cb183-17"></a>error_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb183-18"><a href="#cb183-18"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb183-19"><a href="#cb183-19"></a></span>
<span id="cb183-20"><a href="#cb183-20"></a><span class="kw">boxplot</span>(</span>
<span id="cb183-21"><a href="#cb183-21"></a> <span class="kw">list</span>(<span class="st">`</span><span class="dt">1se</span><span class="st">`</span>=error_1se_vec, <span class="dt">min =</span> error_min_vec), </span>
<span id="cb183-22"><a href="#cb183-22"></a> <span class="dt">ylab=</span>brcaRna_cv_lassoAll_lst[[<span class="dv">1</span>]]<span class="op">$</span>name,</span>
<span id="cb183-23"><a href="#cb183-23"></a> <span class="dt">ylim=</span><span class="kw">c</span>(<span class="fl">0.10</span>, <span class="fl">.13</span>)</span>
<span id="cb183-24"><a href="#cb183-24"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brca-rnaseq-model-size-lassoAll-1.png" alt="Feature selection and estimated error by repeated cv lasso models" width="768" />
<p class="caption">
(#fig:brca-rnaseq-model-size-lassoAll)Feature selection and estimated error by repeated cv lasso models
</p>
</div>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="#cb184-1"></a><span class="co"># tabular format</span></span>
<span id="cb184-2"><a href="#cb184-2"></a>tmp &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">rbind</span>(</span>
<span id="cb184-3"><a href="#cb184-3"></a> <span class="st">`</span><span class="dt">features_1se</span><span class="st">`</span> =<span class="st"> </span><span class="kw">summary</span>(nzero_1se_vec),</span>
<span id="cb184-4"><a href="#cb184-4"></a> <span class="dt">features_min =</span> <span class="kw">summary</span>(nzero_min_vec),</span>
<span id="cb184-5"><a href="#cb184-5"></a> <span class="st">`</span><span class="dt">features:min-1se</span><span class="st">`</span> =<span class="st"> </span><span class="kw">summary</span>(nzero_min_vec <span class="op">-</span><span class="st"> </span>nzero_1se_vec),</span>
<span id="cb184-6"><a href="#cb184-6"></a> <span class="st">`</span><span class="dt">cv_error_1se</span><span class="st">`</span> =<span class="st"> </span><span class="kw">summary</span>(<span class="dv">100</span><span class="op">*</span>error_1se_vec),</span>
<span id="cb184-7"><a href="#cb184-7"></a> <span class="dt">cv_error_min =</span> <span class="kw">summary</span>(<span class="dv">100</span><span class="op">*</span>error_min_vec),</span>
<span id="cb184-8"><a href="#cb184-8"></a> <span class="st">`</span><span class="dt">cv_error:1se-min</span><span class="st">`</span> =<span class="st"> </span><span class="kw">summary</span>(<span class="dv">100</span><span class="op">*</span>(error_1se_vec<span class="op">-</span>error_min_vec))</span>
<span id="cb184-9"><a href="#cb184-9"></a>))</span>
<span id="cb184-10"><a href="#cb184-10"></a></span>
<span id="cb184-11"><a href="#cb184-11"></a>knitr<span class="op">::</span><span class="kw">kable</span>(tmp <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(<span class="op">-</span>Mean),</span>
<span id="cb184-12"><a href="#cb184-12"></a>  <span class="dt">caption =</span> <span class="st">&quot;Number of selected features&quot;</span>,</span>
<span id="cb184-13"><a href="#cb184-13"></a>  <span class="dt">digits=</span><span class="dv">1</span>) <span class="op">%&gt;%</span></span>
<span id="cb184-14"><a href="#cb184-14"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:brca-rnaseq-model-size-lassoAll)Number of selected features
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Min.
</th>
<th style="text-align:right;">
X1st.Qu.
</th>
<th style="text-align:right;">
Median
</th>
<th style="text-align:right;">
X3rd.Qu.
</th>
<th style="text-align:right;">
Max.
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
features_1se
</td>
<td style="text-align:right;">
97.0
</td>
<td style="text-align:right;">
118.5
</td>
<td style="text-align:right;">
144.0
</td>
<td style="text-align:right;">
162.0
</td>
<td style="text-align:right;">
208.0
</td>
</tr>
<tr>
<td style="text-align:left;">
features_min
</td>
<td style="text-align:right;">
144.0
</td>
<td style="text-align:right;">
221.2
</td>
<td style="text-align:right;">
258.0
</td>
<td style="text-align:right;">
291.0
</td>
<td style="text-align:right;">
451.0
</td>
</tr>
<tr>
<td style="text-align:left;">
features:min-1se
</td>
<td style="text-align:right;">
27.0
</td>
<td style="text-align:right;">
59.0
</td>
<td style="text-align:right;">
94.5
</td>
<td style="text-align:right;">
153.2
</td>
<td style="text-align:right;">
344.0
</td>
</tr>
<tr>
<td style="text-align:left;">
cv_error_1se
</td>
<td style="text-align:right;">
10.9
</td>
<td style="text-align:right;">
11.3
</td>
<td style="text-align:right;">
11.5
</td>
<td style="text-align:right;">
11.7
</td>
<td style="text-align:right;">
12.0
</td>
</tr>
<tr>
<td style="text-align:left;">
cv_error_min
</td>
<td style="text-align:right;">
10.6
</td>
<td style="text-align:right;">
10.8
</td>
<td style="text-align:right;">
10.9
</td>
<td style="text-align:right;">
11.1
</td>
<td style="text-align:right;">
11.5
</td>
</tr>
<tr>
<td style="text-align:left;">
cv_error:1se-min
</td>
<td style="text-align:right;">
0.2
</td>
<td style="text-align:right;">
0.4
</td>
<td style="text-align:right;">
0.5
</td>
<td style="text-align:right;">
0.7
</td>
<td style="text-align:right;">
0.8
</td>
</tr>
</tbody>
</table>
<p><br/></p>
<p>The number of features selected by the minimum lambda models are larger
than the number selected by the “one standard error” rule models by a median
of 94.5.
The cv error rates obtained from the minimum lambda models are lower
then “one standard error” rule models error rates by a median of
0.5%.</p>
<p>The cv error rates observed in this set are comparable to the
rates oberved in the lasso models fitted to the training sample set
which consisted of 80% of the samples in this set. In other words,
there is no obvious gain in performance in moving from
a data set with
1194 vs 746 samples
to a data set with
1492 vs 933 samples.
See Table @ref(tab:printErrors).</p>
<p>It’s not clear at this point whether the minimum lambda model is truly better than
the “one standard error” rule model. We would need and external validation
set to make this determination. We can compare the two sets
of out-of-fold predicted values, averaged across cv replicates, to see if
there is a meaningful difference between the two.</p>
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb185-1"><a href="#cb185-1"></a><span class="co"># predicted probs - 1se</span></span>
<span id="cb185-2"><a href="#cb185-2"></a>lassoAll_predResp_1se_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst, <span class="cf">function</span>(cv_fit) { </span>
<span id="cb185-3"><a href="#cb185-3"></a>  ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se,cv_fit<span class="op">$</span>lambda)</span>
<span id="cb185-4"><a href="#cb185-4"></a>  <span class="kw">logistic_f</span>(cv_fit<span class="op">$</span>fit.preval[,ndx_1se])</span>
<span id="cb185-5"><a href="#cb185-5"></a> })</span>
<span id="cb185-6"><a href="#cb185-6"></a>lassoAll_predResp_1se_vec &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(lassoAll_predResp_1se_mtx)</span>
<span id="cb185-7"><a href="#cb185-7"></a></span>
<span id="cb185-8"><a href="#cb185-8"></a><span class="co"># predicted probs - min</span></span>
<span id="cb185-9"><a href="#cb185-9"></a>lassoAll_predResp_min_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst, <span class="cf">function</span>(cv_fit) { </span>
<span id="cb185-10"><a href="#cb185-10"></a>  ndx_min &lt;-<span class="st"> </span><span class="kw">match</span>(cv_fit<span class="op">$</span>lambda.min,cv_fit<span class="op">$</span>lambda)</span>
<span id="cb185-11"><a href="#cb185-11"></a>  <span class="kw">logistic_f</span>(cv_fit<span class="op">$</span>fit.preval[,ndx_min])</span>
<span id="cb185-12"><a href="#cb185-12"></a> })</span>
<span id="cb185-13"><a href="#cb185-13"></a>lassoAll_predResp_min_vec &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(lassoAll_predResp_min_mtx)</span>
<span id="cb185-14"><a href="#cb185-14"></a></span>
<span id="cb185-15"><a href="#cb185-15"></a><span class="co"># plot</span></span>
<span id="cb185-16"><a href="#cb185-16"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">5</span>,<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb185-17"><a href="#cb185-17"></a>tmp &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb185-18"><a href="#cb185-18"></a> <span class="st">`</span><span class="dt">1se</span><span class="st">`</span> =<span class="st"> </span><span class="kw">split</span>(lassoAll_predResp_1se_vec, brcaRna_all_group_vec),</span>
<span id="cb185-19"><a href="#cb185-19"></a> <span class="dt">min =</span> <span class="kw">split</span>(lassoAll_predResp_min_vec, brcaRna_all_group_vec)</span>
<span id="cb185-20"><a href="#cb185-20"></a>)</span>
<span id="cb185-21"><a href="#cb185-21"></a><span class="kw">names</span>(tmp) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&#39;</span><span class="ch">\\</span><span class="st">.&#39;</span>, <span class="st">&#39;</span><span class="ch">\n</span><span class="st">&#39;</span>, <span class="kw">names</span>(tmp))</span>
<span id="cb185-22"><a href="#cb185-22"></a></span>
<span id="cb185-23"><a href="#cb185-23"></a><span class="kw">boxplot</span>(</span>
<span id="cb185-24"><a href="#cb185-24"></a> tmp,</span>
<span id="cb185-25"><a href="#cb185-25"></a> <span class="dt">ylab=</span><span class="st">&#39;Predicted oof probability&#39;</span>,</span>
<span id="cb185-26"><a href="#cb185-26"></a> <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;green&#39;</span>, <span class="st">&#39;red&#39;</span>),</span>
<span id="cb185-27"><a href="#cb185-27"></a> <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb185-28"><a href="#cb185-28"></a>)</span>
<span id="cb185-29"><a href="#cb185-29"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(tmp), <span class="dt">tick=</span>F, <span class="kw">names</span>(tmp))</span>
<span id="cb185-30"><a href="#cb185-30"></a></span>
<span id="cb185-31"><a href="#cb185-31"></a></span>
<span id="cb185-32"><a href="#cb185-32"></a><span class="co"># compare the two</span></span>
<span id="cb185-33"><a href="#cb185-33"></a><span class="kw">plot</span>(</span>
<span id="cb185-34"><a href="#cb185-34"></a> <span class="dt">x =</span> lassoAll_predResp_1se_vec, <span class="dt">xlab=</span><span class="st">&#39;1se model oof Prob&#39;</span>,</span>
<span id="cb185-35"><a href="#cb185-35"></a> <span class="dt">y =</span> lassoAll_predResp_min_vec, <span class="dt">ylab=</span><span class="st">&#39;min lambda model oof Prob&#39;</span>,</span>
<span id="cb185-36"><a href="#cb185-36"></a> <span class="dt">col =</span> <span class="kw">ifelse</span>(brcaRna_all_group_vec <span class="op">==</span><span class="st"> &#39;LumA&#39;</span>, <span class="st">&#39;red&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb185-37"><a href="#cb185-37"></a>)</span>
<span id="cb185-38"><a href="#cb185-38"></a> </span>
<span id="cb185-39"><a href="#cb185-39"></a><span class="co"># Add reference lines at 10% false positive</span></span>
<span id="cb185-40"><a href="#cb185-40"></a>thres_1se &lt;-<span class="st"> </span><span class="kw">quantile</span>(lassoAll_predResp_1se_vec[brcaRna_all_group_vec <span class="op">==</span><span class="st"> &#39;Other&#39;</span>], <span class="dt">prob=</span>.<span class="dv">9</span>)</span>
<span id="cb185-41"><a href="#cb185-41"></a>thres_min &lt;-<span class="st"> </span><span class="kw">quantile</span>(lassoAll_predResp_min_vec[brcaRna_all_group_vec <span class="op">==</span><span class="st"> &#39;Other&#39;</span>], <span class="dt">prob=</span>.<span class="dv">9</span>)</span>
<span id="cb185-42"><a href="#cb185-42"></a><span class="kw">abline</span>(<span class="dt">v =</span> thres_1se, <span class="dt">h =</span> thres_min, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brca-rnaseq-get-sample-pred-1.png" alt="Predicted probabilities - averaged over cv replicates" width="960" />
<p class="caption">
(#fig:brca-rnaseq-get-sample-pred)Predicted probabilities - averaged over cv replicates
</p>
</div>
<!-- THIS PARAGRAPH REFERRED TO THE FITTED PROBS; NOT THE OOF PRED PROBS
We see that the minimum lambda models provide a better fit to the data,
which is to be expected as the minimum lambda models have more estimated
parameters than the one standard error rule models.  
-->
<p>We see that there isn’t a big difference in out-of-fold predicted
probabilities between the one-standard-error rule and the minimum lamda models.
One way to quantify
the difference in classification errors is to classify samples
according to each vector of predicted probabilities, setting
the thresholds to achieve a fixed false positive rate, 10% say.
These thresholds are indicated by the grey lines in the scatter plot
on the right side of Figure @ref(fig:get-sample-pred).</p>
<!-- APPLIED TO THE FITTED VALUES
We note
that althouth predicted probability distributions are quite different
for the two models, the the class predictions at a 10% false discovery threshold
are largely in agreement.
-->
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="#cb186-1"></a><span class="co">### </span><span class="al">NOTE</span><span class="co"> THAT HERE MODEL PRODUCES PROB(Other)</span></span>
<span id="cb186-2"><a href="#cb186-2"></a>lassoAll_predClass_1se_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb186-3"><a href="#cb186-3"></a> lassoAll_predResp_1se_vec <span class="op">&gt;</span><span class="st"> </span>thres_1se, <span class="st">&#39;LumA&#39;</span>, <span class="st">&#39;Other&#39;</span>)</span>
<span id="cb186-4"><a href="#cb186-4"></a></span>
<span id="cb186-5"><a href="#cb186-5"></a>lassoAll_predClass_min_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb186-6"><a href="#cb186-6"></a> lassoAll_predResp_min_vec <span class="op">&gt;</span><span class="st"> </span>thres_min, <span class="st">&#39;LumA&#39;</span>, <span class="st">&#39;Other&#39;</span>)</span>
<span id="cb186-7"><a href="#cb186-7"></a></span>
<span id="cb186-8"><a href="#cb186-8"></a>tmp &lt;-<span class="st"> </span><span class="kw">cbind</span>(</span>
<span id="cb186-9"><a href="#cb186-9"></a> <span class="kw">table</span>(<span class="dt">truth=</span>brcaRna_all_group_vec, <span class="st">`</span><span class="dt">1se-pred</span><span class="st">`</span>=lassoAll_predClass_1se_vec),</span>
<span id="cb186-10"><a href="#cb186-10"></a> <span class="kw">table</span>(<span class="dt">truth=</span>brcaRna_all_group_vec, <span class="st">`</span><span class="dt">min-pred</span><span class="st">`</span>=lassoAll_predClass_min_vec)</span>
<span id="cb186-11"><a href="#cb186-11"></a>) </span>
<span id="cb186-12"><a href="#cb186-12"></a><span class="co"># HARD-WIRED Hack for printing!!!!</span></span>
<span id="cb186-13"><a href="#cb186-13"></a><span class="kw">colnames</span>(tmp) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;1se-LumA&#39;</span>, <span class="st">&#39;1se-Other&#39;</span>, <span class="st">&#39;min-LumA&#39;</span>, <span class="st">&#39;min-Other&#39;</span>)</span>
<span id="cb186-14"><a href="#cb186-14"></a></span>
<span id="cb186-15"><a href="#cb186-15"></a>knitr<span class="op">::</span><span class="kw">kable</span>(tmp,</span>
<span id="cb186-16"><a href="#cb186-16"></a>  <span class="dt">caption =</span> <span class="st">&quot;Classifications: rows are truth&quot;</span>,</span>
<span id="cb186-17"><a href="#cb186-17"></a>  <span class="dt">digits=</span><span class="dv">1</span>) <span class="op">%&gt;%</span></span>
<span id="cb186-18"><a href="#cb186-18"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:brca-rnaseq-get-sample-class)Classifications: rows are truth
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
1se-LumA
</th>
<th style="text-align:right;">
1se-Other
</th>
<th style="text-align:right;">
min-LumA
</th>
<th style="text-align:right;">
min-Other
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
LumA
</td>
<td style="text-align:right;">
1288
</td>
<td style="text-align:right;">
204
</td>
<td style="text-align:right;">
1304
</td>
<td style="text-align:right;">
188
</td>
</tr>
<tr>
<td style="text-align:left;">
Other
</td>
<td style="text-align:right;">
94
</td>
<td style="text-align:right;">
839
</td>
<td style="text-align:right;">
94
</td>
<td style="text-align:right;">
839
</td>
</tr>
</tbody>
</table>
<p>When we fix the false positive rate at 10% (ie. the other samples error rates are fixed),
the <code>1se</code> model makes 204 false negative calls whereas the
minimum lambda model makes 188. A difference of 1.1%</p>
<!-- APPLIED TO THE FITTED PROBABILITIES
We see that the min lambda model, makes no false negative calls at a 90% sensitivity
setting, and the sensitivity could be increased substantially at no false negative
cost.  This is definitely over-fitting the data set.  For the purpose
of computing sample consistency scores - what do these differences mean? 
-->
<div id="get-sample-consistency-scores-1" class="section level3 unnumbered" number="">
<h3 class="unnumbered" number="">Get sample consistency scores</h3>
<p>To compute consistency scores, we will use the out-of-fold predicted probabilities.</p>
<div class="sourceCode" id="cb187"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb187-1"><a href="#cb187-1"></a><span class="co"># get qual scores</span></span>
<span id="cb187-2"><a href="#cb187-2"></a></span>
<span id="cb187-3"><a href="#cb187-3"></a>y &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(brcaRna_all_group_vec <span class="op">==</span><span class="st"> &#39;LumA&#39;</span>)</span>
<span id="cb187-4"><a href="#cb187-4"></a><span class="co"># 1se</span></span>
<span id="cb187-5"><a href="#cb187-5"></a>p &lt;-<span class="st"> </span>lassoAll_predResp_1se_vec</span>
<span id="cb187-6"><a href="#cb187-6"></a>brcaRna_sample_1se_qual_vec &lt;-<span class="st"> </span>p<span class="op">^</span>y<span class="op">*</span>(<span class="dv">1</span><span class="op">-</span>p)<span class="op">^</span>(<span class="dv">1</span><span class="op">-</span>y)</span>
<span id="cb187-7"><a href="#cb187-7"></a></span>
<span id="cb187-8"><a href="#cb187-8"></a><span class="co"># min</span></span>
<span id="cb187-9"><a href="#cb187-9"></a>p &lt;-<span class="st"> </span>lassoAll_predResp_min_vec</span>
<span id="cb187-10"><a href="#cb187-10"></a>brcaRna_sample_min_qual_vec &lt;-<span class="st"> </span>p<span class="op">^</span>y<span class="op">*</span>(<span class="dv">1</span><span class="op">-</span>p)<span class="op">^</span>(<span class="dv">1</span><span class="op">-</span>y)</span></code></pre></div>
<p>We can examine consistency scores as a function of classification bin.</p>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="#cb188-1"></a>y &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(brcaRna_all_group_vec <span class="op">==</span><span class="st"> &#39;LumA&#39;</span>)</span>
<span id="cb188-2"><a href="#cb188-2"></a></span>
<span id="cb188-3"><a href="#cb188-3"></a><span class="co"># 1se</span></span>
<span id="cb188-4"><a href="#cb188-4"></a>lassoAll_1se_conf_vec &lt;-<span class="st"> </span><span class="kw">paste</span>(</span>
<span id="cb188-5"><a href="#cb188-5"></a> y, </span>
<span id="cb188-6"><a href="#cb188-6"></a> <span class="kw">as.numeric</span>(lassoAll_predClass_1se_vec<span class="op">==</span><span class="st">&#39;LumA&#39;</span>),</span>
<span id="cb188-7"><a href="#cb188-7"></a> <span class="dt">sep =</span> <span class="st">&#39;:&#39;</span></span>
<span id="cb188-8"><a href="#cb188-8"></a>)</span>
<span id="cb188-9"><a href="#cb188-9"></a></span>
<span id="cb188-10"><a href="#cb188-10"></a><span class="co"># min</span></span>
<span id="cb188-11"><a href="#cb188-11"></a>lassoAll_min_conf_vec &lt;-<span class="st"> </span><span class="kw">paste</span>(</span>
<span id="cb188-12"><a href="#cb188-12"></a> y, </span>
<span id="cb188-13"><a href="#cb188-13"></a> <span class="kw">as.numeric</span>(lassoAll_predClass_min_vec<span class="op">==</span><span class="st">&#39;LumA&#39;</span>),</span>
<span id="cb188-14"><a href="#cb188-14"></a> <span class="dt">sep =</span> <span class="st">&#39;:&#39;</span></span>
<span id="cb188-15"><a href="#cb188-15"></a>)</span>
<span id="cb188-16"><a href="#cb188-16"></a></span>
<span id="cb188-17"><a href="#cb188-17"></a></span>
<span id="cb188-18"><a href="#cb188-18"></a>tmp &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb188-19"><a href="#cb188-19"></a> <span class="kw">split</span>(brcaRna_sample_1se_qual_vec, lassoAll_1se_conf_vec), </span>
<span id="cb188-20"><a href="#cb188-20"></a> <span class="kw">split</span>(brcaRna_sample_min_qual_vec, lassoAll_min_conf_vec)</span>
<span id="cb188-21"><a href="#cb188-21"></a>)</span>
<span id="cb188-22"><a href="#cb188-22"></a></span>
<span id="cb188-23"><a href="#cb188-23"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">3</span>,<span class="dv">3</span>,<span class="dv">2</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb188-24"><a href="#cb188-24"></a>gplots<span class="op">::</span><span class="kw">boxplot2</span>(<span class="kw">split</span>(brcaRna_sample_1se_qual_vec, lassoAll_1se_conf_vec), </span>
<span id="cb188-25"><a href="#cb188-25"></a>  <span class="dt">outline=</span>F, <span class="dt">ylab =</span> <span class="st">&#39;&#39;</span>, </span>
<span id="cb188-26"><a href="#cb188-26"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;green&#39;</span>, <span class="st">&#39;green&#39;</span>, <span class="st">&#39;red&#39;</span>, <span class="st">&#39;red&#39;</span>),</span>
<span id="cb188-27"><a href="#cb188-27"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>))</span>
<span id="cb188-28"><a href="#cb188-28"></a><span class="kw">title</span>(<span class="st">&#39;1se Model&#39;</span>)</span>
<span id="cb188-29"><a href="#cb188-29"></a></span>
<span id="cb188-30"><a href="#cb188-30"></a>gplots<span class="op">::</span><span class="kw">boxplot2</span>(<span class="kw">split</span>(brcaRna_sample_min_qual_vec, lassoAll_min_conf_vec), </span>
<span id="cb188-31"><a href="#cb188-31"></a>  <span class="dt">outline=</span>F, <span class="dt">ylab =</span> <span class="st">&#39;&#39;</span>,</span>
<span id="cb188-32"><a href="#cb188-32"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;green&#39;</span>, <span class="st">&#39;green&#39;</span>, <span class="st">&#39;red&#39;</span>, <span class="st">&#39;red&#39;</span>),</span>
<span id="cb188-33"><a href="#cb188-33"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>))</span>
<span id="cb188-34"><a href="#cb188-34"></a><span class="kw">title</span>(<span class="st">&#39;min lambda Model&#39;</span>)</span>
<span id="cb188-35"><a href="#cb188-35"></a></span>
<span id="cb188-36"><a href="#cb188-36"></a></span>
<span id="cb188-37"><a href="#cb188-37"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.5</span>, <span class="st">&#39;Classification - Truth:Predicted&#39;</span>)</span>
<span id="cb188-38"><a href="#cb188-38"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.5</span>, <span class="st">&#39;Consistency Score&#39;</span>)</span>
<span id="cb188-39"><a href="#cb188-39"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.5</span>, <span class="st">&#39;Sample Consistency vs Classification Outcome&#39;</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brca-rnaseq-plot-qual-conf-1.png" alt="consistency scores by classification - Other=0, LumA=1" width="768" />
<p class="caption">
(#fig:brca-rnaseq-plot-qual-conf)consistency scores by classification - Other=0, LumA=1
</p>
</div>
<p>This figure shows that for false positive cases (0:1 or classifying a
other as an LumA case), the algorithm is <em>more certain</em> of its predicted
outcome than for the false negative cases (1:0 or classifying an LumA case as Other).
ie. the misclassified Other samples are quite similar to LumA sample, whereas there
is more ambiguity in the misclassified LumA samples.</p>
<p>We will use the minimum lambda model to provide
the fitted probabilities used to compute consistency scores,
but we could have used either one.</p>
<div class="sourceCode" id="cb189"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb189-1"><a href="#cb189-1"></a>brcaRna_sample_qual_vec &lt;-<span class="st"> </span>brcaRna_sample_min_qual_vec</span></code></pre></div>
</div>
</div>
<div id="selected-feature-list-stability-1" class="section level2" number="8.2">
<h2 number="8.2"><span class="header-section-number">8.2</span> Selected feature list stability</h2>
<p>Before moving on to the simulation, let’s examine gene selection stability on the
full data set. We have two sets of sellected features - one for the
one standard deviation rule model, and one for the mimimum lambda model.
We saw in Table @ref(tab:model-size-lassoAll) that the number of features
selected by the minimum lambda models had an IQR of
221.25-291,
while the one standard error rule models had an IQR of
118.5-162.</p>
<p>Let’s examine the stability of the gene lists across cv replicates.</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="#cb190-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb190-2"><a href="#cb190-2"></a></span>
<span id="cb190-3"><a href="#cb190-3"></a></span>
<span id="cb190-4"><a href="#cb190-4"></a><span class="co"># 1se</span></span>
<span id="cb190-5"><a href="#cb190-5"></a>lassoAll_coef_1se_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(brcaRna_cv_lassoAll_lst, <span class="cf">function</span>(cv_fit){</span>
<span id="cb190-6"><a href="#cb190-6"></a> cv_fit_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb190-7"><a href="#cb190-7"></a> cv_fit,</span>
<span id="cb190-8"><a href="#cb190-8"></a> <span class="dt">s =</span> <span class="st">&quot;lambda.1se&quot;</span></span>
<span id="cb190-9"><a href="#cb190-9"></a> )</span>
<span id="cb190-10"><a href="#cb190-10"></a> cv_fit_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][cv_fit_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>]]</span>
<span id="cb190-11"><a href="#cb190-11"></a> })</span>
<span id="cb190-12"><a href="#cb190-12"></a></span>
<span id="cb190-13"><a href="#cb190-13"></a><span class="co"># put into matrix</span></span>
<span id="cb190-14"><a href="#cb190-14"></a>lassoAll_coef_1se_all &lt;-<span class="st"> </span><span class="kw">Reduce</span>(union, lassoAll_coef_1se_lst)</span>
<span id="cb190-15"><a href="#cb190-15"></a>lassoAll_coef_1se_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(lassoAll_coef_1se_lst, </span>
<span id="cb190-16"><a href="#cb190-16"></a>  <span class="cf">function</span>(LL) <span class="kw">is.element</span>(lassoAll_coef_1se_all, LL)</span>
<span id="cb190-17"><a href="#cb190-17"></a>)</span>
<span id="cb190-18"><a href="#cb190-18"></a><span class="kw">rownames</span>(lassoAll_coef_1se_mtx) &lt;-<span class="st"> </span>lassoAll_coef_1se_all</span>
<span id="cb190-19"><a href="#cb190-19"></a></span>
<span id="cb190-20"><a href="#cb190-20"></a>genes_by_rep_1se_tbl &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="kw">rowSums</span>(lassoAll_coef_1se_mtx))</span>
<span id="cb190-21"><a href="#cb190-21"></a><span class="kw">barplot</span>(</span>
<span id="cb190-22"><a href="#cb190-22"></a> genes_by_rep_1se_tbl,</span>
<span id="cb190-23"><a href="#cb190-23"></a> <span class="dt">xlab=</span><span class="st">&#39;Number of Replicates&#39;</span>,</span>
<span id="cb190-24"><a href="#cb190-24"></a> <span class="dt">ylab=</span><span class="st">&#39;Number of features&#39;</span></span>
<span id="cb190-25"><a href="#cb190-25"></a></span>
<span id="cb190-26"><a href="#cb190-26"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brca-rnaseq-feature-list-1se-1.png" alt="Feature list stability for one standard error rule models" width="768" />
<p class="caption">
(#fig:brca-rnaseq-feature-list-1se)Feature list stability for one standard error rule models
</p>
</div>
<p>We see that 83 features are included in every
cv replicate. These make up between
51%
and
70%
(Q1 and Q3) of the cv replicate one standard error rule models feature lists.</p>
<div class="sourceCode" id="cb191"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb191-1"><a href="#cb191-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb191-2"><a href="#cb191-2"></a></span>
<span id="cb191-3"><a href="#cb191-3"></a></span>
<span id="cb191-4"><a href="#cb191-4"></a><span class="co"># min</span></span>
<span id="cb191-5"><a href="#cb191-5"></a>lassoAll_coef_min_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(brcaRna_cv_lassoAll_lst, <span class="cf">function</span>(cv_fit){</span>
<span id="cb191-6"><a href="#cb191-6"></a> cv_fit_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb191-7"><a href="#cb191-7"></a> cv_fit,</span>
<span id="cb191-8"><a href="#cb191-8"></a> <span class="dt">s =</span> <span class="st">&quot;lambda.min&quot;</span></span>
<span id="cb191-9"><a href="#cb191-9"></a> )</span>
<span id="cb191-10"><a href="#cb191-10"></a> cv_fit_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][cv_fit_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>]]</span>
<span id="cb191-11"><a href="#cb191-11"></a> })</span>
<span id="cb191-12"><a href="#cb191-12"></a></span>
<span id="cb191-13"><a href="#cb191-13"></a><span class="co"># put into matrix</span></span>
<span id="cb191-14"><a href="#cb191-14"></a>lassoAll_coef_min_all &lt;-<span class="st"> </span><span class="kw">Reduce</span>(union, lassoAll_coef_min_lst)</span>
<span id="cb191-15"><a href="#cb191-15"></a>lassoAll_coef_min_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(lassoAll_coef_min_lst, </span>
<span id="cb191-16"><a href="#cb191-16"></a>  <span class="cf">function</span>(LL) <span class="kw">is.element</span>(lassoAll_coef_min_all, LL)</span>
<span id="cb191-17"><a href="#cb191-17"></a>)</span>
<span id="cb191-18"><a href="#cb191-18"></a><span class="kw">rownames</span>(lassoAll_coef_min_mtx) &lt;-<span class="st"> </span>lassoAll_coef_min_all</span>
<span id="cb191-19"><a href="#cb191-19"></a></span>
<span id="cb191-20"><a href="#cb191-20"></a>genes_by_rep_min_tbl &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="kw">rowSums</span>(lassoAll_coef_min_mtx))</span>
<span id="cb191-21"><a href="#cb191-21"></a><span class="kw">barplot</span>(</span>
<span id="cb191-22"><a href="#cb191-22"></a> genes_by_rep_min_tbl,</span>
<span id="cb191-23"><a href="#cb191-23"></a> <span class="dt">xlab=</span><span class="st">&#39;Number of Replicates&#39;</span>,</span>
<span id="cb191-24"><a href="#cb191-24"></a> <span class="dt">ylab=</span><span class="st">&#39;Number of features&#39;</span></span>
<span id="cb191-25"><a href="#cb191-25"></a></span>
<span id="cb191-26"><a href="#cb191-26"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brca-rnaseq-feature-list-min-1.png" alt="Feature list stability for minimum lambda models" width="768" />
<p class="caption">
(#fig:brca-rnaseq-feature-list-min)Feature list stability for minimum lambda models
</p>
</div>
<p>We see that 110 features are included in every
cv replicate. These make up between
38%
and
50%
(Q1 and Q3) of the cv replicate min feature lists.
We will consider the genes that are selected in all cv replicates as a
gene signature produced by each model.</p>
<div class="sourceCode" id="cb192"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb192-1"><a href="#cb192-1"></a>lasso_gene_sign_1se_vec &lt;-<span class="st"> </span><span class="kw">rownames</span>(lassoAll_coef_1se_mtx)[<span class="kw">rowSums</span>(lassoAll_coef_1se_mtx)<span class="op">==</span><span class="dv">30</span>]</span>
<span id="cb192-2"><a href="#cb192-2"></a>lasso_gene_sign_min_vec &lt;-<span class="st"> </span><span class="kw">rownames</span>(lassoAll_coef_min_mtx)[<span class="kw">rowSums</span>(lassoAll_coef_min_mtx)<span class="op">==</span><span class="dv">30</span>]</span></code></pre></div>
<p>68 out of
83 of the genes in the 1se model gene signature
are contained in the min lambda model gene signature.</p>
</div>
<div id="simulation-design-1" class="section level2" number="8.3">
<h2 number="8.3"><span class="header-section-number">8.3</span> Simulation Design</h2>
<p>We are now ready to run the simulations.</p>
<div class="sourceCode" id="cb193"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb193-1"><a href="#cb193-1"></a> SIM &lt;-<span class="st"> </span><span class="dv">30</span></span>
<span id="cb193-2"><a href="#cb193-2"></a> SIZE &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">25</span>, <span class="dv">50</span>, <span class="dv">100</span>, <span class="dv">200</span>, <span class="dv">300</span>)</span>
<span id="cb193-3"><a href="#cb193-3"></a> CV_REP &lt;-<span class="st"> </span><span class="dv">30</span></span></code></pre></div>
<p>Simluation parameters:</p>
<ul>
<li><p>Number of simulations : SIM = 30</p></li>
<li><p>Sample sizes: SIZE = 25, 50, 100, 200, 300</p></li>
<li><p>Number of CV Replicates: CV_REP = 30</p></li>
</ul>
<p>We will repeat the simulation process SIM = 30 times.
For each simulation iteration, we will select 300 Other and
300 LumA samples at random. Models will be fitted and analyzed
to balanced subsets of SIZE = 25, 50, 100, 200, 300, in a <code>Matryoshka doll</code> manner to
emulate a typical sample accrual process.</p>
<p>For a given simulation and a given sample size, we will obtain
CV_REP = 30 cross-validated lasso fits. From these fits,
we can obtain 30 out-of-fold assessments of classification accuracy
to get a sense if its variability. From each cv replicate, we also obtain
an estimated model size and a set of selected features. We will want
to examine how these stabilize as the sample size increases.</p>
<p>Note that we limit the simulations to a maximum of sample size of 300 in
order to to have simulations with low overlap. With 300
randomly selected LumA samples, the expected overlap between two randomly
selected sets of LumA samples is 4%.
For Others the expected overlap is 10.3%.</p>
</div>
<div id="setup-simulation-1" class="section level2" number="8.4">
<h2 number="8.4"><span class="header-section-number">8.4</span> Setup simulation</h2>
<p>To setup the simulation, we only need two master tables: one for the selection of Others
and one for the selection of LumA samples.</p>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb194-1"><a href="#cb194-1"></a>brcaRna_all_other_vec &lt;-<span class="st"> </span><span class="kw">names</span>(brcaRna_all_group_vec[brcaRna_all_group_vec<span class="op">==</span><span class="st">&#39;Other&#39;</span>]) </span>
<span id="cb194-2"><a href="#cb194-2"></a>brcaRna_all_LumA_vec &lt;-<span class="st"> </span><span class="kw">names</span>(brcaRna_all_group_vec[brcaRna_all_group_vec<span class="op">==</span><span class="st">&#39;LumA&#39;</span>])  </span></code></pre></div>
<p>We have 933 other sample IDs stored in <code>brcaRna_all_other_vec</code>
and 1492 LumA sample IDs stored in <code>brcaRna_all_LumA_vec</code>.
To create a suite of random samples from these, we only need to randomly select indices from
each vector.</p>
<div class="sourceCode" id="cb195"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb195-1"><a href="#cb195-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb195-2"><a href="#cb195-2"></a></span>
<span id="cb195-3"><a href="#cb195-3"></a><span class="kw">set.seed</span>(<span class="dv">12379</span>)</span>
<span id="cb195-4"><a href="#cb195-4"></a></span>
<span id="cb195-5"><a href="#cb195-5"></a>brcaRna_sim_other_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb195-6"><a href="#cb195-6"></a> <span class="dv">1</span><span class="op">:</span>SIM, </span>
<span id="cb195-7"><a href="#cb195-7"></a> <span class="cf">function</span>(dummy) </span>
<span id="cb195-8"><a href="#cb195-8"></a>   <span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(brcaRna_all_other_vec), <span class="dt">size =</span>  <span class="kw">max</span>(SIZE))</span>
<span id="cb195-9"><a href="#cb195-9"></a>)</span>
<span id="cb195-10"><a href="#cb195-10"></a></span>
<span id="cb195-11"><a href="#cb195-11"></a></span>
<span id="cb195-12"><a href="#cb195-12"></a>brcaRna_sim_LumA_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb195-13"><a href="#cb195-13"></a> <span class="dv">1</span><span class="op">:</span>SIM, </span>
<span id="cb195-14"><a href="#cb195-14"></a> <span class="cf">function</span>(dummy) </span>
<span id="cb195-15"><a href="#cb195-15"></a>   <span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(brcaRna_all_LumA_vec), <span class="dt">size =</span>  <span class="kw">max</span>(SIZE))</span>
<span id="cb195-16"><a href="#cb195-16"></a>)</span></code></pre></div>
<p>Each simulation is specified by a given column of the simulation design matrices:
<code>brcaRna_sim_other_mtx</code> and <code>brcaRna_sim_LumA_mtx</code>, each with domensions 300, 30.
Within each simulation, we can run the analyses of size 25, 50, 100, 200, 300 by simply selecting
samples specified in the appropriate rows of each design matrix.</p>
<p>We can examine how much variability we have in the consistency scores of the selected samples.
Here we show results for the smalle sample sizes where variability will be the greatest.</p>
<div class="sourceCode" id="cb196"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb196-1"><a href="#cb196-1"></a>brcaRna_all_other_qual_vec &lt;-<span class="st"> </span>brcaRna_sample_qual_vec[brcaRna_all_other_vec]</span>
<span id="cb196-2"><a href="#cb196-2"></a>brcaRna_sim_other_qual_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb196-3"><a href="#cb196-3"></a>  <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(brcaRna_sim_other_mtx), </span>
<span id="cb196-4"><a href="#cb196-4"></a>  <span class="cf">function</span>(CC) brcaRna_all_other_qual_vec[brcaRna_sim_other_mtx[,CC]]</span>
<span id="cb196-5"><a href="#cb196-5"></a> )</span>
<span id="cb196-6"><a href="#cb196-6"></a></span>
<span id="cb196-7"><a href="#cb196-7"></a>brcaRna_all_LumA_qual_vec &lt;-<span class="st"> </span>brcaRna_sample_qual_vec[brcaRna_all_LumA_vec]</span>
<span id="cb196-8"><a href="#cb196-8"></a>brcaRna_sim_LumA_qual_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb196-9"><a href="#cb196-9"></a>  <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(brcaRna_sim_LumA_mtx),  </span>
<span id="cb196-10"><a href="#cb196-10"></a>  <span class="cf">function</span>(CC) brcaRna_all_LumA_qual_vec[brcaRna_sim_LumA_mtx[,CC]]</span>
<span id="cb196-11"><a href="#cb196-11"></a> )</span>
<span id="cb196-12"><a href="#cb196-12"></a></span>
<span id="cb196-13"><a href="#cb196-13"></a><span class="co"># ONLY LOOK AT SAMPLE SIZE == 50</span></span>
<span id="cb196-14"><a href="#cb196-14"></a>NN &lt;-<span class="st"> </span><span class="dv">50</span></span>
<span id="cb196-15"><a href="#cb196-15"></a></span>
<span id="cb196-16"><a href="#cb196-16"></a><span class="co"># PLOT</span></span>
<span id="cb196-17"><a href="#cb196-17"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">1</span>), <span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">5</span>,<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb196-18"><a href="#cb196-18"></a><span class="co"># other</span></span>
<span id="cb196-19"><a href="#cb196-19"></a><span class="kw">boxplot</span>(</span>
<span id="cb196-20"><a href="#cb196-20"></a>  brcaRna_sim_other_qual_mtx[<span class="dv">1</span><span class="op">:</span>NN,],</span>
<span id="cb196-21"><a href="#cb196-21"></a>  <span class="dt">outline =</span> T, </span>
<span id="cb196-22"><a href="#cb196-22"></a>  <span class="dt">ylab =</span> <span class="st">&#39;Consistency Score&#39;</span>,</span>
<span id="cb196-23"><a href="#cb196-23"></a>  <span class="dt">xaxt =</span> <span class="st">&#39;n&#39;</span></span>
<span id="cb196-24"><a href="#cb196-24"></a>)</span>
<span id="cb196-25"><a href="#cb196-25"></a><span class="kw">title</span>(<span class="st">&quot;Other sample consistency across simulations&quot;</span>)</span>
<span id="cb196-26"><a href="#cb196-26"></a></span>
<span id="cb196-27"><a href="#cb196-27"></a><span class="co"># LumA</span></span>
<span id="cb196-28"><a href="#cb196-28"></a><span class="kw">boxplot</span>(</span>
<span id="cb196-29"><a href="#cb196-29"></a>  brcaRna_sim_LumA_qual_mtx[<span class="dv">1</span><span class="op">:</span>NN,],</span>
<span id="cb196-30"><a href="#cb196-30"></a>  <span class="dt">outline =</span> T, </span>
<span id="cb196-31"><a href="#cb196-31"></a>  <span class="dt">ylab =</span> <span class="st">&#39;Consistency Score&#39;</span></span>
<span id="cb196-32"><a href="#cb196-32"></a>)</span>
<span id="cb196-33"><a href="#cb196-33"></a><span class="kw">title</span>(<span class="st">&quot;LumA sample consistency across simulations&quot;</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brca-rnaseq-look-sim-qual-0-50ONLY-1.png" alt="sample consistency by simulation run for size = 50 " width="960" />
<p class="caption">
(#fig:brca-rnaseq-look-sim-qual-0-50ONLY)sample consistency by simulation run for size = 50
</p>
</div>
<p>In this figure, we are summarizing the consistency measures of 50 samples per group acress
30 simulations, or random selections of other and LumA samples.
We see significant variability in sample consistency, especially in the Other cases.
This may lead an unwary observer to be overly optimistic, or overly pessimistic,
in the early accrual stages of a study.</p>
</div>
<div id="run-simulations-1" class="section level2" number="8.5">
<h2 number="8.5"><span class="header-section-number">8.5</span> Run simulations</h2>
<p>As these take a while to run,
we will save the results of each similation to a different
object and store to disk. These can be easily read from disk
when needed for analysis.</p>
<p>The simulation results are saved to the file system and
only needs to be run once. The simulation takes <span class="math inline">\(\approx\)</span> 14 minutes
per iteration, or <span class="math inline">\(\approx\)</span> 7 hours of run time on a laptop.
(Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Mojave 10.14.6)</p>
<!-- RUN ONCE - THEN GET FROM MEMORY -->
<div class="sourceCode" id="cb197"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb197-1"><a href="#cb197-1"></a>start_time &lt;-<span class="st"> </span><span class="kw">proc.time</span>()</span>
<span id="cb197-2"><a href="#cb197-2"></a></span>
<span id="cb197-3"><a href="#cb197-3"></a><span class="co"># Get stage from SIZE</span></span>
<span id="cb197-4"><a href="#cb197-4"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(brcaRna_sim_other_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>, SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb197-5"><a href="#cb197-5"></a></span>
<span id="cb197-6"><a href="#cb197-6"></a><span class="cf">for</span> (SIMno <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(brcaRna_sim_other_qual_mtx)) {</span>
<span id="cb197-7"><a href="#cb197-7"></a></span>
<span id="cb197-8"><a href="#cb197-8"></a>  <span class="co">#cat(&quot;Running simulation &quot;, SIMno, &quot;\n&quot;)</span></span>
<span id="cb197-9"><a href="#cb197-9"></a></span>
<span id="cb197-10"><a href="#cb197-10"></a>  brca_sim_cv_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(<span class="kw">levels</span>(stage_vec)), <span class="cf">function</span>(STGno) {</span>
<span id="cb197-11"><a href="#cb197-11"></a>    Stage_rows_vec &lt;-<span class="st"> </span><span class="kw">which</span>(stage_vec <span class="op">%in%</span><span class="st"> </span><span class="kw">levels</span>(stage_vec)[<span class="dv">1</span><span class="op">:</span>STGno])</span>
<span id="cb197-12"><a href="#cb197-12"></a>    <span class="co">#cat(&quot;Stage &quot;, STGno, &quot;- analyzing&quot;, length(Stage_rows_vec), &quot;paired samples.\n&quot;)</span></span>
<span id="cb197-13"><a href="#cb197-13"></a></span>
<span id="cb197-14"><a href="#cb197-14"></a>    sim_stage_samples_vec &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb197-15"><a href="#cb197-15"></a>      brcaRna_all_other_vec[brcaRna_sim_other_mtx[Stage_rows_vec, SIMno]],</span>
<span id="cb197-16"><a href="#cb197-16"></a>      brcaRna_all_LumA_vec[brcaRna_sim_LumA_mtx[Stage_rows_vec, SIMno]]</span>
<span id="cb197-17"><a href="#cb197-17"></a>    )</span>
<span id="cb197-18"><a href="#cb197-18"></a>    sim_stage_geneExpr_mtx &lt;-<span class="st"> </span>brcaRna_all_geneExpr_mtx[sim_stage_samples_vec, ]</span>
<span id="cb197-19"><a href="#cb197-19"></a>    sim_stage_group_vec &lt;-<span class="st"> </span>brcaRna_all_group_vec[sim_stage_samples_vec]</span>
<span id="cb197-20"><a href="#cb197-20"></a>    <span class="co">#print(table(sim_stage_group_vec))</span></span>
<span id="cb197-21"><a href="#cb197-21"></a></span>
<span id="cb197-22"><a href="#cb197-22"></a>    sim_stage_cv_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span>CV_REP, <span class="cf">function</span>(CV) {</span>
<span id="cb197-23"><a href="#cb197-23"></a>      cv_fit &lt;-<span class="st"> </span>glmnet<span class="op">::</span><span class="kw">cv.glmnet</span>(</span>
<span id="cb197-24"><a href="#cb197-24"></a>        <span class="dt">x =</span> sim_stage_geneExpr_mtx,</span>
<span id="cb197-25"><a href="#cb197-25"></a>        <span class="dt">y =</span> sim_stage_group_vec,</span>
<span id="cb197-26"><a href="#cb197-26"></a>        <span class="dt">alpha =</span> <span class="dv">1</span>,</span>
<span id="cb197-27"><a href="#cb197-27"></a>        <span class="dt">family =</span> <span class="st">&quot;binomial&quot;</span>,</span>
<span id="cb197-28"><a href="#cb197-28"></a>        <span class="dt">type.measure =</span> <span class="st">&quot;class&quot;</span>,</span>
<span id="cb197-29"><a href="#cb197-29"></a>        <span class="dt">keep =</span> T,</span>
<span id="cb197-30"><a href="#cb197-30"></a>        <span class="dt">nlambda =</span> <span class="dv">30</span></span>
<span id="cb197-31"><a href="#cb197-31"></a>      )</span>
<span id="cb197-32"><a href="#cb197-32"></a></span>
<span id="cb197-33"><a href="#cb197-33"></a>      <span class="co"># Extract 1se metrics from cv_fit</span></span>
<span id="cb197-34"><a href="#cb197-34"></a>      <span class="co">#######################</span></span>
<span id="cb197-35"><a href="#cb197-35"></a>      ndx_1se &lt;-<span class="st"> </span><span class="kw">which</span>(cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se)</span>
<span id="cb197-36"><a href="#cb197-36"></a></span>
<span id="cb197-37"><a href="#cb197-37"></a>      nzero_1se &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>nzero[ndx_1se]</span>
<span id="cb197-38"><a href="#cb197-38"></a>      cvm_1se &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>cvm[ndx_1se]</span>
<span id="cb197-39"><a href="#cb197-39"></a></span>
<span id="cb197-40"><a href="#cb197-40"></a>      <span class="co"># test error</span></span>
<span id="cb197-41"><a href="#cb197-41"></a>      sim_stage_test_samples_vec &lt;-<span class="st"> </span><span class="kw">setdiff</span>(<span class="kw">rownames</span>(brcaRna_all_geneExpr_mtx), sim_stage_samples_vec)</span>
<span id="cb197-42"><a href="#cb197-42"></a>      sim_stage_brcaRna_test_geneExpr_mtx &lt;-<span class="st"> </span>brcaRna_all_geneExpr_mtx[sim_stage_test_samples_vec,]</span>
<span id="cb197-43"><a href="#cb197-43"></a>      sim_stage_brcaRna_test_group_vec &lt;-<span class="st"> </span>brcaRna_all_group_vec[sim_stage_test_samples_vec]</span>
<span id="cb197-44"><a href="#cb197-44"></a></span>
<span id="cb197-45"><a href="#cb197-45"></a>      test_pred_1se_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb197-46"><a href="#cb197-46"></a>       cv_fit,</span>
<span id="cb197-47"><a href="#cb197-47"></a>       <span class="dt">newx=</span>sim_stage_brcaRna_test_geneExpr_mtx,</span>
<span id="cb197-48"><a href="#cb197-48"></a>       <span class="dt">s=</span><span class="st">&quot;lambda.1se&quot;</span>,</span>
<span id="cb197-49"><a href="#cb197-49"></a>       <span class="dt">type=</span><span class="st">&quot;class&quot;</span></span>
<span id="cb197-50"><a href="#cb197-50"></a>      )</span>
<span id="cb197-51"><a href="#cb197-51"></a>      test_1se_error &lt;-<span class="st"> </span><span class="kw">mean</span>(test_pred_1se_vec <span class="op">!=</span><span class="st"> </span>sim_stage_brcaRna_test_group_vec)</span>
<span id="cb197-52"><a href="#cb197-52"></a></span>
<span id="cb197-53"><a href="#cb197-53"></a>      <span class="co"># genes</span></span>
<span id="cb197-54"><a href="#cb197-54"></a>      coef_1se &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb197-55"><a href="#cb197-55"></a>        cv_fit,</span>
<span id="cb197-56"><a href="#cb197-56"></a>        <span class="dt">s =</span> <span class="st">&quot;lambda.1se&quot;</span></span>
<span id="cb197-57"><a href="#cb197-57"></a>      )</span>
<span id="cb197-58"><a href="#cb197-58"></a>      genes_1se &lt;-<span class="st"> </span>coef_1se<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][coef_1se<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>]]</span>
<span id="cb197-59"><a href="#cb197-59"></a></span>
<span id="cb197-60"><a href="#cb197-60"></a>      <span class="co"># Extract min metrics from cv_fit</span></span>
<span id="cb197-61"><a href="#cb197-61"></a>      <span class="co">#######################</span></span>
<span id="cb197-62"><a href="#cb197-62"></a>      ndx_min &lt;-<span class="st"> </span><span class="kw">which</span>(cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min)</span>
<span id="cb197-63"><a href="#cb197-63"></a></span>
<span id="cb197-64"><a href="#cb197-64"></a>      nzero_min &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>nzero[ndx_min]</span>
<span id="cb197-65"><a href="#cb197-65"></a>      cvm_min &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>cvm[ndx_min]</span>
<span id="cb197-66"><a href="#cb197-66"></a></span>
<span id="cb197-67"><a href="#cb197-67"></a>      <span class="co"># test error</span></span>
<span id="cb197-68"><a href="#cb197-68"></a>      sim_stage_test_samples_vec &lt;-<span class="st"> </span><span class="kw">setdiff</span>(<span class="kw">rownames</span>(brcaRna_all_geneExpr_mtx), sim_stage_samples_vec)</span>
<span id="cb197-69"><a href="#cb197-69"></a>      sim_stage_brcaRna_test_geneExpr_mtx &lt;-<span class="st"> </span>brcaRna_all_geneExpr_mtx[sim_stage_test_samples_vec,]</span>
<span id="cb197-70"><a href="#cb197-70"></a>      sim_stage_brcaRna_test_group_vec &lt;-<span class="st"> </span>brcaRna_all_group_vec[sim_stage_test_samples_vec]</span>
<span id="cb197-71"><a href="#cb197-71"></a></span>
<span id="cb197-72"><a href="#cb197-72"></a>      test_pred_min_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb197-73"><a href="#cb197-73"></a>       cv_fit,</span>
<span id="cb197-74"><a href="#cb197-74"></a>       <span class="dt">newx=</span>sim_stage_brcaRna_test_geneExpr_mtx,</span>
<span id="cb197-75"><a href="#cb197-75"></a>       <span class="dt">s=</span><span class="st">&quot;lambda.min&quot;</span>,</span>
<span id="cb197-76"><a href="#cb197-76"></a>       <span class="dt">type=</span><span class="st">&quot;class&quot;</span></span>
<span id="cb197-77"><a href="#cb197-77"></a>      )</span>
<span id="cb197-78"><a href="#cb197-78"></a>      test_min_error &lt;-<span class="st"> </span><span class="kw">mean</span>(test_pred_min_vec <span class="op">!=</span><span class="st"> </span>sim_stage_brcaRna_test_group_vec)</span>
<span id="cb197-79"><a href="#cb197-79"></a></span>
<span id="cb197-80"><a href="#cb197-80"></a>      <span class="co"># genes</span></span>
<span id="cb197-81"><a href="#cb197-81"></a>      coef_min &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb197-82"><a href="#cb197-82"></a>        cv_fit,</span>
<span id="cb197-83"><a href="#cb197-83"></a>        <span class="dt">s =</span> <span class="st">&quot;lambda.min&quot;</span></span>
<span id="cb197-84"><a href="#cb197-84"></a>      )</span>
<span id="cb197-85"><a href="#cb197-85"></a>      genes_min &lt;-<span class="st"> </span>coef_min<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][coef_min<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>]]</span>
<span id="cb197-86"><a href="#cb197-86"></a></span>
<span id="cb197-87"><a href="#cb197-87"></a>      <span class="co"># return cv_fit summary metrics</span></span>
<span id="cb197-88"><a href="#cb197-88"></a>      <span class="kw">list</span>(</span>
<span id="cb197-89"><a href="#cb197-89"></a>       <span class="dt">p_1se =</span> nzero_1se, </span>
<span id="cb197-90"><a href="#cb197-90"></a>       <span class="dt">p_min =</span> nzero_min, </span>
<span id="cb197-91"><a href="#cb197-91"></a>       <span class="dt">cv_1se =</span> cvm_1se, </span>
<span id="cb197-92"><a href="#cb197-92"></a>       <span class="dt">cv_min =</span> cvm_min, </span>
<span id="cb197-93"><a href="#cb197-93"></a>       <span class="dt">test_1se=</span>test_1se_error, </span>
<span id="cb197-94"><a href="#cb197-94"></a>       <span class="dt">test_min=</span>test_min_error, </span>
<span id="cb197-95"><a href="#cb197-95"></a>       <span class="dt">genes_1se =</span> genes_1se,</span>
<span id="cb197-96"><a href="#cb197-96"></a>       <span class="dt">genes_min =</span> genes_min)</span>
<span id="cb197-97"><a href="#cb197-97"></a>    })</span>
<span id="cb197-98"><a href="#cb197-98"></a>    sim_stage_cv_lst</span>
<span id="cb197-99"><a href="#cb197-99"></a>  })</span>
<span id="cb197-100"><a href="#cb197-100"></a></span>
<span id="cb197-101"><a href="#cb197-101"></a>  <span class="co"># save  brcaRna_sim_cv_lst</span></span>
<span id="cb197-102"><a href="#cb197-102"></a>  fName &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;brcaRna_sim_&quot;</span>, SIMno, <span class="st">&quot;_cv_lst&quot;</span>)</span>
<span id="cb197-103"><a href="#cb197-103"></a>  <span class="kw">assign</span>(fName, brca_sim_cv_lst)</span>
<span id="cb197-104"><a href="#cb197-104"></a>  <span class="kw">save</span>(<span class="dt">list =</span> fName, <span class="dt">file=</span><span class="kw">file.path</span>(<span class="st">&quot;RData&quot;</span>, fName))</span>
<span id="cb197-105"><a href="#cb197-105"></a></span>
<span id="cb197-106"><a href="#cb197-106"></a>}</span></code></pre></div>
<!-- DEBUG - rename after the fact -->
</div>
<div id="simulation-results-1" class="section level2" number="8.6">
<h2 number="8.6"><span class="header-section-number">8.6</span> Simulation results</h2>
<p>Recall the we have 30 simulations, or randomly selected sets of LumA and Other samples,
analyzed in inreasing sizes of 25, 50, 100, 200, 300, with
30 repeated cross-validated lasso fits:</p>
<ul>
<li><p>Sample sizes: SIZE = 25, 50, 100, 200, 300</p></li>
<li><p>Number of CV Replicates: CV_REP = 30</p></li>
</ul>
<p>First we extract simluation results and store into one big table (only showing the top of table shere):</p>
<div class="sourceCode" id="cb198"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb198-1"><a href="#cb198-1"></a>brcaRna_sim_files_vec &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="st">&#39;RData&#39;</span>, <span class="st">&#39;^brcaRna_sim_&#39;</span>)</span>
<span id="cb198-2"><a href="#cb198-2"></a></span>
<span id="cb198-3"><a href="#cb198-3"></a></span>
<span id="cb198-4"><a href="#cb198-4"></a><span class="co"># define extraction methods</span></span>
<span id="cb198-5"><a href="#cb198-5"></a></span>
<span id="cb198-6"><a href="#cb198-6"></a><span class="co"># Each sumulation is a list of cv results </span></span>
<span id="cb198-7"><a href="#cb198-7"></a><span class="co">## nested in a list of replicates</span></span>
<span id="cb198-8"><a href="#cb198-8"></a><span class="co">##############################################</span></span>
<span id="cb198-9"><a href="#cb198-9"></a></span>
<span id="cb198-10"><a href="#cb198-10"></a><span class="co"># cvList2frm_f makes a frame out of the inner list</span></span>
<span id="cb198-11"><a href="#cb198-11"></a>cvList2frm_f &lt;-<span class="st"> </span><span class="cf">function</span>(cv_lst) {</span>
<span id="cb198-12"><a href="#cb198-12"></a> frm1 &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">t</span>(<span class="kw">sapply</span>(cv_lst, <span class="cf">function</span>(x) x)))</span>
<span id="cb198-13"><a href="#cb198-13"></a> frm2 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb198-14"><a href="#cb198-14"></a>  <span class="kw">unlist</span>(frm1[[<span class="dv">1</span>]]), <span class="kw">unlist</span>(frm1[[<span class="dv">2</span>]]),</span>
<span id="cb198-15"><a href="#cb198-15"></a>  <span class="kw">unlist</span>(frm1[[<span class="dv">3</span>]]), <span class="kw">unlist</span>(frm1[[<span class="dv">4</span>]]),</span>
<span id="cb198-16"><a href="#cb198-16"></a>  <span class="kw">unlist</span>(frm1[[<span class="dv">5</span>]]), <span class="kw">unlist</span>(frm1[[<span class="dv">6</span>]]),</span>
<span id="cb198-17"><a href="#cb198-17"></a>  frm1[<span class="dv">7</span>], frm1[<span class="dv">8</span>])</span>
<span id="cb198-18"><a href="#cb198-18"></a>  <span class="kw">names</span>(frm2) &lt;-<span class="st"> </span><span class="kw">names</span>(frm1)</span>
<span id="cb198-19"><a href="#cb198-19"></a>  <span class="kw">data.frame</span>(<span class="dt">Rep=</span><span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(frm2), frm2)}</span>
<span id="cb198-20"><a href="#cb198-20"></a></span>
<span id="cb198-21"><a href="#cb198-21"></a><span class="co"># cv_lst_to_frm loop over replicates, concatenating the inner list frames</span></span>
<span id="cb198-22"><a href="#cb198-22"></a>cv_lst_to_frm &lt;-<span class="st"> </span><span class="cf">function</span>(brcaRna_sim_cv_lst) {</span>
<span id="cb198-23"><a href="#cb198-23"></a> <span class="kw">do.call</span>(<span class="st">&#39;rbind&#39;</span>, <span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(brcaRna_sim_cv_lst),</span>
<span id="cb198-24"><a href="#cb198-24"></a>  <span class="cf">function</span>(JJ) {</span>
<span id="cb198-25"><a href="#cb198-25"></a>    siz_frm &lt;-<span class="st"> </span><span class="kw">cvList2frm_f</span>(brcaRna_sim_cv_lst[[JJ]])</span>
<span id="cb198-26"><a href="#cb198-26"></a>    <span class="kw">data.frame</span>(<span class="dt">Size=</span>SIZE[JJ], siz_frm)</span>
<span id="cb198-27"><a href="#cb198-27"></a>  }))</span>
<span id="cb198-28"><a href="#cb198-28"></a>}</span>
<span id="cb198-29"><a href="#cb198-29"></a></span>
<span id="cb198-30"><a href="#cb198-30"></a><span class="co"># we loop across simulations to combine all results into one big table</span></span>
<span id="cb198-31"><a href="#cb198-31"></a>brcaRna_lasso_sim_results_frm &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="st">&#39;rbind&#39;</span>, <span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(brcaRna_sim_files_vec),</span>
<span id="cb198-32"><a href="#cb198-32"></a> <span class="cf">function</span>(SIM_NO) {</span>
<span id="cb198-33"><a href="#cb198-33"></a>  <span class="kw">load</span>(<span class="dt">file=</span><span class="kw">file.path</span>(<span class="st">&#39;RData&#39;</span>, brcaRna_sim_files_vec[SIM_NO]))</span>
<span id="cb198-34"><a href="#cb198-34"></a>  <span class="kw">assign</span>(<span class="st">&#39;brcaRna_sim_cv_lst&#39;</span>, <span class="kw">get</span>(brcaRna_sim_files_vec[SIM_NO]))</span>
<span id="cb198-35"><a href="#cb198-35"></a>  <span class="kw">rm</span>(<span class="dt">list=</span>brcaRna_sim_files_vec[SIM_NO])</span>
<span id="cb198-36"><a href="#cb198-36"></a>  </span>
<span id="cb198-37"><a href="#cb198-37"></a>  <span class="kw">data.frame</span>(<span class="dt">SimNo=</span><span class="kw">paste0</span>(<span class="st">&#39;Sim_&#39;</span>,<span class="kw">formatC</span>(SIM_NO,<span class="dt">width =</span> <span class="dv">2</span>,<span class="dt">flag =</span> <span class="dv">0</span>)), <span class="kw">cv_lst_to_frm</span>(brcaRna_sim_cv_lst))</span>
<span id="cb198-38"><a href="#cb198-38"></a>} </span>
<span id="cb198-39"><a href="#cb198-39"></a>)) </span></code></pre></div>
<!-- 
Have a table of simulation results - `brcaRna_lasso_sim_results_frm`:
-->
<div id="simulation-results---look-at-one-simulation-1" class="section level3" number="8.6.1">
<h3 number="8.6.1"><span class="header-section-number">8.6.1</span> Simulation Results - look at one simulation</h3>
<div id="model-accuracy-assessment-2" class="section level4" number="8.6.1.1">
<h4 number="8.6.1.1"><span class="header-section-number">8.6.1.1</span> Model Accuracy Assessment</h4>
<p>First examine results for one simulation run. In the figures that follow,
each boxplot summarized 30 repreated cross validation runs performed on a
fixed random selection of Other and Affeced samples. Recall that as
we move from 25 to 50, etc., the sample sets are growing to emulate an
accrual of samples over time.</p>
<div class="sourceCode" id="cb199"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb199-1"><a href="#cb199-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb199-2"><a href="#cb199-2"></a></span>
<span id="cb199-3"><a href="#cb199-3"></a><span class="co"># get full model cv error ref</span></span>
<span id="cb199-4"><a href="#cb199-4"></a>error_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb199-5"><a href="#cb199-5"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb199-6"><a href="#cb199-6"></a>error_1se_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(error_1se_vec, <span class="dt">prob=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>)        </span>
<span id="cb199-7"><a href="#cb199-7"></a></span>
<span id="cb199-8"><a href="#cb199-8"></a>error_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb199-9"><a href="#cb199-9"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb199-10"><a href="#cb199-10"></a>error_min_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(error_min_vec, <span class="dt">prob=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>)        </span>
<span id="cb199-11"><a href="#cb199-11"></a></span>
<span id="cb199-12"><a href="#cb199-12"></a><span class="co"># Utility objects</span></span>
<span id="cb199-13"><a href="#cb199-13"></a>SIZE0 &lt;-<span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_pad</span>(SIZE, <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>)</span>
<span id="cb199-14"><a href="#cb199-14"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(brcaRna_sim_other_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>,SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb199-15"><a href="#cb199-15"></a></span>
<span id="cb199-16"><a href="#cb199-16"></a></span>
<span id="cb199-17"><a href="#cb199-17"></a><span class="co">#SIM &lt;- &quot;Sim_01&quot;</span></span>
<span id="cb199-18"><a href="#cb199-18"></a></span>
<span id="cb199-19"><a href="#cb199-19"></a><span class="cf">for</span>(SIM <span class="cf">in</span> <span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>SimNo)[<span class="dv">1</span>]){</span>
<span id="cb199-20"><a href="#cb199-20"></a></span>
<span id="cb199-21"><a href="#cb199-21"></a>SimNum &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">sub</span>(<span class="st">&#39;Sim_&#39;</span>,<span class="st">&#39;&#39;</span>,SIM))</span>
<span id="cb199-22"><a href="#cb199-22"></a></span>
<span id="cb199-23"><a href="#cb199-23"></a>simNo_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(SimNo<span class="op">==</span>SIM)</span>
<span id="cb199-24"><a href="#cb199-24"></a></span>
<span id="cb199-25"><a href="#cb199-25"></a></span>
<span id="cb199-26"><a href="#cb199-26"></a><span class="co"># errors</span></span>
<span id="cb199-27"><a href="#cb199-27"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb199-28"><a href="#cb199-28"></a><span class="co">###################</span></span>
<span id="cb199-29"><a href="#cb199-29"></a><span class="co"># 1se</span></span>
<span id="cb199-30"><a href="#cb199-30"></a><span class="co">####################</span></span>
<span id="cb199-31"><a href="#cb199-31"></a>cv_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb199-32"><a href="#cb199-32"></a> <span class="kw">split</span>(cv_1se, Size))</span>
<span id="cb199-33"><a href="#cb199-33"></a><span class="kw">names</span>(cv_1se_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(cv_1se_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb199-34"><a href="#cb199-34"></a></span>
<span id="cb199-35"><a href="#cb199-35"></a>test_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb199-36"><a href="#cb199-36"></a> <span class="kw">split</span>(test_1se, Size))</span>
<span id="cb199-37"><a href="#cb199-37"></a><span class="kw">names</span>(test_1se_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(test_1se_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb199-38"><a href="#cb199-38"></a></span>
<span id="cb199-39"><a href="#cb199-39"></a>error_1se_lst &lt;-<span class="st"> </span><span class="kw">c</span>(cv_1se_lst, test_1se_lst)</span>
<span id="cb199-40"><a href="#cb199-40"></a>error_1se_lst &lt;-<span class="st"> </span>error_1se_lst[<span class="kw">order</span>(<span class="kw">names</span>(error_1se_lst))]</span>
<span id="cb199-41"><a href="#cb199-41"></a></span>
<span id="cb199-42"><a href="#cb199-42"></a><span class="kw">boxplot</span>(error_1se_lst, </span>
<span id="cb199-43"><a href="#cb199-43"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>), </span>
<span id="cb199-44"><a href="#cb199-44"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>), </span>
<span id="cb199-45"><a href="#cb199-45"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb199-46"><a href="#cb199-46"></a>)</span>
<span id="cb199-47"><a href="#cb199-47"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T,  <span class="st">&#39;Misclassification Error&#39;</span>)</span>
<span id="cb199-48"><a href="#cb199-48"></a></span>
<span id="cb199-49"><a href="#cb199-49"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb199-50"><a href="#cb199-50"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb199-51"><a href="#cb199-51"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_lst)), </span>
<span id="cb199-52"><a href="#cb199-52"></a>  SIZE0</span>
<span id="cb199-53"><a href="#cb199-53"></a> )</span>
<span id="cb199-54"><a href="#cb199-54"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb199-55"><a href="#cb199-55"></a><span class="kw">abline</span>(<span class="dt">h=</span> error_1se_q2, <span class="dt">col =</span> <span class="st">&#39;red&#39;</span>)</span>
<span id="cb199-56"><a href="#cb199-56"></a><span class="kw">legend</span>(<span class="st">&#39;topright&#39;</span>, </span>
<span id="cb199-57"><a href="#cb199-57"></a>   <span class="co">#title=&#39;1se errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb199-58"><a href="#cb199-58"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb199-59"><a href="#cb199-59"></a>   <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&#39;cv error&#39;</span>, <span class="st">&#39;test set&#39;</span>),</span>
<span id="cb199-60"><a href="#cb199-60"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb199-61"><a href="#cb199-61"></a> )</span>
<span id="cb199-62"><a href="#cb199-62"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;one se lambda models&#39;</span>))</span>
<span id="cb199-63"><a href="#cb199-63"></a></span>
<span id="cb199-64"><a href="#cb199-64"></a>SKIP  &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb199-65"><a href="#cb199-65"></a><span class="co"># Add qual annotation</span></span>
<span id="cb199-66"><a href="#cb199-66"></a>other_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(brcaRna_sim_other_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb199-67"><a href="#cb199-67"></a>LumA_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(brcaRna_sim_LumA_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb199-68"><a href="#cb199-68"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb199-69"><a href="#cb199-69"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb199-70"><a href="#cb199-70"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_lst)),</span>
<span id="cb199-71"><a href="#cb199-71"></a>  <span class="kw">round</span>(other_qual_vec, <span class="dv">2</span>)</span>
<span id="cb199-72"><a href="#cb199-72"></a> )</span>
<span id="cb199-73"><a href="#cb199-73"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb199-74"><a href="#cb199-74"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb199-75"><a href="#cb199-75"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_lst)),</span>
<span id="cb199-76"><a href="#cb199-76"></a>  <span class="kw">round</span>(LumA_qual_vec, <span class="dv">2</span>)</span>
<span id="cb199-77"><a href="#cb199-77"></a> )</span>
<span id="cb199-78"><a href="#cb199-78"></a>}<span class="co">#SKIP</span></span>
<span id="cb199-79"><a href="#cb199-79"></a></span>
<span id="cb199-80"><a href="#cb199-80"></a><span class="co"># min</span></span>
<span id="cb199-81"><a href="#cb199-81"></a><span class="co">####################</span></span>
<span id="cb199-82"><a href="#cb199-82"></a>cv_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb199-83"><a href="#cb199-83"></a> <span class="kw">split</span>(cv_min, Size))</span>
<span id="cb199-84"><a href="#cb199-84"></a><span class="kw">names</span>(cv_min_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(cv_min_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb199-85"><a href="#cb199-85"></a></span>
<span id="cb199-86"><a href="#cb199-86"></a>test_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb199-87"><a href="#cb199-87"></a> <span class="kw">split</span>(test_min, Size))</span>
<span id="cb199-88"><a href="#cb199-88"></a><span class="kw">names</span>(test_min_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(test_min_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb199-89"><a href="#cb199-89"></a></span>
<span id="cb199-90"><a href="#cb199-90"></a>error_min_lst &lt;-<span class="st"> </span><span class="kw">c</span>(cv_min_lst, test_min_lst)</span>
<span id="cb199-91"><a href="#cb199-91"></a>error_min_lst &lt;-<span class="st"> </span>error_min_lst[<span class="kw">order</span>(<span class="kw">names</span>(error_min_lst))]</span>
<span id="cb199-92"><a href="#cb199-92"></a></span>
<span id="cb199-93"><a href="#cb199-93"></a><span class="kw">boxplot</span>(error_min_lst, </span>
<span id="cb199-94"><a href="#cb199-94"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>), </span>
<span id="cb199-95"><a href="#cb199-95"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>),</span>
<span id="cb199-96"><a href="#cb199-96"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb199-97"><a href="#cb199-97"></a>)</span>
<span id="cb199-98"><a href="#cb199-98"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T,  <span class="st">&#39;Misclassification Error&#39;</span>)</span>
<span id="cb199-99"><a href="#cb199-99"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb199-100"><a href="#cb199-100"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb199-101"><a href="#cb199-101"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_lst)), </span>
<span id="cb199-102"><a href="#cb199-102"></a>  SIZE0</span>
<span id="cb199-103"><a href="#cb199-103"></a> )</span>
<span id="cb199-104"><a href="#cb199-104"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb199-105"><a href="#cb199-105"></a><span class="kw">abline</span>(<span class="dt">h=</span> error_min_q2, <span class="dt">col =</span> <span class="st">&#39;red&#39;</span>)</span>
<span id="cb199-106"><a href="#cb199-106"></a><span class="kw">legend</span>(<span class="st">&#39;topright&#39;</span>, </span>
<span id="cb199-107"><a href="#cb199-107"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb199-108"><a href="#cb199-108"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb199-109"><a href="#cb199-109"></a>   <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&#39;cv error&#39;</span>, <span class="st">&#39;test set&#39;</span>),</span>
<span id="cb199-110"><a href="#cb199-110"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb199-111"><a href="#cb199-111"></a> )</span>
<span id="cb199-112"><a href="#cb199-112"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;min lambda models&#39;</span>))</span>
<span id="cb199-113"><a href="#cb199-113"></a></span>
<span id="cb199-114"><a href="#cb199-114"></a>SKIP  &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb199-115"><a href="#cb199-115"></a><span class="co"># Add qual annotation</span></span>
<span id="cb199-116"><a href="#cb199-116"></a>other_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(brcaRna_sim_other_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb199-117"><a href="#cb199-117"></a>LumA_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(brcaRna_sim_LumA_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb199-118"><a href="#cb199-118"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb199-119"><a href="#cb199-119"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb199-120"><a href="#cb199-120"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_lst)),</span>
<span id="cb199-121"><a href="#cb199-121"></a>  <span class="kw">round</span>(other_qual_vec, <span class="dv">2</span>)</span>
<span id="cb199-122"><a href="#cb199-122"></a> )</span>
<span id="cb199-123"><a href="#cb199-123"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb199-124"><a href="#cb199-124"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb199-125"><a href="#cb199-125"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_lst)),</span>
<span id="cb199-126"><a href="#cb199-126"></a>  <span class="kw">round</span>(LumA_qual_vec, <span class="dv">2</span>)</span>
<span id="cb199-127"><a href="#cb199-127"></a> )</span>
<span id="cb199-128"><a href="#cb199-128"></a>}<span class="co">#SKIP</span></span>
<span id="cb199-129"><a href="#cb199-129"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, <span class="kw">paste</span>(<span class="st">&#39;Sim =&#39;</span>,  SIM))</span>
<span id="cb199-130"><a href="#cb199-130"></a></span>
<span id="cb199-131"><a href="#cb199-131"></a>} <span class="co"># for(SIM</span></span></code></pre></div>
<div class="figure">
<img src="Static/figures/brca-rnaseq-lasso-simRes-errors-bySim-1.png" alt="lasso Model Errors by Sample Size" width="960" />
<p class="caption">
(#fig:brca-rnaseq-lasso-simRes-errors-bySim)lasso Model Errors by Sample Size
</p>
</div>
<p>In this one simulation, we see:</p>
<ul>
<li><p>Model accuracy steadily increases with sample size</p></li>
<li><p>CV error rates tend to be pessimistic, expecially for the small sample sizes
and more so for the 1se models.</p></li>
<li><p>There isn’t much to chose from between the one standard error and the minimum lambda models.</p></li>
</ul>
</div>
<div id="feature-selection-2" class="section level4" number="8.6.1.2">
<h4 number="8.6.1.2"><span class="header-section-number">8.6.1.2</span> Feature Selection</h4>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb200-1"><a href="#cb200-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb200-2"><a href="#cb200-2"></a></span>
<span id="cb200-3"><a href="#cb200-3"></a><span class="co"># get full model nzero ref</span></span>
<span id="cb200-4"><a href="#cb200-4"></a>nzero_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb200-5"><a href="#cb200-5"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb200-6"><a href="#cb200-6"></a>nzero_1se_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(nzero_1se_vec, <span class="dt">prob=</span><span class="kw">c</span>(<span class="dv">2</span>)<span class="op">/</span><span class="dv">4</span>)</span>
<span id="cb200-7"><a href="#cb200-7"></a></span>
<span id="cb200-8"><a href="#cb200-8"></a>nzero_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb200-9"><a href="#cb200-9"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb200-10"><a href="#cb200-10"></a>nzero_min_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(nzero_min_vec, <span class="dt">prob=</span><span class="kw">c</span>(<span class="dv">2</span>)<span class="op">/</span><span class="dv">4</span>)</span>
<span id="cb200-11"><a href="#cb200-11"></a></span>
<span id="cb200-12"><a href="#cb200-12"></a><span class="co"># Utility objects</span></span>
<span id="cb200-13"><a href="#cb200-13"></a>SIZE0 &lt;-<span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_pad</span>(SIZE, <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>)</span>
<span id="cb200-14"><a href="#cb200-14"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(brcaRna_sim_other_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>,SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb200-15"><a href="#cb200-15"></a></span>
<span id="cb200-16"><a href="#cb200-16"></a></span>
<span id="cb200-17"><a href="#cb200-17"></a><span class="co">#SIM &lt;- &quot;Sim_01&quot;</span></span>
<span id="cb200-18"><a href="#cb200-18"></a></span>
<span id="cb200-19"><a href="#cb200-19"></a><span class="cf">for</span>(SIM <span class="cf">in</span> <span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>SimNo)[<span class="dv">1</span>]){</span>
<span id="cb200-20"><a href="#cb200-20"></a></span>
<span id="cb200-21"><a href="#cb200-21"></a>SimNum &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">sub</span>(<span class="st">&#39;Sim_&#39;</span>,<span class="st">&#39;&#39;</span>,SIM))</span>
<span id="cb200-22"><a href="#cb200-22"></a></span>
<span id="cb200-23"><a href="#cb200-23"></a>simNo_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(SimNo<span class="op">==</span>SIM)</span>
<span id="cb200-24"><a href="#cb200-24"></a></span>
<span id="cb200-25"><a href="#cb200-25"></a></span>
<span id="cb200-26"><a href="#cb200-26"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb200-27"><a href="#cb200-27"></a><span class="co">###################</span></span>
<span id="cb200-28"><a href="#cb200-28"></a><span class="co"># 1se</span></span>
<span id="cb200-29"><a href="#cb200-29"></a><span class="co">####################</span></span>
<span id="cb200-30"><a href="#cb200-30"></a><span class="co"># selected feature counts</span></span>
<span id="cb200-31"><a href="#cb200-31"></a>p_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb200-32"><a href="#cb200-32"></a> <span class="kw">split</span>(p_1se, Size))</span>
<span id="cb200-33"><a href="#cb200-33"></a><span class="kw">names</span>(p_1se_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(p_1se_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_p&#39;</span>)</span>
<span id="cb200-34"><a href="#cb200-34"></a></span>
<span id="cb200-35"><a href="#cb200-35"></a><span class="co"># get selected features that are part of lasso_gene_sign_1se_vec</span></span>
<span id="cb200-36"><a href="#cb200-36"></a><span class="co"># - the signature selected genes</span></span>
<span id="cb200-37"><a href="#cb200-37"></a>sign_genes_1se_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(simNo_results_frm), <span class="cf">function</span>(RR)</span>
<span id="cb200-38"><a href="#cb200-38"></a>    <span class="kw">intersect</span>(<span class="kw">unlist</span>(simNo_results_frm[RR, <span class="st">&#39;genes_1se&#39;</span>]), lasso_gene_sign_1se_vec))</span>
<span id="cb200-39"><a href="#cb200-39"></a></span>
<span id="cb200-40"><a href="#cb200-40"></a>sign_p_1se_lst &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="kw">sapply</span>(sign_genes_1se_lst, length), simNo_results_frm<span class="op">$</span>Size)</span>
<span id="cb200-41"><a href="#cb200-41"></a><span class="kw">names</span>(sign_p_1se_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(sign_p_1se_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_signP&#39;</span>)</span>
<span id="cb200-42"><a href="#cb200-42"></a></span>
<span id="cb200-43"><a href="#cb200-43"></a></span>
<span id="cb200-44"><a href="#cb200-44"></a>p_singP_1se_lst &lt;-<span class="st"> </span><span class="kw">c</span>(p_1se_lst, sign_p_1se_lst)</span>
<span id="cb200-45"><a href="#cb200-45"></a>p_singP_1se_lst &lt;-<span class="st"> </span>p_singP_1se_lst[<span class="kw">order</span>(<span class="kw">names</span>(p_singP_1se_lst))]</span>
<span id="cb200-46"><a href="#cb200-46"></a></span>
<span id="cb200-47"><a href="#cb200-47"></a><span class="kw">boxplot</span>(p_singP_1se_lst,</span>
<span id="cb200-48"><a href="#cb200-48"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb200-49"><a href="#cb200-49"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">200</span>),</span>
<span id="cb200-50"><a href="#cb200-50"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb200-51"><a href="#cb200-51"></a>)</span>
<span id="cb200-52"><a href="#cb200-52"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T,  <span class="st">&#39;number of selected features&#39;</span>)</span>
<span id="cb200-53"><a href="#cb200-53"></a></span>
<span id="cb200-54"><a href="#cb200-54"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb200-55"><a href="#cb200-55"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb200-56"><a href="#cb200-56"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_lst)),</span>
<span id="cb200-57"><a href="#cb200-57"></a>  SIZE0</span>
<span id="cb200-58"><a href="#cb200-58"></a> )</span>
<span id="cb200-59"><a href="#cb200-59"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb200-60"><a href="#cb200-60"></a><span class="co">#abline(h= nzero_1se_q2, col = &#39;red&#39;)</span></span>
<span id="cb200-61"><a href="#cb200-61"></a><span class="kw">legend</span>(<span class="st">&#39;topleft&#39;</span>,</span>
<span id="cb200-62"><a href="#cb200-62"></a>   <span class="co">#title=&#39;1se errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb200-63"><a href="#cb200-63"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>),</span>
<span id="cb200-64"><a href="#cb200-64"></a>   <span class="dt">legend=</span> <span class="kw">c</span>(<span class="st">&#39;selected genes&#39;</span>,<span class="st">&#39;signature genes&#39;</span>),</span>
<span id="cb200-65"><a href="#cb200-65"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb200-66"><a href="#cb200-66"></a> )</span>
<span id="cb200-67"><a href="#cb200-67"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;one se lambda models&#39;</span>) )</span>
<span id="cb200-68"><a href="#cb200-68"></a></span>
<span id="cb200-69"><a href="#cb200-69"></a>SKIP  &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb200-70"><a href="#cb200-70"></a><span class="co"># Add qual annotation</span></span>
<span id="cb200-71"><a href="#cb200-71"></a>other_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(brcaRna_sim_other_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb200-72"><a href="#cb200-72"></a>LumA_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(brcaRna_sim_LumA_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb200-73"><a href="#cb200-73"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb200-74"><a href="#cb200-74"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb200-75"><a href="#cb200-75"></a>  <span class="dt">at =</span>  <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_lst)),</span>
<span id="cb200-76"><a href="#cb200-76"></a>  <span class="kw">round</span>(other_qual_vec, <span class="dv">2</span>)</span>
<span id="cb200-77"><a href="#cb200-77"></a> )</span>
<span id="cb200-78"><a href="#cb200-78"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb200-79"><a href="#cb200-79"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb200-80"><a href="#cb200-80"></a>  <span class="dt">at =</span>  <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_lst)),</span>
<span id="cb200-81"><a href="#cb200-81"></a>  <span class="kw">round</span>(LumA_qual_vec, <span class="dv">2</span>)</span>
<span id="cb200-82"><a href="#cb200-82"></a> )</span>
<span id="cb200-83"><a href="#cb200-83"></a>}<span class="co">#SKIP</span></span>
<span id="cb200-84"><a href="#cb200-84"></a></span>
<span id="cb200-85"><a href="#cb200-85"></a><span class="co">###################</span></span>
<span id="cb200-86"><a href="#cb200-86"></a><span class="co"># min</span></span>
<span id="cb200-87"><a href="#cb200-87"></a><span class="co">####################</span></span>
<span id="cb200-88"><a href="#cb200-88"></a><span class="co"># selected feature counts</span></span>
<span id="cb200-89"><a href="#cb200-89"></a>p_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb200-90"><a href="#cb200-90"></a> <span class="kw">split</span>(p_min, Size))</span>
<span id="cb200-91"><a href="#cb200-91"></a><span class="kw">names</span>(p_min_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(p_min_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_p&#39;</span>)</span>
<span id="cb200-92"><a href="#cb200-92"></a></span>
<span id="cb200-93"><a href="#cb200-93"></a><span class="co"># get selected features that are part of lasso_gene_sign_min_vec</span></span>
<span id="cb200-94"><a href="#cb200-94"></a><span class="co"># - the signature selected genes</span></span>
<span id="cb200-95"><a href="#cb200-95"></a>sign_genes_min_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(simNo_results_frm), <span class="cf">function</span>(RR)</span>
<span id="cb200-96"><a href="#cb200-96"></a>    <span class="kw">intersect</span>(<span class="kw">unlist</span>(simNo_results_frm[RR, <span class="st">&#39;genes_min&#39;</span>]), lasso_gene_sign_min_vec))</span>
<span id="cb200-97"><a href="#cb200-97"></a></span>
<span id="cb200-98"><a href="#cb200-98"></a>sign_p_min_lst &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="kw">sapply</span>(sign_genes_min_lst, length), simNo_results_frm<span class="op">$</span>Size)</span>
<span id="cb200-99"><a href="#cb200-99"></a><span class="kw">names</span>(sign_p_min_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(sign_p_min_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_signP&#39;</span>)</span>
<span id="cb200-100"><a href="#cb200-100"></a></span>
<span id="cb200-101"><a href="#cb200-101"></a></span>
<span id="cb200-102"><a href="#cb200-102"></a>p_singP_min_lst &lt;-<span class="st"> </span><span class="kw">c</span>(p_min_lst, sign_p_min_lst)</span>
<span id="cb200-103"><a href="#cb200-103"></a>p_singP_min_lst &lt;-<span class="st"> </span>p_singP_min_lst[<span class="kw">order</span>(<span class="kw">names</span>(p_singP_min_lst))]</span>
<span id="cb200-104"><a href="#cb200-104"></a></span>
<span id="cb200-105"><a href="#cb200-105"></a><span class="kw">boxplot</span>(p_singP_min_lst,</span>
<span id="cb200-106"><a href="#cb200-106"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb200-107"><a href="#cb200-107"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">200</span>),</span>
<span id="cb200-108"><a href="#cb200-108"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb200-109"><a href="#cb200-109"></a>)</span>
<span id="cb200-110"><a href="#cb200-110"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb200-111"><a href="#cb200-111"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb200-112"><a href="#cb200-112"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_lst)),</span>
<span id="cb200-113"><a href="#cb200-113"></a>  SIZE0</span>
<span id="cb200-114"><a href="#cb200-114"></a> )</span>
<span id="cb200-115"><a href="#cb200-115"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb200-116"><a href="#cb200-116"></a><span class="co">#abline(h= nzero_min_q2, col = &#39;red&#39;)</span></span>
<span id="cb200-117"><a href="#cb200-117"></a><span class="kw">legend</span>(<span class="st">&#39;topleft&#39;</span>,</span>
<span id="cb200-118"><a href="#cb200-118"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb200-119"><a href="#cb200-119"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>),</span>
<span id="cb200-120"><a href="#cb200-120"></a>   <span class="dt">legend=</span> <span class="kw">c</span>(<span class="st">&#39;selected genes&#39;</span>,<span class="st">&#39;signature genes&#39;</span>),</span>
<span id="cb200-121"><a href="#cb200-121"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb200-122"><a href="#cb200-122"></a> )</span>
<span id="cb200-123"><a href="#cb200-123"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;min lambda models&#39;</span>))</span>
<span id="cb200-124"><a href="#cb200-124"></a></span>
<span id="cb200-125"><a href="#cb200-125"></a>SKIP  &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb200-126"><a href="#cb200-126"></a><span class="co"># Add qual annotation</span></span>
<span id="cb200-127"><a href="#cb200-127"></a>other_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(brcaRna_sim_other_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb200-128"><a href="#cb200-128"></a>LumA_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(brcaRna_sim_LumA_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb200-129"><a href="#cb200-129"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb200-130"><a href="#cb200-130"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb200-131"><a href="#cb200-131"></a>  <span class="dt">at =</span>  <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_lst)),</span>
<span id="cb200-132"><a href="#cb200-132"></a>  <span class="kw">round</span>(other_qual_vec, <span class="dv">2</span>)</span>
<span id="cb200-133"><a href="#cb200-133"></a> )</span>
<span id="cb200-134"><a href="#cb200-134"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb200-135"><a href="#cb200-135"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb200-136"><a href="#cb200-136"></a>  <span class="dt">at =</span>  <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_lst)),</span>
<span id="cb200-137"><a href="#cb200-137"></a>  <span class="kw">round</span>(LumA_qual_vec, <span class="dv">2</span>)</span>
<span id="cb200-138"><a href="#cb200-138"></a> )</span>
<span id="cb200-139"><a href="#cb200-139"></a>}<span class="co">#SKIP</span></span>
<span id="cb200-140"><a href="#cb200-140"></a></span>
<span id="cb200-141"><a href="#cb200-141"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, <span class="kw">paste</span>(<span class="st">&#39;Sim =&#39;</span>,  SIM))</span>
<span id="cb200-142"><a href="#cb200-142"></a></span>
<span id="cb200-143"><a href="#cb200-143"></a>} <span class="co"># for(SIM</span></span></code></pre></div>
<div class="figure">
<img src="Static/figures/brca-rnaseq-lasso-simRes-features-bySim-1.png" alt="lasso Models Selected Features by Sample Size" width="960" />
<p class="caption">
(#fig:brca-rnaseq-lasso-simRes-features-bySim)lasso Models Selected Features by Sample Size
</p>
</div>
<p>Compare with Figures
@ref(fig:hcc5hmC-glmnetSuite-lasso-simRes-errors-bySim)
and
@ref(fig:hcc5hmC-glmnetSuite-lasso-simRes-features-bySim)
for the HCC 5hmC data.</p>
<p><strong>Reader Note</strong></p>
<pre><code>IN the case of discriminating between LumA and other breast cancer subtypes,
we could use features in the LumA breast cancer subtype signature as
positive controls for feature selection.

To Do.</code></pre>
<!-- 
In this one simulation, we see:

* The selected number of features in the smaller sample size analyses
are low, with few feaures belonging to the core signature identified in the
full data set.  

* As the sample size increases the number of features selected in the minimum 
lambda model remains variable, but the number of core signature features
selected in the samples of sizes 200 and 300 is stable and between 40 and 50.  
-->
</div>
</div>
<div id="summarize-results-across-simulation-runs.-1" class="section level3" number="8.6.2">
<h3 number="8.6.2"><span class="header-section-number">8.6.2</span> Summarize results across simulation runs.</h3>
<div id="model-accuracy-assessment-3" class="section level4" number="8.6.2.1">
<h4 number="8.6.2.1"><span class="header-section-number">8.6.2.1</span> Model Accuracy Assessment</h4>
<p>Now look acoss all simulations. In the figures that follow, each boxplot
summarizes the results of 30 simulations. For a give sample size and a
given simulation, each data point is the median across 30 repeated cv runs.</p>
<div class="sourceCode" id="cb202"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb202-1"><a href="#cb202-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb202-2"><a href="#cb202-2"></a></span>
<span id="cb202-3"><a href="#cb202-3"></a><span class="co"># get full model cv error ref</span></span>
<span id="cb202-4"><a href="#cb202-4"></a>error_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb202-5"><a href="#cb202-5"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb202-6"><a href="#cb202-6"></a>error_1se_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(error_1se_vec, <span class="dt">prob=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>)        </span>
<span id="cb202-7"><a href="#cb202-7"></a></span>
<span id="cb202-8"><a href="#cb202-8"></a>error_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb202-9"><a href="#cb202-9"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb202-10"><a href="#cb202-10"></a>error_min_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(error_min_vec, <span class="dt">prob=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>)        </span>
<span id="cb202-11"><a href="#cb202-11"></a></span>
<span id="cb202-12"><a href="#cb202-12"></a><span class="co"># Utility objects</span></span>
<span id="cb202-13"><a href="#cb202-13"></a>SIZE0 &lt;-<span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_pad</span>(SIZE, <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>)</span>
<span id="cb202-14"><a href="#cb202-14"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(brcaRna_sim_other_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>,SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb202-15"><a href="#cb202-15"></a></span>
<span id="cb202-16"><a href="#cb202-16"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb202-17"><a href="#cb202-17"></a><span class="co"># 1se</span></span>
<span id="cb202-18"><a href="#cb202-18"></a><span class="co">#########################################</span></span>
<span id="cb202-19"><a href="#cb202-19"></a><span class="co">## cv</span></span>
<span id="cb202-20"><a href="#cb202-20"></a>cv_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb202-21"><a href="#cb202-21"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb202-22"><a href="#cb202-22"></a> sizeVal_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb202-23"><a href="#cb202-23"></a> sizeVal_cv_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(cv_1se, SimNo))</span>
<span id="cb202-24"><a href="#cb202-24"></a> <span class="kw">sapply</span>(sizeVal_cv_1se_lst, median)</span>
<span id="cb202-25"><a href="#cb202-25"></a>})</span>
<span id="cb202-26"><a href="#cb202-26"></a><span class="kw">names</span>(cv_1se_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb202-27"><a href="#cb202-27"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb202-28"><a href="#cb202-28"></a></span>
<span id="cb202-29"><a href="#cb202-29"></a><span class="co">## test</span></span>
<span id="cb202-30"><a href="#cb202-30"></a>test_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb202-31"><a href="#cb202-31"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb202-32"><a href="#cb202-32"></a> sizeVal_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb202-33"><a href="#cb202-33"></a> sizeVal_test_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(test_1se, SimNo))</span>
<span id="cb202-34"><a href="#cb202-34"></a> <span class="kw">sapply</span>(sizeVal_test_1se_lst, median)</span>
<span id="cb202-35"><a href="#cb202-35"></a>})</span>
<span id="cb202-36"><a href="#cb202-36"></a><span class="kw">names</span>(test_1se_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb202-37"><a href="#cb202-37"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_test&#39;</span>)</span>
<span id="cb202-38"><a href="#cb202-38"></a></span>
<span id="cb202-39"><a href="#cb202-39"></a></span>
<span id="cb202-40"><a href="#cb202-40"></a>error_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">c</span>(cv_1se_Bysize_lst, test_1se_Bysize_lst)</span>
<span id="cb202-41"><a href="#cb202-41"></a>error_1se_Bysize_lst &lt;-<span class="st"> </span>error_1se_Bysize_lst[<span class="kw">order</span>(<span class="kw">names</span>(error_1se_Bysize_lst))]</span>
<span id="cb202-42"><a href="#cb202-42"></a></span>
<span id="cb202-43"><a href="#cb202-43"></a><span class="kw">boxplot</span>(error_1se_Bysize_lst,</span>
<span id="cb202-44"><a href="#cb202-44"></a>  <span class="dt">col=</span><span class="dv">0</span>,</span>
<span id="cb202-45"><a href="#cb202-45"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb202-46"><a href="#cb202-46"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">.5</span>),</span>
<span id="cb202-47"><a href="#cb202-47"></a>  <span class="dt">outline=</span>F,</span>
<span id="cb202-48"><a href="#cb202-48"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb202-49"><a href="#cb202-49"></a>)</span>
<span id="cb202-50"><a href="#cb202-50"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T,  <span class="st">&#39;Misclassification Error&#39;</span>)</span>
<span id="cb202-51"><a href="#cb202-51"></a></span>
<span id="cb202-52"><a href="#cb202-52"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(error_1se_Bysize_lst))</span>
<span id="cb202-53"><a href="#cb202-53"></a><span class="kw">points</span>(</span>
<span id="cb202-54"><a href="#cb202-54"></a>   <span class="dt">x=</span><span class="kw">jitter</span>(<span class="kw">rep</span>(JJ, <span class="kw">length</span>(error_1se_Bysize_lst[[JJ]])), <span class="dt">amount=</span><span class="fl">0.25</span>), </span>
<span id="cb202-55"><a href="#cb202-55"></a>   <span class="dt">y=</span>error_1se_Bysize_lst[[JJ]], <span class="dt">cex=</span><span class="fl">0.5</span>,</span>
<span id="cb202-56"><a href="#cb202-56"></a>   <span class="dt">col=</span><span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">&#39;cv&#39;</span>, <span class="kw">names</span>(error_1se_Bysize_lst)[JJ]),<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb202-57"><a href="#cb202-57"></a>)</span>
<span id="cb202-58"><a href="#cb202-58"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb202-59"><a href="#cb202-59"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb202-60"><a href="#cb202-60"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_Bysize_lst)),</span>
<span id="cb202-61"><a href="#cb202-61"></a>  SIZE0</span>
<span id="cb202-62"><a href="#cb202-62"></a> )</span>
<span id="cb202-63"><a href="#cb202-63"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_Bysize_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb202-64"><a href="#cb202-64"></a><span class="kw">abline</span>(<span class="dt">h=</span> error_min_q2, <span class="dt">col =</span> <span class="st">&#39;red&#39;</span>)</span>
<span id="cb202-65"><a href="#cb202-65"></a><span class="kw">legend</span>(<span class="st">&#39;topright&#39;</span>,</span>
<span id="cb202-66"><a href="#cb202-66"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb202-67"><a href="#cb202-67"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb202-68"><a href="#cb202-68"></a>   <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&#39;cv error&#39;</span>, <span class="st">&#39;test set&#39;</span>),</span>
<span id="cb202-69"><a href="#cb202-69"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb202-70"><a href="#cb202-70"></a> )</span>
<span id="cb202-71"><a href="#cb202-71"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;one se lambda models&#39;</span>))</span>
<span id="cb202-72"><a href="#cb202-72"></a></span>
<span id="cb202-73"><a href="#cb202-73"></a></span>
<span id="cb202-74"><a href="#cb202-74"></a><span class="co"># min</span></span>
<span id="cb202-75"><a href="#cb202-75"></a><span class="co">#########################################</span></span>
<span id="cb202-76"><a href="#cb202-76"></a><span class="co">## cv</span></span>
<span id="cb202-77"><a href="#cb202-77"></a>cv_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb202-78"><a href="#cb202-78"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb202-79"><a href="#cb202-79"></a> sizeVal_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb202-80"><a href="#cb202-80"></a> sizeVal_cv_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(cv_min, SimNo))</span>
<span id="cb202-81"><a href="#cb202-81"></a> <span class="kw">sapply</span>(sizeVal_cv_min_lst, median)</span>
<span id="cb202-82"><a href="#cb202-82"></a>})</span>
<span id="cb202-83"><a href="#cb202-83"></a><span class="kw">names</span>(cv_min_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb202-84"><a href="#cb202-84"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb202-85"><a href="#cb202-85"></a></span>
<span id="cb202-86"><a href="#cb202-86"></a><span class="co">## test</span></span>
<span id="cb202-87"><a href="#cb202-87"></a>test_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb202-88"><a href="#cb202-88"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb202-89"><a href="#cb202-89"></a> sizeVal_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb202-90"><a href="#cb202-90"></a> sizeVal_test_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(test_min, SimNo))</span>
<span id="cb202-91"><a href="#cb202-91"></a> <span class="kw">sapply</span>(sizeVal_test_min_lst, median)</span>
<span id="cb202-92"><a href="#cb202-92"></a>})</span>
<span id="cb202-93"><a href="#cb202-93"></a><span class="kw">names</span>(test_min_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb202-94"><a href="#cb202-94"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_test&#39;</span>)</span>
<span id="cb202-95"><a href="#cb202-95"></a></span>
<span id="cb202-96"><a href="#cb202-96"></a></span>
<span id="cb202-97"><a href="#cb202-97"></a>error_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">c</span>(cv_min_Bysize_lst, test_min_Bysize_lst)</span>
<span id="cb202-98"><a href="#cb202-98"></a>error_min_Bysize_lst &lt;-<span class="st"> </span>error_min_Bysize_lst[<span class="kw">order</span>(<span class="kw">names</span>(error_min_Bysize_lst))]</span>
<span id="cb202-99"><a href="#cb202-99"></a></span>
<span id="cb202-100"><a href="#cb202-100"></a><span class="kw">boxplot</span>(error_min_Bysize_lst,</span>
<span id="cb202-101"><a href="#cb202-101"></a>  <span class="dt">col=</span><span class="dv">0</span>,</span>
<span id="cb202-102"><a href="#cb202-102"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb202-103"><a href="#cb202-103"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">.5</span>),</span>
<span id="cb202-104"><a href="#cb202-104"></a>  <span class="dt">outline=</span>F,</span>
<span id="cb202-105"><a href="#cb202-105"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb202-106"><a href="#cb202-106"></a>)</span>
<span id="cb202-107"><a href="#cb202-107"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(error_min_Bysize_lst))</span>
<span id="cb202-108"><a href="#cb202-108"></a><span class="kw">points</span>(</span>
<span id="cb202-109"><a href="#cb202-109"></a>   <span class="dt">x=</span><span class="kw">jitter</span>(<span class="kw">rep</span>(JJ, <span class="kw">length</span>(error_min_Bysize_lst[[JJ]])), <span class="dt">amount=</span><span class="fl">0.25</span>), </span>
<span id="cb202-110"><a href="#cb202-110"></a>   <span class="dt">y=</span>error_min_Bysize_lst[[JJ]], <span class="dt">cex=</span><span class="fl">0.5</span>,</span>
<span id="cb202-111"><a href="#cb202-111"></a>   <span class="dt">col=</span><span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">&#39;cv&#39;</span>, <span class="kw">names</span>(error_min_Bysize_lst)[JJ]),<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb202-112"><a href="#cb202-112"></a>)</span>
<span id="cb202-113"><a href="#cb202-113"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb202-114"><a href="#cb202-114"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb202-115"><a href="#cb202-115"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_Bysize_lst)),</span>
<span id="cb202-116"><a href="#cb202-116"></a>  SIZE0</span>
<span id="cb202-117"><a href="#cb202-117"></a> )</span>
<span id="cb202-118"><a href="#cb202-118"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_Bysize_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb202-119"><a href="#cb202-119"></a><span class="kw">abline</span>(<span class="dt">h=</span> error_min_q2, <span class="dt">col =</span> <span class="st">&#39;red&#39;</span>)</span>
<span id="cb202-120"><a href="#cb202-120"></a><span class="kw">legend</span>(<span class="st">&#39;topright&#39;</span>,</span>
<span id="cb202-121"><a href="#cb202-121"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb202-122"><a href="#cb202-122"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb202-123"><a href="#cb202-123"></a>   <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&#39;cv error&#39;</span>, <span class="st">&#39;test set&#39;</span>),</span>
<span id="cb202-124"><a href="#cb202-124"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb202-125"><a href="#cb202-125"></a> )</span>
<span id="cb202-126"><a href="#cb202-126"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;min lambda models&#39;</span>))</span>
<span id="cb202-127"><a href="#cb202-127"></a></span>
<span id="cb202-128"><a href="#cb202-128"></a></span>
<span id="cb202-129"><a href="#cb202-129"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, <span class="kw">paste</span>(<span class="st">&#39;lasso fit error rates summarized across simulations&#39;</span>))</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brca-rnaseq-lasso-simRes-errors-overSim-1.png" alt="lasso Model Errors by Sample Size" width="960" />
<p class="caption">
(#fig:brca-rnaseq-lasso-simRes-errors-overSim)lasso Model Errors by Sample Size
</p>
</div>
<p><br/></p>
<p>As in the case of the analysis of the HCC 5hmC data, we see that at lower
samples sizes the assessed performance is quite variable.
In this example cohorts of N=25 LumA and Other subtype samples
can yield classification error estimates ranging from less than %10
to over 50%.</p>
<p><br/></p>
<p>To appreciate how much variability can be encountered as samples are accrued over time
we need to look at a typical path the assessed model accuracy estimates might take.</p>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb203-1"><a href="#cb203-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb203-2"><a href="#cb203-2"></a></span>
<span id="cb203-3"><a href="#cb203-3"></a>error_1se_Bysize_mtx &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="st">&#39;cbind&#39;</span>, <span class="kw">lapply</span>(error_1se_Bysize_lst, <span class="cf">function</span>(LL) LL))</span>
<span id="cb203-4"><a href="#cb203-4"></a></span>
<span id="cb203-5"><a href="#cb203-5"></a>cv_error_1se_Bysize_mtx &lt;-<span class="st"> </span>error_1se_Bysize_mtx[,<span class="kw">grep</span>(<span class="st">&#39;_cv&#39;</span>, <span class="kw">colnames</span>(error_1se_Bysize_mtx))]</span>
<span id="cb203-6"><a href="#cb203-6"></a></span>
<span id="cb203-7"><a href="#cb203-7"></a><span class="kw">plot</span>(<span class="dt">x=</span><span class="kw">c</span>(<span class="dv">1</span>, <span class="kw">ncol</span>(cv_error_1se_Bysize_mtx)), <span class="dt">y=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.6</span>), </span>
<span id="cb203-8"><a href="#cb203-8"></a>  <span class="dt">xlab=</span><span class="st">&#39;sample size&#39;</span>, <span class="dt">ylab=</span><span class="st">&#39;Misclassification Error&#39;</span>, </span>
<span id="cb203-9"><a href="#cb203-9"></a>  <span class="dt">type=</span><span class="st">&#39;n&#39;</span>, <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>)</span>
<span id="cb203-10"><a href="#cb203-10"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(cv_error_1se_Bysize_mtx),</span>
<span id="cb203-11"><a href="#cb203-11"></a> <span class="dt">labels=</span><span class="kw">sub</span>(<span class="st">&#39;_cv&#39;</span>,<span class="st">&#39;&#39;</span>,<span class="kw">colnames</span>(cv_error_1se_Bysize_mtx)))</span>
<span id="cb203-12"><a href="#cb203-12"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">15</span>)</span>
<span id="cb203-13"><a href="#cb203-13"></a><span class="kw">lines</span>(<span class="dt">x=</span><span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(cv_error_1se_Bysize_mtx), <span class="dt">y=</span>cv_error_1se_Bysize_mtx[JJ,],</span>
<span id="cb203-14"><a href="#cb203-14"></a> <span class="dt">type=</span><span class="st">&#39;b&#39;</span>, <span class="dt">pch=</span>JJ, <span class="dt">col=</span>JJ)</span>
<span id="cb203-15"><a href="#cb203-15"></a><span class="kw">title</span>(<span class="st">&#39;Example Misclassification Error Paths&#39;</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brca-glmnetSuite-lasso-simRes-errorsPath-overSim-1.png" alt="lasso Model Error Paths" width="960" />
<p class="caption">
(#fig:brca-glmnetSuite-lasso-simRes-errorsPath-overSim)lasso Model Error Paths
</p>
</div>
<p>We see how erratic the assessed model accuracy can be when sample sizes are small,
and that it would be hard to guess the ultimate level of accuracy the
is achievable, or the number of samples required to get a reasonable
estimate of the achievable level of accuracy.</p>
<p><br/></p>
<!--
* For the smaller samples sizes, cv error rates for the minimum lambda models tend to be
optimistic.  

* For lower sample sizes, assess performance is quite variable.
-->
</div>
<div id="feature-selection-3" class="section level4" number="8.6.2.2">
<h4 number="8.6.2.2"><span class="header-section-number">8.6.2.2</span> Feature Selection</h4>
<div class="sourceCode" id="cb204"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb204-1"><a href="#cb204-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb204-2"><a href="#cb204-2"></a></span>
<span id="cb204-3"><a href="#cb204-3"></a><span class="co"># Utility objects</span></span>
<span id="cb204-4"><a href="#cb204-4"></a>SIZE0 &lt;-<span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_pad</span>(SIZE, <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>)</span>
<span id="cb204-5"><a href="#cb204-5"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(brcaRna_sim_other_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>,SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb204-6"><a href="#cb204-6"></a></span>
<span id="cb204-7"><a href="#cb204-7"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb204-8"><a href="#cb204-8"></a><span class="co"># 1se</span></span>
<span id="cb204-9"><a href="#cb204-9"></a><span class="co">#########################################</span></span>
<span id="cb204-10"><a href="#cb204-10"></a><span class="co"># selected features</span></span>
<span id="cb204-11"><a href="#cb204-11"></a>p_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb204-12"><a href="#cb204-12"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb204-13"><a href="#cb204-13"></a> sizeVal_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb204-14"><a href="#cb204-14"></a> sizeVal_p_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(p_1se, SimNo))</span>
<span id="cb204-15"><a href="#cb204-15"></a> <span class="kw">sapply</span>(sizeVal_p_1se_lst, median)</span>
<span id="cb204-16"><a href="#cb204-16"></a>})</span>
<span id="cb204-17"><a href="#cb204-17"></a><span class="kw">names</span>(p_1se_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb204-18"><a href="#cb204-18"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_p&#39;</span>)</span>
<span id="cb204-19"><a href="#cb204-19"></a></span>
<span id="cb204-20"><a href="#cb204-20"></a><span class="co"># selected signatue features</span></span>
<span id="cb204-21"><a href="#cb204-21"></a>sign_p_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb204-22"><a href="#cb204-22"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb204-23"><a href="#cb204-23"></a> sizeVal_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb204-24"><a href="#cb204-24"></a> </span>
<span id="cb204-25"><a href="#cb204-25"></a>  </span>
<span id="cb204-26"><a href="#cb204-26"></a> sizeVal_sign_genes_1se_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sizeVal_results_frm), <span class="cf">function</span>(RR)</span>
<span id="cb204-27"><a href="#cb204-27"></a>    <span class="kw">intersect</span>(<span class="kw">unlist</span>(sizeVal_results_frm[RR, <span class="st">&#39;genes_1se&#39;</span>]), lasso_gene_sign_1se_vec))</span>
<span id="cb204-28"><a href="#cb204-28"></a></span>
<span id="cb204-29"><a href="#cb204-29"></a> sizeVal_sign_p_1se_lst &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="kw">sapply</span>(sizeVal_sign_genes_1se_lst, length),</span>
<span id="cb204-30"><a href="#cb204-30"></a>    sizeVal_results_frm<span class="op">$</span>SimNo)</span>
<span id="cb204-31"><a href="#cb204-31"></a> </span>
<span id="cb204-32"><a href="#cb204-32"></a> <span class="kw">sapply</span>(sizeVal_sign_p_1se_lst, median)</span>
<span id="cb204-33"><a href="#cb204-33"></a>})</span>
<span id="cb204-34"><a href="#cb204-34"></a><span class="kw">names</span>(sign_p_1se_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb204-35"><a href="#cb204-35"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_signP&#39;</span>)</span>
<span id="cb204-36"><a href="#cb204-36"></a></span>
<span id="cb204-37"><a href="#cb204-37"></a></span>
<span id="cb204-38"><a href="#cb204-38"></a>p_singP_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">c</span>(p_1se_Bysize_lst, sign_p_1se_Bysize_lst)</span>
<span id="cb204-39"><a href="#cb204-39"></a>p_singP_1se_Bysize_lst &lt;-<span class="st"> </span>p_singP_1se_Bysize_lst[<span class="kw">order</span>(<span class="kw">names</span>(p_singP_1se_Bysize_lst))]</span>
<span id="cb204-40"><a href="#cb204-40"></a></span>
<span id="cb204-41"><a href="#cb204-41"></a><span class="kw">boxplot</span>(p_singP_1se_Bysize_lst,</span>
<span id="cb204-42"><a href="#cb204-42"></a>  <span class="dt">col=</span><span class="dv">0</span>,</span>
<span id="cb204-43"><a href="#cb204-43"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb204-44"><a href="#cb204-44"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">300</span>),</span>
<span id="cb204-45"><a href="#cb204-45"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb204-46"><a href="#cb204-46"></a>)</span>
<span id="cb204-47"><a href="#cb204-47"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T,  <span class="st">&#39;number of selected features&#39;</span>)</span>
<span id="cb204-48"><a href="#cb204-48"></a></span>
<span id="cb204-49"><a href="#cb204-49"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(p_singP_1se_Bysize_lst))</span>
<span id="cb204-50"><a href="#cb204-50"></a><span class="kw">points</span>(</span>
<span id="cb204-51"><a href="#cb204-51"></a>   <span class="dt">x=</span><span class="kw">jitter</span>(<span class="kw">rep</span>(JJ, <span class="kw">length</span>(p_singP_1se_Bysize_lst[[JJ]])), <span class="dt">amount=</span><span class="fl">0.25</span>),</span>
<span id="cb204-52"><a href="#cb204-52"></a>   <span class="dt">y=</span>p_singP_1se_Bysize_lst[[JJ]], <span class="dt">cex=</span><span class="fl">0.5</span>,</span>
<span id="cb204-53"><a href="#cb204-53"></a>   <span class="dt">col=</span><span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">&#39;_p&#39;</span>, <span class="kw">names</span>(p_singP_1se_Bysize_lst)[JJ]),<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb204-54"><a href="#cb204-54"></a>)</span>
<span id="cb204-55"><a href="#cb204-55"></a></span>
<span id="cb204-56"><a href="#cb204-56"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb204-57"><a href="#cb204-57"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb204-58"><a href="#cb204-58"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_Bysize_lst)),</span>
<span id="cb204-59"><a href="#cb204-59"></a>  SIZE0</span>
<span id="cb204-60"><a href="#cb204-60"></a> )</span>
<span id="cb204-61"><a href="#cb204-61"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_Bysize_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb204-62"><a href="#cb204-62"></a><span class="co">#abline(h= nzero_1se_q2, col = &#39;red&#39;)</span></span>
<span id="cb204-63"><a href="#cb204-63"></a><span class="kw">legend</span>(<span class="st">&#39;topleft&#39;</span>,</span>
<span id="cb204-64"><a href="#cb204-64"></a>   <span class="co">#title=&#39;1se errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb204-65"><a href="#cb204-65"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>),</span>
<span id="cb204-66"><a href="#cb204-66"></a>   <span class="dt">legend=</span> <span class="kw">c</span>(<span class="st">&#39;selected genes&#39;</span>,<span class="st">&#39;signature genes&#39;</span>),</span>
<span id="cb204-67"><a href="#cb204-67"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb204-68"><a href="#cb204-68"></a> )</span>
<span id="cb204-69"><a href="#cb204-69"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;one se lamdba models&#39;</span>))</span>
<span id="cb204-70"><a href="#cb204-70"></a></span>
<span id="cb204-71"><a href="#cb204-71"></a></span>
<span id="cb204-72"><a href="#cb204-72"></a><span class="co"># min</span></span>
<span id="cb204-73"><a href="#cb204-73"></a><span class="co">#########################################</span></span>
<span id="cb204-74"><a href="#cb204-74"></a><span class="co"># selected features</span></span>
<span id="cb204-75"><a href="#cb204-75"></a>p_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb204-76"><a href="#cb204-76"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb204-77"><a href="#cb204-77"></a> sizeVal_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb204-78"><a href="#cb204-78"></a> sizeVal_p_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(p_min, SimNo))</span>
<span id="cb204-79"><a href="#cb204-79"></a> <span class="kw">sapply</span>(sizeVal_p_min_lst, median)</span>
<span id="cb204-80"><a href="#cb204-80"></a>})</span>
<span id="cb204-81"><a href="#cb204-81"></a><span class="kw">names</span>(p_min_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb204-82"><a href="#cb204-82"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_p&#39;</span>)</span>
<span id="cb204-83"><a href="#cb204-83"></a></span>
<span id="cb204-84"><a href="#cb204-84"></a><span class="co"># selected signatue features</span></span>
<span id="cb204-85"><a href="#cb204-85"></a>sign_p_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb204-86"><a href="#cb204-86"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb204-87"><a href="#cb204-87"></a> sizeVal_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb204-88"><a href="#cb204-88"></a> </span>
<span id="cb204-89"><a href="#cb204-89"></a>  </span>
<span id="cb204-90"><a href="#cb204-90"></a> sizeVal_sign_genes_min_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sizeVal_results_frm), <span class="cf">function</span>(RR)</span>
<span id="cb204-91"><a href="#cb204-91"></a>    <span class="kw">intersect</span>(<span class="kw">unlist</span>(sizeVal_results_frm[RR, <span class="st">&#39;genes_min&#39;</span>]), lasso_gene_sign_min_vec))</span>
<span id="cb204-92"><a href="#cb204-92"></a></span>
<span id="cb204-93"><a href="#cb204-93"></a> sizeVal_sign_p_min_lst &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="kw">sapply</span>(sizeVal_sign_genes_min_lst, length),</span>
<span id="cb204-94"><a href="#cb204-94"></a>    sizeVal_results_frm<span class="op">$</span>SimNo)</span>
<span id="cb204-95"><a href="#cb204-95"></a> </span>
<span id="cb204-96"><a href="#cb204-96"></a> <span class="kw">sapply</span>(sizeVal_sign_p_min_lst, median)</span>
<span id="cb204-97"><a href="#cb204-97"></a>})</span>
<span id="cb204-98"><a href="#cb204-98"></a><span class="kw">names</span>(sign_p_min_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb204-99"><a href="#cb204-99"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_signP&#39;</span>)</span>
<span id="cb204-100"><a href="#cb204-100"></a></span>
<span id="cb204-101"><a href="#cb204-101"></a></span>
<span id="cb204-102"><a href="#cb204-102"></a>p_singP_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">c</span>(p_min_Bysize_lst, sign_p_min_Bysize_lst)</span>
<span id="cb204-103"><a href="#cb204-103"></a>p_singP_min_Bysize_lst &lt;-<span class="st"> </span>p_singP_min_Bysize_lst[<span class="kw">order</span>(<span class="kw">names</span>(p_singP_min_Bysize_lst))]</span>
<span id="cb204-104"><a href="#cb204-104"></a></span>
<span id="cb204-105"><a href="#cb204-105"></a><span class="kw">boxplot</span>(p_singP_min_Bysize_lst,</span>
<span id="cb204-106"><a href="#cb204-106"></a>  <span class="dt">col=</span><span class="dv">0</span>,</span>
<span id="cb204-107"><a href="#cb204-107"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb204-108"><a href="#cb204-108"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">300</span>),</span>
<span id="cb204-109"><a href="#cb204-109"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb204-110"><a href="#cb204-110"></a>)</span>
<span id="cb204-111"><a href="#cb204-111"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(p_singP_min_Bysize_lst))</span>
<span id="cb204-112"><a href="#cb204-112"></a><span class="kw">points</span>(</span>
<span id="cb204-113"><a href="#cb204-113"></a>   <span class="dt">x=</span><span class="kw">jitter</span>(<span class="kw">rep</span>(JJ, <span class="kw">length</span>(p_singP_min_Bysize_lst[[JJ]])), <span class="dt">amount=</span><span class="fl">0.25</span>),</span>
<span id="cb204-114"><a href="#cb204-114"></a>   <span class="dt">y=</span>p_singP_min_Bysize_lst[[JJ]], <span class="dt">cex=</span><span class="fl">0.5</span>,</span>
<span id="cb204-115"><a href="#cb204-115"></a>   <span class="dt">col=</span><span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">&#39;_p&#39;</span>, <span class="kw">names</span>(p_singP_min_Bysize_lst)[JJ]),<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb204-116"><a href="#cb204-116"></a>)</span>
<span id="cb204-117"><a href="#cb204-117"></a></span>
<span id="cb204-118"><a href="#cb204-118"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb204-119"><a href="#cb204-119"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb204-120"><a href="#cb204-120"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_Bysize_lst)),</span>
<span id="cb204-121"><a href="#cb204-121"></a>  SIZE0</span>
<span id="cb204-122"><a href="#cb204-122"></a> )</span>
<span id="cb204-123"><a href="#cb204-123"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_Bysize_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb204-124"><a href="#cb204-124"></a><span class="co">#abline(h= nzero_min_q2, col = &#39;red&#39;)</span></span>
<span id="cb204-125"><a href="#cb204-125"></a><span class="kw">legend</span>(<span class="st">&#39;topleft&#39;</span>,</span>
<span id="cb204-126"><a href="#cb204-126"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb204-127"><a href="#cb204-127"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>),</span>
<span id="cb204-128"><a href="#cb204-128"></a>   <span class="dt">legend=</span> <span class="kw">c</span>(<span class="st">&#39;selected genes&#39;</span>,<span class="st">&#39;signature genes&#39;</span>),</span>
<span id="cb204-129"><a href="#cb204-129"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb204-130"><a href="#cb204-130"></a> )</span>
<span id="cb204-131"><a href="#cb204-131"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;min lambda models&#39;</span>))</span>
<span id="cb204-132"><a href="#cb204-132"></a></span>
<span id="cb204-133"><a href="#cb204-133"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, <span class="kw">paste</span>(<span class="st">&#39;lasso fit feature selection summarized across simulations&#39;</span>))</span></code></pre></div>
<div class="figure">
<img src="Static/figures/brca-rnaseq-lasso-simRes-features-OverSim-1.png" alt="lasso Models Selected Features by Sample Size" width="960" />
<p class="caption">
(#fig:brca-rnaseq-lasso-simRes-features-OverSim)lasso Models Selected Features by Sample Size
</p>
</div>
<p>Compare with Figures
@ref(fig:hcc5hmC-glmnetSuite-lasso-simRes-errors-overSim)
and
@ref(fig:hcc5hmC-glmnetSuite-lasso-simRes-features-OverSim)
for the HCC 5hmC data.</p>
<!--

* The number of selected features incerase with sample size.  

*  The number of selected features is quite variable, even for larger sample sizes.

* The number of core signature features among selected features is stable for larger sample
sizes and represenst 30 to 40% of the selected features (for N=200 and 300 respectively).

-->
<!-- SKIP THIS
## Effect of sample consistency

**This is a bust ...**

To see the effect of sample consistency on classification results, we will
focus on the performance of the small sample fits where variability is the
greatest, and the context where investigators should be most aware of the
effect of sample selection over and beyond sample size concerns.
-->
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb205-1"><a href="#cb205-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb205-2"><a href="#cb205-2"></a></span>
<span id="cb205-3"><a href="#cb205-3"></a></span>
<span id="cb205-4"><a href="#cb205-4"></a>other_samp25_qual_vec &lt;-<span class="st"> </span><span class="kw">apply</span>(brcaRna_sim_other_qual_mtx[<span class="dv">1</span><span class="op">:</span><span class="dv">25</span>, ], <span class="dv">2</span>, median)</span>
<span id="cb205-5"><a href="#cb205-5"></a>LumA_samp25_qual_vec &lt;-<span class="st"> </span><span class="kw">apply</span>(brcaRna_sim_LumA_qual_mtx[<span class="dv">1</span><span class="op">:</span><span class="dv">25</span>, ], <span class="dv">2</span>, median)</span>
<span id="cb205-6"><a href="#cb205-6"></a></span>
<span id="cb205-7"><a href="#cb205-7"></a>samp25_qual_error_mtx &lt;-<span class="st"> </span><span class="kw">cbind</span>(</span>
<span id="cb205-8"><a href="#cb205-8"></a>   <span class="st">`</span><span class="dt">1se_cv</span><span class="st">`</span> =<span class="st"> </span>error_1se_Bysize_lst[[<span class="st">&#39;025_cv&#39;</span>]],</span>
<span id="cb205-9"><a href="#cb205-9"></a>   <span class="st">`</span><span class="dt">1se_test</span><span class="st">`</span> =<span class="st"> </span>error_1se_Bysize_lst[[<span class="st">&#39;025_test&#39;</span>]],</span>
<span id="cb205-10"><a href="#cb205-10"></a>   <span class="st">`</span><span class="dt">other_Q</span><span class="st">`</span> =<span class="st"> </span>other_samp25_qual_vec,</span>
<span id="cb205-11"><a href="#cb205-11"></a>   <span class="st">`</span><span class="dt">LumA_Q</span><span class="st">`</span> =<span class="st"> </span>LumA_samp25_qual_vec)</span>
<span id="cb205-12"><a href="#cb205-12"></a> </span>
<span id="cb205-13"><a href="#cb205-13"></a></span>
<span id="cb205-14"><a href="#cb205-14"></a><span class="co"># Correlation panel</span></span>
<span id="cb205-15"><a href="#cb205-15"></a>panel.cor &lt;-<span class="st"> </span><span class="cf">function</span>(x, y){</span>
<span id="cb205-16"><a href="#cb205-16"></a>    usr &lt;-<span class="st"> </span><span class="kw">par</span>(<span class="st">&quot;usr&quot;</span>); <span class="kw">on.exit</span>(<span class="kw">par</span>(usr))</span>
<span id="cb205-17"><a href="#cb205-17"></a>    <span class="kw">par</span>(<span class="dt">usr =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>))</span>
<span id="cb205-18"><a href="#cb205-18"></a>    r &lt;-<span class="st"> </span><span class="kw">round</span>(<span class="kw">cor</span>(x, y), <span class="dt">digits=</span><span class="dv">2</span>)</span>
<span id="cb205-19"><a href="#cb205-19"></a>    txt &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;R = &quot;</span>, r)</span>
<span id="cb205-20"><a href="#cb205-20"></a>    cex.cor &lt;-<span class="st"> </span><span class="fl">0.8</span><span class="op">/</span><span class="kw">strwidth</span>(txt)</span>
<span id="cb205-21"><a href="#cb205-21"></a>    <span class="kw">text</span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>, txt, <span class="dt">cex =</span> <span class="fl">1.5</span>) <span class="co">###cex.cor * r)</span></span>
<span id="cb205-22"><a href="#cb205-22"></a>}</span>
<span id="cb205-23"><a href="#cb205-23"></a></span>
<span id="cb205-24"><a href="#cb205-24"></a><span class="kw">pairs</span>(samp25_qual_error_mtx,</span>
<span id="cb205-25"><a href="#cb205-25"></a> <span class="dt">lower.panel =</span> panel.cor)</span></code></pre></div>
<!-- for the HCC data we had:
We see that although the sample consistency effect is not strong, the consistency of the
LumA sample cohorts does have a measurable impact on the 1se cv and test set error
rates (cor = -0.32 and -0.27, respectively).  
-->
<!--chapter:end:08-glmnnetSuiteBrCaRNASeq.Rmd-->
</div>
</div>
</div>
</div>
<div id="conclusions" class="section level1" number="9">
<h1 number="9"><span class="header-section-number">9</span> Conclusions</h1>
<p>We have found that …</p>
<p>Other questions …</p>
<!--chapter:end:A1-summary.Rmd-->
</div>
<div id="references" class="section level1 unnumbered" number="">
<h1 class="unnumbered" number="">References</h1>
<div id="refs" class="references">
<div id="ref-Gai:2019aa">
<p>1. Gai, W., and Sun, K. Epigenetic biomarkers in cell-free dna and applications in liquid biopsy. Genes <em>10</em>, 32. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/30634483">https://pubmed.ncbi.nlm.nih.gov/30634483</a>.</p>
</div>
<div id="ref-Cai:2019aa">
<p>2. Cai, J., Chen, L., Zhang, Z., Zhang, X., Lu, X., Liu, W., Shi, G., Ge, Y., Gao, P., and Yang, Y. <em>et al.</em> Genome-wide mapping of 5-hydroxymethylcytosines in circulating cell-free dna as a non-invasive approach for early detection of hepatocellular carcinoma. Gut, gutjnl–2019–318882. Available at: <a href="http://gut.bmj.com/content/early/2019/07/28/gutjnl-2019-318882.abstract">http://gut.bmj.com/content/early/2019/07/28/gutjnl-2019-318882.abstract</a>.</p>
</div>
<div id="ref-Li:2017aa">
<p>3. Li, W., Zhang, X., Lu, X., You, L., Song, Y., Luo, Z., Zhang, J., Nie, J., Zheng, W., and Xu, D. <em>et al.</em> DNA 5-hydroxymethylcytosines from cell-free circulating dna as diagnostic biomarkers for human cancers. bioRxiv, 163204. Available at: <a href="http://biorxiv.org/content/early/2017/07/13/163204.abstract">http://biorxiv.org/content/early/2017/07/13/163204.abstract</a>.</p>
</div>
<div id="ref-Song:2017aa">
<p>4. Song, C.-X., Yin, S., Ma, L., Wheeler, A., Chen, Y., Zhang, Y., Liu, B., Xiong, J., Zhang, W., and Hu, J. <em>et al.</em> (2017). 5-hydroxymethylcytosine signatures in cell-free dna provide information about tumor types and stages. Cell Research <em>27</em>, 1231–1242. Available at: <a href="https://doi.org/10.1038/cr.2017.106">https://doi.org/10.1038/cr.2017.106</a>.</p>
</div>
<div id="ref-Collin:2018aa">
<p>5. Collin, F., Ning, Y., Phillips, T., McCarthy, E., Scott, A., Ellison, C., Ku, C.-J., Guler, G.D., Chau, K., and Ashworth, A. <em>et al.</em> Detection of early stage pancreatic cancer using 5-hydroxymethylcytosine signatures in circulating cell free dna. bioRxiv, 422675. Available at: <a href="http://biorxiv.org/content/early/2018/09/26/422675.abstract">http://biorxiv.org/content/early/2018/09/26/422675.abstract</a>.</p>
</div>
<div id="ref-Friedman:2010aa">
<p>6. Friedman, J., Hastie, T., and Tibshirani, R. (2010). Regularization paths for generalized linear models via coordinate descent. J Stat Softw <em>33</em>, 1–22.</p>
</div>
<div id="ref-Hastie:2017aa">
<p>7. Hastie, T., and Tibshirani, R. (2017). Extended comparisons of best subset selection, forward stepwise selection, and the lasso. arXiv: Methodology.</p>
</div>
<div id="ref-Tibshirani:2012aa">
<p>8. Tibshirani, R., Bien, J., Friedman, J., Hastie, T., Simon, N., Taylor, J., and Tibshirani, R.J. Strong rules for discarding predictors in lasso-type problems. Journal of the Royal Statistical Society. Series B, Statistical methodology <em>74</em>, 245–266. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/25506256">https://pubmed.ncbi.nlm.nih.gov/25506256</a>.</p>
</div>
<div id="ref-Simon:2011aa">
<p>9. Simon, N., Friedman, J., Hastie, T., and Tibshirani, R. Regularization paths for cox’s proportional hazards model via coordinate descent. J Stat Softw <em>39</em>, 1–13.</p>
</div>
<div id="ref-Simon:2013aa">
<p>10. Simon, N., Friedman, J., and Hastie, T. (2013). A blockwise descent algorithm for group-penalized multiresponse and multinomial regression. arXiv preprint arXiv:1311.6529.</p>
</div>
<div id="ref-Xiang:2020aa">
<p>11. Xiang, G., Keller, C.A., Giardine, B., An, L., Li, Q., Zhang, Y., and Hardison, R.C. (2020). S3norm: Simultaneous normalization of sequencing depth and signal-to-noise ratio in epigenomic data. Nucleic Acids Research <em>48</em>, e43–e43. Available at: <a href="https://doi.org/10.1093/nar/gkaa105">https://doi.org/10.1093/nar/gkaa105</a>.</p>
</div>
<div id="ref-Lozoya:2018aa">
<p>12. Lozoya, O.A., Santos, J.H., and Woychik, R.P. A leveraged signal-to-noise ratio (lstnr) method to extract differentially expressed genes and multivariate patterns of expression from noisy and low-replication rnaseq data. Frontiers in genetics <em>9</em>, 176–176. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/29868123">https://pubmed.ncbi.nlm.nih.gov/29868123</a>.</p>
</div>
<div id="ref-Simonsen:2018aa">
<p>13. Simonsen, A.T., Hansen, M.C., Kjeldsen, E., Møller, P.L., Hindkjær, J.J., Hokland, P., and Aggerholm, A. (2018). Systematic evaluation of signal-to-noise ratio in variant detection from single cell genome multiple displacement amplification and exome sequencing. BMC Genomics <em>19</em>, 681. Available at: <a href="https://doi.org/10.1186/s12864-018-5063-5">https://doi.org/10.1186/s12864-018-5063-5</a>.</p>
</div>
<div id="ref-Rapaport:2013aa">
<p>14. Rapaport, F., Khanin, R., Liang, Y., Pirun, M., Krek, A., Zumbo, P., Mason, C.E., Socci, N.D., and Betel, D. (2013). Comprehensive evaluation of differential gene expression analysis methods for rna-seq data. Genome biology <em>14</em>, R95–R95. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/24020486">https://pubmed.ncbi.nlm.nih.gov/24020486</a>.</p>
</div>
<div id="ref-Bertsimas:2016aa">
<p>15. Bertsimas, D., King, A., and Mazumder, R. (2016). Best subset selection via a modern optimization lens. Ann. Statist. <em>44</em>, 813–852. Available at: <a href="https://projecteuclid.org:443/euclid.aos/1458245736">https://projecteuclid.org:443/euclid.aos/1458245736</a>.</p>
</div>
<div id="ref-Huang:2017aa">
<p>16. Huang, L.-H., Lin, P.-H., Tsai, K.-W., Wang, L.-J., Huang, Y.-H., Kuo, H.-C., and Li, S.-C. The effects of storage temperature and duration of blood samples on dna and rna qualities. PloS one <em>12</em>, e0184692–e0184692. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/28926588">https://pubmed.ncbi.nlm.nih.gov/28926588</a>.</p>
</div>
<div id="ref-Permenter:2015aa">
<p>17. Permenter, J., Ishwar, A., Rounsavall, A., Smith, M., Faske, J., Sailey, C.J., and Alfaro, M.P. (2015). Quantitative analysis of genomic dna degradation in whole blood under various storage conditions for molecular diagnostic testing. Molecular and Cellular Probes <em>29</em>, 449–453. Available at: <a href="http://www.sciencedirect.com/science/article/pii/S0890850815300207">http://www.sciencedirect.com/science/article/pii/S0890850815300207</a>.</p>
</div>
<div id="ref-Law:2018aa">
<p>18. Law, C., Alhamdoosh, M., Su, S., Dong, X., Tian, L., Smyth, G., and Ritchie, M. (2018). RNA-seq analysis is easy as 1-2-3 with limma, glimma and edgeR [version 3; peer review: 3 approved]. F1000Research <em>5</em>. Available at: <a href="https://dx.doi.org/10.12688%2Ff1000research.9005.3">https://dx.doi.org/10.12688%2Ff1000research.9005.3</a>.</p>
</div>
<div id="ref-Ritchie:2015aa">
<p>19. Ritchie, M.E., Phipson, B., Wu, D., Hu, Y., Law, C.W., Shi, W., and Smyth, G.K. Limma powers differential expression analyses for rna-sequencing and microarray studies. Nucleic acids research <em>43</em>, e47–e47. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/25605792">https://pubmed.ncbi.nlm.nih.gov/25605792</a>.</p>
</div>
<div id="ref-Peixoto:2015aa">
<p>20. Peixoto, L., Risso, D., Poplawski, S.G., Wimmer, M.E., Speed, T.P., Wood, M.A., and Abel, T. How data analysis affects power, reproducibility and biological insight of rna-seq studies in complex datasets. Nucleic acids research <em>43</em>, 7664–7674. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/26202970">https://pubmed.ncbi.nlm.nih.gov/26202970</a>.</p>
</div>
<div id="ref-Gandolfo:2018aa">
<p>21. Gandolfo, L.C., and Speed, T.P. RLE plots: Visualizing unwanted variation in high dimensional data. PloS one <em>13</em>, e0191629–e0191629. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/29401521">https://pubmed.ncbi.nlm.nih.gov/29401521</a>.</p>
</div>
<div id="ref-Risso:2014aa">
<p>22. Risso, D., Ngai, J., Speed, T.P., and Dudoit, S. (2014). Normalization of rna-seq data using factor analysis of control genes or samples. Nature Biotechnology <em>32</em>, 896–902. Available at: <a href="https://doi.org/10.1038/nbt.2931">https://doi.org/10.1038/nbt.2931</a>.</p>
</div>
<div id="ref-McCarthy:2009aa">
<p>23. McCarthy, D.J., and Smyth, G.K. Testing significance relative to a fold-change threshold is a treat. Bioinformatics <em>25</em>, 765–771. Available at: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2654802/">http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2654802/</a>.</p>
</div>
<div id="ref-Lockhart:2014aa">
<p>24. Lockhart, R., Taylor, J., Tibshirani, R.J., and Tibshirani, R. A significance test for the lasso. Ann Stat <em>42</em>, 413–468.</p>
</div>
<div id="ref-Wasserman:2014aa">
<p>25. Wasserman, L. (2014). Discussion: "A significance test for the lasso". Ann. Statist. <em>42</em>, 501–508. Available at: <a href="https://projecteuclid.org:443/euclid.aos/1400592166">https://projecteuclid.org:443/euclid.aos/1400592166</a>.</p>
</div>
<div id="ref-Engebretsen:2019aa">
<p>26. Engebretsen, S., and Bohlin, J. Statistical predictions with glmnet. Clinical epigenetics <em>11</em>, 123–123. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/31443682">https://pubmed.ncbi.nlm.nih.gov/31443682</a>.</p>
</div>
<div id="ref-Hofling:2008aa">
<p>27. Höfling, H., and Tibshirani, R. (2008). A study of pre-validation. <em>2</em>, 643–664. Available at: <a href="http://www.jstor.org/stable/30244221">http://www.jstor.org/stable/30244221</a>.</p>
</div>
</div>
</div>
<div id="appendix-1" class="section level1 unnumbered" number="">
<h1 class="unnumbered" number="">Appendix 1 - Sample size in elastic net fits to HCC 5hmC-Seq Data</h1>
<p>Repeat the analyses from Section @ref(model-suite), but using
the elastic net as classfication model.</p>
<!-- RUN ONCE -->
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb206-1"><a href="#cb206-1"></a><span class="kw">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb206-2"><a href="#cb206-2"></a></span>
<span id="cb206-3"><a href="#cb206-3"></a>start_time &lt;-<span class="st">  </span><span class="kw">proc.time</span>()</span>
<span id="cb206-4"><a href="#cb206-4"></a></span>
<span id="cb206-5"><a href="#cb206-5"></a>hcc5hmC_cv_enetAll_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">30</span>, <span class="cf">function</span>(REP) {</span>
<span id="cb206-6"><a href="#cb206-6"></a>glmnet<span class="op">::</span><span class="kw">cv.glmnet</span>(</span>
<span id="cb206-7"><a href="#cb206-7"></a> <span class="dt">x =</span> all_lcpm_mtx,</span>
<span id="cb206-8"><a href="#cb206-8"></a> <span class="dt">y =</span> <span class="kw">factor</span>(hcc5hmC_hcc5hmC,<span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&#39;Control&#39;</span>, <span class="st">&#39;HCC&#39;</span>)),</span>
<span id="cb206-9"><a href="#cb206-9"></a> <span class="dt">alpha =</span> <span class="fl">0.5</span>,</span>
<span id="cb206-10"><a href="#cb206-10"></a> <span class="dt">family =</span> <span class="st">&#39;binomial&#39;</span>,</span>
<span id="cb206-11"><a href="#cb206-11"></a> <span class="dt">type.measure  =</span>  <span class="st">&quot;class&quot;</span>,</span>
<span id="cb206-12"><a href="#cb206-12"></a> <span class="dt">keep =</span> T,</span>
<span id="cb206-13"><a href="#cb206-13"></a> <span class="dt">nlambda =</span> <span class="dv">100</span></span>
<span id="cb206-14"><a href="#cb206-14"></a>)</span>
<span id="cb206-15"><a href="#cb206-15"></a>}</span>
<span id="cb206-16"><a href="#cb206-16"></a>)</span>
<span id="cb206-17"><a href="#cb206-17"></a></span>
<span id="cb206-18"><a href="#cb206-18"></a><span class="kw">message</span>(<span class="st">&quot;enetAll time: &quot;</span>, <span class="kw">round</span>((<span class="kw">proc.time</span>() <span class="op">-</span><span class="st"> </span>start_time)[<span class="dv">3</span>],<span class="dv">2</span>),<span class="st">&quot;s&quot;</span>)</span></code></pre></div>
<!-- enet-fit-all takes a while - save results -->
<!-- DO THIS ONCE -->
<p>Examine the fits.</p>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb207-1"><a href="#cb207-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb207-2"><a href="#cb207-2"></a><span class="kw">plot</span>(</span>
<span id="cb207-3"><a href="#cb207-3"></a> <span class="kw">log</span>(hcc5hmC_cv_enetAll_lst[[<span class="dv">1</span>]]<span class="op">$</span>lambda),</span>
<span id="cb207-4"><a href="#cb207-4"></a> hcc5hmC_cv_enetAll_lst[[<span class="dv">1</span>]]<span class="op">$</span>cvm,</span>
<span id="cb207-5"><a href="#cb207-5"></a> <span class="dt">lwd=</span><span class="dv">2</span>,</span>
<span id="cb207-6"><a href="#cb207-6"></a> <span class="dt">xlab=</span><span class="st">&#39;log(Lambda)&#39;</span>, <span class="dt">ylab=</span><span class="st">&#39;CV Misclassification Error&#39;</span>, <span class="dt">type=</span><span class="st">&#39;l&#39;</span>, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">.5</span>)</span>
<span id="cb207-7"><a href="#cb207-7"></a>)</span>
<span id="cb207-8"><a href="#cb207-8"></a></span>
<span id="cb207-9"><a href="#cb207-9"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">2</span><span class="op">:</span><span class="kw">length</span>(hcc5hmC_cv_enetAll_lst))</span>
<span id="cb207-10"><a href="#cb207-10"></a> <span class="kw">lines</span>(</span>
<span id="cb207-11"><a href="#cb207-11"></a>  <span class="kw">log</span>(hcc5hmC_cv_enetAll_lst[[JJ]]<span class="op">$</span>lambda),</span>
<span id="cb207-12"><a href="#cb207-12"></a>  hcc5hmC_cv_enetAll_lst[[JJ]]<span class="op">$</span>cvm,</span>
<span id="cb207-13"><a href="#cb207-13"></a>  <span class="dt">lwd=</span><span class="dv">2</span></span>
<span id="cb207-14"><a href="#cb207-14"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuiteB-enet-plot-enetAll-1.png" alt="Repeated cv enet models fitted to all samples" width="576" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuiteB-enet-plot-enetAll)Repeated cv enet models fitted to all samples
</p>
</div>
<p>These cv curves are again remarkably consistent meaning that the determination of the size or sparsity
of the model through cross validation is fairly precise:</p>
<!-- DONT CACHE THIS ??? -->
<div class="sourceCode" id="cb208"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb208-1"><a href="#cb208-1"></a><span class="kw">library</span>(magrittr)</span>
<span id="cb208-2"><a href="#cb208-2"></a></span>
<span id="cb208-3"><a href="#cb208-3"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb208-4"><a href="#cb208-4"></a></span>
<span id="cb208-5"><a href="#cb208-5"></a><span class="co"># nzero</span></span>
<span id="cb208-6"><a href="#cb208-6"></a>nzero_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_enetAll_lst,</span>
<span id="cb208-7"><a href="#cb208-7"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb208-8"><a href="#cb208-8"></a></span>
<span id="cb208-9"><a href="#cb208-9"></a>nzero_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_enetAll_lst,</span>
<span id="cb208-10"><a href="#cb208-10"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb208-11"><a href="#cb208-11"></a></span>
<span id="cb208-12"><a href="#cb208-12"></a><span class="kw">boxplot</span>(<span class="kw">list</span>(<span class="st">`</span><span class="dt">1se</span><span class="st">`</span>=nzero_1se_vec, <span class="dt">min =</span> nzero_min_vec), <span class="dt">ylab=</span><span class="st">&quot;Full Model cv Summary&quot;</span>)</span>
<span id="cb208-13"><a href="#cb208-13"></a></span>
<span id="cb208-14"><a href="#cb208-14"></a><span class="co"># error</span></span>
<span id="cb208-15"><a href="#cb208-15"></a>error_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_enetAll_lst,</span>
<span id="cb208-16"><a href="#cb208-16"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb208-17"><a href="#cb208-17"></a></span>
<span id="cb208-18"><a href="#cb208-18"></a>error_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_enetAll_lst,</span>
<span id="cb208-19"><a href="#cb208-19"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb208-20"><a href="#cb208-20"></a></span>
<span id="cb208-21"><a href="#cb208-21"></a><span class="kw">boxplot</span>(</span>
<span id="cb208-22"><a href="#cb208-22"></a> <span class="kw">list</span>(<span class="st">`</span><span class="dt">1se</span><span class="st">`</span>=error_1se_vec, <span class="dt">min =</span> error_min_vec), </span>
<span id="cb208-23"><a href="#cb208-23"></a> <span class="dt">ylab=</span>hcc5hmC_cv_enetAll_lst[[<span class="dv">1</span>]]<span class="op">$</span>name,</span>
<span id="cb208-24"><a href="#cb208-24"></a> <span class="dt">ylim=</span><span class="kw">c</span>(<span class="fl">0.06</span>, <span class="fl">.10</span>)</span>
<span id="cb208-25"><a href="#cb208-25"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuiteB-model-size-enetAll-1.png" alt="Feature selection and estimated error by repeated cv enet models" width="768" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuiteB-model-size-enetAll)Feature selection and estimated error by repeated cv enet models
</p>
</div>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb209-1"><a href="#cb209-1"></a><span class="co"># tabular format</span></span>
<span id="cb209-2"><a href="#cb209-2"></a>tmp &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">rbind</span>(</span>
<span id="cb209-3"><a href="#cb209-3"></a> <span class="st">`</span><span class="dt">features_1se</span><span class="st">`</span> =<span class="st"> </span><span class="kw">summary</span>(nzero_1se_vec),</span>
<span id="cb209-4"><a href="#cb209-4"></a> <span class="dt">features_min =</span> <span class="kw">summary</span>(nzero_min_vec),</span>
<span id="cb209-5"><a href="#cb209-5"></a> <span class="st">`</span><span class="dt">features:min-1se</span><span class="st">`</span> =<span class="st"> </span><span class="kw">summary</span>(nzero_min_vec <span class="op">-</span><span class="st"> </span>nzero_1se_vec),</span>
<span id="cb209-6"><a href="#cb209-6"></a> <span class="st">`</span><span class="dt">cv_error_1se</span><span class="st">`</span> =<span class="st"> </span><span class="kw">summary</span>(<span class="dv">100</span><span class="op">*</span>error_1se_vec),</span>
<span id="cb209-7"><a href="#cb209-7"></a> <span class="dt">cv_error_min =</span> <span class="kw">summary</span>(<span class="dv">100</span><span class="op">*</span>error_min_vec),</span>
<span id="cb209-8"><a href="#cb209-8"></a> <span class="st">`</span><span class="dt">cv_error:1se-min</span><span class="st">`</span> =<span class="st"> </span><span class="kw">summary</span>(<span class="dv">100</span><span class="op">*</span>(error_1se_vec<span class="op">-</span>error_min_vec))</span>
<span id="cb209-9"><a href="#cb209-9"></a>))</span>
<span id="cb209-10"><a href="#cb209-10"></a></span>
<span id="cb209-11"><a href="#cb209-11"></a>knitr<span class="op">::</span><span class="kw">kable</span>(tmp <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(<span class="op">-</span>Mean),</span>
<span id="cb209-12"><a href="#cb209-12"></a>  <span class="dt">caption =</span> <span class="st">&quot;Number of selected features&quot;</span>,</span>
<span id="cb209-13"><a href="#cb209-13"></a>  <span class="dt">digits=</span><span class="dv">1</span>) <span class="op">%&gt;%</span></span>
<span id="cb209-14"><a href="#cb209-14"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-glmnetSuiteB-model-size-enetAll)Number of selected features
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Min.
</th>
<th style="text-align:right;">
X1st.Qu.
</th>
<th style="text-align:right;">
Median
</th>
<th style="text-align:right;">
X3rd.Qu.
</th>
<th style="text-align:right;">
Max.
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
features_1se
</td>
<td style="text-align:right;">
94.0
</td>
<td style="text-align:right;">
174.0
</td>
<td style="text-align:right;">
183.0
</td>
<td style="text-align:right;">
215.0
</td>
<td style="text-align:right;">
406.0
</td>
</tr>
<tr>
<td style="text-align:left;">
features_min
</td>
<td style="text-align:right;">
183.0
</td>
<td style="text-align:right;">
271.5
</td>
<td style="text-align:right;">
357.5
</td>
<td style="text-align:right;">
457.2
</td>
<td style="text-align:right;">
501.0
</td>
</tr>
<tr>
<td style="text-align:left;">
features:min-1se
</td>
<td style="text-align:right;">
23.0
</td>
<td style="text-align:right;">
81.5
</td>
<td style="text-align:right;">
168.5
</td>
<td style="text-align:right;">
237.5
</td>
<td style="text-align:right;">
331.0
</td>
</tr>
<tr>
<td style="text-align:left;">
cv_error_1se
</td>
<td style="text-align:right;">
7.1
</td>
<td style="text-align:right;">
7.4
</td>
<td style="text-align:right;">
7.6
</td>
<td style="text-align:right;">
7.7
</td>
<td style="text-align:right;">
8.3
</td>
</tr>
<tr>
<td style="text-align:left;">
cv_error_min
</td>
<td style="text-align:right;">
6.5
</td>
<td style="text-align:right;">
6.8
</td>
<td style="text-align:right;">
7.0
</td>
<td style="text-align:right;">
7.2
</td>
<td style="text-align:right;">
7.5
</td>
</tr>
<tr>
<td style="text-align:left;">
cv_error:1se-min
</td>
<td style="text-align:right;">
0.2
</td>
<td style="text-align:right;">
0.5
</td>
<td style="text-align:right;">
0.6
</td>
<td style="text-align:right;">
0.7
</td>
<td style="text-align:right;">
0.9
</td>
</tr>
</tbody>
</table>
<p>The number of features selected by the minimum lambda model is larger
than the number selected by the “one standard error” rule by a median
of <span class="math inline">\(168.5\)</span> and results on
a median reduction in cv error rates of
<span class="math inline">\(0.6\)</span>%.</p>
<p>The cv error rates observed in this set are comparable to the
rates oberved in the enet models fitted to the training sample set
which consisted of 80% of the samples in this set. See Table @ref(tab:printErrors).
It’s not clear at this point whether the minimum lambda model is better than
the “one standard error” rule model. We would need and external validation
set to make this determination. We can compare the two sets
of out-of-fold predicted values, averaged across cv replicates, to see if
there is a meaningful difference between the two.</p>
<div class="sourceCode" id="cb210"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb210-1"><a href="#cb210-1"></a><span class="co"># predicted probs - 1se</span></span>
<span id="cb210-2"><a href="#cb210-2"></a>enetAll_predResp_1se_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_enetAll_lst, <span class="cf">function</span>(cv_fit) { </span>
<span id="cb210-3"><a href="#cb210-3"></a>  ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se,cv_fit<span class="op">$</span>lambda)</span>
<span id="cb210-4"><a href="#cb210-4"></a>  <span class="kw">logistic_f</span>(cv_fit<span class="op">$</span>fit.preval[,ndx_1se])</span>
<span id="cb210-5"><a href="#cb210-5"></a> })</span>
<span id="cb210-6"><a href="#cb210-6"></a>enetAll_predResp_1se_vec &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(enetAll_predResp_1se_mtx)</span>
<span id="cb210-7"><a href="#cb210-7"></a></span>
<span id="cb210-8"><a href="#cb210-8"></a><span class="co"># predicted probs - min</span></span>
<span id="cb210-9"><a href="#cb210-9"></a>enetAll_predResp_min_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_enetAll_lst, <span class="cf">function</span>(cv_fit) { </span>
<span id="cb210-10"><a href="#cb210-10"></a>  ndx_min &lt;-<span class="st"> </span><span class="kw">match</span>(cv_fit<span class="op">$</span>lambda.min,cv_fit<span class="op">$</span>lambda)</span>
<span id="cb210-11"><a href="#cb210-11"></a>  <span class="kw">logistic_f</span>(cv_fit<span class="op">$</span>fit.preval[,ndx_min])</span>
<span id="cb210-12"><a href="#cb210-12"></a> })</span>
<span id="cb210-13"><a href="#cb210-13"></a>enetAll_predResp_min_vec &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(enetAll_predResp_min_mtx)</span>
<span id="cb210-14"><a href="#cb210-14"></a></span>
<span id="cb210-15"><a href="#cb210-15"></a><span class="co"># plot</span></span>
<span id="cb210-16"><a href="#cb210-16"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">5</span>,<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb210-17"><a href="#cb210-17"></a>tmp &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb210-18"><a href="#cb210-18"></a> <span class="st">`</span><span class="dt">1se</span><span class="st">`</span> =<span class="st"> </span><span class="kw">split</span>(enetAll_predResp_1se_vec, hcc5hmC_all_group_vec),</span>
<span id="cb210-19"><a href="#cb210-19"></a> <span class="dt">min =</span> <span class="kw">split</span>(enetAll_predResp_min_vec, hcc5hmC_all_group_vec)</span>
<span id="cb210-20"><a href="#cb210-20"></a>)</span>
<span id="cb210-21"><a href="#cb210-21"></a><span class="kw">names</span>(tmp) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&#39;</span><span class="ch">\\</span><span class="st">.&#39;</span>, <span class="st">&#39;</span><span class="ch">\n</span><span class="st">&#39;</span>, <span class="kw">names</span>(tmp))</span>
<span id="cb210-22"><a href="#cb210-22"></a></span>
<span id="cb210-23"><a href="#cb210-23"></a><span class="kw">boxplot</span>(</span>
<span id="cb210-24"><a href="#cb210-24"></a> tmp,</span>
<span id="cb210-25"><a href="#cb210-25"></a> <span class="dt">ylab=</span><span class="st">&#39;Predicted oof probability&#39;</span>,</span>
<span id="cb210-26"><a href="#cb210-26"></a> <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;green&#39;</span>, <span class="st">&#39;red&#39;</span>),</span>
<span id="cb210-27"><a href="#cb210-27"></a> <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb210-28"><a href="#cb210-28"></a>)</span>
<span id="cb210-29"><a href="#cb210-29"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(tmp), <span class="dt">tick=</span>F, <span class="kw">names</span>(tmp))</span>
<span id="cb210-30"><a href="#cb210-30"></a></span>
<span id="cb210-31"><a href="#cb210-31"></a></span>
<span id="cb210-32"><a href="#cb210-32"></a><span class="co"># compare the two</span></span>
<span id="cb210-33"><a href="#cb210-33"></a><span class="kw">plot</span>(</span>
<span id="cb210-34"><a href="#cb210-34"></a> <span class="dt">x =</span> enetAll_predResp_1se_vec, <span class="dt">xlab=</span><span class="st">&#39;1se model oof Prob&#39;</span>,</span>
<span id="cb210-35"><a href="#cb210-35"></a> <span class="dt">y =</span> enetAll_predResp_min_vec, <span class="dt">ylab=</span><span class="st">&#39;min lambda model oof Prob&#39;</span>,</span>
<span id="cb210-36"><a href="#cb210-36"></a> <span class="dt">col =</span> <span class="kw">ifelse</span>(hcc5hmC_all_group_vec <span class="op">==</span><span class="st"> &#39;HCC&#39;</span>, <span class="st">&#39;red&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb210-37"><a href="#cb210-37"></a>)</span>
<span id="cb210-38"><a href="#cb210-38"></a> </span>
<span id="cb210-39"><a href="#cb210-39"></a><span class="co"># Add referecne lines at 10% false positive</span></span>
<span id="cb210-40"><a href="#cb210-40"></a>thres_1se &lt;-<span class="st"> </span><span class="kw">quantile</span>(enetAll_predResp_1se_vec[hcc5hmC_all_group_vec <span class="op">==</span><span class="st"> &#39;Control&#39;</span>], <span class="dt">prob=</span>.<span class="dv">9</span>)</span>
<span id="cb210-41"><a href="#cb210-41"></a>thres_min &lt;-<span class="st"> </span><span class="kw">quantile</span>(enetAll_predResp_min_vec[hcc5hmC_all_group_vec <span class="op">==</span><span class="st"> &#39;Control&#39;</span>], <span class="dt">prob=</span>.<span class="dv">9</span>)</span>
<span id="cb210-42"><a href="#cb210-42"></a><span class="kw">abline</span>(<span class="dt">v =</span> thres_1se, <span class="dt">h =</span> thres_min, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuiteB-enet-get-sample-pred-1.png" alt="Predicted probabilities - averaged over cv replicates" width="960" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuiteB-enet-get-sample-pred)Predicted probabilities - averaged over cv replicates
</p>
</div>
<!-- THIS PARAGRAPH REFERRED TO THE FITTED PROBS; NOT THE OOF PRED PROBS
We see that the minimum lambda models provide a better fit to the data,
which is to be expected as the minimum lambda models have more estimated
parameters than the one standard error rule models.  
-->
<p>We see that there isn’t a big difference in out-of-fold predicted
probabilities between the one-standard-error rule ans minimum lamda models.
One way to quantify
the difference in classification errors is to classify samples
according to each vector of predicted probabilities, setting
the thresholds to achieve a fixed false positive rate, 10% say.
These thresholds are indicated by the grey lines in the scatter plot
on the right side of Figure @ref(fig:get-sample-pred).</p>
<!-- APPLIED TO THE FITTED VALUES
We note
that althouth predicted probability distributions are quite different
for the two models, the the class predictions at a 10% false discovery threshold
are largely in agreement.
-->
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="#cb211-1"></a>enetAll_predClass_1se_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb211-2"><a href="#cb211-2"></a> enetAll_predResp_1se_vec <span class="op">&gt;</span><span class="st"> </span>thres_1se, <span class="st">&#39;HCC&#39;</span>, <span class="st">&#39;Control&#39;</span>)</span>
<span id="cb211-3"><a href="#cb211-3"></a></span>
<span id="cb211-4"><a href="#cb211-4"></a>enetAll_predClass_min_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb211-5"><a href="#cb211-5"></a> enetAll_predResp_min_vec <span class="op">&gt;</span><span class="st"> </span>thres_min, <span class="st">&#39;HCC&#39;</span>, <span class="st">&#39;Control&#39;</span>)</span>
<span id="cb211-6"><a href="#cb211-6"></a></span>
<span id="cb211-7"><a href="#cb211-7"></a>tmp &lt;-<span class="st"> </span><span class="kw">cbind</span>(</span>
<span id="cb211-8"><a href="#cb211-8"></a> <span class="kw">table</span>(<span class="dt">truth=</span>hcc5hmC_all_group_vec, <span class="st">`</span><span class="dt">1se-pred</span><span class="st">`</span>=enetAll_predClass_1se_vec),</span>
<span id="cb211-9"><a href="#cb211-9"></a> <span class="kw">table</span>(<span class="dt">truth=</span>hcc5hmC_all_group_vec, <span class="st">`</span><span class="dt">min-pred</span><span class="st">`</span>=enetAll_predClass_min_vec)</span>
<span id="cb211-10"><a href="#cb211-10"></a>) </span>
<span id="cb211-11"><a href="#cb211-11"></a><span class="co"># Hack for printing</span></span>
<span id="cb211-12"><a href="#cb211-12"></a><span class="kw">colnames</span>(tmp) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;1se-Control&#39;</span>, <span class="st">&#39;1se-HCC&#39;</span>, <span class="st">&#39;min-Control&#39;</span>, <span class="st">&#39;min-HCC&#39;</span>)</span>
<span id="cb211-13"><a href="#cb211-13"></a></span>
<span id="cb211-14"><a href="#cb211-14"></a>knitr<span class="op">::</span><span class="kw">kable</span>(tmp,</span>
<span id="cb211-15"><a href="#cb211-15"></a>  <span class="dt">caption =</span> <span class="st">&quot;Classifications: rows are truth&quot;</span>,</span>
<span id="cb211-16"><a href="#cb211-16"></a>  <span class="dt">digits=</span><span class="dv">1</span>) <span class="op">%&gt;%</span></span>
<span id="cb211-17"><a href="#cb211-17"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-glmnetSuiteB-enet-get-sample-class)Classifications: rows are truth
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
1se-Control
</th>
<th style="text-align:right;">
1se-HCC
</th>
<th style="text-align:right;">
min-Control
</th>
<th style="text-align:right;">
min-HCC
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Control
</td>
<td style="text-align:right;">
700
</td>
<td style="text-align:right;">
78
</td>
<td style="text-align:right;">
700
</td>
<td style="text-align:right;">
78
</td>
</tr>
<tr>
<td style="text-align:left;">
HCC
</td>
<td style="text-align:right;">
34
</td>
<td style="text-align:right;">
521
</td>
<td style="text-align:right;">
25
</td>
<td style="text-align:right;">
530
</td>
</tr>
</tbody>
</table>
<p>When we fix the false positive rate at 10%, the <code>1se</code> model makes 39 false
negative calls whereas the minimum lambda model makes 32. A difference
of <span class="math inline">\(1.3\)</span>%</p>
<!-- APPLIED TO THE FITTED PROBABILITIES
We see that the min lambda model, makes no false negative calls at a 90% sensitivity
setting, and the sensitivity could be increased substantially at no false negative
cost.  This is definitely over-fitting the data set.  For the purpose
of computing sample quality scores - what do these differences mean? 
-->
<!-- SKIP THIS - can use lasso scores 
### Get quality scores {.unnumbered}  

To compute quality scores, we will use the out-of-fold predicted probabilities.
-->
<!-- HAVING TROUBLE TURNING NUMBERING OFF - add '###'  
-->
<div id="selected-feature-list-stability-2" class="section level2 unnumbered" number="">
<h2 class="unnumbered" number="">Selected feature list stability</h2>
<p>Before moving on to the simulation, let’s examine gene selection stability on the
full data set. We have two sets of sellected features - one for the
one standard deviation rile model, and one for the mimimum lambda model.
We saw in Table @ref(tab:model-size-enetAll) that the number of features
selected by the minimum lambda models had an IQR of
<span class="math inline">\(271.5-457.25\)</span>,
while the one standard error rule models had an IQR of
<span class="math inline">\(174-215\)</span>.</p>
<p>Let’s examine the stability of the gene lists across cv replicates.</p>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb212-1"><a href="#cb212-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb212-2"><a href="#cb212-2"></a></span>
<span id="cb212-3"><a href="#cb212-3"></a></span>
<span id="cb212-4"><a href="#cb212-4"></a><span class="co"># 1se</span></span>
<span id="cb212-5"><a href="#cb212-5"></a>enetAll_coef_1se_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(hcc5hmC_cv_enetAll_lst, <span class="cf">function</span>(cv_fit){</span>
<span id="cb212-6"><a href="#cb212-6"></a> cv_fit_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb212-7"><a href="#cb212-7"></a> cv_fit,</span>
<span id="cb212-8"><a href="#cb212-8"></a> <span class="dt">s =</span> <span class="st">&quot;lambda.1se&quot;</span></span>
<span id="cb212-9"><a href="#cb212-9"></a> )</span>
<span id="cb212-10"><a href="#cb212-10"></a> cv_fit_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][cv_fit_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>]]</span>
<span id="cb212-11"><a href="#cb212-11"></a> })</span>
<span id="cb212-12"><a href="#cb212-12"></a></span>
<span id="cb212-13"><a href="#cb212-13"></a><span class="co"># put into matrix</span></span>
<span id="cb212-14"><a href="#cb212-14"></a>enetAll_coef_1se_all &lt;-<span class="st"> </span><span class="kw">Reduce</span>(union, enetAll_coef_1se_lst)</span>
<span id="cb212-15"><a href="#cb212-15"></a>enetAll_coef_1se_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(enetAll_coef_1se_lst, </span>
<span id="cb212-16"><a href="#cb212-16"></a>  <span class="cf">function</span>(LL) <span class="kw">is.element</span>(enetAll_coef_1se_all, LL)</span>
<span id="cb212-17"><a href="#cb212-17"></a>)</span>
<span id="cb212-18"><a href="#cb212-18"></a><span class="kw">rownames</span>(enetAll_coef_1se_mtx) &lt;-<span class="st"> </span>enetAll_coef_1se_all</span>
<span id="cb212-19"><a href="#cb212-19"></a></span>
<span id="cb212-20"><a href="#cb212-20"></a>genes_by_rep_1se_tbl &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="kw">rowSums</span>(enetAll_coef_1se_mtx))</span>
<span id="cb212-21"><a href="#cb212-21"></a><span class="kw">barplot</span>(</span>
<span id="cb212-22"><a href="#cb212-22"></a> genes_by_rep_1se_tbl,</span>
<span id="cb212-23"><a href="#cb212-23"></a> <span class="dt">xlab=</span><span class="st">&#39;Number of Replicates&#39;</span>,</span>
<span id="cb212-24"><a href="#cb212-24"></a> <span class="dt">ylab=</span><span class="st">&#39;Number of features&#39;</span></span>
<span id="cb212-25"><a href="#cb212-25"></a></span>
<span id="cb212-26"><a href="#cb212-26"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuiteB-enet-feature-list-1se-1.png" alt="Feature list stability for one standard error rule models" width="768" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuiteB-enet-feature-list-1se)Feature list stability for one standard error rule models
</p>
</div>
<p>We see that <span class="math inline">\(76\)</span> features are included in every
cv replicate. These make up between
<span class="math inline">\(35\)</span>%
and
<span class="math inline">\(44\)</span>%
(Q1 and Q3) of the cv replicate one standard error rule models feature lists.</p>
<div class="sourceCode" id="cb213"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb213-1"><a href="#cb213-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb213-2"><a href="#cb213-2"></a></span>
<span id="cb213-3"><a href="#cb213-3"></a></span>
<span id="cb213-4"><a href="#cb213-4"></a><span class="co"># min</span></span>
<span id="cb213-5"><a href="#cb213-5"></a>enetAll_coef_min_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(hcc5hmC_cv_enetAll_lst, <span class="cf">function</span>(cv_fit){</span>
<span id="cb213-6"><a href="#cb213-6"></a> cv_fit_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb213-7"><a href="#cb213-7"></a> cv_fit,</span>
<span id="cb213-8"><a href="#cb213-8"></a> <span class="dt">s =</span> <span class="st">&quot;lambda.min&quot;</span></span>
<span id="cb213-9"><a href="#cb213-9"></a> )</span>
<span id="cb213-10"><a href="#cb213-10"></a> cv_fit_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][cv_fit_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>]]</span>
<span id="cb213-11"><a href="#cb213-11"></a> })</span>
<span id="cb213-12"><a href="#cb213-12"></a></span>
<span id="cb213-13"><a href="#cb213-13"></a><span class="co"># put into matrix</span></span>
<span id="cb213-14"><a href="#cb213-14"></a>enetAll_coef_min_all &lt;-<span class="st"> </span><span class="kw">Reduce</span>(union, enetAll_coef_min_lst)</span>
<span id="cb213-15"><a href="#cb213-15"></a>enetAll_coef_min_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(enetAll_coef_min_lst, </span>
<span id="cb213-16"><a href="#cb213-16"></a>  <span class="cf">function</span>(LL) <span class="kw">is.element</span>(enetAll_coef_min_all, LL)</span>
<span id="cb213-17"><a href="#cb213-17"></a>)</span>
<span id="cb213-18"><a href="#cb213-18"></a><span class="kw">rownames</span>(enetAll_coef_min_mtx) &lt;-<span class="st"> </span>enetAll_coef_min_all</span>
<span id="cb213-19"><a href="#cb213-19"></a></span>
<span id="cb213-20"><a href="#cb213-20"></a>genes_by_rep_min_tbl &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="kw">rowSums</span>(enetAll_coef_min_mtx))</span>
<span id="cb213-21"><a href="#cb213-21"></a><span class="kw">barplot</span>(</span>
<span id="cb213-22"><a href="#cb213-22"></a> genes_by_rep_min_tbl,</span>
<span id="cb213-23"><a href="#cb213-23"></a> <span class="dt">xlab=</span><span class="st">&#39;Number of Replicates&#39;</span>,</span>
<span id="cb213-24"><a href="#cb213-24"></a> <span class="dt">ylab=</span><span class="st">&#39;Number of features&#39;</span></span>
<span id="cb213-25"><a href="#cb213-25"></a></span>
<span id="cb213-26"><a href="#cb213-26"></a>)</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuiteB-enet-feature-list-min-1.png" alt="Feature list stability for minimum lambda models" width="768" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuiteB-enet-feature-list-min)Feature list stability for minimum lambda models
</p>
</div>
<p>We see that <span class="math inline">\(168\)</span> features are included in every
cv replicate. These make up between
<span class="math inline">\(37\)</span>%
and
<span class="math inline">\(62\)</span>%
(Q1 and Q3) of the cv replicate min feature lists.
We will consider the genes that are selected in all cv replicates as a
gene signature produced by each model.</p>
<div class="sourceCode" id="cb214"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb214-1"><a href="#cb214-1"></a>enet_gene_sign_1se_vec &lt;-<span class="st"> </span><span class="kw">rownames</span>(enetAll_coef_1se_mtx)[<span class="kw">rowSums</span>(enetAll_coef_1se_mtx)<span class="op">==</span><span class="dv">30</span>]</span>
<span id="cb214-2"><a href="#cb214-2"></a>enet_gene_sign_min_vec &lt;-<span class="st"> </span><span class="kw">rownames</span>(enetAll_coef_min_mtx)[<span class="kw">rowSums</span>(enetAll_coef_min_mtx)<span class="op">==</span><span class="dv">30</span>]</span></code></pre></div>
<p>76 out of
76 of the genes in the 1se model gene signature
are contained in the min lambda model gene signature.</p>
<!-- USE SAME DESIGN AS lasso
## Simulation Design

We are now ready to run the simulations.

-->
</div>
<div id="run-simulations---enet" class="section level2 unnumbered" number="">
<h2 class="unnumbered" number="">Run simulations - enet</h2>
<p>As these make take a while to run,
we will save the results of each similation to a different
object and store to disk. These can be easily read from disk
when needed for analysis.</p>
<p>The simulation saves results to the file system and
only needs to be run once. The simulation takes <span class="math inline">\(\approx\)</span> 8 minutes
per iteration, or 4 hours of run time on a laptop.
(Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Mojave 10.14.6)</p>
<!-- RUN ONCE - THEN GET FROM MEMORY -->
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb215-1"><a href="#cb215-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb215-2"><a href="#cb215-2"></a>start_time &lt;-<span class="st"> </span><span class="kw">proc.time</span>()</span>
<span id="cb215-3"><a href="#cb215-3"></a></span>
<span id="cb215-4"><a href="#cb215-4"></a><span class="co"># Get stage from SIZE</span></span>
<span id="cb215-5"><a href="#cb215-5"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sim_control_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>, SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb215-6"><a href="#cb215-6"></a></span>
<span id="cb215-7"><a href="#cb215-7"></a><span class="co"># ran in two runs 1:7, 8:ncol</span></span>
<span id="cb215-8"><a href="#cb215-8"></a><span class="cf">for</span> (SIMno <span class="cf">in</span> <span class="dv">8</span><span class="op">:</span><span class="kw">ncol</span>(sim_control_qual_mtx)) {</span>
<span id="cb215-9"><a href="#cb215-9"></a></span>
<span id="cb215-10"><a href="#cb215-10"></a>  <span class="co">#cat(&quot;Running simulation &quot;, SIMno, &quot;\n&quot;)</span></span>
<span id="cb215-11"><a href="#cb215-11"></a></span>
<span id="cb215-12"><a href="#cb215-12"></a>  sim_cv_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(<span class="kw">levels</span>(stage_vec)), <span class="cf">function</span>(STGno) {</span>
<span id="cb215-13"><a href="#cb215-13"></a>    Stage_rows_vec &lt;-<span class="st"> </span><span class="kw">which</span>(stage_vec <span class="op">%in%</span><span class="st"> </span><span class="kw">levels</span>(stage_vec)[<span class="dv">1</span><span class="op">:</span>STGno])</span>
<span id="cb215-14"><a href="#cb215-14"></a>    <span class="co">#cat(&quot;Stage &quot;, STGno, &quot;- analyzing&quot;, length(Stage_rows_vec), &quot;paired samples.\n&quot;)</span></span>
<span id="cb215-15"><a href="#cb215-15"></a></span>
<span id="cb215-16"><a href="#cb215-16"></a>    sim_stage_samples_vec &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb215-17"><a href="#cb215-17"></a>      all_control_vec[sim_control_mtx[Stage_rows_vec, SIMno]],</span>
<span id="cb215-18"><a href="#cb215-18"></a>      all_affected_vec[sim_affected_mtx[Stage_rows_vec, SIMno]]</span>
<span id="cb215-19"><a href="#cb215-19"></a>    )</span>
<span id="cb215-20"><a href="#cb215-20"></a>    sim_stage_lcpm_mtx &lt;-<span class="st"> </span>all_lcpm_mtx[sim_stage_samples_vec, ]</span>
<span id="cb215-21"><a href="#cb215-21"></a>    sim_stage_group_vec &lt;-<span class="st"> </span>hcc5hmC_all_group_vec[sim_stage_samples_vec]</span>
<span id="cb215-22"><a href="#cb215-22"></a>    <span class="co">#print(table(sim_stage_group_vec))</span></span>
<span id="cb215-23"><a href="#cb215-23"></a></span>
<span id="cb215-24"><a href="#cb215-24"></a>    sim_stage_cv_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span>CV_REP, <span class="cf">function</span>(CV) {</span>
<span id="cb215-25"><a href="#cb215-25"></a>      cv_fit &lt;-<span class="st"> </span>glmnet<span class="op">::</span><span class="kw">cv.glmnet</span>(</span>
<span id="cb215-26"><a href="#cb215-26"></a>        <span class="dt">x =</span> sim_stage_lcpm_mtx,</span>
<span id="cb215-27"><a href="#cb215-27"></a>        <span class="dt">y =</span> sim_stage_group_vec,</span>
<span id="cb215-28"><a href="#cb215-28"></a>        <span class="dt">alpha =</span> <span class="dv">1</span>,</span>
<span id="cb215-29"><a href="#cb215-29"></a>        <span class="dt">family =</span> <span class="st">&quot;binomial&quot;</span>,</span>
<span id="cb215-30"><a href="#cb215-30"></a>        <span class="dt">type.measure =</span> <span class="st">&quot;class&quot;</span>,</span>
<span id="cb215-31"><a href="#cb215-31"></a>        <span class="dt">keep =</span> T,</span>
<span id="cb215-32"><a href="#cb215-32"></a>        <span class="dt">nlambda =</span> <span class="dv">30</span></span>
<span id="cb215-33"><a href="#cb215-33"></a>      )</span>
<span id="cb215-34"><a href="#cb215-34"></a></span>
<span id="cb215-35"><a href="#cb215-35"></a>      <span class="co"># Extract 1se metrics from cv_fit</span></span>
<span id="cb215-36"><a href="#cb215-36"></a>      <span class="co">#######################</span></span>
<span id="cb215-37"><a href="#cb215-37"></a>      ndx_1se &lt;-<span class="st"> </span><span class="kw">which</span>(cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se)</span>
<span id="cb215-38"><a href="#cb215-38"></a></span>
<span id="cb215-39"><a href="#cb215-39"></a>      nzero_1se &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>nzero[ndx_1se]</span>
<span id="cb215-40"><a href="#cb215-40"></a>      cvm_1se &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>cvm[ndx_1se]</span>
<span id="cb215-41"><a href="#cb215-41"></a></span>
<span id="cb215-42"><a href="#cb215-42"></a>      <span class="co"># test error</span></span>
<span id="cb215-43"><a href="#cb215-43"></a>      sim_stage_test_samples_vec &lt;-<span class="st"> </span><span class="kw">setdiff</span>(<span class="kw">rownames</span>(all_lcpm_mtx), sim_stage_samples_vec)</span>
<span id="cb215-44"><a href="#cb215-44"></a>      sim_stage_test_lcpm_mtx &lt;-<span class="st"> </span>all_lcpm_mtx[sim_stage_test_samples_vec,]</span>
<span id="cb215-45"><a href="#cb215-45"></a>      sim_stage_test_group_vec &lt;-<span class="st"> </span>hcc5hmC_all_group_vec[sim_stage_test_samples_vec]</span>
<span id="cb215-46"><a href="#cb215-46"></a></span>
<span id="cb215-47"><a href="#cb215-47"></a>      test_pred_1se_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb215-48"><a href="#cb215-48"></a>       cv_fit,</span>
<span id="cb215-49"><a href="#cb215-49"></a>       <span class="dt">newx=</span>sim_stage_test_lcpm_mtx,</span>
<span id="cb215-50"><a href="#cb215-50"></a>       <span class="dt">s=</span><span class="st">&quot;lambda.1se&quot;</span>,</span>
<span id="cb215-51"><a href="#cb215-51"></a>       <span class="dt">type=</span><span class="st">&quot;class&quot;</span></span>
<span id="cb215-52"><a href="#cb215-52"></a>      )</span>
<span id="cb215-53"><a href="#cb215-53"></a>      test_1se_error &lt;-<span class="st"> </span><span class="kw">mean</span>(test_pred_1se_vec <span class="op">!=</span><span class="st"> </span>sim_stage_test_group_vec)</span>
<span id="cb215-54"><a href="#cb215-54"></a></span>
<span id="cb215-55"><a href="#cb215-55"></a>      <span class="co"># genes</span></span>
<span id="cb215-56"><a href="#cb215-56"></a>      coef_1se &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb215-57"><a href="#cb215-57"></a>        cv_fit,</span>
<span id="cb215-58"><a href="#cb215-58"></a>        <span class="dt">s =</span> <span class="st">&quot;lambda.1se&quot;</span></span>
<span id="cb215-59"><a href="#cb215-59"></a>      )</span>
<span id="cb215-60"><a href="#cb215-60"></a>      genes_1se &lt;-<span class="st"> </span>coef_1se<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][coef_1se<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>]]</span>
<span id="cb215-61"><a href="#cb215-61"></a></span>
<span id="cb215-62"><a href="#cb215-62"></a>      <span class="co"># Extract min metrics from cv_fit</span></span>
<span id="cb215-63"><a href="#cb215-63"></a>      <span class="co">#######################</span></span>
<span id="cb215-64"><a href="#cb215-64"></a>      ndx_min &lt;-<span class="st"> </span><span class="kw">which</span>(cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min)</span>
<span id="cb215-65"><a href="#cb215-65"></a></span>
<span id="cb215-66"><a href="#cb215-66"></a>      nzero_min &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>nzero[ndx_min]</span>
<span id="cb215-67"><a href="#cb215-67"></a>      cvm_min &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>cvm[ndx_min]</span>
<span id="cb215-68"><a href="#cb215-68"></a></span>
<span id="cb215-69"><a href="#cb215-69"></a>      <span class="co"># test error</span></span>
<span id="cb215-70"><a href="#cb215-70"></a>      sim_stage_test_samples_vec &lt;-<span class="st"> </span><span class="kw">setdiff</span>(<span class="kw">rownames</span>(all_lcpm_mtx), sim_stage_samples_vec)</span>
<span id="cb215-71"><a href="#cb215-71"></a>      sim_stage_test_lcpm_mtx &lt;-<span class="st"> </span>all_lcpm_mtx[sim_stage_test_samples_vec,]</span>
<span id="cb215-72"><a href="#cb215-72"></a>      sim_stage_test_group_vec &lt;-<span class="st"> </span>hcc5hmC_all_group_vec[sim_stage_test_samples_vec]</span>
<span id="cb215-73"><a href="#cb215-73"></a></span>
<span id="cb215-74"><a href="#cb215-74"></a>      test_pred_min_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb215-75"><a href="#cb215-75"></a>       cv_fit,</span>
<span id="cb215-76"><a href="#cb215-76"></a>       <span class="dt">newx=</span>sim_stage_test_lcpm_mtx,</span>
<span id="cb215-77"><a href="#cb215-77"></a>       <span class="dt">s=</span><span class="st">&quot;lambda.min&quot;</span>,</span>
<span id="cb215-78"><a href="#cb215-78"></a>       <span class="dt">type=</span><span class="st">&quot;class&quot;</span></span>
<span id="cb215-79"><a href="#cb215-79"></a>      )</span>
<span id="cb215-80"><a href="#cb215-80"></a>      test_min_error &lt;-<span class="st"> </span><span class="kw">mean</span>(test_pred_min_vec <span class="op">!=</span><span class="st"> </span>sim_stage_test_group_vec)</span>
<span id="cb215-81"><a href="#cb215-81"></a></span>
<span id="cb215-82"><a href="#cb215-82"></a>      <span class="co"># genes</span></span>
<span id="cb215-83"><a href="#cb215-83"></a>      coef_min &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb215-84"><a href="#cb215-84"></a>        cv_fit,</span>
<span id="cb215-85"><a href="#cb215-85"></a>        <span class="dt">s =</span> <span class="st">&quot;lambda.min&quot;</span></span>
<span id="cb215-86"><a href="#cb215-86"></a>      )</span>
<span id="cb215-87"><a href="#cb215-87"></a>      genes_min &lt;-<span class="st"> </span>coef_min<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][coef_min<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>]]</span>
<span id="cb215-88"><a href="#cb215-88"></a></span>
<span id="cb215-89"><a href="#cb215-89"></a>      <span class="co"># return cv_fit summary metrics</span></span>
<span id="cb215-90"><a href="#cb215-90"></a>      <span class="kw">list</span>(</span>
<span id="cb215-91"><a href="#cb215-91"></a>       <span class="dt">p_1se =</span> nzero_1se, </span>
<span id="cb215-92"><a href="#cb215-92"></a>       <span class="dt">p_min =</span> nzero_min, </span>
<span id="cb215-93"><a href="#cb215-93"></a>       <span class="dt">cv_1se =</span> cvm_1se, </span>
<span id="cb215-94"><a href="#cb215-94"></a>       <span class="dt">cv_min =</span> cvm_min, </span>
<span id="cb215-95"><a href="#cb215-95"></a>       <span class="dt">test_1se=</span>test_1se_error, </span>
<span id="cb215-96"><a href="#cb215-96"></a>       <span class="dt">test_min=</span>test_min_error, </span>
<span id="cb215-97"><a href="#cb215-97"></a>       <span class="dt">genes_1se =</span> genes_1se,</span>
<span id="cb215-98"><a href="#cb215-98"></a>       <span class="dt">genes_min =</span> genes_min)</span>
<span id="cb215-99"><a href="#cb215-99"></a>    })</span>
<span id="cb215-100"><a href="#cb215-100"></a>    sim_stage_cv_lst</span>
<span id="cb215-101"><a href="#cb215-101"></a>  })</span>
<span id="cb215-102"><a href="#cb215-102"></a></span>
<span id="cb215-103"><a href="#cb215-103"></a>  <span class="co"># save  sim_cv_lst</span></span>
<span id="cb215-104"><a href="#cb215-104"></a>  fName &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;enet_sim_&quot;</span>, SIMno, <span class="st">&quot;_cv_lst&quot;</span>)</span>
<span id="cb215-105"><a href="#cb215-105"></a>  <span class="kw">assign</span>(fName, sim_cv_lst)</span>
<span id="cb215-106"><a href="#cb215-106"></a>  <span class="kw">save</span>(<span class="dt">list =</span> fName, <span class="dt">file=</span><span class="kw">file.path</span>(<span class="st">&quot;RData&quot;</span>, fName))</span>
<span id="cb215-107"><a href="#cb215-107"></a></span>
<span id="cb215-108"><a href="#cb215-108"></a>}</span>
<span id="cb215-109"><a href="#cb215-109"></a>  <span class="kw">message</span>(<span class="st">&quot;simulation time: &quot;</span>, <span class="kw">round</span>((<span class="kw">proc.time</span>() <span class="op">-</span><span class="st"> </span>start_time)[<span class="dv">3</span>], <span class="dv">2</span>), <span class="st">&quot;s&quot;</span>)</span></code></pre></div>
</div>
<div id="enet-simulation-results" class="section level2 unnumbered" number="">
<h2 class="unnumbered" number="">enet Simulation results</h2>
<!--
First we extract simluation results and store into one big table:
-->
<!-- 
Have a table of simulation results - `enet_sim_results_frm`:
-->
<div id="simulation-results---look-at-one-simulation-2" class="section level3 unnumbered" number="">
<h3 class="unnumbered" number="">Simulation Results - look at one simulation</h3>
<p>First examine results for one simulation run.</p>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="#cb216-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb216-2"><a href="#cb216-2"></a></span>
<span id="cb216-3"><a href="#cb216-3"></a><span class="co"># get full model cv error ref</span></span>
<span id="cb216-4"><a href="#cb216-4"></a>error_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_enetAll_lst,</span>
<span id="cb216-5"><a href="#cb216-5"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb216-6"><a href="#cb216-6"></a>error_1se_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(error_1se_vec, <span class="dt">prob=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>)        </span>
<span id="cb216-7"><a href="#cb216-7"></a></span>
<span id="cb216-8"><a href="#cb216-8"></a>error_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_enetAll_lst,</span>
<span id="cb216-9"><a href="#cb216-9"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb216-10"><a href="#cb216-10"></a>error_min_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(error_min_vec, <span class="dt">prob=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>)        </span>
<span id="cb216-11"><a href="#cb216-11"></a></span>
<span id="cb216-12"><a href="#cb216-12"></a><span class="co"># Utility objects</span></span>
<span id="cb216-13"><a href="#cb216-13"></a>SIZE0 &lt;-<span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_pad</span>(SIZE, <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>)</span>
<span id="cb216-14"><a href="#cb216-14"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sim_control_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>,SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb216-15"><a href="#cb216-15"></a></span>
<span id="cb216-16"><a href="#cb216-16"></a></span>
<span id="cb216-17"><a href="#cb216-17"></a><span class="co">#SIM &lt;- &quot;Sim_01&quot;</span></span>
<span id="cb216-18"><a href="#cb216-18"></a></span>
<span id="cb216-19"><a href="#cb216-19"></a><span class="cf">for</span>(SIM <span class="cf">in</span> <span class="kw">unique</span>(enet_sim_results_frm<span class="op">$</span>SimNo)[<span class="dv">1</span>]){</span>
<span id="cb216-20"><a href="#cb216-20"></a></span>
<span id="cb216-21"><a href="#cb216-21"></a>SimNum &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">sub</span>(<span class="st">&#39;Sim_&#39;</span>,<span class="st">&#39;&#39;</span>,SIM))</span>
<span id="cb216-22"><a href="#cb216-22"></a></span>
<span id="cb216-23"><a href="#cb216-23"></a>simNo_results_frm &lt;-<span class="st"> </span>enet_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(SimNo<span class="op">==</span>SIM)</span>
<span id="cb216-24"><a href="#cb216-24"></a></span>
<span id="cb216-25"><a href="#cb216-25"></a></span>
<span id="cb216-26"><a href="#cb216-26"></a><span class="co"># errors</span></span>
<span id="cb216-27"><a href="#cb216-27"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb216-28"><a href="#cb216-28"></a><span class="co">###################</span></span>
<span id="cb216-29"><a href="#cb216-29"></a><span class="co"># 1se</span></span>
<span id="cb216-30"><a href="#cb216-30"></a><span class="co">####################</span></span>
<span id="cb216-31"><a href="#cb216-31"></a>cv_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb216-32"><a href="#cb216-32"></a> <span class="kw">split</span>(cv_1se, Size))</span>
<span id="cb216-33"><a href="#cb216-33"></a><span class="kw">names</span>(cv_1se_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(cv_1se_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb216-34"><a href="#cb216-34"></a></span>
<span id="cb216-35"><a href="#cb216-35"></a>test_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb216-36"><a href="#cb216-36"></a> <span class="kw">split</span>(test_1se, Size))</span>
<span id="cb216-37"><a href="#cb216-37"></a><span class="kw">names</span>(test_1se_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(test_1se_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb216-38"><a href="#cb216-38"></a></span>
<span id="cb216-39"><a href="#cb216-39"></a>error_1se_lst &lt;-<span class="st"> </span><span class="kw">c</span>(cv_1se_lst, test_1se_lst)</span>
<span id="cb216-40"><a href="#cb216-40"></a>error_1se_lst &lt;-<span class="st"> </span>error_1se_lst[<span class="kw">order</span>(<span class="kw">names</span>(error_1se_lst))]</span>
<span id="cb216-41"><a href="#cb216-41"></a></span>
<span id="cb216-42"><a href="#cb216-42"></a><span class="kw">boxplot</span>(error_1se_lst, </span>
<span id="cb216-43"><a href="#cb216-43"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>), </span>
<span id="cb216-44"><a href="#cb216-44"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="fl">0.05</span>, <span class="fl">.4</span>),</span>
<span id="cb216-45"><a href="#cb216-45"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb216-46"><a href="#cb216-46"></a>)</span>
<span id="cb216-47"><a href="#cb216-47"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb216-48"><a href="#cb216-48"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb216-49"><a href="#cb216-49"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_lst)), </span>
<span id="cb216-50"><a href="#cb216-50"></a>  SIZE0</span>
<span id="cb216-51"><a href="#cb216-51"></a> )</span>
<span id="cb216-52"><a href="#cb216-52"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb216-53"><a href="#cb216-53"></a><span class="kw">abline</span>(<span class="dt">h=</span> error_1se_q2, <span class="dt">col =</span> <span class="st">&#39;red&#39;</span>)</span>
<span id="cb216-54"><a href="#cb216-54"></a><span class="kw">legend</span>(<span class="st">&#39;topright&#39;</span>, </span>
<span id="cb216-55"><a href="#cb216-55"></a>   <span class="co">#title=&#39;1se errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb216-56"><a href="#cb216-56"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb216-57"><a href="#cb216-57"></a>   <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&#39;cv error&#39;</span>, <span class="st">&#39;test set&#39;</span>),</span>
<span id="cb216-58"><a href="#cb216-58"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb216-59"><a href="#cb216-59"></a> )</span>
<span id="cb216-60"><a href="#cb216-60"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;one se lambda - error rates&#39;</span>))</span>
<span id="cb216-61"><a href="#cb216-61"></a></span>
<span id="cb216-62"><a href="#cb216-62"></a>SKIP  &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb216-63"><a href="#cb216-63"></a><span class="co"># Add qual annotation</span></span>
<span id="cb216-64"><a href="#cb216-64"></a>control_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(sim_control_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb216-65"><a href="#cb216-65"></a>affected_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(sim_affected_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb216-66"><a href="#cb216-66"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb216-67"><a href="#cb216-67"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb216-68"><a href="#cb216-68"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_lst)),</span>
<span id="cb216-69"><a href="#cb216-69"></a>  <span class="kw">round</span>(control_qual_vec, <span class="dv">2</span>)</span>
<span id="cb216-70"><a href="#cb216-70"></a> )</span>
<span id="cb216-71"><a href="#cb216-71"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb216-72"><a href="#cb216-72"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb216-73"><a href="#cb216-73"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_lst)),</span>
<span id="cb216-74"><a href="#cb216-74"></a>  <span class="kw">round</span>(affected_qual_vec, <span class="dv">2</span>)</span>
<span id="cb216-75"><a href="#cb216-75"></a> )</span>
<span id="cb216-76"><a href="#cb216-76"></a>}<span class="co">#SKIP</span></span>
<span id="cb216-77"><a href="#cb216-77"></a></span>
<span id="cb216-78"><a href="#cb216-78"></a><span class="co"># min</span></span>
<span id="cb216-79"><a href="#cb216-79"></a><span class="co">####################</span></span>
<span id="cb216-80"><a href="#cb216-80"></a>cv_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb216-81"><a href="#cb216-81"></a> <span class="kw">split</span>(cv_min, Size))</span>
<span id="cb216-82"><a href="#cb216-82"></a><span class="kw">names</span>(cv_min_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(cv_min_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb216-83"><a href="#cb216-83"></a></span>
<span id="cb216-84"><a href="#cb216-84"></a>test_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb216-85"><a href="#cb216-85"></a> <span class="kw">split</span>(test_min, Size))</span>
<span id="cb216-86"><a href="#cb216-86"></a><span class="kw">names</span>(test_min_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(test_min_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb216-87"><a href="#cb216-87"></a></span>
<span id="cb216-88"><a href="#cb216-88"></a>error_min_lst &lt;-<span class="st"> </span><span class="kw">c</span>(cv_min_lst, test_min_lst)</span>
<span id="cb216-89"><a href="#cb216-89"></a>error_min_lst &lt;-<span class="st"> </span>error_min_lst[<span class="kw">order</span>(<span class="kw">names</span>(error_min_lst))]</span>
<span id="cb216-90"><a href="#cb216-90"></a></span>
<span id="cb216-91"><a href="#cb216-91"></a><span class="kw">boxplot</span>(error_min_lst, </span>
<span id="cb216-92"><a href="#cb216-92"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>), </span>
<span id="cb216-93"><a href="#cb216-93"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="fl">0.05</span>, <span class="fl">.4</span>),</span>
<span id="cb216-94"><a href="#cb216-94"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb216-95"><a href="#cb216-95"></a>)</span>
<span id="cb216-96"><a href="#cb216-96"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb216-97"><a href="#cb216-97"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb216-98"><a href="#cb216-98"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_lst)), </span>
<span id="cb216-99"><a href="#cb216-99"></a>  SIZE0</span>
<span id="cb216-100"><a href="#cb216-100"></a> )</span>
<span id="cb216-101"><a href="#cb216-101"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb216-102"><a href="#cb216-102"></a><span class="kw">abline</span>(<span class="dt">h=</span> error_min_q2, <span class="dt">col =</span> <span class="st">&#39;red&#39;</span>)</span>
<span id="cb216-103"><a href="#cb216-103"></a><span class="kw">legend</span>(<span class="st">&#39;topright&#39;</span>, </span>
<span id="cb216-104"><a href="#cb216-104"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb216-105"><a href="#cb216-105"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb216-106"><a href="#cb216-106"></a>   <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&#39;cv error&#39;</span>, <span class="st">&#39;test set&#39;</span>),</span>
<span id="cb216-107"><a href="#cb216-107"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb216-108"><a href="#cb216-108"></a> )</span>
<span id="cb216-109"><a href="#cb216-109"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;min lambda - error rates&#39;</span>))</span>
<span id="cb216-110"><a href="#cb216-110"></a></span>
<span id="cb216-111"><a href="#cb216-111"></a>SKIP  &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb216-112"><a href="#cb216-112"></a><span class="co"># Add qual annotation</span></span>
<span id="cb216-113"><a href="#cb216-113"></a>control_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(sim_control_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb216-114"><a href="#cb216-114"></a>affected_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(sim_affected_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb216-115"><a href="#cb216-115"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb216-116"><a href="#cb216-116"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb216-117"><a href="#cb216-117"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_lst)),</span>
<span id="cb216-118"><a href="#cb216-118"></a>  <span class="kw">round</span>(control_qual_vec, <span class="dv">2</span>)</span>
<span id="cb216-119"><a href="#cb216-119"></a> )</span>
<span id="cb216-120"><a href="#cb216-120"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb216-121"><a href="#cb216-121"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb216-122"><a href="#cb216-122"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_lst)),</span>
<span id="cb216-123"><a href="#cb216-123"></a>  <span class="kw">round</span>(affected_qual_vec, <span class="dv">2</span>)</span>
<span id="cb216-124"><a href="#cb216-124"></a> )</span>
<span id="cb216-125"><a href="#cb216-125"></a>}<span class="co">#SKIP</span></span>
<span id="cb216-126"><a href="#cb216-126"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, <span class="kw">paste</span>(<span class="st">&#39;Sim =&#39;</span>,  SIM))</span>
<span id="cb216-127"><a href="#cb216-127"></a></span>
<span id="cb216-128"><a href="#cb216-128"></a>} <span class="co"># for(SIM</span></span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuiteB-enet-simRes-errors-bySim-1.png" alt="enet Model Errors by Sample Size" width="960" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuiteB-enet-simRes-errors-bySim)enet Model Errors by Sample Size
</p>
</div>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb217-1"><a href="#cb217-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb217-2"><a href="#cb217-2"></a></span>
<span id="cb217-3"><a href="#cb217-3"></a><span class="co"># get full model nzero ref</span></span>
<span id="cb217-4"><a href="#cb217-4"></a>nzero_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_enetAll_lst,</span>
<span id="cb217-5"><a href="#cb217-5"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb217-6"><a href="#cb217-6"></a>nzero_1se_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(nzero_1se_vec, <span class="dt">prob=</span><span class="kw">c</span>(<span class="dv">2</span>)<span class="op">/</span><span class="dv">4</span>)</span>
<span id="cb217-7"><a href="#cb217-7"></a></span>
<span id="cb217-8"><a href="#cb217-8"></a>nzero_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_enetAll_lst,</span>
<span id="cb217-9"><a href="#cb217-9"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb217-10"><a href="#cb217-10"></a>nzero_min_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(nzero_min_vec, <span class="dt">prob=</span><span class="kw">c</span>(<span class="dv">2</span>)<span class="op">/</span><span class="dv">4</span>)</span>
<span id="cb217-11"><a href="#cb217-11"></a></span>
<span id="cb217-12"><a href="#cb217-12"></a><span class="co"># Utility objects</span></span>
<span id="cb217-13"><a href="#cb217-13"></a>SIZE0 &lt;-<span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_pad</span>(SIZE, <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>)</span>
<span id="cb217-14"><a href="#cb217-14"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sim_control_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>,SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb217-15"><a href="#cb217-15"></a></span>
<span id="cb217-16"><a href="#cb217-16"></a></span>
<span id="cb217-17"><a href="#cb217-17"></a><span class="co">#SIM &lt;- &quot;Sim_01&quot;</span></span>
<span id="cb217-18"><a href="#cb217-18"></a></span>
<span id="cb217-19"><a href="#cb217-19"></a><span class="cf">for</span>(SIM <span class="cf">in</span> <span class="kw">unique</span>(enet_sim_results_frm<span class="op">$</span>SimNo)[<span class="dv">1</span>]){</span>
<span id="cb217-20"><a href="#cb217-20"></a></span>
<span id="cb217-21"><a href="#cb217-21"></a>SimNum &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">sub</span>(<span class="st">&#39;Sim_&#39;</span>,<span class="st">&#39;&#39;</span>,SIM))</span>
<span id="cb217-22"><a href="#cb217-22"></a></span>
<span id="cb217-23"><a href="#cb217-23"></a>simNo_results_frm &lt;-<span class="st"> </span>enet_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(SimNo<span class="op">==</span>SIM)</span>
<span id="cb217-24"><a href="#cb217-24"></a></span>
<span id="cb217-25"><a href="#cb217-25"></a></span>
<span id="cb217-26"><a href="#cb217-26"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb217-27"><a href="#cb217-27"></a><span class="co">###################</span></span>
<span id="cb217-28"><a href="#cb217-28"></a><span class="co"># 1se</span></span>
<span id="cb217-29"><a href="#cb217-29"></a><span class="co">####################</span></span>
<span id="cb217-30"><a href="#cb217-30"></a><span class="co"># selected feature counts</span></span>
<span id="cb217-31"><a href="#cb217-31"></a>p_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb217-32"><a href="#cb217-32"></a> <span class="kw">split</span>(p_1se, Size))</span>
<span id="cb217-33"><a href="#cb217-33"></a><span class="kw">names</span>(p_1se_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(p_1se_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_p&#39;</span>)</span>
<span id="cb217-34"><a href="#cb217-34"></a></span>
<span id="cb217-35"><a href="#cb217-35"></a><span class="co"># get selected features that are part of enet_gene_sign_1se_vec</span></span>
<span id="cb217-36"><a href="#cb217-36"></a><span class="co"># - the signature selected genes</span></span>
<span id="cb217-37"><a href="#cb217-37"></a>sign_genes_1se_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(simNo_results_frm), <span class="cf">function</span>(RR)</span>
<span id="cb217-38"><a href="#cb217-38"></a>    <span class="kw">intersect</span>(<span class="kw">unlist</span>(simNo_results_frm[RR, <span class="st">&#39;genes_1se&#39;</span>]), enet_gene_sign_1se_vec))</span>
<span id="cb217-39"><a href="#cb217-39"></a></span>
<span id="cb217-40"><a href="#cb217-40"></a>sign_p_1se_lst &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="kw">sapply</span>(sign_genes_1se_lst, length), simNo_results_frm<span class="op">$</span>Size)</span>
<span id="cb217-41"><a href="#cb217-41"></a><span class="kw">names</span>(sign_p_1se_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(sign_p_1se_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_signP&#39;</span>)</span>
<span id="cb217-42"><a href="#cb217-42"></a></span>
<span id="cb217-43"><a href="#cb217-43"></a></span>
<span id="cb217-44"><a href="#cb217-44"></a>p_singP_1se_lst &lt;-<span class="st"> </span><span class="kw">c</span>(p_1se_lst, sign_p_1se_lst)</span>
<span id="cb217-45"><a href="#cb217-45"></a>p_singP_1se_lst &lt;-<span class="st"> </span>p_singP_1se_lst[<span class="kw">order</span>(<span class="kw">names</span>(p_singP_1se_lst))]</span>
<span id="cb217-46"><a href="#cb217-46"></a></span>
<span id="cb217-47"><a href="#cb217-47"></a><span class="kw">boxplot</span>(p_singP_1se_lst,</span>
<span id="cb217-48"><a href="#cb217-48"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb217-49"><a href="#cb217-49"></a>  <span class="co">#ylim=c(0, 300),</span></span>
<span id="cb217-50"><a href="#cb217-50"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb217-51"><a href="#cb217-51"></a>)</span>
<span id="cb217-52"><a href="#cb217-52"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb217-53"><a href="#cb217-53"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb217-54"><a href="#cb217-54"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_lst)),</span>
<span id="cb217-55"><a href="#cb217-55"></a>  SIZE0</span>
<span id="cb217-56"><a href="#cb217-56"></a> )</span>
<span id="cb217-57"><a href="#cb217-57"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb217-58"><a href="#cb217-58"></a><span class="co">#abline(h= nzero_1se_q2, col = &#39;red&#39;)</span></span>
<span id="cb217-59"><a href="#cb217-59"></a><span class="kw">legend</span>(<span class="st">&#39;topleft&#39;</span>,</span>
<span id="cb217-60"><a href="#cb217-60"></a>   <span class="co">#title=&#39;1se errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb217-61"><a href="#cb217-61"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>),</span>
<span id="cb217-62"><a href="#cb217-62"></a>   <span class="dt">legend=</span> <span class="kw">c</span>(<span class="st">&#39;selected genes&#39;</span>,<span class="st">&#39;signature genes&#39;</span>),</span>
<span id="cb217-63"><a href="#cb217-63"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb217-64"><a href="#cb217-64"></a> )</span>
<span id="cb217-65"><a href="#cb217-65"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;one se lamdba - selected gene counts&#39;</span>))</span>
<span id="cb217-66"><a href="#cb217-66"></a></span>
<span id="cb217-67"><a href="#cb217-67"></a>SKIP  &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb217-68"><a href="#cb217-68"></a><span class="co"># Add qual annotation</span></span>
<span id="cb217-69"><a href="#cb217-69"></a>control_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(sim_control_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb217-70"><a href="#cb217-70"></a>affected_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(sim_affected_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb217-71"><a href="#cb217-71"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb217-72"><a href="#cb217-72"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb217-73"><a href="#cb217-73"></a>  <span class="dt">at =</span>  <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_lst)),</span>
<span id="cb217-74"><a href="#cb217-74"></a>  <span class="kw">round</span>(control_qual_vec, <span class="dv">2</span>)</span>
<span id="cb217-75"><a href="#cb217-75"></a> )</span>
<span id="cb217-76"><a href="#cb217-76"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb217-77"><a href="#cb217-77"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb217-78"><a href="#cb217-78"></a>  <span class="dt">at =</span>  <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_lst)),</span>
<span id="cb217-79"><a href="#cb217-79"></a>  <span class="kw">round</span>(affected_qual_vec, <span class="dv">2</span>)</span>
<span id="cb217-80"><a href="#cb217-80"></a> )</span>
<span id="cb217-81"><a href="#cb217-81"></a>}<span class="co">#SKIP</span></span>
<span id="cb217-82"><a href="#cb217-82"></a></span>
<span id="cb217-83"><a href="#cb217-83"></a><span class="co">###################</span></span>
<span id="cb217-84"><a href="#cb217-84"></a><span class="co"># min</span></span>
<span id="cb217-85"><a href="#cb217-85"></a><span class="co">####################</span></span>
<span id="cb217-86"><a href="#cb217-86"></a><span class="co"># selected feature counts</span></span>
<span id="cb217-87"><a href="#cb217-87"></a>p_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb217-88"><a href="#cb217-88"></a> <span class="kw">split</span>(p_min, Size))</span>
<span id="cb217-89"><a href="#cb217-89"></a><span class="kw">names</span>(p_min_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(p_min_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_p&#39;</span>)</span>
<span id="cb217-90"><a href="#cb217-90"></a></span>
<span id="cb217-91"><a href="#cb217-91"></a><span class="co"># get selected features that are part of enet_gene_sign_min_vec</span></span>
<span id="cb217-92"><a href="#cb217-92"></a><span class="co"># - the signature selected genes</span></span>
<span id="cb217-93"><a href="#cb217-93"></a>sign_genes_min_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(simNo_results_frm), <span class="cf">function</span>(RR)</span>
<span id="cb217-94"><a href="#cb217-94"></a>    <span class="kw">intersect</span>(<span class="kw">unlist</span>(simNo_results_frm[RR, <span class="st">&#39;genes_min&#39;</span>]), enet_gene_sign_min_vec))</span>
<span id="cb217-95"><a href="#cb217-95"></a></span>
<span id="cb217-96"><a href="#cb217-96"></a>sign_p_min_lst &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="kw">sapply</span>(sign_genes_min_lst, length), simNo_results_frm<span class="op">$</span>Size)</span>
<span id="cb217-97"><a href="#cb217-97"></a><span class="kw">names</span>(sign_p_min_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(sign_p_min_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_signP&#39;</span>)</span>
<span id="cb217-98"><a href="#cb217-98"></a></span>
<span id="cb217-99"><a href="#cb217-99"></a></span>
<span id="cb217-100"><a href="#cb217-100"></a>p_singP_min_lst &lt;-<span class="st"> </span><span class="kw">c</span>(p_min_lst, sign_p_min_lst)</span>
<span id="cb217-101"><a href="#cb217-101"></a>p_singP_min_lst &lt;-<span class="st"> </span>p_singP_min_lst[<span class="kw">order</span>(<span class="kw">names</span>(p_singP_min_lst))]</span>
<span id="cb217-102"><a href="#cb217-102"></a></span>
<span id="cb217-103"><a href="#cb217-103"></a><span class="kw">boxplot</span>(p_singP_min_lst,</span>
<span id="cb217-104"><a href="#cb217-104"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb217-105"><a href="#cb217-105"></a>  <span class="co">#ylim=c(0, 300),</span></span>
<span id="cb217-106"><a href="#cb217-106"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb217-107"><a href="#cb217-107"></a>)</span>
<span id="cb217-108"><a href="#cb217-108"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb217-109"><a href="#cb217-109"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb217-110"><a href="#cb217-110"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_lst)),</span>
<span id="cb217-111"><a href="#cb217-111"></a>  SIZE0</span>
<span id="cb217-112"><a href="#cb217-112"></a> )</span>
<span id="cb217-113"><a href="#cb217-113"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb217-114"><a href="#cb217-114"></a><span class="co">#abline(h= nzero_min_q2, col = &#39;red&#39;)</span></span>
<span id="cb217-115"><a href="#cb217-115"></a><span class="kw">legend</span>(<span class="st">&#39;topleft&#39;</span>,</span>
<span id="cb217-116"><a href="#cb217-116"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb217-117"><a href="#cb217-117"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>),</span>
<span id="cb217-118"><a href="#cb217-118"></a>   <span class="dt">legend=</span> <span class="kw">c</span>(<span class="st">&#39;selected genes&#39;</span>,<span class="st">&#39;signature genes&#39;</span>),</span>
<span id="cb217-119"><a href="#cb217-119"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb217-120"><a href="#cb217-120"></a> )</span>
<span id="cb217-121"><a href="#cb217-121"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;min lambda - selected gene counts&#39;</span>))</span>
<span id="cb217-122"><a href="#cb217-122"></a></span>
<span id="cb217-123"><a href="#cb217-123"></a>SKIP  &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb217-124"><a href="#cb217-124"></a><span class="co"># Add qual annotation</span></span>
<span id="cb217-125"><a href="#cb217-125"></a>control_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(sim_control_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb217-126"><a href="#cb217-126"></a>affected_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(sim_affected_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb217-127"><a href="#cb217-127"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb217-128"><a href="#cb217-128"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb217-129"><a href="#cb217-129"></a>  <span class="dt">at =</span>  <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_lst)),</span>
<span id="cb217-130"><a href="#cb217-130"></a>  <span class="kw">round</span>(control_qual_vec, <span class="dv">2</span>)</span>
<span id="cb217-131"><a href="#cb217-131"></a> )</span>
<span id="cb217-132"><a href="#cb217-132"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb217-133"><a href="#cb217-133"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb217-134"><a href="#cb217-134"></a>  <span class="dt">at =</span>  <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_lst)),</span>
<span id="cb217-135"><a href="#cb217-135"></a>  <span class="kw">round</span>(affected_qual_vec, <span class="dv">2</span>)</span>
<span id="cb217-136"><a href="#cb217-136"></a> )</span>
<span id="cb217-137"><a href="#cb217-137"></a>}<span class="co">#SKIP</span></span>
<span id="cb217-138"><a href="#cb217-138"></a></span>
<span id="cb217-139"><a href="#cb217-139"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, <span class="kw">paste</span>(<span class="st">&#39;Sim =&#39;</span>,  SIM))</span>
<span id="cb217-140"><a href="#cb217-140"></a></span>
<span id="cb217-141"><a href="#cb217-141"></a>} <span class="co"># for(SIM</span></span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuiteB-enet-simRes-features-bySim-1.png" alt="enet Models Selected Features by Sample Size" width="960" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuiteB-enet-simRes-features-bySim)enet Models Selected Features by Sample Size
</p>
</div>
</div>
<div id="summarize-results-across-simulation-runs" class="section level3 unnumbered" number="">
<h3 class="unnumbered" number="">Summarize results across simulation runs</h3>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb218-1"><a href="#cb218-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb218-2"><a href="#cb218-2"></a></span>
<span id="cb218-3"><a href="#cb218-3"></a><span class="co"># get full model cv error ref</span></span>
<span id="cb218-4"><a href="#cb218-4"></a>error_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_enetAll_lst,</span>
<span id="cb218-5"><a href="#cb218-5"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb218-6"><a href="#cb218-6"></a>error_1se_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(error_1se_vec, <span class="dt">prob=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>)        </span>
<span id="cb218-7"><a href="#cb218-7"></a></span>
<span id="cb218-8"><a href="#cb218-8"></a>error_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(hcc5hmC_cv_enetAll_lst,</span>
<span id="cb218-9"><a href="#cb218-9"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb218-10"><a href="#cb218-10"></a>error_min_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(error_min_vec, <span class="dt">prob=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>)        </span>
<span id="cb218-11"><a href="#cb218-11"></a></span>
<span id="cb218-12"><a href="#cb218-12"></a><span class="co"># Utility objects</span></span>
<span id="cb218-13"><a href="#cb218-13"></a>SIZE0 &lt;-<span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_pad</span>(SIZE, <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>)</span>
<span id="cb218-14"><a href="#cb218-14"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sim_control_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>,SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb218-15"><a href="#cb218-15"></a></span>
<span id="cb218-16"><a href="#cb218-16"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb218-17"><a href="#cb218-17"></a><span class="co"># 1se</span></span>
<span id="cb218-18"><a href="#cb218-18"></a><span class="co">#########################################</span></span>
<span id="cb218-19"><a href="#cb218-19"></a><span class="co">## cv</span></span>
<span id="cb218-20"><a href="#cb218-20"></a>cv_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(enet_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb218-21"><a href="#cb218-21"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb218-22"><a href="#cb218-22"></a> sizeVal_results_frm &lt;-<span class="st"> </span>enet_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb218-23"><a href="#cb218-23"></a> sizeVal_cv_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(cv_1se, SimNo))</span>
<span id="cb218-24"><a href="#cb218-24"></a> <span class="kw">sapply</span>(sizeVal_cv_1se_lst, median)</span>
<span id="cb218-25"><a href="#cb218-25"></a>})</span>
<span id="cb218-26"><a href="#cb218-26"></a><span class="kw">names</span>(cv_1se_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb218-27"><a href="#cb218-27"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(enet_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb218-28"><a href="#cb218-28"></a></span>
<span id="cb218-29"><a href="#cb218-29"></a><span class="co">## test</span></span>
<span id="cb218-30"><a href="#cb218-30"></a>test_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(enet_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb218-31"><a href="#cb218-31"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb218-32"><a href="#cb218-32"></a> sizeVal_results_frm &lt;-<span class="st"> </span>enet_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb218-33"><a href="#cb218-33"></a> sizeVal_test_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(test_1se, SimNo))</span>
<span id="cb218-34"><a href="#cb218-34"></a> <span class="kw">sapply</span>(sizeVal_test_1se_lst, median)</span>
<span id="cb218-35"><a href="#cb218-35"></a>})</span>
<span id="cb218-36"><a href="#cb218-36"></a><span class="kw">names</span>(test_1se_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb218-37"><a href="#cb218-37"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(enet_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_test&#39;</span>)</span>
<span id="cb218-38"><a href="#cb218-38"></a></span>
<span id="cb218-39"><a href="#cb218-39"></a></span>
<span id="cb218-40"><a href="#cb218-40"></a>error_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">c</span>(cv_1se_Bysize_lst, test_1se_Bysize_lst)</span>
<span id="cb218-41"><a href="#cb218-41"></a>error_1se_Bysize_lst &lt;-<span class="st"> </span>error_1se_Bysize_lst[<span class="kw">order</span>(<span class="kw">names</span>(error_1se_Bysize_lst))]</span>
<span id="cb218-42"><a href="#cb218-42"></a></span>
<span id="cb218-43"><a href="#cb218-43"></a><span class="kw">boxplot</span>(error_1se_Bysize_lst,</span>
<span id="cb218-44"><a href="#cb218-44"></a>  <span class="dt">col=</span><span class="dv">0</span>,</span>
<span id="cb218-45"><a href="#cb218-45"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb218-46"><a href="#cb218-46"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="fl">0.05</span>, <span class="fl">.5</span>),</span>
<span id="cb218-47"><a href="#cb218-47"></a>  <span class="dt">outline=</span>F,</span>
<span id="cb218-48"><a href="#cb218-48"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb218-49"><a href="#cb218-49"></a>)</span>
<span id="cb218-50"><a href="#cb218-50"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(error_1se_Bysize_lst))</span>
<span id="cb218-51"><a href="#cb218-51"></a><span class="kw">points</span>(</span>
<span id="cb218-52"><a href="#cb218-52"></a>   <span class="dt">x=</span><span class="kw">jitter</span>(<span class="kw">rep</span>(JJ, <span class="kw">length</span>(error_1se_Bysize_lst[[JJ]])), <span class="dt">amount=</span><span class="fl">0.25</span>), </span>
<span id="cb218-53"><a href="#cb218-53"></a>   <span class="dt">y=</span>error_1se_Bysize_lst[[JJ]], <span class="dt">cex=</span><span class="fl">0.5</span>,</span>
<span id="cb218-54"><a href="#cb218-54"></a>   <span class="dt">col=</span><span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">&#39;cv&#39;</span>, <span class="kw">names</span>(error_1se_Bysize_lst)[JJ]),<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb218-55"><a href="#cb218-55"></a>)</span>
<span id="cb218-56"><a href="#cb218-56"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb218-57"><a href="#cb218-57"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb218-58"><a href="#cb218-58"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_Bysize_lst)),</span>
<span id="cb218-59"><a href="#cb218-59"></a>  SIZE0</span>
<span id="cb218-60"><a href="#cb218-60"></a> )</span>
<span id="cb218-61"><a href="#cb218-61"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_Bysize_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb218-62"><a href="#cb218-62"></a><span class="kw">abline</span>(<span class="dt">h=</span> error_min_q2, <span class="dt">col =</span> <span class="st">&#39;red&#39;</span>)</span>
<span id="cb218-63"><a href="#cb218-63"></a><span class="kw">legend</span>(<span class="st">&#39;topright&#39;</span>,</span>
<span id="cb218-64"><a href="#cb218-64"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb218-65"><a href="#cb218-65"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb218-66"><a href="#cb218-66"></a>   <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&#39;cv error&#39;</span>, <span class="st">&#39;test set&#39;</span>),</span>
<span id="cb218-67"><a href="#cb218-67"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb218-68"><a href="#cb218-68"></a> )</span>
<span id="cb218-69"><a href="#cb218-69"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;one se lambda - error rates&#39;</span>))</span>
<span id="cb218-70"><a href="#cb218-70"></a></span>
<span id="cb218-71"><a href="#cb218-71"></a></span>
<span id="cb218-72"><a href="#cb218-72"></a><span class="co"># min</span></span>
<span id="cb218-73"><a href="#cb218-73"></a><span class="co">#########################################</span></span>
<span id="cb218-74"><a href="#cb218-74"></a><span class="co">## cv</span></span>
<span id="cb218-75"><a href="#cb218-75"></a>cv_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(enet_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb218-76"><a href="#cb218-76"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb218-77"><a href="#cb218-77"></a> sizeVal_results_frm &lt;-<span class="st"> </span>enet_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb218-78"><a href="#cb218-78"></a> sizeVal_cv_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(cv_min, SimNo))</span>
<span id="cb218-79"><a href="#cb218-79"></a> <span class="kw">sapply</span>(sizeVal_cv_min_lst, median)</span>
<span id="cb218-80"><a href="#cb218-80"></a>})</span>
<span id="cb218-81"><a href="#cb218-81"></a><span class="kw">names</span>(cv_min_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb218-82"><a href="#cb218-82"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(enet_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb218-83"><a href="#cb218-83"></a></span>
<span id="cb218-84"><a href="#cb218-84"></a><span class="co">## test</span></span>
<span id="cb218-85"><a href="#cb218-85"></a>test_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(enet_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb218-86"><a href="#cb218-86"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb218-87"><a href="#cb218-87"></a> sizeVal_results_frm &lt;-<span class="st"> </span>enet_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb218-88"><a href="#cb218-88"></a> sizeVal_test_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(test_min, SimNo))</span>
<span id="cb218-89"><a href="#cb218-89"></a> <span class="kw">sapply</span>(sizeVal_test_min_lst, median)</span>
<span id="cb218-90"><a href="#cb218-90"></a>})</span>
<span id="cb218-91"><a href="#cb218-91"></a><span class="kw">names</span>(test_min_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb218-92"><a href="#cb218-92"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(enet_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_test&#39;</span>)</span>
<span id="cb218-93"><a href="#cb218-93"></a></span>
<span id="cb218-94"><a href="#cb218-94"></a></span>
<span id="cb218-95"><a href="#cb218-95"></a>error_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">c</span>(cv_min_Bysize_lst, test_min_Bysize_lst)</span>
<span id="cb218-96"><a href="#cb218-96"></a>error_min_Bysize_lst &lt;-<span class="st"> </span>error_min_Bysize_lst[<span class="kw">order</span>(<span class="kw">names</span>(error_min_Bysize_lst))]</span>
<span id="cb218-97"><a href="#cb218-97"></a></span>
<span id="cb218-98"><a href="#cb218-98"></a><span class="kw">boxplot</span>(error_min_Bysize_lst,</span>
<span id="cb218-99"><a href="#cb218-99"></a>  <span class="dt">col=</span><span class="dv">0</span>,</span>
<span id="cb218-100"><a href="#cb218-100"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb218-101"><a href="#cb218-101"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="fl">0.05</span>, <span class="fl">.5</span>),</span>
<span id="cb218-102"><a href="#cb218-102"></a>  <span class="dt">outline=</span>F,</span>
<span id="cb218-103"><a href="#cb218-103"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb218-104"><a href="#cb218-104"></a>)</span>
<span id="cb218-105"><a href="#cb218-105"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(error_min_Bysize_lst))</span>
<span id="cb218-106"><a href="#cb218-106"></a><span class="kw">points</span>(</span>
<span id="cb218-107"><a href="#cb218-107"></a>   <span class="dt">x=</span><span class="kw">jitter</span>(<span class="kw">rep</span>(JJ, <span class="kw">length</span>(error_min_Bysize_lst[[JJ]])), <span class="dt">amount=</span><span class="fl">0.25</span>), </span>
<span id="cb218-108"><a href="#cb218-108"></a>   <span class="dt">y=</span>error_min_Bysize_lst[[JJ]], <span class="dt">cex=</span><span class="fl">0.5</span>,</span>
<span id="cb218-109"><a href="#cb218-109"></a>   <span class="dt">col=</span><span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">&#39;cv&#39;</span>, <span class="kw">names</span>(error_min_Bysize_lst)[JJ]),<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb218-110"><a href="#cb218-110"></a>)</span>
<span id="cb218-111"><a href="#cb218-111"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb218-112"><a href="#cb218-112"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb218-113"><a href="#cb218-113"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_Bysize_lst)),</span>
<span id="cb218-114"><a href="#cb218-114"></a>  SIZE0</span>
<span id="cb218-115"><a href="#cb218-115"></a> )</span>
<span id="cb218-116"><a href="#cb218-116"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_Bysize_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb218-117"><a href="#cb218-117"></a><span class="kw">abline</span>(<span class="dt">h=</span> error_min_q2, <span class="dt">col =</span> <span class="st">&#39;red&#39;</span>)</span>
<span id="cb218-118"><a href="#cb218-118"></a><span class="kw">legend</span>(<span class="st">&#39;topright&#39;</span>,</span>
<span id="cb218-119"><a href="#cb218-119"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb218-120"><a href="#cb218-120"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb218-121"><a href="#cb218-121"></a>   <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&#39;cv error&#39;</span>, <span class="st">&#39;test set&#39;</span>),</span>
<span id="cb218-122"><a href="#cb218-122"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb218-123"><a href="#cb218-123"></a> )</span>
<span id="cb218-124"><a href="#cb218-124"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;min lambda - error rates&#39;</span>))</span>
<span id="cb218-125"><a href="#cb218-125"></a></span>
<span id="cb218-126"><a href="#cb218-126"></a></span>
<span id="cb218-127"><a href="#cb218-127"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, <span class="kw">paste</span>(<span class="st">&#39;enet fit error rates summarized across simulations&#39;</span>))</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuiteB-enet-simRes-errors-overSim-1.png" alt="enet Model Errors by Sample Size" width="960" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuiteB-enet-simRes-errors-overSim)enet Model Errors by Sample Size
</p>
</div>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb219-1"><a href="#cb219-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb219-2"><a href="#cb219-2"></a></span>
<span id="cb219-3"><a href="#cb219-3"></a>error_1se_Bysize_sum_frm &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">sapply</span>(error_1se_Bysize_lst, <span class="cf">function</span>(LL) <span class="kw">quantile</span>(LL, <span class="dt">prob=</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>)<span class="op">/</span><span class="dv">4</span>)))</span>
<span id="cb219-4"><a href="#cb219-4"></a><span class="kw">colnames</span>(error_1se_Bysize_sum_frm) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&#39;1se_&#39;</span>, <span class="kw">colnames</span>(error_1se_Bysize_sum_frm))</span>
<span id="cb219-5"><a href="#cb219-5"></a></span>
<span id="cb219-6"><a href="#cb219-6"></a></span>
<span id="cb219-7"><a href="#cb219-7"></a>error_min_Bysize_sum_frm &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">sapply</span>(error_min_Bysize_lst, <span class="cf">function</span>(LL) <span class="kw">quantile</span>(LL, <span class="dt">prob=</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>)<span class="op">/</span><span class="dv">4</span>)))</span>
<span id="cb219-8"><a href="#cb219-8"></a><span class="kw">colnames</span>(error_min_Bysize_sum_frm) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&#39;min_&#39;</span>, <span class="kw">colnames</span>(error_min_Bysize_sum_frm))</span>
<span id="cb219-9"><a href="#cb219-9"></a></span>
<span id="cb219-10"><a href="#cb219-10"></a></span>
<span id="cb219-11"><a href="#cb219-11"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">cbind</span>(<span class="st">`</span><span class="dt">1se</span><span class="st">`</span>=error_1se_Bysize_sum_frm, <span class="dt">min=</span>error_min_Bysize_sum_frm),</span>
<span id="cb219-12"><a href="#cb219-12"></a>      <span class="dt">caption =</span> <span class="kw">paste</span>(<span class="st">&quot;elastic net error rates by sample size across simulations&quot;</span>),</span>
<span id="cb219-13"><a href="#cb219-13"></a>    <span class="dt">digits=</span><span class="dv">2</span>) <span class="op">%&gt;%</span></span>
<span id="cb219-14"><a href="#cb219-14"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-glmnetSuiteB-print-enet-simRes-errors-overSim)elastic net error rates by sample size across simulations
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
1se_25%
</th>
<th style="text-align:right;">
1se_50%
</th>
<th style="text-align:right;">
1se_75%
</th>
<th style="text-align:right;">
min_25%
</th>
<th style="text-align:right;">
min_50%
</th>
<th style="text-align:right;">
min_75%
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
025_cv
</td>
<td style="text-align:right;">
0.30
</td>
<td style="text-align:right;">
0.32
</td>
<td style="text-align:right;">
0.40
</td>
<td style="text-align:right;">
0.26
</td>
<td style="text-align:right;">
0.29
</td>
<td style="text-align:right;">
0.38
</td>
</tr>
<tr>
<td style="text-align:left;">
025_test
</td>
<td style="text-align:right;">
0.27
</td>
<td style="text-align:right;">
0.30
</td>
<td style="text-align:right;">
0.36
</td>
<td style="text-align:right;">
0.28
</td>
<td style="text-align:right;">
0.32
</td>
<td style="text-align:right;">
0.37
</td>
</tr>
<tr>
<td style="text-align:left;">
050_cv
</td>
<td style="text-align:right;">
0.20
</td>
<td style="text-align:right;">
0.23
</td>
<td style="text-align:right;">
0.28
</td>
<td style="text-align:right;">
0.18
</td>
<td style="text-align:right;">
0.21
</td>
<td style="text-align:right;">
0.25
</td>
</tr>
<tr>
<td style="text-align:left;">
050_test
</td>
<td style="text-align:right;">
0.22
</td>
<td style="text-align:right;">
0.24
</td>
<td style="text-align:right;">
0.26
</td>
<td style="text-align:right;">
0.21
</td>
<td style="text-align:right;">
0.24
</td>
<td style="text-align:right;">
0.26
</td>
</tr>
<tr>
<td style="text-align:left;">
100_cv
</td>
<td style="text-align:right;">
0.16
</td>
<td style="text-align:right;">
0.18
</td>
<td style="text-align:right;">
0.20
</td>
<td style="text-align:right;">
0.15
</td>
<td style="text-align:right;">
0.17
</td>
<td style="text-align:right;">
0.18
</td>
</tr>
<tr>
<td style="text-align:left;">
100_test
</td>
<td style="text-align:right;">
0.17
</td>
<td style="text-align:right;">
0.18
</td>
<td style="text-align:right;">
0.19
</td>
<td style="text-align:right;">
0.16
</td>
<td style="text-align:right;">
0.17
</td>
<td style="text-align:right;">
0.18
</td>
</tr>
<tr>
<td style="text-align:left;">
200_cv
</td>
<td style="text-align:right;">
0.12
</td>
<td style="text-align:right;">
0.13
</td>
<td style="text-align:right;">
0.13
</td>
<td style="text-align:right;">
0.11
</td>
<td style="text-align:right;">
0.12
</td>
<td style="text-align:right;">
0.12
</td>
</tr>
<tr>
<td style="text-align:left;">
200_test
</td>
<td style="text-align:right;">
0.11
</td>
<td style="text-align:right;">
0.12
</td>
<td style="text-align:right;">
0.13
</td>
<td style="text-align:right;">
0.11
</td>
<td style="text-align:right;">
0.12
</td>
<td style="text-align:right;">
0.12
</td>
</tr>
<tr>
<td style="text-align:left;">
300_cv
</td>
<td style="text-align:right;">
0.10
</td>
<td style="text-align:right;">
0.10
</td>
<td style="text-align:right;">
0.11
</td>
<td style="text-align:right;">
0.09
</td>
<td style="text-align:right;">
0.10
</td>
<td style="text-align:right;">
0.10
</td>
</tr>
<tr>
<td style="text-align:left;">
300_test
</td>
<td style="text-align:right;">
0.09
</td>
<td style="text-align:right;">
0.10
</td>
<td style="text-align:right;">
0.10
</td>
<td style="text-align:right;">
0.08
</td>
<td style="text-align:right;">
0.09
</td>
<td style="text-align:right;">
0.10
</td>
</tr>
</tbody>
</table>
<p>Now look at feature selection.</p>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb220-1"><a href="#cb220-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb220-2"><a href="#cb220-2"></a></span>
<span id="cb220-3"><a href="#cb220-3"></a><span class="co"># Utility objects</span></span>
<span id="cb220-4"><a href="#cb220-4"></a>SIZE0 &lt;-<span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_pad</span>(SIZE, <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>)</span>
<span id="cb220-5"><a href="#cb220-5"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sim_control_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>,SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb220-6"><a href="#cb220-6"></a></span>
<span id="cb220-7"><a href="#cb220-7"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb220-8"><a href="#cb220-8"></a><span class="co"># 1se</span></span>
<span id="cb220-9"><a href="#cb220-9"></a><span class="co">#########################################</span></span>
<span id="cb220-10"><a href="#cb220-10"></a><span class="co"># selected features</span></span>
<span id="cb220-11"><a href="#cb220-11"></a>p_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(enet_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb220-12"><a href="#cb220-12"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb220-13"><a href="#cb220-13"></a> sizeVal_results_frm &lt;-<span class="st"> </span>enet_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb220-14"><a href="#cb220-14"></a> sizeVal_p_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(p_1se, SimNo))</span>
<span id="cb220-15"><a href="#cb220-15"></a> <span class="kw">sapply</span>(sizeVal_p_1se_lst, median)</span>
<span id="cb220-16"><a href="#cb220-16"></a>})</span>
<span id="cb220-17"><a href="#cb220-17"></a><span class="kw">names</span>(p_1se_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb220-18"><a href="#cb220-18"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(enet_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_p&#39;</span>)</span>
<span id="cb220-19"><a href="#cb220-19"></a></span>
<span id="cb220-20"><a href="#cb220-20"></a><span class="co"># selected signatue features</span></span>
<span id="cb220-21"><a href="#cb220-21"></a>sign_p_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(enet_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb220-22"><a href="#cb220-22"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb220-23"><a href="#cb220-23"></a> sizeVal_results_frm &lt;-<span class="st"> </span>enet_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb220-24"><a href="#cb220-24"></a> </span>
<span id="cb220-25"><a href="#cb220-25"></a>  </span>
<span id="cb220-26"><a href="#cb220-26"></a> sizeVal_sign_genes_1se_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sizeVal_results_frm), <span class="cf">function</span>(RR)</span>
<span id="cb220-27"><a href="#cb220-27"></a>    <span class="kw">intersect</span>(<span class="kw">unlist</span>(sizeVal_results_frm[RR, <span class="st">&#39;genes_1se&#39;</span>]), enet_gene_sign_1se_vec))</span>
<span id="cb220-28"><a href="#cb220-28"></a></span>
<span id="cb220-29"><a href="#cb220-29"></a> sizeVal_sign_p_1se_lst &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="kw">sapply</span>(sizeVal_sign_genes_1se_lst, length),</span>
<span id="cb220-30"><a href="#cb220-30"></a>    sizeVal_results_frm<span class="op">$</span>SimNo)</span>
<span id="cb220-31"><a href="#cb220-31"></a> </span>
<span id="cb220-32"><a href="#cb220-32"></a> <span class="kw">sapply</span>(sizeVal_sign_p_1se_lst, median)</span>
<span id="cb220-33"><a href="#cb220-33"></a>})</span>
<span id="cb220-34"><a href="#cb220-34"></a><span class="kw">names</span>(sign_p_1se_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb220-35"><a href="#cb220-35"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(enet_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_signP&#39;</span>)</span>
<span id="cb220-36"><a href="#cb220-36"></a></span>
<span id="cb220-37"><a href="#cb220-37"></a></span>
<span id="cb220-38"><a href="#cb220-38"></a>p_singP_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">c</span>(p_1se_Bysize_lst, sign_p_1se_Bysize_lst)</span>
<span id="cb220-39"><a href="#cb220-39"></a>p_singP_1se_Bysize_lst &lt;-<span class="st"> </span>p_singP_1se_Bysize_lst[<span class="kw">order</span>(<span class="kw">names</span>(p_singP_1se_Bysize_lst))]</span>
<span id="cb220-40"><a href="#cb220-40"></a></span>
<span id="cb220-41"><a href="#cb220-41"></a><span class="kw">boxplot</span>(p_singP_1se_Bysize_lst,</span>
<span id="cb220-42"><a href="#cb220-42"></a>  <span class="dt">col=</span><span class="dv">0</span>,</span>
<span id="cb220-43"><a href="#cb220-43"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb220-44"><a href="#cb220-44"></a>  <span class="co">#ylim=c(0, 300),</span></span>
<span id="cb220-45"><a href="#cb220-45"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb220-46"><a href="#cb220-46"></a>)</span>
<span id="cb220-47"><a href="#cb220-47"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(p_singP_1se_Bysize_lst))</span>
<span id="cb220-48"><a href="#cb220-48"></a><span class="kw">points</span>(</span>
<span id="cb220-49"><a href="#cb220-49"></a>   <span class="dt">x=</span><span class="kw">jitter</span>(<span class="kw">rep</span>(JJ, <span class="kw">length</span>(p_singP_1se_Bysize_lst[[JJ]])), <span class="dt">amount=</span><span class="fl">0.25</span>),</span>
<span id="cb220-50"><a href="#cb220-50"></a>   <span class="dt">y=</span>p_singP_1se_Bysize_lst[[JJ]], <span class="dt">cex=</span><span class="fl">0.5</span>,</span>
<span id="cb220-51"><a href="#cb220-51"></a>   <span class="dt">col=</span><span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">&#39;_p&#39;</span>, <span class="kw">names</span>(p_singP_1se_Bysize_lst)[JJ]),<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb220-52"><a href="#cb220-52"></a>)</span>
<span id="cb220-53"><a href="#cb220-53"></a></span>
<span id="cb220-54"><a href="#cb220-54"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb220-55"><a href="#cb220-55"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb220-56"><a href="#cb220-56"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_Bysize_lst)),</span>
<span id="cb220-57"><a href="#cb220-57"></a>  SIZE0</span>
<span id="cb220-58"><a href="#cb220-58"></a> )</span>
<span id="cb220-59"><a href="#cb220-59"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_Bysize_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb220-60"><a href="#cb220-60"></a><span class="co">#abline(h= nzero_1se_q2, col = &#39;red&#39;)</span></span>
<span id="cb220-61"><a href="#cb220-61"></a><span class="kw">legend</span>(<span class="st">&#39;topleft&#39;</span>,</span>
<span id="cb220-62"><a href="#cb220-62"></a>   <span class="co">#title=&#39;1se errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb220-63"><a href="#cb220-63"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>),</span>
<span id="cb220-64"><a href="#cb220-64"></a>   <span class="dt">legend=</span> <span class="kw">c</span>(<span class="st">&#39;selected genes&#39;</span>,<span class="st">&#39;signature genes&#39;</span>),</span>
<span id="cb220-65"><a href="#cb220-65"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb220-66"><a href="#cb220-66"></a> )</span>
<span id="cb220-67"><a href="#cb220-67"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;one se lamdba - selected gene counts&#39;</span>))</span>
<span id="cb220-68"><a href="#cb220-68"></a></span>
<span id="cb220-69"><a href="#cb220-69"></a></span>
<span id="cb220-70"><a href="#cb220-70"></a><span class="co"># min</span></span>
<span id="cb220-71"><a href="#cb220-71"></a><span class="co">#########################################</span></span>
<span id="cb220-72"><a href="#cb220-72"></a><span class="co"># selected features</span></span>
<span id="cb220-73"><a href="#cb220-73"></a>p_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(enet_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb220-74"><a href="#cb220-74"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb220-75"><a href="#cb220-75"></a> sizeVal_results_frm &lt;-<span class="st"> </span>enet_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb220-76"><a href="#cb220-76"></a> sizeVal_p_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(p_min, SimNo))</span>
<span id="cb220-77"><a href="#cb220-77"></a> <span class="kw">sapply</span>(sizeVal_p_min_lst, median)</span>
<span id="cb220-78"><a href="#cb220-78"></a>})</span>
<span id="cb220-79"><a href="#cb220-79"></a><span class="kw">names</span>(p_min_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb220-80"><a href="#cb220-80"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(enet_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_p&#39;</span>)</span>
<span id="cb220-81"><a href="#cb220-81"></a></span>
<span id="cb220-82"><a href="#cb220-82"></a><span class="co"># selected signatue features</span></span>
<span id="cb220-83"><a href="#cb220-83"></a>sign_p_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(enet_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb220-84"><a href="#cb220-84"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb220-85"><a href="#cb220-85"></a> sizeVal_results_frm &lt;-<span class="st"> </span>enet_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb220-86"><a href="#cb220-86"></a> </span>
<span id="cb220-87"><a href="#cb220-87"></a>  </span>
<span id="cb220-88"><a href="#cb220-88"></a> sizeVal_sign_genes_min_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sizeVal_results_frm), <span class="cf">function</span>(RR)</span>
<span id="cb220-89"><a href="#cb220-89"></a>    <span class="kw">intersect</span>(<span class="kw">unlist</span>(sizeVal_results_frm[RR, <span class="st">&#39;genes_min&#39;</span>]), enet_gene_sign_min_vec))</span>
<span id="cb220-90"><a href="#cb220-90"></a></span>
<span id="cb220-91"><a href="#cb220-91"></a> sizeVal_sign_p_min_lst &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="kw">sapply</span>(sizeVal_sign_genes_min_lst, length),</span>
<span id="cb220-92"><a href="#cb220-92"></a>    sizeVal_results_frm<span class="op">$</span>SimNo)</span>
<span id="cb220-93"><a href="#cb220-93"></a> </span>
<span id="cb220-94"><a href="#cb220-94"></a> <span class="kw">sapply</span>(sizeVal_sign_p_min_lst, median)</span>
<span id="cb220-95"><a href="#cb220-95"></a>})</span>
<span id="cb220-96"><a href="#cb220-96"></a><span class="kw">names</span>(sign_p_min_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb220-97"><a href="#cb220-97"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(enet_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_signP&#39;</span>)</span>
<span id="cb220-98"><a href="#cb220-98"></a></span>
<span id="cb220-99"><a href="#cb220-99"></a></span>
<span id="cb220-100"><a href="#cb220-100"></a>p_singP_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">c</span>(p_min_Bysize_lst, sign_p_min_Bysize_lst)</span>
<span id="cb220-101"><a href="#cb220-101"></a>p_singP_min_Bysize_lst &lt;-<span class="st"> </span>p_singP_min_Bysize_lst[<span class="kw">order</span>(<span class="kw">names</span>(p_singP_min_Bysize_lst))]</span>
<span id="cb220-102"><a href="#cb220-102"></a></span>
<span id="cb220-103"><a href="#cb220-103"></a><span class="kw">boxplot</span>(p_singP_min_Bysize_lst,</span>
<span id="cb220-104"><a href="#cb220-104"></a>  <span class="dt">col=</span><span class="dv">0</span>,</span>
<span id="cb220-105"><a href="#cb220-105"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb220-106"><a href="#cb220-106"></a>  <span class="co">#ylim=c(0, 300),</span></span>
<span id="cb220-107"><a href="#cb220-107"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb220-108"><a href="#cb220-108"></a>)</span>
<span id="cb220-109"><a href="#cb220-109"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(p_singP_min_Bysize_lst))</span>
<span id="cb220-110"><a href="#cb220-110"></a><span class="kw">points</span>(</span>
<span id="cb220-111"><a href="#cb220-111"></a>   <span class="dt">x=</span><span class="kw">jitter</span>(<span class="kw">rep</span>(JJ, <span class="kw">length</span>(p_singP_min_Bysize_lst[[JJ]])), <span class="dt">amount=</span><span class="fl">0.25</span>),</span>
<span id="cb220-112"><a href="#cb220-112"></a>   <span class="dt">y=</span>p_singP_min_Bysize_lst[[JJ]], <span class="dt">cex=</span><span class="fl">0.5</span>,</span>
<span id="cb220-113"><a href="#cb220-113"></a>   <span class="dt">col=</span><span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">&#39;_p&#39;</span>, <span class="kw">names</span>(p_singP_min_Bysize_lst)[JJ]),<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb220-114"><a href="#cb220-114"></a>)</span>
<span id="cb220-115"><a href="#cb220-115"></a></span>
<span id="cb220-116"><a href="#cb220-116"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb220-117"><a href="#cb220-117"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb220-118"><a href="#cb220-118"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_Bysize_lst)),</span>
<span id="cb220-119"><a href="#cb220-119"></a>  SIZE0</span>
<span id="cb220-120"><a href="#cb220-120"></a> )</span>
<span id="cb220-121"><a href="#cb220-121"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_Bysize_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb220-122"><a href="#cb220-122"></a><span class="co">#abline(h= nzero_min_q2, col = &#39;red&#39;)</span></span>
<span id="cb220-123"><a href="#cb220-123"></a><span class="kw">legend</span>(<span class="st">&#39;topleft&#39;</span>,</span>
<span id="cb220-124"><a href="#cb220-124"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb220-125"><a href="#cb220-125"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>),</span>
<span id="cb220-126"><a href="#cb220-126"></a>   <span class="dt">legend=</span> <span class="kw">c</span>(<span class="st">&#39;selected genes&#39;</span>,<span class="st">&#39;signature genes&#39;</span>),</span>
<span id="cb220-127"><a href="#cb220-127"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb220-128"><a href="#cb220-128"></a> )</span>
<span id="cb220-129"><a href="#cb220-129"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;min lambda - selected gene counts&#39;</span>))</span>
<span id="cb220-130"><a href="#cb220-130"></a></span>
<span id="cb220-131"><a href="#cb220-131"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, <span class="kw">paste</span>(<span class="st">&#39;enet fit feature selection summarized across simulations&#39;</span>))</span></code></pre></div>
<div class="figure">
<img src="Static/figures/hcc5hmC-glmnetSuiteB-enet-simRes-features-OverSim-1.png" alt="enet Models Selected Features by Sample Size" width="960" />
<p class="caption">
(#fig:hcc5hmC-glmnetSuiteB-enet-simRes-features-OverSim)enet Models Selected Features by Sample Size
</p>
</div>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb221-1"><a href="#cb221-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb221-2"><a href="#cb221-2"></a></span>
<span id="cb221-3"><a href="#cb221-3"></a>p_sing_1se_Bysize_sum_frm &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">sapply</span>(p_singP_1se_Bysize_lst, <span class="cf">function</span>(LL) <span class="kw">quantile</span>(LL, <span class="dt">prob=</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>)<span class="op">/</span><span class="dv">4</span>)))</span>
<span id="cb221-4"><a href="#cb221-4"></a><span class="kw">colnames</span>(p_sing_1se_Bysize_sum_frm) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&#39;1se_&#39;</span>, <span class="kw">colnames</span>(p_sing_1se_Bysize_sum_frm))</span>
<span id="cb221-5"><a href="#cb221-5"></a></span>
<span id="cb221-6"><a href="#cb221-6"></a>p_sing_min_Bysize_sum_frm &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">sapply</span>(p_singP_min_Bysize_lst, <span class="cf">function</span>(LL) <span class="kw">quantile</span>(LL, <span class="dt">prob=</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>)<span class="op">/</span><span class="dv">4</span>)))</span>
<span id="cb221-7"><a href="#cb221-7"></a><span class="kw">colnames</span>(p_sing_min_Bysize_sum_frm) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&#39;min_&#39;</span>, <span class="kw">colnames</span>(p_sing_min_Bysize_sum_frm))</span>
<span id="cb221-8"><a href="#cb221-8"></a></span>
<span id="cb221-9"><a href="#cb221-9"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">cbind</span>(p_sing_1se_Bysize_sum_frm, p_sing_min_Bysize_sum_frm),</span>
<span id="cb221-10"><a href="#cb221-10"></a>    <span class="dt">caption =</span> <span class="kw">paste</span>(<span class="st">&quot;elastic net feature selection by sample size across simulations&quot;</span>),</span>
<span id="cb221-11"><a href="#cb221-11"></a>    <span class="dt">digits=</span><span class="dv">2</span>) <span class="op">%&gt;%</span></span>
<span id="cb221-12"><a href="#cb221-12"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
(#tab:hcc5hmC-glmnetSuiteB-print-enet-simRes-features-OverSim)elastic net feature selection by sample size across simulations
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
1se_25%
</th>
<th style="text-align:right;">
1se_50%
</th>
<th style="text-align:right;">
1se_75%
</th>
<th style="text-align:right;">
min_25%
</th>
<th style="text-align:right;">
min_50%
</th>
<th style="text-align:right;">
min_75%
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
025_p
</td>
<td style="text-align:right;">
3.00
</td>
<td style="text-align:right;">
8.50
</td>
<td style="text-align:right;">
14.75
</td>
<td style="text-align:right;">
6.25
</td>
<td style="text-align:right;">
15.0
</td>
<td style="text-align:right;">
27.75
</td>
</tr>
<tr>
<td style="text-align:left;">
025_signP
</td>
<td style="text-align:right;">
1.00
</td>
<td style="text-align:right;">
1.25
</td>
<td style="text-align:right;">
2.75
</td>
<td style="text-align:right;">
2.00
</td>
<td style="text-align:right;">
3.0
</td>
<td style="text-align:right;">
3.75
</td>
</tr>
<tr>
<td style="text-align:left;">
050_p
</td>
<td style="text-align:right;">
5.00
</td>
<td style="text-align:right;">
8.50
</td>
<td style="text-align:right;">
16.00
</td>
<td style="text-align:right;">
18.25
</td>
<td style="text-align:right;">
31.0
</td>
<td style="text-align:right;">
44.12
</td>
</tr>
<tr>
<td style="text-align:left;">
050_signP
</td>
<td style="text-align:right;">
3.00
</td>
<td style="text-align:right;">
4.00
</td>
<td style="text-align:right;">
5.75
</td>
<td style="text-align:right;">
5.00
</td>
<td style="text-align:right;">
7.5
</td>
<td style="text-align:right;">
9.38
</td>
</tr>
<tr>
<td style="text-align:left;">
100_p
</td>
<td style="text-align:right;">
17.25
</td>
<td style="text-align:right;">
22.50
</td>
<td style="text-align:right;">
36.00
</td>
<td style="text-align:right;">
42.25
</td>
<td style="text-align:right;">
55.0
</td>
<td style="text-align:right;">
65.88
</td>
</tr>
<tr>
<td style="text-align:left;">
100_signP
</td>
<td style="text-align:right;">
8.12
</td>
<td style="text-align:right;">
10.50
</td>
<td style="text-align:right;">
14.00
</td>
<td style="text-align:right;">
15.25
</td>
<td style="text-align:right;">
16.0
</td>
<td style="text-align:right;">
19.00
</td>
</tr>
<tr>
<td style="text-align:left;">
200_p
</td>
<td style="text-align:right;">
38.00
</td>
<td style="text-align:right;">
54.50
</td>
<td style="text-align:right;">
63.25
</td>
<td style="text-align:right;">
95.00
</td>
<td style="text-align:right;">
105.0
</td>
<td style="text-align:right;">
128.75
</td>
</tr>
<tr>
<td style="text-align:left;">
200_signP
</td>
<td style="text-align:right;">
20.00
</td>
<td style="text-align:right;">
22.50
</td>
<td style="text-align:right;">
24.88
</td>
<td style="text-align:right;">
34.00
</td>
<td style="text-align:right;">
37.0
</td>
<td style="text-align:right;">
41.38
</td>
</tr>
<tr>
<td style="text-align:left;">
300_p
</td>
<td style="text-align:right;">
50.00
</td>
<td style="text-align:right;">
67.75
</td>
<td style="text-align:right;">
71.88
</td>
<td style="text-align:right;">
95.88
</td>
<td style="text-align:right;">
111.5
</td>
<td style="text-align:right;">
152.75
</td>
</tr>
<tr>
<td style="text-align:left;">
300_signP
</td>
<td style="text-align:right;">
30.00
</td>
<td style="text-align:right;">
32.50
</td>
<td style="text-align:right;">
36.75
</td>
<td style="text-align:right;">
48.12
</td>
<td style="text-align:right;">
55.0
</td>
<td style="text-align:right;">
61.50
</td>
</tr>
</tbody>
</table>
<!--chapter:end:A2-references.Rmd-->
</div>
</div>
</div>
<div class="footnotes">
<hr />
<ol>
<li id="fn1"><p>See <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE96058">NCBI GEO GSE96058</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
<!--bookdown:body:end-->
            </section>

          </div>
        </div>
      </div>
<!--bookdown:link_prev-->
<!--bookdown:link_next-->
    </div>
  </div>
<!--bookdown:config-->

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
