--- 
title: "DNA Hydroxymethylation in Hepatocellular Carcinoma"
author: "Francois Collin"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
### see https://community.rstudio.com/t/bookdown-pdf-generation/12359
knit: "bookdown::render_book"  
documentclass: book
bibliography: [bib/HCC5hmc.bib]
#biblio-style: apalike
csl: csl/cell-numeric.csl
#csl: csl/american-medical-association-alphabetical.csl
link-citations: yes
description: "Data from Cai et al. (2019) paper are explored"
---

<!-- ONLY THIS FILE SHOULD HAVE YAML -->

<!-- THIS FILE DOESN'T HAVE TO HAVE ANY CONTENT ... -->
 

<style>

.watermark {
  opacity: 0.2;
  position: fixed;
  top: 50%;
  left: 50%;
  font-size: 500%;
  color: #00407d;
}

</style>

<!-- THIS DIDN'T DO ANYTHING
<div class="watermark">DRAFT</div>
-->

```{r setup}
   # file rmarkdown file management options: cache, figures
 figures_DIR <- file.path('Static', 'figures/')
 suppressMessages(dir.create(figures_DIR, recursive=T))
 knitr::opts_chunk$set(fig.path=paste0(figures_DIR))
```


<!--chapter:end:index.Rmd-->

# Preamble {.unnumbered #index} 


This vignette offers some exploratory data analyses of data available from 
the [NCBI GEO web site](https://www.ncbi.nlm.nih.gov/geo/).  

## License {-}

<!-- From https://github.com/santisoler/cc-licenses -->
<!-- THis doesnt work with pdf -->
<!-- COMMENT OUT FOR bookdown::pdf_book ????
![](CC_4_0.png)
![](https://i.creativecommons.org/l/by/4.0/88x31.png)

-->

`r knitr::include_graphics(
  "Static/images/CC_4_0.png",  dpi=100)`


This work by Francois Collin is licensed under a
[Creative Commons Attribution 4.0 International License](http://creativecommons.org/licenses/by/4.0/)



# Introduction {#intro}


The goal of detecting
cancer at the earliest stage of development with a non-invasive procedure
has busied many groups with the task of perfecting techniques to support
what has become commonly known as a 
liquid biopsy - the analysis of biomarkers circulating in fluids such as blood,
saliva or urine.  Epigenetic biomarkers present themselves as good candidates for this application
(Gai and Sun (2019) [@Gai:2019aa]).  In particular,
given their prevalence in the human genome, 
close correlation with gene expression and high chemical stability,
DNA modifications such as 5-methylcytosine (5mC) and 5-hydroxymethylcytosine (5hmC)
are DNA epigenetic marks that provide much promise as
cancer diagnosis biomarkers that could be profitably analyzed in liquid biopsies
[@Cai:2019aa; @Li:2017aa; @Song:2017aa; @Collin:2018aa].


<!--
This work has already led to some commercial products.
Thrive Earlier Detection Corp.  launched CancerSEEK, a liquid biopsy test to detect multiple cancers early in 2019.  Guardant Health offers a liquid biopsy test, Guardant360, for advanced solid 
tumor cancers. Others in the space include Karius, which focuses on a liquid biopsy for 
infectious disease, and GRAIL Bio, which was launched by Illumina in January 2016. 
In 2017, Verily Life Sciences, one of Google/Alphabet’s companies, invested in Freenome, 
another liquid biopsy company.
-->
Li et al. (2017) [@Li:2017aa] used a sensitive and selective chemical labeling technology
to extract genome-wide 5hmC profiles from circulating cell-free DNA (cfDNA) 
as well as from genomic DNA (gDNA) 
collected from a cohort of 260 patients recently diagnosed with colorectal, 
gastric, pancreatic, liver or thyroid cancer and normal tissues from 90 healthy individuals
They found 5hmC-based biomarkers of circulating cfDNA to be highly predictive of some cancer types.
Similar small sample size findings were reported in Song et al. (2017) [@Song:2017aa].  

Focusing on hepatocellular carcinoma, Cai et al. (2019) [@Cai:2019aa] assembled a sizable dataset
to demonstrate the feasibility of using features derived from 
5-hydroxymethylcytosines marks in circulating cell-free DNA as 
a non-invasive approach for the early detection of
hepatocellular carcinoma.  The data that are the basis of that
report are available on the NCBI GEO web site 
([Series GSE112679](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE112679)).
The data have also been bundled in a R data package which can be installed from github:

```
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
devtools::install_github("12379Monty/GSE112679")
```

An important question in the early development of classifiers of the sorts
that are the basis of any liquid biopsy diagnostic tool is how many samples
should be collected to make properly informed decisions.  In this
report we will explore the GSE112679 data to shed some light on
the relationship between sample size and model performance
in the context classifying samples based on 5hmC data.


In Section \@ref(preproc) we preprocess the data that
we will use for the classification analysis and perform some light QC analyses.
In Section \@ref(modeling-background) we provide some background to our
modeling approach.
In Section \@ref(explore-sparsity-fits) we explore some glmnet fits
that discriminate between early stage HCC and control samples.
In Section \@ref(model-suite) we examine the results of fitting a suite of models to
investigate the effect of sample size on model performance.



<!--chapter:end:00-intro.Rmd-->

# Preprocessing {#preproc}

<!--
 FN <- 'tmp'
 # Shotcuts for knitting and redering while in R session (Invoke interactive R from R/Scripts folder)
 kk <- function(n='') knitr::knit2html(paste("t", n, sep=''), envir=globalenv(),
       output=paste(FN,".html", sep=''))

 rr <- function(n='') rmarkdown::render(paste("t", n, sep=''), envir=globalenv(),
       output_file=paste(FN,".html", sep='')) ##, output_dir='Scripts')

 bb <- function(n='') browseURL(paste(FN,".html", sep=''))

 # The usual shotcuts
 zz <- function(n='') source(paste("t", n, sep=''))
-->



```{r setup-preproc, include=F}
   # file rmarkdown file management options: cache, figures
 figures_DIR <- file.path('Static', 'figures/')
 suppressMessages(dir.create(figures_DIR, recursive=T))
 knitr::opts_chunk$set(fig.path=paste0(figures_DIR))
```

## Load the data

<!-- THIS ENSURES NO EVALUATION TAKES PLACE BY DEFAULT -->
<!-- TO TURN ON, SET eval=T                            -->
```{r chunk-options, include=FALSE, eval=F}
library("knitr")
opts_chunk$set(eval = FALSE)
```

<!-- Add base libraries -->
```{r libraries, include=FALSE, eval=T}
library("magrittr")
```


The data that are available from NCBI GEO
[Series GSE112679](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE112679)
can be conveniently accessed through an R data package.
Attaching the GSE112679 package makes the count data tables 
available as well as a gene annotation table and a sample description table.
See [GSE112679 R Data Package page](https://12379monty.github.io/GSE112679/).
For the Cai et al. [@Cai:2019aa] model fitting and analysis, samples were separated into
`Train` and `Val-1` subsets.  `Val-2` was an external validation set.

```{r loadData, cache=F}

if (!("GSE112679" %in% rownames(installed.packages()))) {
  if (!requireNamespace("devtools", quietly = TRUE)) {
    install.packages("devtools")
  }
  devtools::install_github("12379Monty/GSE112679")
}
library(GSE112679)
sampDesc$DxStage <- with(sampDesc, ifelse(outcome=='HCC', 
   paste0(outcome,':',stage), outcome))

with(
  sampDesc %>% dplyr::filter(sampType == "blood"),
  knitr::kable(table(DxStage, trainValGroup, exclude = NULL),
   caption="GSE112679 Samples by Dx Group and Subset") %>% 
  kableExtra::kable_styling(full_width = F)
)

```

For this analysis, we will consider early stage cancer samples
and healthy or benign samples from the `Train` or `Val-1` subsets.
The appropriate outcome variable will be renamed or aliased `group`

```{r subsetSamples, cache=T, cache.vars=c('sampDescA','groupCol'), echo=F}

# Use suffix 'A' for Analysis samples
sampDescA <-
  sampDesc %>%
  dplyr::filter(sampType == "blood" &
    (trainValGroup %in% c("Train", "Val-1")) &
    ((outcome2 == "BenignHealthy") |
      (outcome2 == "HCC" & stage == "Early"))) %>%
  dplyr::rename(group = outcome2) %>%
  dplyr::arrange(group, sampID)
# Recode group
sampDescA$group <- with(
  sampDescA,
  ifelse(group == "BenignHealthy", "Control", group)
)
# set groupCol for later
groupCol <- c("#F3C300", "#875692")
names(groupCol) <- unique(sampDescA$group)

with(sampDescA, 
 knitr::kable(table(group, exclude = NULL),
  caption="Samples used in this analysis") %>%
  kableExtra::kable_styling(full_width = F)
)

```


The features are counts of reads captured by chemical labeling, and indicate
the level of 5-hydroxymethylcytosines within each gene body.  Cai et al. (2019),
Li et al. (2017) and Song et al. (2017) [@Cai:2019aa;@Li:2017aa;@Song:2017aa]
all analyze 5hmC gene body counts using standard RNA-Seq methodologies, and we will
do the same here.  

Note that before conducting any substantive analyses, the data would normally
be very carefully examined for any sign of quality variation between groups
of samples.  This analysis would integrate sample meta data - where and when were
the blood samples collected - as well as library preparation and sequencing metrics
in order to detect any sign of processing artifacts that may be present in the dataset.
This is particularly important when dealing with blood samples as variable
DNA quality degradation is a well known challenge that is encountered when dealing with
such samples [@Huang:2017aa].  Although blood specimen handling protocols can be 
put in place to minimize quality variation [@Permenter:2015aa], variability
can never be completely eradicated, especially in the context of blood samples
collected by different groups, working in different environments.  The problem
of variable DNA quality becomes paricularly pernicuous when it is compounded
with a confounding factor that sneaks in when the control sample collection
events are separated in time and space from the cancer sample collection events;
an all too common occurence.  

As proper data QC requires an intimate familiarity with the details of
data collection and processing, such a task cannot be untertaken here.
We will simply run a *minimal set of QC sanity checks* to make sure that
there are no apparent systematic effects in the data.


```{r getfeatures, cache=T, cache.vars=c('featureCountsA'), echo=F}

featureCountsA <- cbind(
  Train_featureCount,
  Val1_featureCount,
  Val2_featureCount
)[, rownames(sampDescA)]

```

We first look at coverage - make sure there isn't too much disparity of coverage 
across samples. To detect shared variability, samples can be annotated and ordered
according to sample features that may be linked to sample batch processing.  Here we 
the samples have been ordered by group and sample id (an alias of geoAcc).

```{r lcpmBxp, cache=T, fig.height=4, fig.width=10, fig.cap='Sample log2 count boxplots', echo=F}

par(mar = c(1, 3, 2, 1))
boxplot(log2(featureCountsA + 1),
  ylim = c(3, 11), ylab='log2 Count',
  staplewex = 0,       # remove horizontal whisker lines
  staplecol = "white", # just to be totally sure :)
  outline = F,         # remove outlying points
  whisklty = 0,        # remove vertical whisker lines
  las = 2, horizontal = F, xaxt = "n",
  border = groupCol[sampDescA$group]
)
legend("top", legend = names(groupCol), text.col = groupCol, 
  ncol = 2, bty = "n")
# Add reference lines
SampleMedian <- apply(log2(featureCountsA + 1), 2, median)
abline(h = median(SampleMedian), col = "grey")
axis(side = 4, at = round(median(SampleMedian), 1), 
  las = 2, col = "grey", line = -1, tick = F)

```

<!--
Coverage level looks fairly comparable across samples.  It is sometimes helpful to
keep track of the actual coverage which can be adequetely tracked by distribution
quantiles.
-->

```{r quantCountsA, echo=F}

featureCountsA_quant <- apply(featureCountsA, 2, function(CC) {
  c(quantile(CC, prob = c(.15, (1:3) / 4)), totCovM = sum(CC) / 1e6)
})

featureCountsA_quant2 <- apply(featureCountsA_quant, 1, function(RR) {
  quantile(RR, prob = (1:3) / 4)
})

knitr::kable(featureCountsA_quant2,
  digits = 1,
  caption = paste(
    "Coverage Summary - Columns are sample coverage quantiles and total coverage",
    "\nRows are quartiles across samples"
  )
) %>% kableExtra::kable_styling(full_width = F)

```


From this table, we see that 25% of the samples have total coverage exceeding
`r round(featureCountsA_quant2["75%", "totCovM"],1)`M reads, 25% of samples
have a 15 percentile of coverage lower than
`r featureCountsA_quant2["25%", "15%"]`, etc.  


<!-- SKIP
We next look at relative log representation (RLR) (in the context of measuring the density of 
5hmC marks in genes, we refer to `representation` as opposed to `expression`; the
two can be used interchangibly) -
make sure the shapes of the distributions are not widely different.
-->
```{r rlr, cache=T, cache.vars='lcpm_mtx', fig.height=4, fig.width=10, fig.cap='Sample RLR', eval=T, echo=F,include=F}

lcpm_mtx <- edgeR::cpm(featureCountsA, log = T)
median_vec <- apply(lcpm_mtx, 1, median)
RLR_mtx <- sweep(lcpm_mtx, 1, median_vec, "-")

par(mar = c(1, 3, 2, 1))
boxplot(RLR_mtx,
  xlab = "", ylab='Relative Log Representation', ylim = c(-.6, .6),
  staplewex = 0, # remove horizontal whisker lines
  staplecol = "white", # just to be totally sure :)
  outline = F, # remove outlying points
  whisklty = 0, # remove vertical whisker lines
  las = 2, horizontal = F, xaxt = "n",
  border = groupCol[sampDescA$group]
)
legend("top", legend = names(groupCol), 
  text.col = groupCol, ncol = 2, bty = "n")
# Add group Q1, Q3
for (GRP in unique(sampDescA$group)) {
  group_ndx <- which(sampDescA$group == GRP)
  group_Q1Q3_mtx <- apply(RLR_mtx[, group_ndx], 2, 
     quantile, prob = c(.25, .75))
  abline(h = apply(group_Q1Q3_mtx, 1, median), 
     col = groupCol[GRP], lwd = 2)
}


```

<!-- SKIPPED
We note that the HCC samples have slightly more variable coverage distribution.
A few samples are quite different.
-->

## Differential representation analysis {#dra}


In the remainder of this section, we will process the data and
perform differential expression analysis as outlined in 
Law et al. (2018) [@Law:2018aa].   The main analysis steps are: 

* remove lowly expressed genes
* normalize gene expression distributions
* remove heteroscedascity
* fit linear models and examine DE results


It is good practice to perform this differential expression analysis prior to 
fitting models to get an idea of how difficult it will be to discriminate 
between samples belonging to the different subgroups. The pipeline
outlined in Law et al. (2018) [@Law:2018aa] also provides some
basic quality assessment opportunities.


### Remove lowly expressed genes {-}

Genes that are not expressed at a biologically 
meaningful level in any condition should be discarded to reduce the 
subset of genes to those that are of interest, and to reduce the number of tests 
carried out downstream when looking at differential expression.  Carrying
un-informative genes may also be a hindrance to classification and other
downtream analyses.  

To determine a sensible threshold we can begin by examining the shapes of the distributions.

```{r densityLcpm, fig.height=4, fig.width=10, fig.cap='Sample $log_2$ CPM densities', eval=T, echo=F}

par(mar = c(4, 3, 2, 1))
plot(density(lcpm_mtx[, 1]),
  col = groupCol[sampDescA$group[1]],
  lwd = 2, ylim = c(0, .25), las = 2, main = "", xlab = "log2 CPM"
)
abline(v = 0, col = 3)
# After verifying no outliers, can plot a random subset 
for (JJ in sample(2:ncol(lcpm_mtx), size = 100)) {
  den <- density(lcpm_mtx[, JJ])
  lines(den$x, den$y, col = groupCol[sampDescA$group[JJ]], lwd = 2)
} # for(JJ
legend("topright", legend = names(groupCol), 
  text.col = groupCol, bty = "n")

```

`r  LibSizeSum <- summary( colSums(featureCountsA) / 1e6 )`
`r CPM_THR <- 3; SAMP_THR <- 25` 

As is typically the case with RNA-Seq data, we notice many weakly represented genes 
in this dataset.  A cpm value of 1 appears to adequatly separate 
the expressed from the un-expressed genes, but we will be slightly more strict here
and require a CPM threshold of `r CPM_THR` .  Using a nominal CPM value of 
`r CPM_THR`, genes are deeemed to be `represented` if their expression is 
above this threshold, and not represented otherwise. 
For this analysis we will require that genes be `represented` in at least 
`r SAMP_THR` samples across the entire dataset to be retained for downstream analysis.
Here, a CPM value of `r CPM_THR` means that a gene is represented if it 
has at least `r round(CPM_THR*LibSizeSum['Min.'])` reads in the sample with the 
lowest sequencing depth (library size `r round(LibSizeSum['Min.'],1)` million).
Note that the thresholds used here are arbitrary as there are no hard and fast 
rules to set these by.
The voom-plot, which is part of analyses done to remove heteroscedasticity,
can be examined to verify that the filtering performed is adequate.


<!-- or at least 
`r round(CPM_THR*LibSizeSum['Max.'])` counts in the sample with the 
greatest sequencing depth (library size `r round(LibSizeSum['Max.'],1)` million).
-->

Remove weakly represented genes and replot densities.

`r weak_flg <- rowSums(edgeR::cpm(featureCountsA) > CPM_THR) < SAMP_THR `
Removing `r round(100 * mean(weak_flg), 1)`%  of genes...

```{r removeWeak, cache=T, cache.vars=c('featureCountsAF', 'genes_annotAF', 'lcpm_mtx'), echo=F, include=F}

# Use suffix 'F' for Filtered genes
featureCountsAF <- featureCountsA[!weak_flg, ]

genes.ndx <- match(rownames(featureCountsAF), genes_annot$Symbol)
if(sum(is.na(genes.ndx))) stop("featureCountsA/genes_annot mismatch")
genes_annotAF <- genes_annot[genes.ndx,]

lcpm_mtx <- edgeR::cpm(featureCountsAF, log = T)
dim(lcpm_mtx)

rm(featureCountsA)

```


```{r densityLcpm2, fig.height=4, fig.width=10, fig.cap='Sample $log_2$ CPM densities after removing weak genes', eval=T, echo=F}

par(mar = c(4, 3, 2, 1))
plot(density(lcpm_mtx[, 1]),
  col = groupCol[sampDescA$group[1]],
  lwd = 2, ylim = c(0, .25), las = 2, main = "", xlab = "log2 CPM"
)
#abline(v = 0, col = 3)
# After verifying no outliers, can plot a random subset 
for (JJ in sample(2:ncol(lcpm_mtx), size = 100)) {
  den <- density(lcpm_mtx[, JJ])
  lines(den$x, den$y, col = groupCol[sampDescA$group[JJ]], lwd = 2)
} # for(JJ
legend("topright", legend = names(groupCol), 
  text.col = groupCol, bty = "n")

```

<!--
Note that the $log_2(CMP)$ distribution is not quite symmetric.
-->

As another sanity check, we will look at a 
multidimensional scaling plot of distances between gene expression
profiles.  We use `plotMDS` in limma package [@Ritchie:2015aa]),
which plots samples on a two-dimensional scatterplot so that distances on
the plot approximate the typical log2 fold changes between the
samples.   

Before producing the MDS plot we will normalize the distributions.
We will store the data into s `DGEList` object as this is convenient
when running many of the analyses implemented in the edgeR and limma packages.
Call the set 'AF', for set 'A', 'Filtered'.

```{r getDGEL, cache=T, cache.var=c('AF_dgel', 'AF_lcmp_mtx')}

AF_dgel <- edgeR::DGEList(
  counts = featureCountsAF,
  genes = genes_annotAF,
  samples = sampDescA,
  group = sampDescA$group
)
AF_dgel <- edgeR::calcNormFactors(AF_dgel)
AF_lcmp_mtx <- edgeR::cpm(AF_dgel, log = T)

# Save AF_dgel to facilitate restarting
# remove from final version
save(list = "AF_dgel", file = "RData/AF_dgel")
```

Verify that the counts are properly normalized.


```{r normedLcpmBxp, cache=T, fig.height=4, fig.width=10, fig.cap='Sample log2 count boxplots', echo=F}

par(mar = c(1, 3, 2, 1))
boxplot(AF_lcmp_mtx,
  ylim = c(1, 8), ylab='Normalized Log CPM',
  staplewex = 0,       # remove horizontal whisker lines
  staplecol = "white", # just to be totally sure :)
  outline = F,         # remove outlying points
  whisklty = 0,        # remove vertical whisker lines
  las = 2, horizontal = F, xaxt = "n",
  border = groupCol[sampDescA$group]
)
legend("top", legend = names(groupCol), text.col = groupCol,
  ncol = 2, bty = "n")
# Add reference lines
SampleMedian <- apply(AF_lcmp_mtx, 2, median)
abline(h = median(SampleMedian), col = "grey")
axis(side = 4, at = round(median(SampleMedian), 1),
  las = 2, col = "grey", line = -1, tick = F)

```

Proceed with MDS plots.

```{r plotMDS, cache=T, fig.height=5, fig.width=10, fig.cap='MDS plots of log-CPM values', echo=F}

par(mfcol = c(1, 2), mar = c(4, 4, 2, 1), xpd = NA, oma = c(0, 0, 2, 0))

# wo loss of generality, sample 500 samples
# simply a matter of convenience to save time
# remove from final version
set.seed(1)
samp_ndx <- sample(1:ncol(AF_lcmp_mtx), size = 500)
MDS.out <- limma::plotMDS(AF_lcmp_mtx[, samp_ndx],
  col = groupCol[sampDescA$group[samp_ndx]], pch = 1
)
legend("topleft",
  legend = names(groupCol),
  text.col = groupCol, bty = "n"
)

MDS.out <- limma::plotMDS(AF_lcmp_mtx[, samp_ndx],
  col = groupCol[sampDescA$group[samp_ndx]], pch = 1,
  dim.plot = 3:4
)
```

The MDS plot, which is analogous to a PCA plot adapted to gene exression data,
does not indicate strong clustering of samples.  The fanning pattern observed in the
first two dimensions indicates that a few samples are drifting way from the
core set, but in no particular direction.  There is some structure in the
3rd and 4th dimension plot which should be investigated.  
`glMDSPlot` from package `Glimma` provides an interactive MDS 
plot that can extremely usful for exploration

```{r GlMDSplot, echo=T,cache=T, cache.vars='', fig.height=6, fig.width=11,fig.cap="MDS plots of log-CPM values", echo=F}

Glimma::glMDSPlot(AF_dgel[, samp_ndx],
  groups = AF_dgel$samples[
    samp_ndx,
    c("group", "trainValGroup", "sampType", "tissue", "title", "stage")
  ],
  main = paste("MDS plot: filtered counts"), #### , Excluding outlier samples"),
  path = ".", folder = figures_DIR,
  html = paste0("GlMDSplot"), launch = F
)
```

Link to glMDSPlot: 
[Here](`r file.path(figures_DIR, paste0("GlMDSplot.html"))`)  

No obvious factor links the samples in the 3 clusters observed on the
4th MDS dimensions. The percent of variance exaplained by this dimension or 
$\approx$ 4%.   The glMDSPlot indicates further segregation along
the 6th dimension.  The percent of variance exaplained by this dimension or 
$\approx$ 2%.  Tracking down this source of variability may be quite challenging,
especially without having the complete information about the sample attributes 
and provenance.  

Unwanted variability is a well-documented problem in the analysis of RNA-Seq data
(see Peixoto et al. (2015) [@Peixoto:2015aa]), and many procedures have been proposed
to reduce the effect of unwanted variation on RNA-Seq analsys results 
([@Gandolfo:2018aa;@Peixoto:2015aa;@Risso:2014aa]).  There are undoubtedly
some similar sources of systematic variation in the 5hmC data, but it is
beyond the scope of this work to investigate these in this particular dataset.
Given that the clustering of samples occurs in MDS dimensions that explain
a small fraction of variability, and that these is no assocation with the
factor of interest, HCC vs Control, these sources of variability should not
interfere too much with our classification analysis.  It would nonetheless be interesting
to assess whether downstream results can be improved by removing this variability.



### Creating a design matrix and contrasts  {-}

Before proceeding with the statistical modeling used for the 
differential expression analysis, we need to set up a
model design matrix.

```{r DEADesign, cache=F, include=T, echo=F, include=T}

Design_mtx <- model.matrix( ~  -1 + group, data=AF_dgel$samples)
colnames(Design_mtx) <- sub('group', '', colnames(Design_mtx))

cat("colSums(Design_mtx):\n")
colSums(Design_mtx)

Contrasts_mtx <- limma::makeContrasts(
  HCCvsControl = HCC  - Control,
  levels=colnames(Design_mtx))

cat("Contrasts:\n")
Contrasts_mtx

```

```{r printDesign, echo=F, include=F}
 knitr::kable(head(Design_mtx), caption='Design Matrix') %>%
  kableExtra::kable_styling(full_width = F)
```


### Removing heteroscedasticity from the count data {-}


As for RNA-Seq data, for 5hmC count data the variance is not independent of the mean.
In `limma`, the R package we are using for our analyses, 
linear modeling is carried out on the log-CPM values which are assumed to be 
normally distributed and the mean-variance relationship is accommodated using precision 
weights calculated by the voom function.  We apply this transformation next.


```{r Voom1, cache=T, cache.vars=c('filteredCountsAF_voom'), fig.height=6, fig.width=11, fig.cap="Removing heteroscedascity", echo=T}

par(mfrow=c(1,1))
filteredCountsAF_voom <- limma::voom(AF_dgel, Design_mtx, plot=T)

```

Note that the voom-plot provides a visual check on the level of filtering performed upstream.
If filtering of lowly-expressed genes is insufficient, a drop in variance levels can be 
observed at the low end of the expression scale due to very small counts. 

<!--
Means (x-axis) and variances (y-axis) of each gene are plotted to show the dependence between the two before voom is applied to the data ( A) and how the trend is removed after voom precision weights are applied to the data ( B). The plot on the left is created within the voom function which extracts residual variances from fitting linear models to log-CPM transformed data. Variances are then re-scaled to quarter-root variances (or square-root of standard deviations) and plotted against the mean expression of each gene. The means are log 2-transformed mean-counts with an offset of 2. The plot on the right is created using plotSA which plots log 2 residual standard deviations against mean log-CPM values. The average log 2 residual standard deviation is marked by a horizontal blue line. In both plots, each black dot represents a gene and a red curve is fitted to these points.
-->

<!--

To get a sense of how this compares with RNA-Seq dsta, we can take a look at Figure 4 in Law et al. [@Law:2018aa]:

TICKr knitr::include_graphics(
  "Static/images/Law-fig4.gif",  dpi=100)TICK

We observe that the variability in the 5hmC data is quite a bit lower.  Statistical summaries would give
a better idea.  

-->


### Fit linear models and examine the results {-}

Having properly filtered and  normalized the data,
the linear models can be fitted to each gene and the results
examined to assess differential expression between the two groups
of interest, in our case HCC vs Control.

Table \@ref(tab:lmFit) displays the counts of genes in each DE category:

```{r lmFit, cache=T, echo=T, cache.vars=c('filteredCountsAF_voom_efit','filteredCountsAF_voom_efit_dt'),echo=F}


 filteredCountsAF_voom_fit <- limma::lmFit(filteredCountsAF_voom, Design_mtx)
 colnames(filteredCountsAF_voom_fit$coefficients) <- sub("\\(Intercept\\)", "Intercept",
 colnames(filteredCountsAF_voom_fit$coefficients) )

 filteredCountsAF_voom_fit <- limma::contrasts.fit(
    filteredCountsAF_voom_fit, contrasts=Contrasts_mtx)

 filteredCountsAF_voom_efit <- limma::eBayes(filteredCountsAF_voom_fit)

 filteredCountsAF_voom_efit_dt <-
 limma::decideTests(filteredCountsAF_voom_efit,adjust.method = "BH", p.value = 0.05)
 
 knitr::kable(t(summary(filteredCountsAF_voom_efit_dt)),
  caption="DE Results at FDR = 0.05") %>% 
  kableExtra::kable_styling(full_width = F)

```

### Graphical representations of DE results: MD Plots {-}

To summarise results for all genes visually, mean-difference plots
(aka MA plot), which display log-FCs from the linear model fit against 
the average log-CPM values can be generated using the plotMD function,
with the differentially expressed genes highlighted.

We may also be interested in whether certain gene features are 
related to gene identification.  Gene GC content, for example, might be
of interest.

```{r mdPlotEfit, cache=T, cache.vars='', fig.height=5, fig.width=11, fig.cap="HCC vs Control - Genes Identified at FDR = 0,05", echo=F}

par(mfrow=c(1,3), mar=c(4.5,4.5,2,1),oma=c(1,1,2,0))

# log-fold-change vs ave-expr
limma::plotMD(filteredCountsAF_voom_efit,
 ylim = c(-0.4, 0.4),
 column='HCCvsControl',
 status=filteredCountsAF_voom_efit_dt[,'HCCvsControl'],
 hl.pch = 16, hl.col = c("lightblue", "pink"), hl.cex = .5,
 bg.pch = 16, bg.col = "grey", bg.cex = 0.5,
 main = '',
 xlab = paste0(
    "Average log-expression: IQR=",
    paste(round(quantile(filteredCountsAF_voom_efit$Amean, prob = c(1, 3) / 4), 2),
      collapse = ", "
    )
  ),
  ylab = paste0(
    "log-fold-change: IQR=",
    paste(round(quantile(filteredCountsAF_voom_efit$coefficients[, 'HCCvsControl'], prob = c(1, 3) / 4), 2),
      collapse = ", "
    )
  ),
  legend = F, cex.lab=1.5
)
abline(h = 0, col = "black")
rug(quantile(filteredCountsAF_voom_efit$coefficients[, 'HCCvsControl'], prob = c(1, 2, 3) / 4),
  col = "purple",
  ticksize = .03, side = 2, lwd = 2
)
rug(quantile(filteredCountsAF_voom_efit$Amean, prob = c(1, 2, 3) / 4),
  col = "purple",
  ticksize = .03, side = 1, lwd = 2
)

# log-fold-change vs identification

boxplot(split(
 filteredCountsAF_voom_efit$coefficients[, 'HCCvsControl'],
 filteredCountsAF_voom_efit_dt[,'HCCvsControl']),
 outline=F,
 border=c("pink", "grey", "lightblue"), xaxt='n',
 ylab='log-fold-change', ylim=c(-.4, .4),
 cex.lab=1.5
)
axis(side=1, at=1:3, c('down', 'notDE', 'up'), cex.axis=1.5)

REPLACE.THIS <- function(){
tmp.lst <- lapply( 
split(
 filteredCountsAF_voom_efit$coefficients[, 'HCCvsControl'],
 filteredCountsAF_voom_efit_dt[,'HCCvsControl']
),
density)

plot(tmp.lst[["-1"]]$y, tmp.lst[["-1"]]$x, col='pink', 
     lwd=2, type='l', ylim=c(-.4,.4), xlim=c(0, 17),
 xlab='density', ylab='log-fold-change', cex.lab=1.5)
#lines(tmp.lst[["0"]]$y, tmp.lst[["0"]]$x, col='grey', lwd=2)
lines(tmp.lst[["1"]]$y, tmp.lst[["1"]]$x, col='lightblue', lwd=2)

}#REPLACE.THIS


# gc vs identification
genes_ndx <- match(rownames(filteredCountsAF_voom_efit), genes_annotAF$Symbol)
if(sum(is.na(genes_ndx))) stop("filteredCountsAF_voom_efit/genes_annotAF: genes mismatch")
GC_vec <- with(genes_annotAF[genes_ndx,],(G+C)/(A+C+G+T))


boxplot(split(
 GC_vec,
 filteredCountsAF_voom_efit_dt[,'HCCvsControl']),
 outline=F,
 border=c("pink", "grey", "lightblue"), xaxt='n',
 ylab='gene-gc', cex.lab=1.5
)
axis(side=1, at=1:3, c('down', 'notDE', 'up'), cex.axis=1.5)

 #mtext(side=3, outer=T, cex=1.25, "Genes identified at adjusted p-value=0.05")
```

```{r quantlogFC,echo=F}

featureCountsAF_logFC_sum <- sapply(
 split(
 filteredCountsAF_voom_efit$coefficients[, 'HCCvsControl'],
 filteredCountsAF_voom_efit_dt[,'HCCvsControl']),
 quantile, prob = (1:3) / 4)

colnames(featureCountsAF_logFC_sum) <- as.character(factor(
 colnames(featureCountsAF_logFC_sum), 
 levels=c("-1", "0", "1"),
 labels=c('down', 'notDE', 'up')
))


knitr::kable(featureCountsAF_logFC_sum,
  digits = 2,
  caption = "log FC quartiles by gene identification") %>%
  kableExtra::kable_styling(full_width = F)

```


While many genes are identified, the effect sizes are quite small,
which results in a low signal-to-noise ratio context.   See 
Section \@ref(snr-regime) below.

The log-fold-change distribution for up-represented genes is long-tailed,
with many high log fold-change values.
By contrast, log-fold-change distribution for down-represented genes
closer to symmetric and has few genes with low log fold-change values.
We will see how this affects the results of identifying genes with
an effect size requirement.

The GC content of down regulated genes tends to be slightly lower than the
rest of the genes.  A statistical test would find that the difference
between the mean of the down regulated gene population is singificantly different
than the mean of the other gene population even though the difference is
quite small
(`r round( 
mean(GC_vec[filteredCountsAF_voom_efit_dt[,'HCCvsControl']=='-1']) -
mean(GC_vec[filteredCountsAF_voom_efit_dt[,'HCCvsControl']!='-1']),
3)`).

These asymmetries are minor, but it would still be good to establish that
they relfect biology rather than processing artifacts.  


### DE genes at 10% fold change {-}

For a stricter definition on significance, one may require log-fold-changes 
(log-FCs) to be above a minimum value. The treat method 
(McCarthy and Smyth 2009 [@McCarthy:2009aa]) can be used to calculate p-values 
from empirical Bayes moderated t-statistics with a minimum log-FC requirement. 
The number of differentially expressed genes are greatly reduced if we 
impose a minimal fold-change requirement of 10%.

```{r mdPlotTfit, cache=T, cache.vars='', fig.height=5, fig.width=11, fig.cap="HCC vs Control - Identified Genes at FDR = 0,05 and logFC > 10%",echo=F}

filteredCountsAF_voom_tfit <- limma::treat(filteredCountsAF_voom_fit, lfc=log2(1.10))
filteredCountsAF_voom_tfit_dt <- limma::decideTests(filteredCountsAF_voom_tfit)

cat("10% FC Gene Identification Summary - voom, adjust.method = BH, p.value = 0.05:\n")
summary(filteredCountsAF_voom_tfit_dt)

# log-fold-change vs ave-expr
limma::plotMD(filteredCountsAF_voom_efit,
 ylim = c(-0.5, 0.5),
 column='HCCvsControl',
 status=filteredCountsAF_voom_tfit_dt[,'HCCvsControl'],
 hl.pch = 16, hl.col = c("blue", "red"), hl.cex = .7,
 bg.pch = 16, bg.col = "grey", bg.cex = 0.5,
 main = '',
 xlab = paste0(
    "Average log-expression: IQR=",
    paste(round(quantile(filteredCountsAF_voom_efit$Amean, prob = c(1, 3) / 4), 2),
      collapse = ", "
    )
  ),
  ylab = paste0(
    "log-fold-change: IQR=",
    paste(round(quantile(filteredCountsAF_voom_efit$coefficients[, 'HCCvsControl'], prob = c(1, 3) / 4), 2),
      collapse = ", "
    )
  ),
  legend = F
)
abline(h = 0, col = "black")
rug(quantile(filteredCountsAF_voom_efit$coefficients[, 'HCCvsControl'], prob = c(1, 2, 3) / 4),
  col = "purple",
  ticksize = .03, side = 2, lwd = 2
)
rug(quantile(filteredCountsAF_voom_efit$Amean, prob = c(1, 2, 3) / 4),
  col = "purple",
  ticksize = .03, side = 1, lwd = 2
)


```

As noted above, the log-fold-change distribution for the up-represented genes
is long-tailes in  comparison to log-fold-change distribution for the down-represented genes.
As a result fewer down-represented than up-regulated genes are identified when a 
minimum log-FC requirement is imposed.


## Signal-to-noise ratio regime {#snr-regime}

In Hastie et al. (2017) [@Hastie:2017aa]) results from `lasso` fits are
compared with `best subset` and `forward selection` fits and it is argued
that while `best subset` is optimal for high signal-to-noise regimes, 
the lasso gains some competitive advantage when the prevailing signal-to-noise
ratio of the dataset is lowered.  

We can extract sigma and signal from the fit objects to get SNR values for each gene
to see in what SNR regime the 5hmC gene body data are.

 
```{r altCV, cache=T, cache.vars='',fig.height=4, fig.width=6, fig.cap="Alternative CV Calculation"}
lib.size <- colSums(AF_dgel$counts)

fit <- filteredCountsAF_voom_efit
sx <- fit$Amean + mean(log2(lib.size + 1)) - log2(1e+06)
sy <- sqrt(fit$sigma)

CV <- sy/sx    


```

<!-- DEBUG BCV from Section \@ref(analysis-of-coverage-variability) vs CV
pairs(cbind(BCV_mtx, CV)) 
boxplot(cbind(BCV_mtx, CV), outline=F) 
-->


```{r plotSNR,  fig.height=5, fig.width=10, fig.cap="Cumulative Distribution of SNR - rug = 25, 50, 75 and 90th percentile",message=F, echo=F,include=T}

Effect <- abs(filteredCountsAF_voom_efit$coefficients[,'HCCvsControl'])
Noise <- filteredCountsAF_voom_efit$sigma
SNR <- Effect/Noise

plot(spatstat::CDF(density(SNR)),
  col = 1, lwd = 2, ylab = "Prob(SNR<x)",
  xlim = c(0, 0.2)
)

SNR_quant <- quantile(SNR, prob=c((1:3)/4,.9))
rug(SNR_quant,
    lwd = 2, ticksize = 0.05, col = 1
  )


knitr::kable(t(SNR_quant),
  digits = 3,
  caption = paste(
    "SNR Quantiles") 
) %>% kableExtra::kable_styling(full_width = F)



```

These SNR values are in the range where the lasso and relaxed lasso gain some advantage over
best subset and forward selection fits (see  Hastie et al. (2017) [@Hastie:2017aa]).


<!--chapter:end:00-preprocessing.Rmd-->

# Modeling - Background {#modeling-background}

Refer to [first pass study](https://hcc-5hmc-analysis.netlify.app/) for
relevant exploratory data  analyis results.


In the section we look at  some models fitted to discriminate between
early stage HCC and healthy and benign samples (grouped as Controls here)
from the GSE112679 data set.  


* Some questions to addredd with the baseline model 
   - how separable are the data: what accuracy do we expect 
   - individual sample quality scores: which samples are hard to classify?  Compute a score
in [0, 1], where 1 is perfectly good classification and 0 is poerfectly bad.

## Predictive modeling for genomic data

The main challenge in calibrating predictive models to genomic data is that
there are many more features than there are example cases to fit to;
the  now classic $n << p$ problem.  
In this scenario, fitting methods tend to overfit.  The problem
can be addressed by selecting variables, regularizng the fit or both.
See the Trevor Hastie talk: 
[Statistical Learning with Big Data - Trevor Hastie](https://web.stanford.edu/~hastie/TALKS/SLBD_new.pdf)
for a good discussion of this problem and potential solutions.



### caret for model evaluation

[The `caret` Package](https://topepo.github.io/caret/index.html)
provide a set of functions that streamline the process for fitting and
evalluating a large numbet of predictive models in parallel. The package contains tools for:

* data splitting  
* pre-processing  
* feature selection  
* model tuning using resampling  
* variable importance estimation  

The tools facilitate the process of automating randomly spliting data sets into training, 
testing and evaluating so that predictive models can be evaluated on a comparable and
exhaustive basis.  Especially useful is the functionality that is provided to
repeatedly randomly stratify samples into train and test set so that any
sample selection bias is removed.  


What makes the `caret` package extremely useful is that a common interface is
provided to an exhaustive collection of fitting procedures.  Without
this common interface, one has to learn the programming interfaces that
are used in all fitting procedures to be included in a comparative analysis,
which can be quite burdenful.  

Some of the models which can be evaluated with caret include: 
(only some of these can be used with multinomial responses)

* FDA - Flexible Discriminant Analysis  
* stepLDA - Linear Discriminant Analysis with Stepwise Feature Selection  
* stepQDA - Quadratic Discriminant Analysis with Stepwise Feature Selection  
* knn - k nearest neighbors  
* pam - Nearest shrunken centroids  
* rf - Random forests  
* svmRadial - Support vector machines (RBF kernel)  
* gbm - Boosted trees  
* xgbLinear - eXtreme Gradient Boosting  
* xgbTree - eXtreme Gradient Boosting  
* neuralnet - neural network  

Many more models can be implemented and evaluated with `caret`, 
including some `deep learning` methods.

`Simulated Annealing Feature Selection` and `Genetic Algorithms`.
Many methods found [here](https://topepo.github.io/caret/available-models.html)
are also worth investigating.

We only mention `caret` here because it is an exteremely useful tool for 
anyone interested in comparing many predictive models.  We have done that
in the past and have found that regularized regression models perform
as welll as any in the context of classifiction based on genomic scale data.



## glmnet

In this investigation we will focus on models that can be
analyzed with the the `glmnet` R  package [@Friedman:2010aa].  Several
factors favor this choice:  

* the glmnet package is a well supported package providing
extensive functionality for regularized regression and classification models.

* the hyper-parameters of the elastic net enable us to explore
the relationship between model size, or sparsity, and predictive accuracy.
ie. we can investigate the "bet on sparsity" principle:
*Use a procedure that does well in sparse problems, since no procedure
does well in dense problems*.

* in our experience building classifiers from genomic scale data, regularized
classification models using the elastic net penalty do as well as any other,
and are more economical in terms of computing time, espacially in comparison to
the more exotic boosting algorithms.

* the `lasso` has been shown to be near optimal for the $n<<p$ problem
over a wide range of signal-to-noise regiments. 


***


Much of the following comes from the 
[Glmnet Vignette](https://web.stanford.edu/~hastie/glmnet/glmnet_alpha.html).


Glmnet is a package that fits a generalized linear model via penalized maximum likelihood. 
The regularization path is computed for the lasso or elasticnet penalty at a 
grid of values for the regularization parameter lambda 
([@Friedman:2010aa;@Tibshirani:2012aa;@Simon:2011aa;@Simon:2013aa]). 

`glmnet` solves the following problem:

$$\min_{\beta_0,\beta} \frac{1}{N} \sum_{i=1}^{N} w_i l(y_i,\beta_0+\beta^T x_i) + \lambda\left[(1-\alpha)||\beta||_2^2/2 + \alpha ||\beta||_1\right],$$

over a grid of values of $\lambda$.
Here $l(y,\eta)$ is the negative log-likelihood contribution for observation i; 
e.g. for the Gaussian case it is $\frac{1}{2}(y-\eta)^2$.


###  **alpha** hyper-parameter {-}

The elastic-net penalty is controlled by $\alpha$, and bridges the gap between 
lasso ($\alpha$=1, the default) and ridge ($\alpha$=0). 
The tuning parameter $\lambda$ controls the overall strength of the penalty. 

It is known that the ridge penalty shrinks the coefficients of correlated predictors 
towards each other while the lasso tends to pick one of them and discard the others. 
The elastic-net penalty mixes these two; if predictors are correlated in groups, 
an $\alpha$=0.5 tends to select the groups in or out together. 
This is a higher level parameter, and users might pick a value upfront, 
else experiment with a few different values. One use of $\alpha$ is for numerical stability; 
for example, the *elastic net with $\alpha = 1 - \epsilon$ for some small $\epsilon$>0 
performs much like the lasso, but removes any degeneracies and wild behavior caused 
by extreme correlations*.



### Lasso vs Best Subset  {-}

* Best subset selection 

 $$\min_{\beta \in \mathcal{R}^p} ||Y - X\beta||^2_2 \, \, subject \, to \, \, ||\beta||_0 \leq k$$


* lasso 

 $$\min_{\beta \in \mathcal{R}^p} ||Y - X\beta||^2_2 \, \, subject \, to \, \, ||\beta||_1 \leq t$$


* Bertsimas et al. (2016) [@Bertsimas:2016aa]
    - presented a mixed integer optimization (MIO) formulation for the best subset selection problem   
    - Using these MIO solvers, can solve problems with p in the hundreds and even thousands
    - demonstrated that best subset selection generally gives superior prediction accuracy compared to forward stepwise selection and the lasso, over a variety of problem setups. 

* Hastie et al. (2017) [@Hastie:2017aa]
    - neither best subset selection nor the lasso uniformly dominate the other, with best subset selection generally performing better in high signal-to-noise (SNR) ratio regimes, and the lasso better in low SNR regimes;
    - best subset selection and forward stepwise perform quite similarly throughout;
    - the relaxed lasso is the overall winner, performing just about as well as the lasso in low SNR scenarios, 
and as well as best subset selection in high SNR scenarios. We conclude that it is 
able to use its auxiliary shrinkage parameter ($\gamma$) to get the “best of both worlds”: 
it accepts the heavy shrinkage from the lasso when such shrinkage is helpful, and reverses it when it is not.
   

* relaxed lasso

$$\hat{\beta}^{relax}(\lambda, \gamma) = \gamma \beta^{lasso}(\lambda) + (1 - \gamma)(\beta^{LS}(\lambda)$$

* shrunken relaxed lasso (aka the blended fit)

Suppose the **glmnet** fitted linear predictor at $\lambda$ is $\hat{\eta}_\lambda(x)$
and the relaxed version is $\tilde{\eta}_\lambda(x)$, then the shrunken relaxed lasso fit is

$$\tilde{\eta}_{\lambda,\gamma}(x)=(1-\gamma)\tilde{\eta}_\lambda(x) + \gamma \hat{\eta}_\lambda(x)$$

$\gamma \in [0,\, 1]$ is an additional tuning parameter which can be selected by cross validation.

The debiasing will potentially improve prediction performance, and CV will typically select a model 
with a smaller number of variables. This procedure is very competitive with forward-stepwise and 
best-subset regression, and has a considerable speed advantage when the number of variables is large. 
This is especially true for best-subset, but even so for forward stepwise. 
The latter has to plod through the variables one-at-a-time,
while glmnet will just plunge in and find a good active set.

Further details may be found in 
Friedman, Hastie, and Tibshirani (2010),
Tibshirani et al. (2012),
Simon et al. (2011),
Simon, Friedman, and Hastie (2013) and 
Hastie, Tibshirani, and Tibshirani (2017)
([@Friedman:2010aa;@Tibshirani:2012aa;@Simon:2011aa;@Simon:2013aa;@Hastie:2017aa]). 


* SNR  
    - $y_0=f(x_0) + \epsilon_0$
    - $SNR=\frac{var(f(x_0))}{var(\epsilon_0)}$
    - $PVE(g)=1 - \frac{\mathbb{E}(y_0-g(x_0))^2}{Var(y_0)}$
    - $PVE(f) = 1 - \frac{Var(\epsilon_0)}{Var(y_0)} = \frac{SNR}{1+SNR}$
    - $SNR = \frac{PVE}{1-PVE}$
    - $c_v = \frac{\sigma}{\mu}=\frac{Var(y)}{\mathbb{E}(y)}$
        - a PVE of 0.5 is rare for noisy observational data, and 0.2 may be more typical
        - A PVE of 0.86, corresponding to an SNR of 6, is unheard of!
        - For small SNR, SNR $\approx$ PVE
        - See Xiang et al. (2020) [@Xiang:2020aa], Lozoya et al. (2018) [@Lozoya:2018aa], 
Simonson et al. (2018) [@Simonsen:2018aa] and
Rapaport et al. (2013) [@Rapaport:2013aa] for SNR in RNA-Seq

<!--chapter:end:00-modelingBackground.Rmd-->

# The bet on sparsity {#explore-sparsity}

In this section we explore various fits that can be computed 
ans analyzed with tools provided in the `glmnet` package.
Refer to the [Glmnet Vignette](https://web.stanford.edu/~hastie/glmnet/glmnet_alpha.html)
for a quick reference guide.

## CV analysis setup 

```{r setParameters}

K_FOLD <- 10
trainP <- 0.8
EPS <- 0.02    # Have no idea what "small" epsilon means

```

First we divide the analysis dataset into `train` and `test` in a `r trainP/(1-trainP)`:1 ratio.  

```{r getTrainVal, cache=T, cache.vars=c('train_sampID_vec', 'test_sampID_vec','train_group_vec','test_group_vec','train_lcpm_mtx','test_lcpm_mtx')}

set.seed(1)
train_sampID_vec <- with(AF_dgel$samples,
AF_dgel$samples$sampID[caret::createDataPartition(y=group, p=trainP, list=F)]
)

test_sampID_vec <- with(AF_dgel$samples,
setdiff(sampID, train_sampID_vec)
)

train_group_vec <- AF_dgel$samples[train_sampID_vec, 'group']
test_group_vec <- AF_dgel$samples[test_sampID_vec, 'group']

knitr::kable(table(train_group_vec),
  caption="Train set") %>%
   kableExtra::kable_styling(full_width = F)

knitr::kable(table(test_group_vec),
  caption="Test set") %>%
   kableExtra::kable_styling(full_width = F)

train_lcpm_mtx <- t(lcpm_mtx[,train_sampID_vec])
test_lcpm_mtx <- t(lcpm_mtx[,test_sampID_vec])

```

We explore some glmnet fits and the "bet on sparsity"    
  
* Consider models:  
    - lasso: $\alpha = 1.0$ - sparse model  
    - ridge $\alpha = 0$ - shrunken coefficients model
    - elastic net:  $\alpha = 0.5$  - semi sparse model
    - lassoC: $\alpha = 1-\epsilon =$ `r 1- EPS` - lasso for correlated predictors  
* Does the relaxed lasso improve performance?    
* Does the shrunken relaxed lasso (aka the blended mix) improve performance  
* How sparse is the model undelying best 5hmC classifier for Early HCC vs Control?    
* Is the degree of sparsity, or the size of the model, a stable feature of the problem and data set?  

In this analysis, we will only evaluate models in terms of 
model size, stability and performance.  We leave the question
of significance testing of hypotheses about model parameters
completely out.  See Lockhart et al. (2014) [@Lockhart:2014aa]
and Wassermam (2014) [@Wasserman:2014aa] for a discussion of this topic.


Next we create folds for `r K_FOLD`-fold cross-validation of models fitted to
training data.  We'll use caret::createFolds to assign samples
to folds while keeping the outcome ratios constant across folds.


```{r getTrainFolds, cache=T, cache.vars='train_foldid_vec'}
# This is too variable, both in terms of fold size And composition
#foldid_vec <- sample(1:10, size=length(train_group_vec), replace=T)

set.seed(1)
train_foldid_vec <- caret::createFolds(
 factor(train_group_vec), 
 k=K_FOLD,
 list=F)

knitr::kable(sapply(split(train_group_vec, train_foldid_vec), 
  table), caption="training samples fold composition") %>%
   kableExtra::kable_styling(full_width = F)
 
```

Note that the folds identify samples that are left-out of the training
data for each fold fit.


## Fit and compare models 


* cross-validated accuracy
* test set accuracy  
* sparsity
    - for lasso, enet and lassoC, examine number of selected variables

Although “the one standard error rule” can produce a model with fewer predictors, it usually results in increased MSE and more biased parameter estimates
(see Engebretsen et al. (2019) [@Engebretsen:2019aa] for example).
We will look at both the minimum cv error and the one standard error rule model 
preformance.

```{r fitModels, cache=T, cache.vars=c('cv_lasso', 'cv_ridge', 'cv_enet', 'cv_lassoC')}

require(doMC)
registerDoMC(cores=14)

start_time <-  proc.time()

cv_lasso <- glmnet::cv.glmnet(
 x=train_lcpm_mtx,
 y=train_group_vec,
 foldid=train_foldid_vec,
 alpha=1,
 family='binomial', 
 type.measure = "class")

message("lasso time: ", round((proc.time() - start_time)[3],2),"s")

start_time <-  proc.time()

cv_ridge <- glmnet::cv.glmnet(
 x=train_lcpm_mtx,
 y=train_group_vec,
 foldid=train_foldid_vec,
 alpha=0,
 family='binomial', 
 type.measure = "class")

message("ridge time: ", round((proc.time() - start_time)[3],2),"s")

start_time <-  proc.time()

cv_enet <- glmnet::cv.glmnet(
 x=train_lcpm_mtx,
 y=train_group_vec,
 foldid=train_foldid_vec,
 alpha=0.5,
 family='binomial',
 type.measure = "class")

message("enet time: ", round((proc.time() - start_time)[3],2),"s")

start_time <-  proc.time()

cv_lassoC <-  glmnet::cv.glmnet(
 x=train_lcpm_mtx,
 y=train_group_vec,
 foldid=train_foldid_vec,
 alpha=1-EPS,
 family='binomial',
 type.measure = "class")

message("lassoC time: ", round((proc.time() - start_time)[3],2),"s")

```

The ridge regression model takes over 10 times longer to compute.


<!-- do not show
Define plotting function.
Maybe show in appendix??
-->
```{r plot_cv_f,echo=F}

plot_cv_f <- function(cv_fit, Nzero=T, ...) {
 
 lambda.1se_p <- cv_fit$nzero[cv_fit$lambda==cv_fit$lambda.1se]
 lambda.min_p <- cv_fit$nzero[cv_fit$lambda==cv_fit$lambda.min]
 
 test_pred_1se_vec <- predict(
  cv_fit, 
  newx=test_lcpm_mtx, 
  s="lambda.1se",
  type="class"
 )
 test_pred_1se_error <- mean(test_pred_1se_vec!=test_group_vec)
 
 test_pred_min_vec <- predict(
  cv_fit, 
  newx=test_lcpm_mtx, 
  s="lambda.min",
  type="class"
 )
 test_pred_min_error <- mean(test_pred_min_vec!=test_group_vec)
 
  
 plot(
  log(cv_fit$lambda),
  cv_fit$cvm,
  pch=16,col="red",
  xlab='',ylab='',
  ...
 )
 abline(v=log(c(cv_fit$lambda.1se, cv_fit$lambda.min)))
 if(Nzero)
 axis(side=3, tick=F, at=log(cv_fit$lambda), 
  labels=cv_fit$nzero, line=-1
 )
 LL <- 2
 #mtext(side=1, outer=F, line=LL, "log(Lambda)")
 #LL <- LL+1
 mtext(side=1, outer=F, line=LL, paste(
  #ifelse(Nzero, paste("1se p =", lambda.1se_p),''),
  "1se: cv =", round(100*cv_fit$cvm[cv_fit$lambda==cv_fit$lambda.1se],1),
  "test =", round(100*test_pred_1se_error,1)
 ))
 LL <- LL+1
 mtext(side=1, outer=F, line=LL, paste(
  #ifelse(Nzero, paste("min p =", lambda.min_p),''),
  "min: cv =", round(100*cv_fit$cvm[cv_fit$lambda==cv_fit$lambda.min],1),
  "test =", round(100*test_pred_min_error,1)
 ))
 
}

```

Examine model performance.

```{r lookFits, cache=T, cache.vars='', fig.height=5, fig.width=11, fig.cap="compare fits", echo=F}
 par(mfrow=c(1,3), mar=c(5, 2, 3, 1), oma=c(3,2,0,0)) 
 plot_cv_f(cv_lasso, ylim=c(0,.5))
 title('lasso')

 plot_cv_f(cv_ridge, Nzero=F, ylim=c(0,.5))
 title('ridge')

 plot_cv_f(cv_enet, ylim=c(0,.5))
 title('enet')

 mtext(side=1, outer=T, cex=1.25, 'log(Lambda)')
 mtext(side=2, outer=T, cex=1.25, cv_lasso$name)

```

All models produce cv assessments of `r cv_lasso$name` that are substnatially lower than
the test set assessments.  `lasso` performs comparably to `enet` and
better than the `ridge` model.

## Relaxed lasso and blended mix

```{r fitLassoR, cache=T, cache.vars=c('cv_lassoR'), include=F}

require(doMC)
registerDoMC(cores=14)


start_time <-  proc.time()

cv_lassoR <-  glmnet::cv.glmnet(
 x=train_lcpm_mtx,
 y=train_group_vec,
 foldid=train_foldid_vec,
 alpha=1,
 relax=T,
 family='binomial',
 type.measure = "class")

message("lassoR time: ", round((proc.time() - start_time)[3],2),"s")

```

<!--
The relaxed fit takes quite a bit longer.  
-->

```{r lookLassoR, cache=T, cache.vars='', fig.height=5, fig.width=5, fig.cap="lassoR fit"}
 library(glmnet)

 cv_lassoR_sum <- print(cv_lassoR)

 plot(cv_lassoR)

test_pred_1se_vec <- predict(
  cv_lassoR, 
  newx=test_lcpm_mtx, 
  s="lambda.1se",
  type="class"
)
test_pred_1se_error <- mean(test_pred_1se_vec!=test_group_vec)
 
test_pred_min_vec <- predict(
  cv_lassoR, 
  newx=test_lcpm_mtx, 
  s="lambda.min",
  type="class"
)
test_pred_min_error <- mean(test_pred_min_vec!=test_group_vec)
  
cv_lassoR_1se_error <- cv_lassoR_sum['1se','Measure']
cv_lassoR_min_error <- cv_lassoR_sum['min','Measure']

knitr::kable(rbind(
 onese=c(cv_error=cv_lassoR_1se_error, test_error=test_pred_1se_error)*100,
 min=c(cv_error=cv_lassoR_min_error, test_error=test_pred_min_error)*100
),
 caption="CV vs test Errors", digits=1
) %>%
  kableExtra::kable_styling(full_width = F)


```

Look at confusion matrices

```{r confMtxTrainLasso, cache=T, cache.vars='', fig.cap="Train set confusion", echo=F}

cv_lasso_cnf <- glmnet::confusion.glmnet(
 cv_lasso, 
 newx=train_lcpm_mtx,
 newy=train_group_vec
)

knitr::kable(cv_lasso_cnf, caption="cv lasso confusion matrix: train set") %>%
  kableExtra::kable_styling(full_width = F)

```

```{r confMtxTrainLassoR, cache=T, cache.vars='', fig.cap="Train set confusion", echo=F}
cv_lassoR_cnf <- glmnet::confusion.glmnet(
 cv_lassoR, 
 newx=train_lcpm_mtx,
 newy=train_group_vec
)

knitr::kable(cv_lassoR_cnf, caption="cv lassoR confusion matrix: train set")  %>%
  kableExtra::kable_styling(full_width = F)


```


```{r confMtxTestLasso, cache=T, cache.vars='', fig.cap="Test set confusion", echo=F}

cv_lasso_cnf <- glmnet::confusion.glmnet(
 cv_lasso, 
 newx=test_lcpm_mtx,
 newy=test_group_vec
)

knitr::kable(cv_lasso_cnf, caption="cv lasso confusion matrix: test set") %>%
  kableExtra::kable_styling(full_width = F)

```

```{r confMtxTestlassoR, cache=T, cache.vars='', fig.cap="Test set confusion", echo=F}

cv_lassoR_cnf <- glmnet::confusion.glmnet(
 cv_lassoR, 
 newx=test_lcpm_mtx,
 newy=test_group_vec
)

knitr::kable(cv_lassoR_cnf, caption="cv lassoR confusion matrix: test set")  %>%
  kableExtra::kable_styling(full_width = F)


```


In all models the sensitivity weak compared to the sensitivity.  Let's examine the 
ROC curves to see where the trade-off is.


<!--chapter:end:00-glmnetFits.Rmd-->

## Refit with "auc" as optimization

```{r fitModels2, cache=T, cache.vars=c('cv_lasso2', 'cv_ridge2', 'cv_enet2', 'cv_lassoC2')}

require(doMC)
registerDoMC(cores=14)

start_time <-  proc.time()

cv_lasso2 <- glmnet::cv.glmnet(
 x=train_lcpm_mtx,
 y=train_group_vec,
 foldid=train_foldid_vec,
 alpha=1,
 family='binomial', 
 type.measure = "auc")

message("lasso time: ", round((proc.time() - start_time)[3],2),"s")

start_time <-  proc.time()

cv_ridge2 <- glmnet::cv.glmnet(
 x=train_lcpm_mtx,
 y=train_group_vec,
 foldid=train_foldid_vec,
 alpha=0,
 family='binomial', 
 type.measure = "auc")

message("ridge time: ", round((proc.time() - start_time)[3],2),"s")

start_time <-  proc.time()

cv_enet2 <- glmnet::cv.glmnet(
 x=train_lcpm_mtx,
 y=train_group_vec,
 foldid=train_foldid_vec,
 alpha=0.5,
 family='binomial',
 type.measure = "auc")

message("enet time: ", round((proc.time() - start_time)[3],2),"s")

start_time <-  proc.time()

cv_lassoC2 <-  glmnet::cv.glmnet(
 x=train_lcpm_mtx,
 y=train_group_vec,
 foldid=train_foldid_vec,
 alpha=1-EPS,
 family='binomial',
 type.measure = "class")

message("lassoC time: ", round((proc.time() - start_time)[3],2),"s")

```

The ridge regression model takes over 10 times longer to compute.



Examine model performance.

```{r lookFits2, cache=T, cache.vars='', fig.height=5, fig.width=11, fig.cap="compare fits", echo=F}
 par(mfrow=c(1,3), mar=c(5, 2, 3, 1), oma=c(3,2,0,0)) 
 plot_cv_f(cv_lasso2, ylim=c(0,.5))
 title('lasso')

 plot_cv_f(cv_ridge2, Nzero=F, ylim=c(0,.5))
 title('ridge')

 plot_cv_f(cv_enet2, ylim=c(0,.5))
 title('enet')

 mtext(side=1, outer=T, cex=1.25, 'log(Lambda)')
 mtext(side=2, outer=T, cex=1.25, cv_lasso$name)

```

All models produce cv assessments of `r cv_lasso$name` that are substnatially lower than
the test set assessments.  `lasso` performs comparably to `enet` and
better than the `ridge` model.

## Relaxed lasso and blended mix

```{r fitLassoR2, cache=T, cache.vars=c('cv_lassoR2'), include=F}

require(doMC)
registerDoMC(cores=14)


start_time <-  proc.time()

cv_lassoR <-  glmnet::cv.glmnet(
 x=train_lcpm_mtx,
 y=train_group_vec,
 foldid=train_foldid_vec,
 alpha=1,
 relax=T,
 family='binomial',
 type.measure = "auc")

message("lassoR time: ", round((proc.time() - start_time)[3],2),"s")

```

<!--
The relaxed fit takes quite a bit longer.  
-->

```{r lookLassoR2, cache=T, cache.vars='', fig.height=5, fig.width=5, fig.cap="lassoR fit"}
 library(glmnet)

 cv_lassoR2_sum <- print(cv_lassoR2)

 plot(cv_lassoR2)

test_pred_1se_vec <- predict(
  cv_lassoR2, 
  newx=test_lcpm_mtx, 
  s="lambda.1se",
  type="class"
)
test_pred_1se_error <- mean(test_pred_1se_vec!=test_group_vec)
 
test_pred_min_vec <- predict(
  cv_lassoR2, 
  newx=test_lcpm_mtx, 
  s="lambda.min",
  type="class"
)
test_pred_min_error <- mean(test_pred_min_vec!=test_group_vec)
  
cv_lassoR2_1se_error <- cv_lassoR2_sum['1se','Measure']
cv_lassoR2_min_error <- cv_lassoR2_sum['min','Measure']

knitr::kable(rbind(
 onese=c(cv_error=cv_lassoR2_1se_error, test_error=test_pred_1se_error)*100,
 min=c(cv_error=cv_lassoR2_min_error, test_error=test_pred_min_error)*100
),
 caption="CV vs test Errors", digits=1
) %>%
  kableExtra::kable_styling(full_width = F)


```

Look at confusion matrices

```{r confMtxTrainLasso2, cache=T, cache.vars='', fig.cap="Train set confusion", echo=F}

cv_lasso_cnf <- glmnet::confusion.glmnet(
 cv_lasso2, 
 newx=train_lcpm_mtx,
 newy=train_group_vec
)

knitr::kable(cv_lasso_cnf, caption="cv lasso confusion matrix: train set") %>%
  kableExtra::kable_styling(full_width = F)

```

```{r confMtxTrainLassoR2, cache=T, cache.vars='', fig.cap="Train set confusion", echo=F}
cv_lassoR_cnf <- glmnet::confusion.glmnet(
 cv_lassoR2, 
 newx=train_lcpm_mtx,
 newy=train_group_vec
)

knitr::kable(cv_lassoR2_cnf, caption="cv lassoR2 confusion matrix: train set")  %>%
  kableExtra::kable_styling(full_width = F)


```


```{r confMtxTestLasso2, cache=T, cache.vars='', fig.cap="Test set confusion", echo=F}

cv_lasso_cnf <- glmnet::confusion.glmnet(
 cv_lasso2, 
 newx=test_lcpm_mtx,
 newy=test_group_vec
)

knitr::kable(cv_lasso_cnf, caption="cv lasso confusion matrix: test set") %>%
  kableExtra::kable_styling(full_width = F)

```

```{r confMtxTestlassoR2, cache=T, cache.vars='', fig.cap="Test set confusion", echo=F}

cv_lassoR_cnf <- glmnet::confusion.glmnet(
 cv_lassoR2, 
 newx=test_lcpm_mtx,
 newy=test_group_vec
)

knitr::kable(cv_lassoR2_cnf, caption="cv lassoR2 confusion matrix: test set")  %>%
  kableExtra::kable_styling(full_width = F)


```


In all models the sensitivity weak compared to the sensitivity.  Let's examine the 
ROC curves to see where the trade-off is.




<!--chapter:end:00-glmnetFits2.Rmd-->

# Examine feature selection

Recall

```
It is known that the ridge penalty shrinks the coefficients of correlated predictors
towards each other while the lasso tends to pick one of them and discard the others.
The elastic-net penalty mixes these two; if predictors are correlated in groups,
an $\alpha$=0.5 tends to select the groups in or out together.
This is a higher level parameter, and users might pick a value upfront,
else experiment with a few different values. One use of $\alpha$ is for numerical stability;
for example, the *elastic net with $\alpha = 1 - \epsilon$ for some small $\epsilon$>0
performs much like the lasso, but removes any degeneracies and wild behavior caused
by extreme correlations*.
```

## Sparsity stability



<!--chapter:end:00-featureSel.Rmd-->

# Fitted Model Suite {#model-suite}

We examine the results of fitting a suite of models to
investigate the effect of sample size on model performance.



<!--chapter:end:00-glmnetSuite.Rmd-->

# Conclusions

We have found that ...

Other questions ...


<!--chapter:end:00-summary.Rmd-->

`r if (knitr::is_html_output()) '
# References {-}
'`

<!--chapter:end:00-references.Rmd-->

