<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Section 8 BrCa RNA-Seq: Sample size investigation | DNA Hydroxymethylation in Hepatocellular Carcinoma</title>
  <meta name="description" content="Data from Cai et al. (2019) paper are explored" />
  <meta name="generator" content="bookdown 0.20.6 and GitBook 2.6.7" />

  <meta property="og:title" content="Section 8 BrCa RNA-Seq: Sample size investigation | DNA Hydroxymethylation in Hepatocellular Carcinoma" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Data from Cai et al. (2019) paper are explored" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Section 8 BrCa RNA-Seq: Sample size investigation | DNA Hydroxymethylation in Hepatocellular Carcinoma" />
  
  <meta name="twitter:description" content="Data from Cai et al. (2019) paper are explored" />
  

<meta name="author" content="Francois Collin" />


<meta name="date" content="2020-01-01" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="brca-rnaseseq-explore-sparsity.html"/>
<link rel="next" href="conclusions.html"/>
<script src="libs/header-attrs-2.4/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<script type="text/javascript">
// source:https://stackoverflow.com/questions/45360998/code-folding-in-bookdown
// toggle visibility of R source blocks in R Markdown output
function toggle_R() {
  var x = document.getElementsByClassName('r');
  if (x.length == 0) return;
  function toggle_vis(o) {
    var d = o.style.display;
    o.style.display = (d == 'block' || d == '') ? 'none':'block';
  }

  for (i = 0; i < x.length; i++) {
    var y = x[i];
    if (y.tagName.toLowerCase() === 'pre') toggle_vis(y);
  }

    var elem = document.getElementById("myButton1");
    if (elem.value === "Hide Global") elem.value = "Show Global";
    else elem.value = "Hide Global";
}

document.write('<input onclick="toggle_R();" type="button" value="Hide Global" id="myButton1" style="position: absolute; top: 10%; right: 2%; z-index: 200"></input>')

</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">DNA Hydroxymethylation in Hepatocellular Carcinoma</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preamble</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i>License</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="modeling-background.html"><a href="modeling-background.html"><i class="fa fa-check"></i><b>2</b> Classification Analysis in the <code>n &lt;&lt; p</code> Context</a>
<ul>
<li class="chapter" data-level="2.1" data-path="modeling-background.html"><a href="modeling-background.html#caret-for-model-evaluation"><i class="fa fa-check"></i><b>2.1</b> caret for model evaluation</a></li>
<li class="chapter" data-level="2.2" data-path="modeling-background.html"><a href="modeling-background.html#the-glmnet-r-package"><i class="fa fa-check"></i><b>2.2</b> The <code>glmnet</code> R package</a>
<ul>
<li><a href="modeling-background.html#alpha-hyper-parameter"><strong>alpha</strong> hyper-parameter</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="modeling-background.html"><a href="modeling-background.html#signal-to-noise"><i class="fa fa-check"></i><b>2.3</b> Signal-to-noise Ratio</a></li>
<li class="chapter" data-level="2.4" data-path="modeling-background.html"><a href="modeling-background.html#lasso-vs-best-sub"><i class="fa fa-check"></i><b>2.4</b> Lasso vs Best Subset</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html"><i class="fa fa-check"></i><b>3</b> HCC 5hmC-Seq : Preprocessing</a>
<ul>
<li class="chapter" data-level="3.1" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#load-the-data"><i class="fa fa-check"></i><b>3.1</b> Load the data</a></li>
<li class="chapter" data-level="3.2" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#dra"><i class="fa fa-check"></i><b>3.2</b> Differential representation analysis</a>
<ul>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#remove-lowly-expressed-genes"><i class="fa fa-check"></i>Remove lowly expressed genes</a></li>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#creating-a-design-matrix-and-contrasts"><i class="fa fa-check"></i>Creating a design matrix and contrasts</a></li>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#removing-heteroscedasticity-from-the-count-data"><i class="fa fa-check"></i>Removing heteroscedasticity from the count data</a></li>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#fit-linear-models-and-examine-the-results"><i class="fa fa-check"></i>Fit linear models and examine the results</a></li>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#graphical-representations-of-de-results-md-plots"><i class="fa fa-check"></i>Graphical representations of DE results: MD Plots</a></li>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#de-genes-at-10-fold-change"><i class="fa fa-check"></i>DE genes at 10% fold change</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#snr-regime"><i class="fa fa-check"></i><b>3.3</b> Signal-to-noise ratio regime</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html"><i class="fa fa-check"></i><b>4</b> HCC 5hmC-Seq: Exploring Sparsity</a>
<ul>
<li class="chapter" data-level="4.1" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#cross-validation-analysis-setup"><i class="fa fa-check"></i><b>4.1</b> Cross-validation analysis setup</a></li>
<li class="chapter" data-level="4.2" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#fit-and-compare-models"><i class="fa fa-check"></i><b>4.2</b> Fit and compare models</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#logistic-regression-in-glmnet"><i class="fa fa-check"></i><b>4.2.1</b> Logistic regression in <code>glmnet</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#the-relaxed-lasso-and-blended-mix-models"><i class="fa fa-check"></i><b>4.3</b> The relaxed lasso and blended mix models</a></li>
<li class="chapter" data-level="4.4" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#examination-of-sensitivity-vs-specificity"><i class="fa fa-check"></i><b>4.4</b> Examination of sensitivity vs specificity</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#training-data-out-of-fold-roc-curves"><i class="fa fa-check"></i><b>4.4.1</b> Training data out-of-fold ROC curves</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#compare-predictions-at-misclassified-samples"><i class="fa fa-check"></i><b>4.5</b> Compare predictions at misclassified samples</a></li>
<li class="chapter" data-level="4.6" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#compare-coefficient-profiles"><i class="fa fa-check"></i><b>4.6</b> Compare coefficient profiles</a></li>
<li class="chapter" data-level="4.7" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#examine-feature-selection"><i class="fa fa-check"></i><b>4.7</b> Examine feature selection</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html"><i class="fa fa-check"></i><b>5</b> HCC 5hmC-Seq: Sample size investigation</a>
<ul>
<li class="chapter" data-level="5.1" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#full-data-set-fit"><i class="fa fa-check"></i><b>5.1</b> Full data set fit</a>
<ul>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#get-sample-consistency-scores"><i class="fa fa-check"></i>Get sample consistency scores</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#selected-feature-list-stability"><i class="fa fa-check"></i><b>5.2</b> Selected feature list stability</a></li>
<li class="chapter" data-level="5.3" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#simulation-design"><i class="fa fa-check"></i><b>5.3</b> Simulation Design</a></li>
<li class="chapter" data-level="5.4" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#setup-simulation"><i class="fa fa-check"></i><b>5.4</b> Setup simulation</a></li>
<li class="chapter" data-level="5.5" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#run-simulations"><i class="fa fa-check"></i><b>5.5</b> Run simulations</a></li>
<li class="chapter" data-level="5.6" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#simulation-results"><i class="fa fa-check"></i><b>5.6</b> Simulation results</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#results-for-one-simulation"><i class="fa fa-check"></i><b>5.6.1</b> Results for one simulation</a></li>
<li class="chapter" data-level="5.6.2" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#results-across-simulation-runs."><i class="fa fa-check"></i><b>5.6.2</b> Results across simulation runs.</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#effect-of-sample-consistency"><i class="fa fa-check"></i><b>5.7</b> Effect of sample consistency</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html"><i class="fa fa-check"></i><b>6</b> BrCa RNA-Seq: Preprocessing</a>
<ul>
<li class="chapter" data-level="6.1" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#load-the-data-1"><i class="fa fa-check"></i><b>6.1</b> Load the data</a></li>
<li class="chapter" data-level="6.2" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#breast-rnaseq-dra"><i class="fa fa-check"></i><b>6.2</b> Differential representation analysis</a>
<ul>
<li class="chapter" data-level="" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#remove-lowly-expressed-genes-1"><i class="fa fa-check"></i>Remove lowly expressed genes</a></li>
<li class="chapter" data-level="" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#creating-a-design-matrix-and-contrasts-1"><i class="fa fa-check"></i>Creating a design matrix and contrasts</a></li>
<li class="chapter" data-level="" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#fit-linear-models-and-examine-the-results-1"><i class="fa fa-check"></i>Fit linear models and examine the results</a></li>
<li class="chapter" data-level="" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#graphical-representations-of-de-results-md-plots-1"><i class="fa fa-check"></i>Graphical representations of DE results: MD Plots</a></li>
<li class="chapter" data-level="" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#de-genes-at-10-fold-change-1"><i class="fa fa-check"></i>DE genes at 10% fold change</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#breast-rnaseq-snr-regime"><i class="fa fa-check"></i><b>6.3</b> Signal-to-noise ratio regime</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html"><i class="fa fa-check"></i><b>7</b> BrCa RNA-Seq: Exploring Sparsity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#cross-validation-analysis-setup-1"><i class="fa fa-check"></i><b>7.1</b> Cross-validation analysis setup</a></li>
<li class="chapter" data-level="7.2" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#fit-and-compare-models-1"><i class="fa fa-check"></i><b>7.2</b> Fit and compare models</a></li>
<li class="chapter" data-level="7.3" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#the-relaxed-lasso-and-blended-mix-models-1"><i class="fa fa-check"></i><b>7.3</b> The relaxed lasso and blended mix models</a></li>
<li class="chapter" data-level="7.4" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#examination-of-sensitivity-vs-specificity-1"><i class="fa fa-check"></i><b>7.4</b> Examination of sensitivity vs specificity</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#training-data-out-of-fold-roc-curves-1"><i class="fa fa-check"></i><b>7.4.1</b> Training data out-of-fold ROC curves</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#compare-predictions-at-misclassified-samples-1"><i class="fa fa-check"></i><b>7.5</b> Compare predictions at misclassified samples</a></li>
<li class="chapter" data-level="7.6" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#compare-coefficient-profiles-1"><i class="fa fa-check"></i><b>7.6</b> Compare coefficient profiles</a></li>
<li class="chapter" data-level="7.7" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#examine-feature-selection-1"><i class="fa fa-check"></i><b>7.7</b> Examine feature selection</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html"><i class="fa fa-check"></i><b>8</b> BrCa RNA-Seq: Sample size investigation</a>
<ul>
<li class="chapter" data-level="8.1" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#full-data-set-fit-1"><i class="fa fa-check"></i><b>8.1</b> Full data set fit</a>
<ul>
<li class="chapter" data-level="" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#get-sample-consistency-scores-1"><i class="fa fa-check"></i>Get sample consistency scores</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#selected-feature-list-stability-1"><i class="fa fa-check"></i><b>8.2</b> Selected feature list stability</a></li>
<li class="chapter" data-level="8.3" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#simulation-design-1"><i class="fa fa-check"></i><b>8.3</b> Simulation Design</a></li>
<li class="chapter" data-level="8.4" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#setup-simulation-1"><i class="fa fa-check"></i><b>8.4</b> Setup simulation</a></li>
<li class="chapter" data-level="8.5" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#run-simulations-1"><i class="fa fa-check"></i><b>8.5</b> Run simulations</a></li>
<li class="chapter" data-level="8.6" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#simulation-results-1"><i class="fa fa-check"></i><b>8.6</b> Simulation results</a>
<ul>
<li class="chapter" data-level="8.6.1" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#results-for-one-simulation-1"><i class="fa fa-check"></i><b>8.6.1</b> Results for one simulation</a></li>
<li class="chapter" data-level="8.6.2" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#results-across-simulation-runs.-1"><i class="fa fa-check"></i><b>8.6.2</b> Results across simulation runs.</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="conclusions.html"><a href="conclusions.html"><i class="fa fa-check"></i><b>9</b> Conclusions</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html"><i class="fa fa-check"></i>Appendix 1 - Sample size in elastic net fits to HCC 5hmC-Seq Data</a>
<ul>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html#selected-feature-list-stability-2"><i class="fa fa-check"></i>Selected feature list stability</a></li>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html#run-simulations---enet"><i class="fa fa-check"></i>Run simulations - enet</a></li>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html#enet-simulation-results"><i class="fa fa-check"></i>enet Simulation results</a>
<ul>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html#simulation-results---look-at-one-simulation"><i class="fa fa-check"></i>Simulation Results - look at one simulation</a></li>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html#summarize-results-across-simulation-runs"><i class="fa fa-check"></i>Summarize results across simulation runs</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="LICENSE.md" style="font:em;" target="blank">Copyright (c) 2020 Francois Collin</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">DNA Hydroxymethylation in Hepatocellular Carcinoma</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="brca-rnaseq-model-suite" class="section level1" number="8">
<h1><span class="header-section-number">Section 8</span> BrCa RNA-Seq: Sample size investigation</h1>
<p>In this section we repeat the analyses ran on the HCC 5hmC data in
Section <a href="hcc-5hmcseq-model-suite.html#hcc-5hmcseq-model-suite">5</a> on the breast cancer
RNA-Seq data set to provide a contrasting SNR regime context.</p>
<div id="full-data-set-fit-1" class="section level2" number="8.1">
<h2><span class="header-section-number">8.1</span> Full data set fit</h2>
<p>We begin by fitting a model to the entire data set in order to:</p>
<ul>
<li><p>obtain a baseline clssification performance against which to judge the performance
obtained from the fits to smaller sample sets,</p></li>
<li><p>obtain sample consistency scores which can be used to explain variability
in the performance of model fitted to data sets of a fixed size, and</p></li>
<li><p>produce a <em>full model</em> gene signature which can be used to evaluate
the stability of selected features in models fitted to data sets of diffferent
sizes.</p></li>
</ul>
<p>First assemble the data set. This entails simply re-combining the
train and test data.</p>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="brca-rnaseq-model-suite.html#cb178-1" aria-hidden="true"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb178-2"><a href="brca-rnaseq-model-suite.html#cb178-2" aria-hidden="true"></a></span>
<span id="cb178-3"><a href="brca-rnaseq-model-suite.html#cb178-3" aria-hidden="true"></a><span class="co"># combine train and test </span></span>
<span id="cb178-4"><a href="brca-rnaseq-model-suite.html#cb178-4" aria-hidden="true"></a>brcaRna_all_geneExpr_mtx &lt;-<span class="st"> </span><span class="kw">rbind</span>(brcaRna_train_geneExpr_mtx, brcaRna_test_geneExpr_mtx)</span>
<span id="cb178-5"><a href="brca-rnaseq-model-suite.html#cb178-5" aria-hidden="true"></a></span>
<span id="cb178-6"><a href="brca-rnaseq-model-suite.html#cb178-6" aria-hidden="true"></a><span class="co"># we have to be careful with factors!</span></span>
<span id="cb178-7"><a href="brca-rnaseq-model-suite.html#cb178-7" aria-hidden="true"></a><span class="co"># We&#39;ll keep as a character and change to factor when needed</span></span>
<span id="cb178-8"><a href="brca-rnaseq-model-suite.html#cb178-8" aria-hidden="true"></a>brcaRna_all_group_vec &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb178-9"><a href="brca-rnaseq-model-suite.html#cb178-9" aria-hidden="true"></a> <span class="kw">as.character</span>(brcaRna_train_group_vec), </span>
<span id="cb178-10"><a href="brca-rnaseq-model-suite.html#cb178-10" aria-hidden="true"></a> <span class="kw">as.character</span>(brcaRna_test_group_vec)</span>
<span id="cb178-11"><a href="brca-rnaseq-model-suite.html#cb178-11" aria-hidden="true"></a>)</span>
<span id="cb178-12"><a href="brca-rnaseq-model-suite.html#cb178-12" aria-hidden="true"></a><span class="co"># I suspect adding names to vectors breaks one of the tidy commandments,</span></span>
<span id="cb178-13"><a href="brca-rnaseq-model-suite.html#cb178-13" aria-hidden="true"></a><span class="co"># but then again I am sure I have already offended the creed beyond salvation</span></span>
<span id="cb178-14"><a href="brca-rnaseq-model-suite.html#cb178-14" aria-hidden="true"></a><span class="kw">names</span>(brcaRna_all_group_vec) &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb178-15"><a href="brca-rnaseq-model-suite.html#cb178-15" aria-hidden="true"></a> <span class="kw">names</span>(brcaRna_train_group_vec),</span>
<span id="cb178-16"><a href="brca-rnaseq-model-suite.html#cb178-16" aria-hidden="true"></a> <span class="kw">names</span>(brcaRna_test_group_vec)</span>
<span id="cb178-17"><a href="brca-rnaseq-model-suite.html#cb178-17" aria-hidden="true"></a>)</span>
<span id="cb178-18"><a href="brca-rnaseq-model-suite.html#cb178-18" aria-hidden="true"></a></span>
<span id="cb178-19"><a href="brca-rnaseq-model-suite.html#cb178-19" aria-hidden="true"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">table</span>(<span class="dt">group =</span> brcaRna_all_group_vec),</span>
<span id="cb178-20"><a href="brca-rnaseq-model-suite.html#cb178-20" aria-hidden="true"></a>  <span class="dt">caption =</span> <span class="st">&quot;samples by group&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb178-21"><a href="brca-rnaseq-model-suite.html#cb178-21" aria-hidden="true"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:brca-rnaseq-get-all-data">Table 8.1: </span>samples by group
</caption>
<thead>
<tr>
<th style="text-align:left;">
group
</th>
<th style="text-align:right;">
Freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
LumA
</td>
<td style="text-align:right;">
1492
</td>
</tr>
<tr>
<td style="text-align:left;">
Other
</td>
<td style="text-align:right;">
933
</td>
</tr>
</tbody>
</table>
<p>Now fit the losso model through cross-validation.
Note that the results of a cv fit are random due to the
random allocation of samples to folds. We can reduce this
varibility by properly averaging results over repeated cv fits.
Here we will obtain sample consistency scores by averaging results
over 30 cv runs.</p>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="brca-rnaseq-model-suite.html#cb179-1" aria-hidden="true"></a><span class="kw">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb179-2"><a href="brca-rnaseq-model-suite.html#cb179-2" aria-hidden="true"></a></span>
<span id="cb179-3"><a href="brca-rnaseq-model-suite.html#cb179-3" aria-hidden="true"></a>start_time &lt;-<span class="st">  </span><span class="kw">proc.time</span>()</span>
<span id="cb179-4"><a href="brca-rnaseq-model-suite.html#cb179-4" aria-hidden="true"></a></span>
<span id="cb179-5"><a href="brca-rnaseq-model-suite.html#cb179-5" aria-hidden="true"></a>brcaRna_cv_lassoAll_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">30</span>, <span class="cf">function</span>(REP) {</span>
<span id="cb179-6"><a href="brca-rnaseq-model-suite.html#cb179-6" aria-hidden="true"></a>glmnet<span class="op">::</span><span class="kw">cv.glmnet</span>(</span>
<span id="cb179-7"><a href="brca-rnaseq-model-suite.html#cb179-7" aria-hidden="true"></a> <span class="dt">x =</span> brcaRna_all_geneExpr_mtx,</span>
<span id="cb179-8"><a href="brca-rnaseq-model-suite.html#cb179-8" aria-hidden="true"></a> <span class="dt">y =</span> <span class="kw">factor</span>(brcaRna_all_group_vec,<span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&#39;Other&#39;</span>, <span class="st">&#39;LumA&#39;</span>)),</span>
<span id="cb179-9"><a href="brca-rnaseq-model-suite.html#cb179-9" aria-hidden="true"></a> <span class="dt">alpha =</span> <span class="dv">1</span>,</span>
<span id="cb179-10"><a href="brca-rnaseq-model-suite.html#cb179-10" aria-hidden="true"></a> <span class="dt">family =</span> <span class="st">&#39;binomial&#39;</span>,</span>
<span id="cb179-11"><a href="brca-rnaseq-model-suite.html#cb179-11" aria-hidden="true"></a> <span class="dt">type.measure  =</span>  <span class="st">&quot;class&quot;</span>,</span>
<span id="cb179-12"><a href="brca-rnaseq-model-suite.html#cb179-12" aria-hidden="true"></a> <span class="dt">keep =</span> T,</span>
<span id="cb179-13"><a href="brca-rnaseq-model-suite.html#cb179-13" aria-hidden="true"></a> <span class="dt">nlambda =</span> <span class="dv">100</span></span>
<span id="cb179-14"><a href="brca-rnaseq-model-suite.html#cb179-14" aria-hidden="true"></a>)</span>
<span id="cb179-15"><a href="brca-rnaseq-model-suite.html#cb179-15" aria-hidden="true"></a>}</span>
<span id="cb179-16"><a href="brca-rnaseq-model-suite.html#cb179-16" aria-hidden="true"></a>)</span>
<span id="cb179-17"><a href="brca-rnaseq-model-suite.html#cb179-17" aria-hidden="true"></a></span>
<span id="cb179-18"><a href="brca-rnaseq-model-suite.html#cb179-18" aria-hidden="true"></a><span class="kw">message</span>(<span class="st">&quot;lassoAll time: &quot;</span>, <span class="kw">round</span>((<span class="kw">proc.time</span>() <span class="op">-</span><span class="st"> </span>start_time)[<span class="dv">3</span>],<span class="dv">2</span>),<span class="st">&quot;s&quot;</span>)</span></code></pre></div>
<!-- lasso-fit-all takes a while - save results -->
<!-- DO THIS ONCE -->
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="brca-rnaseq-model-suite.html#cb180-1" aria-hidden="true"></a> <span class="kw">load</span>(<span class="dt">file=</span><span class="kw">file.path</span>(<span class="st">&quot;RData&quot;</span>,<span class="st">&#39;brcaRna_cv_lassoAll_lst&#39;</span>))</span></code></pre></div>
<p>Examine the fits.</p>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="brca-rnaseq-model-suite.html#cb181-1" aria-hidden="true"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb181-2"><a href="brca-rnaseq-model-suite.html#cb181-2" aria-hidden="true"></a><span class="kw">plot</span>(</span>
<span id="cb181-3"><a href="brca-rnaseq-model-suite.html#cb181-3" aria-hidden="true"></a> <span class="kw">log</span>(brcaRna_cv_lassoAll_lst[[<span class="dv">1</span>]]<span class="op">$</span>lambda),</span>
<span id="cb181-4"><a href="brca-rnaseq-model-suite.html#cb181-4" aria-hidden="true"></a> brcaRna_cv_lassoAll_lst[[<span class="dv">1</span>]]<span class="op">$</span>cvm,</span>
<span id="cb181-5"><a href="brca-rnaseq-model-suite.html#cb181-5" aria-hidden="true"></a> <span class="dt">lwd=</span><span class="dv">2</span>,</span>
<span id="cb181-6"><a href="brca-rnaseq-model-suite.html#cb181-6" aria-hidden="true"></a> <span class="dt">xlab=</span><span class="st">&#39;log(Lambda)&#39;</span>, <span class="dt">ylab=</span><span class="st">&#39;CV Misclassification Error&#39;</span>, <span class="dt">type=</span><span class="st">&#39;l&#39;</span>, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">.5</span>)</span>
<span id="cb181-7"><a href="brca-rnaseq-model-suite.html#cb181-7" aria-hidden="true"></a>)</span>
<span id="cb181-8"><a href="brca-rnaseq-model-suite.html#cb181-8" aria-hidden="true"></a></span>
<span id="cb181-9"><a href="brca-rnaseq-model-suite.html#cb181-9" aria-hidden="true"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">2</span><span class="op">:</span><span class="kw">length</span>(brcaRna_cv_lassoAll_lst))</span>
<span id="cb181-10"><a href="brca-rnaseq-model-suite.html#cb181-10" aria-hidden="true"></a> <span class="kw">lines</span>(</span>
<span id="cb181-11"><a href="brca-rnaseq-model-suite.html#cb181-11" aria-hidden="true"></a>  <span class="kw">log</span>(brcaRna_cv_lassoAll_lst[[JJ]]<span class="op">$</span>lambda),</span>
<span id="cb181-12"><a href="brca-rnaseq-model-suite.html#cb181-12" aria-hidden="true"></a>  brcaRna_cv_lassoAll_lst[[JJ]]<span class="op">$</span>cvm,</span>
<span id="cb181-13"><a href="brca-rnaseq-model-suite.html#cb181-13" aria-hidden="true"></a>  <span class="dt">lwd=</span><span class="dv">2</span></span>
<span id="cb181-14"><a href="brca-rnaseq-model-suite.html#cb181-14" aria-hidden="true"></a>)</span></code></pre></div>
<div class="figure"><span id="fig:brca-rnaseq-plot-lassoAll"></span>
<img src="Static/figures/brca-rnaseq-plot-lassoAll-1.png" alt="Repeated cv lasso models fitted to all samples" width="576" />
<p class="caption">
Figure 8.1: Repeated cv lasso models fitted to all samples
</p>
</div>
<p>These cv curves are remarkably consistent meaning that the determination of the size or sparsity
of the model fitted through cross validation to the full data set is fairly precise:</p>
<!-- DONT CACHE THIS ??? -->
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="brca-rnaseq-model-suite.html#cb182-1" aria-hidden="true"></a><span class="kw">library</span>(magrittr)</span>
<span id="cb182-2"><a href="brca-rnaseq-model-suite.html#cb182-2" aria-hidden="true"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb182-3"><a href="brca-rnaseq-model-suite.html#cb182-3" aria-hidden="true"></a></span>
<span id="cb182-4"><a href="brca-rnaseq-model-suite.html#cb182-4" aria-hidden="true"></a><span class="co"># nzero</span></span>
<span id="cb182-5"><a href="brca-rnaseq-model-suite.html#cb182-5" aria-hidden="true"></a>nzero_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb182-6"><a href="brca-rnaseq-model-suite.html#cb182-6" aria-hidden="true"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb182-7"><a href="brca-rnaseq-model-suite.html#cb182-7" aria-hidden="true"></a></span>
<span id="cb182-8"><a href="brca-rnaseq-model-suite.html#cb182-8" aria-hidden="true"></a>nzero_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb182-9"><a href="brca-rnaseq-model-suite.html#cb182-9" aria-hidden="true"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb182-10"><a href="brca-rnaseq-model-suite.html#cb182-10" aria-hidden="true"></a></span>
<span id="cb182-11"><a href="brca-rnaseq-model-suite.html#cb182-11" aria-hidden="true"></a><span class="kw">boxplot</span>(<span class="kw">list</span>(<span class="st">`</span><span class="dt">1se</span><span class="st">`</span>=nzero_1se_vec, <span class="dt">min =</span> nzero_min_vec), <span class="dt">ylab=</span><span class="st">&quot;Selected Features&quot;</span>)</span>
<span id="cb182-12"><a href="brca-rnaseq-model-suite.html#cb182-12" aria-hidden="true"></a></span>
<span id="cb182-13"><a href="brca-rnaseq-model-suite.html#cb182-13" aria-hidden="true"></a><span class="co"># error</span></span>
<span id="cb182-14"><a href="brca-rnaseq-model-suite.html#cb182-14" aria-hidden="true"></a>error_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb182-15"><a href="brca-rnaseq-model-suite.html#cb182-15" aria-hidden="true"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb182-16"><a href="brca-rnaseq-model-suite.html#cb182-16" aria-hidden="true"></a></span>
<span id="cb182-17"><a href="brca-rnaseq-model-suite.html#cb182-17" aria-hidden="true"></a>error_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb182-18"><a href="brca-rnaseq-model-suite.html#cb182-18" aria-hidden="true"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb182-19"><a href="brca-rnaseq-model-suite.html#cb182-19" aria-hidden="true"></a></span>
<span id="cb182-20"><a href="brca-rnaseq-model-suite.html#cb182-20" aria-hidden="true"></a><span class="kw">boxplot</span>(</span>
<span id="cb182-21"><a href="brca-rnaseq-model-suite.html#cb182-21" aria-hidden="true"></a> <span class="kw">list</span>(<span class="st">`</span><span class="dt">1se</span><span class="st">`</span>=error_1se_vec, <span class="dt">min =</span> error_min_vec), </span>
<span id="cb182-22"><a href="brca-rnaseq-model-suite.html#cb182-22" aria-hidden="true"></a> <span class="dt">ylab=</span>brcaRna_cv_lassoAll_lst[[<span class="dv">1</span>]]<span class="op">$</span>name,</span>
<span id="cb182-23"><a href="brca-rnaseq-model-suite.html#cb182-23" aria-hidden="true"></a> <span class="dt">ylim=</span><span class="kw">c</span>(<span class="fl">0.10</span>, <span class="fl">.13</span>)</span>
<span id="cb182-24"><a href="brca-rnaseq-model-suite.html#cb182-24" aria-hidden="true"></a>)</span></code></pre></div>
<div class="figure"><span id="fig:brca-rnaseq-model-size-lassoAll"></span>
<img src="Static/figures/brca-rnaseq-model-size-lassoAll-1.png" alt="Feature selection and estimated error by repeated cv lasso models" width="768" />
<p class="caption">
Figure 8.2: Feature selection and estimated error by repeated cv lasso models
</p>
</div>
<div class="sourceCode" id="cb183"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb183-1"><a href="brca-rnaseq-model-suite.html#cb183-1" aria-hidden="true"></a><span class="co"># tabular format</span></span>
<span id="cb183-2"><a href="brca-rnaseq-model-suite.html#cb183-2" aria-hidden="true"></a>tmp &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">rbind</span>(</span>
<span id="cb183-3"><a href="brca-rnaseq-model-suite.html#cb183-3" aria-hidden="true"></a> <span class="st">`</span><span class="dt">features_1se</span><span class="st">`</span> =<span class="st"> </span><span class="kw">summary</span>(nzero_1se_vec),</span>
<span id="cb183-4"><a href="brca-rnaseq-model-suite.html#cb183-4" aria-hidden="true"></a> <span class="dt">features_min =</span> <span class="kw">summary</span>(nzero_min_vec),</span>
<span id="cb183-5"><a href="brca-rnaseq-model-suite.html#cb183-5" aria-hidden="true"></a> <span class="st">`</span><span class="dt">features:min-1se</span><span class="st">`</span> =<span class="st"> </span><span class="kw">summary</span>(nzero_min_vec <span class="op">-</span><span class="st"> </span>nzero_1se_vec),</span>
<span id="cb183-6"><a href="brca-rnaseq-model-suite.html#cb183-6" aria-hidden="true"></a> <span class="st">`</span><span class="dt">cv_error_1se</span><span class="st">`</span> =<span class="st"> </span><span class="kw">summary</span>(<span class="dv">100</span><span class="op">*</span>error_1se_vec),</span>
<span id="cb183-7"><a href="brca-rnaseq-model-suite.html#cb183-7" aria-hidden="true"></a> <span class="dt">cv_error_min =</span> <span class="kw">summary</span>(<span class="dv">100</span><span class="op">*</span>error_min_vec),</span>
<span id="cb183-8"><a href="brca-rnaseq-model-suite.html#cb183-8" aria-hidden="true"></a> <span class="st">`</span><span class="dt">cv_error:1se-min</span><span class="st">`</span> =<span class="st"> </span><span class="kw">summary</span>(<span class="dv">100</span><span class="op">*</span>(error_1se_vec<span class="op">-</span>error_min_vec))</span>
<span id="cb183-9"><a href="brca-rnaseq-model-suite.html#cb183-9" aria-hidden="true"></a>))</span>
<span id="cb183-10"><a href="brca-rnaseq-model-suite.html#cb183-10" aria-hidden="true"></a></span>
<span id="cb183-11"><a href="brca-rnaseq-model-suite.html#cb183-11" aria-hidden="true"></a>knitr<span class="op">::</span><span class="kw">kable</span>(tmp <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(<span class="op">-</span>Mean),</span>
<span id="cb183-12"><a href="brca-rnaseq-model-suite.html#cb183-12" aria-hidden="true"></a>  <span class="dt">caption =</span> <span class="st">&quot;Number of selected features&quot;</span>,</span>
<span id="cb183-13"><a href="brca-rnaseq-model-suite.html#cb183-13" aria-hidden="true"></a>  <span class="dt">digits=</span><span class="dv">1</span>) <span class="op">%&gt;%</span></span>
<span id="cb183-14"><a href="brca-rnaseq-model-suite.html#cb183-14" aria-hidden="true"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:brca-rnaseq-model-size-lassoAll">Table 8.2: </span>Number of selected features
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Min.
</th>
<th style="text-align:right;">
X1st.Qu.
</th>
<th style="text-align:right;">
Median
</th>
<th style="text-align:right;">
X3rd.Qu.
</th>
<th style="text-align:right;">
Max.
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
features_1se
</td>
<td style="text-align:right;">
97.0
</td>
<td style="text-align:right;">
118.5
</td>
<td style="text-align:right;">
144.0
</td>
<td style="text-align:right;">
162.0
</td>
<td style="text-align:right;">
208.0
</td>
</tr>
<tr>
<td style="text-align:left;">
features_min
</td>
<td style="text-align:right;">
144.0
</td>
<td style="text-align:right;">
221.2
</td>
<td style="text-align:right;">
258.0
</td>
<td style="text-align:right;">
291.0
</td>
<td style="text-align:right;">
451.0
</td>
</tr>
<tr>
<td style="text-align:left;">
features:min-1se
</td>
<td style="text-align:right;">
27.0
</td>
<td style="text-align:right;">
59.0
</td>
<td style="text-align:right;">
94.5
</td>
<td style="text-align:right;">
153.2
</td>
<td style="text-align:right;">
344.0
</td>
</tr>
<tr>
<td style="text-align:left;">
cv_error_1se
</td>
<td style="text-align:right;">
10.9
</td>
<td style="text-align:right;">
11.3
</td>
<td style="text-align:right;">
11.5
</td>
<td style="text-align:right;">
11.7
</td>
<td style="text-align:right;">
12.0
</td>
</tr>
<tr>
<td style="text-align:left;">
cv_error_min
</td>
<td style="text-align:right;">
10.6
</td>
<td style="text-align:right;">
10.8
</td>
<td style="text-align:right;">
10.9
</td>
<td style="text-align:right;">
11.1
</td>
<td style="text-align:right;">
11.5
</td>
</tr>
<tr>
<td style="text-align:left;">
cv_error:1se-min
</td>
<td style="text-align:right;">
0.2
</td>
<td style="text-align:right;">
0.4
</td>
<td style="text-align:right;">
0.5
</td>
<td style="text-align:right;">
0.7
</td>
<td style="text-align:right;">
0.8
</td>
</tr>
</tbody>
</table>
<p><br/></p>
<p>The number of features selected by the minimum lambda models are larger
than the number selected by the “one standard error” rule models by a median
of 94.5.
The cv error rates obtained from the minimum lambda models are lower
then “one standard error” rule models error rates by a median of
0.5%.</p>
<p>The cv error rates observed in this set are comparable to the
rates oberved in the lasso models fitted to the training sample set
which consisted of 80% of the samples in this set. In other words,
there is no obvious gain in performance in moving from
a data set with
1194 vs 746 samples
to a data set with
1492 vs 933 samples.
See Table <a href="#tab:printErrors"><strong>??</strong></a>.</p>
<p>It’s not clear at this point whether the minimum lambda model is truly better than
the “one standard error” rule model. We would need and external validation
set to make this determination. We can compare the two sets
of out-of-fold predicted values, averaged across cv replicates, to see if
there is a meaningful difference between the two.</p>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="brca-rnaseq-model-suite.html#cb184-1" aria-hidden="true"></a><span class="co"># predicted probs - 1se</span></span>
<span id="cb184-2"><a href="brca-rnaseq-model-suite.html#cb184-2" aria-hidden="true"></a>lassoAll_predResp_1se_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst, <span class="cf">function</span>(cv_fit) { </span>
<span id="cb184-3"><a href="brca-rnaseq-model-suite.html#cb184-3" aria-hidden="true"></a>  ndx_1se &lt;-<span class="st"> </span><span class="kw">match</span>(cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se,cv_fit<span class="op">$</span>lambda)</span>
<span id="cb184-4"><a href="brca-rnaseq-model-suite.html#cb184-4" aria-hidden="true"></a>  <span class="kw">logistic_f</span>(cv_fit<span class="op">$</span>fit.preval[,ndx_1se])</span>
<span id="cb184-5"><a href="brca-rnaseq-model-suite.html#cb184-5" aria-hidden="true"></a> })</span>
<span id="cb184-6"><a href="brca-rnaseq-model-suite.html#cb184-6" aria-hidden="true"></a>lassoAll_predResp_1se_vec &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(lassoAll_predResp_1se_mtx)</span>
<span id="cb184-7"><a href="brca-rnaseq-model-suite.html#cb184-7" aria-hidden="true"></a></span>
<span id="cb184-8"><a href="brca-rnaseq-model-suite.html#cb184-8" aria-hidden="true"></a><span class="co"># predicted probs - min</span></span>
<span id="cb184-9"><a href="brca-rnaseq-model-suite.html#cb184-9" aria-hidden="true"></a>lassoAll_predResp_min_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst, <span class="cf">function</span>(cv_fit) { </span>
<span id="cb184-10"><a href="brca-rnaseq-model-suite.html#cb184-10" aria-hidden="true"></a>  ndx_min &lt;-<span class="st"> </span><span class="kw">match</span>(cv_fit<span class="op">$</span>lambda.min,cv_fit<span class="op">$</span>lambda)</span>
<span id="cb184-11"><a href="brca-rnaseq-model-suite.html#cb184-11" aria-hidden="true"></a>  <span class="kw">logistic_f</span>(cv_fit<span class="op">$</span>fit.preval[,ndx_min])</span>
<span id="cb184-12"><a href="brca-rnaseq-model-suite.html#cb184-12" aria-hidden="true"></a> })</span>
<span id="cb184-13"><a href="brca-rnaseq-model-suite.html#cb184-13" aria-hidden="true"></a>lassoAll_predResp_min_vec &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(lassoAll_predResp_min_mtx)</span>
<span id="cb184-14"><a href="brca-rnaseq-model-suite.html#cb184-14" aria-hidden="true"></a></span>
<span id="cb184-15"><a href="brca-rnaseq-model-suite.html#cb184-15" aria-hidden="true"></a><span class="co"># plot</span></span>
<span id="cb184-16"><a href="brca-rnaseq-model-suite.html#cb184-16" aria-hidden="true"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">5</span>,<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb184-17"><a href="brca-rnaseq-model-suite.html#cb184-17" aria-hidden="true"></a>tmp &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb184-18"><a href="brca-rnaseq-model-suite.html#cb184-18" aria-hidden="true"></a> <span class="st">`</span><span class="dt">1se</span><span class="st">`</span> =<span class="st"> </span><span class="kw">split</span>(lassoAll_predResp_1se_vec, brcaRna_all_group_vec),</span>
<span id="cb184-19"><a href="brca-rnaseq-model-suite.html#cb184-19" aria-hidden="true"></a> <span class="dt">min =</span> <span class="kw">split</span>(lassoAll_predResp_min_vec, brcaRna_all_group_vec)</span>
<span id="cb184-20"><a href="brca-rnaseq-model-suite.html#cb184-20" aria-hidden="true"></a>)</span>
<span id="cb184-21"><a href="brca-rnaseq-model-suite.html#cb184-21" aria-hidden="true"></a><span class="kw">names</span>(tmp) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&#39;</span><span class="ch">\\</span><span class="st">.&#39;</span>, <span class="st">&#39;</span><span class="ch">\n</span><span class="st">&#39;</span>, <span class="kw">names</span>(tmp))</span>
<span id="cb184-22"><a href="brca-rnaseq-model-suite.html#cb184-22" aria-hidden="true"></a></span>
<span id="cb184-23"><a href="brca-rnaseq-model-suite.html#cb184-23" aria-hidden="true"></a><span class="kw">boxplot</span>(</span>
<span id="cb184-24"><a href="brca-rnaseq-model-suite.html#cb184-24" aria-hidden="true"></a> tmp,</span>
<span id="cb184-25"><a href="brca-rnaseq-model-suite.html#cb184-25" aria-hidden="true"></a> <span class="dt">ylab=</span><span class="st">&#39;Predicted oof probability&#39;</span>,</span>
<span id="cb184-26"><a href="brca-rnaseq-model-suite.html#cb184-26" aria-hidden="true"></a> <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;green&#39;</span>, <span class="st">&#39;red&#39;</span>),</span>
<span id="cb184-27"><a href="brca-rnaseq-model-suite.html#cb184-27" aria-hidden="true"></a> <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb184-28"><a href="brca-rnaseq-model-suite.html#cb184-28" aria-hidden="true"></a>)</span>
<span id="cb184-29"><a href="brca-rnaseq-model-suite.html#cb184-29" aria-hidden="true"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(tmp), <span class="dt">tick=</span>F, <span class="kw">names</span>(tmp))</span>
<span id="cb184-30"><a href="brca-rnaseq-model-suite.html#cb184-30" aria-hidden="true"></a></span>
<span id="cb184-31"><a href="brca-rnaseq-model-suite.html#cb184-31" aria-hidden="true"></a></span>
<span id="cb184-32"><a href="brca-rnaseq-model-suite.html#cb184-32" aria-hidden="true"></a><span class="co"># compare the two</span></span>
<span id="cb184-33"><a href="brca-rnaseq-model-suite.html#cb184-33" aria-hidden="true"></a><span class="kw">plot</span>(</span>
<span id="cb184-34"><a href="brca-rnaseq-model-suite.html#cb184-34" aria-hidden="true"></a> <span class="dt">x =</span> lassoAll_predResp_1se_vec, <span class="dt">xlab=</span><span class="st">&#39;1se model oof Prob&#39;</span>,</span>
<span id="cb184-35"><a href="brca-rnaseq-model-suite.html#cb184-35" aria-hidden="true"></a> <span class="dt">y =</span> lassoAll_predResp_min_vec, <span class="dt">ylab=</span><span class="st">&#39;min lambda model oof Prob&#39;</span>,</span>
<span id="cb184-36"><a href="brca-rnaseq-model-suite.html#cb184-36" aria-hidden="true"></a> <span class="dt">col =</span> <span class="kw">ifelse</span>(brcaRna_all_group_vec <span class="op">==</span><span class="st"> &#39;LumA&#39;</span>, <span class="st">&#39;red&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb184-37"><a href="brca-rnaseq-model-suite.html#cb184-37" aria-hidden="true"></a>)</span>
<span id="cb184-38"><a href="brca-rnaseq-model-suite.html#cb184-38" aria-hidden="true"></a> </span>
<span id="cb184-39"><a href="brca-rnaseq-model-suite.html#cb184-39" aria-hidden="true"></a><span class="co"># Add reference lines at 10% false positive</span></span>
<span id="cb184-40"><a href="brca-rnaseq-model-suite.html#cb184-40" aria-hidden="true"></a>thres_1se &lt;-<span class="st"> </span><span class="kw">quantile</span>(lassoAll_predResp_1se_vec[brcaRna_all_group_vec <span class="op">==</span><span class="st"> &#39;Other&#39;</span>], <span class="dt">prob=</span>.<span class="dv">9</span>)</span>
<span id="cb184-41"><a href="brca-rnaseq-model-suite.html#cb184-41" aria-hidden="true"></a>thres_min &lt;-<span class="st"> </span><span class="kw">quantile</span>(lassoAll_predResp_min_vec[brcaRna_all_group_vec <span class="op">==</span><span class="st"> &#39;Other&#39;</span>], <span class="dt">prob=</span>.<span class="dv">9</span>)</span>
<span id="cb184-42"><a href="brca-rnaseq-model-suite.html#cb184-42" aria-hidden="true"></a><span class="kw">abline</span>(<span class="dt">v =</span> thres_1se, <span class="dt">h =</span> thres_min, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span></code></pre></div>
<div class="figure"><span id="fig:brca-rnaseq-get-sample-pred"></span>
<img src="Static/figures/brca-rnaseq-get-sample-pred-1.png" alt="Predicted probabilities - averaged over cv replicates" width="960" />
<p class="caption">
Figure 8.3: Predicted probabilities - averaged over cv replicates
</p>
</div>
<!-- THIS PARAGRAPH REFERRED TO THE FITTED PROBS; NOT THE OOF PRED PROBS
We see that the minimum lambda models provide a better fit to the data,
which is to be expected as the minimum lambda models have more estimated
parameters than the one standard error rule models.  
-->
<p>We see that there isn’t a big difference in out-of-fold predicted
probabilities between the one-standard-error rule and the minimum lamda models.
One way to quantify
the difference in classification errors is to classify samples
according to each vector of predicted probabilities, setting
the thresholds to achieve a fixed false positive rate, 10% say.
These thresholds are indicated by the grey lines in the scatter plot
on the right side of Figure <a href="#fig:get-sample-pred"><strong>??</strong></a>.</p>
<!-- APPLIED TO THE FITTED VALUES
We note
that althouth predicted probability distributions are quite different
for the two models, the the class predictions at a 10% false discovery threshold
are largely in agreement.
-->
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb185-1"><a href="brca-rnaseq-model-suite.html#cb185-1" aria-hidden="true"></a><span class="co">### </span><span class="al">NOTE</span><span class="co"> THAT HERE MODEL PRODUCES PROB(Other)</span></span>
<span id="cb185-2"><a href="brca-rnaseq-model-suite.html#cb185-2" aria-hidden="true"></a>lassoAll_predClass_1se_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb185-3"><a href="brca-rnaseq-model-suite.html#cb185-3" aria-hidden="true"></a> lassoAll_predResp_1se_vec <span class="op">&gt;</span><span class="st"> </span>thres_1se, <span class="st">&#39;LumA&#39;</span>, <span class="st">&#39;Other&#39;</span>)</span>
<span id="cb185-4"><a href="brca-rnaseq-model-suite.html#cb185-4" aria-hidden="true"></a></span>
<span id="cb185-5"><a href="brca-rnaseq-model-suite.html#cb185-5" aria-hidden="true"></a>lassoAll_predClass_min_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb185-6"><a href="brca-rnaseq-model-suite.html#cb185-6" aria-hidden="true"></a> lassoAll_predResp_min_vec <span class="op">&gt;</span><span class="st"> </span>thres_min, <span class="st">&#39;LumA&#39;</span>, <span class="st">&#39;Other&#39;</span>)</span>
<span id="cb185-7"><a href="brca-rnaseq-model-suite.html#cb185-7" aria-hidden="true"></a></span>
<span id="cb185-8"><a href="brca-rnaseq-model-suite.html#cb185-8" aria-hidden="true"></a>tmp &lt;-<span class="st"> </span><span class="kw">cbind</span>(</span>
<span id="cb185-9"><a href="brca-rnaseq-model-suite.html#cb185-9" aria-hidden="true"></a> <span class="kw">table</span>(<span class="dt">truth=</span>brcaRna_all_group_vec, <span class="st">`</span><span class="dt">1se-pred</span><span class="st">`</span>=lassoAll_predClass_1se_vec),</span>
<span id="cb185-10"><a href="brca-rnaseq-model-suite.html#cb185-10" aria-hidden="true"></a> <span class="kw">table</span>(<span class="dt">truth=</span>brcaRna_all_group_vec, <span class="st">`</span><span class="dt">min-pred</span><span class="st">`</span>=lassoAll_predClass_min_vec)</span>
<span id="cb185-11"><a href="brca-rnaseq-model-suite.html#cb185-11" aria-hidden="true"></a>) </span>
<span id="cb185-12"><a href="brca-rnaseq-model-suite.html#cb185-12" aria-hidden="true"></a><span class="co"># HARD-WIRED Hack for printing!!!!</span></span>
<span id="cb185-13"><a href="brca-rnaseq-model-suite.html#cb185-13" aria-hidden="true"></a><span class="kw">colnames</span>(tmp) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;1se-LumA&#39;</span>, <span class="st">&#39;1se-Other&#39;</span>, <span class="st">&#39;min-LumA&#39;</span>, <span class="st">&#39;min-Other&#39;</span>)</span>
<span id="cb185-14"><a href="brca-rnaseq-model-suite.html#cb185-14" aria-hidden="true"></a></span>
<span id="cb185-15"><a href="brca-rnaseq-model-suite.html#cb185-15" aria-hidden="true"></a>knitr<span class="op">::</span><span class="kw">kable</span>(tmp,</span>
<span id="cb185-16"><a href="brca-rnaseq-model-suite.html#cb185-16" aria-hidden="true"></a>  <span class="dt">caption =</span> <span class="st">&quot;Classifications: rows are truth&quot;</span>,</span>
<span id="cb185-17"><a href="brca-rnaseq-model-suite.html#cb185-17" aria-hidden="true"></a>  <span class="dt">digits=</span><span class="dv">1</span>) <span class="op">%&gt;%</span></span>
<span id="cb185-18"><a href="brca-rnaseq-model-suite.html#cb185-18" aria-hidden="true"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:brca-rnaseq-get-sample-class">Table 8.3: </span>Classifications: rows are truth
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
1se-LumA
</th>
<th style="text-align:right;">
1se-Other
</th>
<th style="text-align:right;">
min-LumA
</th>
<th style="text-align:right;">
min-Other
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
LumA
</td>
<td style="text-align:right;">
1288
</td>
<td style="text-align:right;">
204
</td>
<td style="text-align:right;">
1304
</td>
<td style="text-align:right;">
188
</td>
</tr>
<tr>
<td style="text-align:left;">
Other
</td>
<td style="text-align:right;">
94
</td>
<td style="text-align:right;">
839
</td>
<td style="text-align:right;">
94
</td>
<td style="text-align:right;">
839
</td>
</tr>
</tbody>
</table>
<p>When we fix the false positive rate at 10% (ie. the other samples error rates are fixed),
the <code>1se</code> model makes 204 false negative calls whereas the
minimum lambda model makes 188. A difference of 1.1%</p>
<!-- APPLIED TO THE FITTED PROBABILITIES
We see that the min lambda model, makes no false negative calls at a 90% sensitivity
setting, and the sensitivity could be increased substantially at no false negative
cost.  This is definitely over-fitting the data set.  For the purpose
of computing sample consistency scores - what do these differences mean? 
-->
<div id="get-sample-consistency-scores-1" class="section level3 unnumbered">
<h3>Get sample consistency scores</h3>
<p>To compute consistency scores, we will use the out-of-fold predicted probabilities.</p>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="brca-rnaseq-model-suite.html#cb186-1" aria-hidden="true"></a><span class="co"># get qual scores</span></span>
<span id="cb186-2"><a href="brca-rnaseq-model-suite.html#cb186-2" aria-hidden="true"></a></span>
<span id="cb186-3"><a href="brca-rnaseq-model-suite.html#cb186-3" aria-hidden="true"></a>y &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(brcaRna_all_group_vec <span class="op">==</span><span class="st"> &#39;LumA&#39;</span>)</span>
<span id="cb186-4"><a href="brca-rnaseq-model-suite.html#cb186-4" aria-hidden="true"></a><span class="co"># 1se</span></span>
<span id="cb186-5"><a href="brca-rnaseq-model-suite.html#cb186-5" aria-hidden="true"></a>p &lt;-<span class="st"> </span>lassoAll_predResp_1se_vec</span>
<span id="cb186-6"><a href="brca-rnaseq-model-suite.html#cb186-6" aria-hidden="true"></a>brcaRna_sample_1se_qual_vec &lt;-<span class="st"> </span>p<span class="op">^</span>y<span class="op">*</span>(<span class="dv">1</span><span class="op">-</span>p)<span class="op">^</span>(<span class="dv">1</span><span class="op">-</span>y)</span>
<span id="cb186-7"><a href="brca-rnaseq-model-suite.html#cb186-7" aria-hidden="true"></a></span>
<span id="cb186-8"><a href="brca-rnaseq-model-suite.html#cb186-8" aria-hidden="true"></a><span class="co"># min</span></span>
<span id="cb186-9"><a href="brca-rnaseq-model-suite.html#cb186-9" aria-hidden="true"></a>p &lt;-<span class="st"> </span>lassoAll_predResp_min_vec</span>
<span id="cb186-10"><a href="brca-rnaseq-model-suite.html#cb186-10" aria-hidden="true"></a>brcaRna_sample_min_qual_vec &lt;-<span class="st"> </span>p<span class="op">^</span>y<span class="op">*</span>(<span class="dv">1</span><span class="op">-</span>p)<span class="op">^</span>(<span class="dv">1</span><span class="op">-</span>y)</span></code></pre></div>
<p>We can examine consistency scores as a function of classification bin.</p>
<div class="sourceCode" id="cb187"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb187-1"><a href="brca-rnaseq-model-suite.html#cb187-1" aria-hidden="true"></a>y &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(brcaRna_all_group_vec <span class="op">==</span><span class="st"> &#39;LumA&#39;</span>)</span>
<span id="cb187-2"><a href="brca-rnaseq-model-suite.html#cb187-2" aria-hidden="true"></a></span>
<span id="cb187-3"><a href="brca-rnaseq-model-suite.html#cb187-3" aria-hidden="true"></a><span class="co"># 1se</span></span>
<span id="cb187-4"><a href="brca-rnaseq-model-suite.html#cb187-4" aria-hidden="true"></a>lassoAll_1se_conf_vec &lt;-<span class="st"> </span><span class="kw">paste</span>(</span>
<span id="cb187-5"><a href="brca-rnaseq-model-suite.html#cb187-5" aria-hidden="true"></a> y, </span>
<span id="cb187-6"><a href="brca-rnaseq-model-suite.html#cb187-6" aria-hidden="true"></a> <span class="kw">as.numeric</span>(lassoAll_predClass_1se_vec<span class="op">==</span><span class="st">&#39;LumA&#39;</span>),</span>
<span id="cb187-7"><a href="brca-rnaseq-model-suite.html#cb187-7" aria-hidden="true"></a> <span class="dt">sep =</span> <span class="st">&#39;:&#39;</span></span>
<span id="cb187-8"><a href="brca-rnaseq-model-suite.html#cb187-8" aria-hidden="true"></a>)</span>
<span id="cb187-9"><a href="brca-rnaseq-model-suite.html#cb187-9" aria-hidden="true"></a></span>
<span id="cb187-10"><a href="brca-rnaseq-model-suite.html#cb187-10" aria-hidden="true"></a><span class="co"># min</span></span>
<span id="cb187-11"><a href="brca-rnaseq-model-suite.html#cb187-11" aria-hidden="true"></a>lassoAll_min_conf_vec &lt;-<span class="st"> </span><span class="kw">paste</span>(</span>
<span id="cb187-12"><a href="brca-rnaseq-model-suite.html#cb187-12" aria-hidden="true"></a> y, </span>
<span id="cb187-13"><a href="brca-rnaseq-model-suite.html#cb187-13" aria-hidden="true"></a> <span class="kw">as.numeric</span>(lassoAll_predClass_min_vec<span class="op">==</span><span class="st">&#39;LumA&#39;</span>),</span>
<span id="cb187-14"><a href="brca-rnaseq-model-suite.html#cb187-14" aria-hidden="true"></a> <span class="dt">sep =</span> <span class="st">&#39;:&#39;</span></span>
<span id="cb187-15"><a href="brca-rnaseq-model-suite.html#cb187-15" aria-hidden="true"></a>)</span>
<span id="cb187-16"><a href="brca-rnaseq-model-suite.html#cb187-16" aria-hidden="true"></a></span>
<span id="cb187-17"><a href="brca-rnaseq-model-suite.html#cb187-17" aria-hidden="true"></a></span>
<span id="cb187-18"><a href="brca-rnaseq-model-suite.html#cb187-18" aria-hidden="true"></a>tmp &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb187-19"><a href="brca-rnaseq-model-suite.html#cb187-19" aria-hidden="true"></a> <span class="kw">split</span>(brcaRna_sample_1se_qual_vec, lassoAll_1se_conf_vec), </span>
<span id="cb187-20"><a href="brca-rnaseq-model-suite.html#cb187-20" aria-hidden="true"></a> <span class="kw">split</span>(brcaRna_sample_min_qual_vec, lassoAll_min_conf_vec)</span>
<span id="cb187-21"><a href="brca-rnaseq-model-suite.html#cb187-21" aria-hidden="true"></a>)</span>
<span id="cb187-22"><a href="brca-rnaseq-model-suite.html#cb187-22" aria-hidden="true"></a></span>
<span id="cb187-23"><a href="brca-rnaseq-model-suite.html#cb187-23" aria-hidden="true"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">3</span>,<span class="dv">3</span>,<span class="dv">2</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb187-24"><a href="brca-rnaseq-model-suite.html#cb187-24" aria-hidden="true"></a>gplots<span class="op">::</span><span class="kw">boxplot2</span>(<span class="kw">split</span>(brcaRna_sample_1se_qual_vec, lassoAll_1se_conf_vec), </span>
<span id="cb187-25"><a href="brca-rnaseq-model-suite.html#cb187-25" aria-hidden="true"></a>  <span class="dt">outline=</span>F, <span class="dt">ylab =</span> <span class="st">&#39;&#39;</span>, </span>
<span id="cb187-26"><a href="brca-rnaseq-model-suite.html#cb187-26" aria-hidden="true"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;green&#39;</span>, <span class="st">&#39;green&#39;</span>, <span class="st">&#39;red&#39;</span>, <span class="st">&#39;red&#39;</span>),</span>
<span id="cb187-27"><a href="brca-rnaseq-model-suite.html#cb187-27" aria-hidden="true"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>))</span>
<span id="cb187-28"><a href="brca-rnaseq-model-suite.html#cb187-28" aria-hidden="true"></a><span class="kw">title</span>(<span class="st">&#39;1se Model&#39;</span>)</span>
<span id="cb187-29"><a href="brca-rnaseq-model-suite.html#cb187-29" aria-hidden="true"></a></span>
<span id="cb187-30"><a href="brca-rnaseq-model-suite.html#cb187-30" aria-hidden="true"></a>gplots<span class="op">::</span><span class="kw">boxplot2</span>(<span class="kw">split</span>(brcaRna_sample_min_qual_vec, lassoAll_min_conf_vec), </span>
<span id="cb187-31"><a href="brca-rnaseq-model-suite.html#cb187-31" aria-hidden="true"></a>  <span class="dt">outline=</span>F, <span class="dt">ylab =</span> <span class="st">&#39;&#39;</span>,</span>
<span id="cb187-32"><a href="brca-rnaseq-model-suite.html#cb187-32" aria-hidden="true"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;green&#39;</span>, <span class="st">&#39;green&#39;</span>, <span class="st">&#39;red&#39;</span>, <span class="st">&#39;red&#39;</span>),</span>
<span id="cb187-33"><a href="brca-rnaseq-model-suite.html#cb187-33" aria-hidden="true"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>))</span>
<span id="cb187-34"><a href="brca-rnaseq-model-suite.html#cb187-34" aria-hidden="true"></a><span class="kw">title</span>(<span class="st">&#39;min lambda Model&#39;</span>)</span>
<span id="cb187-35"><a href="brca-rnaseq-model-suite.html#cb187-35" aria-hidden="true"></a></span>
<span id="cb187-36"><a href="brca-rnaseq-model-suite.html#cb187-36" aria-hidden="true"></a></span>
<span id="cb187-37"><a href="brca-rnaseq-model-suite.html#cb187-37" aria-hidden="true"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.5</span>, <span class="st">&#39;Classification - Truth:Predicted&#39;</span>)</span>
<span id="cb187-38"><a href="brca-rnaseq-model-suite.html#cb187-38" aria-hidden="true"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.5</span>, <span class="st">&#39;Consistency Score&#39;</span>)</span>
<span id="cb187-39"><a href="brca-rnaseq-model-suite.html#cb187-39" aria-hidden="true"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.5</span>, <span class="st">&#39;Sample Consistency vs Classification Outcome&#39;</span>)</span></code></pre></div>
<div class="figure"><span id="fig:brca-rnaseq-plot-qual-conf"></span>
<img src="Static/figures/brca-rnaseq-plot-qual-conf-1.png" alt="consistency scores by classification - Other=0, LumA=1" width="768" />
<p class="caption">
Figure 8.4: consistency scores by classification - Other=0, LumA=1
</p>
</div>
<p>This figure shows that for false positive cases (0:1 or classifying a
other as an LumA case), the algorithm is <em>more certain</em> of its predicted
outcome than for the false negative cases (1:0 or classifying an LumA case as Other).
ie. the misclassified Other samples are quite similar to LumA sample, whereas there
is more ambiguity in the misclassified LumA samples.</p>
<p>We will use the minimum lambda model to provide
the fitted probabilities used to compute consistency scores,
but we could have used either one.</p>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="brca-rnaseq-model-suite.html#cb188-1" aria-hidden="true"></a>brcaRna_sample_qual_vec &lt;-<span class="st"> </span>brcaRna_sample_min_qual_vec</span></code></pre></div>
</div>
</div>
<div id="selected-feature-list-stability-1" class="section level2" number="8.2">
<h2><span class="header-section-number">8.2</span> Selected feature list stability</h2>
<p>Before moving on to the simulation, let’s examine gene selection stability on the
full data set. We have two sets of sellected features - one for the
one standard deviation rule model, and one for the mimimum lambda model.
We saw in Table <a href="#tab:model-size-lassoAll"><strong>??</strong></a> that the number of features
selected by the minimum lambda models had an IQR of
221.25-291,
while the one standard error rule models had an IQR of
118.5-162.</p>
<p>Let’s examine the stability of the gene lists across cv replicates.</p>
<div class="sourceCode" id="cb189"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb189-1"><a href="brca-rnaseq-model-suite.html#cb189-1" aria-hidden="true"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb189-2"><a href="brca-rnaseq-model-suite.html#cb189-2" aria-hidden="true"></a></span>
<span id="cb189-3"><a href="brca-rnaseq-model-suite.html#cb189-3" aria-hidden="true"></a></span>
<span id="cb189-4"><a href="brca-rnaseq-model-suite.html#cb189-4" aria-hidden="true"></a><span class="co"># 1se</span></span>
<span id="cb189-5"><a href="brca-rnaseq-model-suite.html#cb189-5" aria-hidden="true"></a>lassoAll_coef_1se_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(brcaRna_cv_lassoAll_lst, <span class="cf">function</span>(cv_fit){</span>
<span id="cb189-6"><a href="brca-rnaseq-model-suite.html#cb189-6" aria-hidden="true"></a> cv_fit_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb189-7"><a href="brca-rnaseq-model-suite.html#cb189-7" aria-hidden="true"></a> cv_fit,</span>
<span id="cb189-8"><a href="brca-rnaseq-model-suite.html#cb189-8" aria-hidden="true"></a> <span class="dt">s =</span> <span class="st">&quot;lambda.1se&quot;</span></span>
<span id="cb189-9"><a href="brca-rnaseq-model-suite.html#cb189-9" aria-hidden="true"></a> )</span>
<span id="cb189-10"><a href="brca-rnaseq-model-suite.html#cb189-10" aria-hidden="true"></a> cv_fit_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][cv_fit_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>]]</span>
<span id="cb189-11"><a href="brca-rnaseq-model-suite.html#cb189-11" aria-hidden="true"></a> })</span>
<span id="cb189-12"><a href="brca-rnaseq-model-suite.html#cb189-12" aria-hidden="true"></a></span>
<span id="cb189-13"><a href="brca-rnaseq-model-suite.html#cb189-13" aria-hidden="true"></a><span class="co"># put into matrix</span></span>
<span id="cb189-14"><a href="brca-rnaseq-model-suite.html#cb189-14" aria-hidden="true"></a>lassoAll_coef_1se_all &lt;-<span class="st"> </span><span class="kw">Reduce</span>(union, lassoAll_coef_1se_lst)</span>
<span id="cb189-15"><a href="brca-rnaseq-model-suite.html#cb189-15" aria-hidden="true"></a>lassoAll_coef_1se_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(lassoAll_coef_1se_lst, </span>
<span id="cb189-16"><a href="brca-rnaseq-model-suite.html#cb189-16" aria-hidden="true"></a>  <span class="cf">function</span>(LL) <span class="kw">is.element</span>(lassoAll_coef_1se_all, LL)</span>
<span id="cb189-17"><a href="brca-rnaseq-model-suite.html#cb189-17" aria-hidden="true"></a>)</span>
<span id="cb189-18"><a href="brca-rnaseq-model-suite.html#cb189-18" aria-hidden="true"></a><span class="kw">rownames</span>(lassoAll_coef_1se_mtx) &lt;-<span class="st"> </span>lassoAll_coef_1se_all</span>
<span id="cb189-19"><a href="brca-rnaseq-model-suite.html#cb189-19" aria-hidden="true"></a></span>
<span id="cb189-20"><a href="brca-rnaseq-model-suite.html#cb189-20" aria-hidden="true"></a>genes_by_rep_1se_tbl &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="kw">rowSums</span>(lassoAll_coef_1se_mtx))</span>
<span id="cb189-21"><a href="brca-rnaseq-model-suite.html#cb189-21" aria-hidden="true"></a><span class="kw">barplot</span>(</span>
<span id="cb189-22"><a href="brca-rnaseq-model-suite.html#cb189-22" aria-hidden="true"></a> genes_by_rep_1se_tbl,</span>
<span id="cb189-23"><a href="brca-rnaseq-model-suite.html#cb189-23" aria-hidden="true"></a> <span class="dt">xlab=</span><span class="st">&#39;Number of Replicates&#39;</span>,</span>
<span id="cb189-24"><a href="brca-rnaseq-model-suite.html#cb189-24" aria-hidden="true"></a> <span class="dt">ylab=</span><span class="st">&#39;Number of features&#39;</span></span>
<span id="cb189-25"><a href="brca-rnaseq-model-suite.html#cb189-25" aria-hidden="true"></a></span>
<span id="cb189-26"><a href="brca-rnaseq-model-suite.html#cb189-26" aria-hidden="true"></a>)</span></code></pre></div>
<div class="figure"><span id="fig:brca-rnaseq-feature-list-1se"></span>
<img src="Static/figures/brca-rnaseq-feature-list-1se-1.png" alt="Feature list stability for one standard error rule models" width="768" />
<p class="caption">
Figure 8.5: Feature list stability for one standard error rule models
</p>
</div>
<p>We see that 83 features are included in every
cv replicate. These make up between
51%
and
70%
(Q1 and Q3) of the cv replicate one standard error rule models feature lists.</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="brca-rnaseq-model-suite.html#cb190-1" aria-hidden="true"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb190-2"><a href="brca-rnaseq-model-suite.html#cb190-2" aria-hidden="true"></a></span>
<span id="cb190-3"><a href="brca-rnaseq-model-suite.html#cb190-3" aria-hidden="true"></a></span>
<span id="cb190-4"><a href="brca-rnaseq-model-suite.html#cb190-4" aria-hidden="true"></a><span class="co"># min</span></span>
<span id="cb190-5"><a href="brca-rnaseq-model-suite.html#cb190-5" aria-hidden="true"></a>lassoAll_coef_min_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(brcaRna_cv_lassoAll_lst, <span class="cf">function</span>(cv_fit){</span>
<span id="cb190-6"><a href="brca-rnaseq-model-suite.html#cb190-6" aria-hidden="true"></a> cv_fit_coef &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb190-7"><a href="brca-rnaseq-model-suite.html#cb190-7" aria-hidden="true"></a> cv_fit,</span>
<span id="cb190-8"><a href="brca-rnaseq-model-suite.html#cb190-8" aria-hidden="true"></a> <span class="dt">s =</span> <span class="st">&quot;lambda.min&quot;</span></span>
<span id="cb190-9"><a href="brca-rnaseq-model-suite.html#cb190-9" aria-hidden="true"></a> )</span>
<span id="cb190-10"><a href="brca-rnaseq-model-suite.html#cb190-10" aria-hidden="true"></a> cv_fit_coef<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][cv_fit_coef<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>]]</span>
<span id="cb190-11"><a href="brca-rnaseq-model-suite.html#cb190-11" aria-hidden="true"></a> })</span>
<span id="cb190-12"><a href="brca-rnaseq-model-suite.html#cb190-12" aria-hidden="true"></a></span>
<span id="cb190-13"><a href="brca-rnaseq-model-suite.html#cb190-13" aria-hidden="true"></a><span class="co"># put into matrix</span></span>
<span id="cb190-14"><a href="brca-rnaseq-model-suite.html#cb190-14" aria-hidden="true"></a>lassoAll_coef_min_all &lt;-<span class="st"> </span><span class="kw">Reduce</span>(union, lassoAll_coef_min_lst)</span>
<span id="cb190-15"><a href="brca-rnaseq-model-suite.html#cb190-15" aria-hidden="true"></a>lassoAll_coef_min_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(lassoAll_coef_min_lst, </span>
<span id="cb190-16"><a href="brca-rnaseq-model-suite.html#cb190-16" aria-hidden="true"></a>  <span class="cf">function</span>(LL) <span class="kw">is.element</span>(lassoAll_coef_min_all, LL)</span>
<span id="cb190-17"><a href="brca-rnaseq-model-suite.html#cb190-17" aria-hidden="true"></a>)</span>
<span id="cb190-18"><a href="brca-rnaseq-model-suite.html#cb190-18" aria-hidden="true"></a><span class="kw">rownames</span>(lassoAll_coef_min_mtx) &lt;-<span class="st"> </span>lassoAll_coef_min_all</span>
<span id="cb190-19"><a href="brca-rnaseq-model-suite.html#cb190-19" aria-hidden="true"></a></span>
<span id="cb190-20"><a href="brca-rnaseq-model-suite.html#cb190-20" aria-hidden="true"></a>genes_by_rep_min_tbl &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="kw">rowSums</span>(lassoAll_coef_min_mtx))</span>
<span id="cb190-21"><a href="brca-rnaseq-model-suite.html#cb190-21" aria-hidden="true"></a><span class="kw">barplot</span>(</span>
<span id="cb190-22"><a href="brca-rnaseq-model-suite.html#cb190-22" aria-hidden="true"></a> genes_by_rep_min_tbl,</span>
<span id="cb190-23"><a href="brca-rnaseq-model-suite.html#cb190-23" aria-hidden="true"></a> <span class="dt">xlab=</span><span class="st">&#39;Number of Replicates&#39;</span>,</span>
<span id="cb190-24"><a href="brca-rnaseq-model-suite.html#cb190-24" aria-hidden="true"></a> <span class="dt">ylab=</span><span class="st">&#39;Number of features&#39;</span></span>
<span id="cb190-25"><a href="brca-rnaseq-model-suite.html#cb190-25" aria-hidden="true"></a></span>
<span id="cb190-26"><a href="brca-rnaseq-model-suite.html#cb190-26" aria-hidden="true"></a>)</span></code></pre></div>
<div class="figure"><span id="fig:brca-rnaseq-feature-list-min"></span>
<img src="Static/figures/brca-rnaseq-feature-list-min-1.png" alt="Feature list stability for minimum lambda models" width="768" />
<p class="caption">
Figure 8.6: Feature list stability for minimum lambda models
</p>
</div>
<p>We see that 110 features are included in every
cv replicate. These make up between
38%
and
50%
(Q1 and Q3) of the cv replicate min feature lists.
We will consider the genes that are selected in all cv replicates as a
gene signature produced by each model.</p>
<div class="sourceCode" id="cb191"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb191-1"><a href="brca-rnaseq-model-suite.html#cb191-1" aria-hidden="true"></a>lasso_gene_sign_1se_vec &lt;-<span class="st"> </span><span class="kw">rownames</span>(lassoAll_coef_1se_mtx)[<span class="kw">rowSums</span>(lassoAll_coef_1se_mtx)<span class="op">==</span><span class="dv">30</span>]</span>
<span id="cb191-2"><a href="brca-rnaseq-model-suite.html#cb191-2" aria-hidden="true"></a>lasso_gene_sign_min_vec &lt;-<span class="st"> </span><span class="kw">rownames</span>(lassoAll_coef_min_mtx)[<span class="kw">rowSums</span>(lassoAll_coef_min_mtx)<span class="op">==</span><span class="dv">30</span>]</span></code></pre></div>
<p>68 out of
83 of the genes in the 1se model gene signature
are contained in the min lambda model gene signature.</p>
</div>
<div id="simulation-design-1" class="section level2" number="8.3">
<h2><span class="header-section-number">8.3</span> Simulation Design</h2>
<p>We are now ready to run the simulations.</p>
<div class="sourceCode" id="cb192"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb192-1"><a href="brca-rnaseq-model-suite.html#cb192-1" aria-hidden="true"></a> SIM &lt;-<span class="st"> </span><span class="dv">30</span></span>
<span id="cb192-2"><a href="brca-rnaseq-model-suite.html#cb192-2" aria-hidden="true"></a> SIZE &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">25</span>, <span class="dv">50</span>, <span class="dv">100</span>, <span class="dv">200</span>, <span class="dv">300</span>)</span>
<span id="cb192-3"><a href="brca-rnaseq-model-suite.html#cb192-3" aria-hidden="true"></a> CV_REP &lt;-<span class="st"> </span><span class="dv">30</span></span></code></pre></div>
<p>Simluation parameters:</p>
<ul>
<li><p>Number of simulations : SIM = 30</p></li>
<li><p>Sample sizes: SIZE = 25, 50, 100, 200, 300</p></li>
<li><p>Number of CV Replicates: CV_REP = 30</p></li>
</ul>
<p>We will repeat the simulation process SIM = 30 times.
For each simulation iteration, we will select 300 Other and
300 LumA samples at random. Models will be fitted and analyzed
to balanced subsets of SIZE = 25, 50, 100, 200, 300, in a <code>Matryoshka doll</code> manner to
emulate a typical sample accrual process.</p>
<p>For a given simulation and a given sample size, we will obtain
CV_REP = 30 cross-validated lasso fits. From these fits,
we can obtain 30 out-of-fold assessments of classification accuracy
to get a sense if its variability. From each cv replicate, we also obtain
an estimated model size and a set of selected features. We will want
to examine how these stabilize as the sample size increases.</p>
<p>Note that we limit the simulations to a maximum of sample size of 300 in
order to to have simulations with low overlap. With 300
randomly selected LumA samples, the expected overlap between two randomly
selected sets of LumA samples is 4%.
For Others the expected overlap is 10.3%.</p>
</div>
<div id="setup-simulation-1" class="section level2" number="8.4">
<h2><span class="header-section-number">8.4</span> Setup simulation</h2>
<p>To setup the simulation, we only need two master tables: one for the selection of Others
and one for the selection of LumA samples.</p>
<div class="sourceCode" id="cb193"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb193-1"><a href="brca-rnaseq-model-suite.html#cb193-1" aria-hidden="true"></a>brcaRna_all_other_vec &lt;-<span class="st"> </span><span class="kw">names</span>(brcaRna_all_group_vec[brcaRna_all_group_vec<span class="op">==</span><span class="st">&#39;Other&#39;</span>]) </span>
<span id="cb193-2"><a href="brca-rnaseq-model-suite.html#cb193-2" aria-hidden="true"></a>brcaRna_all_LumA_vec &lt;-<span class="st"> </span><span class="kw">names</span>(brcaRna_all_group_vec[brcaRna_all_group_vec<span class="op">==</span><span class="st">&#39;LumA&#39;</span>])  </span></code></pre></div>
<p>We have 933 other sample IDs stored in <code>brcaRna_all_other_vec</code>
and 1492 LumA sample IDs stored in <code>brcaRna_all_LumA_vec</code>.
To create a suite of random samples from these, we only need to randomly select indices from
each vector.</p>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb194-1"><a href="brca-rnaseq-model-suite.html#cb194-1" aria-hidden="true"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb194-2"><a href="brca-rnaseq-model-suite.html#cb194-2" aria-hidden="true"></a></span>
<span id="cb194-3"><a href="brca-rnaseq-model-suite.html#cb194-3" aria-hidden="true"></a><span class="kw">set.seed</span>(<span class="dv">12379</span>)</span>
<span id="cb194-4"><a href="brca-rnaseq-model-suite.html#cb194-4" aria-hidden="true"></a></span>
<span id="cb194-5"><a href="brca-rnaseq-model-suite.html#cb194-5" aria-hidden="true"></a>brcaRna_sim_other_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb194-6"><a href="brca-rnaseq-model-suite.html#cb194-6" aria-hidden="true"></a> <span class="dv">1</span><span class="op">:</span>SIM, </span>
<span id="cb194-7"><a href="brca-rnaseq-model-suite.html#cb194-7" aria-hidden="true"></a> <span class="cf">function</span>(dummy) </span>
<span id="cb194-8"><a href="brca-rnaseq-model-suite.html#cb194-8" aria-hidden="true"></a>   <span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(brcaRna_all_other_vec), <span class="dt">size =</span>  <span class="kw">max</span>(SIZE))</span>
<span id="cb194-9"><a href="brca-rnaseq-model-suite.html#cb194-9" aria-hidden="true"></a>)</span>
<span id="cb194-10"><a href="brca-rnaseq-model-suite.html#cb194-10" aria-hidden="true"></a></span>
<span id="cb194-11"><a href="brca-rnaseq-model-suite.html#cb194-11" aria-hidden="true"></a></span>
<span id="cb194-12"><a href="brca-rnaseq-model-suite.html#cb194-12" aria-hidden="true"></a>brcaRna_sim_LumA_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb194-13"><a href="brca-rnaseq-model-suite.html#cb194-13" aria-hidden="true"></a> <span class="dv">1</span><span class="op">:</span>SIM, </span>
<span id="cb194-14"><a href="brca-rnaseq-model-suite.html#cb194-14" aria-hidden="true"></a> <span class="cf">function</span>(dummy) </span>
<span id="cb194-15"><a href="brca-rnaseq-model-suite.html#cb194-15" aria-hidden="true"></a>   <span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(brcaRna_all_LumA_vec), <span class="dt">size =</span>  <span class="kw">max</span>(SIZE))</span>
<span id="cb194-16"><a href="brca-rnaseq-model-suite.html#cb194-16" aria-hidden="true"></a>)</span></code></pre></div>
<p>Each simulation is specified by a given column of the simulation design matrices:
<code>brcaRna_sim_other_mtx</code> and <code>brcaRna_sim_LumA_mtx</code>, each with domensions 300, 30.
Within each simulation, we can run the analyses of size 25, 50, 100, 200, 300 by simply selecting
samples specified in the appropriate rows of each design matrix.</p>
<p>We can examine how much variability we have in the consistency scores of the selected samples.
Here we show results for the smalle sample sizes where variability will be the greatest.</p>
<div class="sourceCode" id="cb195"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb195-1"><a href="brca-rnaseq-model-suite.html#cb195-1" aria-hidden="true"></a>brcaRna_all_other_qual_vec &lt;-<span class="st"> </span>brcaRna_sample_qual_vec[brcaRna_all_other_vec]</span>
<span id="cb195-2"><a href="brca-rnaseq-model-suite.html#cb195-2" aria-hidden="true"></a>brcaRna_sim_other_qual_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb195-3"><a href="brca-rnaseq-model-suite.html#cb195-3" aria-hidden="true"></a>  <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(brcaRna_sim_other_mtx), </span>
<span id="cb195-4"><a href="brca-rnaseq-model-suite.html#cb195-4" aria-hidden="true"></a>  <span class="cf">function</span>(CC) brcaRna_all_other_qual_vec[brcaRna_sim_other_mtx[,CC]]</span>
<span id="cb195-5"><a href="brca-rnaseq-model-suite.html#cb195-5" aria-hidden="true"></a> )</span>
<span id="cb195-6"><a href="brca-rnaseq-model-suite.html#cb195-6" aria-hidden="true"></a></span>
<span id="cb195-7"><a href="brca-rnaseq-model-suite.html#cb195-7" aria-hidden="true"></a>brcaRna_all_LumA_qual_vec &lt;-<span class="st"> </span>brcaRna_sample_qual_vec[brcaRna_all_LumA_vec]</span>
<span id="cb195-8"><a href="brca-rnaseq-model-suite.html#cb195-8" aria-hidden="true"></a>brcaRna_sim_LumA_qual_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb195-9"><a href="brca-rnaseq-model-suite.html#cb195-9" aria-hidden="true"></a>  <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(brcaRna_sim_LumA_mtx),  </span>
<span id="cb195-10"><a href="brca-rnaseq-model-suite.html#cb195-10" aria-hidden="true"></a>  <span class="cf">function</span>(CC) brcaRna_all_LumA_qual_vec[brcaRna_sim_LumA_mtx[,CC]]</span>
<span id="cb195-11"><a href="brca-rnaseq-model-suite.html#cb195-11" aria-hidden="true"></a> )</span>
<span id="cb195-12"><a href="brca-rnaseq-model-suite.html#cb195-12" aria-hidden="true"></a></span>
<span id="cb195-13"><a href="brca-rnaseq-model-suite.html#cb195-13" aria-hidden="true"></a><span class="co"># ONLY LOOK AT SAMPLE SIZE == 50</span></span>
<span id="cb195-14"><a href="brca-rnaseq-model-suite.html#cb195-14" aria-hidden="true"></a>NN &lt;-<span class="st"> </span><span class="dv">50</span></span>
<span id="cb195-15"><a href="brca-rnaseq-model-suite.html#cb195-15" aria-hidden="true"></a></span>
<span id="cb195-16"><a href="brca-rnaseq-model-suite.html#cb195-16" aria-hidden="true"></a><span class="co"># PLOT</span></span>
<span id="cb195-17"><a href="brca-rnaseq-model-suite.html#cb195-17" aria-hidden="true"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">1</span>), <span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">5</span>,<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb195-18"><a href="brca-rnaseq-model-suite.html#cb195-18" aria-hidden="true"></a><span class="co"># other</span></span>
<span id="cb195-19"><a href="brca-rnaseq-model-suite.html#cb195-19" aria-hidden="true"></a><span class="kw">boxplot</span>(</span>
<span id="cb195-20"><a href="brca-rnaseq-model-suite.html#cb195-20" aria-hidden="true"></a>  brcaRna_sim_other_qual_mtx[<span class="dv">1</span><span class="op">:</span>NN,],</span>
<span id="cb195-21"><a href="brca-rnaseq-model-suite.html#cb195-21" aria-hidden="true"></a>  <span class="dt">outline =</span> T, </span>
<span id="cb195-22"><a href="brca-rnaseq-model-suite.html#cb195-22" aria-hidden="true"></a>  <span class="dt">ylab =</span> <span class="st">&#39;Consistency Score&#39;</span>,</span>
<span id="cb195-23"><a href="brca-rnaseq-model-suite.html#cb195-23" aria-hidden="true"></a>  <span class="dt">xaxt =</span> <span class="st">&#39;n&#39;</span></span>
<span id="cb195-24"><a href="brca-rnaseq-model-suite.html#cb195-24" aria-hidden="true"></a>)</span>
<span id="cb195-25"><a href="brca-rnaseq-model-suite.html#cb195-25" aria-hidden="true"></a><span class="kw">title</span>(<span class="st">&quot;Other sample consistency across simulations&quot;</span>)</span>
<span id="cb195-26"><a href="brca-rnaseq-model-suite.html#cb195-26" aria-hidden="true"></a></span>
<span id="cb195-27"><a href="brca-rnaseq-model-suite.html#cb195-27" aria-hidden="true"></a><span class="co"># LumA</span></span>
<span id="cb195-28"><a href="brca-rnaseq-model-suite.html#cb195-28" aria-hidden="true"></a><span class="kw">boxplot</span>(</span>
<span id="cb195-29"><a href="brca-rnaseq-model-suite.html#cb195-29" aria-hidden="true"></a>  brcaRna_sim_LumA_qual_mtx[<span class="dv">1</span><span class="op">:</span>NN,],</span>
<span id="cb195-30"><a href="brca-rnaseq-model-suite.html#cb195-30" aria-hidden="true"></a>  <span class="dt">outline =</span> T, </span>
<span id="cb195-31"><a href="brca-rnaseq-model-suite.html#cb195-31" aria-hidden="true"></a>  <span class="dt">ylab =</span> <span class="st">&#39;Consistency Score&#39;</span></span>
<span id="cb195-32"><a href="brca-rnaseq-model-suite.html#cb195-32" aria-hidden="true"></a>)</span>
<span id="cb195-33"><a href="brca-rnaseq-model-suite.html#cb195-33" aria-hidden="true"></a><span class="kw">title</span>(<span class="st">&quot;LumA sample consistency across simulations&quot;</span>)</span></code></pre></div>
<div class="figure"><span id="fig:brca-rnaseq-look-sim-qual-0-50ONLY"></span>
<img src="Static/figures/brca-rnaseq-look-sim-qual-0-50ONLY-1.png" alt="sample consistency by simulation run for size = 50 " width="960" />
<p class="caption">
Figure 8.7: sample consistency by simulation run for size = 50
</p>
</div>
<p>In this figure, we are summarizing the consistency measures of 50 samples per group acress
30 simulations, or random selections of other and LumA samples.
We see significant variability in sample consistency, especially in the Other cases.
This may lead an unwary observer to be overly optimistic, or overly pessimistic,
in the early accrual stages of a study.</p>
</div>
<div id="run-simulations-1" class="section level2" number="8.5">
<h2><span class="header-section-number">8.5</span> Run simulations</h2>
<p>As these take a while to run,
we will save the results of each similation to a different
object and store to disk. These can be easily read from disk
when needed for analysis.</p>
<p>The simulation results are saved to the file system and
only needs to be run once. The simulation takes <span class="math inline">\(\approx\)</span> 14 minutes
per iteration, or <span class="math inline">\(\approx\)</span> 7 hours of run time on a laptop.
(Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Mojave 10.14.6)</p>
<!-- RUN ONCE - THEN GET FROM MEMORY -->
<div class="sourceCode" id="cb196"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb196-1"><a href="brca-rnaseq-model-suite.html#cb196-1" aria-hidden="true"></a>start_time &lt;-<span class="st"> </span><span class="kw">proc.time</span>()</span>
<span id="cb196-2"><a href="brca-rnaseq-model-suite.html#cb196-2" aria-hidden="true"></a></span>
<span id="cb196-3"><a href="brca-rnaseq-model-suite.html#cb196-3" aria-hidden="true"></a><span class="co"># Get stage from SIZE</span></span>
<span id="cb196-4"><a href="brca-rnaseq-model-suite.html#cb196-4" aria-hidden="true"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(brcaRna_sim_other_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>, SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb196-5"><a href="brca-rnaseq-model-suite.html#cb196-5" aria-hidden="true"></a></span>
<span id="cb196-6"><a href="brca-rnaseq-model-suite.html#cb196-6" aria-hidden="true"></a><span class="cf">for</span> (SIMno <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(brcaRna_sim_other_qual_mtx)) {</span>
<span id="cb196-7"><a href="brca-rnaseq-model-suite.html#cb196-7" aria-hidden="true"></a></span>
<span id="cb196-8"><a href="brca-rnaseq-model-suite.html#cb196-8" aria-hidden="true"></a>  <span class="co">#cat(&quot;Running simulation &quot;, SIMno, &quot;\n&quot;)</span></span>
<span id="cb196-9"><a href="brca-rnaseq-model-suite.html#cb196-9" aria-hidden="true"></a></span>
<span id="cb196-10"><a href="brca-rnaseq-model-suite.html#cb196-10" aria-hidden="true"></a>  brca_sim_cv_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(<span class="kw">levels</span>(stage_vec)), <span class="cf">function</span>(STGno) {</span>
<span id="cb196-11"><a href="brca-rnaseq-model-suite.html#cb196-11" aria-hidden="true"></a>    Stage_rows_vec &lt;-<span class="st"> </span><span class="kw">which</span>(stage_vec <span class="op">%in%</span><span class="st"> </span><span class="kw">levels</span>(stage_vec)[<span class="dv">1</span><span class="op">:</span>STGno])</span>
<span id="cb196-12"><a href="brca-rnaseq-model-suite.html#cb196-12" aria-hidden="true"></a>    <span class="co">#cat(&quot;Stage &quot;, STGno, &quot;- analyzing&quot;, length(Stage_rows_vec), &quot;paired samples.\n&quot;)</span></span>
<span id="cb196-13"><a href="brca-rnaseq-model-suite.html#cb196-13" aria-hidden="true"></a></span>
<span id="cb196-14"><a href="brca-rnaseq-model-suite.html#cb196-14" aria-hidden="true"></a>    sim_stage_samples_vec &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb196-15"><a href="brca-rnaseq-model-suite.html#cb196-15" aria-hidden="true"></a>      brcaRna_all_other_vec[brcaRna_sim_other_mtx[Stage_rows_vec, SIMno]],</span>
<span id="cb196-16"><a href="brca-rnaseq-model-suite.html#cb196-16" aria-hidden="true"></a>      brcaRna_all_LumA_vec[brcaRna_sim_LumA_mtx[Stage_rows_vec, SIMno]]</span>
<span id="cb196-17"><a href="brca-rnaseq-model-suite.html#cb196-17" aria-hidden="true"></a>    )</span>
<span id="cb196-18"><a href="brca-rnaseq-model-suite.html#cb196-18" aria-hidden="true"></a>    sim_stage_geneExpr_mtx &lt;-<span class="st"> </span>brcaRna_all_geneExpr_mtx[sim_stage_samples_vec, ]</span>
<span id="cb196-19"><a href="brca-rnaseq-model-suite.html#cb196-19" aria-hidden="true"></a>    sim_stage_group_vec &lt;-<span class="st"> </span>brcaRna_all_group_vec[sim_stage_samples_vec]</span>
<span id="cb196-20"><a href="brca-rnaseq-model-suite.html#cb196-20" aria-hidden="true"></a>    <span class="co">#print(table(sim_stage_group_vec))</span></span>
<span id="cb196-21"><a href="brca-rnaseq-model-suite.html#cb196-21" aria-hidden="true"></a></span>
<span id="cb196-22"><a href="brca-rnaseq-model-suite.html#cb196-22" aria-hidden="true"></a>    sim_stage_cv_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span>CV_REP, <span class="cf">function</span>(CV) {</span>
<span id="cb196-23"><a href="brca-rnaseq-model-suite.html#cb196-23" aria-hidden="true"></a>      cv_fit &lt;-<span class="st"> </span>glmnet<span class="op">::</span><span class="kw">cv.glmnet</span>(</span>
<span id="cb196-24"><a href="brca-rnaseq-model-suite.html#cb196-24" aria-hidden="true"></a>        <span class="dt">x =</span> sim_stage_geneExpr_mtx,</span>
<span id="cb196-25"><a href="brca-rnaseq-model-suite.html#cb196-25" aria-hidden="true"></a>        <span class="dt">y =</span> sim_stage_group_vec,</span>
<span id="cb196-26"><a href="brca-rnaseq-model-suite.html#cb196-26" aria-hidden="true"></a>        <span class="dt">alpha =</span> <span class="dv">1</span>,</span>
<span id="cb196-27"><a href="brca-rnaseq-model-suite.html#cb196-27" aria-hidden="true"></a>        <span class="dt">family =</span> <span class="st">&quot;binomial&quot;</span>,</span>
<span id="cb196-28"><a href="brca-rnaseq-model-suite.html#cb196-28" aria-hidden="true"></a>        <span class="dt">type.measure =</span> <span class="st">&quot;class&quot;</span>,</span>
<span id="cb196-29"><a href="brca-rnaseq-model-suite.html#cb196-29" aria-hidden="true"></a>        <span class="dt">keep =</span> T,</span>
<span id="cb196-30"><a href="brca-rnaseq-model-suite.html#cb196-30" aria-hidden="true"></a>        <span class="dt">nlambda =</span> <span class="dv">30</span></span>
<span id="cb196-31"><a href="brca-rnaseq-model-suite.html#cb196-31" aria-hidden="true"></a>      )</span>
<span id="cb196-32"><a href="brca-rnaseq-model-suite.html#cb196-32" aria-hidden="true"></a></span>
<span id="cb196-33"><a href="brca-rnaseq-model-suite.html#cb196-33" aria-hidden="true"></a>      <span class="co"># Extract 1se metrics from cv_fit</span></span>
<span id="cb196-34"><a href="brca-rnaseq-model-suite.html#cb196-34" aria-hidden="true"></a>      <span class="co">#######################</span></span>
<span id="cb196-35"><a href="brca-rnaseq-model-suite.html#cb196-35" aria-hidden="true"></a>      ndx_1se &lt;-<span class="st"> </span><span class="kw">which</span>(cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se)</span>
<span id="cb196-36"><a href="brca-rnaseq-model-suite.html#cb196-36" aria-hidden="true"></a></span>
<span id="cb196-37"><a href="brca-rnaseq-model-suite.html#cb196-37" aria-hidden="true"></a>      nzero_1se &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>nzero[ndx_1se]</span>
<span id="cb196-38"><a href="brca-rnaseq-model-suite.html#cb196-38" aria-hidden="true"></a>      cvm_1se &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>cvm[ndx_1se]</span>
<span id="cb196-39"><a href="brca-rnaseq-model-suite.html#cb196-39" aria-hidden="true"></a></span>
<span id="cb196-40"><a href="brca-rnaseq-model-suite.html#cb196-40" aria-hidden="true"></a>      <span class="co"># test error</span></span>
<span id="cb196-41"><a href="brca-rnaseq-model-suite.html#cb196-41" aria-hidden="true"></a>      sim_stage_test_samples_vec &lt;-<span class="st"> </span><span class="kw">setdiff</span>(<span class="kw">rownames</span>(brcaRna_all_geneExpr_mtx), sim_stage_samples_vec)</span>
<span id="cb196-42"><a href="brca-rnaseq-model-suite.html#cb196-42" aria-hidden="true"></a>      sim_stage_brcaRna_test_geneExpr_mtx &lt;-<span class="st"> </span>brcaRna_all_geneExpr_mtx[sim_stage_test_samples_vec,]</span>
<span id="cb196-43"><a href="brca-rnaseq-model-suite.html#cb196-43" aria-hidden="true"></a>      sim_stage_brcaRna_test_group_vec &lt;-<span class="st"> </span>brcaRna_all_group_vec[sim_stage_test_samples_vec]</span>
<span id="cb196-44"><a href="brca-rnaseq-model-suite.html#cb196-44" aria-hidden="true"></a></span>
<span id="cb196-45"><a href="brca-rnaseq-model-suite.html#cb196-45" aria-hidden="true"></a>      test_pred_1se_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb196-46"><a href="brca-rnaseq-model-suite.html#cb196-46" aria-hidden="true"></a>       cv_fit,</span>
<span id="cb196-47"><a href="brca-rnaseq-model-suite.html#cb196-47" aria-hidden="true"></a>       <span class="dt">newx=</span>sim_stage_brcaRna_test_geneExpr_mtx,</span>
<span id="cb196-48"><a href="brca-rnaseq-model-suite.html#cb196-48" aria-hidden="true"></a>       <span class="dt">s=</span><span class="st">&quot;lambda.1se&quot;</span>,</span>
<span id="cb196-49"><a href="brca-rnaseq-model-suite.html#cb196-49" aria-hidden="true"></a>       <span class="dt">type=</span><span class="st">&quot;class&quot;</span></span>
<span id="cb196-50"><a href="brca-rnaseq-model-suite.html#cb196-50" aria-hidden="true"></a>      )</span>
<span id="cb196-51"><a href="brca-rnaseq-model-suite.html#cb196-51" aria-hidden="true"></a>      test_1se_error &lt;-<span class="st"> </span><span class="kw">mean</span>(test_pred_1se_vec <span class="op">!=</span><span class="st"> </span>sim_stage_brcaRna_test_group_vec)</span>
<span id="cb196-52"><a href="brca-rnaseq-model-suite.html#cb196-52" aria-hidden="true"></a></span>
<span id="cb196-53"><a href="brca-rnaseq-model-suite.html#cb196-53" aria-hidden="true"></a>      <span class="co"># genes</span></span>
<span id="cb196-54"><a href="brca-rnaseq-model-suite.html#cb196-54" aria-hidden="true"></a>      coef_1se &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb196-55"><a href="brca-rnaseq-model-suite.html#cb196-55" aria-hidden="true"></a>        cv_fit,</span>
<span id="cb196-56"><a href="brca-rnaseq-model-suite.html#cb196-56" aria-hidden="true"></a>        <span class="dt">s =</span> <span class="st">&quot;lambda.1se&quot;</span></span>
<span id="cb196-57"><a href="brca-rnaseq-model-suite.html#cb196-57" aria-hidden="true"></a>      )</span>
<span id="cb196-58"><a href="brca-rnaseq-model-suite.html#cb196-58" aria-hidden="true"></a>      genes_1se &lt;-<span class="st"> </span>coef_1se<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][coef_1se<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>]]</span>
<span id="cb196-59"><a href="brca-rnaseq-model-suite.html#cb196-59" aria-hidden="true"></a></span>
<span id="cb196-60"><a href="brca-rnaseq-model-suite.html#cb196-60" aria-hidden="true"></a>      <span class="co"># Extract min metrics from cv_fit</span></span>
<span id="cb196-61"><a href="brca-rnaseq-model-suite.html#cb196-61" aria-hidden="true"></a>      <span class="co">#######################</span></span>
<span id="cb196-62"><a href="brca-rnaseq-model-suite.html#cb196-62" aria-hidden="true"></a>      ndx_min &lt;-<span class="st"> </span><span class="kw">which</span>(cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min)</span>
<span id="cb196-63"><a href="brca-rnaseq-model-suite.html#cb196-63" aria-hidden="true"></a></span>
<span id="cb196-64"><a href="brca-rnaseq-model-suite.html#cb196-64" aria-hidden="true"></a>      nzero_min &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>nzero[ndx_min]</span>
<span id="cb196-65"><a href="brca-rnaseq-model-suite.html#cb196-65" aria-hidden="true"></a>      cvm_min &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>cvm[ndx_min]</span>
<span id="cb196-66"><a href="brca-rnaseq-model-suite.html#cb196-66" aria-hidden="true"></a></span>
<span id="cb196-67"><a href="brca-rnaseq-model-suite.html#cb196-67" aria-hidden="true"></a>      <span class="co"># test error</span></span>
<span id="cb196-68"><a href="brca-rnaseq-model-suite.html#cb196-68" aria-hidden="true"></a>      sim_stage_test_samples_vec &lt;-<span class="st"> </span><span class="kw">setdiff</span>(<span class="kw">rownames</span>(brcaRna_all_geneExpr_mtx), sim_stage_samples_vec)</span>
<span id="cb196-69"><a href="brca-rnaseq-model-suite.html#cb196-69" aria-hidden="true"></a>      sim_stage_brcaRna_test_geneExpr_mtx &lt;-<span class="st"> </span>brcaRna_all_geneExpr_mtx[sim_stage_test_samples_vec,]</span>
<span id="cb196-70"><a href="brca-rnaseq-model-suite.html#cb196-70" aria-hidden="true"></a>      sim_stage_brcaRna_test_group_vec &lt;-<span class="st"> </span>brcaRna_all_group_vec[sim_stage_test_samples_vec]</span>
<span id="cb196-71"><a href="brca-rnaseq-model-suite.html#cb196-71" aria-hidden="true"></a></span>
<span id="cb196-72"><a href="brca-rnaseq-model-suite.html#cb196-72" aria-hidden="true"></a>      test_pred_min_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb196-73"><a href="brca-rnaseq-model-suite.html#cb196-73" aria-hidden="true"></a>       cv_fit,</span>
<span id="cb196-74"><a href="brca-rnaseq-model-suite.html#cb196-74" aria-hidden="true"></a>       <span class="dt">newx=</span>sim_stage_brcaRna_test_geneExpr_mtx,</span>
<span id="cb196-75"><a href="brca-rnaseq-model-suite.html#cb196-75" aria-hidden="true"></a>       <span class="dt">s=</span><span class="st">&quot;lambda.min&quot;</span>,</span>
<span id="cb196-76"><a href="brca-rnaseq-model-suite.html#cb196-76" aria-hidden="true"></a>       <span class="dt">type=</span><span class="st">&quot;class&quot;</span></span>
<span id="cb196-77"><a href="brca-rnaseq-model-suite.html#cb196-77" aria-hidden="true"></a>      )</span>
<span id="cb196-78"><a href="brca-rnaseq-model-suite.html#cb196-78" aria-hidden="true"></a>      test_min_error &lt;-<span class="st"> </span><span class="kw">mean</span>(test_pred_min_vec <span class="op">!=</span><span class="st"> </span>sim_stage_brcaRna_test_group_vec)</span>
<span id="cb196-79"><a href="brca-rnaseq-model-suite.html#cb196-79" aria-hidden="true"></a></span>
<span id="cb196-80"><a href="brca-rnaseq-model-suite.html#cb196-80" aria-hidden="true"></a>      <span class="co"># genes</span></span>
<span id="cb196-81"><a href="brca-rnaseq-model-suite.html#cb196-81" aria-hidden="true"></a>      coef_min &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb196-82"><a href="brca-rnaseq-model-suite.html#cb196-82" aria-hidden="true"></a>        cv_fit,</span>
<span id="cb196-83"><a href="brca-rnaseq-model-suite.html#cb196-83" aria-hidden="true"></a>        <span class="dt">s =</span> <span class="st">&quot;lambda.min&quot;</span></span>
<span id="cb196-84"><a href="brca-rnaseq-model-suite.html#cb196-84" aria-hidden="true"></a>      )</span>
<span id="cb196-85"><a href="brca-rnaseq-model-suite.html#cb196-85" aria-hidden="true"></a>      genes_min &lt;-<span class="st"> </span>coef_min<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][coef_min<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>]]</span>
<span id="cb196-86"><a href="brca-rnaseq-model-suite.html#cb196-86" aria-hidden="true"></a></span>
<span id="cb196-87"><a href="brca-rnaseq-model-suite.html#cb196-87" aria-hidden="true"></a>      <span class="co"># return cv_fit summary metrics</span></span>
<span id="cb196-88"><a href="brca-rnaseq-model-suite.html#cb196-88" aria-hidden="true"></a>      <span class="kw">list</span>(</span>
<span id="cb196-89"><a href="brca-rnaseq-model-suite.html#cb196-89" aria-hidden="true"></a>       <span class="dt">p_1se =</span> nzero_1se, </span>
<span id="cb196-90"><a href="brca-rnaseq-model-suite.html#cb196-90" aria-hidden="true"></a>       <span class="dt">p_min =</span> nzero_min, </span>
<span id="cb196-91"><a href="brca-rnaseq-model-suite.html#cb196-91" aria-hidden="true"></a>       <span class="dt">cv_1se =</span> cvm_1se, </span>
<span id="cb196-92"><a href="brca-rnaseq-model-suite.html#cb196-92" aria-hidden="true"></a>       <span class="dt">cv_min =</span> cvm_min, </span>
<span id="cb196-93"><a href="brca-rnaseq-model-suite.html#cb196-93" aria-hidden="true"></a>       <span class="dt">test_1se=</span>test_1se_error, </span>
<span id="cb196-94"><a href="brca-rnaseq-model-suite.html#cb196-94" aria-hidden="true"></a>       <span class="dt">test_min=</span>test_min_error, </span>
<span id="cb196-95"><a href="brca-rnaseq-model-suite.html#cb196-95" aria-hidden="true"></a>       <span class="dt">genes_1se =</span> genes_1se,</span>
<span id="cb196-96"><a href="brca-rnaseq-model-suite.html#cb196-96" aria-hidden="true"></a>       <span class="dt">genes_min =</span> genes_min)</span>
<span id="cb196-97"><a href="brca-rnaseq-model-suite.html#cb196-97" aria-hidden="true"></a>    })</span>
<span id="cb196-98"><a href="brca-rnaseq-model-suite.html#cb196-98" aria-hidden="true"></a>    sim_stage_cv_lst</span>
<span id="cb196-99"><a href="brca-rnaseq-model-suite.html#cb196-99" aria-hidden="true"></a>  })</span>
<span id="cb196-100"><a href="brca-rnaseq-model-suite.html#cb196-100" aria-hidden="true"></a></span>
<span id="cb196-101"><a href="brca-rnaseq-model-suite.html#cb196-101" aria-hidden="true"></a>  <span class="co"># save  brcaRna_sim_cv_lst</span></span>
<span id="cb196-102"><a href="brca-rnaseq-model-suite.html#cb196-102" aria-hidden="true"></a>  fName &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;brcaRna_sim_&quot;</span>, SIMno, <span class="st">&quot;_cv_lst&quot;</span>)</span>
<span id="cb196-103"><a href="brca-rnaseq-model-suite.html#cb196-103" aria-hidden="true"></a>  <span class="kw">assign</span>(fName, brca_sim_cv_lst)</span>
<span id="cb196-104"><a href="brca-rnaseq-model-suite.html#cb196-104" aria-hidden="true"></a>  <span class="kw">save</span>(<span class="dt">list =</span> fName, <span class="dt">file=</span><span class="kw">file.path</span>(<span class="st">&quot;RData&quot;</span>, fName))</span>
<span id="cb196-105"><a href="brca-rnaseq-model-suite.html#cb196-105" aria-hidden="true"></a></span>
<span id="cb196-106"><a href="brca-rnaseq-model-suite.html#cb196-106" aria-hidden="true"></a>}</span></code></pre></div>
<!-- DEBUG - rename after the fact -->
</div>
<div id="simulation-results-1" class="section level2" number="8.6">
<h2><span class="header-section-number">8.6</span> Simulation results</h2>
<p>Recall the we have 30 simulations, or randomly selected sets of LumA and Other samples,
analyzed in inreasing sizes of 25, 50, 100, 200, 300, with
30 repeated cross-validated lasso fits:</p>
<ul>
<li><p>Sample sizes: SIZE = 25, 50, 100, 200, 300</p></li>
<li><p>Number of CV Replicates: CV_REP = 30</p></li>
</ul>
<p>First we extract simluation results and store into one big table (only showing the top of table shere):</p>
<div class="sourceCode" id="cb197"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb197-1"><a href="brca-rnaseq-model-suite.html#cb197-1" aria-hidden="true"></a>brcaRna_sim_files_vec &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="st">&#39;RData&#39;</span>, <span class="st">&#39;^brcaRna_sim_&#39;</span>)</span>
<span id="cb197-2"><a href="brca-rnaseq-model-suite.html#cb197-2" aria-hidden="true"></a></span>
<span id="cb197-3"><a href="brca-rnaseq-model-suite.html#cb197-3" aria-hidden="true"></a></span>
<span id="cb197-4"><a href="brca-rnaseq-model-suite.html#cb197-4" aria-hidden="true"></a><span class="co"># define extraction methods</span></span>
<span id="cb197-5"><a href="brca-rnaseq-model-suite.html#cb197-5" aria-hidden="true"></a></span>
<span id="cb197-6"><a href="brca-rnaseq-model-suite.html#cb197-6" aria-hidden="true"></a><span class="co"># Each sumulation is a list of cv results </span></span>
<span id="cb197-7"><a href="brca-rnaseq-model-suite.html#cb197-7" aria-hidden="true"></a><span class="co">## nested in a list of replicates</span></span>
<span id="cb197-8"><a href="brca-rnaseq-model-suite.html#cb197-8" aria-hidden="true"></a><span class="co">##############################################</span></span>
<span id="cb197-9"><a href="brca-rnaseq-model-suite.html#cb197-9" aria-hidden="true"></a></span>
<span id="cb197-10"><a href="brca-rnaseq-model-suite.html#cb197-10" aria-hidden="true"></a><span class="co"># cvList2frm_f makes a frame out of the inner list</span></span>
<span id="cb197-11"><a href="brca-rnaseq-model-suite.html#cb197-11" aria-hidden="true"></a>cvList2frm_f &lt;-<span class="st"> </span><span class="cf">function</span>(cv_lst) {</span>
<span id="cb197-12"><a href="brca-rnaseq-model-suite.html#cb197-12" aria-hidden="true"></a> frm1 &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">t</span>(<span class="kw">sapply</span>(cv_lst, <span class="cf">function</span>(x) x)))</span>
<span id="cb197-13"><a href="brca-rnaseq-model-suite.html#cb197-13" aria-hidden="true"></a> frm2 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb197-14"><a href="brca-rnaseq-model-suite.html#cb197-14" aria-hidden="true"></a>  <span class="kw">unlist</span>(frm1[[<span class="dv">1</span>]]), <span class="kw">unlist</span>(frm1[[<span class="dv">2</span>]]),</span>
<span id="cb197-15"><a href="brca-rnaseq-model-suite.html#cb197-15" aria-hidden="true"></a>  <span class="kw">unlist</span>(frm1[[<span class="dv">3</span>]]), <span class="kw">unlist</span>(frm1[[<span class="dv">4</span>]]),</span>
<span id="cb197-16"><a href="brca-rnaseq-model-suite.html#cb197-16" aria-hidden="true"></a>  <span class="kw">unlist</span>(frm1[[<span class="dv">5</span>]]), <span class="kw">unlist</span>(frm1[[<span class="dv">6</span>]]),</span>
<span id="cb197-17"><a href="brca-rnaseq-model-suite.html#cb197-17" aria-hidden="true"></a>  frm1[<span class="dv">7</span>], frm1[<span class="dv">8</span>])</span>
<span id="cb197-18"><a href="brca-rnaseq-model-suite.html#cb197-18" aria-hidden="true"></a>  <span class="kw">names</span>(frm2) &lt;-<span class="st"> </span><span class="kw">names</span>(frm1)</span>
<span id="cb197-19"><a href="brca-rnaseq-model-suite.html#cb197-19" aria-hidden="true"></a>  <span class="kw">data.frame</span>(<span class="dt">Rep=</span><span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(frm2), frm2)}</span>
<span id="cb197-20"><a href="brca-rnaseq-model-suite.html#cb197-20" aria-hidden="true"></a></span>
<span id="cb197-21"><a href="brca-rnaseq-model-suite.html#cb197-21" aria-hidden="true"></a><span class="co"># cv_lst_to_frm loop over replicates, concatenating the inner list frames</span></span>
<span id="cb197-22"><a href="brca-rnaseq-model-suite.html#cb197-22" aria-hidden="true"></a>cv_lst_to_frm &lt;-<span class="st"> </span><span class="cf">function</span>(brcaRna_sim_cv_lst) {</span>
<span id="cb197-23"><a href="brca-rnaseq-model-suite.html#cb197-23" aria-hidden="true"></a> <span class="kw">do.call</span>(<span class="st">&#39;rbind&#39;</span>, <span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(brcaRna_sim_cv_lst),</span>
<span id="cb197-24"><a href="brca-rnaseq-model-suite.html#cb197-24" aria-hidden="true"></a>  <span class="cf">function</span>(JJ) {</span>
<span id="cb197-25"><a href="brca-rnaseq-model-suite.html#cb197-25" aria-hidden="true"></a>    siz_frm &lt;-<span class="st"> </span><span class="kw">cvList2frm_f</span>(brcaRna_sim_cv_lst[[JJ]])</span>
<span id="cb197-26"><a href="brca-rnaseq-model-suite.html#cb197-26" aria-hidden="true"></a>    <span class="kw">data.frame</span>(<span class="dt">Size=</span>SIZE[JJ], siz_frm)</span>
<span id="cb197-27"><a href="brca-rnaseq-model-suite.html#cb197-27" aria-hidden="true"></a>  }))</span>
<span id="cb197-28"><a href="brca-rnaseq-model-suite.html#cb197-28" aria-hidden="true"></a>}</span>
<span id="cb197-29"><a href="brca-rnaseq-model-suite.html#cb197-29" aria-hidden="true"></a></span>
<span id="cb197-30"><a href="brca-rnaseq-model-suite.html#cb197-30" aria-hidden="true"></a><span class="co"># we loop across simulations to combine all results into one big table</span></span>
<span id="cb197-31"><a href="brca-rnaseq-model-suite.html#cb197-31" aria-hidden="true"></a>brcaRna_lasso_sim_results_frm &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="st">&#39;rbind&#39;</span>, <span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(brcaRna_sim_files_vec),</span>
<span id="cb197-32"><a href="brca-rnaseq-model-suite.html#cb197-32" aria-hidden="true"></a> <span class="cf">function</span>(SIM_NO) {</span>
<span id="cb197-33"><a href="brca-rnaseq-model-suite.html#cb197-33" aria-hidden="true"></a>  <span class="kw">load</span>(<span class="dt">file=</span><span class="kw">file.path</span>(<span class="st">&#39;RData&#39;</span>, brcaRna_sim_files_vec[SIM_NO]))</span>
<span id="cb197-34"><a href="brca-rnaseq-model-suite.html#cb197-34" aria-hidden="true"></a>  <span class="kw">assign</span>(<span class="st">&#39;brcaRna_sim_cv_lst&#39;</span>, <span class="kw">get</span>(brcaRna_sim_files_vec[SIM_NO]))</span>
<span id="cb197-35"><a href="brca-rnaseq-model-suite.html#cb197-35" aria-hidden="true"></a>  <span class="kw">rm</span>(<span class="dt">list=</span>brcaRna_sim_files_vec[SIM_NO])</span>
<span id="cb197-36"><a href="brca-rnaseq-model-suite.html#cb197-36" aria-hidden="true"></a>  </span>
<span id="cb197-37"><a href="brca-rnaseq-model-suite.html#cb197-37" aria-hidden="true"></a>  <span class="kw">data.frame</span>(<span class="dt">SimNo=</span><span class="kw">paste0</span>(<span class="st">&#39;Sim_&#39;</span>,<span class="kw">formatC</span>(SIM_NO,<span class="dt">width =</span> <span class="dv">2</span>,<span class="dt">flag =</span> <span class="dv">0</span>)), <span class="kw">cv_lst_to_frm</span>(brcaRna_sim_cv_lst))</span>
<span id="cb197-38"><a href="brca-rnaseq-model-suite.html#cb197-38" aria-hidden="true"></a>} </span>
<span id="cb197-39"><a href="brca-rnaseq-model-suite.html#cb197-39" aria-hidden="true"></a>)) </span></code></pre></div>
<!-- 
Have a table of simulation results - `brcaRna_lasso_sim_results_frm`:
-->
<div id="results-for-one-simulation-1" class="section level3" number="8.6.1">
<h3><span class="header-section-number">8.6.1</span> Results for one simulation</h3>
<div id="model-accuracy-assessment-2" class="section level4" number="8.6.1.1">
<h4><span class="header-section-number">8.6.1.1</span> Model Accuracy Assessment</h4>
<p>First examine results for one simulation run. In the figures that follow,
each boxplot summarized 30 repreated cross validation runs performed on a
fixed random selection of Other and Affeced samples. Recall that as
we move from 25 to 50, etc., the sample sets are growing to emulate an
accrual of samples over time.</p>
<div class="sourceCode" id="cb198"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb198-1"><a href="brca-rnaseq-model-suite.html#cb198-1" aria-hidden="true"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb198-2"><a href="brca-rnaseq-model-suite.html#cb198-2" aria-hidden="true"></a></span>
<span id="cb198-3"><a href="brca-rnaseq-model-suite.html#cb198-3" aria-hidden="true"></a><span class="co"># get full model cv error ref</span></span>
<span id="cb198-4"><a href="brca-rnaseq-model-suite.html#cb198-4" aria-hidden="true"></a>error_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb198-5"><a href="brca-rnaseq-model-suite.html#cb198-5" aria-hidden="true"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb198-6"><a href="brca-rnaseq-model-suite.html#cb198-6" aria-hidden="true"></a>error_1se_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(error_1se_vec, <span class="dt">prob=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>)        </span>
<span id="cb198-7"><a href="brca-rnaseq-model-suite.html#cb198-7" aria-hidden="true"></a></span>
<span id="cb198-8"><a href="brca-rnaseq-model-suite.html#cb198-8" aria-hidden="true"></a>error_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb198-9"><a href="brca-rnaseq-model-suite.html#cb198-9" aria-hidden="true"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb198-10"><a href="brca-rnaseq-model-suite.html#cb198-10" aria-hidden="true"></a>error_min_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(error_min_vec, <span class="dt">prob=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>)        </span>
<span id="cb198-11"><a href="brca-rnaseq-model-suite.html#cb198-11" aria-hidden="true"></a></span>
<span id="cb198-12"><a href="brca-rnaseq-model-suite.html#cb198-12" aria-hidden="true"></a><span class="co"># Utility objects</span></span>
<span id="cb198-13"><a href="brca-rnaseq-model-suite.html#cb198-13" aria-hidden="true"></a>SIZE0 &lt;-<span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_pad</span>(SIZE, <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>)</span>
<span id="cb198-14"><a href="brca-rnaseq-model-suite.html#cb198-14" aria-hidden="true"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(brcaRna_sim_other_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>,SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb198-15"><a href="brca-rnaseq-model-suite.html#cb198-15" aria-hidden="true"></a></span>
<span id="cb198-16"><a href="brca-rnaseq-model-suite.html#cb198-16" aria-hidden="true"></a></span>
<span id="cb198-17"><a href="brca-rnaseq-model-suite.html#cb198-17" aria-hidden="true"></a><span class="co">#SIM &lt;- &quot;Sim_01&quot;</span></span>
<span id="cb198-18"><a href="brca-rnaseq-model-suite.html#cb198-18" aria-hidden="true"></a></span>
<span id="cb198-19"><a href="brca-rnaseq-model-suite.html#cb198-19" aria-hidden="true"></a><span class="cf">for</span>(SIM <span class="cf">in</span> <span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>SimNo)[<span class="dv">1</span>]){</span>
<span id="cb198-20"><a href="brca-rnaseq-model-suite.html#cb198-20" aria-hidden="true"></a></span>
<span id="cb198-21"><a href="brca-rnaseq-model-suite.html#cb198-21" aria-hidden="true"></a>SimNum &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">sub</span>(<span class="st">&#39;Sim_&#39;</span>,<span class="st">&#39;&#39;</span>,SIM))</span>
<span id="cb198-22"><a href="brca-rnaseq-model-suite.html#cb198-22" aria-hidden="true"></a></span>
<span id="cb198-23"><a href="brca-rnaseq-model-suite.html#cb198-23" aria-hidden="true"></a>simNo_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(SimNo<span class="op">==</span>SIM)</span>
<span id="cb198-24"><a href="brca-rnaseq-model-suite.html#cb198-24" aria-hidden="true"></a></span>
<span id="cb198-25"><a href="brca-rnaseq-model-suite.html#cb198-25" aria-hidden="true"></a></span>
<span id="cb198-26"><a href="brca-rnaseq-model-suite.html#cb198-26" aria-hidden="true"></a><span class="co"># errors</span></span>
<span id="cb198-27"><a href="brca-rnaseq-model-suite.html#cb198-27" aria-hidden="true"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb198-28"><a href="brca-rnaseq-model-suite.html#cb198-28" aria-hidden="true"></a><span class="co">###################</span></span>
<span id="cb198-29"><a href="brca-rnaseq-model-suite.html#cb198-29" aria-hidden="true"></a><span class="co"># 1se</span></span>
<span id="cb198-30"><a href="brca-rnaseq-model-suite.html#cb198-30" aria-hidden="true"></a><span class="co">####################</span></span>
<span id="cb198-31"><a href="brca-rnaseq-model-suite.html#cb198-31" aria-hidden="true"></a>cv_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb198-32"><a href="brca-rnaseq-model-suite.html#cb198-32" aria-hidden="true"></a> <span class="kw">split</span>(cv_1se, Size))</span>
<span id="cb198-33"><a href="brca-rnaseq-model-suite.html#cb198-33" aria-hidden="true"></a><span class="kw">names</span>(cv_1se_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(cv_1se_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb198-34"><a href="brca-rnaseq-model-suite.html#cb198-34" aria-hidden="true"></a></span>
<span id="cb198-35"><a href="brca-rnaseq-model-suite.html#cb198-35" aria-hidden="true"></a>test_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb198-36"><a href="brca-rnaseq-model-suite.html#cb198-36" aria-hidden="true"></a> <span class="kw">split</span>(test_1se, Size))</span>
<span id="cb198-37"><a href="brca-rnaseq-model-suite.html#cb198-37" aria-hidden="true"></a><span class="kw">names</span>(test_1se_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(test_1se_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb198-38"><a href="brca-rnaseq-model-suite.html#cb198-38" aria-hidden="true"></a></span>
<span id="cb198-39"><a href="brca-rnaseq-model-suite.html#cb198-39" aria-hidden="true"></a>error_1se_lst &lt;-<span class="st"> </span><span class="kw">c</span>(cv_1se_lst, test_1se_lst)</span>
<span id="cb198-40"><a href="brca-rnaseq-model-suite.html#cb198-40" aria-hidden="true"></a>error_1se_lst &lt;-<span class="st"> </span>error_1se_lst[<span class="kw">order</span>(<span class="kw">names</span>(error_1se_lst))]</span>
<span id="cb198-41"><a href="brca-rnaseq-model-suite.html#cb198-41" aria-hidden="true"></a></span>
<span id="cb198-42"><a href="brca-rnaseq-model-suite.html#cb198-42" aria-hidden="true"></a><span class="kw">boxplot</span>(error_1se_lst, </span>
<span id="cb198-43"><a href="brca-rnaseq-model-suite.html#cb198-43" aria-hidden="true"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>), </span>
<span id="cb198-44"><a href="brca-rnaseq-model-suite.html#cb198-44" aria-hidden="true"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>), </span>
<span id="cb198-45"><a href="brca-rnaseq-model-suite.html#cb198-45" aria-hidden="true"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb198-46"><a href="brca-rnaseq-model-suite.html#cb198-46" aria-hidden="true"></a>)</span>
<span id="cb198-47"><a href="brca-rnaseq-model-suite.html#cb198-47" aria-hidden="true"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T,  <span class="st">&#39;Misclassification Error&#39;</span>)</span>
<span id="cb198-48"><a href="brca-rnaseq-model-suite.html#cb198-48" aria-hidden="true"></a></span>
<span id="cb198-49"><a href="brca-rnaseq-model-suite.html#cb198-49" aria-hidden="true"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb198-50"><a href="brca-rnaseq-model-suite.html#cb198-50" aria-hidden="true"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb198-51"><a href="brca-rnaseq-model-suite.html#cb198-51" aria-hidden="true"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_lst)), </span>
<span id="cb198-52"><a href="brca-rnaseq-model-suite.html#cb198-52" aria-hidden="true"></a>  SIZE0</span>
<span id="cb198-53"><a href="brca-rnaseq-model-suite.html#cb198-53" aria-hidden="true"></a> )</span>
<span id="cb198-54"><a href="brca-rnaseq-model-suite.html#cb198-54" aria-hidden="true"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb198-55"><a href="brca-rnaseq-model-suite.html#cb198-55" aria-hidden="true"></a><span class="kw">abline</span>(<span class="dt">h=</span> error_1se_q2, <span class="dt">col =</span> <span class="st">&#39;red&#39;</span>)</span>
<span id="cb198-56"><a href="brca-rnaseq-model-suite.html#cb198-56" aria-hidden="true"></a><span class="kw">legend</span>(<span class="st">&#39;topright&#39;</span>, </span>
<span id="cb198-57"><a href="brca-rnaseq-model-suite.html#cb198-57" aria-hidden="true"></a>   <span class="co">#title=&#39;1se errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb198-58"><a href="brca-rnaseq-model-suite.html#cb198-58" aria-hidden="true"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb198-59"><a href="brca-rnaseq-model-suite.html#cb198-59" aria-hidden="true"></a>   <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&#39;cv error&#39;</span>, <span class="st">&#39;test set&#39;</span>),</span>
<span id="cb198-60"><a href="brca-rnaseq-model-suite.html#cb198-60" aria-hidden="true"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb198-61"><a href="brca-rnaseq-model-suite.html#cb198-61" aria-hidden="true"></a> )</span>
<span id="cb198-62"><a href="brca-rnaseq-model-suite.html#cb198-62" aria-hidden="true"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;one se lambda models&#39;</span>))</span>
<span id="cb198-63"><a href="brca-rnaseq-model-suite.html#cb198-63" aria-hidden="true"></a></span>
<span id="cb198-64"><a href="brca-rnaseq-model-suite.html#cb198-64" aria-hidden="true"></a>SKIP  &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb198-65"><a href="brca-rnaseq-model-suite.html#cb198-65" aria-hidden="true"></a><span class="co"># Add qual annotation</span></span>
<span id="cb198-66"><a href="brca-rnaseq-model-suite.html#cb198-66" aria-hidden="true"></a>other_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(brcaRna_sim_other_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb198-67"><a href="brca-rnaseq-model-suite.html#cb198-67" aria-hidden="true"></a>LumA_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(brcaRna_sim_LumA_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb198-68"><a href="brca-rnaseq-model-suite.html#cb198-68" aria-hidden="true"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb198-69"><a href="brca-rnaseq-model-suite.html#cb198-69" aria-hidden="true"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb198-70"><a href="brca-rnaseq-model-suite.html#cb198-70" aria-hidden="true"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_lst)),</span>
<span id="cb198-71"><a href="brca-rnaseq-model-suite.html#cb198-71" aria-hidden="true"></a>  <span class="kw">round</span>(other_qual_vec, <span class="dv">2</span>)</span>
<span id="cb198-72"><a href="brca-rnaseq-model-suite.html#cb198-72" aria-hidden="true"></a> )</span>
<span id="cb198-73"><a href="brca-rnaseq-model-suite.html#cb198-73" aria-hidden="true"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb198-74"><a href="brca-rnaseq-model-suite.html#cb198-74" aria-hidden="true"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb198-75"><a href="brca-rnaseq-model-suite.html#cb198-75" aria-hidden="true"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_lst)),</span>
<span id="cb198-76"><a href="brca-rnaseq-model-suite.html#cb198-76" aria-hidden="true"></a>  <span class="kw">round</span>(LumA_qual_vec, <span class="dv">2</span>)</span>
<span id="cb198-77"><a href="brca-rnaseq-model-suite.html#cb198-77" aria-hidden="true"></a> )</span>
<span id="cb198-78"><a href="brca-rnaseq-model-suite.html#cb198-78" aria-hidden="true"></a>}<span class="co">#SKIP</span></span>
<span id="cb198-79"><a href="brca-rnaseq-model-suite.html#cb198-79" aria-hidden="true"></a></span>
<span id="cb198-80"><a href="brca-rnaseq-model-suite.html#cb198-80" aria-hidden="true"></a><span class="co"># min</span></span>
<span id="cb198-81"><a href="brca-rnaseq-model-suite.html#cb198-81" aria-hidden="true"></a><span class="co">####################</span></span>
<span id="cb198-82"><a href="brca-rnaseq-model-suite.html#cb198-82" aria-hidden="true"></a>cv_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb198-83"><a href="brca-rnaseq-model-suite.html#cb198-83" aria-hidden="true"></a> <span class="kw">split</span>(cv_min, Size))</span>
<span id="cb198-84"><a href="brca-rnaseq-model-suite.html#cb198-84" aria-hidden="true"></a><span class="kw">names</span>(cv_min_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(cv_min_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb198-85"><a href="brca-rnaseq-model-suite.html#cb198-85" aria-hidden="true"></a></span>
<span id="cb198-86"><a href="brca-rnaseq-model-suite.html#cb198-86" aria-hidden="true"></a>test_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb198-87"><a href="brca-rnaseq-model-suite.html#cb198-87" aria-hidden="true"></a> <span class="kw">split</span>(test_min, Size))</span>
<span id="cb198-88"><a href="brca-rnaseq-model-suite.html#cb198-88" aria-hidden="true"></a><span class="kw">names</span>(test_min_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(test_min_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb198-89"><a href="brca-rnaseq-model-suite.html#cb198-89" aria-hidden="true"></a></span>
<span id="cb198-90"><a href="brca-rnaseq-model-suite.html#cb198-90" aria-hidden="true"></a>error_min_lst &lt;-<span class="st"> </span><span class="kw">c</span>(cv_min_lst, test_min_lst)</span>
<span id="cb198-91"><a href="brca-rnaseq-model-suite.html#cb198-91" aria-hidden="true"></a>error_min_lst &lt;-<span class="st"> </span>error_min_lst[<span class="kw">order</span>(<span class="kw">names</span>(error_min_lst))]</span>
<span id="cb198-92"><a href="brca-rnaseq-model-suite.html#cb198-92" aria-hidden="true"></a></span>
<span id="cb198-93"><a href="brca-rnaseq-model-suite.html#cb198-93" aria-hidden="true"></a><span class="kw">boxplot</span>(error_min_lst, </span>
<span id="cb198-94"><a href="brca-rnaseq-model-suite.html#cb198-94" aria-hidden="true"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>), </span>
<span id="cb198-95"><a href="brca-rnaseq-model-suite.html#cb198-95" aria-hidden="true"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>),</span>
<span id="cb198-96"><a href="brca-rnaseq-model-suite.html#cb198-96" aria-hidden="true"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb198-97"><a href="brca-rnaseq-model-suite.html#cb198-97" aria-hidden="true"></a>)</span>
<span id="cb198-98"><a href="brca-rnaseq-model-suite.html#cb198-98" aria-hidden="true"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T,  <span class="st">&#39;Misclassification Error&#39;</span>)</span>
<span id="cb198-99"><a href="brca-rnaseq-model-suite.html#cb198-99" aria-hidden="true"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb198-100"><a href="brca-rnaseq-model-suite.html#cb198-100" aria-hidden="true"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb198-101"><a href="brca-rnaseq-model-suite.html#cb198-101" aria-hidden="true"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_lst)), </span>
<span id="cb198-102"><a href="brca-rnaseq-model-suite.html#cb198-102" aria-hidden="true"></a>  SIZE0</span>
<span id="cb198-103"><a href="brca-rnaseq-model-suite.html#cb198-103" aria-hidden="true"></a> )</span>
<span id="cb198-104"><a href="brca-rnaseq-model-suite.html#cb198-104" aria-hidden="true"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb198-105"><a href="brca-rnaseq-model-suite.html#cb198-105" aria-hidden="true"></a><span class="kw">abline</span>(<span class="dt">h=</span> error_min_q2, <span class="dt">col =</span> <span class="st">&#39;red&#39;</span>)</span>
<span id="cb198-106"><a href="brca-rnaseq-model-suite.html#cb198-106" aria-hidden="true"></a><span class="kw">legend</span>(<span class="st">&#39;topright&#39;</span>, </span>
<span id="cb198-107"><a href="brca-rnaseq-model-suite.html#cb198-107" aria-hidden="true"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb198-108"><a href="brca-rnaseq-model-suite.html#cb198-108" aria-hidden="true"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb198-109"><a href="brca-rnaseq-model-suite.html#cb198-109" aria-hidden="true"></a>   <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&#39;cv error&#39;</span>, <span class="st">&#39;test set&#39;</span>),</span>
<span id="cb198-110"><a href="brca-rnaseq-model-suite.html#cb198-110" aria-hidden="true"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb198-111"><a href="brca-rnaseq-model-suite.html#cb198-111" aria-hidden="true"></a> )</span>
<span id="cb198-112"><a href="brca-rnaseq-model-suite.html#cb198-112" aria-hidden="true"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;min lambda models&#39;</span>))</span>
<span id="cb198-113"><a href="brca-rnaseq-model-suite.html#cb198-113" aria-hidden="true"></a></span>
<span id="cb198-114"><a href="brca-rnaseq-model-suite.html#cb198-114" aria-hidden="true"></a>SKIP  &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb198-115"><a href="brca-rnaseq-model-suite.html#cb198-115" aria-hidden="true"></a><span class="co"># Add qual annotation</span></span>
<span id="cb198-116"><a href="brca-rnaseq-model-suite.html#cb198-116" aria-hidden="true"></a>other_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(brcaRna_sim_other_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb198-117"><a href="brca-rnaseq-model-suite.html#cb198-117" aria-hidden="true"></a>LumA_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(brcaRna_sim_LumA_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb198-118"><a href="brca-rnaseq-model-suite.html#cb198-118" aria-hidden="true"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb198-119"><a href="brca-rnaseq-model-suite.html#cb198-119" aria-hidden="true"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb198-120"><a href="brca-rnaseq-model-suite.html#cb198-120" aria-hidden="true"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_lst)),</span>
<span id="cb198-121"><a href="brca-rnaseq-model-suite.html#cb198-121" aria-hidden="true"></a>  <span class="kw">round</span>(other_qual_vec, <span class="dv">2</span>)</span>
<span id="cb198-122"><a href="brca-rnaseq-model-suite.html#cb198-122" aria-hidden="true"></a> )</span>
<span id="cb198-123"><a href="brca-rnaseq-model-suite.html#cb198-123" aria-hidden="true"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb198-124"><a href="brca-rnaseq-model-suite.html#cb198-124" aria-hidden="true"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb198-125"><a href="brca-rnaseq-model-suite.html#cb198-125" aria-hidden="true"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_lst)),</span>
<span id="cb198-126"><a href="brca-rnaseq-model-suite.html#cb198-126" aria-hidden="true"></a>  <span class="kw">round</span>(LumA_qual_vec, <span class="dv">2</span>)</span>
<span id="cb198-127"><a href="brca-rnaseq-model-suite.html#cb198-127" aria-hidden="true"></a> )</span>
<span id="cb198-128"><a href="brca-rnaseq-model-suite.html#cb198-128" aria-hidden="true"></a>}<span class="co">#SKIP</span></span>
<span id="cb198-129"><a href="brca-rnaseq-model-suite.html#cb198-129" aria-hidden="true"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, <span class="kw">paste</span>(<span class="st">&#39;Sim =&#39;</span>,  SIM))</span>
<span id="cb198-130"><a href="brca-rnaseq-model-suite.html#cb198-130" aria-hidden="true"></a></span>
<span id="cb198-131"><a href="brca-rnaseq-model-suite.html#cb198-131" aria-hidden="true"></a>} <span class="co"># for(SIM</span></span></code></pre></div>
<div class="figure"><span id="fig:brca-rnaseq-lasso-simRes-errors-bySim"></span>
<img src="Static/figures/brca-rnaseq-lasso-simRes-errors-bySim-1.png" alt="lasso Model Errors by Sample Size" width="960" />
<p class="caption">
Figure 8.8: lasso Model Errors by Sample Size
</p>
</div>
<p>In this one simulation, we see:</p>
<ul>
<li><p>Model accuracy steadily increases with sample size</p></li>
<li><p>CV error rates tend to be pessimistic, expecially for the small sample sizes
and more so for the 1se models.</p></li>
<li><p>There isn’t much to chose from between the one standard error and the minimum lambda models.</p></li>
</ul>
</div>
<div id="feature-selection-2" class="section level4" number="8.6.1.2">
<h4><span class="header-section-number">8.6.1.2</span> Feature Selection</h4>
<div class="sourceCode" id="cb199"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb199-1"><a href="brca-rnaseq-model-suite.html#cb199-1" aria-hidden="true"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb199-2"><a href="brca-rnaseq-model-suite.html#cb199-2" aria-hidden="true"></a></span>
<span id="cb199-3"><a href="brca-rnaseq-model-suite.html#cb199-3" aria-hidden="true"></a><span class="co"># get full model nzero ref</span></span>
<span id="cb199-4"><a href="brca-rnaseq-model-suite.html#cb199-4" aria-hidden="true"></a>nzero_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb199-5"><a href="brca-rnaseq-model-suite.html#cb199-5" aria-hidden="true"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb199-6"><a href="brca-rnaseq-model-suite.html#cb199-6" aria-hidden="true"></a>nzero_1se_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(nzero_1se_vec, <span class="dt">prob=</span><span class="kw">c</span>(<span class="dv">2</span>)<span class="op">/</span><span class="dv">4</span>)</span>
<span id="cb199-7"><a href="brca-rnaseq-model-suite.html#cb199-7" aria-hidden="true"></a></span>
<span id="cb199-8"><a href="brca-rnaseq-model-suite.html#cb199-8" aria-hidden="true"></a>nzero_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb199-9"><a href="brca-rnaseq-model-suite.html#cb199-9" aria-hidden="true"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>nzero[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb199-10"><a href="brca-rnaseq-model-suite.html#cb199-10" aria-hidden="true"></a>nzero_min_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(nzero_min_vec, <span class="dt">prob=</span><span class="kw">c</span>(<span class="dv">2</span>)<span class="op">/</span><span class="dv">4</span>)</span>
<span id="cb199-11"><a href="brca-rnaseq-model-suite.html#cb199-11" aria-hidden="true"></a></span>
<span id="cb199-12"><a href="brca-rnaseq-model-suite.html#cb199-12" aria-hidden="true"></a><span class="co"># Utility objects</span></span>
<span id="cb199-13"><a href="brca-rnaseq-model-suite.html#cb199-13" aria-hidden="true"></a>SIZE0 &lt;-<span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_pad</span>(SIZE, <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>)</span>
<span id="cb199-14"><a href="brca-rnaseq-model-suite.html#cb199-14" aria-hidden="true"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(brcaRna_sim_other_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>,SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb199-15"><a href="brca-rnaseq-model-suite.html#cb199-15" aria-hidden="true"></a></span>
<span id="cb199-16"><a href="brca-rnaseq-model-suite.html#cb199-16" aria-hidden="true"></a></span>
<span id="cb199-17"><a href="brca-rnaseq-model-suite.html#cb199-17" aria-hidden="true"></a><span class="co">#SIM &lt;- &quot;Sim_01&quot;</span></span>
<span id="cb199-18"><a href="brca-rnaseq-model-suite.html#cb199-18" aria-hidden="true"></a></span>
<span id="cb199-19"><a href="brca-rnaseq-model-suite.html#cb199-19" aria-hidden="true"></a><span class="cf">for</span>(SIM <span class="cf">in</span> <span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>SimNo)[<span class="dv">1</span>]){</span>
<span id="cb199-20"><a href="brca-rnaseq-model-suite.html#cb199-20" aria-hidden="true"></a></span>
<span id="cb199-21"><a href="brca-rnaseq-model-suite.html#cb199-21" aria-hidden="true"></a>SimNum &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">sub</span>(<span class="st">&#39;Sim_&#39;</span>,<span class="st">&#39;&#39;</span>,SIM))</span>
<span id="cb199-22"><a href="brca-rnaseq-model-suite.html#cb199-22" aria-hidden="true"></a></span>
<span id="cb199-23"><a href="brca-rnaseq-model-suite.html#cb199-23" aria-hidden="true"></a>simNo_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(SimNo<span class="op">==</span>SIM)</span>
<span id="cb199-24"><a href="brca-rnaseq-model-suite.html#cb199-24" aria-hidden="true"></a></span>
<span id="cb199-25"><a href="brca-rnaseq-model-suite.html#cb199-25" aria-hidden="true"></a></span>
<span id="cb199-26"><a href="brca-rnaseq-model-suite.html#cb199-26" aria-hidden="true"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb199-27"><a href="brca-rnaseq-model-suite.html#cb199-27" aria-hidden="true"></a><span class="co">###################</span></span>
<span id="cb199-28"><a href="brca-rnaseq-model-suite.html#cb199-28" aria-hidden="true"></a><span class="co"># 1se</span></span>
<span id="cb199-29"><a href="brca-rnaseq-model-suite.html#cb199-29" aria-hidden="true"></a><span class="co">####################</span></span>
<span id="cb199-30"><a href="brca-rnaseq-model-suite.html#cb199-30" aria-hidden="true"></a><span class="co"># selected feature counts</span></span>
<span id="cb199-31"><a href="brca-rnaseq-model-suite.html#cb199-31" aria-hidden="true"></a>p_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb199-32"><a href="brca-rnaseq-model-suite.html#cb199-32" aria-hidden="true"></a> <span class="kw">split</span>(p_1se, Size))</span>
<span id="cb199-33"><a href="brca-rnaseq-model-suite.html#cb199-33" aria-hidden="true"></a><span class="kw">names</span>(p_1se_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(p_1se_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_p&#39;</span>)</span>
<span id="cb199-34"><a href="brca-rnaseq-model-suite.html#cb199-34" aria-hidden="true"></a></span>
<span id="cb199-35"><a href="brca-rnaseq-model-suite.html#cb199-35" aria-hidden="true"></a><span class="co"># get selected features that are part of lasso_gene_sign_1se_vec</span></span>
<span id="cb199-36"><a href="brca-rnaseq-model-suite.html#cb199-36" aria-hidden="true"></a><span class="co"># - the signature selected genes</span></span>
<span id="cb199-37"><a href="brca-rnaseq-model-suite.html#cb199-37" aria-hidden="true"></a>sign_genes_1se_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(simNo_results_frm), <span class="cf">function</span>(RR)</span>
<span id="cb199-38"><a href="brca-rnaseq-model-suite.html#cb199-38" aria-hidden="true"></a>    <span class="kw">intersect</span>(<span class="kw">unlist</span>(simNo_results_frm[RR, <span class="st">&#39;genes_1se&#39;</span>]), lasso_gene_sign_1se_vec))</span>
<span id="cb199-39"><a href="brca-rnaseq-model-suite.html#cb199-39" aria-hidden="true"></a></span>
<span id="cb199-40"><a href="brca-rnaseq-model-suite.html#cb199-40" aria-hidden="true"></a>sign_p_1se_lst &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="kw">sapply</span>(sign_genes_1se_lst, length), simNo_results_frm<span class="op">$</span>Size)</span>
<span id="cb199-41"><a href="brca-rnaseq-model-suite.html#cb199-41" aria-hidden="true"></a><span class="kw">names</span>(sign_p_1se_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(sign_p_1se_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_signP&#39;</span>)</span>
<span id="cb199-42"><a href="brca-rnaseq-model-suite.html#cb199-42" aria-hidden="true"></a></span>
<span id="cb199-43"><a href="brca-rnaseq-model-suite.html#cb199-43" aria-hidden="true"></a></span>
<span id="cb199-44"><a href="brca-rnaseq-model-suite.html#cb199-44" aria-hidden="true"></a>p_singP_1se_lst &lt;-<span class="st"> </span><span class="kw">c</span>(p_1se_lst, sign_p_1se_lst)</span>
<span id="cb199-45"><a href="brca-rnaseq-model-suite.html#cb199-45" aria-hidden="true"></a>p_singP_1se_lst &lt;-<span class="st"> </span>p_singP_1se_lst[<span class="kw">order</span>(<span class="kw">names</span>(p_singP_1se_lst))]</span>
<span id="cb199-46"><a href="brca-rnaseq-model-suite.html#cb199-46" aria-hidden="true"></a></span>
<span id="cb199-47"><a href="brca-rnaseq-model-suite.html#cb199-47" aria-hidden="true"></a><span class="kw">boxplot</span>(p_singP_1se_lst,</span>
<span id="cb199-48"><a href="brca-rnaseq-model-suite.html#cb199-48" aria-hidden="true"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb199-49"><a href="brca-rnaseq-model-suite.html#cb199-49" aria-hidden="true"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">200</span>),</span>
<span id="cb199-50"><a href="brca-rnaseq-model-suite.html#cb199-50" aria-hidden="true"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb199-51"><a href="brca-rnaseq-model-suite.html#cb199-51" aria-hidden="true"></a>)</span>
<span id="cb199-52"><a href="brca-rnaseq-model-suite.html#cb199-52" aria-hidden="true"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T,  <span class="st">&#39;number of selected features&#39;</span>)</span>
<span id="cb199-53"><a href="brca-rnaseq-model-suite.html#cb199-53" aria-hidden="true"></a></span>
<span id="cb199-54"><a href="brca-rnaseq-model-suite.html#cb199-54" aria-hidden="true"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb199-55"><a href="brca-rnaseq-model-suite.html#cb199-55" aria-hidden="true"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb199-56"><a href="brca-rnaseq-model-suite.html#cb199-56" aria-hidden="true"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_lst)),</span>
<span id="cb199-57"><a href="brca-rnaseq-model-suite.html#cb199-57" aria-hidden="true"></a>  SIZE0</span>
<span id="cb199-58"><a href="brca-rnaseq-model-suite.html#cb199-58" aria-hidden="true"></a> )</span>
<span id="cb199-59"><a href="brca-rnaseq-model-suite.html#cb199-59" aria-hidden="true"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb199-60"><a href="brca-rnaseq-model-suite.html#cb199-60" aria-hidden="true"></a><span class="co">#abline(h= nzero_1se_q2, col = &#39;red&#39;)</span></span>
<span id="cb199-61"><a href="brca-rnaseq-model-suite.html#cb199-61" aria-hidden="true"></a><span class="kw">legend</span>(<span class="st">&#39;topleft&#39;</span>,</span>
<span id="cb199-62"><a href="brca-rnaseq-model-suite.html#cb199-62" aria-hidden="true"></a>   <span class="co">#title=&#39;1se errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb199-63"><a href="brca-rnaseq-model-suite.html#cb199-63" aria-hidden="true"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>),</span>
<span id="cb199-64"><a href="brca-rnaseq-model-suite.html#cb199-64" aria-hidden="true"></a>   <span class="dt">legend=</span> <span class="kw">c</span>(<span class="st">&#39;selected genes&#39;</span>,<span class="st">&#39;signature genes&#39;</span>),</span>
<span id="cb199-65"><a href="brca-rnaseq-model-suite.html#cb199-65" aria-hidden="true"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb199-66"><a href="brca-rnaseq-model-suite.html#cb199-66" aria-hidden="true"></a> )</span>
<span id="cb199-67"><a href="brca-rnaseq-model-suite.html#cb199-67" aria-hidden="true"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;one se lambda models&#39;</span>) )</span>
<span id="cb199-68"><a href="brca-rnaseq-model-suite.html#cb199-68" aria-hidden="true"></a></span>
<span id="cb199-69"><a href="brca-rnaseq-model-suite.html#cb199-69" aria-hidden="true"></a>SKIP  &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb199-70"><a href="brca-rnaseq-model-suite.html#cb199-70" aria-hidden="true"></a><span class="co"># Add qual annotation</span></span>
<span id="cb199-71"><a href="brca-rnaseq-model-suite.html#cb199-71" aria-hidden="true"></a>other_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(brcaRna_sim_other_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb199-72"><a href="brca-rnaseq-model-suite.html#cb199-72" aria-hidden="true"></a>LumA_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(brcaRna_sim_LumA_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb199-73"><a href="brca-rnaseq-model-suite.html#cb199-73" aria-hidden="true"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb199-74"><a href="brca-rnaseq-model-suite.html#cb199-74" aria-hidden="true"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb199-75"><a href="brca-rnaseq-model-suite.html#cb199-75" aria-hidden="true"></a>  <span class="dt">at =</span>  <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_lst)),</span>
<span id="cb199-76"><a href="brca-rnaseq-model-suite.html#cb199-76" aria-hidden="true"></a>  <span class="kw">round</span>(other_qual_vec, <span class="dv">2</span>)</span>
<span id="cb199-77"><a href="brca-rnaseq-model-suite.html#cb199-77" aria-hidden="true"></a> )</span>
<span id="cb199-78"><a href="brca-rnaseq-model-suite.html#cb199-78" aria-hidden="true"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb199-79"><a href="brca-rnaseq-model-suite.html#cb199-79" aria-hidden="true"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb199-80"><a href="brca-rnaseq-model-suite.html#cb199-80" aria-hidden="true"></a>  <span class="dt">at =</span>  <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_lst)),</span>
<span id="cb199-81"><a href="brca-rnaseq-model-suite.html#cb199-81" aria-hidden="true"></a>  <span class="kw">round</span>(LumA_qual_vec, <span class="dv">2</span>)</span>
<span id="cb199-82"><a href="brca-rnaseq-model-suite.html#cb199-82" aria-hidden="true"></a> )</span>
<span id="cb199-83"><a href="brca-rnaseq-model-suite.html#cb199-83" aria-hidden="true"></a>}<span class="co">#SKIP</span></span>
<span id="cb199-84"><a href="brca-rnaseq-model-suite.html#cb199-84" aria-hidden="true"></a></span>
<span id="cb199-85"><a href="brca-rnaseq-model-suite.html#cb199-85" aria-hidden="true"></a><span class="co">###################</span></span>
<span id="cb199-86"><a href="brca-rnaseq-model-suite.html#cb199-86" aria-hidden="true"></a><span class="co"># min</span></span>
<span id="cb199-87"><a href="brca-rnaseq-model-suite.html#cb199-87" aria-hidden="true"></a><span class="co">####################</span></span>
<span id="cb199-88"><a href="brca-rnaseq-model-suite.html#cb199-88" aria-hidden="true"></a><span class="co"># selected feature counts</span></span>
<span id="cb199-89"><a href="brca-rnaseq-model-suite.html#cb199-89" aria-hidden="true"></a>p_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(simNo_results_frm,</span>
<span id="cb199-90"><a href="brca-rnaseq-model-suite.html#cb199-90" aria-hidden="true"></a> <span class="kw">split</span>(p_min, Size))</span>
<span id="cb199-91"><a href="brca-rnaseq-model-suite.html#cb199-91" aria-hidden="true"></a><span class="kw">names</span>(p_min_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(p_min_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_p&#39;</span>)</span>
<span id="cb199-92"><a href="brca-rnaseq-model-suite.html#cb199-92" aria-hidden="true"></a></span>
<span id="cb199-93"><a href="brca-rnaseq-model-suite.html#cb199-93" aria-hidden="true"></a><span class="co"># get selected features that are part of lasso_gene_sign_min_vec</span></span>
<span id="cb199-94"><a href="brca-rnaseq-model-suite.html#cb199-94" aria-hidden="true"></a><span class="co"># - the signature selected genes</span></span>
<span id="cb199-95"><a href="brca-rnaseq-model-suite.html#cb199-95" aria-hidden="true"></a>sign_genes_min_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(simNo_results_frm), <span class="cf">function</span>(RR)</span>
<span id="cb199-96"><a href="brca-rnaseq-model-suite.html#cb199-96" aria-hidden="true"></a>    <span class="kw">intersect</span>(<span class="kw">unlist</span>(simNo_results_frm[RR, <span class="st">&#39;genes_min&#39;</span>]), lasso_gene_sign_min_vec))</span>
<span id="cb199-97"><a href="brca-rnaseq-model-suite.html#cb199-97" aria-hidden="true"></a></span>
<span id="cb199-98"><a href="brca-rnaseq-model-suite.html#cb199-98" aria-hidden="true"></a>sign_p_min_lst &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="kw">sapply</span>(sign_genes_min_lst, length), simNo_results_frm<span class="op">$</span>Size)</span>
<span id="cb199-99"><a href="brca-rnaseq-model-suite.html#cb199-99" aria-hidden="true"></a><span class="kw">names</span>(sign_p_min_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">names</span>(sign_p_min_lst), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>),<span class="st">&#39;_signP&#39;</span>)</span>
<span id="cb199-100"><a href="brca-rnaseq-model-suite.html#cb199-100" aria-hidden="true"></a></span>
<span id="cb199-101"><a href="brca-rnaseq-model-suite.html#cb199-101" aria-hidden="true"></a></span>
<span id="cb199-102"><a href="brca-rnaseq-model-suite.html#cb199-102" aria-hidden="true"></a>p_singP_min_lst &lt;-<span class="st"> </span><span class="kw">c</span>(p_min_lst, sign_p_min_lst)</span>
<span id="cb199-103"><a href="brca-rnaseq-model-suite.html#cb199-103" aria-hidden="true"></a>p_singP_min_lst &lt;-<span class="st"> </span>p_singP_min_lst[<span class="kw">order</span>(<span class="kw">names</span>(p_singP_min_lst))]</span>
<span id="cb199-104"><a href="brca-rnaseq-model-suite.html#cb199-104" aria-hidden="true"></a></span>
<span id="cb199-105"><a href="brca-rnaseq-model-suite.html#cb199-105" aria-hidden="true"></a><span class="kw">boxplot</span>(p_singP_min_lst,</span>
<span id="cb199-106"><a href="brca-rnaseq-model-suite.html#cb199-106" aria-hidden="true"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb199-107"><a href="brca-rnaseq-model-suite.html#cb199-107" aria-hidden="true"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">200</span>),</span>
<span id="cb199-108"><a href="brca-rnaseq-model-suite.html#cb199-108" aria-hidden="true"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb199-109"><a href="brca-rnaseq-model-suite.html#cb199-109" aria-hidden="true"></a>)</span>
<span id="cb199-110"><a href="brca-rnaseq-model-suite.html#cb199-110" aria-hidden="true"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb199-111"><a href="brca-rnaseq-model-suite.html#cb199-111" aria-hidden="true"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb199-112"><a href="brca-rnaseq-model-suite.html#cb199-112" aria-hidden="true"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_lst)),</span>
<span id="cb199-113"><a href="brca-rnaseq-model-suite.html#cb199-113" aria-hidden="true"></a>  SIZE0</span>
<span id="cb199-114"><a href="brca-rnaseq-model-suite.html#cb199-114" aria-hidden="true"></a> )</span>
<span id="cb199-115"><a href="brca-rnaseq-model-suite.html#cb199-115" aria-hidden="true"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb199-116"><a href="brca-rnaseq-model-suite.html#cb199-116" aria-hidden="true"></a><span class="co">#abline(h= nzero_min_q2, col = &#39;red&#39;)</span></span>
<span id="cb199-117"><a href="brca-rnaseq-model-suite.html#cb199-117" aria-hidden="true"></a><span class="kw">legend</span>(<span class="st">&#39;topleft&#39;</span>,</span>
<span id="cb199-118"><a href="brca-rnaseq-model-suite.html#cb199-118" aria-hidden="true"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb199-119"><a href="brca-rnaseq-model-suite.html#cb199-119" aria-hidden="true"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>),</span>
<span id="cb199-120"><a href="brca-rnaseq-model-suite.html#cb199-120" aria-hidden="true"></a>   <span class="dt">legend=</span> <span class="kw">c</span>(<span class="st">&#39;selected genes&#39;</span>,<span class="st">&#39;signature genes&#39;</span>),</span>
<span id="cb199-121"><a href="brca-rnaseq-model-suite.html#cb199-121" aria-hidden="true"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb199-122"><a href="brca-rnaseq-model-suite.html#cb199-122" aria-hidden="true"></a> )</span>
<span id="cb199-123"><a href="brca-rnaseq-model-suite.html#cb199-123" aria-hidden="true"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;min lambda models&#39;</span>))</span>
<span id="cb199-124"><a href="brca-rnaseq-model-suite.html#cb199-124" aria-hidden="true"></a></span>
<span id="cb199-125"><a href="brca-rnaseq-model-suite.html#cb199-125" aria-hidden="true"></a>SKIP  &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb199-126"><a href="brca-rnaseq-model-suite.html#cb199-126" aria-hidden="true"></a><span class="co"># Add qual annotation</span></span>
<span id="cb199-127"><a href="brca-rnaseq-model-suite.html#cb199-127" aria-hidden="true"></a>other_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(brcaRna_sim_other_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb199-128"><a href="brca-rnaseq-model-suite.html#cb199-128" aria-hidden="true"></a>LumA_qual_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">split</span>(brcaRna_sim_LumA_qual_mtx[,SimNum], stage_vec), median)</span>
<span id="cb199-129"><a href="brca-rnaseq-model-suite.html#cb199-129" aria-hidden="true"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb199-130"><a href="brca-rnaseq-model-suite.html#cb199-130" aria-hidden="true"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb199-131"><a href="brca-rnaseq-model-suite.html#cb199-131" aria-hidden="true"></a>  <span class="dt">at =</span>  <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_lst)),</span>
<span id="cb199-132"><a href="brca-rnaseq-model-suite.html#cb199-132" aria-hidden="true"></a>  <span class="kw">round</span>(other_qual_vec, <span class="dv">2</span>)</span>
<span id="cb199-133"><a href="brca-rnaseq-model-suite.html#cb199-133" aria-hidden="true"></a> )</span>
<span id="cb199-134"><a href="brca-rnaseq-model-suite.html#cb199-134" aria-hidden="true"></a>LL &lt;-<span class="st"> </span>LL <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb199-135"><a href="brca-rnaseq-model-suite.html#cb199-135" aria-hidden="true"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb199-136"><a href="brca-rnaseq-model-suite.html#cb199-136" aria-hidden="true"></a>  <span class="dt">at =</span>  <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_lst)),</span>
<span id="cb199-137"><a href="brca-rnaseq-model-suite.html#cb199-137" aria-hidden="true"></a>  <span class="kw">round</span>(LumA_qual_vec, <span class="dv">2</span>)</span>
<span id="cb199-138"><a href="brca-rnaseq-model-suite.html#cb199-138" aria-hidden="true"></a> )</span>
<span id="cb199-139"><a href="brca-rnaseq-model-suite.html#cb199-139" aria-hidden="true"></a>}<span class="co">#SKIP</span></span>
<span id="cb199-140"><a href="brca-rnaseq-model-suite.html#cb199-140" aria-hidden="true"></a></span>
<span id="cb199-141"><a href="brca-rnaseq-model-suite.html#cb199-141" aria-hidden="true"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, <span class="kw">paste</span>(<span class="st">&#39;Sim =&#39;</span>,  SIM))</span>
<span id="cb199-142"><a href="brca-rnaseq-model-suite.html#cb199-142" aria-hidden="true"></a></span>
<span id="cb199-143"><a href="brca-rnaseq-model-suite.html#cb199-143" aria-hidden="true"></a>} <span class="co"># for(SIM</span></span></code></pre></div>
<div class="figure"><span id="fig:brca-rnaseq-lasso-simRes-features-bySim"></span>
<img src="Static/figures/brca-rnaseq-lasso-simRes-features-bySim-1.png" alt="lasso Models Selected Features by Sample Size" width="960" />
<p class="caption">
Figure 8.9: lasso Models Selected Features by Sample Size
</p>
</div>
<p>Compare with Figures
<a href="hcc-5hmcseq-model-suite.html#fig:hcc5hmC-glmnetSuite-lasso-simRes-errors-bySim">5.8</a>
and
<a href="hcc-5hmcseq-model-suite.html#fig:hcc5hmC-glmnetSuite-lasso-simRes-features-bySim">5.9</a>
for the HCC 5hmC data.</p>
<p><strong>Reader Note</strong></p>
<pre><code>IN the case of discriminating between LumA and other breast cancer subtypes,
we could use features in the LumA breast cancer subtype signature as
positive controls for feature selection.

To Do.</code></pre>
<!-- 
In this one simulation, we see:

* The selected number of features in the smaller sample size analyses
are low, with few feaures belonging to the core signature identified in the
full data set.  

* As the sample size increases the number of features selected in the minimum 
lambda model remains variable, but the number of core signature features
selected in the samples of sizes 200 and 300 is stable and between 40 and 50.  
-->
</div>
</div>
<div id="results-across-simulation-runs.-1" class="section level3" number="8.6.2">
<h3><span class="header-section-number">8.6.2</span> Results across simulation runs.</h3>
<div id="model-accuracy-assessment-3" class="section level4" number="8.6.2.1">
<h4><span class="header-section-number">8.6.2.1</span> Model Accuracy Assessment</h4>
<p>Now look acoss all simulations. In the figures that follow, each boxplot
summarizes the results of 30 simulations. For a give sample size and a
given simulation, each data point is the median across 30 repeated cv runs.</p>
<div class="sourceCode" id="cb201"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb201-1"><a href="brca-rnaseq-model-suite.html#cb201-1" aria-hidden="true"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb201-2"><a href="brca-rnaseq-model-suite.html#cb201-2" aria-hidden="true"></a></span>
<span id="cb201-3"><a href="brca-rnaseq-model-suite.html#cb201-3" aria-hidden="true"></a><span class="co"># get full model cv error ref</span></span>
<span id="cb201-4"><a href="brca-rnaseq-model-suite.html#cb201-4" aria-hidden="true"></a>error_1se_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb201-5"><a href="brca-rnaseq-model-suite.html#cb201-5" aria-hidden="true"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se])</span>
<span id="cb201-6"><a href="brca-rnaseq-model-suite.html#cb201-6" aria-hidden="true"></a>error_1se_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(error_1se_vec, <span class="dt">prob=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>)        </span>
<span id="cb201-7"><a href="brca-rnaseq-model-suite.html#cb201-7" aria-hidden="true"></a></span>
<span id="cb201-8"><a href="brca-rnaseq-model-suite.html#cb201-8" aria-hidden="true"></a>error_min_vec &lt;-<span class="st"> </span><span class="kw">sapply</span>(brcaRna_cv_lassoAll_lst,</span>
<span id="cb201-9"><a href="brca-rnaseq-model-suite.html#cb201-9" aria-hidden="true"></a> <span class="cf">function</span>(cv_fit) cv_fit<span class="op">$</span>cvm[cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda.min])</span>
<span id="cb201-10"><a href="brca-rnaseq-model-suite.html#cb201-10" aria-hidden="true"></a>error_min_q2 &lt;-<span class="st"> </span><span class="kw">quantile</span>(error_min_vec, <span class="dt">prob=</span><span class="dv">1</span><span class="op">/</span><span class="dv">2</span>)        </span>
<span id="cb201-11"><a href="brca-rnaseq-model-suite.html#cb201-11" aria-hidden="true"></a></span>
<span id="cb201-12"><a href="brca-rnaseq-model-suite.html#cb201-12" aria-hidden="true"></a><span class="co"># Utility objects</span></span>
<span id="cb201-13"><a href="brca-rnaseq-model-suite.html#cb201-13" aria-hidden="true"></a>SIZE0 &lt;-<span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_pad</span>(SIZE, <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>)</span>
<span id="cb201-14"><a href="brca-rnaseq-model-suite.html#cb201-14" aria-hidden="true"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(brcaRna_sim_other_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>,SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb201-15"><a href="brca-rnaseq-model-suite.html#cb201-15" aria-hidden="true"></a></span>
<span id="cb201-16"><a href="brca-rnaseq-model-suite.html#cb201-16" aria-hidden="true"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb201-17"><a href="brca-rnaseq-model-suite.html#cb201-17" aria-hidden="true"></a><span class="co"># 1se</span></span>
<span id="cb201-18"><a href="brca-rnaseq-model-suite.html#cb201-18" aria-hidden="true"></a><span class="co">#########################################</span></span>
<span id="cb201-19"><a href="brca-rnaseq-model-suite.html#cb201-19" aria-hidden="true"></a><span class="co">## cv</span></span>
<span id="cb201-20"><a href="brca-rnaseq-model-suite.html#cb201-20" aria-hidden="true"></a>cv_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb201-21"><a href="brca-rnaseq-model-suite.html#cb201-21" aria-hidden="true"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb201-22"><a href="brca-rnaseq-model-suite.html#cb201-22" aria-hidden="true"></a> sizeVal_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb201-23"><a href="brca-rnaseq-model-suite.html#cb201-23" aria-hidden="true"></a> sizeVal_cv_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(cv_1se, SimNo))</span>
<span id="cb201-24"><a href="brca-rnaseq-model-suite.html#cb201-24" aria-hidden="true"></a> <span class="kw">sapply</span>(sizeVal_cv_1se_lst, median)</span>
<span id="cb201-25"><a href="brca-rnaseq-model-suite.html#cb201-25" aria-hidden="true"></a>})</span>
<span id="cb201-26"><a href="brca-rnaseq-model-suite.html#cb201-26" aria-hidden="true"></a><span class="kw">names</span>(cv_1se_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb201-27"><a href="brca-rnaseq-model-suite.html#cb201-27" aria-hidden="true"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb201-28"><a href="brca-rnaseq-model-suite.html#cb201-28" aria-hidden="true"></a></span>
<span id="cb201-29"><a href="brca-rnaseq-model-suite.html#cb201-29" aria-hidden="true"></a><span class="co">## test</span></span>
<span id="cb201-30"><a href="brca-rnaseq-model-suite.html#cb201-30" aria-hidden="true"></a>test_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb201-31"><a href="brca-rnaseq-model-suite.html#cb201-31" aria-hidden="true"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb201-32"><a href="brca-rnaseq-model-suite.html#cb201-32" aria-hidden="true"></a> sizeVal_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb201-33"><a href="brca-rnaseq-model-suite.html#cb201-33" aria-hidden="true"></a> sizeVal_test_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(test_1se, SimNo))</span>
<span id="cb201-34"><a href="brca-rnaseq-model-suite.html#cb201-34" aria-hidden="true"></a> <span class="kw">sapply</span>(sizeVal_test_1se_lst, median)</span>
<span id="cb201-35"><a href="brca-rnaseq-model-suite.html#cb201-35" aria-hidden="true"></a>})</span>
<span id="cb201-36"><a href="brca-rnaseq-model-suite.html#cb201-36" aria-hidden="true"></a><span class="kw">names</span>(test_1se_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb201-37"><a href="brca-rnaseq-model-suite.html#cb201-37" aria-hidden="true"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_test&#39;</span>)</span>
<span id="cb201-38"><a href="brca-rnaseq-model-suite.html#cb201-38" aria-hidden="true"></a></span>
<span id="cb201-39"><a href="brca-rnaseq-model-suite.html#cb201-39" aria-hidden="true"></a></span>
<span id="cb201-40"><a href="brca-rnaseq-model-suite.html#cb201-40" aria-hidden="true"></a>error_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">c</span>(cv_1se_Bysize_lst, test_1se_Bysize_lst)</span>
<span id="cb201-41"><a href="brca-rnaseq-model-suite.html#cb201-41" aria-hidden="true"></a>error_1se_Bysize_lst &lt;-<span class="st"> </span>error_1se_Bysize_lst[<span class="kw">order</span>(<span class="kw">names</span>(error_1se_Bysize_lst))]</span>
<span id="cb201-42"><a href="brca-rnaseq-model-suite.html#cb201-42" aria-hidden="true"></a></span>
<span id="cb201-43"><a href="brca-rnaseq-model-suite.html#cb201-43" aria-hidden="true"></a><span class="kw">boxplot</span>(error_1se_Bysize_lst,</span>
<span id="cb201-44"><a href="brca-rnaseq-model-suite.html#cb201-44" aria-hidden="true"></a>  <span class="dt">col=</span><span class="dv">0</span>,</span>
<span id="cb201-45"><a href="brca-rnaseq-model-suite.html#cb201-45" aria-hidden="true"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb201-46"><a href="brca-rnaseq-model-suite.html#cb201-46" aria-hidden="true"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">.5</span>),</span>
<span id="cb201-47"><a href="brca-rnaseq-model-suite.html#cb201-47" aria-hidden="true"></a>  <span class="dt">outline=</span>F,</span>
<span id="cb201-48"><a href="brca-rnaseq-model-suite.html#cb201-48" aria-hidden="true"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb201-49"><a href="brca-rnaseq-model-suite.html#cb201-49" aria-hidden="true"></a>)</span>
<span id="cb201-50"><a href="brca-rnaseq-model-suite.html#cb201-50" aria-hidden="true"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T,  <span class="st">&#39;Misclassification Error&#39;</span>)</span>
<span id="cb201-51"><a href="brca-rnaseq-model-suite.html#cb201-51" aria-hidden="true"></a></span>
<span id="cb201-52"><a href="brca-rnaseq-model-suite.html#cb201-52" aria-hidden="true"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(error_1se_Bysize_lst))</span>
<span id="cb201-53"><a href="brca-rnaseq-model-suite.html#cb201-53" aria-hidden="true"></a><span class="kw">points</span>(</span>
<span id="cb201-54"><a href="brca-rnaseq-model-suite.html#cb201-54" aria-hidden="true"></a>   <span class="dt">x=</span><span class="kw">jitter</span>(<span class="kw">rep</span>(JJ, <span class="kw">length</span>(error_1se_Bysize_lst[[JJ]])), <span class="dt">amount=</span><span class="fl">0.25</span>), </span>
<span id="cb201-55"><a href="brca-rnaseq-model-suite.html#cb201-55" aria-hidden="true"></a>   <span class="dt">y=</span>error_1se_Bysize_lst[[JJ]], <span class="dt">cex=</span><span class="fl">0.5</span>,</span>
<span id="cb201-56"><a href="brca-rnaseq-model-suite.html#cb201-56" aria-hidden="true"></a>   <span class="dt">col=</span><span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">&#39;cv&#39;</span>, <span class="kw">names</span>(error_1se_Bysize_lst)[JJ]),<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb201-57"><a href="brca-rnaseq-model-suite.html#cb201-57" aria-hidden="true"></a>)</span>
<span id="cb201-58"><a href="brca-rnaseq-model-suite.html#cb201-58" aria-hidden="true"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb201-59"><a href="brca-rnaseq-model-suite.html#cb201-59" aria-hidden="true"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb201-60"><a href="brca-rnaseq-model-suite.html#cb201-60" aria-hidden="true"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_Bysize_lst)),</span>
<span id="cb201-61"><a href="brca-rnaseq-model-suite.html#cb201-61" aria-hidden="true"></a>  SIZE0</span>
<span id="cb201-62"><a href="brca-rnaseq-model-suite.html#cb201-62" aria-hidden="true"></a> )</span>
<span id="cb201-63"><a href="brca-rnaseq-model-suite.html#cb201-63" aria-hidden="true"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_1se_Bysize_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb201-64"><a href="brca-rnaseq-model-suite.html#cb201-64" aria-hidden="true"></a><span class="kw">abline</span>(<span class="dt">h=</span> error_min_q2, <span class="dt">col =</span> <span class="st">&#39;red&#39;</span>)</span>
<span id="cb201-65"><a href="brca-rnaseq-model-suite.html#cb201-65" aria-hidden="true"></a><span class="kw">legend</span>(<span class="st">&#39;topright&#39;</span>,</span>
<span id="cb201-66"><a href="brca-rnaseq-model-suite.html#cb201-66" aria-hidden="true"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb201-67"><a href="brca-rnaseq-model-suite.html#cb201-67" aria-hidden="true"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb201-68"><a href="brca-rnaseq-model-suite.html#cb201-68" aria-hidden="true"></a>   <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&#39;cv error&#39;</span>, <span class="st">&#39;test set&#39;</span>),</span>
<span id="cb201-69"><a href="brca-rnaseq-model-suite.html#cb201-69" aria-hidden="true"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb201-70"><a href="brca-rnaseq-model-suite.html#cb201-70" aria-hidden="true"></a> )</span>
<span id="cb201-71"><a href="brca-rnaseq-model-suite.html#cb201-71" aria-hidden="true"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;one se lambda models&#39;</span>))</span>
<span id="cb201-72"><a href="brca-rnaseq-model-suite.html#cb201-72" aria-hidden="true"></a></span>
<span id="cb201-73"><a href="brca-rnaseq-model-suite.html#cb201-73" aria-hidden="true"></a></span>
<span id="cb201-74"><a href="brca-rnaseq-model-suite.html#cb201-74" aria-hidden="true"></a><span class="co"># min</span></span>
<span id="cb201-75"><a href="brca-rnaseq-model-suite.html#cb201-75" aria-hidden="true"></a><span class="co">#########################################</span></span>
<span id="cb201-76"><a href="brca-rnaseq-model-suite.html#cb201-76" aria-hidden="true"></a><span class="co">## cv</span></span>
<span id="cb201-77"><a href="brca-rnaseq-model-suite.html#cb201-77" aria-hidden="true"></a>cv_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb201-78"><a href="brca-rnaseq-model-suite.html#cb201-78" aria-hidden="true"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb201-79"><a href="brca-rnaseq-model-suite.html#cb201-79" aria-hidden="true"></a> sizeVal_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb201-80"><a href="brca-rnaseq-model-suite.html#cb201-80" aria-hidden="true"></a> sizeVal_cv_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(cv_min, SimNo))</span>
<span id="cb201-81"><a href="brca-rnaseq-model-suite.html#cb201-81" aria-hidden="true"></a> <span class="kw">sapply</span>(sizeVal_cv_min_lst, median)</span>
<span id="cb201-82"><a href="brca-rnaseq-model-suite.html#cb201-82" aria-hidden="true"></a>})</span>
<span id="cb201-83"><a href="brca-rnaseq-model-suite.html#cb201-83" aria-hidden="true"></a><span class="kw">names</span>(cv_min_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb201-84"><a href="brca-rnaseq-model-suite.html#cb201-84" aria-hidden="true"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_cv&#39;</span>)</span>
<span id="cb201-85"><a href="brca-rnaseq-model-suite.html#cb201-85" aria-hidden="true"></a></span>
<span id="cb201-86"><a href="brca-rnaseq-model-suite.html#cb201-86" aria-hidden="true"></a><span class="co">## test</span></span>
<span id="cb201-87"><a href="brca-rnaseq-model-suite.html#cb201-87" aria-hidden="true"></a>test_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb201-88"><a href="brca-rnaseq-model-suite.html#cb201-88" aria-hidden="true"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb201-89"><a href="brca-rnaseq-model-suite.html#cb201-89" aria-hidden="true"></a> sizeVal_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb201-90"><a href="brca-rnaseq-model-suite.html#cb201-90" aria-hidden="true"></a> sizeVal_test_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(test_min, SimNo))</span>
<span id="cb201-91"><a href="brca-rnaseq-model-suite.html#cb201-91" aria-hidden="true"></a> <span class="kw">sapply</span>(sizeVal_test_min_lst, median)</span>
<span id="cb201-92"><a href="brca-rnaseq-model-suite.html#cb201-92" aria-hidden="true"></a>})</span>
<span id="cb201-93"><a href="brca-rnaseq-model-suite.html#cb201-93" aria-hidden="true"></a><span class="kw">names</span>(test_min_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb201-94"><a href="brca-rnaseq-model-suite.html#cb201-94" aria-hidden="true"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_test&#39;</span>)</span>
<span id="cb201-95"><a href="brca-rnaseq-model-suite.html#cb201-95" aria-hidden="true"></a></span>
<span id="cb201-96"><a href="brca-rnaseq-model-suite.html#cb201-96" aria-hidden="true"></a></span>
<span id="cb201-97"><a href="brca-rnaseq-model-suite.html#cb201-97" aria-hidden="true"></a>error_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">c</span>(cv_min_Bysize_lst, test_min_Bysize_lst)</span>
<span id="cb201-98"><a href="brca-rnaseq-model-suite.html#cb201-98" aria-hidden="true"></a>error_min_Bysize_lst &lt;-<span class="st"> </span>error_min_Bysize_lst[<span class="kw">order</span>(<span class="kw">names</span>(error_min_Bysize_lst))]</span>
<span id="cb201-99"><a href="brca-rnaseq-model-suite.html#cb201-99" aria-hidden="true"></a></span>
<span id="cb201-100"><a href="brca-rnaseq-model-suite.html#cb201-100" aria-hidden="true"></a><span class="kw">boxplot</span>(error_min_Bysize_lst,</span>
<span id="cb201-101"><a href="brca-rnaseq-model-suite.html#cb201-101" aria-hidden="true"></a>  <span class="dt">col=</span><span class="dv">0</span>,</span>
<span id="cb201-102"><a href="brca-rnaseq-model-suite.html#cb201-102" aria-hidden="true"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb201-103"><a href="brca-rnaseq-model-suite.html#cb201-103" aria-hidden="true"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">.5</span>),</span>
<span id="cb201-104"><a href="brca-rnaseq-model-suite.html#cb201-104" aria-hidden="true"></a>  <span class="dt">outline=</span>F,</span>
<span id="cb201-105"><a href="brca-rnaseq-model-suite.html#cb201-105" aria-hidden="true"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb201-106"><a href="brca-rnaseq-model-suite.html#cb201-106" aria-hidden="true"></a>)</span>
<span id="cb201-107"><a href="brca-rnaseq-model-suite.html#cb201-107" aria-hidden="true"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(error_min_Bysize_lst))</span>
<span id="cb201-108"><a href="brca-rnaseq-model-suite.html#cb201-108" aria-hidden="true"></a><span class="kw">points</span>(</span>
<span id="cb201-109"><a href="brca-rnaseq-model-suite.html#cb201-109" aria-hidden="true"></a>   <span class="dt">x=</span><span class="kw">jitter</span>(<span class="kw">rep</span>(JJ, <span class="kw">length</span>(error_min_Bysize_lst[[JJ]])), <span class="dt">amount=</span><span class="fl">0.25</span>), </span>
<span id="cb201-110"><a href="brca-rnaseq-model-suite.html#cb201-110" aria-hidden="true"></a>   <span class="dt">y=</span>error_min_Bysize_lst[[JJ]], <span class="dt">cex=</span><span class="fl">0.5</span>,</span>
<span id="cb201-111"><a href="brca-rnaseq-model-suite.html#cb201-111" aria-hidden="true"></a>   <span class="dt">col=</span><span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">&#39;cv&#39;</span>, <span class="kw">names</span>(error_min_Bysize_lst)[JJ]),<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb201-112"><a href="brca-rnaseq-model-suite.html#cb201-112" aria-hidden="true"></a>)</span>
<span id="cb201-113"><a href="brca-rnaseq-model-suite.html#cb201-113" aria-hidden="true"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb201-114"><a href="brca-rnaseq-model-suite.html#cb201-114" aria-hidden="true"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb201-115"><a href="brca-rnaseq-model-suite.html#cb201-115" aria-hidden="true"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_Bysize_lst)),</span>
<span id="cb201-116"><a href="brca-rnaseq-model-suite.html#cb201-116" aria-hidden="true"></a>  SIZE0</span>
<span id="cb201-117"><a href="brca-rnaseq-model-suite.html#cb201-117" aria-hidden="true"></a> )</span>
<span id="cb201-118"><a href="brca-rnaseq-model-suite.html#cb201-118" aria-hidden="true"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_cv&#39;</span>),<span class="kw">names</span>(error_min_Bysize_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb201-119"><a href="brca-rnaseq-model-suite.html#cb201-119" aria-hidden="true"></a><span class="kw">abline</span>(<span class="dt">h=</span> error_min_q2, <span class="dt">col =</span> <span class="st">&#39;red&#39;</span>)</span>
<span id="cb201-120"><a href="brca-rnaseq-model-suite.html#cb201-120" aria-hidden="true"></a><span class="kw">legend</span>(<span class="st">&#39;topright&#39;</span>,</span>
<span id="cb201-121"><a href="brca-rnaseq-model-suite.html#cb201-121" aria-hidden="true"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb201-122"><a href="brca-rnaseq-model-suite.html#cb201-122" aria-hidden="true"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb201-123"><a href="brca-rnaseq-model-suite.html#cb201-123" aria-hidden="true"></a>   <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&#39;cv error&#39;</span>, <span class="st">&#39;test set&#39;</span>),</span>
<span id="cb201-124"><a href="brca-rnaseq-model-suite.html#cb201-124" aria-hidden="true"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb201-125"><a href="brca-rnaseq-model-suite.html#cb201-125" aria-hidden="true"></a> )</span>
<span id="cb201-126"><a href="brca-rnaseq-model-suite.html#cb201-126" aria-hidden="true"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;min lambda models&#39;</span>))</span>
<span id="cb201-127"><a href="brca-rnaseq-model-suite.html#cb201-127" aria-hidden="true"></a></span>
<span id="cb201-128"><a href="brca-rnaseq-model-suite.html#cb201-128" aria-hidden="true"></a></span>
<span id="cb201-129"><a href="brca-rnaseq-model-suite.html#cb201-129" aria-hidden="true"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, <span class="kw">paste</span>(<span class="st">&#39;lasso fit error rates summarized across simulations&#39;</span>))</span></code></pre></div>
<div class="figure"><span id="fig:brca-rnaseq-lasso-simRes-errors-overSim"></span>
<img src="Static/figures/brca-rnaseq-lasso-simRes-errors-overSim-1.png" alt="lasso Model Errors by Sample Size" width="960" />
<p class="caption">
Figure 8.10: lasso Model Errors by Sample Size
</p>
</div>
<p><br/></p>
<p>As in the case of the analysis of the HCC 5hmC data, we see that at lower
samples sizes the assessed performance is quite variable.
In this example cohorts of N=25 LumA and Other subtype samples
can yield classification error estimates ranging from less than %10
to over 50%.</p>
<p><br/></p>
<p>To appreciate how much variability can be encountered as samples are accrued over time
we need to look at a typical path the assessed model accuracy estimates might take.</p>
<div class="sourceCode" id="cb202"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb202-1"><a href="brca-rnaseq-model-suite.html#cb202-1" aria-hidden="true"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb202-2"><a href="brca-rnaseq-model-suite.html#cb202-2" aria-hidden="true"></a></span>
<span id="cb202-3"><a href="brca-rnaseq-model-suite.html#cb202-3" aria-hidden="true"></a>error_1se_Bysize_mtx &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="st">&#39;cbind&#39;</span>, <span class="kw">lapply</span>(error_1se_Bysize_lst, <span class="cf">function</span>(LL) LL))</span>
<span id="cb202-4"><a href="brca-rnaseq-model-suite.html#cb202-4" aria-hidden="true"></a></span>
<span id="cb202-5"><a href="brca-rnaseq-model-suite.html#cb202-5" aria-hidden="true"></a>cv_error_1se_Bysize_mtx &lt;-<span class="st"> </span>error_1se_Bysize_mtx[,<span class="kw">grep</span>(<span class="st">&#39;_cv&#39;</span>, <span class="kw">colnames</span>(error_1se_Bysize_mtx))]</span>
<span id="cb202-6"><a href="brca-rnaseq-model-suite.html#cb202-6" aria-hidden="true"></a></span>
<span id="cb202-7"><a href="brca-rnaseq-model-suite.html#cb202-7" aria-hidden="true"></a><span class="kw">plot</span>(<span class="dt">x=</span><span class="kw">c</span>(<span class="dv">1</span>, <span class="kw">ncol</span>(cv_error_1se_Bysize_mtx)), <span class="dt">y=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.6</span>), </span>
<span id="cb202-8"><a href="brca-rnaseq-model-suite.html#cb202-8" aria-hidden="true"></a>  <span class="dt">xlab=</span><span class="st">&#39;sample size&#39;</span>, <span class="dt">ylab=</span><span class="st">&#39;Misclassification Error&#39;</span>, </span>
<span id="cb202-9"><a href="brca-rnaseq-model-suite.html#cb202-9" aria-hidden="true"></a>  <span class="dt">type=</span><span class="st">&#39;n&#39;</span>, <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>)</span>
<span id="cb202-10"><a href="brca-rnaseq-model-suite.html#cb202-10" aria-hidden="true"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(cv_error_1se_Bysize_mtx),</span>
<span id="cb202-11"><a href="brca-rnaseq-model-suite.html#cb202-11" aria-hidden="true"></a> <span class="dt">labels=</span><span class="kw">sub</span>(<span class="st">&#39;_cv&#39;</span>,<span class="st">&#39;&#39;</span>,<span class="kw">colnames</span>(cv_error_1se_Bysize_mtx)))</span>
<span id="cb202-12"><a href="brca-rnaseq-model-suite.html#cb202-12" aria-hidden="true"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">15</span>)</span>
<span id="cb202-13"><a href="brca-rnaseq-model-suite.html#cb202-13" aria-hidden="true"></a><span class="kw">lines</span>(<span class="dt">x=</span><span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(cv_error_1se_Bysize_mtx), <span class="dt">y=</span>cv_error_1se_Bysize_mtx[JJ,],</span>
<span id="cb202-14"><a href="brca-rnaseq-model-suite.html#cb202-14" aria-hidden="true"></a> <span class="dt">type=</span><span class="st">&#39;b&#39;</span>, <span class="dt">pch=</span>JJ, <span class="dt">col=</span>JJ)</span>
<span id="cb202-15"><a href="brca-rnaseq-model-suite.html#cb202-15" aria-hidden="true"></a><span class="kw">title</span>(<span class="st">&#39;Example Misclassification Error Paths&#39;</span>)</span></code></pre></div>
<div class="figure"><span id="fig:brca-glmnetSuite-lasso-simRes-errorsPath-overSim"></span>
<img src="Static/figures/brca-glmnetSuite-lasso-simRes-errorsPath-overSim-1.png" alt="lasso Model Error Paths" width="960" />
<p class="caption">
Figure 8.11: lasso Model Error Paths
</p>
</div>
<p>We see how erratic the assessed model accuracy can be when sample sizes are small,
and that it would be hard to guess the ultimate level of accuracy the
is achievable, or the number of samples required to get a reasonable
estimate of the achievable level of accuracy.</p>
<p><br/></p>
<!--
* For the smaller samples sizes, cv error rates for the minimum lambda models tend to be
optimistic.  

* For lower sample sizes, assess performance is quite variable.
-->
</div>
<div id="feature-selection-3" class="section level4" number="8.6.2.2">
<h4><span class="header-section-number">8.6.2.2</span> Feature Selection</h4>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb203-1"><a href="brca-rnaseq-model-suite.html#cb203-1" aria-hidden="true"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb203-2"><a href="brca-rnaseq-model-suite.html#cb203-2" aria-hidden="true"></a></span>
<span id="cb203-3"><a href="brca-rnaseq-model-suite.html#cb203-3" aria-hidden="true"></a><span class="co"># Utility objects</span></span>
<span id="cb203-4"><a href="brca-rnaseq-model-suite.html#cb203-4" aria-hidden="true"></a>SIZE0 &lt;-<span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_pad</span>(SIZE, <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>)</span>
<span id="cb203-5"><a href="brca-rnaseq-model-suite.html#cb203-5" aria-hidden="true"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(brcaRna_sim_other_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>,SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb203-6"><a href="brca-rnaseq-model-suite.html#cb203-6" aria-hidden="true"></a></span>
<span id="cb203-7"><a href="brca-rnaseq-model-suite.html#cb203-7" aria-hidden="true"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb203-8"><a href="brca-rnaseq-model-suite.html#cb203-8" aria-hidden="true"></a><span class="co"># 1se</span></span>
<span id="cb203-9"><a href="brca-rnaseq-model-suite.html#cb203-9" aria-hidden="true"></a><span class="co">#########################################</span></span>
<span id="cb203-10"><a href="brca-rnaseq-model-suite.html#cb203-10" aria-hidden="true"></a><span class="co"># selected features</span></span>
<span id="cb203-11"><a href="brca-rnaseq-model-suite.html#cb203-11" aria-hidden="true"></a>p_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb203-12"><a href="brca-rnaseq-model-suite.html#cb203-12" aria-hidden="true"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb203-13"><a href="brca-rnaseq-model-suite.html#cb203-13" aria-hidden="true"></a> sizeVal_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb203-14"><a href="brca-rnaseq-model-suite.html#cb203-14" aria-hidden="true"></a> sizeVal_p_1se_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(p_1se, SimNo))</span>
<span id="cb203-15"><a href="brca-rnaseq-model-suite.html#cb203-15" aria-hidden="true"></a> <span class="kw">sapply</span>(sizeVal_p_1se_lst, median)</span>
<span id="cb203-16"><a href="brca-rnaseq-model-suite.html#cb203-16" aria-hidden="true"></a>})</span>
<span id="cb203-17"><a href="brca-rnaseq-model-suite.html#cb203-17" aria-hidden="true"></a><span class="kw">names</span>(p_1se_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb203-18"><a href="brca-rnaseq-model-suite.html#cb203-18" aria-hidden="true"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_p&#39;</span>)</span>
<span id="cb203-19"><a href="brca-rnaseq-model-suite.html#cb203-19" aria-hidden="true"></a></span>
<span id="cb203-20"><a href="brca-rnaseq-model-suite.html#cb203-20" aria-hidden="true"></a><span class="co"># selected signatue features</span></span>
<span id="cb203-21"><a href="brca-rnaseq-model-suite.html#cb203-21" aria-hidden="true"></a>sign_p_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb203-22"><a href="brca-rnaseq-model-suite.html#cb203-22" aria-hidden="true"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb203-23"><a href="brca-rnaseq-model-suite.html#cb203-23" aria-hidden="true"></a> sizeVal_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb203-24"><a href="brca-rnaseq-model-suite.html#cb203-24" aria-hidden="true"></a> </span>
<span id="cb203-25"><a href="brca-rnaseq-model-suite.html#cb203-25" aria-hidden="true"></a>  </span>
<span id="cb203-26"><a href="brca-rnaseq-model-suite.html#cb203-26" aria-hidden="true"></a> sizeVal_sign_genes_1se_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sizeVal_results_frm), <span class="cf">function</span>(RR)</span>
<span id="cb203-27"><a href="brca-rnaseq-model-suite.html#cb203-27" aria-hidden="true"></a>    <span class="kw">intersect</span>(<span class="kw">unlist</span>(sizeVal_results_frm[RR, <span class="st">&#39;genes_1se&#39;</span>]), lasso_gene_sign_1se_vec))</span>
<span id="cb203-28"><a href="brca-rnaseq-model-suite.html#cb203-28" aria-hidden="true"></a></span>
<span id="cb203-29"><a href="brca-rnaseq-model-suite.html#cb203-29" aria-hidden="true"></a> sizeVal_sign_p_1se_lst &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="kw">sapply</span>(sizeVal_sign_genes_1se_lst, length),</span>
<span id="cb203-30"><a href="brca-rnaseq-model-suite.html#cb203-30" aria-hidden="true"></a>    sizeVal_results_frm<span class="op">$</span>SimNo)</span>
<span id="cb203-31"><a href="brca-rnaseq-model-suite.html#cb203-31" aria-hidden="true"></a> </span>
<span id="cb203-32"><a href="brca-rnaseq-model-suite.html#cb203-32" aria-hidden="true"></a> <span class="kw">sapply</span>(sizeVal_sign_p_1se_lst, median)</span>
<span id="cb203-33"><a href="brca-rnaseq-model-suite.html#cb203-33" aria-hidden="true"></a>})</span>
<span id="cb203-34"><a href="brca-rnaseq-model-suite.html#cb203-34" aria-hidden="true"></a><span class="kw">names</span>(sign_p_1se_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb203-35"><a href="brca-rnaseq-model-suite.html#cb203-35" aria-hidden="true"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_signP&#39;</span>)</span>
<span id="cb203-36"><a href="brca-rnaseq-model-suite.html#cb203-36" aria-hidden="true"></a></span>
<span id="cb203-37"><a href="brca-rnaseq-model-suite.html#cb203-37" aria-hidden="true"></a></span>
<span id="cb203-38"><a href="brca-rnaseq-model-suite.html#cb203-38" aria-hidden="true"></a>p_singP_1se_Bysize_lst &lt;-<span class="st"> </span><span class="kw">c</span>(p_1se_Bysize_lst, sign_p_1se_Bysize_lst)</span>
<span id="cb203-39"><a href="brca-rnaseq-model-suite.html#cb203-39" aria-hidden="true"></a>p_singP_1se_Bysize_lst &lt;-<span class="st"> </span>p_singP_1se_Bysize_lst[<span class="kw">order</span>(<span class="kw">names</span>(p_singP_1se_Bysize_lst))]</span>
<span id="cb203-40"><a href="brca-rnaseq-model-suite.html#cb203-40" aria-hidden="true"></a></span>
<span id="cb203-41"><a href="brca-rnaseq-model-suite.html#cb203-41" aria-hidden="true"></a><span class="kw">boxplot</span>(p_singP_1se_Bysize_lst,</span>
<span id="cb203-42"><a href="brca-rnaseq-model-suite.html#cb203-42" aria-hidden="true"></a>  <span class="dt">col=</span><span class="dv">0</span>,</span>
<span id="cb203-43"><a href="brca-rnaseq-model-suite.html#cb203-43" aria-hidden="true"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb203-44"><a href="brca-rnaseq-model-suite.html#cb203-44" aria-hidden="true"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">300</span>),</span>
<span id="cb203-45"><a href="brca-rnaseq-model-suite.html#cb203-45" aria-hidden="true"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb203-46"><a href="brca-rnaseq-model-suite.html#cb203-46" aria-hidden="true"></a>)</span>
<span id="cb203-47"><a href="brca-rnaseq-model-suite.html#cb203-47" aria-hidden="true"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">2</span>, <span class="dt">outer=</span>T,  <span class="st">&#39;number of selected features&#39;</span>)</span>
<span id="cb203-48"><a href="brca-rnaseq-model-suite.html#cb203-48" aria-hidden="true"></a></span>
<span id="cb203-49"><a href="brca-rnaseq-model-suite.html#cb203-49" aria-hidden="true"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(p_singP_1se_Bysize_lst))</span>
<span id="cb203-50"><a href="brca-rnaseq-model-suite.html#cb203-50" aria-hidden="true"></a><span class="kw">points</span>(</span>
<span id="cb203-51"><a href="brca-rnaseq-model-suite.html#cb203-51" aria-hidden="true"></a>   <span class="dt">x=</span><span class="kw">jitter</span>(<span class="kw">rep</span>(JJ, <span class="kw">length</span>(p_singP_1se_Bysize_lst[[JJ]])), <span class="dt">amount=</span><span class="fl">0.25</span>),</span>
<span id="cb203-52"><a href="brca-rnaseq-model-suite.html#cb203-52" aria-hidden="true"></a>   <span class="dt">y=</span>p_singP_1se_Bysize_lst[[JJ]], <span class="dt">cex=</span><span class="fl">0.5</span>,</span>
<span id="cb203-53"><a href="brca-rnaseq-model-suite.html#cb203-53" aria-hidden="true"></a>   <span class="dt">col=</span><span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">&#39;_p&#39;</span>, <span class="kw">names</span>(p_singP_1se_Bysize_lst)[JJ]),<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb203-54"><a href="brca-rnaseq-model-suite.html#cb203-54" aria-hidden="true"></a>)</span>
<span id="cb203-55"><a href="brca-rnaseq-model-suite.html#cb203-55" aria-hidden="true"></a></span>
<span id="cb203-56"><a href="brca-rnaseq-model-suite.html#cb203-56" aria-hidden="true"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb203-57"><a href="brca-rnaseq-model-suite.html#cb203-57" aria-hidden="true"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb203-58"><a href="brca-rnaseq-model-suite.html#cb203-58" aria-hidden="true"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_Bysize_lst)),</span>
<span id="cb203-59"><a href="brca-rnaseq-model-suite.html#cb203-59" aria-hidden="true"></a>  SIZE0</span>
<span id="cb203-60"><a href="brca-rnaseq-model-suite.html#cb203-60" aria-hidden="true"></a> )</span>
<span id="cb203-61"><a href="brca-rnaseq-model-suite.html#cb203-61" aria-hidden="true"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_1se_Bysize_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb203-62"><a href="brca-rnaseq-model-suite.html#cb203-62" aria-hidden="true"></a><span class="co">#abline(h= nzero_1se_q2, col = &#39;red&#39;)</span></span>
<span id="cb203-63"><a href="brca-rnaseq-model-suite.html#cb203-63" aria-hidden="true"></a><span class="kw">legend</span>(<span class="st">&#39;topleft&#39;</span>,</span>
<span id="cb203-64"><a href="brca-rnaseq-model-suite.html#cb203-64" aria-hidden="true"></a>   <span class="co">#title=&#39;1se errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb203-65"><a href="brca-rnaseq-model-suite.html#cb203-65" aria-hidden="true"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>),</span>
<span id="cb203-66"><a href="brca-rnaseq-model-suite.html#cb203-66" aria-hidden="true"></a>   <span class="dt">legend=</span> <span class="kw">c</span>(<span class="st">&#39;selected genes&#39;</span>,<span class="st">&#39;signature genes&#39;</span>),</span>
<span id="cb203-67"><a href="brca-rnaseq-model-suite.html#cb203-67" aria-hidden="true"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb203-68"><a href="brca-rnaseq-model-suite.html#cb203-68" aria-hidden="true"></a> )</span>
<span id="cb203-69"><a href="brca-rnaseq-model-suite.html#cb203-69" aria-hidden="true"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;one se lamdba models&#39;</span>))</span>
<span id="cb203-70"><a href="brca-rnaseq-model-suite.html#cb203-70" aria-hidden="true"></a></span>
<span id="cb203-71"><a href="brca-rnaseq-model-suite.html#cb203-71" aria-hidden="true"></a></span>
<span id="cb203-72"><a href="brca-rnaseq-model-suite.html#cb203-72" aria-hidden="true"></a><span class="co"># min</span></span>
<span id="cb203-73"><a href="brca-rnaseq-model-suite.html#cb203-73" aria-hidden="true"></a><span class="co">#########################################</span></span>
<span id="cb203-74"><a href="brca-rnaseq-model-suite.html#cb203-74" aria-hidden="true"></a><span class="co"># selected features</span></span>
<span id="cb203-75"><a href="brca-rnaseq-model-suite.html#cb203-75" aria-hidden="true"></a>p_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb203-76"><a href="brca-rnaseq-model-suite.html#cb203-76" aria-hidden="true"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb203-77"><a href="brca-rnaseq-model-suite.html#cb203-77" aria-hidden="true"></a> sizeVal_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb203-78"><a href="brca-rnaseq-model-suite.html#cb203-78" aria-hidden="true"></a> sizeVal_p_min_lst &lt;-<span class="st"> </span><span class="kw">with</span>(sizeVal_results_frm, <span class="kw">split</span>(p_min, SimNo))</span>
<span id="cb203-79"><a href="brca-rnaseq-model-suite.html#cb203-79" aria-hidden="true"></a> <span class="kw">sapply</span>(sizeVal_p_min_lst, median)</span>
<span id="cb203-80"><a href="brca-rnaseq-model-suite.html#cb203-80" aria-hidden="true"></a>})</span>
<span id="cb203-81"><a href="brca-rnaseq-model-suite.html#cb203-81" aria-hidden="true"></a><span class="kw">names</span>(p_min_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb203-82"><a href="brca-rnaseq-model-suite.html#cb203-82" aria-hidden="true"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_p&#39;</span>)</span>
<span id="cb203-83"><a href="brca-rnaseq-model-suite.html#cb203-83" aria-hidden="true"></a></span>
<span id="cb203-84"><a href="brca-rnaseq-model-suite.html#cb203-84" aria-hidden="true"></a><span class="co"># selected signatue features</span></span>
<span id="cb203-85"><a href="brca-rnaseq-model-suite.html#cb203-85" aria-hidden="true"></a>sign_p_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size),</span>
<span id="cb203-86"><a href="brca-rnaseq-model-suite.html#cb203-86" aria-hidden="true"></a><span class="cf">function</span>(SizeVal) {</span>
<span id="cb203-87"><a href="brca-rnaseq-model-suite.html#cb203-87" aria-hidden="true"></a> sizeVal_results_frm &lt;-<span class="st"> </span>brcaRna_lasso_sim_results_frm <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Size<span class="op">==</span>SizeVal)</span>
<span id="cb203-88"><a href="brca-rnaseq-model-suite.html#cb203-88" aria-hidden="true"></a> </span>
<span id="cb203-89"><a href="brca-rnaseq-model-suite.html#cb203-89" aria-hidden="true"></a>  </span>
<span id="cb203-90"><a href="brca-rnaseq-model-suite.html#cb203-90" aria-hidden="true"></a> sizeVal_sign_genes_min_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sizeVal_results_frm), <span class="cf">function</span>(RR)</span>
<span id="cb203-91"><a href="brca-rnaseq-model-suite.html#cb203-91" aria-hidden="true"></a>    <span class="kw">intersect</span>(<span class="kw">unlist</span>(sizeVal_results_frm[RR, <span class="st">&#39;genes_min&#39;</span>]), lasso_gene_sign_min_vec))</span>
<span id="cb203-92"><a href="brca-rnaseq-model-suite.html#cb203-92" aria-hidden="true"></a></span>
<span id="cb203-93"><a href="brca-rnaseq-model-suite.html#cb203-93" aria-hidden="true"></a> sizeVal_sign_p_min_lst &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="kw">sapply</span>(sizeVal_sign_genes_min_lst, length),</span>
<span id="cb203-94"><a href="brca-rnaseq-model-suite.html#cb203-94" aria-hidden="true"></a>    sizeVal_results_frm<span class="op">$</span>SimNo)</span>
<span id="cb203-95"><a href="brca-rnaseq-model-suite.html#cb203-95" aria-hidden="true"></a> </span>
<span id="cb203-96"><a href="brca-rnaseq-model-suite.html#cb203-96" aria-hidden="true"></a> <span class="kw">sapply</span>(sizeVal_sign_p_min_lst, median)</span>
<span id="cb203-97"><a href="brca-rnaseq-model-suite.html#cb203-97" aria-hidden="true"></a>})</span>
<span id="cb203-98"><a href="brca-rnaseq-model-suite.html#cb203-98" aria-hidden="true"></a><span class="kw">names</span>(sign_p_min_Bysize_lst) &lt;-<span class="st"> </span><span class="kw">paste0</span>(</span>
<span id="cb203-99"><a href="brca-rnaseq-model-suite.html#cb203-99" aria-hidden="true"></a> stringr<span class="op">::</span><span class="kw">str_pad</span>(<span class="kw">unique</span>(brcaRna_lasso_sim_results_frm<span class="op">$</span>Size), <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">pad=</span><span class="st">&#39;0&#39;</span>), <span class="st">&#39;_signP&#39;</span>)</span>
<span id="cb203-100"><a href="brca-rnaseq-model-suite.html#cb203-100" aria-hidden="true"></a></span>
<span id="cb203-101"><a href="brca-rnaseq-model-suite.html#cb203-101" aria-hidden="true"></a></span>
<span id="cb203-102"><a href="brca-rnaseq-model-suite.html#cb203-102" aria-hidden="true"></a>p_singP_min_Bysize_lst &lt;-<span class="st"> </span><span class="kw">c</span>(p_min_Bysize_lst, sign_p_min_Bysize_lst)</span>
<span id="cb203-103"><a href="brca-rnaseq-model-suite.html#cb203-103" aria-hidden="true"></a>p_singP_min_Bysize_lst &lt;-<span class="st"> </span>p_singP_min_Bysize_lst[<span class="kw">order</span>(<span class="kw">names</span>(p_singP_min_Bysize_lst))]</span>
<span id="cb203-104"><a href="brca-rnaseq-model-suite.html#cb203-104" aria-hidden="true"></a></span>
<span id="cb203-105"><a href="brca-rnaseq-model-suite.html#cb203-105" aria-hidden="true"></a><span class="kw">boxplot</span>(p_singP_min_Bysize_lst,</span>
<span id="cb203-106"><a href="brca-rnaseq-model-suite.html#cb203-106" aria-hidden="true"></a>  <span class="dt">col=</span><span class="dv">0</span>,</span>
<span id="cb203-107"><a href="brca-rnaseq-model-suite.html#cb203-107" aria-hidden="true"></a>  <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;green&#39;</span>),</span>
<span id="cb203-108"><a href="brca-rnaseq-model-suite.html#cb203-108" aria-hidden="true"></a>  <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">300</span>),</span>
<span id="cb203-109"><a href="brca-rnaseq-model-suite.html#cb203-109" aria-hidden="true"></a>  <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb203-110"><a href="brca-rnaseq-model-suite.html#cb203-110" aria-hidden="true"></a>)</span>
<span id="cb203-111"><a href="brca-rnaseq-model-suite.html#cb203-111" aria-hidden="true"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(p_singP_min_Bysize_lst))</span>
<span id="cb203-112"><a href="brca-rnaseq-model-suite.html#cb203-112" aria-hidden="true"></a><span class="kw">points</span>(</span>
<span id="cb203-113"><a href="brca-rnaseq-model-suite.html#cb203-113" aria-hidden="true"></a>   <span class="dt">x=</span><span class="kw">jitter</span>(<span class="kw">rep</span>(JJ, <span class="kw">length</span>(p_singP_min_Bysize_lst[[JJ]])), <span class="dt">amount=</span><span class="fl">0.25</span>),</span>
<span id="cb203-114"><a href="brca-rnaseq-model-suite.html#cb203-114" aria-hidden="true"></a>   <span class="dt">y=</span>p_singP_min_Bysize_lst[[JJ]], <span class="dt">cex=</span><span class="fl">0.5</span>,</span>
<span id="cb203-115"><a href="brca-rnaseq-model-suite.html#cb203-115" aria-hidden="true"></a>   <span class="dt">col=</span><span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">&#39;_p&#39;</span>, <span class="kw">names</span>(p_singP_min_Bysize_lst)[JJ]),<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>)</span>
<span id="cb203-116"><a href="brca-rnaseq-model-suite.html#cb203-116" aria-hidden="true"></a>)</span>
<span id="cb203-117"><a href="brca-rnaseq-model-suite.html#cb203-117" aria-hidden="true"></a></span>
<span id="cb203-118"><a href="brca-rnaseq-model-suite.html#cb203-118" aria-hidden="true"></a>LL &lt;-<span class="st"> </span><span class="dv">-1</span></span>
<span id="cb203-119"><a href="brca-rnaseq-model-suite.html#cb203-119" aria-hidden="true"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">tick=</span>F, <span class="dt">line =</span> LL,</span>
<span id="cb203-120"><a href="brca-rnaseq-model-suite.html#cb203-120" aria-hidden="true"></a>  <span class="dt">at =</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_Bysize_lst)),</span>
<span id="cb203-121"><a href="brca-rnaseq-model-suite.html#cb203-121" aria-hidden="true"></a>  SIZE0</span>
<span id="cb203-122"><a href="brca-rnaseq-model-suite.html#cb203-122" aria-hidden="true"></a> )</span>
<span id="cb203-123"><a href="brca-rnaseq-model-suite.html#cb203-123" aria-hidden="true"></a><span class="kw">abline</span>(<span class="dt">v=</span> <span class="kw">match</span>(<span class="kw">paste0</span>(SIZE0,<span class="st">&#39;_p&#39;</span>),<span class="kw">names</span>(p_singP_min_Bysize_lst))[<span class="op">-</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="st">&#39;grey&#39;</span>)</span>
<span id="cb203-124"><a href="brca-rnaseq-model-suite.html#cb203-124" aria-hidden="true"></a><span class="co">#abline(h= nzero_min_q2, col = &#39;red&#39;)</span></span>
<span id="cb203-125"><a href="brca-rnaseq-model-suite.html#cb203-125" aria-hidden="true"></a><span class="kw">legend</span>(<span class="st">&#39;topleft&#39;</span>,</span>
<span id="cb203-126"><a href="brca-rnaseq-model-suite.html#cb203-126" aria-hidden="true"></a>   <span class="co">#title=&#39;min errors&#39;, title.col = &#39;black&#39;,</span></span>
<span id="cb203-127"><a href="brca-rnaseq-model-suite.html#cb203-127" aria-hidden="true"></a>   <span class="dt">text.col =</span> <span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>, <span class="st">&#39;green&#39;</span>),</span>
<span id="cb203-128"><a href="brca-rnaseq-model-suite.html#cb203-128" aria-hidden="true"></a>   <span class="dt">legend=</span> <span class="kw">c</span>(<span class="st">&#39;selected genes&#39;</span>,<span class="st">&#39;signature genes&#39;</span>),</span>
<span id="cb203-129"><a href="brca-rnaseq-model-suite.html#cb203-129" aria-hidden="true"></a>   <span class="dt">bty=</span><span class="st">&#39;n&#39;</span></span>
<span id="cb203-130"><a href="brca-rnaseq-model-suite.html#cb203-130" aria-hidden="true"></a> )</span>
<span id="cb203-131"><a href="brca-rnaseq-model-suite.html#cb203-131" aria-hidden="true"></a><span class="kw">title</span>(<span class="kw">paste</span>(<span class="st">&#39;min lambda models&#39;</span>))</span>
<span id="cb203-132"><a href="brca-rnaseq-model-suite.html#cb203-132" aria-hidden="true"></a></span>
<span id="cb203-133"><a href="brca-rnaseq-model-suite.html#cb203-133" aria-hidden="true"></a><span class="kw">mtext</span>(<span class="dt">side=</span><span class="dv">3</span>, <span class="dt">outer=</span>T, <span class="dt">cex=</span><span class="fl">1.25</span>, <span class="kw">paste</span>(<span class="st">&#39;lasso fit feature selection summarized across simulations&#39;</span>))</span></code></pre></div>
<div class="figure"><span id="fig:brca-rnaseq-lasso-simRes-features-OverSim"></span>
<img src="Static/figures/brca-rnaseq-lasso-simRes-features-OverSim-1.png" alt="lasso Models Selected Features by Sample Size" width="960" />
<p class="caption">
Figure 8.12: lasso Models Selected Features by Sample Size
</p>
</div>
<p>Compare with Figures
<a href="hcc-5hmcseq-model-suite.html#fig:hcc5hmC-glmnetSuite-lasso-simRes-errors-overSim">5.10</a>
and
<a href="hcc-5hmcseq-model-suite.html#fig:hcc5hmC-glmnetSuite-lasso-simRes-features-OverSim">5.12</a>
for the HCC 5hmC data.</p>
<!--

* The number of selected features incerase with sample size.  

*  The number of selected features is quite variable, even for larger sample sizes.

* The number of core signature features among selected features is stable for larger sample
sizes and represenst 30 to 40% of the selected features (for N=200 and 300 respectively).

-->
<!-- SKIP THIS
## Effect of sample consistency

**This is a bust ...**

To see the effect of sample consistency on classification results, we will
focus on the performance of the small sample fits where variability is the
greatest, and the context where investigators should be most aware of the
effect of sample selection over and beyond sample size concerns.
-->
<div class="sourceCode" id="cb204"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb204-1"><a href="brca-rnaseq-model-suite.html#cb204-1" aria-hidden="true"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb204-2"><a href="brca-rnaseq-model-suite.html#cb204-2" aria-hidden="true"></a></span>
<span id="cb204-3"><a href="brca-rnaseq-model-suite.html#cb204-3" aria-hidden="true"></a></span>
<span id="cb204-4"><a href="brca-rnaseq-model-suite.html#cb204-4" aria-hidden="true"></a>other_samp25_qual_vec &lt;-<span class="st"> </span><span class="kw">apply</span>(brcaRna_sim_other_qual_mtx[<span class="dv">1</span><span class="op">:</span><span class="dv">25</span>, ], <span class="dv">2</span>, median)</span>
<span id="cb204-5"><a href="brca-rnaseq-model-suite.html#cb204-5" aria-hidden="true"></a>LumA_samp25_qual_vec &lt;-<span class="st"> </span><span class="kw">apply</span>(brcaRna_sim_LumA_qual_mtx[<span class="dv">1</span><span class="op">:</span><span class="dv">25</span>, ], <span class="dv">2</span>, median)</span>
<span id="cb204-6"><a href="brca-rnaseq-model-suite.html#cb204-6" aria-hidden="true"></a></span>
<span id="cb204-7"><a href="brca-rnaseq-model-suite.html#cb204-7" aria-hidden="true"></a>samp25_qual_error_mtx &lt;-<span class="st"> </span><span class="kw">cbind</span>(</span>
<span id="cb204-8"><a href="brca-rnaseq-model-suite.html#cb204-8" aria-hidden="true"></a>   <span class="st">`</span><span class="dt">1se_cv</span><span class="st">`</span> =<span class="st"> </span>error_1se_Bysize_lst[[<span class="st">&#39;025_cv&#39;</span>]],</span>
<span id="cb204-9"><a href="brca-rnaseq-model-suite.html#cb204-9" aria-hidden="true"></a>   <span class="st">`</span><span class="dt">1se_test</span><span class="st">`</span> =<span class="st"> </span>error_1se_Bysize_lst[[<span class="st">&#39;025_test&#39;</span>]],</span>
<span id="cb204-10"><a href="brca-rnaseq-model-suite.html#cb204-10" aria-hidden="true"></a>   <span class="st">`</span><span class="dt">other_Q</span><span class="st">`</span> =<span class="st"> </span>other_samp25_qual_vec,</span>
<span id="cb204-11"><a href="brca-rnaseq-model-suite.html#cb204-11" aria-hidden="true"></a>   <span class="st">`</span><span class="dt">LumA_Q</span><span class="st">`</span> =<span class="st"> </span>LumA_samp25_qual_vec)</span>
<span id="cb204-12"><a href="brca-rnaseq-model-suite.html#cb204-12" aria-hidden="true"></a> </span>
<span id="cb204-13"><a href="brca-rnaseq-model-suite.html#cb204-13" aria-hidden="true"></a></span>
<span id="cb204-14"><a href="brca-rnaseq-model-suite.html#cb204-14" aria-hidden="true"></a><span class="co"># Correlation panel</span></span>
<span id="cb204-15"><a href="brca-rnaseq-model-suite.html#cb204-15" aria-hidden="true"></a>panel.cor &lt;-<span class="st"> </span><span class="cf">function</span>(x, y){</span>
<span id="cb204-16"><a href="brca-rnaseq-model-suite.html#cb204-16" aria-hidden="true"></a>    usr &lt;-<span class="st"> </span><span class="kw">par</span>(<span class="st">&quot;usr&quot;</span>); <span class="kw">on.exit</span>(<span class="kw">par</span>(usr))</span>
<span id="cb204-17"><a href="brca-rnaseq-model-suite.html#cb204-17" aria-hidden="true"></a>    <span class="kw">par</span>(<span class="dt">usr =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>))</span>
<span id="cb204-18"><a href="brca-rnaseq-model-suite.html#cb204-18" aria-hidden="true"></a>    r &lt;-<span class="st"> </span><span class="kw">round</span>(<span class="kw">cor</span>(x, y), <span class="dt">digits=</span><span class="dv">2</span>)</span>
<span id="cb204-19"><a href="brca-rnaseq-model-suite.html#cb204-19" aria-hidden="true"></a>    txt &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;R = &quot;</span>, r)</span>
<span id="cb204-20"><a href="brca-rnaseq-model-suite.html#cb204-20" aria-hidden="true"></a>    cex.cor &lt;-<span class="st"> </span><span class="fl">0.8</span><span class="op">/</span><span class="kw">strwidth</span>(txt)</span>
<span id="cb204-21"><a href="brca-rnaseq-model-suite.html#cb204-21" aria-hidden="true"></a>    <span class="kw">text</span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>, txt, <span class="dt">cex =</span> <span class="fl">1.5</span>) <span class="co">###cex.cor * r)</span></span>
<span id="cb204-22"><a href="brca-rnaseq-model-suite.html#cb204-22" aria-hidden="true"></a>}</span>
<span id="cb204-23"><a href="brca-rnaseq-model-suite.html#cb204-23" aria-hidden="true"></a></span>
<span id="cb204-24"><a href="brca-rnaseq-model-suite.html#cb204-24" aria-hidden="true"></a><span class="kw">pairs</span>(samp25_qual_error_mtx,</span>
<span id="cb204-25"><a href="brca-rnaseq-model-suite.html#cb204-25" aria-hidden="true"></a> <span class="dt">lower.panel =</span> panel.cor)</span></code></pre></div>
<!-- for the HCC data we had:
We see that although the sample consistency effect is not strong, the consistency of the
LumA sample cohorts does have a measurable impact on the 1se cv and test set error
rates (cor = -0.32 and -0.27, respectively).  
-->

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="brca-rnaseseq-explore-sparsity.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="conclusions.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["HCC5hmCExplore.pdf", "HCC5hmCExplore.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
