<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Section 6 BrCa RNA-Seq: Preprocessing | DNA Hydroxymethylation in Hepatocellular Carcinoma</title>
  <meta name="description" content="Data from Cai et al. (2019) paper are explored" />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Section 6 BrCa RNA-Seq: Preprocessing | DNA Hydroxymethylation in Hepatocellular Carcinoma" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Data from Cai et al. (2019) paper are explored" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Section 6 BrCa RNA-Seq: Preprocessing | DNA Hydroxymethylation in Hepatocellular Carcinoma" />
  
  <meta name="twitter:description" content="Data from Cai et al. (2019) paper are explored" />
  

<meta name="author" content="Francois Collin" />


<meta name="date" content="2020-01-01" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="hcc-5hmcseq-model-suite.html"/>
<link rel="next" href="brca-rnaseseq-explore-sparsity.html"/>
<script src="libs/header-attrs-2.2/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<script type="text/javascript">
// source:https://stackoverflow.com/questions/45360998/code-folding-in-bookdown
// toggle visibility of R source blocks in R Markdown output
function toggle_R() {
  var x = document.getElementsByClassName('r');
  if (x.length == 0) return;
  function toggle_vis(o) {
    var d = o.style.display;
    o.style.display = (d == 'block' || d == '') ? 'none':'block';
  }

  for (i = 0; i < x.length; i++) {
    var y = x[i];
    if (y.tagName.toLowerCase() === 'pre') toggle_vis(y);
  }

    var elem = document.getElementById("myButton1");
    if (elem.value === "Hide Global") elem.value = "Show Global";
    else elem.value = "Hide Global";
}

document.write('<input onclick="toggle_R();" type="button" value="Hide Global" id="myButton1" style="position: absolute; top: 10%; right: 2%; z-index: 200"></input>')

</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">DNA Hydroxymethylation in Hepatocellular Carcinoma</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preamble</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i>License</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="modeling-background.html"><a href="modeling-background.html"><i class="fa fa-check"></i><b>2</b> Classification Analysis in the <code>n &lt;&lt; p</code> Context</a>
<ul>
<li class="chapter" data-level="2.1" data-path="modeling-background.html"><a href="modeling-background.html#caret-for-model-evaluation"><i class="fa fa-check"></i><b>2.1</b> caret for model evaluation</a></li>
<li class="chapter" data-level="2.2" data-path="modeling-background.html"><a href="modeling-background.html#the-glmnet-r-package"><i class="fa fa-check"></i><b>2.2</b> The <code>glmnet</code> R package</a>
<ul>
<li><a href="modeling-background.html#alpha-hyper-parameter"><strong>alpha</strong> hyper-parameter</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="modeling-background.html"><a href="modeling-background.html#signal-to-noise"><i class="fa fa-check"></i><b>2.3</b> Signal-to-noise Ratio</a></li>
<li class="chapter" data-level="2.4" data-path="modeling-background.html"><a href="modeling-background.html#lasso-vs-best-sub"><i class="fa fa-check"></i><b>2.4</b> Lasso vs Best Subset</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html"><i class="fa fa-check"></i><b>3</b> HCC 5hmC-Seq : Preprocessing</a>
<ul>
<li class="chapter" data-level="3.1" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#load-the-data"><i class="fa fa-check"></i><b>3.1</b> Load the data</a></li>
<li class="chapter" data-level="3.2" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#dra"><i class="fa fa-check"></i><b>3.2</b> Differential representation analysis</a>
<ul>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#remove-lowly-expressed-genes"><i class="fa fa-check"></i>Remove lowly expressed genes</a></li>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#creating-a-design-matrix-and-contrasts"><i class="fa fa-check"></i>Creating a design matrix and contrasts</a></li>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#removing-heteroscedasticity-from-the-count-data"><i class="fa fa-check"></i>Removing heteroscedasticity from the count data</a></li>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#fit-linear-models-and-examine-the-results"><i class="fa fa-check"></i>Fit linear models and examine the results</a></li>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#graphical-representations-of-de-results-md-plots"><i class="fa fa-check"></i>Graphical representations of DE results: MD Plots</a></li>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#de-genes-at-10-fold-change"><i class="fa fa-check"></i>DE genes at 10% fold change</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#snr-regime"><i class="fa fa-check"></i><b>3.3</b> Signal-to-noise ratio regime</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html"><i class="fa fa-check"></i><b>4</b> HCC 5hmC-Seq: Exploring Sparsity</a>
<ul>
<li class="chapter" data-level="4.1" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#cross-validation-analysis-setup"><i class="fa fa-check"></i><b>4.1</b> Cross-validation analysis setup</a></li>
<li class="chapter" data-level="4.2" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#fit-and-compare-models"><i class="fa fa-check"></i><b>4.2</b> Fit and compare models</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#logistic-regression-in-glmnet"><i class="fa fa-check"></i><b>4.2.1</b> Logistic regression in <code>glmnet</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#the-relaxed-lasso-and-blended-mix-models"><i class="fa fa-check"></i><b>4.3</b> The relaxed lasso and blended mix models</a></li>
<li class="chapter" data-level="4.4" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#examination-of-sensitivity-vs-specificity"><i class="fa fa-check"></i><b>4.4</b> Examination of sensitivity vs specificity</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#training-data-out-of-fold-roc-curves"><i class="fa fa-check"></i><b>4.4.1</b> Training data out-of-fold ROC curves</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#compare-predictions-at-misclassified-samples"><i class="fa fa-check"></i><b>4.5</b> Compare predictions at misclassified samples</a></li>
<li class="chapter" data-level="4.6" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#compare-coefficient-profiles"><i class="fa fa-check"></i><b>4.6</b> Compare coefficient profiles</a></li>
<li class="chapter" data-level="4.7" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#examine-feature-selection"><i class="fa fa-check"></i><b>4.7</b> Examine feature selection</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html"><i class="fa fa-check"></i><b>5</b> HCC 5hmC-Seq: Sample size investigation</a>
<ul>
<li class="chapter" data-level="5.1" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#full-data-set-fit"><i class="fa fa-check"></i><b>5.1</b> Full data set fit</a>
<ul>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#get-sample-consistency-scores"><i class="fa fa-check"></i>Get sample consistency scores</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#selected-feature-list-stability"><i class="fa fa-check"></i><b>5.2</b> Selected feature list stability</a></li>
<li class="chapter" data-level="5.3" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#simulation-design"><i class="fa fa-check"></i><b>5.3</b> Simulation Design</a></li>
<li class="chapter" data-level="5.4" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#setup-simulation"><i class="fa fa-check"></i><b>5.4</b> Setup simulation</a></li>
<li class="chapter" data-level="5.5" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#run-simulations"><i class="fa fa-check"></i><b>5.5</b> Run simulations</a></li>
<li class="chapter" data-level="5.6" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#simulation-results"><i class="fa fa-check"></i><b>5.6</b> Simulation results</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#simulation-results---look-at-one-simulation"><i class="fa fa-check"></i><b>5.6.1</b> Simulation Results - look at one simulation</a></li>
<li class="chapter" data-level="5.6.2" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#summarize-results-across-simulation-runs."><i class="fa fa-check"></i><b>5.6.2</b> Summarize results across simulation runs.</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#effect-of-sample-consistency"><i class="fa fa-check"></i><b>5.7</b> Effect of sample consistency</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html"><i class="fa fa-check"></i><b>6</b> BrCa RNA-Seq: Preprocessing</a>
<ul>
<li class="chapter" data-level="6.1" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#load-the-data-1"><i class="fa fa-check"></i><b>6.1</b> Load the data</a></li>
<li class="chapter" data-level="6.2" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#breast-rnaseq-dra"><i class="fa fa-check"></i><b>6.2</b> Differential representation analysis</a>
<ul>
<li class="chapter" data-level="" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#remove-lowly-expressed-genes-1"><i class="fa fa-check"></i>Remove lowly expressed genes</a></li>
<li class="chapter" data-level="" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#creating-a-design-matrix-and-contrasts-1"><i class="fa fa-check"></i>Creating a design matrix and contrasts</a></li>
<li class="chapter" data-level="" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#fit-linear-models-and-examine-the-results-1"><i class="fa fa-check"></i>Fit linear models and examine the results</a></li>
<li class="chapter" data-level="" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#graphical-representations-of-de-results-md-plots-1"><i class="fa fa-check"></i>Graphical representations of DE results: MD Plots</a></li>
<li class="chapter" data-level="" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#de-genes-at-10-fold-change-1"><i class="fa fa-check"></i>DE genes at 10% fold change</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#breast-rnaseq-snr-regime"><i class="fa fa-check"></i><b>6.3</b> Signal-to-noise ratio regime</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html"><i class="fa fa-check"></i><b>7</b> BrCa RNA-Seq: Exploring Sparsity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#cross-validation-analysis-setup-1"><i class="fa fa-check"></i><b>7.1</b> Cross-validation analysis setup</a></li>
<li class="chapter" data-level="7.2" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#fit-and-compare-models-1"><i class="fa fa-check"></i><b>7.2</b> Fit and compare models</a></li>
<li class="chapter" data-level="7.3" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#the-relaxed-lasso-and-blended-mix-models-1"><i class="fa fa-check"></i><b>7.3</b> The relaxed lasso and blended mix models</a></li>
<li class="chapter" data-level="7.4" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#examination-of-sensitivity-vs-specificity-1"><i class="fa fa-check"></i><b>7.4</b> Examination of sensitivity vs specificity</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#training-data-out-of-fold-roc-curves-1"><i class="fa fa-check"></i><b>7.4.1</b> Training data out-of-fold ROC curves</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#compare-predictions-at-misclassified-samples-1"><i class="fa fa-check"></i><b>7.5</b> Compare predictions at misclassified samples</a></li>
<li class="chapter" data-level="7.6" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#compare-coefficient-profiles-1"><i class="fa fa-check"></i><b>7.6</b> Compare coefficient profiles</a></li>
<li class="chapter" data-level="7.7" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#examine-feature-selection-1"><i class="fa fa-check"></i><b>7.7</b> Examine feature selection</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html"><i class="fa fa-check"></i><b>8</b> BrCa RNA-Seq: Sample size investigation</a>
<ul>
<li class="chapter" data-level="8.1" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#full-data-set-fit-1"><i class="fa fa-check"></i><b>8.1</b> Full data set fit</a>
<ul>
<li class="chapter" data-level="" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#get-sample-consistency-scores-1"><i class="fa fa-check"></i>Get sample consistency scores</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#selected-feature-list-stability-1"><i class="fa fa-check"></i><b>8.2</b> Selected feature list stability</a></li>
<li class="chapter" data-level="8.3" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#simulation-design-1"><i class="fa fa-check"></i><b>8.3</b> Simulation Design</a></li>
<li class="chapter" data-level="8.4" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#setup-simulation-1"><i class="fa fa-check"></i><b>8.4</b> Setup simulation</a></li>
<li class="chapter" data-level="8.5" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#run-simulations-1"><i class="fa fa-check"></i><b>8.5</b> Run simulations</a></li>
<li class="chapter" data-level="8.6" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#simulation-results-1"><i class="fa fa-check"></i><b>8.6</b> Simulation results</a>
<ul>
<li class="chapter" data-level="8.6.1" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#simulation-results---look-at-one-simulation-1"><i class="fa fa-check"></i><b>8.6.1</b> Simulation Results - look at one simulation</a></li>
<li class="chapter" data-level="8.6.2" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#summarize-results-across-simulation-runs.-1"><i class="fa fa-check"></i><b>8.6.2</b> Summarize results across simulation runs.</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="conclusions.html"><a href="conclusions.html"><i class="fa fa-check"></i><b>9</b> Conclusions</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html"><i class="fa fa-check"></i>Appendix 1 - Sample size in elastic net fits to HCC 5hmC-Seq Data</a>
<ul>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html#selected-feature-list-stability-2"><i class="fa fa-check"></i>Selected feature list stability</a></li>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html#run-simulations---enet"><i class="fa fa-check"></i>Run simulations - enet</a></li>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html#enet-simulation-results"><i class="fa fa-check"></i>enet Simulation results</a>
<ul>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html#simulation-results---look-at-one-simulation-2"><i class="fa fa-check"></i>Simulation Results - look at one simulation</a></li>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html#summarize-results-across-simulation-runs"><i class="fa fa-check"></i>Summarize results across simulation runs</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="LICENSE.md" style="font:em;" target="blank">Copyright (c) 2020 Francois Collin</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">DNA Hydroxymethylation in Hepatocellular Carcinoma</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="brca-rnaseq-preproc" class="section level1" number="6">
<h1><span class="header-section-number">Section 6</span> BrCa RNA-Seq: Preprocessing</h1>
<p>In this section we examine some RNA-Seq data from a breast cancer study
to provide a different SNR context from the HCC 5hmC data in which
to examine the relationship between sample size and model performance.
For this exercise, we will look at discriminating between breast
cancer samples of subtype LumA and all other subtypes among ER+/HER2-
breast cancer samples. This may
be a somewhat artificial and uninteresting classification problem,
but it does provide a binary classification problem with many examples
in the two subgroups allowing is to run simulations analogous to
those that were run on the HCC 5hmC data.</p>
<!--
Reader note:  

* The most interesting part of this section is Subsection \@ref(snr-regime)
which tells us where on the SNR spectrum the 5hmC data lies.
In view of the work by  Hastie et al. (2017) [@Hastie:2017aa]
summarized in Subsection \@ref(lasso-vs-best-sub), we would expect to
lasso model, and the relaxed version, to do very well with this
classifiaction problem.
-->
<!--

 FN <- 'tmp'
 # Shotcuts for knitting and redering while in R session (Invoke interactive R from R/Scripts folder)
 kk <- function(n='') knitr::knit2html(paste("t", n, sep=''), envir=globalenv(),
       output=paste(FN,".html", sep=''))

 rr <- function(n='') rmarkdown::render(paste("t", n, sep=''), envir=globalenv(),
       output_file=paste(FN,".html", sep='')) ##, output_dir='Scripts')

 bb <- function(n='') browseURL(paste(FN,".html", sep=''))

 # The usual shotcuts
 zz <- function(n='') source(paste("t", n, sep=''))

-->
<div id="load-the-data-1" class="section level2" number="6.1">
<h2><span class="header-section-number">6.1</span> Load the data</h2>
<!-- THIS ENSURES NO EVALUATION TAKES PLACE BY DEFAULT -->
<!-- TO TURN ON, SET eval=T                            -->
<!-- Add base libraries -->
<p>The data that are available from NCBI GEO
<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE96058">Series GSE96058</a>
can be conveniently accessed through an R data package.
Attaching the GSE96058 package makes the count data tables
available as well as a gene annotation table and a sample description table.
See <a href="https://12379monty.github.io/GSE96058/">GSE96058 R Data Package page</a>.</p>
<p>The data in GSE96058 have been separated into various data frames and
matrices. The separation of the gene expression data into separate matrices
was designed to get around github’s file size restrictions. See
<a href="https://github.com/12379Monty/GSE96058">GSE96058 Package Page</a>
for details.</p>
<!--
**Programming Note**:  

* When the GSE96058 package is attached with the `library(GSE96058)`
function call, any previously loaded object with the same
name will mask the corresponding objects lazy-loaded by GSE96058.
To get around this problem, R data packages should have
unique object names.

-->
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="brca-rnaseq-preproc.html#cb109-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb109-2"><a href="brca-rnaseq-preproc.html#cb109-2"></a><span class="cf">if</span> (<span class="op">!</span>(<span class="st">&quot;GSE96058&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()))) {</span>
<span id="cb109-3"><a href="brca-rnaseq-preproc.html#cb109-3"></a>  <span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;devtools&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)) {</span>
<span id="cb109-4"><a href="brca-rnaseq-preproc.html#cb109-4"></a>    <span class="kw">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span>
<span id="cb109-5"><a href="brca-rnaseq-preproc.html#cb109-5"></a>  }</span>
<span id="cb109-6"><a href="brca-rnaseq-preproc.html#cb109-6"></a>  devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;12379Monty/GSE96058&quot;</span>)</span>
<span id="cb109-7"><a href="brca-rnaseq-preproc.html#cb109-7"></a>}</span>
<span id="cb109-8"><a href="brca-rnaseq-preproc.html#cb109-8"></a></span>
<span id="cb109-9"><a href="brca-rnaseq-preproc.html#cb109-9"></a><span class="kw">library</span>(GSE96058)</span>
<span id="cb109-10"><a href="brca-rnaseq-preproc.html#cb109-10"></a></span>
<span id="cb109-11"><a href="brca-rnaseq-preproc.html#cb109-11"></a><span class="co"># Strip GSE_ID prefix form object names</span></span>
<span id="cb109-12"><a href="brca-rnaseq-preproc.html#cb109-12"></a><span class="co">### MOD: (2020.09.21) keep &#39;brcaRna_&#39; as prefix to object names.</span></span>
<span id="cb109-13"><a href="brca-rnaseq-preproc.html#cb109-13"></a><span class="co">###     Otherwise we run into problems when different datasets are</span></span>
<span id="cb109-14"><a href="brca-rnaseq-preproc.html#cb109-14"></a><span class="co">###     analyzed</span></span>
<span id="cb109-15"><a href="brca-rnaseq-preproc.html#cb109-15"></a></span>
<span id="cb109-16"><a href="brca-rnaseq-preproc.html#cb109-16"></a><span class="cf">for</span>(OBJ <span class="cf">in</span> <span class="kw">data</span>(<span class="dt">package=</span><span class="st">&#39;GSE96058&#39;</span>)<span class="op">$</span>results[, <span class="st">&#39;Item&#39;</span>])</span>
<span id="cb109-17"><a href="brca-rnaseq-preproc.html#cb109-17"></a><span class="kw">assign</span>(<span class="kw">sub</span>(<span class="st">&#39;GSE96058_&#39;</span>,<span class="st">&#39;brcaRna_&#39;</span>,OBJ), <span class="kw">get</span>(OBJ))</span>
<span id="cb109-18"><a href="brca-rnaseq-preproc.html#cb109-18"></a></span>
<span id="cb109-19"><a href="brca-rnaseq-preproc.html#cb109-19"></a><span class="kw">detach</span>(package<span class="op">:</span>GSE96058, <span class="dt">unload =</span> T )</span></code></pre></div>
<p>For this analysis, we will consider ER+/HER2- samples only and
examine <code>pam50_subtype = 'LumA' vs 'Other'</code> as outcome.
This choice is predicated on the desire to have a fairly
balanced binary outcome to work with when we examine the effect
of sample size on classification performance.
The outcome in this dataset will be aliased as <code>group</code>.</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="brca-rnaseq-preproc.html#cb110-1"></a>ER_HER2_tbl &lt;-<span class="st"> </span><span class="kw">with</span>(brcaRna_sampDesc <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb110-2"><a href="brca-rnaseq-preproc.html#cb110-2"></a><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">grepl</span>(<span class="st">&#39;repl$&#39;</span>, title) <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span>isRepl <span class="op">&amp;</span><span class="st"> </span>er_Status<span class="op">!=</span><span class="st">&#39;NA&#39;</span>   <span class="op">&amp;</span><span class="st"> </span>her2_Status<span class="op">!=</span><span class="st">&#39;NA&#39;</span>),</span>
<span id="cb110-3"><a href="brca-rnaseq-preproc.html#cb110-3"></a> <span class="kw">table</span>(<span class="dt">ER_HER2=</span><span class="kw">paste</span>(er_Status, her2_Status, <span class="dt">sep=</span><span class="st">&#39;_&#39;</span>),</span>
<span id="cb110-4"><a href="brca-rnaseq-preproc.html#cb110-4"></a>  <span class="dt">exclude=</span><span class="ot">NULL</span>)</span>
<span id="cb110-5"><a href="brca-rnaseq-preproc.html#cb110-5"></a> )</span>
<span id="cb110-6"><a href="brca-rnaseq-preproc.html#cb110-6"></a></span>
<span id="cb110-7"><a href="brca-rnaseq-preproc.html#cb110-7"></a> knitr<span class="op">::</span><span class="kw">kable</span>(ER_HER2_tbl,</span>
<span id="cb110-8"><a href="brca-rnaseq-preproc.html#cb110-8"></a>  <span class="dt">caption=</span><span class="st">&quot;ER, PR and HER2 Status&quot;</span>)  <span class="op">%&gt;%</span></span>
<span id="cb110-9"><a href="brca-rnaseq-preproc.html#cb110-9"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:brcaRna-preproc-subsetSamples">Table 6.1: </span>ER, PR and HER2 Status
</caption>
<thead>
<tr>
<th style="text-align:left;">
ER_HER2
</th>
<th style="text-align:right;">
Freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
0_0
</td>
<td style="text-align:right;">
165
</td>
</tr>
<tr>
<td style="text-align:left;">
0_1
</td>
<td style="text-align:right;">
63
</td>
</tr>
<tr>
<td style="text-align:left;">
1_0
</td>
<td style="text-align:right;">
2425
</td>
</tr>
<tr>
<td style="text-align:left;">
1_1
</td>
<td style="text-align:right;">
310
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="brca-rnaseq-preproc.html#cb111-1"></a><span class="co"># Subset analysis samples</span></span>
<span id="cb111-2"><a href="brca-rnaseq-preproc.html#cb111-2"></a>brcaRna_sampDesc &lt;-</span>
<span id="cb111-3"><a href="brca-rnaseq-preproc.html#cb111-3"></a><span class="st">  </span>brcaRna_sampDesc <span class="op">%&gt;%</span></span>
<span id="cb111-4"><a href="brca-rnaseq-preproc.html#cb111-4"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">grepl</span>(<span class="st">&#39;repl$&#39;</span>, title) <span class="op">&amp;</span><span class="st"> </span>er_Status <span class="op">==</span><span class="st"> &#39;1&#39;</span> <span class="op">&amp;</span><span class="st"> </span>her2_Status <span class="op">==</span><span class="st"> &#39;0&#39;</span>) <span class="op">%&gt;%</span></span>
<span id="cb111-5"><a href="brca-rnaseq-preproc.html#cb111-5"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">mutate</span>(<span class="dt">group =</span> pam50_subtype) <span class="op">%&gt;%</span></span>
<span id="cb111-6"><a href="brca-rnaseq-preproc.html#cb111-6"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">rename</span>(<span class="dt">sampID =</span> title) <span class="op">%&gt;%</span></span>
<span id="cb111-7"><a href="brca-rnaseq-preproc.html#cb111-7"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">arrange</span>(group, sampID)</span>
<span id="cb111-8"><a href="brca-rnaseq-preproc.html#cb111-8"></a></span>
<span id="cb111-9"><a href="brca-rnaseq-preproc.html#cb111-9"></a><span class="co">#with(brcaRna_sampDesc, table(group, exclude=NULL))</span></span>
<span id="cb111-10"><a href="brca-rnaseq-preproc.html#cb111-10"></a></span>
<span id="cb111-11"><a href="brca-rnaseq-preproc.html#cb111-11"></a><span class="co"># Recode group</span></span>
<span id="cb111-12"><a href="brca-rnaseq-preproc.html#cb111-12"></a>brcaRna_sampDesc<span class="op">$</span>group &lt;-<span class="st"> </span><span class="kw">with</span>(</span>
<span id="cb111-13"><a href="brca-rnaseq-preproc.html#cb111-13"></a>  brcaRna_sampDesc,</span>
<span id="cb111-14"><a href="brca-rnaseq-preproc.html#cb111-14"></a>  <span class="kw">ifelse</span>(group <span class="op">==</span><span class="st"> &quot;LumA&quot;</span>, <span class="st">&quot;LumA&quot;</span>, <span class="st">&quot;Other&quot;</span>)  <span class="co">## Luma &gt; Other</span></span>
<span id="cb111-15"><a href="brca-rnaseq-preproc.html#cb111-15"></a>)</span>
<span id="cb111-16"><a href="brca-rnaseq-preproc.html#cb111-16"></a></span>
<span id="cb111-17"><a href="brca-rnaseq-preproc.html#cb111-17"></a><span class="co"># Re-order and name rows</span></span>
<span id="cb111-18"><a href="brca-rnaseq-preproc.html#cb111-18"></a>o.v &lt;-<span class="st"> </span><span class="kw">with</span>(brcaRna_sampDesc, <span class="kw">order</span>(group, sampID))</span>
<span id="cb111-19"><a href="brca-rnaseq-preproc.html#cb111-19"></a>brcaRna_sampDesc  &lt;-<span class="st"> </span>brcaRna_sampDesc[o.v, ]</span>
<span id="cb111-20"><a href="brca-rnaseq-preproc.html#cb111-20"></a></span>
<span id="cb111-21"><a href="brca-rnaseq-preproc.html#cb111-21"></a><span class="kw">rownames</span>(brcaRna_sampDesc) &lt;-<span class="st"> </span>brcaRna_sampDesc<span class="op">$</span>sampID</span>
<span id="cb111-22"><a href="brca-rnaseq-preproc.html#cb111-22"></a></span>
<span id="cb111-23"><a href="brca-rnaseq-preproc.html#cb111-23"></a><span class="co"># set brcaRna_groupCol for later</span></span>
<span id="cb111-24"><a href="brca-rnaseq-preproc.html#cb111-24"></a>brcaRna_groupCol &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;#F3C300&quot;</span>, <span class="st">&quot;#875692&quot;</span>)</span>
<span id="cb111-25"><a href="brca-rnaseq-preproc.html#cb111-25"></a><span class="kw">names</span>(brcaRna_groupCol) &lt;-<span class="st"> </span><span class="kw">unique</span>(brcaRna_sampDesc<span class="op">$</span>group)</span>
<span id="cb111-26"><a href="brca-rnaseq-preproc.html#cb111-26"></a></span>
<span id="cb111-27"><a href="brca-rnaseq-preproc.html#cb111-27"></a><span class="kw">with</span>(brcaRna_sampDesc, </span>
<span id="cb111-28"><a href="brca-rnaseq-preproc.html#cb111-28"></a> knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">table</span>(group, <span class="dt">exclude =</span> <span class="ot">NULL</span>),</span>
<span id="cb111-29"><a href="brca-rnaseq-preproc.html#cb111-29"></a>  <span class="dt">caption=</span><span class="st">&quot;Samples used in this analysis&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb111-30"><a href="brca-rnaseq-preproc.html#cb111-30"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span>
<span id="cb111-31"><a href="brca-rnaseq-preproc.html#cb111-31"></a>)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:brcaRna-preproc-subsetSamples">Table 6.1: </span>Samples used in this analysis
</caption>
<thead>
<tr>
<th style="text-align:left;">
group
</th>
<th style="text-align:right;">
Freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
LumA
</td>
<td style="text-align:right;">
1492
</td>
</tr>
<tr>
<td style="text-align:left;">
Other
</td>
<td style="text-align:right;">
933
</td>
</tr>
</tbody>
</table>
<p>The features in this dataset are gene expression indicators:</p>
<pre><code>
Gene expression data in FPKM were generated using cufflinks 2.2.1 
(default parameters except –GTF, –frag-bias-correct GRCh38.fa, –multi-read-correct, 
–library-type fr-firststrand, –total-hits-norm, –max-bundle-frags 10000000). 
The resulting data was was post-processed by collapsing on 30,865 unique gene symbols 
(sum of FPKM values of each matching transcript), adding to each expression measurement 
0.1 FPKM, and performing a log2 transformation.
</code></pre>
<p><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>.</p>
<p>We now proceed to run a <em>minimal set of QC sanity checks</em> to make sure that
there are no apparent systematic effects in the data.</p>
<p>Assemble feature matrix for <code>brcaRna_sampDesc</code> samples.</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="brca-rnaseq-preproc.html#cb113-1"></a><span class="co"># Start for geneExpression_repl1 data</span></span>
<span id="cb113-2"><a href="brca-rnaseq-preproc.html#cb113-2"></a>repl1_ndx &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">is.element</span>(<span class="kw">colnames</span>(brcaRna_geneExpression_repl1), brcaRna_sampDesc<span class="op">$</span>sampID))</span>
<span id="cb113-3"><a href="brca-rnaseq-preproc.html#cb113-3"></a>brcaRna_geneExpression &lt;-<span class="st"> </span>brcaRna_geneExpression_repl1[, repl1_ndx]</span>
<span id="cb113-4"><a href="brca-rnaseq-preproc.html#cb113-4"></a></span>
<span id="cb113-5"><a href="brca-rnaseq-preproc.html#cb113-5"></a><span class="co"># add samples from brcaRna_geneExpression_subJJ</span></span>
<span id="cb113-6"><a href="brca-rnaseq-preproc.html#cb113-6"></a><span class="cf">for</span>(JJ <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>) {</span>
<span id="cb113-7"><a href="brca-rnaseq-preproc.html#cb113-7"></a>  sub_ndx &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">is.element</span>(</span>
<span id="cb113-8"><a href="brca-rnaseq-preproc.html#cb113-8"></a>   <span class="kw">colnames</span>(<span class="kw">get</span>(<span class="kw">paste0</span>(<span class="st">&#39;brcaRna_geneExpression_sub&#39;</span>,JJ))),  brcaRna_sampDesc<span class="op">$</span>sampID))</span>
<span id="cb113-9"><a href="brca-rnaseq-preproc.html#cb113-9"></a>  brcaRna_geneExpression &lt;-<span class="st"> </span><span class="kw">cbind</span>(brcaRna_geneExpression,</span>
<span id="cb113-10"><a href="brca-rnaseq-preproc.html#cb113-10"></a>   <span class="kw">get</span>(<span class="kw">paste0</span>(<span class="st">&#39;brcaRna_geneExpression_sub&#39;</span>,JJ))[, sub_ndx])</span>
<span id="cb113-11"><a href="brca-rnaseq-preproc.html#cb113-11"></a>}</span>
<span id="cb113-12"><a href="brca-rnaseq-preproc.html#cb113-12"></a></span>
<span id="cb113-13"><a href="brca-rnaseq-preproc.html#cb113-13"></a><span class="co"># align brcaRna_geneExpression columns to brcaRna_sampDesc samples order</span></span>
<span id="cb113-14"><a href="brca-rnaseq-preproc.html#cb113-14"></a>geneExpr_ndx &lt;-<span class="st"> </span><span class="kw">match</span>(brcaRna_sampDesc<span class="op">$</span>sampID, <span class="kw">colnames</span>(brcaRna_geneExpression))</span>
<span id="cb113-15"><a href="brca-rnaseq-preproc.html#cb113-15"></a><span class="cf">if</span>(<span class="kw">any</span>(<span class="kw">is.na</span>(geneExpr_ndx)))</span>
<span id="cb113-16"><a href="brca-rnaseq-preproc.html#cb113-16"></a><span class="kw">stop</span>(<span class="st">&quot;brcaRna_sampDesc/brcaRna_geneExpression: sample mismatch&quot;</span>) <span class="cf">else</span></span>
<span id="cb113-17"><a href="brca-rnaseq-preproc.html#cb113-17"></a>brcaRna_geneExpression &lt;-<span class="st"> </span>brcaRna_geneExpression[, <span class="kw">rownames</span>(brcaRna_sampDesc)]</span></code></pre></div>
<p>We first look at coverage - make sure there isn’t too much disparity of coverage
across samples. To detect shared variability, samples can be annotated and ordered
according to sample features that may be linked to sample batch processing. Here we
the samples have been ordered by group and sample id (an alias of geoAcc).</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="brca-rnaseq-preproc.html#cb114-1"></a><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">3</span>))</span>
<span id="cb114-2"><a href="brca-rnaseq-preproc.html#cb114-2"></a><span class="kw">boxplot</span>(brcaRna_geneExpression,</span>
<span id="cb114-3"><a href="brca-rnaseq-preproc.html#cb114-3"></a>  <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">4</span>, <span class="dv">4</span>), <span class="dt">ylab=</span><span class="st">&#39;log2 Count&#39;</span>,</span>
<span id="cb114-4"><a href="brca-rnaseq-preproc.html#cb114-4"></a>  <span class="dt">staplewex =</span> <span class="dv">0</span>,       <span class="co"># remove horizontal whisker lines</span></span>
<span id="cb114-5"><a href="brca-rnaseq-preproc.html#cb114-5"></a>  <span class="dt">staplecol =</span> <span class="st">&quot;white&quot;</span>, <span class="co"># just to be totally sure :)</span></span>
<span id="cb114-6"><a href="brca-rnaseq-preproc.html#cb114-6"></a>  <span class="dt">outline =</span> F,         <span class="co"># remove outlying points</span></span>
<span id="cb114-7"><a href="brca-rnaseq-preproc.html#cb114-7"></a>  <span class="dt">whisklty =</span> <span class="dv">0</span>,        <span class="co"># remove vertical whisker lines</span></span>
<span id="cb114-8"><a href="brca-rnaseq-preproc.html#cb114-8"></a>  <span class="dt">las =</span> <span class="dv">2</span>, <span class="dt">horizontal =</span> F, <span class="dt">xaxt =</span> <span class="st">&quot;n&quot;</span>,</span>
<span id="cb114-9"><a href="brca-rnaseq-preproc.html#cb114-9"></a>  <span class="dt">border =</span> brcaRna_groupCol[brcaRna_sampDesc<span class="op">$</span>group]</span>
<span id="cb114-10"><a href="brca-rnaseq-preproc.html#cb114-10"></a>)</span>
<span id="cb114-11"><a href="brca-rnaseq-preproc.html#cb114-11"></a><span class="kw">legend</span>(<span class="st">&quot;top&quot;</span>, <span class="dt">legend =</span> <span class="kw">names</span>(brcaRna_groupCol), <span class="dt">text.col =</span> brcaRna_groupCol, </span>
<span id="cb114-12"><a href="brca-rnaseq-preproc.html#cb114-12"></a>  <span class="dt">ncol =</span> <span class="dv">2</span>, <span class="dt">bty =</span> <span class="st">&quot;n&quot;</span>)</span>
<span id="cb114-13"><a href="brca-rnaseq-preproc.html#cb114-13"></a><span class="co"># Add reference lines</span></span>
<span id="cb114-14"><a href="brca-rnaseq-preproc.html#cb114-14"></a>SampleMedian &lt;-<span class="st"> </span><span class="kw">apply</span>(brcaRna_geneExpression, <span class="dv">2</span>, median)</span>
<span id="cb114-15"><a href="brca-rnaseq-preproc.html#cb114-15"></a><span class="kw">abline</span>(<span class="dt">h =</span> <span class="kw">median</span>(SampleMedian), <span class="dt">col =</span> <span class="st">&quot;grey&quot;</span>)</span>
<span id="cb114-16"><a href="brca-rnaseq-preproc.html#cb114-16"></a><span class="kw">axis</span>(<span class="dt">side =</span> <span class="dv">4</span>, <span class="dt">at =</span> <span class="kw">round</span>(<span class="kw">median</span>(SampleMedian), <span class="dv">2</span>), </span>
<span id="cb114-17"><a href="brca-rnaseq-preproc.html#cb114-17"></a>  <span class="dt">las =</span> <span class="dv">2</span>, <span class="dt">col =</span> <span class="st">&quot;grey&quot;</span>, <span class="dt">line =</span> <span class="dv">0</span>, <span class="dt">tick =</span> F)</span></code></pre></div>
<div class="figure"><span id="fig:brcaRna-preproc-geneExprBxp"></span>
<img src="Static/figures/brcaRna-preproc-geneExprBxp-1.png" alt="Sample log2 count boxplots" width="960" />
<p class="caption">
Figure 6.1: Sample log2 count boxplots
</p>
</div>
<!-- DNA - we're not looking at COVERAGE!!!! 
Coverage/brcaRna_geneExpression level looks fairly comparable across samples.  It is sometimes helpful to
keep track of the actual coverage which can be adequetely tracked by distribution
quantiles.
-->
<!-- SKIP
From this table, we see that 25% of the samples have total coverage exceeding
$TICKr round(brcaRna_geneExpression_quant2["75%", "totCovM"],1)TICK$M reads, 25% of samples
have a 15 percentile of coverage lower than
$TICKr brcaRna_geneExpression_quant2["25%", "15%"]TICK$, etc.  
-->
<!-- SKIP
We next look at relative log representation (RLR) (in the context of measuring the density of 
5hmC marks in genes, we refer to `representation` as opposed to `expression`; the
two can be used interchangibly) -
make sure the shapes of the distributions are not widely different.
-->
<!-- SKIPPED
We note that the HCC samples have slightly more variable coverage distribution.
A few samples are quite different.
-->
</div>
<div id="breast-rnaseq-dra" class="section level2" number="6.2">
<h2><span class="header-section-number">6.2</span> Differential representation analysis</h2>
<p>In the remainder of this section, we will process the data and
perform differential expression analysis. If we had counts,
the analysis pipeline outlined in Law et al. (2018) <span class="citation">[<a href="references.html#ref-Law:2018aa" role="doc-biblioref">18</a>]</span>
would be appropriate. The main analysis steps of the pipeline are:</p>
<ul>
<li>remove lowly expressed genes</li>
<li>normalize gene expression distributions</li>
<li>remove heteroscedascity</li>
<li>fit linear models and examine DE results</li>
</ul>
<p>We will adapt this pipeline to the FPKM data that we have to work with - basically
skip the 2nd and 3rd steps in the above pipeline.</p>
<div id="remove-lowly-expressed-genes-1" class="section level3 unnumbered" number="">
<h3>Remove lowly expressed genes</h3>
<p>To determine a sensible threshold we can begin by examining the shapes of the distributions.</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="brca-rnaseq-preproc.html#cb115-1"></a><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb115-2"><a href="brca-rnaseq-preproc.html#cb115-2"></a><span class="kw">plot</span>(<span class="kw">density</span>(brcaRna_geneExpression[, <span class="dv">1</span>]), </span>
<span id="cb115-3"><a href="brca-rnaseq-preproc.html#cb115-3"></a>  <span class="dt">col =</span> brcaRna_groupCol[brcaRna_sampDesc<span class="op">$</span>group[<span class="dv">1</span>]], <span class="dt">lwd =</span> <span class="dv">2</span>, </span>
<span id="cb115-4"><a href="brca-rnaseq-preproc.html#cb115-4"></a>  <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">.30</span>), <span class="dt">xlim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">5</span>, <span class="dv">10</span>),</span>
<span id="cb115-5"><a href="brca-rnaseq-preproc.html#cb115-5"></a>  <span class="dt">las =</span> <span class="dv">2</span>, <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;gene expression&quot;</span></span>
<span id="cb115-6"><a href="brca-rnaseq-preproc.html#cb115-6"></a>)</span>
<span id="cb115-7"><a href="brca-rnaseq-preproc.html#cb115-7"></a><span class="kw">abline</span>(<span class="dt">v =</span> <span class="dv">0</span>, <span class="dt">col =</span> <span class="dv">3</span>)</span>
<span id="cb115-8"><a href="brca-rnaseq-preproc.html#cb115-8"></a><span class="co"># After verifying no outliers, can plot a random subset </span></span>
<span id="cb115-9"><a href="brca-rnaseq-preproc.html#cb115-9"></a><span class="cf">for</span> (JJ <span class="cf">in</span> <span class="kw">sample</span>(<span class="dv">2</span><span class="op">:</span><span class="kw">ncol</span>(brcaRna_geneExpression), <span class="dt">size =</span> <span class="dv">100</span>)) {</span>
<span id="cb115-10"><a href="brca-rnaseq-preproc.html#cb115-10"></a>  den &lt;-<span class="st"> </span><span class="kw">density</span>(brcaRna_geneExpression[, JJ])</span>
<span id="cb115-11"><a href="brca-rnaseq-preproc.html#cb115-11"></a>  <span class="kw">lines</span>(den<span class="op">$</span>x, den<span class="op">$</span>y, <span class="dt">col =</span> brcaRna_groupCol[brcaRna_sampDesc<span class="op">$</span>group[JJ]], <span class="dt">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb115-12"><a href="brca-rnaseq-preproc.html#cb115-12"></a>} <span class="co"># for(JJ</span></span>
<span id="cb115-13"><a href="brca-rnaseq-preproc.html#cb115-13"></a><span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="dt">legend =</span> <span class="kw">names</span>(brcaRna_groupCol), </span>
<span id="cb115-14"><a href="brca-rnaseq-preproc.html#cb115-14"></a>  <span class="dt">text.col =</span> brcaRna_groupCol, <span class="dt">bty =</span> <span class="st">&quot;n&quot;</span>)</span></code></pre></div>
<div class="figure"><span id="fig:brcaRna-preproc-densityGeneExpr"></span>
<img src="Static/figures/brcaRna-preproc-densityGeneExpr-1.png" alt="Sample gene expression densities" width="960" />
<p class="caption">
Figure 6.2: Sample gene expression densities
</p>
</div>
<p>In this analysis we will be using a nominal gene expression value of
0, genes are deeemed to be <code>represented</code> if their expression is
above this threshold, and not represented otherwise.
For this analysis we will require that genes be <code>represented</code> in at least
25 samples across the entire dataset to be retained for downstream analysis.</p>
<!-- or at least 
TICKr round(EXPR_THR*LibSizeSum['Max.'])TICK counts in the sample with the 
greatest sequencing depth (library size TICKr round(LibSizeSum['Max.'],1)TICK million).
-->
<p>Remove weakly represented genes and replot densities.</p>
<p>Removing 37.1% of genes…</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="brca-rnaseq-preproc.html#cb116-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb116-2"><a href="brca-rnaseq-preproc.html#cb116-2"></a></span>
<span id="cb116-3"><a href="brca-rnaseq-preproc.html#cb116-3"></a><span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb116-4"><a href="brca-rnaseq-preproc.html#cb116-4"></a><span class="kw">plot</span>(<span class="kw">density</span>(brcaRna_geneExpression_F[, <span class="dv">1</span>]),</span>
<span id="cb116-5"><a href="brca-rnaseq-preproc.html#cb116-5"></a>  <span class="dt">col =</span> brcaRna_groupCol[brcaRna_sampDesc<span class="op">$</span>group[<span class="dv">1</span>]],</span>
<span id="cb116-6"><a href="brca-rnaseq-preproc.html#cb116-6"></a>  <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">las =</span> <span class="dv">2</span>, <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;gene expression&quot;</span>,</span>
<span id="cb116-7"><a href="brca-rnaseq-preproc.html#cb116-7"></a>  <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">.30</span>), <span class="dt">xlim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">5</span>, <span class="dv">10</span>)</span>
<span id="cb116-8"><a href="brca-rnaseq-preproc.html#cb116-8"></a>)</span>
<span id="cb116-9"><a href="brca-rnaseq-preproc.html#cb116-9"></a><span class="co">#abline(v = 0, col = 3)</span></span>
<span id="cb116-10"><a href="brca-rnaseq-preproc.html#cb116-10"></a><span class="co"># After verifying no outliers, can plot a random subset </span></span>
<span id="cb116-11"><a href="brca-rnaseq-preproc.html#cb116-11"></a><span class="cf">for</span> (JJ <span class="cf">in</span> <span class="kw">sample</span>(<span class="dv">2</span><span class="op">:</span><span class="kw">ncol</span>(brcaRna_geneExpression_F), <span class="dt">size =</span> <span class="dv">100</span>)) {</span>
<span id="cb116-12"><a href="brca-rnaseq-preproc.html#cb116-12"></a>  den &lt;-<span class="st"> </span><span class="kw">density</span>(brcaRna_geneExpression_F[, JJ])</span>
<span id="cb116-13"><a href="brca-rnaseq-preproc.html#cb116-13"></a>  <span class="kw">lines</span>(den<span class="op">$</span>x, den<span class="op">$</span>y, <span class="dt">col =</span> brcaRna_groupCol[brcaRna_sampDesc<span class="op">$</span>group[JJ]], <span class="dt">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb116-14"><a href="brca-rnaseq-preproc.html#cb116-14"></a>} <span class="co"># for(JJ</span></span>
<span id="cb116-15"><a href="brca-rnaseq-preproc.html#cb116-15"></a><span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="dt">legend =</span> <span class="kw">names</span>(brcaRna_groupCol), </span>
<span id="cb116-16"><a href="brca-rnaseq-preproc.html#cb116-16"></a>  <span class="dt">text.col =</span> brcaRna_groupCol, <span class="dt">bty =</span> <span class="st">&quot;n&quot;</span>)</span></code></pre></div>
<div class="figure"><span id="fig:brcaRna-preproc-densityGeneExpr2"></span>
<img src="Static/figures/brcaRna-preproc-densityGeneExpr2-1.png" alt="Sample gene expression densities after removing weak genes" width="960" />
<p class="caption">
Figure 6.3: Sample gene expression densities after removing weak genes
</p>
</div>
<!--
Note that the $log_2(CMP)$ distribution is not quite symmetric.
-->
<p>As another sanity check, we will look at a
multidimensional scaling plot of distances between gene expression
profiles. We use <code>plotMDS</code> in limma package <span class="citation">[<a href="references.html#ref-Ritchie:2015aa" role="doc-biblioref">19</a>]</span>),
which plots samples on a two-dimensional scatterplot so that distances on
the plot approximate the typical log2 fold changes between the
samples.</p>
<!-- SKIP - data are $\approx$ normalized 
Before producing the MDS plot we will normalize the distributions.
We will store the data into s `DGEList` object as this is convenient
when running many of the analyses implemented in the edgeR and limma packages.
Call the set 'AF', for set 'A', 'Filtered'.
-->
<!-- SKIP
Verify that the counts are properly normalized.
-->
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="brca-rnaseq-preproc.html#cb117-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb117-2"><a href="brca-rnaseq-preproc.html#cb117-2"></a></span>
<span id="cb117-3"><a href="brca-rnaseq-preproc.html#cb117-3"></a><span class="kw">par</span>(<span class="dt">mfcol =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">1</span>), <span class="dt">xpd =</span> <span class="ot">NA</span>, <span class="dt">oma =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">0</span>))</span>
<span id="cb117-4"><a href="brca-rnaseq-preproc.html#cb117-4"></a></span>
<span id="cb117-5"><a href="brca-rnaseq-preproc.html#cb117-5"></a><span class="co"># wo loss of generality, sample 500 samples</span></span>
<span id="cb117-6"><a href="brca-rnaseq-preproc.html#cb117-6"></a><span class="co"># simply a matter of convenience to save time</span></span>
<span id="cb117-7"><a href="brca-rnaseq-preproc.html#cb117-7"></a><span class="co"># remove from final version</span></span>
<span id="cb117-8"><a href="brca-rnaseq-preproc.html#cb117-8"></a><span class="kw">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb117-9"><a href="brca-rnaseq-preproc.html#cb117-9"></a>samp_ndx &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(brcaRna_geneExpression_F), <span class="dt">size =</span> <span class="dv">500</span>)</span>
<span id="cb117-10"><a href="brca-rnaseq-preproc.html#cb117-10"></a>MDS.out &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">plotMDS</span>(brcaRna_geneExpression_F[, samp_ndx],</span>
<span id="cb117-11"><a href="brca-rnaseq-preproc.html#cb117-11"></a>  <span class="dt">col =</span> brcaRna_groupCol[brcaRna_sampDesc<span class="op">$</span>group[samp_ndx]], <span class="dt">pch =</span> <span class="dv">1</span></span>
<span id="cb117-12"><a href="brca-rnaseq-preproc.html#cb117-12"></a>)</span>
<span id="cb117-13"><a href="brca-rnaseq-preproc.html#cb117-13"></a><span class="kw">legend</span>(<span class="st">&quot;topleft&quot;</span>,</span>
<span id="cb117-14"><a href="brca-rnaseq-preproc.html#cb117-14"></a>  <span class="dt">legend =</span> <span class="kw">names</span>(brcaRna_groupCol),</span>
<span id="cb117-15"><a href="brca-rnaseq-preproc.html#cb117-15"></a>  <span class="dt">text.col =</span> brcaRna_groupCol, <span class="dt">bty =</span> <span class="st">&quot;n&quot;</span></span>
<span id="cb117-16"><a href="brca-rnaseq-preproc.html#cb117-16"></a>)</span>
<span id="cb117-17"><a href="brca-rnaseq-preproc.html#cb117-17"></a></span>
<span id="cb117-18"><a href="brca-rnaseq-preproc.html#cb117-18"></a>MDS.out &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">plotMDS</span>(brcaRna_geneExpression_F[, samp_ndx],</span>
<span id="cb117-19"><a href="brca-rnaseq-preproc.html#cb117-19"></a>  <span class="dt">col =</span> brcaRna_groupCol[brcaRna_sampDesc<span class="op">$</span>group[samp_ndx]], <span class="dt">pch =</span> <span class="dv">1</span>,</span>
<span id="cb117-20"><a href="brca-rnaseq-preproc.html#cb117-20"></a>  <span class="dt">dim.plot =</span> <span class="dv">3</span><span class="op">:</span><span class="dv">4</span></span>
<span id="cb117-21"><a href="brca-rnaseq-preproc.html#cb117-21"></a>)</span></code></pre></div>
<div class="figure"><span id="fig:brcaRna-preproc-plotMDS"></span>
<img src="Static/figures/brcaRna-preproc-plotMDS-1.png" alt="MDS plots of gene expression values" width="960" />
<p class="caption">
Figure 6.4: MDS plots of gene expression values
</p>
</div>
<p>The MDS plot, which is analogous to a PCA plot adapted to gene exression data,
does not indicate strong clustering of samples.</p>
<!-- SKIP
`glMDSPlot` from package `Glimma` provides an interactive MDS 
plot that can extremely usful for exploration
-->
<!-- SKIP
Link to glMDSPlot: 
[Here]($TICKr file.path(figures_DIR, paste0("GlMDSplot.html"))TICK  

No obvious factor links the samples in the 3 clusters observed on the
4th MDS dimensions. The percent of variance exaplained by this dimension or 
$\approx$ 4%.   The glMDSPlot indicates further segregation along
the 6th dimension.  The percent of variance exaplained by this dimension or 
$\approx$ 2%.  Tracking down this source of variability may be quite challenging,
especially without having the complete information about the sample attributes 
and provenance.  

Unwanted variability is a well-documented problem in the analysis of RNA-Seq data
(see Peixoto et al. (2015) [@Peixoto:2015aa]), and many procedures have been proposed
to reduce the effect of unwanted variation on RNA-Seq analsys results 
([@Gandolfo:2018aa;@Peixoto:2015aa;@Risso:2014aa]).  There are undoubtedly
some similar sources of systematic variation in the 5hmC data, but it is
beyond the scope of this work to investigate these in this particular dataset.
Given that the clustering of samples occurs in MDS dimensions that explain
a small fraction of variability, and that these is no assocation with the
factor of interest, LumA vs Other, these sources of variability should not
interfere too much with our classification analysis.  It would nonetheless be interesting
to assess whether downstream results can be improved by removing this variability.

-->
</div>
<div id="creating-a-design-matrix-and-contrasts-1" class="section level3 unnumbered" number="">
<h3>Creating a design matrix and contrasts</h3>
<p>Before proceeding with the statistical modeling used for the
differential expression analysis, we need to set up a
model design matrix.</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="brca-rnaseq-preproc.html#cb118-1"></a><span class="co">### CLEAR CACHE</span></span>
<span id="cb118-2"><a href="brca-rnaseq-preproc.html#cb118-2"></a></span>
<span id="cb118-3"><a href="brca-rnaseq-preproc.html#cb118-3"></a>Design_mtx &lt;-<span class="st"> </span><span class="kw">model.matrix</span>( <span class="op">~</span><span class="st">  </span><span class="dv">-1</span> <span class="op">+</span><span class="st"> </span>group, <span class="dt">data=</span>brcaRna_sampDesc)</span>
<span id="cb118-4"><a href="brca-rnaseq-preproc.html#cb118-4"></a><span class="kw">colnames</span>(Design_mtx) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&#39;group&#39;</span>, <span class="st">&#39;&#39;</span>, <span class="kw">colnames</span>(Design_mtx))</span>
<span id="cb118-5"><a href="brca-rnaseq-preproc.html#cb118-5"></a></span>
<span id="cb118-6"><a href="brca-rnaseq-preproc.html#cb118-6"></a><span class="kw">cat</span>(<span class="st">&quot;colSums(Design_mtx):</span><span class="ch">\n</span><span class="st">&quot;</span>)</span></code></pre></div>
<pre><code>## colSums(Design_mtx):</code></pre>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="brca-rnaseq-preproc.html#cb120-1"></a><span class="kw">colSums</span>(Design_mtx)</span></code></pre></div>
<pre><code>##  LumA Other 
##  1492   933</code></pre>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="brca-rnaseq-preproc.html#cb122-1"></a>Contrasts_mtx &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">makeContrasts</span>(</span>
<span id="cb122-2"><a href="brca-rnaseq-preproc.html#cb122-2"></a>  <span class="dt">LumAvsOther =</span> LumA  <span class="op">-</span><span class="st"> </span>Other,</span>
<span id="cb122-3"><a href="brca-rnaseq-preproc.html#cb122-3"></a>  <span class="dt">levels=</span><span class="kw">colnames</span>(Design_mtx))</span>
<span id="cb122-4"><a href="brca-rnaseq-preproc.html#cb122-4"></a></span>
<span id="cb122-5"><a href="brca-rnaseq-preproc.html#cb122-5"></a><span class="kw">cat</span>(<span class="st">&quot;Contrasts:</span><span class="ch">\n</span><span class="st">&quot;</span>)</span></code></pre></div>
<pre><code>## Contrasts:</code></pre>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="brca-rnaseq-preproc.html#cb124-1"></a>Contrasts_mtx</span></code></pre></div>
<pre><code>##        Contrasts
## Levels  LumAvsOther
##   LumA            1
##   Other          -1</code></pre>
<!-- DOES NOT APPLY - NEED COUNTS!!!!
### Removing heteroscedasticity from the count data {-}


As for RNA-Seq data, for 5hmC count data the variance is not independent of the mean.
In `limma`, the R package we are using for our analyses, 
linear modeling is carried out on the gene expression values which are assumed to be 
normally distributed and the mean-variance relationship is accommodated using precision 
weights calculated by the voom function.  We apply this transformation next.

-->
<!--

Note that the voom-plot provides a visual check on the level of filtering performed upstream.
If filtering of lowly-expressed genes is insufficient, a drop in variance levels can be 
observed at the low end of the expression scale due to very small counts. 

-->
<!--
Means (x-axis) and variances (y-axis) of each gene are plotted to show the dependence between the two before voom is applied to the data ( A) and how the trend is removed after voom precision weights are applied to the data ( B). The plot on the left is created within the voom function which extracts residual variances from fitting linear models to gene expression transformed data. Variances are then re-scaled to quarter-root variances (or square-root of standard deviations) and plotted against the mean expression of each gene. The means are log 2-transformed mean-counts with an offset of 2. The plot on the right is created using plotSA which plots log 2 residual standard deviations against mean gene expression values. The average log 2 residual standard deviation is marked by a horizontal blue line. In both plots, each black dot represents a gene and a red curve is fitted to these points.
-->
<!--

To get a sense of how this compares with RNA-Seq dsta, we can take a look at Figure 4 in Law et al. [@Law:2018aa]:

TICKr knitr::include_graphics(
  "Static/images/Law-fig4.gif",  dpi=100)TICK

We observe that the variability in the 5hmC data is quite a bit lower.  Statistical summaries would give
a better idea.  

-->
</div>
<div id="fit-linear-models-and-examine-the-results-1" class="section level3 unnumbered" number="">
<h3>Fit linear models and examine the results</h3>
<p>Having properly filtered and normalized the data,
the linear models can be fitted to each gene and the results
examined to assess differential expression between the two groups
of interest, in our case LumA vs Other.</p>
<p>Table <a href="#tab:lmFit"><strong>??</strong></a> displays the counts of genes in each DE category:</p>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb126-1"><a href="brca-rnaseq-preproc.html#cb126-1"></a> brcaRna_geneExpression_F_fit &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">lmFit</span>(brcaRna_geneExpression_F, Design_mtx)</span>
<span id="cb126-2"><a href="brca-rnaseq-preproc.html#cb126-2"></a> <span class="kw">colnames</span>(brcaRna_geneExpression_F_fit<span class="op">$</span>coefficients) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">(Intercept</span><span class="ch">\\</span><span class="st">)&quot;</span>, <span class="st">&quot;Intercept&quot;</span>,</span>
<span id="cb126-3"><a href="brca-rnaseq-preproc.html#cb126-3"></a> <span class="kw">colnames</span>(brcaRna_geneExpression_F_fit<span class="op">$</span>coefficients) )</span>
<span id="cb126-4"><a href="brca-rnaseq-preproc.html#cb126-4"></a></span>
<span id="cb126-5"><a href="brca-rnaseq-preproc.html#cb126-5"></a> brcaRna_geneExpression_F_fit &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">contrasts.fit</span>(</span>
<span id="cb126-6"><a href="brca-rnaseq-preproc.html#cb126-6"></a>    brcaRna_geneExpression_F_fit, <span class="dt">contrasts=</span>Contrasts_mtx)</span>
<span id="cb126-7"><a href="brca-rnaseq-preproc.html#cb126-7"></a></span>
<span id="cb126-8"><a href="brca-rnaseq-preproc.html#cb126-8"></a> brcaRna_geneExpression_F_efit &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">eBayes</span>(brcaRna_geneExpression_F_fit)</span>
<span id="cb126-9"><a href="brca-rnaseq-preproc.html#cb126-9"></a></span>
<span id="cb126-10"><a href="brca-rnaseq-preproc.html#cb126-10"></a> brcaRna_geneExpression_F_efit_dt &lt;-</span>
<span id="cb126-11"><a href="brca-rnaseq-preproc.html#cb126-11"></a><span class="st"> </span>limma<span class="op">::</span><span class="kw">decideTests</span>(brcaRna_geneExpression_F_efit,<span class="dt">adjust.method =</span> <span class="st">&quot;BH&quot;</span>, <span class="dt">p.value =</span> <span class="fl">0.05</span>)</span>
<span id="cb126-12"><a href="brca-rnaseq-preproc.html#cb126-12"></a> </span>
<span id="cb126-13"><a href="brca-rnaseq-preproc.html#cb126-13"></a> knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">summary</span>(brcaRna_geneExpression_F_efit_dt)),</span>
<span id="cb126-14"><a href="brca-rnaseq-preproc.html#cb126-14"></a>  <span class="dt">caption=</span><span class="st">&quot;DE Results at FDR = 0.05&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb126-15"><a href="brca-rnaseq-preproc.html#cb126-15"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:brcaRna-preproc-lmFit">Table 6.2: </span>DE Results at FDR = 0.05
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Down
</th>
<th style="text-align:right;">
NotSig
</th>
<th style="text-align:right;">
Up
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
LumAvsOther
</td>
<td style="text-align:right;">
6301
</td>
<td style="text-align:right;">
4498
</td>
<td style="text-align:right;">
8607
</td>
</tr>
</tbody>
</table>
</div>
<div id="graphical-representations-of-de-results-md-plots-1" class="section level3 unnumbered" number="">
<h3>Graphical representations of DE results: MD Plots</h3>
<p>To summarise results for all genes visually, mean-difference plots
(aka MA plot), which display log-FCs from the linear model fit against
the average gene expression values can be generated using the plotMD function,
with the differentially expressed genes highlighted.</p>
<p>We may also be interested in whether certain gene features are
related to gene identification. Gene GC content, for example, might be
of interest. We don’t have GC content here - skip this for now.</p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="brca-rnaseq-preproc.html#cb127-1"></a><span class="co">###  CLEAR CACHE</span></span>
<span id="cb127-2"><a href="brca-rnaseq-preproc.html#cb127-2"></a></span>
<span id="cb127-3"><a href="brca-rnaseq-preproc.html#cb127-3"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">mar=</span><span class="kw">c</span>(<span class="fl">4.5</span>,<span class="fl">4.5</span>,<span class="dv">2</span>,<span class="dv">1</span>),<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">0</span>))</span>
<span id="cb127-4"><a href="brca-rnaseq-preproc.html#cb127-4"></a></span>
<span id="cb127-5"><a href="brca-rnaseq-preproc.html#cb127-5"></a><span class="co"># log-fold-change vs ave-expr</span></span>
<span id="cb127-6"><a href="brca-rnaseq-preproc.html#cb127-6"></a>limma<span class="op">::</span><span class="kw">plotMD</span>(brcaRna_geneExpression_F_efit,</span>
<span id="cb127-7"><a href="brca-rnaseq-preproc.html#cb127-7"></a> <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.5</span>, <span class="fl">0.5</span>),</span>
<span id="cb127-8"><a href="brca-rnaseq-preproc.html#cb127-8"></a> <span class="dt">column=</span><span class="st">&#39;LumAvsOther&#39;</span>,</span>
<span id="cb127-9"><a href="brca-rnaseq-preproc.html#cb127-9"></a> <span class="dt">status=</span>brcaRna_geneExpression_F_efit_dt[,<span class="st">&#39;LumAvsOther&#39;</span>],</span>
<span id="cb127-10"><a href="brca-rnaseq-preproc.html#cb127-10"></a> <span class="dt">hl.pch =</span> <span class="dv">16</span>, <span class="dt">hl.col =</span> <span class="kw">c</span>(<span class="st">&quot;lightblue&quot;</span>, <span class="st">&quot;pink&quot;</span>), <span class="dt">hl.cex =</span> <span class="fl">.5</span>,</span>
<span id="cb127-11"><a href="brca-rnaseq-preproc.html#cb127-11"></a> <span class="dt">bg.pch =</span> <span class="dv">16</span>, <span class="dt">bg.col =</span> <span class="st">&quot;grey&quot;</span>, <span class="dt">bg.cex =</span> <span class="fl">0.5</span>,</span>
<span id="cb127-12"><a href="brca-rnaseq-preproc.html#cb127-12"></a> <span class="dt">main =</span> <span class="st">&#39;&#39;</span>,</span>
<span id="cb127-13"><a href="brca-rnaseq-preproc.html#cb127-13"></a> <span class="dt">xlab =</span> <span class="kw">paste0</span>(</span>
<span id="cb127-14"><a href="brca-rnaseq-preproc.html#cb127-14"></a>    <span class="st">&quot;Average log-expression: IQR=&quot;</span>,</span>
<span id="cb127-15"><a href="brca-rnaseq-preproc.html#cb127-15"></a>    <span class="kw">paste</span>(<span class="kw">round</span>(<span class="kw">quantile</span>(brcaRna_geneExpression_F_efit<span class="op">$</span>Amean, <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>), <span class="dv">2</span>),</span>
<span id="cb127-16"><a href="brca-rnaseq-preproc.html#cb127-16"></a>      <span class="dt">collapse =</span> <span class="st">&quot;, &quot;</span></span>
<span id="cb127-17"><a href="brca-rnaseq-preproc.html#cb127-17"></a>    )</span>
<span id="cb127-18"><a href="brca-rnaseq-preproc.html#cb127-18"></a>  ),</span>
<span id="cb127-19"><a href="brca-rnaseq-preproc.html#cb127-19"></a>  <span class="dt">ylab =</span> <span class="kw">paste0</span>(</span>
<span id="cb127-20"><a href="brca-rnaseq-preproc.html#cb127-20"></a>    <span class="st">&quot;log-fold-change: IQR=&quot;</span>,</span>
<span id="cb127-21"><a href="brca-rnaseq-preproc.html#cb127-21"></a>    <span class="kw">paste</span>(<span class="kw">round</span>(<span class="kw">quantile</span>(brcaRna_geneExpression_F_efit<span class="op">$</span>coefficients[, <span class="st">&#39;LumAvsOther&#39;</span>], <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>), <span class="dv">2</span>),</span>
<span id="cb127-22"><a href="brca-rnaseq-preproc.html#cb127-22"></a>      <span class="dt">collapse =</span> <span class="st">&quot;, &quot;</span></span>
<span id="cb127-23"><a href="brca-rnaseq-preproc.html#cb127-23"></a>    )</span>
<span id="cb127-24"><a href="brca-rnaseq-preproc.html#cb127-24"></a>  ),</span>
<span id="cb127-25"><a href="brca-rnaseq-preproc.html#cb127-25"></a>  <span class="dt">legend =</span> F, <span class="dt">cex.lab=</span><span class="fl">1.5</span></span>
<span id="cb127-26"><a href="brca-rnaseq-preproc.html#cb127-26"></a>)</span>
<span id="cb127-27"><a href="brca-rnaseq-preproc.html#cb127-27"></a><span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">0</span>, <span class="dt">col =</span> <span class="st">&quot;black&quot;</span>)</span>
<span id="cb127-28"><a href="brca-rnaseq-preproc.html#cb127-28"></a><span class="kw">rug</span>(<span class="kw">quantile</span>(brcaRna_geneExpression_F_efit<span class="op">$</span>coefficients[, <span class="st">&#39;LumAvsOther&#39;</span>], <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>),</span>
<span id="cb127-29"><a href="brca-rnaseq-preproc.html#cb127-29"></a>  <span class="dt">col =</span> <span class="st">&quot;purple&quot;</span>,</span>
<span id="cb127-30"><a href="brca-rnaseq-preproc.html#cb127-30"></a>  <span class="dt">ticksize =</span> <span class="fl">.03</span>, <span class="dt">side =</span> <span class="dv">2</span>, <span class="dt">lwd =</span> <span class="dv">2</span></span>
<span id="cb127-31"><a href="brca-rnaseq-preproc.html#cb127-31"></a>)</span>
<span id="cb127-32"><a href="brca-rnaseq-preproc.html#cb127-32"></a><span class="kw">rug</span>(<span class="kw">quantile</span>(brcaRna_geneExpression_F_efit<span class="op">$</span>Amean, <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>),</span>
<span id="cb127-33"><a href="brca-rnaseq-preproc.html#cb127-33"></a>  <span class="dt">col =</span> <span class="st">&quot;purple&quot;</span>,</span>
<span id="cb127-34"><a href="brca-rnaseq-preproc.html#cb127-34"></a>  <span class="dt">ticksize =</span> <span class="fl">.03</span>, <span class="dt">side =</span> <span class="dv">1</span>, <span class="dt">lwd =</span> <span class="dv">2</span></span>
<span id="cb127-35"><a href="brca-rnaseq-preproc.html#cb127-35"></a>)</span>
<span id="cb127-36"><a href="brca-rnaseq-preproc.html#cb127-36"></a></span>
<span id="cb127-37"><a href="brca-rnaseq-preproc.html#cb127-37"></a><span class="co"># log-fold-change vs identification</span></span>
<span id="cb127-38"><a href="brca-rnaseq-preproc.html#cb127-38"></a></span>
<span id="cb127-39"><a href="brca-rnaseq-preproc.html#cb127-39"></a><span class="kw">boxplot</span>(<span class="kw">split</span>(</span>
<span id="cb127-40"><a href="brca-rnaseq-preproc.html#cb127-40"></a> brcaRna_geneExpression_F_efit<span class="op">$</span>coefficients[, <span class="st">&#39;LumAvsOther&#39;</span>],</span>
<span id="cb127-41"><a href="brca-rnaseq-preproc.html#cb127-41"></a> brcaRna_geneExpression_F_efit_dt[,<span class="st">&#39;LumAvsOther&#39;</span>]),</span>
<span id="cb127-42"><a href="brca-rnaseq-preproc.html#cb127-42"></a> <span class="dt">outline=</span>F,</span>
<span id="cb127-43"><a href="brca-rnaseq-preproc.html#cb127-43"></a> <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&quot;pink&quot;</span>, <span class="st">&quot;grey&quot;</span>, <span class="st">&quot;lightblue&quot;</span>), <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>,</span>
<span id="cb127-44"><a href="brca-rnaseq-preproc.html#cb127-44"></a> <span class="dt">ylab=</span><span class="st">&#39;log-fold-change&#39;</span>, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span>.<span class="dv">4</span>, <span class="fl">.4</span>),</span>
<span id="cb127-45"><a href="brca-rnaseq-preproc.html#cb127-45"></a> <span class="dt">cex.lab=</span><span class="fl">1.5</span></span>
<span id="cb127-46"><a href="brca-rnaseq-preproc.html#cb127-46"></a>)</span>
<span id="cb127-47"><a href="brca-rnaseq-preproc.html#cb127-47"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="kw">c</span>(<span class="st">&#39;down&#39;</span>, <span class="st">&#39;notDE&#39;</span>, <span class="st">&#39;up&#39;</span>), <span class="dt">cex.axis=</span><span class="fl">1.5</span>)</span></code></pre></div>
<div class="figure"><span id="fig:brcaRna-preproc-mdPlotEfit"></span>
<img src="Static/figures/brcaRna-preproc-mdPlotEfit-1.png" alt="LumA vs Other - Genes Identified at FDR = 0,05" width="1056" />
<p class="caption">
Figure 6.5: LumA vs Other - Genes Identified at FDR = 0,05
</p>
</div>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="brca-rnaseq-preproc.html#cb128-1"></a><span class="co">#SKIP - DONT HAVE ACGT </span></span>
<span id="cb128-2"><a href="brca-rnaseq-preproc.html#cb128-2"></a>SKIP &lt;-<span class="st"> </span><span class="cf">function</span>() {</span>
<span id="cb128-3"><a href="brca-rnaseq-preproc.html#cb128-3"></a><span class="co"># gc vs identification</span></span>
<span id="cb128-4"><a href="brca-rnaseq-preproc.html#cb128-4"></a>genes_ndx &lt;-<span class="st"> </span><span class="kw">match</span>(<span class="kw">rownames</span>(brcaRna_geneExpression_F_efit), brcaRna_genes_annotF<span class="op">$</span>geneSymbol)</span>
<span id="cb128-5"><a href="brca-rnaseq-preproc.html#cb128-5"></a><span class="cf">if</span>(<span class="kw">sum</span>(<span class="kw">is.na</span>(genes_ndx))) <span class="kw">stop</span>(<span class="st">&quot;brcaRna_geneExpression_F_efit/brcaRna_genes_annotF: genes mismatch&quot;</span>)</span>
<span id="cb128-6"><a href="brca-rnaseq-preproc.html#cb128-6"></a>GC_vec &lt;-<span class="st"> </span><span class="kw">with</span>(brcaRna_genes_annotF[genes_ndx,],(G<span class="op">+</span>C)<span class="op">/</span>(A<span class="op">+</span>C<span class="op">+</span>G<span class="op">+</span>T))</span>
<span id="cb128-7"><a href="brca-rnaseq-preproc.html#cb128-7"></a></span>
<span id="cb128-8"><a href="brca-rnaseq-preproc.html#cb128-8"></a></span>
<span id="cb128-9"><a href="brca-rnaseq-preproc.html#cb128-9"></a><span class="kw">boxplot</span>(<span class="kw">split</span>(</span>
<span id="cb128-10"><a href="brca-rnaseq-preproc.html#cb128-10"></a> GC_vec,</span>
<span id="cb128-11"><a href="brca-rnaseq-preproc.html#cb128-11"></a> brcaRna_geneExpression_F_efit_dt[,<span class="st">&#39;LumAvsOther&#39;</span>]),</span>
<span id="cb128-12"><a href="brca-rnaseq-preproc.html#cb128-12"></a> <span class="dt">outline=</span>F,</span>
<span id="cb128-13"><a href="brca-rnaseq-preproc.html#cb128-13"></a> <span class="dt">border=</span><span class="kw">c</span>(<span class="st">&quot;pink&quot;</span>, <span class="st">&quot;grey&quot;</span>, <span class="st">&quot;lightblue&quot;</span>), <span class="dt">xaxt=</span><span class="st">&#39;n&#39;</span>,</span>
<span id="cb128-14"><a href="brca-rnaseq-preproc.html#cb128-14"></a> <span class="dt">ylab=</span><span class="st">&#39;gene-gc&#39;</span>, <span class="dt">cex.lab=</span><span class="fl">1.5</span></span>
<span id="cb128-15"><a href="brca-rnaseq-preproc.html#cb128-15"></a>)</span>
<span id="cb128-16"><a href="brca-rnaseq-preproc.html#cb128-16"></a><span class="kw">axis</span>(<span class="dt">side=</span><span class="dv">1</span>, <span class="dt">at=</span><span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="kw">c</span>(<span class="st">&#39;down&#39;</span>, <span class="st">&#39;notDE&#39;</span>, <span class="st">&#39;up&#39;</span>), <span class="dt">cex.axis=</span><span class="fl">1.5</span>)</span>
<span id="cb128-17"><a href="brca-rnaseq-preproc.html#cb128-17"></a></span>
<span id="cb128-18"><a href="brca-rnaseq-preproc.html#cb128-18"></a> <span class="co">#mtext(side=3, outer=T, cex=1.25, &quot;Genes identified at adjusted p-value=0.05&quot;)</span></span>
<span id="cb128-19"><a href="brca-rnaseq-preproc.html#cb128-19"></a>}<span class="co">#SKIP</span></span></code></pre></div>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="brca-rnaseq-preproc.html#cb129-1"></a><span class="co">###  CLEAR CACHE</span></span>
<span id="cb129-2"><a href="brca-rnaseq-preproc.html#cb129-2"></a></span>
<span id="cb129-3"><a href="brca-rnaseq-preproc.html#cb129-3"></a>brcaRna_geneExpression_F_logFC_sum &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb129-4"><a href="brca-rnaseq-preproc.html#cb129-4"></a> <span class="kw">split</span>(</span>
<span id="cb129-5"><a href="brca-rnaseq-preproc.html#cb129-5"></a> brcaRna_geneExpression_F_efit<span class="op">$</span>coefficients[, <span class="st">&#39;LumAvsOther&#39;</span>],</span>
<span id="cb129-6"><a href="brca-rnaseq-preproc.html#cb129-6"></a> brcaRna_geneExpression_F_efit_dt[,<span class="st">&#39;LumAvsOther&#39;</span>]),</span>
<span id="cb129-7"><a href="brca-rnaseq-preproc.html#cb129-7"></a> quantile, <span class="dt">prob =</span> (<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>)</span>
<span id="cb129-8"><a href="brca-rnaseq-preproc.html#cb129-8"></a></span>
<span id="cb129-9"><a href="brca-rnaseq-preproc.html#cb129-9"></a><span class="kw">colnames</span>(brcaRna_geneExpression_F_logFC_sum) &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">factor</span>(</span>
<span id="cb129-10"><a href="brca-rnaseq-preproc.html#cb129-10"></a> <span class="kw">colnames</span>(brcaRna_geneExpression_F_logFC_sum), </span>
<span id="cb129-11"><a href="brca-rnaseq-preproc.html#cb129-11"></a> <span class="dt">levels=</span><span class="kw">c</span>(<span class="st">&quot;-1&quot;</span>, <span class="st">&quot;0&quot;</span>, <span class="st">&quot;1&quot;</span>),</span>
<span id="cb129-12"><a href="brca-rnaseq-preproc.html#cb129-12"></a> <span class="dt">labels=</span><span class="kw">c</span>(<span class="st">&#39;down&#39;</span>, <span class="st">&#39;notDE&#39;</span>, <span class="st">&#39;up&#39;</span>)</span>
<span id="cb129-13"><a href="brca-rnaseq-preproc.html#cb129-13"></a>))</span>
<span id="cb129-14"><a href="brca-rnaseq-preproc.html#cb129-14"></a></span>
<span id="cb129-15"><a href="brca-rnaseq-preproc.html#cb129-15"></a></span>
<span id="cb129-16"><a href="brca-rnaseq-preproc.html#cb129-16"></a>knitr<span class="op">::</span><span class="kw">kable</span>(brcaRna_geneExpression_F_logFC_sum,</span>
<span id="cb129-17"><a href="brca-rnaseq-preproc.html#cb129-17"></a>  <span class="dt">digits =</span> <span class="dv">2</span>,</span>
<span id="cb129-18"><a href="brca-rnaseq-preproc.html#cb129-18"></a>  <span class="dt">caption =</span> <span class="st">&quot;log FC quartiles by gene identification&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb129-19"><a href="brca-rnaseq-preproc.html#cb129-19"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:brcaRna-preproc-quantlogFC">Table 6.3: </span>log FC quartiles by gene identification
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
down
</th>
<th style="text-align:right;">
notDE
</th>
<th style="text-align:right;">
up
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
25%
</td>
<td style="text-align:right;">
-0.28
</td>
<td style="text-align:right;">
-0.03
</td>
<td style="text-align:right;">
0.11
</td>
</tr>
<tr>
<td style="text-align:left;">
50%
</td>
<td style="text-align:right;">
-0.17
</td>
<td style="text-align:right;">
0.00
</td>
<td style="text-align:right;">
0.20
</td>
</tr>
<tr>
<td style="text-align:left;">
75%
</td>
<td style="text-align:right;">
-0.10
</td>
<td style="text-align:right;">
0.03
</td>
<td style="text-align:right;">
0.37
</td>
</tr>
</tbody>
</table>
<p>Many genes are identified, and the effect sizes are substantial
in comparison to the 5hmc data - compare tables
<a href="#tab:preprocBrCa-quantlogFC"><strong>??</strong></a> and
<a href="#tab:quantlogFC"><strong>??</strong></a>.
This will result in a higher signal-to-noise ratio context. See
Section <a href="brca-rnaseq-preproc.html#breast-rnaseq-snr-regime">6.3</a>.</p>
<!-- 5hmc comments
The log-fold-change distribution for up-represented genes is long-tailed,
with many high log fold-change values.
By contrast, log-fold-change distribution for down-represented genes
closer to symmetric and has few genes with low log fold-change values.
We will see how this affects the results of identifying genes with
an effect size requirement.

The GC content of down regulated genes tends to be slightly lower than the
rest of the genes.  A statistical test would find that the difference
between the mean of the down regulated gene population is singificantly different
than the mean of the other gene population even though the difference is
quite small
(TICKr round( 
mean(GC_vec[brcaRna_geneExpression_F_efit_dt[,'LumAvsOther']=='-1']) -
mean(GC_vec[brcaRna_geneExpression_F_efit_dt[,'LumAvsOther']!='-1']),
3)TICK

These asymmetries are minor, but it would still be good to establish that
they relfect biology rather than processing artifacts.  
-->
</div>
<div id="de-genes-at-10-fold-change-1" class="section level3 unnumbered" number="">
<h3>DE genes at 10% fold change</h3>
<p>For a stricter definition on significance, one may require log-fold-changes
(log-FCs) to be above a minimum value. The treat method
(McCarthy and Smyth 2009 <span class="citation">[<a href="references.html#ref-McCarthy:2009aa" role="doc-biblioref">23</a>]</span>) can be used to calculate p-values
from empirical Bayes moderated t-statistics with a minimum log-FC requirement.
The number of differentially expressed genes are greatly reduced if we
impose a minimal fold-change requirement of 10%.</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="brca-rnaseq-preproc.html#cb130-1"></a><span class="co">###  CLEAR CACHE</span></span>
<span id="cb130-2"><a href="brca-rnaseq-preproc.html#cb130-2"></a></span>
<span id="cb130-3"><a href="brca-rnaseq-preproc.html#cb130-3"></a>brcaRna_geneExpression_F_tfit &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">treat</span>(brcaRna_geneExpression_F_fit, <span class="dt">lfc=</span><span class="kw">log2</span>(<span class="fl">1.10</span>))</span>
<span id="cb130-4"><a href="brca-rnaseq-preproc.html#cb130-4"></a>brcaRna_geneExpression_F_tfit_dt &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">decideTests</span>(brcaRna_geneExpression_F_tfit)</span>
<span id="cb130-5"><a href="brca-rnaseq-preproc.html#cb130-5"></a></span>
<span id="cb130-6"><a href="brca-rnaseq-preproc.html#cb130-6"></a><span class="kw">cat</span>(<span class="st">&quot;10% FC Gene Identification Summary - voom, adjust.method = BH, p.value = 0.05:</span><span class="ch">\n</span><span class="st">&quot;</span>)</span></code></pre></div>
<pre><code>## 10% FC Gene Identification Summary - voom, adjust.method = BH, p.value = 0.05:</code></pre>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="brca-rnaseq-preproc.html#cb132-1"></a><span class="kw">summary</span>(brcaRna_geneExpression_F_tfit_dt)</span></code></pre></div>
<pre><code>##        LumAvsOther
## Down          2569
## NotSig       12519
## Up            4318</code></pre>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="brca-rnaseq-preproc.html#cb134-1"></a><span class="co"># log-fold-change vs ave-expr</span></span>
<span id="cb134-2"><a href="brca-rnaseq-preproc.html#cb134-2"></a>limma<span class="op">::</span><span class="kw">plotMD</span>(brcaRna_geneExpression_F_efit,</span>
<span id="cb134-3"><a href="brca-rnaseq-preproc.html#cb134-3"></a> <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">0.6</span>, <span class="fl">0.6</span>),</span>
<span id="cb134-4"><a href="brca-rnaseq-preproc.html#cb134-4"></a> <span class="dt">column=</span><span class="st">&#39;LumAvsOther&#39;</span>,</span>
<span id="cb134-5"><a href="brca-rnaseq-preproc.html#cb134-5"></a> <span class="dt">status=</span>brcaRna_geneExpression_F_tfit_dt[,<span class="st">&#39;LumAvsOther&#39;</span>],</span>
<span id="cb134-6"><a href="brca-rnaseq-preproc.html#cb134-6"></a> <span class="dt">hl.pch =</span> <span class="dv">16</span>, <span class="dt">hl.col =</span> <span class="kw">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;red&quot;</span>), <span class="dt">hl.cex =</span> <span class="fl">.7</span>,</span>
<span id="cb134-7"><a href="brca-rnaseq-preproc.html#cb134-7"></a> <span class="dt">bg.pch =</span> <span class="dv">16</span>, <span class="dt">bg.col =</span> <span class="st">&quot;grey&quot;</span>, <span class="dt">bg.cex =</span> <span class="fl">0.5</span>,</span>
<span id="cb134-8"><a href="brca-rnaseq-preproc.html#cb134-8"></a> <span class="dt">main =</span> <span class="st">&#39;&#39;</span>,</span>
<span id="cb134-9"><a href="brca-rnaseq-preproc.html#cb134-9"></a> <span class="dt">xlab =</span> <span class="kw">paste0</span>(</span>
<span id="cb134-10"><a href="brca-rnaseq-preproc.html#cb134-10"></a>    <span class="st">&quot;Average log-expression: IQR=&quot;</span>,</span>
<span id="cb134-11"><a href="brca-rnaseq-preproc.html#cb134-11"></a>    <span class="kw">paste</span>(<span class="kw">round</span>(<span class="kw">quantile</span>(brcaRna_geneExpression_F_efit<span class="op">$</span>Amean, <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>), <span class="dv">2</span>),</span>
<span id="cb134-12"><a href="brca-rnaseq-preproc.html#cb134-12"></a>      <span class="dt">collapse =</span> <span class="st">&quot;, &quot;</span></span>
<span id="cb134-13"><a href="brca-rnaseq-preproc.html#cb134-13"></a>    )</span>
<span id="cb134-14"><a href="brca-rnaseq-preproc.html#cb134-14"></a>  ),</span>
<span id="cb134-15"><a href="brca-rnaseq-preproc.html#cb134-15"></a>  <span class="dt">ylab =</span> <span class="kw">paste0</span>(</span>
<span id="cb134-16"><a href="brca-rnaseq-preproc.html#cb134-16"></a>    <span class="st">&quot;log-fold-change: IQR=&quot;</span>,</span>
<span id="cb134-17"><a href="brca-rnaseq-preproc.html#cb134-17"></a>    <span class="kw">paste</span>(<span class="kw">round</span>(<span class="kw">quantile</span>(brcaRna_geneExpression_F_efit<span class="op">$</span>coefficients[, <span class="st">&#39;LumAvsOther&#39;</span>], <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>), <span class="dv">2</span>),</span>
<span id="cb134-18"><a href="brca-rnaseq-preproc.html#cb134-18"></a>      <span class="dt">collapse =</span> <span class="st">&quot;, &quot;</span></span>
<span id="cb134-19"><a href="brca-rnaseq-preproc.html#cb134-19"></a>    )</span>
<span id="cb134-20"><a href="brca-rnaseq-preproc.html#cb134-20"></a>  ),</span>
<span id="cb134-21"><a href="brca-rnaseq-preproc.html#cb134-21"></a>  <span class="dt">legend =</span> F</span>
<span id="cb134-22"><a href="brca-rnaseq-preproc.html#cb134-22"></a>)</span>
<span id="cb134-23"><a href="brca-rnaseq-preproc.html#cb134-23"></a><span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">0</span>, <span class="dt">col =</span> <span class="st">&quot;black&quot;</span>)</span>
<span id="cb134-24"><a href="brca-rnaseq-preproc.html#cb134-24"></a><span class="kw">rug</span>(<span class="kw">quantile</span>(brcaRna_geneExpression_F_efit<span class="op">$</span>coefficients[, <span class="st">&#39;LumAvsOther&#39;</span>], <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>),</span>
<span id="cb134-25"><a href="brca-rnaseq-preproc.html#cb134-25"></a>  <span class="dt">col =</span> <span class="st">&quot;purple&quot;</span>,</span>
<span id="cb134-26"><a href="brca-rnaseq-preproc.html#cb134-26"></a>  <span class="dt">ticksize =</span> <span class="fl">.03</span>, <span class="dt">side =</span> <span class="dv">2</span>, <span class="dt">lwd =</span> <span class="dv">2</span></span>
<span id="cb134-27"><a href="brca-rnaseq-preproc.html#cb134-27"></a>)</span>
<span id="cb134-28"><a href="brca-rnaseq-preproc.html#cb134-28"></a><span class="kw">rug</span>(<span class="kw">quantile</span>(brcaRna_geneExpression_F_efit<span class="op">$</span>Amean, <span class="dt">prob =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>) <span class="op">/</span><span class="st"> </span><span class="dv">4</span>),</span>
<span id="cb134-29"><a href="brca-rnaseq-preproc.html#cb134-29"></a>  <span class="dt">col =</span> <span class="st">&quot;purple&quot;</span>,</span>
<span id="cb134-30"><a href="brca-rnaseq-preproc.html#cb134-30"></a>  <span class="dt">ticksize =</span> <span class="fl">.03</span>, <span class="dt">side =</span> <span class="dv">1</span>, <span class="dt">lwd =</span> <span class="dv">2</span></span>
<span id="cb134-31"><a href="brca-rnaseq-preproc.html#cb134-31"></a>)</span></code></pre></div>
<div class="figure"><span id="fig:brcaRna-preproc-mdPlotTfit"></span>
<img src="Static/figures/brcaRna-preproc-mdPlotTfit-1.png" alt="LumA vs Other - Identified Genes at FDR = 0,05 and logFC &gt; 10%" width="1056" />
<p class="caption">
Figure 6.6: LumA vs Other - Identified Genes at FDR = 0,05 and logFC &gt; 10%
</p>
</div>
<!-- 5hmc comment - does not apply to this RNA-Seq dataset
As noted above, the log-fold-change distribution for the up-represented genes
is long-tailes in  comparison to log-fold-change distribution for the down-represented genes.
As a result fewer down-represented than up-regulated genes are identified when a 
minimum log-FC requirement is imposed.
-->
</div>
</div>
<div id="breast-rnaseq-snr-regime" class="section level2" number="6.3">
<h2><span class="header-section-number">6.3</span> Signal-to-noise ratio regime</h2>
<p>In Hastie et al. (2017) <span class="citation">[<a href="references.html#ref-Hastie:2017aa" role="doc-biblioref">7</a>]</span>) results from <code>lasso</code> fits are
compared with <code>best subset</code> and <code>forward selection</code> fits and it is argued
that while <code>best subset</code> is optimal for high signal-to-noise regimes,
the lasso gains some competitive advantage when the prevailing signal-to-noise
ratio of the dataset is lowered.</p>
<!-- DNA ???
We can extract sigma and signal from the fit objects to get SNR values for each gene
to see in what SNR regime the 5hmC gene body data are.

-->
<!-- DEBUG BCV from Section \@ref(analysis-of-coverage-variability) vs CV
pairs(cbind(BCV_mtx, CV)) 
boxplot(cbind(BCV_mtx, CV), outline=F) 
-->
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="brca-rnaseq-preproc.html#cb135-1"></a><span class="co">###  CLEAR CACHE</span></span>
<span id="cb135-2"><a href="brca-rnaseq-preproc.html#cb135-2"></a></span>
<span id="cb135-3"><a href="brca-rnaseq-preproc.html#cb135-3"></a>Effect &lt;-<span class="st"> </span><span class="kw">abs</span>(brcaRna_geneExpression_F_efit<span class="op">$</span>coefficients[,<span class="st">&#39;LumAvsOther&#39;</span>])</span>
<span id="cb135-4"><a href="brca-rnaseq-preproc.html#cb135-4"></a>Noise &lt;-<span class="st"> </span>brcaRna_geneExpression_F_efit<span class="op">$</span>sigma</span>
<span id="cb135-5"><a href="brca-rnaseq-preproc.html#cb135-5"></a>SNR &lt;-<span class="st"> </span>Effect<span class="op">/</span>Noise</span>
<span id="cb135-6"><a href="brca-rnaseq-preproc.html#cb135-6"></a></span>
<span id="cb135-7"><a href="brca-rnaseq-preproc.html#cb135-7"></a><span class="kw">plot</span>(spatstat<span class="op">::</span><span class="kw">CDF</span>(<span class="kw">density</span>(SNR)),</span>
<span id="cb135-8"><a href="brca-rnaseq-preproc.html#cb135-8"></a>  <span class="dt">col =</span> <span class="dv">1</span>, <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">ylab =</span> <span class="st">&quot;Prob(SNR&lt;x)&quot;</span>,</span>
<span id="cb135-9"><a href="brca-rnaseq-preproc.html#cb135-9"></a>  <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">1.5</span>)</span>
<span id="cb135-10"><a href="brca-rnaseq-preproc.html#cb135-10"></a>)</span>
<span id="cb135-11"><a href="brca-rnaseq-preproc.html#cb135-11"></a></span>
<span id="cb135-12"><a href="brca-rnaseq-preproc.html#cb135-12"></a>SNR_quant &lt;-<span class="st"> </span><span class="kw">quantile</span>(SNR, <span class="dt">prob=</span><span class="kw">c</span>((<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>)<span class="op">/</span><span class="dv">4</span>,.<span class="dv">9</span>))</span>
<span id="cb135-13"><a href="brca-rnaseq-preproc.html#cb135-13"></a><span class="kw">rug</span>(SNR_quant,</span>
<span id="cb135-14"><a href="brca-rnaseq-preproc.html#cb135-14"></a>    <span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">ticksize =</span> <span class="fl">0.05</span>, <span class="dt">col =</span> <span class="dv">1</span></span>
<span id="cb135-15"><a href="brca-rnaseq-preproc.html#cb135-15"></a>  )</span></code></pre></div>
<div class="figure"><span id="fig:brcaRna-preproc-plotSNR"></span>
<img src="Static/figures/brcaRna-preproc-plotSNR-1.png" alt="Cumulative Distribution of SNR - rug = 25, 50, 75 and 90th percentile" width="960" />
<p class="caption">
Figure 6.7: Cumulative Distribution of SNR - rug = 25, 50, 75 and 90th percentile
</p>
</div>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb136-1"><a href="brca-rnaseq-preproc.html#cb136-1"></a>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">t</span>(SNR_quant),</span>
<span id="cb136-2"><a href="brca-rnaseq-preproc.html#cb136-2"></a>  <span class="dt">digits =</span> <span class="dv">3</span>,</span>
<span id="cb136-3"><a href="brca-rnaseq-preproc.html#cb136-3"></a>  <span class="dt">caption =</span> <span class="kw">paste</span>(</span>
<span id="cb136-4"><a href="brca-rnaseq-preproc.html#cb136-4"></a>    <span class="st">&quot;SNR Quantiles&quot;</span>) </span>
<span id="cb136-5"><a href="brca-rnaseq-preproc.html#cb136-5"></a>) <span class="op">%&gt;%</span><span class="st"> </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:brcaRna-preproc-plotSNR">Table 6.4: </span>SNR Quantiles
</caption>
<thead>
<tr>
<th style="text-align:right;">
25%
</th>
<th style="text-align:right;">
50%
</th>
<th style="text-align:right;">
75%
</th>
<th style="text-align:right;">
90%
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
0.094
</td>
<td style="text-align:right;">
0.217
</td>
<td style="text-align:right;">
0.384
</td>
<td style="text-align:right;">
0.56
</td>
</tr>
</tbody>
</table>
<!--
These SNR values are in the range where the lasso and relaxed lasso gain some advantage over
best subset and forward selection fits (see  Hastie et al. (2017) [@Hastie:2017aa]).
-->

</div>
</div>
<div class="footnotes">
<hr />
<ol start="1">
<li id="fn1"><p>See <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE96058">NCBI GEO GSE96058</a><a href="brca-rnaseq-preproc.html#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="hcc-5hmcseq-model-suite.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="brca-rnaseseq-explore-sparsity.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["HCC5hmCExplore.pdf", "HCC5hmCExplore.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
