<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Section 9 Conclusions | DNA Hydroxymethylation in Hepatocellular Carcinoma</title>
  <meta name="description" content="Data from Cai et al. (2019) paper are explored" />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Section 9 Conclusions | DNA Hydroxymethylation in Hepatocellular Carcinoma" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Data from Cai et al. (2019) paper are explored" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Section 9 Conclusions | DNA Hydroxymethylation in Hepatocellular Carcinoma" />
  
  <meta name="twitter:description" content="Data from Cai et al. (2019) paper are explored" />
  

<meta name="author" content="Francois Collin" />


<meta name="date" content="2020-01-01" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="brca-rnaseq-model-suite.html"/>
<link rel="next" href="references.html"/>
<script src="libs/header-attrs-2.2/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<script type="text/javascript">
// source:https://stackoverflow.com/questions/45360998/code-folding-in-bookdown
// toggle visibility of R source blocks in R Markdown output
function toggle_R() {
  var x = document.getElementsByClassName('r');
  if (x.length == 0) return;
  function toggle_vis(o) {
    var d = o.style.display;
    o.style.display = (d == 'block' || d == '') ? 'none':'block';
  }

  for (i = 0; i < x.length; i++) {
    var y = x[i];
    if (y.tagName.toLowerCase() === 'pre') toggle_vis(y);
  }

    var elem = document.getElementById("myButton1");
    if (elem.value === "Hide Global") elem.value = "Show Global";
    else elem.value = "Hide Global";
}

document.write('<input onclick="toggle_R();" type="button" value="Hide Global" id="myButton1" style="position: absolute; top: 10%; right: 2%; z-index: 200"></input>')

</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">DNA Hydroxymethylation in Hepatocellular Carcinoma</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preamble</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i>License</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="modeling-background.html"><a href="modeling-background.html"><i class="fa fa-check"></i><b>2</b> Classification Analysis in the <code>n &lt;&lt; p</code> Context</a>
<ul>
<li class="chapter" data-level="2.1" data-path="modeling-background.html"><a href="modeling-background.html#caret-for-model-evaluation"><i class="fa fa-check"></i><b>2.1</b> caret for model evaluation</a></li>
<li class="chapter" data-level="2.2" data-path="modeling-background.html"><a href="modeling-background.html#the-glmnet-r-package"><i class="fa fa-check"></i><b>2.2</b> The <code>glmnet</code> R package</a>
<ul>
<li><a href="modeling-background.html#alpha-hyper-parameter"><strong>alpha</strong> hyper-parameter</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="modeling-background.html"><a href="modeling-background.html#signal-to-noise"><i class="fa fa-check"></i><b>2.3</b> Signal-to-noise Ratio</a></li>
<li class="chapter" data-level="2.4" data-path="modeling-background.html"><a href="modeling-background.html#lasso-vs-best-sub"><i class="fa fa-check"></i><b>2.4</b> Lasso vs Best Subset</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html"><i class="fa fa-check"></i><b>3</b> HCC 5hmC-Seq : Preprocessing</a>
<ul>
<li class="chapter" data-level="3.1" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#load-the-data"><i class="fa fa-check"></i><b>3.1</b> Load the data</a></li>
<li class="chapter" data-level="3.2" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#dra"><i class="fa fa-check"></i><b>3.2</b> Differential representation analysis</a>
<ul>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#remove-lowly-expressed-genes"><i class="fa fa-check"></i>Remove lowly expressed genes</a></li>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#creating-a-design-matrix-and-contrasts"><i class="fa fa-check"></i>Creating a design matrix and contrasts</a></li>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#removing-heteroscedasticity-from-the-count-data"><i class="fa fa-check"></i>Removing heteroscedasticity from the count data</a></li>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#fit-linear-models-and-examine-the-results"><i class="fa fa-check"></i>Fit linear models and examine the results</a></li>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#graphical-representations-of-de-results-md-plots"><i class="fa fa-check"></i>Graphical representations of DE results: MD Plots</a></li>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#de-genes-at-10-fold-change"><i class="fa fa-check"></i>DE genes at 10% fold change</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="hcc-5hmcseq-preproc.html"><a href="hcc-5hmcseq-preproc.html#snr-regime"><i class="fa fa-check"></i><b>3.3</b> Signal-to-noise ratio regime</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html"><i class="fa fa-check"></i><b>4</b> HCC 5hmC-Seq: Exploring Sparsity</a>
<ul>
<li class="chapter" data-level="4.1" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#cross-validation-analysis-setup"><i class="fa fa-check"></i><b>4.1</b> Cross-validation analysis setup</a></li>
<li class="chapter" data-level="4.2" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#fit-and-compare-models"><i class="fa fa-check"></i><b>4.2</b> Fit and compare models</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#logistic-regression-in-glmnet"><i class="fa fa-check"></i><b>4.2.1</b> Logistic regression in <code>glmnet</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#the-relaxed-lasso-and-blended-mix-models"><i class="fa fa-check"></i><b>4.3</b> The relaxed lasso and blended mix models</a></li>
<li class="chapter" data-level="4.4" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#examination-of-sensitivity-vs-specificity"><i class="fa fa-check"></i><b>4.4</b> Examination of sensitivity vs specificity</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#training-data-out-of-fold-roc-curves"><i class="fa fa-check"></i><b>4.4.1</b> Training data out-of-fold ROC curves</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#compare-predictions-at-misclassified-samples"><i class="fa fa-check"></i><b>4.5</b> Compare predictions at misclassified samples</a></li>
<li class="chapter" data-level="4.6" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#compare-coefficient-profiles"><i class="fa fa-check"></i><b>4.6</b> Compare coefficient profiles</a></li>
<li class="chapter" data-level="4.7" data-path="hcc-5hmcseq-explore-sparsity.html"><a href="hcc-5hmcseq-explore-sparsity.html#examine-feature-selection"><i class="fa fa-check"></i><b>4.7</b> Examine feature selection</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html"><i class="fa fa-check"></i><b>5</b> HCC 5hmC-Seq: Sample size investigation</a>
<ul>
<li class="chapter" data-level="5.1" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#full-data-set-fit"><i class="fa fa-check"></i><b>5.1</b> Full data set fit</a>
<ul>
<li class="chapter" data-level="" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#get-sample-consistency-scores"><i class="fa fa-check"></i>Get sample consistency scores</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#selected-feature-list-stability"><i class="fa fa-check"></i><b>5.2</b> Selected feature list stability</a></li>
<li class="chapter" data-level="5.3" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#simulation-design"><i class="fa fa-check"></i><b>5.3</b> Simulation Design</a></li>
<li class="chapter" data-level="5.4" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#setup-simulation"><i class="fa fa-check"></i><b>5.4</b> Setup simulation</a></li>
<li class="chapter" data-level="5.5" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#run-simulations"><i class="fa fa-check"></i><b>5.5</b> Run simulations</a></li>
<li class="chapter" data-level="5.6" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#simulation-results"><i class="fa fa-check"></i><b>5.6</b> Simulation results</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#results-for-one-simulation"><i class="fa fa-check"></i><b>5.6.1</b> Results for one simulation</a></li>
<li class="chapter" data-level="5.6.2" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#results-across-simulation-runs."><i class="fa fa-check"></i><b>5.6.2</b> Results across simulation runs.</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="hcc-5hmcseq-model-suite.html"><a href="hcc-5hmcseq-model-suite.html#effect-of-sample-consistency"><i class="fa fa-check"></i><b>5.7</b> Effect of sample consistency</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html"><i class="fa fa-check"></i><b>6</b> BrCa RNA-Seq: Preprocessing</a>
<ul>
<li class="chapter" data-level="6.1" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#load-the-data-1"><i class="fa fa-check"></i><b>6.1</b> Load the data</a></li>
<li class="chapter" data-level="6.2" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#breast-rnaseq-dra"><i class="fa fa-check"></i><b>6.2</b> Differential representation analysis</a>
<ul>
<li class="chapter" data-level="" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#remove-lowly-expressed-genes-1"><i class="fa fa-check"></i>Remove lowly expressed genes</a></li>
<li class="chapter" data-level="" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#creating-a-design-matrix-and-contrasts-1"><i class="fa fa-check"></i>Creating a design matrix and contrasts</a></li>
<li class="chapter" data-level="" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#fit-linear-models-and-examine-the-results-1"><i class="fa fa-check"></i>Fit linear models and examine the results</a></li>
<li class="chapter" data-level="" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#graphical-representations-of-de-results-md-plots-1"><i class="fa fa-check"></i>Graphical representations of DE results: MD Plots</a></li>
<li class="chapter" data-level="" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#de-genes-at-10-fold-change-1"><i class="fa fa-check"></i>DE genes at 10% fold change</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="brca-rnaseq-preproc.html"><a href="brca-rnaseq-preproc.html#breast-rnaseq-snr-regime"><i class="fa fa-check"></i><b>6.3</b> Signal-to-noise ratio regime</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html"><i class="fa fa-check"></i><b>7</b> BrCa RNA-Seq: Exploring Sparsity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#cross-validation-analysis-setup-1"><i class="fa fa-check"></i><b>7.1</b> Cross-validation analysis setup</a></li>
<li class="chapter" data-level="7.2" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#fit-and-compare-models-1"><i class="fa fa-check"></i><b>7.2</b> Fit and compare models</a></li>
<li class="chapter" data-level="7.3" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#the-relaxed-lasso-and-blended-mix-models-1"><i class="fa fa-check"></i><b>7.3</b> The relaxed lasso and blended mix models</a></li>
<li class="chapter" data-level="7.4" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#examination-of-sensitivity-vs-specificity-1"><i class="fa fa-check"></i><b>7.4</b> Examination of sensitivity vs specificity</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#training-data-out-of-fold-roc-curves-1"><i class="fa fa-check"></i><b>7.4.1</b> Training data out-of-fold ROC curves</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#compare-predictions-at-misclassified-samples-1"><i class="fa fa-check"></i><b>7.5</b> Compare predictions at misclassified samples</a></li>
<li class="chapter" data-level="7.6" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#compare-coefficient-profiles-1"><i class="fa fa-check"></i><b>7.6</b> Compare coefficient profiles</a></li>
<li class="chapter" data-level="7.7" data-path="brca-rnaseseq-explore-sparsity.html"><a href="brca-rnaseseq-explore-sparsity.html#examine-feature-selection-1"><i class="fa fa-check"></i><b>7.7</b> Examine feature selection</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html"><i class="fa fa-check"></i><b>8</b> BrCa RNA-Seq: Sample size investigation</a>
<ul>
<li class="chapter" data-level="8.1" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#full-data-set-fit-1"><i class="fa fa-check"></i><b>8.1</b> Full data set fit</a>
<ul>
<li class="chapter" data-level="" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#get-sample-consistency-scores-1"><i class="fa fa-check"></i>Get sample consistency scores</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#selected-feature-list-stability-1"><i class="fa fa-check"></i><b>8.2</b> Selected feature list stability</a></li>
<li class="chapter" data-level="8.3" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#simulation-design-1"><i class="fa fa-check"></i><b>8.3</b> Simulation Design</a></li>
<li class="chapter" data-level="8.4" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#setup-simulation-1"><i class="fa fa-check"></i><b>8.4</b> Setup simulation</a></li>
<li class="chapter" data-level="8.5" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#run-simulations-1"><i class="fa fa-check"></i><b>8.5</b> Run simulations</a></li>
<li class="chapter" data-level="8.6" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#simulation-results-1"><i class="fa fa-check"></i><b>8.6</b> Simulation results</a>
<ul>
<li class="chapter" data-level="8.6.1" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#results-for-one-simulation-1"><i class="fa fa-check"></i><b>8.6.1</b> Results for one simulation</a></li>
<li class="chapter" data-level="8.6.2" data-path="brca-rnaseq-model-suite.html"><a href="brca-rnaseq-model-suite.html#results-across-simulation-runs.-1"><i class="fa fa-check"></i><b>8.6.2</b> Results across simulation runs.</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="conclusions.html"><a href="conclusions.html"><i class="fa fa-check"></i><b>9</b> Conclusions</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html"><i class="fa fa-check"></i>Appendix 1 - Sample size in elastic net fits to HCC 5hmC-Seq Data</a>
<ul>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html#selected-feature-list-stability-2"><i class="fa fa-check"></i>Selected feature list stability</a></li>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html#run-simulations---enet"><i class="fa fa-check"></i>Run simulations - enet</a></li>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html#enet-simulation-results"><i class="fa fa-check"></i>enet Simulation results</a>
<ul>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html#simulation-results---look-at-one-simulation"><i class="fa fa-check"></i>Simulation Results - look at one simulation</a></li>
<li class="chapter" data-level="" data-path="appendix-1.html"><a href="appendix-1.html#summarize-results-across-simulation-runs"><i class="fa fa-check"></i>Summarize results across simulation runs</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="LICENSE.md" style="font:em;" target="blank">Copyright (c) 2020 Francois Collin</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">DNA Hydroxymethylation in Hepatocellular Carcinoma</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="conclusions" class="section level1" number="9">
<h1><span class="header-section-number">Section 9</span> Conclusions</h1>
<p>We have found that …</p>
<p>Other questions …</p>

</div>
            </section>

          </div>
        </div>
      </div>
<a href="brca-rnaseq-model-suite.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["HCC5hmCExplore.pdf", "HCC5hmCExplore.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
