<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Section 5 Fitted Model Suite | DNA Hydroxymethylation in Hepatocellular Carcinoma</title>
  <meta name="description" content="Data from Cai et al. (2019) paper are explored" />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Section 5 Fitted Model Suite | DNA Hydroxymethylation in Hepatocellular Carcinoma" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Data from Cai et al. (2019) paper are explored" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Section 5 Fitted Model Suite | DNA Hydroxymethylation in Hepatocellular Carcinoma" />
  
  <meta name="twitter:description" content="Data from Cai et al. (2019) paper are explored" />
  

<meta name="author" content="Francois Collin" />


<meta name="date" content="2020-09-09" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="explore-sparsity.html"/>
<link rel="next" href="variable-importance.html"/>
<script src="libs/header-attrs-2.2/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<script type="text/javascript">
// source:https://stackoverflow.com/questions/45360998/code-folding-in-bookdown
// toggle visibility of R source blocks in R Markdown output
function toggle_R() {
  var x = document.getElementsByClassName('r');
  if (x.length == 0) return;
  function toggle_vis(o) {
    var d = o.style.display;
    o.style.display = (d == 'block' || d == '') ? 'none':'block';
  }

  for (i = 0; i < x.length; i++) {
    var y = x[i];
    if (y.tagName.toLowerCase() === 'pre') toggle_vis(y);
  }

    var elem = document.getElementById("myButton1");
    if (elem.value === "Hide Global") elem.value = "Show Global";
    else elem.value = "Hide Global";
}

document.write('<input onclick="toggle_R();" type="button" value="Hide Global" id="myButton1" style="position: absolute; top: 10%; right: 2%; z-index: 200"></input>')

</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">DNA Hydroxymethylation in Hepatocellular Carcinoma</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preamble</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i>License</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="modeling-background.html"><a href="modeling-background.html"><i class="fa fa-check"></i><b>2</b> Modeling - Background</a>
<ul>
<li class="chapter" data-level="2.1" data-path="modeling-background.html"><a href="modeling-background.html#predictive-modeling-for-genomic-data"><i class="fa fa-check"></i><b>2.1</b> Predictive modeling for genomic data</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="modeling-background.html"><a href="modeling-background.html#caret-for-model-evaluation"><i class="fa fa-check"></i><b>2.1.1</b> caret for model evaluation</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="modeling-background.html"><a href="modeling-background.html#glmnet"><i class="fa fa-check"></i><b>2.2</b> glmnet</a>
<ul>
<li><a href="modeling-background.html#alpha-hyper-parameter"><strong>alpha</strong> hyper-parameter</a></li>
<li class="chapter" data-level="" data-path="modeling-background.html"><a href="modeling-background.html#signal-to-noise-ratio"><i class="fa fa-check"></i>Signal-to-noise Ratio</a></li>
<li class="chapter" data-level="" data-path="modeling-background.html"><a href="modeling-background.html#lasso-vs-best-subset"><i class="fa fa-check"></i>Lasso vs Best Subset</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preproc.html"><a href="preproc.html"><i class="fa fa-check"></i><b>3</b> Preprocessing</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preproc.html"><a href="preproc.html#load-the-data"><i class="fa fa-check"></i><b>3.1</b> Load the data</a></li>
<li class="chapter" data-level="3.2" data-path="preproc.html"><a href="preproc.html#dra"><i class="fa fa-check"></i><b>3.2</b> Differential representation analysis</a>
<ul>
<li class="chapter" data-level="" data-path="preproc.html"><a href="preproc.html#remove-lowly-expressed-genes"><i class="fa fa-check"></i>Remove lowly expressed genes</a></li>
<li class="chapter" data-level="" data-path="preproc.html"><a href="preproc.html#creating-a-design-matrix-and-contrasts"><i class="fa fa-check"></i>Creating a design matrix and contrasts</a></li>
<li class="chapter" data-level="" data-path="preproc.html"><a href="preproc.html#removing-heteroscedasticity-from-the-count-data"><i class="fa fa-check"></i>Removing heteroscedasticity from the count data</a></li>
<li class="chapter" data-level="" data-path="preproc.html"><a href="preproc.html#fit-linear-models-and-examine-the-results"><i class="fa fa-check"></i>Fit linear models and examine the results</a></li>
<li class="chapter" data-level="" data-path="preproc.html"><a href="preproc.html#graphical-representations-of-de-results-md-plots"><i class="fa fa-check"></i>Graphical representations of DE results: MD Plots</a></li>
<li class="chapter" data-level="" data-path="preproc.html"><a href="preproc.html#de-genes-at-10-fold-change"><i class="fa fa-check"></i>DE genes at 10% fold change</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="preproc.html"><a href="preproc.html#snr-regime"><i class="fa fa-check"></i><b>3.3</b> Signal-to-noise ratio regime</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="explore-sparsity.html"><a href="explore-sparsity.html"><i class="fa fa-check"></i><b>4</b> The bet on sparsity</a>
<ul>
<li class="chapter" data-level="4.1" data-path="explore-sparsity.html"><a href="explore-sparsity.html#cv-analysis-setup"><i class="fa fa-check"></i><b>4.1</b> CV analysis setup</a></li>
<li class="chapter" data-level="4.2" data-path="explore-sparsity.html"><a href="explore-sparsity.html#fit-and-compare-models"><i class="fa fa-check"></i><b>4.2</b> Fit and compare models</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="explore-sparsity.html"><a href="explore-sparsity.html#logistic-regression-in-glmnet"><i class="fa fa-check"></i><b>4.2.1</b> Logistic regression in <code>glmnet</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="explore-sparsity.html"><a href="explore-sparsity.html#relaxed-lasso-and-blended-mix"><i class="fa fa-check"></i><b>4.3</b> Relaxed lasso and blended mix</a></li>
<li class="chapter" data-level="4.4" data-path="explore-sparsity.html"><a href="explore-sparsity.html#examination-of-sensitivity-vs-specificity"><i class="fa fa-check"></i><b>4.4</b> Examination of sensitivity vs specificity</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="explore-sparsity.html"><a href="explore-sparsity.html#training-data-out-of-fold-roc-curves"><i class="fa fa-check"></i><b>4.4.1</b> Training data out-of-fold ROC curves</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="explore-sparsity.html"><a href="explore-sparsity.html#compare-predictions-at-misclassified-samples"><i class="fa fa-check"></i><b>4.5</b> Compare predictions at misclassified samples</a></li>
<li class="chapter" data-level="4.6" data-path="explore-sparsity.html"><a href="explore-sparsity.html#compare-coefficient-profiles"><i class="fa fa-check"></i><b>4.6</b> Compare coefficient profiles</a></li>
<li class="chapter" data-level="4.7" data-path="explore-sparsity.html"><a href="explore-sparsity.html#examine-feature-selection"><i class="fa fa-check"></i><b>4.7</b> Examine feature selection</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="model-suite.html"><a href="model-suite.html"><i class="fa fa-check"></i><b>5</b> Fitted Model Suite</a>
<ul>
<li class="chapter" data-level="5.1" data-path="model-suite.html"><a href="model-suite.html#sample-quality-scores"><i class="fa fa-check"></i><b>5.1</b> Sample quality scores</a></li>
<li class="chapter" data-level="5.2" data-path="model-suite.html"><a href="model-suite.html#simulation-design"><i class="fa fa-check"></i><b>5.2</b> Simulation Design</a></li>
<li class="chapter" data-level="5.3" data-path="model-suite.html"><a href="model-suite.html#setup-simulation"><i class="fa fa-check"></i><b>5.3</b> Setup simulation</a></li>
<li class="chapter" data-level="5.4" data-path="model-suite.html"><a href="model-suite.html#run-simulations"><i class="fa fa-check"></i><b>5.4</b> Run simulations</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="variable-importance.html"><a href="variable-importance.html"><i class="fa fa-check"></i><b>6</b> Variable importance</a></li>
<li class="chapter" data-level="7" data-path="conclusions.html"><a href="conclusions.html"><i class="fa fa-check"></i><b>7</b> Conclusions</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="LICENSE.md" style="font:em;" target="blank">Copyright (c) 2020 Francois Collin</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">DNA Hydroxymethylation in Hepatocellular Carcinoma</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="model-suite" class="section level1" number="5">
<h1><span class="header-section-number">Section 5</span> Fitted Model Suite</h1>
<p>We now examine the results of fitting a suite of models to
investigate the effect of sample size on
various aspects of model performance:</p>
<ul>
<li><p>assessed accuracy: out-of-fold estimates of precision and variability and
cv assessed accuracy and bias.</p></li>
<li><p>selected feature profile stability - to what extent does the
feature set implicitly selected by the lasso vary across random
sampling and what is the effect of sample size.</p></li>
</ul>
<p>It is hypothesized that below a certain threshold,
sample sizes are too small to provide reliable estimates
of performance or stable selected feature profiles.</p>
<p>We will attempt to separate variability which is due to
sample size from variability due to sample composition.
To to this we will track sample quality.
Predicted probabilities from fitted model can be transformed into sample
quality scores: <span class="math inline">\(Q_i = p_i^{y_i}(1-p_i)^{1-y_i}\)</span>, where <span class="math inline">\(p_i\)</span> is the
estimated probability of HCC for sample i and <span class="math inline">\(y_i\)</span> is 1 for HCC samples and
0 for Controls. ie. we use the fitted sample contribution to the
likelihood function as a sample quality score. To derive the quality scores,
we will use the predicted response from a lasso model fitted to the entire data set.</p>
<p>Hard to classify samples will have low quality scores.
In the results that we discuss below, when we look at variability across repeated
random sampling of different sizes, we can use sample quality scores to investigate
how much of the variability is due to sample selection.
Note that quality here is not used to say anything about the sample data quality.
Low quality here only means that a sample is different from the
core of the data set in a way that makes it hard to properly classify.
That could happen if the sample were mislabeled, in which case we could
think of this sample as being poor quality of course.</p>
<div id="sample-quality-scores" class="section level2" number="5.1">
<h2><span class="header-section-number">5.1</span> Sample quality scores</h2>
<p>To get sample quality scores, fit a lasso model, extracted predicted probabilities and
convert to a quality score.</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="model-suite.html#cb77-1"></a><span class="co"># combine train and test </span></span>
<span id="cb77-2"><a href="model-suite.html#cb77-2"></a>all_lcpm_mtx &lt;-<span class="st"> </span><span class="kw">rbind</span>(train_lcpm_mtx, test_lcpm_mtx)</span>
<span id="cb77-3"><a href="model-suite.html#cb77-3"></a></span>
<span id="cb77-4"><a href="model-suite.html#cb77-4"></a><span class="co"># we have to be careful with factors!</span></span>
<span id="cb77-5"><a href="model-suite.html#cb77-5"></a><span class="co"># We&#39;ll keep as a character and change to factor when needed</span></span>
<span id="cb77-6"><a href="model-suite.html#cb77-6"></a>all_group_vec &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb77-7"><a href="model-suite.html#cb77-7"></a> <span class="kw">as.character</span>(train_group_vec), </span>
<span id="cb77-8"><a href="model-suite.html#cb77-8"></a> <span class="kw">as.character</span>(test_group_vec)</span>
<span id="cb77-9"><a href="model-suite.html#cb77-9"></a>)</span>
<span id="cb77-10"><a href="model-suite.html#cb77-10"></a><span class="co"># I suspect adding names to vectors breaks one of the tidy commandments,</span></span>
<span id="cb77-11"><a href="model-suite.html#cb77-11"></a><span class="co"># but then again I am sure I have already offended the creed beyond salvation</span></span>
<span id="cb77-12"><a href="model-suite.html#cb77-12"></a><span class="kw">names</span>(all_group_vec) &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb77-13"><a href="model-suite.html#cb77-13"></a> <span class="kw">names</span>(train_group_vec),</span>
<span id="cb77-14"><a href="model-suite.html#cb77-14"></a> <span class="kw">names</span>(test_group_vec)</span>
<span id="cb77-15"><a href="model-suite.html#cb77-15"></a>)</span>
<span id="cb77-16"><a href="model-suite.html#cb77-16"></a></span>
<span id="cb77-17"><a href="model-suite.html#cb77-17"></a>start_time &lt;-<span class="st">  </span><span class="kw">proc.time</span>()</span>
<span id="cb77-18"><a href="model-suite.html#cb77-18"></a></span>
<span id="cb77-19"><a href="model-suite.html#cb77-19"></a><span class="co"># since we have not set the foldid, set seed here</span></span>
<span id="cb77-20"><a href="model-suite.html#cb77-20"></a><span class="kw">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb77-21"><a href="model-suite.html#cb77-21"></a>cv_lassoAll &lt;-<span class="st"> </span>glmnet<span class="op">::</span><span class="kw">cv.glmnet</span>(</span>
<span id="cb77-22"><a href="model-suite.html#cb77-22"></a> <span class="dt">x =</span> all_lcpm_mtx,</span>
<span id="cb77-23"><a href="model-suite.html#cb77-23"></a> <span class="dt">y =</span> <span class="kw">factor</span>(all_group_vec,<span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&#39;Control&#39;</span>, <span class="st">&#39;HCC&#39;</span>)),</span>
<span id="cb77-24"><a href="model-suite.html#cb77-24"></a> <span class="dt">alpha =</span> <span class="dv">1</span>,</span>
<span id="cb77-25"><a href="model-suite.html#cb77-25"></a> <span class="dt">family =</span> <span class="st">&#39;binomial&#39;</span>,</span>
<span id="cb77-26"><a href="model-suite.html#cb77-26"></a> <span class="dt">type.measure  =</span>  <span class="st">&quot;class&quot;</span>,</span>
<span id="cb77-27"><a href="model-suite.html#cb77-27"></a> <span class="dt">keep =</span> F,</span>
<span id="cb77-28"><a href="model-suite.html#cb77-28"></a> <span class="dt">nlambda =</span> <span class="dv">100</span></span>
<span id="cb77-29"><a href="model-suite.html#cb77-29"></a>)</span>
<span id="cb77-30"><a href="model-suite.html#cb77-30"></a></span>
<span id="cb77-31"><a href="model-suite.html#cb77-31"></a><span class="kw">message</span>(<span class="st">&quot;lassoAll time: &quot;</span>, <span class="kw">round</span>((<span class="kw">proc.time</span>() <span class="op">-</span><span class="st"> </span>start_time)[<span class="dv">3</span>],<span class="dv">2</span>),<span class="st">&quot;s&quot;</span>)</span></code></pre></div>
<pre><code>## lassoAll time: 27.34s</code></pre>
<p>We can examine the fit.</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="model-suite.html#cb79-1"></a>cv_lassoAll</span></code></pre></div>
<pre><code>## 
## Call:  glmnet::cv.glmnet(x = all_lcpm_mtx, y = factor(all_group_vec,      levels = c(&quot;Control&quot;, &quot;HCC&quot;)), type.measure = &quot;class&quot;, keep = F,      alpha = 1, family = &quot;binomial&quot;, nlambda = 100) 
## 
## Measure: Misclassification Error 
## 
##      Lambda Measure       SE Nonzero
## min 0.01345 0.07127 0.007375     144
## 1se 0.01863 0.07802 0.007109      93</code></pre>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="model-suite.html#cb81-1"></a><span class="kw">plot</span>(cv_lassoAll)</span></code></pre></div>
<div class="figure"><span id="fig:plot-lassoAll"></span>
<img src="Static/figures/plot-lassoAll-1.png" alt="lasso model fitted to all samples" width="576" />
<p class="caption">
Figure 5.1: lasso model fitted to all samples
</p>
</div>
<p>More features are now selected, but the cv error rate is comparable to the results
of the lasso fit to the training set. We can use this fit to compute sample
quality scores. We will use the minimum lambda model to provide
the fitted probabilities.</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="model-suite.html#cb82-1"></a><span class="co"># predicted probs</span></span>
<span id="cb82-2"><a href="model-suite.html#cb82-2"></a>lassoAll_predResp_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb82-3"><a href="model-suite.html#cb82-3"></a> cv_lassoAll,</span>
<span id="cb82-4"><a href="model-suite.html#cb82-4"></a> <span class="dt">newx =</span> all_lcpm_mtx,</span>
<span id="cb82-5"><a href="model-suite.html#cb82-5"></a> <span class="dt">s =</span> <span class="st">&quot;lambda.min&quot;</span>,</span>
<span id="cb82-6"><a href="model-suite.html#cb82-6"></a> <span class="dt">type =</span> <span class="st">&quot;resp&quot;</span></span>
<span id="cb82-7"><a href="model-suite.html#cb82-7"></a> )</span>
<span id="cb82-8"><a href="model-suite.html#cb82-8"></a></span>
<span id="cb82-9"><a href="model-suite.html#cb82-9"></a><span class="co"># also get predicted class</span></span>
<span id="cb82-10"><a href="model-suite.html#cb82-10"></a>lassoAll_predClass_vec &lt;-<span class="st"> </span><span class="kw">predict</span>( </span>
<span id="cb82-11"><a href="model-suite.html#cb82-11"></a> cv_lassoAll,</span>
<span id="cb82-12"><a href="model-suite.html#cb82-12"></a> <span class="dt">newx =</span> all_lcpm_mtx,</span>
<span id="cb82-13"><a href="model-suite.html#cb82-13"></a> <span class="dt">s =</span> <span class="st">&quot;lambda.min&quot;</span>,</span>
<span id="cb82-14"><a href="model-suite.html#cb82-14"></a> <span class="dt">type =</span> <span class="st">&quot;class&quot;</span></span>
<span id="cb82-15"><a href="model-suite.html#cb82-15"></a> )</span>
<span id="cb82-16"><a href="model-suite.html#cb82-16"></a></span>
<span id="cb82-17"><a href="model-suite.html#cb82-17"></a><span class="co"># get qual scores</span></span>
<span id="cb82-18"><a href="model-suite.html#cb82-18"></a></span>
<span id="cb82-19"><a href="model-suite.html#cb82-19"></a><span class="co"># Note here that factors are a little but of a pain.</span></span>
<span id="cb82-20"><a href="model-suite.html#cb82-20"></a>y &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(all_group_vec <span class="op">==</span><span class="st"> &#39;HCC&#39;</span>)</span>
<span id="cb82-21"><a href="model-suite.html#cb82-21"></a>p &lt;-<span class="st"> </span>lassoAll_predResp_vec</span>
<span id="cb82-22"><a href="model-suite.html#cb82-22"></a>sample_qual_vec &lt;-<span class="st"> </span>p<span class="op">^</span>y<span class="op">*</span>(<span class="dv">1</span><span class="op">-</span>p)<span class="op">^</span>(<span class="dv">1</span><span class="op">-</span>y)</span>
<span id="cb82-23"><a href="model-suite.html#cb82-23"></a>sample_qual_vec &lt;-<span class="st"> </span>sample_qual_vec[,<span class="dv">1</span>] <span class="co"># to drop the matrix structure</span></span>
<span id="cb82-24"><a href="model-suite.html#cb82-24"></a></span>
<span id="cb82-25"><a href="model-suite.html#cb82-25"></a><span class="co"># Look at quality scores as a finction of classification</span></span>
<span id="cb82-26"><a href="model-suite.html#cb82-26"></a>lassoAll_conf_vec &lt;-<span class="st"> </span><span class="kw">paste</span>(</span>
<span id="cb82-27"><a href="model-suite.html#cb82-27"></a> y, </span>
<span id="cb82-28"><a href="model-suite.html#cb82-28"></a> <span class="kw">as.numeric</span>(lassoAll_predClass_vec<span class="op">==</span><span class="st">&#39;HCC&#39;</span>),</span>
<span id="cb82-29"><a href="model-suite.html#cb82-29"></a> <span class="dt">sep =</span> <span class="st">&#39;:&#39;</span></span>
<span id="cb82-30"><a href="model-suite.html#cb82-30"></a>)</span>
<span id="cb82-31"><a href="model-suite.html#cb82-31"></a></span>
<span id="cb82-32"><a href="model-suite.html#cb82-32"></a>lassoAll_conf_mtx &lt;-<span class="st"> </span><span class="kw">table</span>(</span>
<span id="cb82-33"><a href="model-suite.html#cb82-33"></a> <span class="dt">truth =</span> all_group_vec,</span>
<span id="cb82-34"><a href="model-suite.html#cb82-34"></a> <span class="dt">pred =</span> lassoAll_predClass_vec</span>
<span id="cb82-35"><a href="model-suite.html#cb82-35"></a>)</span></code></pre></div>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="model-suite.html#cb83-1"></a>knitr<span class="op">::</span><span class="kw">kable</span>(lassoAll_conf_mtx,</span>
<span id="cb83-2"><a href="model-suite.html#cb83-2"></a>  <span class="dt">caption =</span> <span class="st">&quot;Rows are truth.  Columns are predicted&quot;</span>) <span class="op">%&gt;%</span></span>
<span id="cb83-3"><a href="model-suite.html#cb83-3"></a><span class="st">   </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span></code></pre></div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:print-conf">Table 5.1: </span>Rows are truth. Columns are predicted
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Control
</th>
<th style="text-align:right;">
HCC
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Control
</td>
<td style="text-align:right;">
765
</td>
<td style="text-align:right;">
13
</td>
</tr>
<tr>
<td style="text-align:left;">
HCC
</td>
<td style="text-align:right;">
39
</td>
<td style="text-align:right;">
516
</td>
</tr>
</tbody>
</table>
<p>The model makes few errors (although in most liquid biopsy applications,
this error rate would arguably be too high). We can examine how quality
varies among classification groups.</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="model-suite.html#cb84-1"></a>gplots<span class="op">::</span><span class="kw">boxplot2</span>(<span class="kw">split</span>(sample_qual_vec, lassoAll_conf_vec), <span class="dt">ylab =</span> <span class="st">&#39;Quality Score&#39;</span>)</span>
<span id="cb84-2"><a href="model-suite.html#cb84-2"></a><span class="kw">title</span>(<span class="dt">sub =</span> <span class="st">&#39;Classification - Truth:Predicted&#39;</span>)</span></code></pre></div>
<div class="figure"><span id="fig:plot-qual-conf"></span>
<img src="Static/figures/plot-qual-conf-1.png" alt="quality scores by classification - HCC=1" width="480" />
<p class="caption">
Figure 5.2: quality scores by classification - HCC=1
</p>
</div>
</div>
<div id="simulation-design" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> Simulation Design</h2>
<p>We are now ready to run the simulations.</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="model-suite.html#cb85-1"></a> SIM &lt;-<span class="st"> </span><span class="dv">30</span></span>
<span id="cb85-2"><a href="model-suite.html#cb85-2"></a> SIZE &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">25</span>, <span class="dv">50</span>, <span class="dv">100</span>, <span class="dv">200</span>, <span class="dv">300</span>)</span>
<span id="cb85-3"><a href="model-suite.html#cb85-3"></a> CV_REP &lt;-<span class="st"> </span><span class="dv">30</span></span></code></pre></div>
<p>Simluation parameters:</p>
<ul>
<li><p>Number of simulations : SIM = <span class="math inline">\(30\)</span></p></li>
<li><p>Sample sizes: SIZE = <span class="math inline">\(25, 50, 100, 200, 300\)</span></p></li>
<li><p>Number of CV Replicates: CV_REP = <span class="math inline">\(30\)</span></p></li>
</ul>
<p>We will repeat the simulation process SIM = <span class="math inline">\(30\)</span> times.
For each simulation iteration, we will select <span class="math inline">\(300\)</span> Control and
<span class="math inline">\(300\)</span> HCC samples at random. Models will be fitted and analyzed
to balanced subsets of sise SIZE = <span class="math inline">\(25, 50, 100, 200, 300\)</span>, in a telescopic manner to
emulated a typical sample accrual process. Note that in this accural process
there is no time effect - the accrual process is completely randomized. In practice,
there could be significant time effects. For example, the first 25 HCC samples could come
from Center A, while the next 25 could come from Center B. In other words,
there is no batch effect or shared variability in our simulation,
while these are almost always present in real data, including
batch effects that are associated with class labels - controls being in
different batches than affected samples is an all too common occurence,
for example. One should be especially watchful of potential batch effects
when dealing with blood samples as blood is notoriously finicky in
character [<span class="citation">[<a href="references.html#ref-Huang:2017aa" role="doc-biblioref">16</a>]</span>; <span class="citation">[<a href="references.html#ref-Permenter:2015aa" role="doc-biblioref">17</a>]</span>;].
Presented with results that look impressively good based on a small data set,
one should definitely be skeptical of the promise of future equally good results.</p>
<p>For a given simulation and a given sample size, we will obtain
CV_REP = <span class="math inline">\(30\)</span> cross-validated lasso fits. From these fits,
we can obtain <span class="math inline">\(30\)</span> out-of-fold assessments of classification accuracy
to get a sense if its variability. From each cv replicate, we also obtain
an estimated model size and a set of selected features. We will want
to examine how these stabilize as the sample size increases.</p>
</div>
<div id="setup-simulation" class="section level2" number="5.3">
<h2><span class="header-section-number">5.3</span> Setup simulation</h2>
<p>To setup the simulation, we only need two master tables: one for the selection of Controls
and one for the selection of HCC samples.</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="model-suite.html#cb86-1"></a>all_control_vec &lt;-<span class="st"> </span><span class="kw">names</span>(all_group_vec[all_group_vec<span class="op">==</span><span class="st">&#39;Control&#39;</span>]) </span>
<span id="cb86-2"><a href="model-suite.html#cb86-2"></a>all_affected_vec &lt;-<span class="st"> </span><span class="kw">names</span>(all_group_vec[all_group_vec<span class="op">==</span><span class="st">&#39;HCC&#39;</span>])  </span></code></pre></div>
<p>We have <span class="math inline">\(778\)</span> IDs of control samples in stored in <code>all_control_vec</code>
and <span class="math inline">\(555\)</span> IDs of affected samples in stored in <code>all_affected_vec</code>.
To create random samples from these we only need to randomly select indices from
each vector.</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="model-suite.html#cb87-1"></a><span class="kw">set.seed</span>(<span class="dv">12379</span>)</span>
<span id="cb87-2"><a href="model-suite.html#cb87-2"></a></span>
<span id="cb87-3"><a href="model-suite.html#cb87-3"></a>sim_control_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb87-4"><a href="model-suite.html#cb87-4"></a> <span class="dv">1</span><span class="op">:</span>SIM, </span>
<span id="cb87-5"><a href="model-suite.html#cb87-5"></a> <span class="cf">function</span>(dummy) </span>
<span id="cb87-6"><a href="model-suite.html#cb87-6"></a>   <span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(all_control_vec), <span class="dt">size =</span>  <span class="kw">max</span>(SIZE))</span>
<span id="cb87-7"><a href="model-suite.html#cb87-7"></a>)</span>
<span id="cb87-8"><a href="model-suite.html#cb87-8"></a></span>
<span id="cb87-9"><a href="model-suite.html#cb87-9"></a></span>
<span id="cb87-10"><a href="model-suite.html#cb87-10"></a>sim_affected_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb87-11"><a href="model-suite.html#cb87-11"></a> <span class="dv">1</span><span class="op">:</span>SIM, </span>
<span id="cb87-12"><a href="model-suite.html#cb87-12"></a> <span class="cf">function</span>(dummy) </span>
<span id="cb87-13"><a href="model-suite.html#cb87-13"></a>   <span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(all_affected_vec), <span class="dt">size =</span>  <span class="kw">max</span>(SIZE))</span>
<span id="cb87-14"><a href="model-suite.html#cb87-14"></a>)</span></code></pre></div>
<p>Each simulation is specified by a given column of the simulation design matrices:
<code>sim_control_mtx</code> and <code>sim_affected_mtx</code>. Within each simulation, we can run
the analyses of size <span class="math inline">\(25, 50, 100, 200, 300\)</span>.</p>
<p>We can examine how much variability we have in the quality scores of the selected samples.</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="model-suite.html#cb88-1"></a>all_control_qual_vec &lt;-<span class="st"> </span>sample_qual_vec[all_control_vec]</span>
<span id="cb88-2"><a href="model-suite.html#cb88-2"></a>sim_control_qual_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb88-3"><a href="model-suite.html#cb88-3"></a>  <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(sim_control_mtx), </span>
<span id="cb88-4"><a href="model-suite.html#cb88-4"></a>  <span class="cf">function</span>(CC) all_control_qual_vec[sim_control_mtx[,CC]]</span>
<span id="cb88-5"><a href="model-suite.html#cb88-5"></a> )</span>
<span id="cb88-6"><a href="model-suite.html#cb88-6"></a></span>
<span id="cb88-7"><a href="model-suite.html#cb88-7"></a>all_affected_qual_vec &lt;-<span class="st"> </span>sample_qual_vec[all_affected_vec]</span>
<span id="cb88-8"><a href="model-suite.html#cb88-8"></a>sim_affected_qual_mtx &lt;-<span class="st"> </span><span class="kw">sapply</span>(</span>
<span id="cb88-9"><a href="model-suite.html#cb88-9"></a>  <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(sim_affected_mtx),  </span>
<span id="cb88-10"><a href="model-suite.html#cb88-10"></a>  <span class="cf">function</span>(CC) all_affected_qual_vec[sim_affected_mtx[,CC]]</span>
<span id="cb88-11"><a href="model-suite.html#cb88-11"></a> )</span>
<span id="cb88-12"><a href="model-suite.html#cb88-12"></a></span>
<span id="cb88-13"><a href="model-suite.html#cb88-13"></a><span class="co"># Get stage from SIZE </span></span>
<span id="cb88-14"><a href="model-suite.html#cb88-14"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sim_control_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>,SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb88-15"><a href="model-suite.html#cb88-15"></a></span>
<span id="cb88-16"><a href="model-suite.html#cb88-16"></a>sim_control_qual_byStage_lst &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="st">&#39;c&#39;</span>, </span>
<span id="cb88-17"><a href="model-suite.html#cb88-17"></a> <span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(sim_control_qual_mtx), </span>
<span id="cb88-18"><a href="model-suite.html#cb88-18"></a>  <span class="cf">function</span>(CC) <span class="kw">c</span>(<span class="kw">split</span>(sim_control_qual_mtx[,CC], stage_vec),<span class="ot">NA</span>)</span>
<span id="cb88-19"><a href="model-suite.html#cb88-19"></a> )</span>
<span id="cb88-20"><a href="model-suite.html#cb88-20"></a>)</span>
<span id="cb88-21"><a href="model-suite.html#cb88-21"></a></span>
<span id="cb88-22"><a href="model-suite.html#cb88-22"></a>sim_affected_qual_byStage_lst &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="st">&#39;c&#39;</span>, </span>
<span id="cb88-23"><a href="model-suite.html#cb88-23"></a> <span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(sim_affected_qual_mtx), </span>
<span id="cb88-24"><a href="model-suite.html#cb88-24"></a>  <span class="cf">function</span>(CC) <span class="kw">c</span>(<span class="kw">split</span>(sim_affected_qual_mtx[,CC], stage_vec),<span class="ot">NA</span>)</span>
<span id="cb88-25"><a href="model-suite.html#cb88-25"></a> )</span>
<span id="cb88-26"><a href="model-suite.html#cb88-26"></a>)</span>
<span id="cb88-27"><a href="model-suite.html#cb88-27"></a></span>
<span id="cb88-28"><a href="model-suite.html#cb88-28"></a><span class="co"># PLOT</span></span>
<span id="cb88-29"><a href="model-suite.html#cb88-29"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">1</span>), <span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">5</span>,<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb88-30"><a href="model-suite.html#cb88-30"></a><span class="co"># control</span></span>
<span id="cb88-31"><a href="model-suite.html#cb88-31"></a><span class="kw">boxplot</span>(</span>
<span id="cb88-32"><a href="model-suite.html#cb88-32"></a>  sim_control_qual_byStage_lst, </span>
<span id="cb88-33"><a href="model-suite.html#cb88-33"></a>  <span class="dt">outline =</span> F, </span>
<span id="cb88-34"><a href="model-suite.html#cb88-34"></a>  <span class="dt">border =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">6</span>,</span>
<span id="cb88-35"><a href="model-suite.html#cb88-35"></a>  <span class="dt">ylab =</span> <span class="st">&#39;Quality Score&#39;</span>,</span>
<span id="cb88-36"><a href="model-suite.html#cb88-36"></a>  <span class="dt">xaxt =</span> <span class="st">&#39;n&#39;</span></span>
<span id="cb88-37"><a href="model-suite.html#cb88-37"></a>)</span>
<span id="cb88-38"><a href="model-suite.html#cb88-38"></a><span class="kw">legend</span>(<span class="st">&#39;bottomright&#39;</span>, <span class="dt">title =</span> <span class="st">&#39;Stage&#39;</span>, <span class="dt">ncol =</span> <span class="dv">2</span>,</span>
<span id="cb88-39"><a href="model-suite.html#cb88-39"></a> <span class="dt">legend =</span> <span class="kw">names</span>(sim_control_qual_byStage_lst[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]), </span>
<span id="cb88-40"><a href="model-suite.html#cb88-40"></a> <span class="dt">text.col =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,</span>
<span id="cb88-41"><a href="model-suite.html#cb88-41"></a> <span class="dt">bty =</span> <span class="st">&#39;n&#39;</span>, <span class="dt">horiz =</span> F</span>
<span id="cb88-42"><a href="model-suite.html#cb88-42"></a>)</span>
<span id="cb88-43"><a href="model-suite.html#cb88-43"></a>sim_ndx &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">names</span>(sim_control_qual_byStage_lst) <span class="op">==</span><span class="st">&#39;&#39;</span>)</span>
<span id="cb88-44"><a href="model-suite.html#cb88-44"></a><span class="kw">abline</span>(<span class="dt">v =</span> sim_ndx, <span class="dt">col =</span> <span class="st">&#39;grey&#39;</span>)</span>
<span id="cb88-45"><a href="model-suite.html#cb88-45"></a><span class="kw">axis</span>(</span>
<span id="cb88-46"><a href="model-suite.html#cb88-46"></a>  <span class="dt">side =</span> <span class="dv">1</span>, </span>
<span id="cb88-47"><a href="model-suite.html#cb88-47"></a>  <span class="dt">at =</span> sim_ndx<span class="dv">-2</span>, </span>
<span id="cb88-48"><a href="model-suite.html#cb88-48"></a>  <span class="dt">label =</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(sim_ndx),</span>
<span id="cb88-49"><a href="model-suite.html#cb88-49"></a>  <span class="dt">tick =</span> F, </span>
<span id="cb88-50"><a href="model-suite.html#cb88-50"></a>  <span class="dt">line =</span> <span class="dv">-1</span>, <span class="dt">las =</span> <span class="dv">2</span>,</span>
<span id="cb88-51"><a href="model-suite.html#cb88-51"></a>  <span class="dt">cex.axis =</span> <span class="fl">0.8</span>)</span>
<span id="cb88-52"><a href="model-suite.html#cb88-52"></a><span class="kw">title</span>(<span class="st">&quot;Control sample quality by stage and simulation&quot;</span>)</span>
<span id="cb88-53"><a href="model-suite.html#cb88-53"></a></span>
<span id="cb88-54"><a href="model-suite.html#cb88-54"></a><span class="co"># affected</span></span>
<span id="cb88-55"><a href="model-suite.html#cb88-55"></a><span class="kw">boxplot</span>(</span>
<span id="cb88-56"><a href="model-suite.html#cb88-56"></a>  sim_affected_qual_byStage_lst, </span>
<span id="cb88-57"><a href="model-suite.html#cb88-57"></a>  <span class="dt">outline =</span> F, </span>
<span id="cb88-58"><a href="model-suite.html#cb88-58"></a>  <span class="dt">border =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">6</span>,</span>
<span id="cb88-59"><a href="model-suite.html#cb88-59"></a>  <span class="dt">ylab =</span> <span class="st">&#39;Quality Score&#39;</span>,</span>
<span id="cb88-60"><a href="model-suite.html#cb88-60"></a>  <span class="dt">xaxt =</span> <span class="st">&#39;n&#39;</span></span>
<span id="cb88-61"><a href="model-suite.html#cb88-61"></a>)</span>
<span id="cb88-62"><a href="model-suite.html#cb88-62"></a>sim_ndx &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">names</span>(sim_affected_qual_byStage_lst)<span class="op">==</span><span class="st">&#39;&#39;</span>)</span>
<span id="cb88-63"><a href="model-suite.html#cb88-63"></a><span class="kw">abline</span>(<span class="dt">v =</span> <span class="kw">which</span>(<span class="kw">names</span>(sim_affected_qual_byStage_lst)<span class="op">==</span><span class="st">&#39;&#39;</span>), <span class="dt">col =</span> <span class="st">&#39;grey&#39;</span>)</span>
<span id="cb88-64"><a href="model-suite.html#cb88-64"></a><span class="kw">axis</span>(</span>
<span id="cb88-65"><a href="model-suite.html#cb88-65"></a>  <span class="dt">side=</span><span class="dv">1</span>, </span>
<span id="cb88-66"><a href="model-suite.html#cb88-66"></a>  <span class="dt">at =</span> sim_ndx<span class="dv">-2</span>,        </span>
<span id="cb88-67"><a href="model-suite.html#cb88-67"></a>  <span class="dt">label =</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(sim_ndx),</span>
<span id="cb88-68"><a href="model-suite.html#cb88-68"></a>  <span class="dt">tick =</span> F, </span>
<span id="cb88-69"><a href="model-suite.html#cb88-69"></a>  <span class="dt">line =</span> <span class="dv">-1</span>, <span class="dt">las =</span> <span class="dv">2</span>,</span>
<span id="cb88-70"><a href="model-suite.html#cb88-70"></a>  <span class="dt">cex.axis =</span> <span class="fl">0.8</span>)</span>
<span id="cb88-71"><a href="model-suite.html#cb88-71"></a><span class="kw">title</span>(<span class="st">&quot;Affected sample quality by stage and simulation&quot;</span>)</span></code></pre></div>
<div class="figure"><span id="fig:look-sim-qual"></span>
<img src="Static/figures/look-sim-qual-1.png" alt="sample quality by simulation run" width="960" />
<p class="caption">
Figure 5.3: sample quality by simulation run
</p>
</div>
<p>We see some variability in sample quality in the smaller analysis stages.
This may lead observers to be overly optimistic, or overly pessimistic,
in the early accrual stages.</p>
</div>
<div id="run-simulations" class="section level2" number="5.4">
<h2><span class="header-section-number">5.4</span> Run simulations</h2>
<p>As these make take a while to run, and space to store,
we will save the results of each similation to a different
object and store to disk.</p>
<p>The simulation saves results to the file system and
only needs to be run once. The simulation takes <span class="math inline">\(\approx\)</span> 8 minutes
per iteration, or 4 hours of run time on a laptop.
(Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Mojave 10.14.6)</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="model-suite.html#cb89-1"></a><span class="co"># Get stage from SIZE</span></span>
<span id="cb89-2"><a href="model-suite.html#cb89-2"></a>stage_vec &lt;-<span class="st"> </span><span class="kw">cut</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(sim_control_qual_mtx), <span class="kw">c</span>(<span class="dv">0</span>, SIZE), <span class="dt">include.lowest =</span> T)</span>
<span id="cb89-3"><a href="model-suite.html#cb89-3"></a></span>
<span id="cb89-4"><a href="model-suite.html#cb89-4"></a><span class="cf">for</span> (SIMno <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(sim_control_qual_mtx)) {</span>
<span id="cb89-5"><a href="model-suite.html#cb89-5"></a>  start_time &lt;-<span class="st"> </span><span class="kw">proc.time</span>()</span>
<span id="cb89-6"><a href="model-suite.html#cb89-6"></a></span>
<span id="cb89-7"><a href="model-suite.html#cb89-7"></a>  <span class="kw">cat</span>(<span class="st">&quot;Running simulation &quot;</span>, SIMno, <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)</span>
<span id="cb89-8"><a href="model-suite.html#cb89-8"></a></span>
<span id="cb89-9"><a href="model-suite.html#cb89-9"></a>  sim_cv_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(<span class="kw">levels</span>(stage_vec)), <span class="cf">function</span>(STGno) {</span>
<span id="cb89-10"><a href="model-suite.html#cb89-10"></a>    Stage_rows_vec &lt;-<span class="st"> </span><span class="kw">which</span>(stage_vec <span class="op">%in%</span><span class="st"> </span><span class="kw">levels</span>(stage_vec)[<span class="dv">1</span><span class="op">:</span>STGno])</span>
<span id="cb89-11"><a href="model-suite.html#cb89-11"></a>    <span class="co">#cat(&quot;Stage &quot;, STGno, &quot;- analyzing&quot;, length(Stage_rows_vec), &quot;paired samples.\n&quot;)</span></span>
<span id="cb89-12"><a href="model-suite.html#cb89-12"></a></span>
<span id="cb89-13"><a href="model-suite.html#cb89-13"></a>    sim_stage_samples_vec &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb89-14"><a href="model-suite.html#cb89-14"></a>      all_control_vec[sim_control_mtx[Stage_rows_vec, SIMno]],</span>
<span id="cb89-15"><a href="model-suite.html#cb89-15"></a>      all_affected_vec[sim_affected_mtx[Stage_rows_vec, SIMno]]</span>
<span id="cb89-16"><a href="model-suite.html#cb89-16"></a>    )</span>
<span id="cb89-17"><a href="model-suite.html#cb89-17"></a>    sim_stage_lcpm_mtx &lt;-<span class="st"> </span>all_lcpm_mtx[sim_stage_samples_vec, ]</span>
<span id="cb89-18"><a href="model-suite.html#cb89-18"></a>    sim_stage_group_vec &lt;-<span class="st"> </span>all_group_vec[sim_stage_samples_vec]</span>
<span id="cb89-19"><a href="model-suite.html#cb89-19"></a>    <span class="kw">print</span>(<span class="kw">table</span>(sim_stage_group_vec))</span>
<span id="cb89-20"><a href="model-suite.html#cb89-20"></a></span>
<span id="cb89-21"><a href="model-suite.html#cb89-21"></a>    sim_stage_cv_lst &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span>CV_REP, <span class="cf">function</span>(CV) {</span>
<span id="cb89-22"><a href="model-suite.html#cb89-22"></a>      cv_fit &lt;-<span class="st"> </span>glmnet<span class="op">::</span><span class="kw">cv.glmnet</span>(</span>
<span id="cb89-23"><a href="model-suite.html#cb89-23"></a>        <span class="dt">x =</span> sim_stage_lcpm_mtx,</span>
<span id="cb89-24"><a href="model-suite.html#cb89-24"></a>        <span class="dt">y =</span> sim_stage_group_vec,</span>
<span id="cb89-25"><a href="model-suite.html#cb89-25"></a>        <span class="dt">alpha =</span> <span class="dv">1</span>,</span>
<span id="cb89-26"><a href="model-suite.html#cb89-26"></a>        <span class="dt">family =</span> <span class="st">&quot;binomial&quot;</span>,</span>
<span id="cb89-27"><a href="model-suite.html#cb89-27"></a>        <span class="dt">type.measure =</span> <span class="st">&quot;class&quot;</span>,</span>
<span id="cb89-28"><a href="model-suite.html#cb89-28"></a>        <span class="dt">keep =</span> T,</span>
<span id="cb89-29"><a href="model-suite.html#cb89-29"></a>        <span class="dt">nlambda =</span> <span class="dv">30</span></span>
<span id="cb89-30"><a href="model-suite.html#cb89-30"></a>      )</span>
<span id="cb89-31"><a href="model-suite.html#cb89-31"></a>      ndx_1se &lt;-<span class="st"> </span><span class="kw">which</span>(cv_fit<span class="op">$</span>lambda <span class="op">==</span><span class="st"> </span>cv_fit<span class="op">$</span>lambda<span class="fl">.1</span>se)</span>
<span id="cb89-32"><a href="model-suite.html#cb89-32"></a></span>
<span id="cb89-33"><a href="model-suite.html#cb89-33"></a>      nzero_1se &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>nzero[ndx_1se]</span>
<span id="cb89-34"><a href="model-suite.html#cb89-34"></a>      cvm_1se &lt;-<span class="st"> </span>cv_fit<span class="op">$</span>cvm[ndx_1se]</span>
<span id="cb89-35"><a href="model-suite.html#cb89-35"></a></span>
<span id="cb89-36"><a href="model-suite.html#cb89-36"></a>      <span class="co"># oof error</span></span>
<span id="cb89-37"><a href="model-suite.html#cb89-37"></a>      oofPred_1se_vec &lt;-<span class="st"> </span><span class="kw">ifelse</span>(</span>
<span id="cb89-38"><a href="model-suite.html#cb89-38"></a>        cv_fit<span class="op">$</span>fit.preval[, ndx_1se] <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&quot;HCC&quot;</span>, <span class="st">&quot;Control&quot;</span></span>
<span id="cb89-39"><a href="model-suite.html#cb89-39"></a>      )</span>
<span id="cb89-40"><a href="model-suite.html#cb89-40"></a>      oofPred_1se_error &lt;-<span class="st"> </span><span class="kw">mean</span>(oofPred_1se_vec <span class="op">!=</span><span class="st"> </span>sim_stage_group_vec)</span>
<span id="cb89-41"><a href="model-suite.html#cb89-41"></a></span>
<span id="cb89-42"><a href="model-suite.html#cb89-42"></a>      <span class="co"># test error</span></span>
<span id="cb89-43"><a href="model-suite.html#cb89-43"></a>      sim_stage_test_samples_vec &lt;-<span class="st"> </span><span class="kw">setdiff</span>(<span class="kw">rownames</span>(all_lcpm_mtx), sim_stage_samples_vec)</span>
<span id="cb89-44"><a href="model-suite.html#cb89-44"></a>      sim_stage_test_lcpm_mtx &lt;-<span class="st"> </span>all_lcpm_mtx[sim_stage_test_samples_vec,]</span>
<span id="cb89-45"><a href="model-suite.html#cb89-45"></a>      sim_stage_test_group_vec &lt;-<span class="st"> </span>all_group_vec[sim_stage_test_samples_vec]</span>
<span id="cb89-46"><a href="model-suite.html#cb89-46"></a></span>
<span id="cb89-47"><a href="model-suite.html#cb89-47"></a>      test_pred_1se_vec &lt;-<span class="st"> </span><span class="kw">predict</span>(</span>
<span id="cb89-48"><a href="model-suite.html#cb89-48"></a>       cv_fit,</span>
<span id="cb89-49"><a href="model-suite.html#cb89-49"></a>       <span class="dt">newx=</span>sim_stage_test_lcpm_mtx,</span>
<span id="cb89-50"><a href="model-suite.html#cb89-50"></a>       <span class="dt">s=</span><span class="st">&quot;lambda.1se&quot;</span>,</span>
<span id="cb89-51"><a href="model-suite.html#cb89-51"></a>       <span class="dt">type=</span><span class="st">&quot;class&quot;</span></span>
<span id="cb89-52"><a href="model-suite.html#cb89-52"></a>      )</span>
<span id="cb89-53"><a href="model-suite.html#cb89-53"></a>      test_1se_error &lt;-<span class="st"> </span><span class="kw">mean</span>(test_pred_1se_vec <span class="op">!=</span><span class="st"> </span>sim_stage_test_group_vec)</span>
<span id="cb89-54"><a href="model-suite.html#cb89-54"></a></span>
<span id="cb89-55"><a href="model-suite.html#cb89-55"></a>      <span class="co"># genes</span></span>
<span id="cb89-56"><a href="model-suite.html#cb89-56"></a>      coef_1se &lt;-<span class="st"> </span><span class="kw">coef</span>(</span>
<span id="cb89-57"><a href="model-suite.html#cb89-57"></a>        cv_fit,</span>
<span id="cb89-58"><a href="model-suite.html#cb89-58"></a>        <span class="dt">s =</span> <span class="st">&quot;lambda.1se&quot;</span></span>
<span id="cb89-59"><a href="model-suite.html#cb89-59"></a>      )</span>
<span id="cb89-60"><a href="model-suite.html#cb89-60"></a>      genes &lt;-<span class="st"> </span>coef_1se<span class="op">@</span>Dimnames[[<span class="dv">1</span>]][coef_1se<span class="op">@</span>i[<span class="op">-</span><span class="dv">1</span>]]</span>
<span id="cb89-61"><a href="model-suite.html#cb89-61"></a></span>
<span id="cb89-62"><a href="model-suite.html#cb89-62"></a>      <span class="kw">list</span>(<span class="dt">p =</span> nzero_1se, <span class="dt">cv_error =</span> cvm_1se, <span class="dt">oof_error =</span> oofPred_1se_error, </span>
<span id="cb89-63"><a href="model-suite.html#cb89-63"></a>           <span class="dt">test_error=</span>test_1se_error, <span class="dt">genes =</span> genes)</span>
<span id="cb89-64"><a href="model-suite.html#cb89-64"></a>    })</span>
<span id="cb89-65"><a href="model-suite.html#cb89-65"></a>    sim_stage_cv_lst</span>
<span id="cb89-66"><a href="model-suite.html#cb89-66"></a>  })</span>
<span id="cb89-67"><a href="model-suite.html#cb89-67"></a></span>
<span id="cb89-68"><a href="model-suite.html#cb89-68"></a>  <span class="co"># save  sim_cv_lst</span></span>
<span id="cb89-69"><a href="model-suite.html#cb89-69"></a>  fName &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;sim_&quot;</span>, SIMno, <span class="st">&quot;_cv_lst&quot;</span>)</span>
<span id="cb89-70"><a href="model-suite.html#cb89-70"></a>  <span class="kw">assign</span>(fName, sim_cv_lst)</span>
<span id="cb89-71"><a href="model-suite.html#cb89-71"></a>  <span class="kw">save</span>(<span class="dt">list =</span> fName, <span class="dt">file=</span><span class="kw">file.path</span>(<span class="st">&quot;RData&quot;</span>, fName))</span>
<span id="cb89-72"><a href="model-suite.html#cb89-72"></a></span>
<span id="cb89-73"><a href="model-suite.html#cb89-73"></a>  <span class="kw">message</span>(<span class="st">&quot;simulation time: &quot;</span>, <span class="kw">round</span>((<span class="kw">proc.time</span>() <span class="op">-</span><span class="st"> </span>start_time)[<span class="dv">3</span>], <span class="dv">2</span>), <span class="st">&quot;s&quot;</span>)</span>
<span id="cb89-74"><a href="model-suite.html#cb89-74"></a>}</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="explore-sparsity.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="variable-importance.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["HCC5hmCExplore.pdf", "HCC5hmCExplore.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
