<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Section 2 Preprocessing | DNA Hydroxymethylation in Hepatocellular Carcinoma</title>
  <meta name="description" content="Data from Cai et al. (2019) paper are explored" />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Section 2 Preprocessing | DNA Hydroxymethylation in Hepatocellular Carcinoma" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Data from Cai et al. (2019) paper are explored" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Section 2 Preprocessing | DNA Hydroxymethylation in Hepatocellular Carcinoma" />
  
  <meta name="twitter:description" content="Data from Cai et al. (2019) paper are explored" />
  

<meta name="author" content="Francois Collin" />


<meta name="date" content="2020-09-06" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="intro.html"/>
<link rel="next" href="modeling-background.html"/>
<script src="libs/header-attrs-2.2/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">DNA Hydroxymethylation in Hepatocellular Carcinoma</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preamble</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i>License</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="preproc.html"><a href="preproc.html"><i class="fa fa-check"></i><b>2</b> Preprocessing</a>
<ul>
<li class="chapter" data-level="2.1" data-path="preproc.html"><a href="preproc.html#load-the-data"><i class="fa fa-check"></i><b>2.1</b> Load the data</a></li>
<li class="chapter" data-level="2.2" data-path="preproc.html"><a href="preproc.html#dra"><i class="fa fa-check"></i><b>2.2</b> Differential representation analysis</a>
<ul>
<li class="chapter" data-level="" data-path="preproc.html"><a href="preproc.html#remove-lowly-expressed-genes"><i class="fa fa-check"></i>Remove lowly expressed genes</a></li>
<li class="chapter" data-level="" data-path="preproc.html"><a href="preproc.html#creating-a-design-matrix-and-contrasts"><i class="fa fa-check"></i>Creating a design matrix and contrasts</a></li>
<li class="chapter" data-level="" data-path="preproc.html"><a href="preproc.html#removing-heteroscedasticity-from-the-count-data"><i class="fa fa-check"></i>Removing heteroscedasticity from the count data</a></li>
<li class="chapter" data-level="" data-path="preproc.html"><a href="preproc.html#fit-linear-models-and-examine-the-results"><i class="fa fa-check"></i>Fit linear models and examine the results</a></li>
<li class="chapter" data-level="" data-path="preproc.html"><a href="preproc.html#graphical-representations-of-de-results-md-plots"><i class="fa fa-check"></i>Graphical representations of DE results: MD Plots</a></li>
<li class="chapter" data-level="" data-path="preproc.html"><a href="preproc.html#de-genes-at-10-fold-change"><i class="fa fa-check"></i>DE genes at 10% fold change</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="preproc.html"><a href="preproc.html#snr-regime"><i class="fa fa-check"></i><b>2.3</b> Signal-to-noise ratio regime</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="modeling-background.html"><a href="modeling-background.html"><i class="fa fa-check"></i><b>3</b> Modeling - Background</a>
<ul>
<li class="chapter" data-level="3.1" data-path="modeling-background.html"><a href="modeling-background.html#predictive-modeling-for-genomic-data"><i class="fa fa-check"></i><b>3.1</b> Predictive modeling for genomic data</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="modeling-background.html"><a href="modeling-background.html#caret-for-model-evaluation"><i class="fa fa-check"></i><b>3.1.1</b> caret for model evaluation</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="modeling-background.html"><a href="modeling-background.html#glmnet"><i class="fa fa-check"></i><b>3.2</b> glmnet</a>
<ul>
<li><a href="modeling-background.html#alpha-hyper-parameter"><strong>alpha</strong> hyper-parameter</a></li>
<li class="chapter" data-level="" data-path="modeling-background.html"><a href="modeling-background.html#lasso-vs-best-subset"><i class="fa fa-check"></i>Lasso vs Best Subset</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="explore-sparsity.html"><a href="explore-sparsity.html"><i class="fa fa-check"></i><b>4</b> The bet on sparsity</a>
<ul>
<li class="chapter" data-level="4.1" data-path="explore-sparsity.html"><a href="explore-sparsity.html#cv-analysis-setup"><i class="fa fa-check"></i><b>4.1</b> CV analysis setup</a></li>
<li class="chapter" data-level="4.2" data-path="explore-sparsity.html"><a href="explore-sparsity.html#fit-and-compare-models"><i class="fa fa-check"></i><b>4.2</b> Fit and compare models</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="explore-sparsity.html"><a href="explore-sparsity.html#logistic-regression-in-glmnet"><i class="fa fa-check"></i><b>4.2.1</b> Logistic regression in <code>glmnet</code></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="explore-sparsity.html"><a href="explore-sparsity.html#relaxed-lasso-and-blended-mix"><i class="fa fa-check"></i><b>4.3</b> Relaxed lasso and blended mix</a></li>
<li class="chapter" data-level="4.4" data-path="explore-sparsity.html"><a href="explore-sparsity.html#examination-of-sensitivity-vs-specificity"><i class="fa fa-check"></i><b>4.4</b> Examination of sensitivity vs specificity</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="explore-sparsity.html"><a href="explore-sparsity.html#training-data-out-of-fold-roc-curves"><i class="fa fa-check"></i><b>4.4.1</b> Training data out-of-fold ROC curves</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="explore-sparsity.html"><a href="explore-sparsity.html#refit-with-auc-as-optimization"><i class="fa fa-check"></i><b>4.5</b> Refit with “auc” as optimization</a></li>
<li class="chapter" data-level="4.6" data-path="explore-sparsity.html"><a href="explore-sparsity.html#relaxed-lasso-and-blended-mix-1"><i class="fa fa-check"></i><b>4.6</b> Relaxed lasso and blended mix</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="examine-feature-selection.html"><a href="examine-feature-selection.html"><i class="fa fa-check"></i><b>5</b> Examine feature selection</a>
<ul>
<li class="chapter" data-level="5.1" data-path="examine-feature-selection.html"><a href="examine-feature-selection.html#sparsity-stability"><i class="fa fa-check"></i><b>5.1</b> Sparsity stability</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="model-suite.html"><a href="model-suite.html"><i class="fa fa-check"></i><b>6</b> Fitted Model Suite</a></li>
<li class="chapter" data-level="7" data-path="conclusions.html"><a href="conclusions.html"><i class="fa fa-check"></i><b>7</b> Conclusions</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="LICENSE.md" style="font:em;" target="blank">Copyright (c) 2020 Francois Collin</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">DNA Hydroxymethylation in Hepatocellular Carcinoma</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="preproc" class="section level1" number="2">
<h1><span class="header-section-number">Section 2</span> Preprocessing</h1>
<!--
 FN <- 'tmp'
 # Shotcuts for knitting and redering while in R session (Invoke interactive R from R/Scripts folder)
 kk <- function(n='') knitr::knit2html(paste("t", n, sep=''), envir=globalenv(),
       output=paste(FN,".html", sep=''))

 rr <- function(n='') rmarkdown::render(paste("t", n, sep=''), envir=globalenv(),
       output_file=paste(FN,".html", sep='')) ##, output_dir='Scripts')

 bb <- function(n='') browseURL(paste(FN,".html", sep=''))

 # The usual shotcuts
 zz <- function(n='') source(paste("t", n, sep=''))
-->
<div id="load-the-data" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> Load the data</h2>
<!-- THIS ENSURES NO EVALUATION TAKES PLACE BY DEFAULT -->
<!-- TO TURN ON, SET eval=T                            -->
<!-- Add base libraries -->
<p>The data that are available from NCBI GEO
<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE112679">Series GSE112679</a>
can be conveniently accessed through an R data package.
Attaching the GSE112679 package makes the count data tables
available as well as a gene annotation table and a sample description table.
See <a href="https://12379monty.github.io/GSE112679/">GSE112679 R Data Package page</a>.
For the Cai et al. <span class="citation">[<a href="references.html#ref-Cai:2019aa" role="doc-biblioref">2</a>]</span> model fitting and analysis, samples were separated into
<code>Train</code> and <code>Val-1</code> subsets. <code>Val-2</code> was an external validation set.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="preproc.html#cb5-1"></a><span class="cf">if</span> (<span class="op">!</span>(<span class="st">&quot;GSE112679&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()))) {</span>
<span id="cb5-2"><a href="preproc.html#cb5-2"></a>  <span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;devtools&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)) {</span>
<span id="cb5-3"><a href="preproc.html#cb5-3"></a>    <span class="kw">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span>
<span id="cb5-4"><a href="preproc.html#cb5-4"></a>  }</span>
<span id="cb5-5"><a href="preproc.html#cb5-5"></a>  devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;12379Monty/GSE112679&quot;</span>)</span>
<span id="cb5-6"><a href="preproc.html#cb5-6"></a>}</span>
<span id="cb5-7"><a href="preproc.html#cb5-7"></a><span class="kw">library</span>(GSE112679)</span>
<span id="cb5-8"><a href="preproc.html#cb5-8"></a>sampDesc<span class="op">$</span>DxStage &lt;-<span class="st"> </span><span class="kw">with</span>(sampDesc, <span class="kw">ifelse</span>(outcome<span class="op">==</span><span class="st">&#39;HCC&#39;</span>, </span>
<span id="cb5-9"><a href="preproc.html#cb5-9"></a>   <span class="kw">paste0</span>(outcome,<span class="st">&#39;:&#39;</span>,stage), outcome))</span>
<span id="cb5-10"><a href="preproc.html#cb5-10"></a></span>
<span id="cb5-11"><a href="preproc.html#cb5-11"></a><span class="kw">with</span>(</span>
<span id="cb5-12"><a href="preproc.html#cb5-12"></a>  sampDesc <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(sampType <span class="op">==</span><span class="st"> &quot;blood&quot;</span>),</span>
<span id="cb5-13"><a href="preproc.html#cb5-13"></a>  knitr<span class="op">::</span><span class="kw">kable</span>(<span class="kw">table</span>(DxStage, trainValGroup, <span class="dt">exclude =</span> <span class="ot">NULL</span>),</span>
<span id="cb5-14"><a href="preproc.html#cb5-14"></a>   <span class="dt">caption=</span><span class="st">&quot;GSE112679 Samples by Dx Group and Subset&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb5-15"><a href="preproc.html#cb5-15"></a><span class="st">  </span>kableExtra<span class="op">::</span><span class="kw">kable_styling</span>(<span class="dt">full_width =</span> F)</span>
<span id="cb5-16"><a href="preproc.html#cb5-16"></a>)</span></code></pre></div>
<pre><code>## Warning in kableExtra::kable_styling(., full_width = F): Please specify format
## in kable. kableExtra can customize either HTML or LaTeX outputs. See https://
## haozhu233.github.io/kableExtra/ for details.</code></pre>
<table>
<caption><span id="tab:loadData">Table 2.1: </span>GSE112679 Samples by Dx Group and Subset</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">Train</th>
<th align="right">Val-1</th>
<th align="right">Val-2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Benign</td>
<td align="right">253</td>
<td align="right">132</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td>CHB</td>
<td align="right">190</td>
<td align="right">96</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td>Cirrhosis</td>
<td align="right">73</td>
<td align="right">33</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td>HCC:Early</td>
<td align="right">335</td>
<td align="right">220</td>
<td align="right">24</td>
</tr>
<tr class="odd">
<td>HCC:Late</td>
<td align="right">0</td>
<td align="right">442</td>
<td align="right">13</td>
</tr>
<tr class="even">
<td>HCC:NA</td>
<td align="right">0</td>
<td align="right">147</td>
<td align="right">23</td>
</tr>
<tr class="odd">
<td>Healthy</td>
<td align="right">269</td>
<td align="right">124</td>
<td align="right">177</td>
</tr>
</tbody>
</table>
<p>For this analysis, we will consider early stage cancer samples
and healthy or benign samples from the <code>Train</code> or <code>Val-1</code> subsets.
The appropriate outcome variable will be renamed or aliased <code>group</code></p>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:subsetSamples">Table 2.2: </span>Samples used in this analysis
</caption>
<thead>
<tr>
<th style="text-align:left;">
group
</th>
<th style="text-align:right;">
Freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Control
</td>
<td style="text-align:right;">
778
</td>
</tr>
<tr>
<td style="text-align:left;">
HCC
</td>
<td style="text-align:right;">
555
</td>
</tr>
</tbody>
</table>
<p>The features are counts of reads captured by chemical labeling, and indicate
the level of 5-hydroxymethylcytosines within each gene body. Cai et al. (2019),
Li et al. (2017) and Song et al. (2017) <span class="citation">[<a href="references.html#ref-Cai:2019aa" role="doc-biblioref">2</a>–<a href="references.html#ref-Song:2017aa" role="doc-biblioref">4</a>]</span>
all analyze 5hmC gene body counts using standard RNA-Seq methodologies, and we will
do the same here.</p>
<p>Note that before conducting any substantive analyses, the data would normally
be very carefully examined for any sign of quality variation between groups
of samples. This analysis would integrate sample meta data - where and when were
the blood samples collected - as well as library preparation and sequencing metrics
in order to detect any sign of processing artifacts that may be present in the dataset.
This is particularly important when dealing with blood samples as variable
DNA quality degradation is a well known challenge that is encountered when dealing with
such samples <span class="citation">[<a href="references.html#ref-Huang:2017aa" role="doc-biblioref">6</a>]</span>. Although blood specimen handling protocols can be
put in place to minimize quality variation <span class="citation">[<a href="references.html#ref-Permenter:2015aa" role="doc-biblioref">7</a>]</span>, variability
can never be completely eradicated, especially in the context of blood samples
collected by different groups, working in different environments. The problem
of variable DNA quality becomes paricularly pernicuous when it is compounded
with a confounding factor that sneaks in when the control sample collection
events are separated in time and space from the cancer sample collection events;
an all too common occurence.</p>
<p>As proper data QC requires an intimate familiarity with the details of
data collection and processing, such a task cannot be untertaken here.
We will simply run a <em>minimal set of QC sanity checks</em> to make sure that
there are no apparent systematic effects in the data.</p>
<p>We first look at coverage - make sure there isn’t too much disparity of coverage
across samples. To detect shared variability, samples can be annotated and ordered
according to sample features that may be linked to sample batch processing. Here we
the samples have been ordered by group and sample id (an alias of geoAcc).</p>
<div class="figure"><span id="fig:lcpmBxp"></span>
<img src="Static/figures/lcpmBxp-1.png" alt="Sample log2 count boxplots" width="960" />
<p class="caption">
Figure 2.1: Sample log2 count boxplots
</p>
</div>
<!--
Coverage level looks fairly comparable across samples.  It is sometimes helpful to
keep track of the actual coverage which can be adequetely tracked by distribution
quantiles.
-->
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:quantCountsA">Table 2.3: </span>Coverage Summary - Columns are sample coverage quantiles and total coverage
Rows are quartiles across samples
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
15%
</th>
<th style="text-align:right;">
25%
</th>
<th style="text-align:right;">
50%
</th>
<th style="text-align:right;">
75%
</th>
<th style="text-align:right;">
totCovM
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
25%
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
24
</td>
<td style="text-align:right;">
111
</td>
<td style="text-align:right;">
321
</td>
<td style="text-align:right;">
5.5
</td>
</tr>
<tr>
<td style="text-align:left;">
50%
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
135
</td>
<td style="text-align:right;">
391
</td>
<td style="text-align:right;">
6.7
</td>
</tr>
<tr>
<td style="text-align:left;">
75%
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
35
</td>
<td style="text-align:right;">
162
</td>
<td style="text-align:right;">
468
</td>
<td style="text-align:right;">
8.0
</td>
</tr>
</tbody>
</table>
<p>From this table, we see that 25% of the samples have total coverage exceeding
8M reads, 25% of samples
have a 15 percentile of coverage lower than
4, etc.</p>
<!-- SKIP
We next look at relative log representation (RLR) (in the context of measuring the density of 
5hmC marks in genes, we refer to `representation` as opposed to `expression`; the
two can be used interchangibly) -
make sure the shapes of the distributions are not widely different.
-->
<!-- SKIPPED
We note that the HCC samples have slightly more variable coverage distribution.
A few samples are quite different.
-->
</div>
<div id="dra" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Differential representation analysis</h2>
<p>In the remainder of this section, we will process the data and
perform differential expression analysis as outlined in
Law et al. (2018) <span class="citation">[<a href="references.html#ref-Law:2018aa" role="doc-biblioref">8</a>]</span>. The main analysis steps are:</p>
<ul>
<li>remove lowly expressed genes</li>
<li>normalize gene expression distributions</li>
<li>remove heteroscedascity</li>
<li>fit linear models and examine DE results</li>
</ul>
<p>It is good practice to perform this differential expression analysis prior to
fitting models to get an idea of how difficult it will be to discriminate
between samples belonging to the different subgroups. The pipeline
outlined in Law et al. (2018) <span class="citation">[<a href="references.html#ref-Law:2018aa" role="doc-biblioref">8</a>]</span> also provides some
basic quality assessment opportunities.</p>
<div id="remove-lowly-expressed-genes" class="section level3 unnumbered" number="">
<h3>Remove lowly expressed genes</h3>
<p>Genes that are not expressed at a biologically
meaningful level in any condition should be discarded to reduce the
subset of genes to those that are of interest, and to reduce the number of tests
carried out downstream when looking at differential expression. Carrying
un-informative genes may also be a hindrance to classification and other
downtream analyses.</p>
<p>To determine a sensible threshold we can begin by examining the shapes of the distributions.</p>
<div class="figure"><span id="fig:densityLcpm"></span>
<img src="Static/figures/densityLcpm-1.png" alt="Sample $log_2$ CPM densities" width="960" />
<p class="caption">
Figure 2.2: Sample <span class="math inline">\(log_2\)</span> CPM densities
</p>
</div>
<p>As is typically the case with RNA-Seq data, we notice many weakly represented genes
in this dataset. A cpm value of 1 appears to adequatly separate
the expressed from the un-expressed genes, but we will be slightly more strict here
and require a CPM threshold of 3 . Using a nominal CPM value of
3, genes are deeemed to be <code>represented</code> if their expression is
above this threshold, and not represented otherwise.
For this analysis we will require that genes be <code>represented</code> in at least
25 samples across the entire dataset to be retained for downstream analysis.
Here, a CPM value of 3 means that a gene is represented if it
has at least 9 reads in the sample with the
lowest sequencing depth (library size 2.9 million).
Note that the thresholds used here are arbitrary as there are no hard and fast
rules to set these by.
The voom-plot, which is part of analyses done to remove heteroscedasticity,
can be examined to verify that the filtering performed is adequate.</p>
<!-- or at least 
40 counts in the sample with the 
greatest sequencing depth (library size 13.3 million).
-->
<p>Remove weakly represented genes and replot densities.</p>
<p>Removing 17.5% of genes…</p>
<div class="figure"><span id="fig:densityLcpm2"></span>
<img src="Static/figures/densityLcpm2-1.png" alt="Sample $log_2$ CPM densities after removing weak genes" width="960" />
<p class="caption">
Figure 2.3: Sample <span class="math inline">\(log_2\)</span> CPM densities after removing weak genes
</p>
</div>
<!--
Note that the $log_2(CMP)$ distribution is not quite symmetric.
-->
<p>As another sanity check, we will look at a
multidimensional scaling plot of distances between gene expression
profiles. We use <code>plotMDS</code> in limma package <span class="citation">[<a href="references.html#ref-Ritchie:2015aa" role="doc-biblioref">9</a>]</span>),
which plots samples on a two-dimensional scatterplot so that distances on
the plot approximate the typical log2 fold changes between the
samples.</p>
<p>Before producing the MDS plot we will normalize the distributions.
We will store the data into s <code>DGEList</code> object as this is convenient
when running many of the analyses implemented in the edgeR and limma packages.
Call the set ‘AF’, for set ‘A’, ‘Filtered’.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="preproc.html#cb7-1"></a>AF_dgel &lt;-<span class="st"> </span>edgeR<span class="op">::</span><span class="kw">DGEList</span>(</span>
<span id="cb7-2"><a href="preproc.html#cb7-2"></a>  <span class="dt">counts =</span> featureCountsAF,</span>
<span id="cb7-3"><a href="preproc.html#cb7-3"></a>  <span class="dt">genes =</span> genes_annotAF,</span>
<span id="cb7-4"><a href="preproc.html#cb7-4"></a>  <span class="dt">samples =</span> sampDescA,</span>
<span id="cb7-5"><a href="preproc.html#cb7-5"></a>  <span class="dt">group =</span> sampDescA<span class="op">$</span>group</span>
<span id="cb7-6"><a href="preproc.html#cb7-6"></a>)</span>
<span id="cb7-7"><a href="preproc.html#cb7-7"></a>AF_dgel &lt;-<span class="st"> </span>edgeR<span class="op">::</span><span class="kw">calcNormFactors</span>(AF_dgel)</span>
<span id="cb7-8"><a href="preproc.html#cb7-8"></a>AF_lcmp_mtx &lt;-<span class="st"> </span>edgeR<span class="op">::</span><span class="kw">cpm</span>(AF_dgel, <span class="dt">log =</span> T)</span>
<span id="cb7-9"><a href="preproc.html#cb7-9"></a></span>
<span id="cb7-10"><a href="preproc.html#cb7-10"></a><span class="co"># Save AF_dgel to facilitate restarting</span></span>
<span id="cb7-11"><a href="preproc.html#cb7-11"></a><span class="co"># remove from final version</span></span>
<span id="cb7-12"><a href="preproc.html#cb7-12"></a><span class="kw">save</span>(<span class="dt">list =</span> <span class="st">&quot;AF_dgel&quot;</span>, <span class="dt">file =</span> <span class="st">&quot;RData/AF_dgel&quot;</span>)</span></code></pre></div>
<p>Verify that the counts are properly normalized.</p>
<div class="figure"><span id="fig:normedLcpmBxp"></span>
<img src="Static/figures/normedLcpmBxp-1.png" alt="Sample log2 count boxplots" width="960" />
<p class="caption">
Figure 2.4: Sample log2 count boxplots
</p>
</div>
<p>Proceed with MDS plots.</p>
<div class="figure"><span id="fig:plotMDS"></span>
<img src="Static/figures/plotMDS-1.png" alt="MDS plots of log-CPM values" width="960" />
<p class="caption">
Figure 2.5: MDS plots of log-CPM values
</p>
</div>
<p>The MDS plot, which is analogous to a PCA plot adapted to gene exression data,
does not indicate strong clustering of samples. The fanning pattern observed in the
first two dimensions indicates that a few samples are drifting way from the
core set, but in no particular direction. There is some structure in the
3rd and 4th dimension plot which should be investigated.<br />
<code>glMDSPlot</code> from package <code>Glimma</code> provides an interactive MDS
plot that can extremely usful for exploration</p>
<p>Link to glMDSPlot:
<a href="Static/figures//GlMDSplot.html">Here</a></p>
<p>No obvious factor links the samples in the 3 clusters observed on the
4th MDS dimensions. The percent of variance exaplained by this dimension or
<span class="math inline">\(\approx\)</span> 4%. The glMDSPlot indicates further segregation along
the 6th dimension. The percent of variance exaplained by this dimension or
<span class="math inline">\(\approx\)</span> 2%. Tracking down this source of variability may be quite challenging,
especially without having the complete information about the sample attributes
and provenance.</p>
<p>Unwanted variability is a well-documented problem in the analysis of RNA-Seq data
(see Peixoto et al. (2015) <span class="citation">[<a href="references.html#ref-Peixoto:2015aa" role="doc-biblioref">10</a>]</span>), and many procedures have been proposed
to reduce the effect of unwanted variation on RNA-Seq analsys results
(<span class="citation">[<a href="references.html#ref-Peixoto:2015aa" role="doc-biblioref">10</a>–<a href="references.html#ref-Risso:2014aa" role="doc-biblioref">12</a>]</span>). There are undoubtedly
some similar sources of systematic variation in the 5hmC data, but it is
beyond the scope of this work to investigate these in this particular dataset.
Given that the clustering of samples occurs in MDS dimensions that explain
a small fraction of variability, and that these is no assocation with the
factor of interest, HCC vs Control, these sources of variability should not
interfere too much with our classification analysis. It would nonetheless be interesting
to assess whether downstream results can be improved by removing this variability.</p>
</div>
<div id="creating-a-design-matrix-and-contrasts" class="section level3 unnumbered" number="">
<h3>Creating a design matrix and contrasts</h3>
<p>Before proceeding with the statistical modeling used for the
differential expression analysis, we need to set up a
model design matrix.</p>
<pre><code>## colSums(Design_mtx):</code></pre>
<pre><code>## Control     HCC 
##     778     555</code></pre>
<pre><code>## Contrasts:</code></pre>
<pre><code>##          Contrasts
## Levels    HCCvsControl
##   Control           -1
##   HCC                1</code></pre>
</div>
<div id="removing-heteroscedasticity-from-the-count-data" class="section level3 unnumbered" number="">
<h3>Removing heteroscedasticity from the count data</h3>
<p>As for RNA-Seq data, for 5hmC count data the variance is not independent of the mean.
In <code>limma</code>, the R package we are using for our analyses,
linear modeling is carried out on the log-CPM values which are assumed to be
normally distributed and the mean-variance relationship is accommodated using precision
weights calculated by the voom function. We apply this transformation next.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="preproc.html#cb12-1"></a><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb12-2"><a href="preproc.html#cb12-2"></a>filteredCountsAF_voom &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw">voom</span>(AF_dgel, Design_mtx, <span class="dt">plot=</span>T)</span></code></pre></div>
<div class="figure"><span id="fig:Voom1"></span>
<img src="Static/figures/Voom1-1.png" alt="Removing heteroscedascity" width="1056" />
<p class="caption">
Figure 2.6: Removing heteroscedascity
</p>
</div>
<p>Note that the voom-plot provides a visual check on the level of filtering performed upstream.
If filtering of lowly-expressed genes is insufficient, a drop in variance levels can be
observed at the low end of the expression scale due to very small counts.</p>
<!--
Means (x-axis) and variances (y-axis) of each gene are plotted to show the dependence between the two before voom is applied to the data ( A) and how the trend is removed after voom precision weights are applied to the data ( B). The plot on the left is created within the voom function which extracts residual variances from fitting linear models to log-CPM transformed data. Variances are then re-scaled to quarter-root variances (or square-root of standard deviations) and plotted against the mean expression of each gene. The means are log 2-transformed mean-counts with an offset of 2. The plot on the right is created using plotSA which plots log 2 residual standard deviations against mean log-CPM values. The average log 2 residual standard deviation is marked by a horizontal blue line. In both plots, each black dot represents a gene and a red curve is fitted to these points.
-->
<!--

To get a sense of how this compares with RNA-Seq dsta, we can take a look at Figure 4 in Law et al. [@Law:2018aa]:

TICKr knitr::include_graphics(
  "Static/images/Law-fig4.gif",  dpi=100)TICK

We observe that the variability in the 5hmC data is quite a bit lower.  Statistical summaries would give
a better idea.  

-->
</div>
<div id="fit-linear-models-and-examine-the-results" class="section level3 unnumbered" number="">
<h3>Fit linear models and examine the results</h3>
<p>Having properly filtered and normalized the data,
the linear models can be fitted to each gene and the results
examined to assess differential expression between the two groups
of interest, in our case HCC vs Control.</p>
<p>Table <a href="preproc.html#tab:lmFit">2.4</a> displays the counts of genes in each DE category:</p>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:lmFit">Table 2.4: </span>DE Results at FDR = 0.05
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Down
</th>
<th style="text-align:right;">
NotSig
</th>
<th style="text-align:right;">
Up
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
HCCvsControl
</td>
<td style="text-align:right;">
5214
</td>
<td style="text-align:right;">
5280
</td>
<td style="text-align:right;">
5258
</td>
</tr>
</tbody>
</table>
</div>
<div id="graphical-representations-of-de-results-md-plots" class="section level3 unnumbered" number="">
<h3>Graphical representations of DE results: MD Plots</h3>
<p>To summarise results for all genes visually, mean-difference plots
(aka MA plot), which display log-FCs from the linear model fit against
the average log-CPM values can be generated using the plotMD function,
with the differentially expressed genes highlighted.</p>
<p>We may also be interested in whether certain gene features are
related to gene identification. Gene GC content, for example, might be
of interest.</p>
<div class="figure"><span id="fig:mdPlotEfit"></span>
<img src="Static/figures/mdPlotEfit-1.png" alt="HCC vs Control - Genes Identified at FDR = 0,05" width="1056" />
<p class="caption">
Figure 2.7: HCC vs Control - Genes Identified at FDR = 0,05
</p>
</div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:quantlogFC">Table 2.5: </span>log FC quartiles by gene identification
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
down
</th>
<th style="text-align:right;">
notDE
</th>
<th style="text-align:right;">
up
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
25%
</td>
<td style="text-align:right;">
-0.07
</td>
<td style="text-align:right;">
-0.01
</td>
<td style="text-align:right;">
0.04
</td>
</tr>
<tr>
<td style="text-align:left;">
50%
</td>
<td style="text-align:right;">
-0.05
</td>
<td style="text-align:right;">
0.00
</td>
<td style="text-align:right;">
0.06
</td>
</tr>
<tr>
<td style="text-align:left;">
75%
</td>
<td style="text-align:right;">
-0.03
</td>
<td style="text-align:right;">
0.01
</td>
<td style="text-align:right;">
0.09
</td>
</tr>
</tbody>
</table>
<p>While many genes are identified, the effect sizes are quite small,
which results in a low signal-to-noise ratio context. See
Section <a href="preproc.html#snr-regime">2.3</a> below.</p>
<p>The log-fold-change distribution for up-represented genes is long-tailed,
with many high log fold-change values.
By contrast, log-fold-change distribution for down-represented genes
closer to symmetric and has few genes with low log fold-change values.
We will see how this affects the results of identifying genes with
an effect size requirement.</p>
<p>The GC content of down regulated genes tends to be slightly lower than the
rest of the genes. A statistical test would find that the difference
between the mean of the down regulated gene population is singificantly different
than the mean of the other gene population even though the difference is
quite small
(-0.028).</p>
<p>These asymmetries are minor, but it would still be good to establish that
they relfect biology rather than processing artifacts.</p>
</div>
<div id="de-genes-at-10-fold-change" class="section level3 unnumbered" number="">
<h3>DE genes at 10% fold change</h3>
<p>For a stricter definition on significance, one may require log-fold-changes
(log-FCs) to be above a minimum value. The treat method
(McCarthy and Smyth 2009 <span class="citation">[<a href="references.html#ref-McCarthy:2009aa" role="doc-biblioref">13</a>]</span>) can be used to calculate p-values
from empirical Bayes moderated t-statistics with a minimum log-FC requirement.
The number of differentially expressed genes are greatly reduced if we
impose a minimal fold-change requirement of 10%.</p>
<pre><code>## 10% FC Gene Identification Summary - voom, adjust.method = BH, p.value = 0.05:</code></pre>
<pre><code>##        HCCvsControl
## Down              3
## NotSig        15550
## Up              199</code></pre>
<div class="figure"><span id="fig:mdPlotTfit"></span>
<img src="Static/figures/mdPlotTfit-1.png" alt="HCC vs Control - Identified Genes at FDR = 0,05 and logFC &gt; 10%" width="1056" />
<p class="caption">
Figure 2.8: HCC vs Control - Identified Genes at FDR = 0,05 and logFC &gt; 10%
</p>
</div>
<p>As noted above, the log-fold-change distribution for the up-represented genes
is long-tailes in comparison to log-fold-change distribution for the down-represented genes.
As a result fewer down-represented than up-regulated genes are identified when a
minimum log-FC requirement is imposed.</p>
</div>
</div>
<div id="snr-regime" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> Signal-to-noise ratio regime</h2>
<p>In Hastie et al. (2017) <span class="citation">[<a href="references.html#ref-Hastie:2017aa" role="doc-biblioref">14</a>]</span>) results from <code>lasso</code> fits are
compared with <code>best subset</code> and <code>forward selection</code> fits and it is argued
that while <code>best subset</code> is optimal for high signal-to-noise regimes,
the lasso gains some competitive advantage when the prevailing signal-to-noise
ratio of the dataset is lowered.</p>
<p>We can extract sigma and signal from the fit objects to get SNR values for each gene
to see in what SNR regime the 5hmC gene body data are.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="preproc.html#cb15-1"></a>lib.size &lt;-<span class="st"> </span><span class="kw">colSums</span>(AF_dgel<span class="op">$</span>counts)</span>
<span id="cb15-2"><a href="preproc.html#cb15-2"></a></span>
<span id="cb15-3"><a href="preproc.html#cb15-3"></a>fit &lt;-<span class="st"> </span>filteredCountsAF_voom_efit</span>
<span id="cb15-4"><a href="preproc.html#cb15-4"></a>sx &lt;-<span class="st"> </span>fit<span class="op">$</span>Amean <span class="op">+</span><span class="st"> </span><span class="kw">mean</span>(<span class="kw">log2</span>(lib.size <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)) <span class="op">-</span><span class="st"> </span><span class="kw">log2</span>(<span class="fl">1e+06</span>)</span>
<span id="cb15-5"><a href="preproc.html#cb15-5"></a>sy &lt;-<span class="st"> </span><span class="kw">sqrt</span>(fit<span class="op">$</span>sigma)</span>
<span id="cb15-6"><a href="preproc.html#cb15-6"></a></span>
<span id="cb15-7"><a href="preproc.html#cb15-7"></a>CV &lt;-<span class="st"> </span>sy<span class="op">/</span>sx    </span></code></pre></div>
<!-- DEBUG BCV from Section \@ref(analysis-of-coverage-variability) vs CV
pairs(cbind(BCV_mtx, CV)) 
boxplot(cbind(BCV_mtx, CV), outline=F) 
-->
<div class="figure"><span id="fig:plotSNR"></span>
<img src="Static/figures/plotSNR-1.png" alt="Cumulative Distribution of SNR - rug = 25, 50, 75 and 90th percentile" width="960" />
<p class="caption">
Figure 2.9: Cumulative Distribution of SNR - rug = 25, 50, 75 and 90th percentile
</p>
</div>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:plotSNR">Table 2.6: </span>SNR Quantiles
</caption>
<thead>
<tr>
<th style="text-align:right;">
25%
</th>
<th style="text-align:right;">
50%
</th>
<th style="text-align:right;">
75%
</th>
<th style="text-align:right;">
90%
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
0.018
</td>
<td style="text-align:right;">
0.036
</td>
<td style="text-align:right;">
0.06
</td>
<td style="text-align:right;">
0.082
</td>
</tr>
</tbody>
</table>
<p>These SNR values are in the range where the lasso and relaxed lasso gain some advantage over
best subset and forward selection fits (see Hastie et al. (2017) <span class="citation">[<a href="references.html#ref-Hastie:2017aa" role="doc-biblioref">14</a>]</span>).</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="intro.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="modeling-background.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["HCC5hmCExplore.pdf", "HCC5hmCExplore.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
