<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>References | DNA Hydroxymethylation in Hepatocellular Carcinoma</title>
  <meta name="description" content="Data from Cai et al. (2019) paper are explored" />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="References | DNA Hydroxymethylation in Hepatocellular Carcinoma" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Data from Cai et al. (2019) paper are explored" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="References | DNA Hydroxymethylation in Hepatocellular Carcinoma" />
  
  <meta name="twitter:description" content="Data from Cai et al. (2019) paper are explored" />
  

<meta name="author" content="Francois Collin" />


<meta name="date" content="2020-09-03" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="conclusions.html"/>

<script src="libs/header-attrs-2.2/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">DNA Hydroxymethylation in Hepatocellular Carcinoma</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preamble</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i>License</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="preproc.html"><a href="preproc.html"><i class="fa fa-check"></i><b>2</b> Preprocessing</a>
<ul>
<li class="chapter" data-level="2.1" data-path="preproc.html"><a href="preproc.html#load-the-data"><i class="fa fa-check"></i><b>2.1</b> Load the data</a></li>
<li class="chapter" data-level="2.2" data-path="preproc.html"><a href="preproc.html#dra"><i class="fa fa-check"></i><b>2.2</b> Differential representation analysis</a>
<ul>
<li class="chapter" data-level="" data-path="preproc.html"><a href="preproc.html#remove-lowly-expressed-genes"><i class="fa fa-check"></i>Remove lowly expressed genes</a></li>
<li class="chapter" data-level="" data-path="preproc.html"><a href="preproc.html#creating-a-design-matrix-and-contrasts"><i class="fa fa-check"></i>Creating a design matrix and contrasts</a></li>
<li class="chapter" data-level="" data-path="preproc.html"><a href="preproc.html#removing-heteroscedasticity-from-the-count-data"><i class="fa fa-check"></i>Removing heteroscedasticity from the count data</a></li>
<li class="chapter" data-level="" data-path="preproc.html"><a href="preproc.html#fit-linear-models-and-examine-the-results"><i class="fa fa-check"></i>Fit linear models and examine the results</a></li>
<li class="chapter" data-level="" data-path="preproc.html"><a href="preproc.html#graphical-representations-of-de-results-md-plots"><i class="fa fa-check"></i>Graphical representations of DE results: MD Plots</a></li>
<li class="chapter" data-level="" data-path="preproc.html"><a href="preproc.html#de-genes-at-10-fold-change"><i class="fa fa-check"></i>DE genes at 10% fold change</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="preproc.html"><a href="preproc.html#look-at-signal-to-noise-ratio"><i class="fa fa-check"></i><b>2.3</b> Look at signal-to-noise ratio</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="modeling-background.html"><a href="modeling-background.html"><i class="fa fa-check"></i><b>3</b> Modeling - Background</a>
<ul>
<li class="chapter" data-level="3.1" data-path="modeling-background.html"><a href="modeling-background.html#predictive-modeling-for-genomic-data"><i class="fa fa-check"></i><b>3.1</b> Predictive modeling for genomic data</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="modeling-background.html"><a href="modeling-background.html#caret-for-model-evaluation"><i class="fa fa-check"></i><b>3.1.1</b> caret for model evaluation</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="modeling-background.html"><a href="modeling-background.html#glmnet"><i class="fa fa-check"></i><b>3.2</b> glmnet</a>
<ul>
<li><a href="modeling-background.html#alpha-hyper-parameter"><strong>alpha</strong> hyper-parameter</a></li>
<li class="chapter" data-level="" data-path="modeling-background.html"><a href="modeling-background.html#lasso-vs-best-subset"><i class="fa fa-check"></i>Lasso vs Best Subset</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="explore-sparsity.html"><a href="explore-sparsity.html"><i class="fa fa-check"></i><b>4</b> The bet on sparsity</a>
<ul>
<li class="chapter" data-level="4.1" data-path="explore-sparsity.html"><a href="explore-sparsity.html#baseline-lasso-fit"><i class="fa fa-check"></i><b>4.1</b> Baseline lasso fit</a></li>
<li class="chapter" data-level="4.2" data-path="explore-sparsity.html"><a href="explore-sparsity.html#relaxed-lasso"><i class="fa fa-check"></i><b>4.2</b> Relaxed lasso</a></li>
<li class="chapter" data-level="4.3" data-path="explore-sparsity.html"><a href="explore-sparsity.html#shrunken-relaxed-lasso"><i class="fa fa-check"></i><b>4.3</b> Shrunken relaxed lasso</a></li>
<li class="chapter" data-level="4.4" data-path="explore-sparsity.html"><a href="explore-sparsity.html#ridge"><i class="fa fa-check"></i><b>4.4</b> ridge</a></li>
<li class="chapter" data-level="4.5" data-path="explore-sparsity.html"><a href="explore-sparsity.html#ridge-lasso"><i class="fa fa-check"></i><b>4.5</b> ridge-lasso</a></li>
<li class="chapter" data-level="4.6" data-path="explore-sparsity.html"><a href="explore-sparsity.html#sparsity-stability"><i class="fa fa-check"></i><b>4.6</b> Sparsity stability</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="model-suite.html"><a href="model-suite.html"><i class="fa fa-check"></i><b>5</b> Fitted Model Suite</a></li>
<li class="chapter" data-level="6" data-path="conclusions.html"><a href="conclusions.html"><i class="fa fa-check"></i><b>6</b> Conclusions</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a>
<ul>
<li class="chapter" data-level="6.1" data-path="references.html"><a href="references.html#analysis-of-coverage-variability"><i class="fa fa-check"></i><b>6.1</b> Analysis of coverage variability</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="LICENSE.md" style="font:em;" target="blank">Copyright (c) 2020 Francois Collin</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">DNA Hydroxymethylation in Hepatocellular Carcinoma</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="references" class="section level1 unnumbered" number="">
<h1>References</h1>

<div id="analysis-of-coverage-variability" class="section level2" number="6.1">
<h2><span class="header-section-number">6.1</span> Analysis of coverage variability</h2>
<p>We will use the methods described in Hart et al. (2013) <span class="citation">[<a href="references.html#ref-Hart:2013aa" role="doc-biblioref">24</a>]</span>
to characterize coverage variability in these data.
These methods do not take multiple comparisons into account.
Other tools for sample size calculation in RNA-Seq studies include
Bi and Liu (2016) <span class="citation">[<a href="references.html#ref-Bi:2016aa" role="doc-biblioref">25</a>]</span>, Baccarella (2018) <span class="citation">[<a href="references.html#ref-Baccarella:2018aa" role="doc-biblioref">26</a>]</span>,
Guo (2014) <span class="citation">[<a href="references.html#ref-Guo:2014aa" role="doc-biblioref">27</a>]</span>, Yu (2017) <span class="citation">[<a href="references.html#ref-Yu:2017aa" role="doc-biblioref">28</a>]</span>, and
Zhao (2018) <span class="citation">[<a href="references.html#ref-Zhao:2018ab" role="doc-biblioref">29</a>]</span>.
Poplawski (2018) <span class="citation">[<a href="references.html#ref-Poplawski:2017aa" role="doc-biblioref">30</a>]</span>
evaluated RNA-seq sample size tools identified from a systematic search. They found the six evaluated tools provided widely different answers, which were strongly affected by fold change.</p>
<p>The references listed above aim at providing guidance for RNA-Seq experimental design.
There is much discussion and a wide range of opinion on sample size requirements
to ensure reproducibility in RNA-Seq results. At one end of the spectrum,
Ein-Dor et al. (2006) <span class="citation">[<a href="references.html#ref-Ein-Dor:2006aa" role="doc-biblioref">31</a>]</span> argue that thousands of samples are needed to
generate a robust gene list for predicting outcome in cancer. At the other end,
Dobbin et al. (2007, 2008) <span class="citation">[<a href="references.html#ref-Dobbin:2007aa" role="doc-biblioref">32</a>,<a href="references.html#ref-Dobbin:2008aa" role="doc-biblioref">33</a>]</span> claim that sample sizes in
the range of 20–30 per class may be adequate for building a good predictor in many cases.
Part of the disparity in sample size requirement recomendation comes from
differences of opinion in terms of what constitutes <code>reproducible results</code>.
In the context of sample classification, if we focus on the predicted probabilities
for individual samples, we may find good reproducibility across studies with
moderate samples sizes. If, on the other hand, we closely inspect the gene
signatures reported across studies, much greater sample sizes may be required to
achieve concordance. Kim (2009) <span class="citation">[<a href="references.html#ref-Kim:2009aa" role="doc-biblioref">34</a>]</span>, like Ein-Dor et al., also find
issues in RNA-Seq research in terms of the instability of identified prognostic
gene signatures, few overlap between independently developed prognostic gene signatures, and
poor inter-study applicability of gene signatures. Fan et al. (2006) <span class="citation">[<a href="references.html#ref-Fan:2006aa" role="doc-biblioref">35</a>]</span>,
on the other hand, found good concordance among gene-expression–based predictors
for breast cancer. We will return to this question when we examine
the relationship between classification model results and sample size in this
dataset later on this paper.</p>
<p>For two groups comparisons, the basic formula for the required number of samples per group is:</p>
<p><span class="math display">\[ n = 2(z_{1-\alpha/2} + z_{\beta})^2 \frac{(1/\mu + \sigma^2)}{ln(\Delta^2)} \]</span></p>
<ul>
<li>The parameters <span class="math inline">\(\alpha\)</span> and <span class="math inline">\(\beta\)</span> are <strong>size</strong> and <strong>power</strong> of the test.<br />
</li>
<li><span class="math inline">\(\Delta\)</span> is the targeted <strong>effect size</strong>.<br />
</li>
<li><span class="math inline">\(\mu\)</span> and <span class="math inline">\(\sigma\)</span> are the <strong>mean</strong> and <strong>coefficient of variation</strong>
of the distribution of measurement, gene representation indices in this case.</li>
</ul>
<p>These three parameters will be fixed across genes or a given study, and are often dictated by
external requirements. Typical values might be an effect size of <span class="math inline">\(\Delta = 1.5\)</span>
(a.k.a fold change), corresponding to detection of a 50% change in gene expression
between the two groups.
<span class="math inline">\(z_{1 - .05/2} = 1.96\)</span>, corresponding to a two sided test at size <span class="math inline">\(\alpha = 0.05\)</span>;
and <span class="math inline">\(z_{.90}= 1.28\)</span> corresponding to 90% power.
The other two variables will be gene and experiment dependent: the normalized depth
of coverage <span class="math inline">\(\mu\)</span> of the gene, and the coefficient of variation <span class="math inline">\(\sigma\)</span> in this
gene between biological replicates. The technical variation of the
comparison is inversely proportional to the number of sequenced reads
for the gene and therefore decreases with sequencing depth.
The biological variation is a property of the particular gene/model system/condition under study.
One would expect it to be smaller for uniform systems such as cell culture and/or products that
are under tight regulatory control, and larger for less uniform replicates such
as human subject samples. The dataset under study in this report is the first of its
kind to give us an idea of variability levels in 5hmC representation.</p>
<!-- SKIP THIS
Before examining biological variability lets take a look at
the sequencing rates.

-->
<p>As in Hart et al. (2013) <span class="citation">[<a href="references.html#ref-Hart:2013aa" role="doc-biblioref">24</a>]</span> we estimate the
biological coefficient of variation (CV) in expression across samples in the data set
using a negative binomial model.</p>
<!--
From [edgeR User's Guide](https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf)
, the biolgical coefficient of variation (BCV), is the square root of the dispersion
parameter under the negative binomial model. Hence, it is equivalent to estimating the
dispersion(s) of the negative binomial model.
-->
<div class="figure"><span id="fig:plotBCV"></span>
<img src="Static/figures/plotBCV-1.png" alt="Cumulative Distribution of CV - rug = 90th percentile" width="960" />
<p class="caption">
Figure 6.1: Cumulative Distribution of CV - rug = 90th percentile
</p>
</div>
<p><br/></p>
<p>BCV values are quite low:</p>
<div class="figure"><span id="fig:bspBCV"></span>
<img src="Static/figures/bspBCV-1.png" alt="BCV distributions" width="960" />
<p class="caption">
Figure 6.2: BCV distributions
</p>
</div>
<p><br/></p>
<!-- 
 SampleSize_f <- function(Alpha=.01, Beta=.80, Delta=2, Mu, Disp)
 {temp <- qnorm(1-Alpha/2) + qnorm(Beta)
  vtemp <- 1/Mu + Disp
  2*vtemp*(temp/log(Delta))^2}

 cat("SampleSize_f(Mu=10, Disp=.4^2):\n")
 SampleSize_f(Mu=10, Disp=.4^2)
-->
<p>We can now look at sample size estimates to required to detect
various effect sizes. The effect sizes examined here are selected
based on the differential representation analysis in Section <a href="preproc.html#dra">2.2</a> below.</p>
<div class="figure"><span id="fig:sampleSizeCurve"></span>
<img src="Static/figures/sampleSizeCurve-1.png" alt="Sample Size Estimates" width="1056" />
<p class="caption">
Figure 6.3: Sample Size Estimates
</p>
</div>
<p>For the filtered reads, coverage looks like this:</p>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:quantCountsAF">Table 6.1: </span>Coverage Summary - Columns are sample coverage quantiles and total coverage
Rows are quartiles across samples
</caption>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
15%
</th>
<th style="text-align:right;">
25%
</th>
<th style="text-align:right;">
50%
</th>
<th style="text-align:right;">
75%
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
25%
</td>
<td style="text-align:right;">
37
</td>
<td style="text-align:right;">
63
</td>
<td style="text-align:right;">
163
</td>
<td style="text-align:right;">
396
</td>
</tr>
<tr>
<td style="text-align:left;">
50%
</td>
<td style="text-align:right;">
45
</td>
<td style="text-align:right;">
76
</td>
<td style="text-align:right;">
198
</td>
<td style="text-align:right;">
483
</td>
</tr>
<tr>
<td style="text-align:left;">
75%
</td>
<td style="text-align:right;">
54
</td>
<td style="text-align:right;">
91
</td>
<td style="text-align:right;">
237
</td>
<td style="text-align:right;">
577
</td>
</tr>
</tbody>
</table>
<p>From this table, we see that 25% of the samples have an upper quartile of gene coverage
exceeding 577 reads, 25% of samples
have a 15 percentile of coverage lower than
37, etc.</p>
<p>Note that with these data, moderate sample sizes are adequate
to detect genes with effect sizes as small as 1.05 (5% fold change). This is due to the fact
that the biological variability in gene body 5hmC density is quite low.
For human samples, RNA-Seq within group biological variability is
typically in the 0.4-1.0 range <span class="citation">[<a href="references.html#ref-Hart:2013aa" role="doc-biblioref">24</a>]</span>.</p>

<div id="refs" class="references">
<div id="ref-Gai:2019aa">
<p>1. Gai, W., and Sun, K. Epigenetic biomarkers in cell-free dna and applications in liquid biopsy. Genes <em>10</em>, 32. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/30634483">https://pubmed.ncbi.nlm.nih.gov/30634483</a>.</p>
</div>
<div id="ref-Cai:2019aa">
<p>2. Cai, J., Chen, L., Zhang, Z., Zhang, X., Lu, X., Liu, W., Shi, G., Ge, Y., Gao, P., and Yang, Y. <em>et al.</em> Genome-wide mapping of 5-hydroxymethylcytosines in circulating cell-free dna as a non-invasive approach for early detection of hepatocellular carcinoma. Gut, gutjnl–2019–318882. Available at: <a href="http://gut.bmj.com/content/early/2019/07/28/gutjnl-2019-318882.abstract">http://gut.bmj.com/content/early/2019/07/28/gutjnl-2019-318882.abstract</a>.</p>
</div>
<div id="ref-Li:2017aa">
<p>3. Li, W., Zhang, X., Lu, X., You, L., Song, Y., Luo, Z., Zhang, J., Nie, J., Zheng, W., and Xu, D. <em>et al.</em> DNA 5-hydroxymethylcytosines from cell-free circulating dna as diagnostic biomarkers for human cancers. bioRxiv, 163204. Available at: <a href="http://biorxiv.org/content/early/2017/07/13/163204.abstract">http://biorxiv.org/content/early/2017/07/13/163204.abstract</a>.</p>
</div>
<div id="ref-Song:2017aa">
<p>4. Song, C.-X., Yin, S., Ma, L., Wheeler, A., Chen, Y., Zhang, Y., Liu, B., Xiong, J., Zhang, W., and Hu, J. <em>et al.</em> (2017). 5-hydroxymethylcytosine signatures in cell-free dna provide information about tumor types and stages. Cell Research <em>27</em>, 1231–1242. Available at: <a href="https://doi.org/10.1038/cr.2017.106">https://doi.org/10.1038/cr.2017.106</a>.</p>
</div>
<div id="ref-Collin:2018aa">
<p>5. Collin, F., Ning, Y., Phillips, T., McCarthy, E., Scott, A., Ellison, C., Ku, C.-J., Guler, G.D., Chau, K., and Ashworth, A. <em>et al.</em> Detection of early stage pancreatic cancer using 5-hydroxymethylcytosine signatures in circulating cell free dna. bioRxiv, 422675. Available at: <a href="http://biorxiv.org/content/early/2018/09/26/422675.abstract">http://biorxiv.org/content/early/2018/09/26/422675.abstract</a>.</p>
</div>
<div id="ref-Huang:2017aa">
<p>6. Huang, L.-H., Lin, P.-H., Tsai, K.-W., Wang, L.-J., Huang, Y.-H., Kuo, H.-C., and Li, S.-C. The effects of storage temperature and duration of blood samples on dna and rna qualities. PloS one <em>12</em>, e0184692–e0184692. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/28926588">https://pubmed.ncbi.nlm.nih.gov/28926588</a>.</p>
</div>
<div id="ref-Permenter:2015aa">
<p>7. Permenter, J., Ishwar, A., Rounsavall, A., Smith, M., Faske, J., Sailey, C.J., and Alfaro, M.P. (2015). Quantitative analysis of genomic dna degradation in whole blood under various storage conditions for molecular diagnostic testing. Molecular and Cellular Probes <em>29</em>, 449–453. Available at: <a href="http://www.sciencedirect.com/science/article/pii/S0890850815300207">http://www.sciencedirect.com/science/article/pii/S0890850815300207</a>.</p>
</div>
<div id="ref-Law:2018aa">
<p>8. Law, C., Alhamdoosh, M., Su, S., Dong, X., Tian, L., Smyth, G., and Ritchie, M. (2018). RNA-seq analysis is easy as 1-2-3 with limma, glimma and edgeR [version 3; peer review: 3 approved]. F1000Research <em>5</em>. Available at: <a href="https://dx.doi.org/10.12688%2Ff1000research.9005.3">https://dx.doi.org/10.12688%2Ff1000research.9005.3</a>.</p>
</div>
<div id="ref-Ritchie:2015aa">
<p>9. Ritchie, M.E., Phipson, B., Wu, D., Hu, Y., Law, C.W., Shi, W., and Smyth, G.K. Limma powers differential expression analyses for rna-sequencing and microarray studies. Nucleic acids research <em>43</em>, e47–e47. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/25605792">https://pubmed.ncbi.nlm.nih.gov/25605792</a>.</p>
</div>
<div id="ref-Peixoto:2015aa">
<p>10. Peixoto, L., Risso, D., Poplawski, S.G., Wimmer, M.E., Speed, T.P., Wood, M.A., and Abel, T. How data analysis affects power, reproducibility and biological insight of rna-seq studies in complex datasets. Nucleic acids research <em>43</em>, 7664–7674. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/26202970">https://pubmed.ncbi.nlm.nih.gov/26202970</a>.</p>
</div>
<div id="ref-Gandolfo:2018aa">
<p>11. Gandolfo, L.C., and Speed, T.P. RLE plots: Visualizing unwanted variation in high dimensional data. PloS one <em>13</em>, e0191629–e0191629. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/29401521">https://pubmed.ncbi.nlm.nih.gov/29401521</a>.</p>
</div>
<div id="ref-Risso:2014aa">
<p>12. Risso, D., Ngai, J., Speed, T.P., and Dudoit, S. (2014). Normalization of rna-seq data using factor analysis of control genes or samples. Nature Biotechnology <em>32</em>, 896–902. Available at: <a href="https://doi.org/10.1038/nbt.2931">https://doi.org/10.1038/nbt.2931</a>.</p>
</div>
<div id="ref-McCarthy:2009aa">
<p>13. McCarthy, D.J., and Smyth, G.K. Testing significance relative to a fold-change threshold is a treat. Bioinformatics <em>25</em>, 765–771. Available at: <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2654802/">http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2654802/</a>.</p>
</div>
<div id="ref-Hastie:2017aa">
<p>14. Hastie, T., and Tibshirani, R. (2017). Extended comparisons of best subset selection, forward stepwise selection, and the lasso. arXiv: Methodology.</p>
</div>
<div id="ref-Friedman:2010aa">
<p>15. Friedman, J., Hastie, T., and Tibshirani, R. (2010). Regularization paths for generalized linear models via coordinate descent. J Stat Softw <em>33</em>, 1–22.</p>
</div>
<div id="ref-Tibshirani:2012aa">
<p>16. Tibshirani, R., Bien, J., Friedman, J., Hastie, T., Simon, N., Taylor, J., and Tibshirani, R.J. Strong rules for discarding predictors in lasso-type problems. Journal of the Royal Statistical Society. Series B, Statistical methodology <em>74</em>, 245–266. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/25506256">https://pubmed.ncbi.nlm.nih.gov/25506256</a>.</p>
</div>
<div id="ref-Simon:2011aa">
<p>17. Simon, N., Friedman, J., Hastie, T., and Tibshirani, R. Regularization paths for cox’s proportional hazards model via coordinate descent. J Stat Softw <em>39</em>, 1–13.</p>
</div>
<div id="ref-Simon:2013aa">
<p>18. Simon, N., Friedman, J., and Hastie, T. (2013). A blockwise descent algorithm for group-penalized multiresponse and multinomial regression. arXiv preprint arXiv:1311.6529.</p>
</div>
<div id="ref-Bertsimas:2016aa">
<p>19. Bertsimas, D., King, A., and Mazumder, R. (2016). Best subset selection via a modern optimization lens. Ann. Statist. <em>44</em>, 813–852. Available at: <a href="https://projecteuclid.org:443/euclid.aos/1458245736">https://projecteuclid.org:443/euclid.aos/1458245736</a>.</p>
</div>
<div id="ref-Xiang:2020aa">
<p>20. Xiang, G., Keller, C.A., Giardine, B., An, L., Li, Q., Zhang, Y., and Hardison, R.C. (2020). S3norm: Simultaneous normalization of sequencing depth and signal-to-noise ratio in epigenomic data. Nucleic Acids Research <em>48</em>, e43–e43. Available at: <a href="https://doi.org/10.1093/nar/gkaa105">https://doi.org/10.1093/nar/gkaa105</a>.</p>
</div>
<div id="ref-Lozoya:2018aa">
<p>21. Lozoya, O.A., Santos, J.H., and Woychik, R.P. A leveraged signal-to-noise ratio (lstnr) method to extract differentially expressed genes and multivariate patterns of expression from noisy and low-replication rnaseq data. Frontiers in genetics <em>9</em>, 176–176. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/29868123">https://pubmed.ncbi.nlm.nih.gov/29868123</a>.</p>
</div>
<div id="ref-Simonsen:2018aa">
<p>22. Simonsen, A.T., Hansen, M.C., Kjeldsen, E., Møller, P.L., Hindkjær, J.J., Hokland, P., and Aggerholm, A. (2018). Systematic evaluation of signal-to-noise ratio in variant detection from single cell genome multiple displacement amplification and exome sequencing. BMC Genomics <em>19</em>, 681. Available at: <a href="https://doi.org/10.1186/s12864-018-5063-5">https://doi.org/10.1186/s12864-018-5063-5</a>.</p>
</div>
<div id="ref-Rapaport:2013aa">
<p>23. Rapaport, F., Khanin, R., Liang, Y., Pirun, M., Krek, A., Zumbo, P., Mason, C.E., Socci, N.D., and Betel, D. (2013). Comprehensive evaluation of differential gene expression analysis methods for rna-seq data. Genome biology <em>14</em>, R95–R95. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/24020486">https://pubmed.ncbi.nlm.nih.gov/24020486</a>.</p>
</div>
<div id="ref-Hart:2013aa">
<p>24. Hart, S.N., Therneau, T.M., Zhang, Y., Poland, G.A., and Kocher, J.-P. Calculating sample size estimates for rna sequencing data. Journal of Computational Biology <em>20</em>, 970–978. Available at: <a href="https://doi.org/10.1089/cmb.2012.0283">https://doi.org/10.1089/cmb.2012.0283</a>.</p>
</div>
<div id="ref-Bi:2016aa">
<p>25. Bi, R., and Liu, P. Sample size calculation while controlling false discovery rate for differential expression analysis with rna-sequencing experiments. BMC bioinformatics <em>17</em>, 146–146. Available at: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27029470">https://www.ncbi.nlm.nih.gov/pubmed/27029470</a>.</p>
</div>
<div id="ref-Baccarella:2018aa">
<p>26. Baccarella, A., Williams, C.R., Parrish, J.Z., and Kim, C.C. Empirical assessment of the impact of sample number and read depth on rna-seq analysis workflow performance. BMC bioinformatics <em>19</em>, 423–423. Available at: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30428853">https://www.ncbi.nlm.nih.gov/pubmed/30428853</a>.</p>
</div>
<div id="ref-Guo:2014aa">
<p>27. Guo, Y., Zhao, S., Li, C.-I., Sheng, Q., and Shyr, Y. RNAseqPS: A web tool for estimating sample size and power for rnaseq experiment. Cancer informatics <em>13</em>, 1–5. Available at: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25374457">https://www.ncbi.nlm.nih.gov/pubmed/25374457</a>.</p>
</div>
<div id="ref-Yu:2017aa">
<p>28. Yu, L., Fernandez, S., and Brock, G. (2017). Power analysis for rna-seq differential expression studies. BMC Bioinformatics <em>18</em>, 234. Available at: <a href="https://doi.org/10.1186/s12859-017-1648-2">https://doi.org/10.1186/s12859-017-1648-2</a>.</p>
</div>
<div id="ref-Zhao:2018ab">
<p>29. Zhao, S., Li, C.-I., Guo, Y., Sheng, Q., and Shyr, Y. RnaSeqSampleSize: Real data based sample size estimation for rna sequencing. BMC bioinformatics <em>19</em>, 191–191. Available at: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29843589">https://www.ncbi.nlm.nih.gov/pubmed/29843589</a>.</p>
</div>
<div id="ref-Poplawski:2017aa">
<p>30. Poplawski, A., and Binder, H. (2017). Feasibility of sample size calculation for rna-seq studies. Briefings in Bioinformatics <em>19</em>, 713–720. Available at: <a href="https://doi.org/10.1093/bib/bbw144">https://doi.org/10.1093/bib/bbw144</a>.</p>
</div>
<div id="ref-Ein-Dor:2006aa">
<p>31. Ein-Dor, L., Zuk, O., and Domany, E. Thousands of samples are needed to generate a robust gene list for predicting outcome in cancer. Proceedings of the National Academy of Sciences of the United States of America <em>103</em>, 5923–5928. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/16585533">https://pubmed.ncbi.nlm.nih.gov/16585533</a>.</p>
</div>
<div id="ref-Dobbin:2007aa">
<p>32. Dobbin, K.K., and Simon, R.M. (2007). Sample size planning for developing classifiers using high-dimensional dna microarray data. Biostatistics <em>8</em>, 101–117. Available at: <a href="https://doi.org/10.1093/biostatistics/kxj036">https://doi.org/10.1093/biostatistics/kxj036</a>.</p>
</div>
<div id="ref-Dobbin:2008aa">
<p>33. Dobbin, K.K., Zhao, Y., and Simon, R.M. How large a training set is needed to develop a classifier for microarray data? Clin Cancer Res <em>14</em>, 108–114.</p>
</div>
<div id="ref-Kim:2009aa">
<p>34. Kim, S.-Y. Effects of sample size on robustness and prediction accuracy of a prognostic gene signature. BMC bioinformatics <em>10</em>, 147–147. Available at: <a href="https://pubmed.ncbi.nlm.nih.gov/19445687">https://pubmed.ncbi.nlm.nih.gov/19445687</a>.</p>
</div>
<div id="ref-Fan:2006aa">
<p>35. Fan, C., Oh, D.S., Wessels, L., Weigelt, B., Nuyten, D.S.A., Nobel, A.B., Veer, L.J. van’t, and Perou, C.M. Concordance among gene-expression–based predictors for breast cancer. New England Journal of Medicine <em>355</em>, 560–569. Available at: <a href="https://doi.org/10.1056/NEJMoa052933">https://doi.org/10.1056/NEJMoa052933</a>.</p>
</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="conclusions.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["HCC5hmCExplore.pdf", "HCC5hmCExplore.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
